
#load("pbmc.RData")

#input <- list(data_source_genetic=c("Melanoma-Van Allen","Melanoma-Hugo"),treatment_genetic=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_genetic=c("Pre","On","Unknown"))
#input <- list(data_source_single_cell=c("Melanoma-Sade_Feldman","Melanoma-Krieg_panel_3"))

#input <- list(data_source_pan_cancer=c("Melanoma-Van Allen","Melanoma-Hugo","Melanoma-Snyder-Nathanson"),treatment_pan_cancer=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_pan_cancer=c("Pre","On"))

#input <- list(data_source_cibersort=c("Van Allen","Hugo"),treatment_cibersort=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_cibersort=c("Pre","on"),immucell_cibersort="T.cells.CD8",slider_cibersort_OS=0.5)
#input <- list(data_source_expression=c("Melanoma-Auslander","Melanoma-Chen","Melanoma-Hugo","Melanoma-Prat","Melanoma-Riaz","Melanoma-Van Allen"),treatment_expression=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression=c("Pre","on"),gene_expression="CD19",slider_expression_OS=0.5,slider_expression_PFS=0.5,clinical_expression="treatment type")
#input <- list(data_source_expression=c("Melanoma-Gide"),treatment_expression=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression=c("Pre","On"),gene_expression="ALL",slider_expression_OS=0.5,slider_expression_PFS=0.5,clinical_expression="treatment type")
#input <- list(data_source_expression_sum=c("Van Allen","Prat"),treatment_expression_sum=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression_sum=c("Pre","on"),gene_expression_sum=c("CTLA4","PDCD1"),slider_expression_sum_OS=0.5)
#input <- list(gene_expression_pairs_1=c("CD28","CD276"),gene_expression_pairs_2=c("PDCD1","TNFRSF4"),gene_expression_pairs_3=c("CTLA4","TNFRSF4"),data_source_expression_pairs=c("Melanoma-Auslander","Melanoma-Chen","Melanoma-Hugo","Melanoma-Prat","Melanoma-Riaz","Melanoma-Van Allen"),treatment_expression_pairs=c("anti-CTLA-4","anti-PD-1/L1"),description_expression_pairs=c("Pre","On"),slider_expression_pairs_OS=1,slider_expression_pairs_PFS=1)

#input <- list(data_source_mutation_signature=c("Van Allen","Hugo"),treatment_mutation_signature=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_mutation_signature=c("Pre","On"),signature_mutation_signature="Signature.12",slider_mutation_signature_PFS=0.5)

# input <- list(data_source_mutation_loads=c("Melanoma-Samstein" ,"Non-Small Cell Lung Cancer-Samstein" ,"Bladder Cancer-Samstein" ,
#                                            "Breast Cancer-Samstein" ,"Cancer of Unknown Primary-Samstein" ,"Colorectal Cancer-Samstein" ,
#                                            "Esophagogastric Cancer-Samstein" ,"Glioma-Samstein" ,"Head and Neck Cancer-Samstein" ,
#                                            "Renal Cell Carcinoma-Samstein" ,"Skin Cancer, Non-Melanoma-Samstein" ),
# treatment_mutation_loads=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
# description_mutation_loads=c("Pre","On"),slider_mutation_loads_OS=0.5)

#input <- list(data_source_mutation=c("Melanoma-Van Allen","Melanoma-Hugo","Melanoma-Riaz","Melanoma-Snyder-Nathanson","Non-Small Cell Lung Cancer-Naiyer Rizvi","Renal Cell Carcinoma-Miao","Hepatocellular Carcinoma-Harding"),treatment_mutation=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_mutation=c("Pre","On"),gene_mutation=c("RYR1"),clinical_mutation="Pre- or On-treatment")
#input <- list(data_source_mutation=c("Melanoma-Samstein"),treatment_mutation=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_mutation=c("Pre","On"),gene_mutation=c("SCYL3"),clinical_mutation="Pre- or On-treatment")
#input <- list(data_source_mutation_loads=c("Melanoma-Hugo"),treatment_mutation_loads=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_mutation_loads=c("Pre","On","Unknown"),slider_mutation_loads_OS=0.5,slider_mutation_loads_PFS=0.5,clinical_mutation_loads="treatment type")
#input <- list(mutation_plots_result="Overall survival status",data_source_mutation=c("Hugo","Van Allen"),treatment_mutation=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_mutation=c("Pre","On"),gene_mutation=c("KRAS","HRAS","NRAS","NF1"))
#input <- list(data_source_multi_dimensions="Samstein",treatment_multi_dimensions=c("anti-CTLA-4","anti-PD-1/L1"),description_multi_dimensions=c("Pre","On"), yaxis_multi_dimensions="mutation load",xaxis_multi_dimensions="mutation load",yaxis_gene_multi_dimensions="Mutation.Load",xaxis_gene_multi_dimensions="Mutation.Load")
#input <- list(data_source_multi_dimensions=c("Melanoma-Auslander","Melanoma-Chen","Melanoma-Hugo","Melanoma-Prat","Melanoma-Riaz","Melanoma-Van Allen"),treatment_multi_dimensions=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_multi_dimensions=c("Pre","On"), xaxis_multi_dimensions="mutational loads",yaxis_multi_dimensions="mutation",xaxis_gene_multi_dimensions="Mutation.Load",yaxis_gene_multi_dimensions=c("TAP1","TAP2","TAPBP","CIITA","RFX5","RFXANK"),slider_multi_dimensions_x=0.5,slider_multi_dimensions_y=0.5,do_multi_dimensions=TRUE,clinical_multi_dimensions="RECIST")

#input <- list(data_source_single_cell_gene=c("Melanoma-Sade_Feldman","Melanoma-Krieg_panel_3"),treatment_single_cell_gene="anti-PD-1/L1",description_single_cell_gene=c("Pre","On"),gene_single_cell_gene="CD19")
#input <- list(data_source_single_cell_cell="Melanoma-Krieg_panel_3",treatment_single_cell_cell=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_single_cell_cell=c("Pre","On"),cluster_single_cell_cell=4,slider_single_cell_cell_OS=0.5,slider_single_cell_cell_PFS=0.5)

#input <- list(file_clinical=list(datapath="J:/YJ/VUMC-Project/ImmuPortal/code/code for Immu-Mela version 2.0/code/www/Example_clinical_data.txt"),file_transcriptome=list(datapath="J:/YJ/VUMC-Project/ImmuPortal/code/code for Immu-Mela version 2.0/code/www/Example_transcriptome_data.txt"),file_mutation=list(datapath="J:/YJ/VUMC-Project/ImmuPortal/code/code for Immu-Mela version 2.0/code/www/Example_mutation_data.txt"),do_user_data=TRUE)
#input <- list(file_clinical=NULL,file_transcriptome=NULL,file_mutation=NULL,sep_clinical="\t",sep_transcriptome=",",sep_mutation="\t")
#sig_matrix<-"H:/YJ/VUMC-Project/ImmuPortal/code/shiny example/www/LM22.txt"
#mixture_file<-"H:/YJ/VUMC-Project/ImmuPortal/code/shiny example/www/Example_transcriptome_data.txt"


#input <- list(data_source_pan_cancer=c("Melanoma-Van Allen","Melanoma-Hugo","Melanoma-Snyder-Nathanson"),treatment_pan_cancer=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_pan_cancer=c("Pre","On"))


shinyServer(function(input, output, session) {

  
  ###############################
  #### Start
  ##############################
  
  
  output$start_plot <- renderPlotly({
    
    bar.data <- data.frame(table(database.test.process()$Disease),stringsAsFactors = FALSE)
    colnames(bar.data) <-c("Cancers","Count")

    pie.1.data <- data.frame(table(database.test.process()$treatment),stringsAsFactors = FALSE)
    colnames(pie.1.data) <- c("treatment","number.1")

    pie.2.data <- data.frame(table(database.test.process()$PFS.status),stringsAsFactors = FALSE)
    colnames(pie.2.data) <- c("response","number.2")
    
    pie.3.data <- data.frame(table(database.test.process()$description),stringsAsFactors = FALSE)
    colnames(pie.3.data) <- c("description","number.3")
    pie.3.data[,"description"] <- as.character(pie.3.data[,"description"])
    pie.3.data[pie.3.data[,"description"]=="Pre","description"]<-"Pre-treatment"
    pie.3.data[pie.3.data[,"description"]=="On","description"]<-"On-treatment"
    
    plot.data <- rbind.fill(bar.data,pie.1.data,pie.2.data,pie.3.data)
    plot.data$Cancers <- factor(plot.data$Cancers, levels = unique(plot.data$Cancers)[order(plot.data$Count, decreasing = TRUE)])
    plot_ly(plot.data, x = ~Cancers, y = ~Count, type = 'bar',
            marker = list(color="#7CDBD5")) %>%
      # marker = list(color=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
      #                       '#CF59AE', '#000000', '#006666', '#006666', '#009933',
      #                       '#ace7b9', '#cdf0d5'))) %>%
      
      add_pie(plot.data, labels= ~treatment,value= ~number.1,type="pie",
              textposition='inside',textinfo='label',
              domain=list(x=c(0.2,0.4),y=c(0.5,1)),hole = 0.2) %>%
      add_pie(plot.data, labels= ~response,value= ~number.2,type="pie",
              textposition='inside',textinfo='label',
              domain=list(x=c(0.5,0.7),y=c(0.5,1)),hole = 0.2) %>%
      add_pie(plot.data, labels= ~description,value= ~number.3,type="pie",
              textposition='inside',textinfo='label',
              domain=list(x=c(0.8,1),y=c(0.5,1)),hole = 0.2) %>%
      layout(showlegend = F,
             xaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = TRUE,
                          tickangle =-45,
                          tickfont=list(size=20,color="black"),
                          title=""),
             yaxis = list(showgrid = FALSE, zeroline = FALSE, showticklabels = FALSE,
                          title="patient number",titlefont=list(size=20,color="black"))) %>%
      config(displayModeBar = F) %>% 
      add_annotations(x=as.character(na.omit(plot.data$Cancers)),
                      y=na.omit(plot.data$Count)+20,
                      text=as.character(na.omit(plot.data$Count)),
                      xref="x",yref="y",font=list(size=14,color="black"),showarrow=F)
    
  })
  
  
  output$start_left_table <- DT::renderDataTable({
    #    bar.data <- data.frame(table(database.test.process()$Disease),stringsAsFactors = FALSE)
    #    colnames(bar.data) <-c("Cancers","Count")
    start.table$datasource <- paste0("<a href='", start.table$reference, "' target='_blank'>", start.table$datasource, "</a>")
    colnames(start.table)[2] <- "data source"
#    start.table <- start.table[order(start.table[,"Disease"]),]

    sketch <- withTags(
      table(
        class = "display",
        thead(
          tr(
            th(colspan = 3, "Cohorts", style = "border-right: solid 2px;"),
            th(colspan = 3, "Treatments", style = "border-right: solid 2px;"),
            th(colspan = 2, "Biopsy time", style = "border-right: solid 2px;"),
            th(colspan = 3, "ICB outcomes", style = "border-right: solid 2px;"),
            th(colspan = 3, "-omics data")
          ),
          # tr(
          #   th(colspan = 3, "", style = "border-right: solid 2px;"),
          #   th(colspan = 4, "AQS AMP450 (4-10-19)", style = "border-right: solid 2px;"),
          #   th("Envista", style = "border-right: solid 2px;"),
          #   th(colspan = 3, "")
          # ),
          tr(
            th("Disease"),
            th("datasource"),
            th("Num.", style = "border-right: solid 2px;"),
            th("anti-CTLA-4"),
            th("anti-PD-1/L1"),
            th("anti-PD-1/L1+anti-CTLA-4", style = "border-right: solid 2px;"),
            th("On-treatment"),
            th("Pre-treatment", style = "border-right: solid 2px;"),
            th("immunotherapy response"),
            th("OS"),
            th("PFS", style = "border-right: solid 2px;"),
            th("genetic data"),
            th("transcriptomic data"),
            th("single-cell data")
          )
        )
      )
    )
    DT::datatable(start.table[,1:14],container=sketch, options = list(columnDefs=list(list(targets="_all",className="dt-center")),searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) %>% 
    formatStyle(c(3,6,8,11), `border-right` = "solid 2px")
    #        formatStyle("Disease",#target="row",
# #                   backgroundColor = styleEqual(unique(start.table$Disease), c( '#d9d9f3','#ceefe4','#fef4a9', '#9dd3a8','#3b9a9c','#78fee0','#a696c8','#fad3cf','#d4edf4','#f9fbba','#e2f2d5','#f6c2c2')))
#                     backgroundColor = styleEqual(unique(start.table$Disease), c( '#9BD7D5')))
    
  })

  Features = c("Signatures prioritization","Mutation","Mutational loads/burden","Mutational signatures","Expression","Expression sum","Expression relation pairs","Immune cell components","Integration","Gene expression","Cell populations","Pan-cancer study")
  Features_feature = c("Signatures prioritization","Genetic: Mutation","Genetic: Mutational loads/burden","Genetic: Mutational signatures","Transcriptomic: Expression","Transcriptomic: Expression sum","Transcriptomic: Expression relation pairs","Transcriptomic: Immune cell components","Integration","Single-cell: Gene expression","Single-cell: Cell populations","Pan-cancer study")
  # names_in_code =c("prio","mutation","mutation_loads","mutation_signature","expression","expression_sum","expression_pairs","cibersort",
  #                  "multi_dimensions","single_cell_gene","single_cell_cell","pan_cancer")
  data_source_single_cell <-  c("Melanoma-Sade_Feldman (048 samples/+++)" ,
                                "Melanoma-Krieg_panel_1 (040 samples/-+-)" ,
                                "Melanoma-Krieg_panel_2 (040 samples/-+-)" ,
                                "Melanoma-Krieg_panel_3 (039 samples/-+-)")

  data_source_genetic <- c("Melanoma-Hugo (040 samples/-+-)",
                           "Melanoma-Riaz (139 samples/-+-)",
                           "Melanoma-Samstein (320 samples/+++)",
                           "Melanoma-Snyder-Nathanson (064 samples/+--)",
                           "Melanoma-Van Allen (112 samples/+--)",
                           "Hepatocellular Carcinoma-Harding (031 samples/-+-)",
                           "Non-Small Cell Lung Cancer-Hira Rizvi (240 samples/-++)",
                           "Non-Small Cell Lung Cancer-Naiyer Rizvi (035 samples/-+-)",
                           "Renal Cell Carcinoma-Miao (098 samples/-++)",
                           "Non-Small Cell Lung Cancer-Samstein (350 samples/-++)",
                           "Bladder Cancer-Samstein (215 samples/-++)",
                           "Breast Cancer-Samstein (044 samples/+++)",
                           "Cancer of Unknown Primary-Samstein (088 samples/+++)",
                           "Colorectal Cancer-Samstein (110 samples/+++)",
                           "Colorectal Cancer-Le (031 samples/-+-)",
                           "Esophagogastric Cancer-Samstein (126 samples/+++)",
                           "Glioma-Samstein (117 samples/-++)",
                           "Head and Neck Cancer-Samstein (139 samples/-++)",
                           "Renal Cell Carcinoma-Samstein (151 samples/-++)",
                           "Basal cell carcinoma-Yost (015 samples/-+-)")
  data_source_transcriptomic <- c("Melanoma-Auslander (037 samples/+++)",
                                  "Melanoma-Chen (104 samples/++-)",
                                  "Melanoma-Gide (091 samples/-++)",
                                  "Melanoma-Snyder-Nathanson (064 samples/+--)",
                                  "Melanoma-Hugo (040 samples/-+-)",
                                  "Melanoma-Prat (025 samples/-+-)",
                                  "Melanoma-Riaz (139 samples/-+-)",
                                  "Melanoma-Van Allen (112 samples/+--)",
                                  "Melanoma-Jerby-Arnon-ValCo2 (112 samples/-+-)",
                                  "Melanoma-Chen-Pre-On (023 samples)",
                                  "Metastatic Urothelial Cancer-Mariathasan (348 samples/-+-)",
                                  "Glioblastoma-Cloughesy (029 samples/-+-)",
                                  "Prostate Cancer-Kwek (022 samples/+--)",
                                  "Renal Cell Carcinoma-Miao (098 samples/-++)",
                                  "Gastric Cancer-Kim (061 samples/-+-)")
  
  data_source_integration <- unique(c(data_source_genetic,data_source_transcriptomic))

 
  observe(
    if (!is.null(input$start_left_table_rows_selected)  ) {
      enable("explore_selected_datasets")
    })

  
  observeEvent(input$explore_selected_datasets,{
    input$explore_selected_datasets
    #updateTabsetPanel(session=session,inputId = "main_tabs",selected="--Rank signatures")
    updateTabsetPanel(session=session,inputId = "main_tabs",selected="--Genetic: Mutation")
      data_source_list <- data_source_integration

      
    # updateSelectizeInput(session=session,inputId="data_source_prio",
    #                      choices=c(data_source_list),
    #                      #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
    #                      #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
    #                      selected=as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)"))),
    #                      
    #                      server=T)

    updateSelectizeInput(session=session,inputId="data_source_mutation",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         
                         server=T)
    # names_in_code =c("prio","mutation","mutation_loads","mutation_signature","expression","expression_sum","expression_pairs","cibersort",
    #                  "multi_dimensions","single_cell_gene","single_cell_cell","pan_cancer")
    
    updateSelectizeInput(session=session,inputId="data_source_mutation_loads",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_mutation_signature",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_expression",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_expression_sum",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_expression_pairs",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
      
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_cibersort",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),

                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_multi_dimensions",
                         choices=c(data_source_list),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),

                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_single_cell_gene",
                         choices = c("Melanoma-Sade_Feldman (048 samples/+++)" ,
                                     "Melanoma-Krieg_panel_1 (040 samples/-+-)" ,
                                     "Melanoma-Krieg_panel_2 (040 samples/-+-)" ,
                                     "Melanoma-Krieg_panel_3 (039 samples/-+-)" ),
                           #c("Melanoma-Sade_Feldman" ,"Melanoma-Krieg_panel_1" ,"Melanoma-Krieg_panel_2" ,"Melanoma-Krieg_panel_3" ),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),

                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_single_cell_cell",
                         choices = c("Melanoma-Sade_Feldman (048 samples/+++)" ,
                                     "Melanoma-Krieg_panel_1 (040 samples/-+-)" ,
                                     "Melanoma-Krieg_panel_2 (040 samples/-+-)" ,
                                     "Melanoma-Krieg_panel_3 (039 samples/-+-)" ),
                         #choices = c("Melanoma-Sade_Feldman" ,"Melanoma-Krieg_panel_1" ,"Melanoma-Krieg_panel_2" ,"Melanoma-Krieg_panel_3" ),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),

                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),

                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_mutation",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_mutation_loads",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_mutation_signature",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression_sum",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression_pairs",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_cibersort",
                         choices=c(data_source_list),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         #selected=as.character(unique(paste(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),start.table[input$start_left_table_rows_selected,"datasource"],sep="-"))),
                         selected=as.character(update_data_source_name[as.character(unique(paste0(as.character(start.table[input$start_left_table_rows_selected,"Disease"]),"-",start.table[input$start_left_table_rows_selected,"datasource"]," (",start.table[input$start_left_table_rows_selected,"Num."]," samples)")))]),
                         server=T)

  })
  

  ############################################
  #### Start Users' data
  ############################################
  user_clinical <- function () {
    if (is.null(input$file_clinical)) {
      return(NULL)
    } else {
      df_clinical <- fread(input$file_clinical$datapath,data.table = FALSE)
      df_clinical
    }
  }
  user_transcriptome <- function() {
    if (is.null(input$file_transcriptome)) {
      return(NULL)
    } else {
      df_transcriptome <- fread(input$file_transcriptome$datapath,data.table = FALSE)
      df_transcriptome
    }
  }
  user_mutation <- function () {
    if (is.null(input$file_mutation)) {
      return(NULL)
    } else {
      df_mutation <- fread(input$file_mutation$datapath,data.table = FALSE)
      df_mutation
    }
  }
  
  user_gene_cell <- function () {
    if (is.null(input$file_gene_cell)) {
      return(NULL)
    } else {
      #df_gene_cell <- fread(input$file_gene_cell$datapath,data.table = FALSE) ##looks  like read.table() can read rownames
      df_gene_cell <- read.table(input$file_gene_cell$datapath,header=T)
      df_gene_cell
    }
  }
  user_patient_cell <- function () {
    if (is.null(input$file_patient_cell)) {
      return(NULL)
    } else {
      df_patient_cell <- fread(input$file_patient_cell$datapath,data.table = FALSE)
      df_patient_cell
    }
  }
  
  database.test.process <- reactive({
    try.input <- try(input)
    if (try.input=="try-error") {
      database.test
    } else {
      #    if (input$do_user_data!=0) {
      if (!is.null(input$do_user_data)) {
        if (!is.null(input$file_clinical) | !is.null(input$file_transcriptome) | !is.null(input$file_mutation) |
            !is.null(input$file_gene_cell) | !is.null(input$file_patient_cell)) {
          
          #user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
          user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])),as.character(user_patient_cell()[,"PatientID"]))))
          
          cal_clinical <- eventReactive (input$do_user_data, {
            if (is.null(input$file_clinical)) {
              # data.frame(Disease=NA,PatientID=user_sample,data_source=NA,treatment=NA,description=NA,PFS.status=NA,RECIST=NA,
              #            OS.status=NA,OS=NA,PFS=NA,Age=NA) 
              data.frame(Disease=NA,PatientID=user_sample,data_source=NA,treatment=NA,description=NA,PFS.status=NA,RECIST=NA,
                         OS.status=NA,OS=NA,PFS=NA,Age=NA,Gender=NA) 
            } else {
              user_c <- user_clinical()
              colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
              colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
              user_c
              
            }
          })
          cal_transcriptome <- eventReactive (input$do_user_data, {
            if (is.null(input$file_transcriptome)) {
              data.frame(PatientID=user_sample) 
            } else {
              user_transcriptome()
            }
          })
          cal_cibersort <- eventReactive (input$do_user_data, {
            if (is.null(input$file_transcriptome)) {
              data.frame(PatientID=user_sample,B.cells.naive=NA,B.cells.memory=NA,Plasma.cells=NA,T.cells.CD8=NA,
                         T.cells.CD4.naive=NA,T.cells.CD4.memory.resting=NA,T.cells.CD4.memory.activated=NA,
                         T.cells.follicular.helper=NA,T.cells.regulatory..Tregs.=NA,T.cells.gamma.delta=NA,NK.cells.resting=NA,
                         NK.cells.activated=NA,Monocytes=NA,Macrophages.M0=NA,Macrophages.M1=NA,Macrophages.M2=NA,
                         Dendritic.cells.resting=NA,Dendritic.cells.activated=NA,Mast.cells.resting=NA,Mast.cells.activated=NA,
                         Eosinophils=NA,Neutrophils=NA)
            } else {
              
              #sig_matrix<-"H:/YJ/VUMC-Project/ImmuPortal/code/shiny example/www/LM22.txt"
              sig_matrix<-readRDS("LM22.rds")
              a<-user_transcriptome()
              b<-t(a)
              colnames(b)<-b[1,]
              b<-b[-1,]
              c<-apply(b,2,as.numeric)
              na.test <-  function (x) {
                w <- apply(x, 2, function(x)all(is.na(x)))
                if (any(w)) {
                  x<-x[,-which(w)]
                  #    stop(paste("All NA in columns", paste(which(w), collapse=", ")))
                }
                x
              }
              d<-na.test(c)
              rownames(d)<-rownames(b)
              e<-CIBERSORT(sig_matrix, d, perm=0, QN=FALSE, absolute=FALSE, abs_method='sig.score')
              data.frame(PatientID=rownames(e),e[,1:22])
            }
          })
          cal_mutation_loads <- eventReactive (input$do_user_data, {
            if (is.null(input$file_mutation)) {
              data.frame(PatientID=user_sample,Mutation.Load=NA)
            } else {
              df_mutation <- user_mutation()
              df_mutation_loads <- as.data.frame(table(df_mutation[,1]))
              colnames(df_mutation_loads) <- c("PatientID","Mutation.Load")
              df_mutation_loads
            }
          })
          cal_mutation_signature <- eventReactive (input$do_user_data,{
            if (is.null(input$file_mutation)) {
              data.frame(PatientID=user_sample,Signature.1=NA,Signature.2=NA,Signature.3=NA,Signature.4=NA,Signature.5=NA,
                         Signature.6=NA,Signature.7=NA,Signature.8=NA,Signature.9=NA,Signature.10=NA,Signature.11=NA,
                         Signature.12=NA,Signature.13=NA,Signature.14=NA,Signature.15=NA,Signature.16=NA,Signature.17=NA,
                         Signature.18=NA,Signature.19=NA,Signature.20=NA,Signature.21=NA,Signature.22=NA,Signature.23=NA,
                         Signature.24=NA,Signature.25=NA,Signature.26=NA,Signature.27=NA,Signature.28=NA,Signature.29=NA,
                         Signature.30=NA,unknown=NA)
            } else {
              df_mutation <- user_mutation()
              library(deconstructSigs)
              if (length(grep("chr",df_mutation[,"Chromosome"]))==0) {
                df_mutation[,"Chromosome"] <- paste0("chr",df_mutation[,"Chromosome"]) 
              } 
              chr <- c("chr1","chr2","chr3","chr4","chr5","chr6","chr7","chr8","chr9",
                       "chr10","chr11","chr12","chr13","chr14","chr15","chr16","chr17","chr18",
                       "chr19","chr20","chr21","chr22","chrX","chrY")
              df_mutation <- df_mutation[df_mutation[,"Chromosome"] %in% chr,]
              df_mutation.sigs.input <- mut.to.sigs.input(mut.ref = df_mutation,
                                                          sample.id = "PatientID", 
                                                          chr = "Chromosome", 
                                                          pos = "pos", 
                                                          ref = "ref", 
                                                          alt = "alt")
              df_mutation.signature <- NA
              for (i in rownames(df_mutation.sigs.input)){
                df_mutation.sample.single = whichSignatures(tumor.ref = df_mutation.sigs.input, 
                                                            signatures.ref = signatures.cosmic, 
                                                            sample.id = i, 
                                                            contexts.needed = TRUE,
                                                            tri.counts.method = 'default')
                df_mutation.signature <- rbind(df_mutation.signature,cbind(df_mutation.sample.single$weight,unknown=df_mutation.sample.single$unknown))
              }
              df_mutation.signature <- df_mutation.signature[-1,]
              df_mutation.signature
            }
          })
          
          
          
          
          user_combine <- function() {
            a<-merge(cal_clinical(),cal_transcriptome(),by="PatientID",all=T)
            b<-merge(cal_cibersort(),cal_mutation_loads(),by="PatientID",all=T)
            c<-merge(a,b,by="PatientID",all=T)
            d<-merge(c,cal_mutation_signature(),by.x="PatientID",by.y="row.names",all=T)
            d<-d[!is.na(d[,"PatientID"]) & !is.na(d[,"data_source"]),]
          }
          library(plyr)
          data.test.adduser <- rbind.fill(database.test,user_combine())
          data.test.adduser <- data.test.adduser[!is.na(data.test.adduser[,"PatientID"]) & !is.na(data.test.adduser[,"data_source"]),]
          data.test.adduser
        } else {
          database.test
        }
      } else {
        database.test
      } 
    }
  })

  all.mutations.process <- reactive({
    if (!is.null(input$file_clinical) | !is.null(input$file_transcriptome) | !is.null(input$file_mutation)) {
      if (!is.null(user_mutation())) {
        a <- user_mutation()
        colnames(a)[1] <- "sample"
        b<-data.frame(a[,1:5],PatientID=a[,1],data_source="User")
        rbind.fill(all.mutations,b)
      } else {
        all.mutations
      }
    } else {
      all.mutations
    }
  })
  
  all.cols <- isolate(database.test.process())
  a.colnames<-colnames(all.cols)[13:ncol(all.cols)]
  all.cols.mutation <- isolate(all.mutations.process())
  a.colnames.mutation<-unique(all.cols.mutation$gene)
  
  
  ############################################
  #### update input selections in all features
  #### only for disease, data source, 
  #### treatment, description
  ############################################
  # observe({
  #   database.tmp <- database.test[,1:12]
  #   updateSelectizeInput(session=session,inputId="treatment_single_cell_gene",
  #                        choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
  #                        selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
  #                        server=T)
  #   updateSelectizeInput(session=session,inputId="description_single_cell_gene",
  #                        choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "description"])),
  #                        selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "description"])),
  #                        server=T)
  #   updateSelectizeInput(session=session,inputId="treatment_single_cell_cell",
  #                        choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "treatment"])),
  #                        selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "treatment"])),
  #                        server=T)
  #   updateSelectizeInput(session=session,inputId="description_single_cell_cell",
  #                        choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "description"])),
  #                        selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "description"])),
  #                        server=T)
  # 
  # })
  # Browse <- c("Genetic","Transcriptomic","Single-cell")
  # #  names_in_code_browse <- c("genetic","transcriptomic","single_cell")
  # Features = c("Signatures prioritization","Mutation","Mutational loads/burden","Mutational signatures","Expression","Expression sum","Expression relation pairs","Immune cell components","Integration","Gene expression","Cell populations","Pan-cancer study")
  # Features_feature = c("Signatures prioritization","Genetic: Mutation","Genetic: Mutational loads/burden","Genetic: Mutational signatures","Transcriptomic: Expression","Transcriptomic: Expression sum","Transcriptomic: Expression relation pairs","Transcriptomic: Immune cell components","Integration","Single-cell: Gene expression","Single-cell: Cell populations","Pan-cancer study")
  # names_in_code =c("prio","mutation","mutation_loads","mutation_signature","expression","expression_sum","expression_pairs","cibersort",
  #                  "multi_dimensions","single_cell_gene","single_cell_cell","pan_cancer_assessment_mutation","pan_cancer")
  
  observe({
    ### due to dynaimc 
    database.tmp <- database.test.process()[,1:12]
    
    updateSelectizeInput(session=session,inputId="treatment_prio",
                         choices=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         selected=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_prio",
                         choices=c("Pre","On","Unknown"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         selected=c("Pre","On","Unknown"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_mutation",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_mutation",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_mutation_loads",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_loads,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_loads,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_mutation_loads",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_loads,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_loads,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_mutation_signature",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_signature,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_signature,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_mutation_signature",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_signature,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_mutation_signature,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_expression",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression,end=-19), "treatment"]))]),
                         server=T)
    if (input$dynamic_check_expression) {
      updateSelectizeInput(session=session,inputId="description_expression",
                               choices="Dynamic: (On-Pre)",
                               selected="Dynamic: (On-Pre)",
                               server=T)

    } else {
      updateSelectizeInput(session=session,inputId="description_expression",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression,end=-19), "description"])),
                         server=T)
    }
    updateSelectizeInput(session=session,inputId="treatment_expression_sum",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "treatment"]))]),
                         server=T)
    if (input$dynamic_check_expression_sum) {
      updateSelectizeInput(session=session,inputId="description_expression_sum",
                           choices="Dynamic: (On-Pre)",
                           selected="Dynamic: (On-Pre)",
                           server=T)
      
    } else {
      updateSelectizeInput(session=session,inputId="description_expression_sum",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "description"])),
                         server=T)
    }
    updateSelectizeInput(session=session,inputId="treatment_expression_pairs",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_pairs,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_pairs,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_expression_pairs",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_pairs,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_pairs,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_cibersort",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "treatment"]))]),
                         server=T)
    if (input$dynamic_check_cibersort) {
      updateSelectizeInput(session=session,inputId="description_cibersort",
                           choices="Dynamic: (On-Pre)",
                           selected="Dynamic: (On-Pre)",
                           server=T)
      
    } else {
      updateSelectizeInput(session=session,inputId="description_cibersort",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "description"])),
                         server=T)
    }
    updateSelectizeInput(session=session,inputId="treatment_multi_dimensions",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_multi_dimensions,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_multi_dimensions,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_multi_dimensions",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_multi_dimensions,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_multi_dimensions,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_single_cell_gene",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_single_cell_gene",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_single_cell_cell",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_single_cell_cell",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_cell,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                         choices=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         selected=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_pan_cancer",
                         choices=c("Pre","On","Unknown"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         selected=c("Pre","On","Unknown"),#as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_single_cell_gene,end=-19), "treatment"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_mutation",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_mutation",
                             choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19), "description"])),
                             selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19), "description"])),
                             server=T)
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_mutation_loads",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_mutation_loads",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_mutation_signature",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_mutation_signature",
                             choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19), "description"])),
                             selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19), "description"])),
                             server=T)
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_expression",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression,end=-19), "treatment"]))]),
                         server=T)
    if (input$dynamic_check_pan_cancer_assessment_expression) {
      updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_expression",
                           choices="Dynamic: (On-Pre)",
                           selected="Dynamic: (On-Pre)",
                           server=T)
      
    } else {
      updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_expression",
                           choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression,end=-19), "description"])),
                           selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression,end=-19), "description"])),
                           server=T)
    }
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_expression_sum",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19), "treatment"]))]),
                         server=T)
    if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
      updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_expression_sum",
                           choices="Dynamic: (On-Pre)",
                           selected="Dynamic: (On-Pre)",
                           server=T)
      
    } else {
      updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_expression_sum",
                           choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19), "description"])),
                           selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19), "description"])),
                           server=T)
    }
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_expression_pairs",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19), "treatment"]))]),
                         server=T)
    updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_expression_pairs",
                         choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19), "description"])),
                         selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19), "description"])),
                         server=T)
    updateSelectizeInput(session=session,inputId="treatment_pan_cancer_assessment_cibersort",
                         choices=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19), "treatment"]))]),
                         selected=unique(update_treatment_name[as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19), "treatment"]))]),
                         server=T)
    if (input$dynamic_check_pan_cancer_assessment_cibersort) {
      updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_cibersort",
                           choices="Dynamic: (On-Pre)",
                           selected="Dynamic: (On-Pre)",
                           server=T)

    } else {
    updateSelectizeInput(session=session,inputId="description_pan_cancer_assessment_cibersort",
                             choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19), "description"])),
                             selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19), "description"])),
                             server=T)
    }

  })
  

  
  
  observeEvent(input$do_user_data,{
    input$do_user_data
    data_source_single_cell_simple <- str_sub(data_source_single_cell,end=-19)
    data_source_integration_simple <- str_sub(data_source_integration,end=-19)
    
    all.disease <- as.character(unique(database.test.process()[,"Disease"]))
    all.data.source <- as.character(unique(database.test.process()[,"data_source"]))
    all.treatment <- as.character(unique(database.test.process()[,"treatment"]))
    all.description <- as.character(unique(database.test.process()[,"description"]))
    user.data.source <- all.data.source[!(all.data.source %in% c(data_source_integration_simple,data_source_single_cell_simple))]

    #### selection update

    updateSelectizeInput(session=session,inputId="data_source_mutation",
                         choices=c(data_source_genetic,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_mutation_loads",
                         choices=c(data_source_genetic,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_mutation_signature",
                         choices=c(data_source_genetic,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_expression",
                           choices=c(data_source_transcriptomic,user.data.source),
                           selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                           server=T) 
    updateSelectizeInput(session=session,inputId="data_source_expression_sum",
                         choices=c(data_source_transcriptomic,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_expression_pairs",
                         choices=c(data_source_transcriptomic,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_cibersort",
                         choices=c(data_source_transcriptomic,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_multi_dimensions",
                         choices=c(data_source_integration,user.data.source),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_single_cell_gene",
                         choices=c(data_source_single_cell,user.data.source),
                         selected="Melanoma-Sade_Feldman (048 samples/+++)",
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_single_cell_cell",
                         choices=c(data_source_single_cell,user.data.source),
                         selected="Melanoma-Sade_Feldman (048 samples/+++)",
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                         choices = c(data_source_integration,user.data.source),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_mutation",
                         choices = c(data_source_genetic,user.data.source),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_mutation_signature",
                         choices = c(data_source_genetic,user.data.source),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression",
                         choices = c(data_source_transcriptomic,user.data.source),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression_sum",
                         choices = c(data_source_transcriptomic,user.data.source),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)
    updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_cibersort",
                         choices = c(data_source_transcriptomic,user.data.source),
                         #choices=c("Hugo","Riaz","Snyder","Van Allen","Samstein","Harding","Hira Rizvi","Naiyer Rizvi"),
                         selected=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
                         server=T)

  })



  summary_plot_browse <- function(inputdata){
    withProgress(message = 'Making plot', value = 0, {
      if (nrow(inputdata)==0) {
        incProgress(100, detail = "Doing part %")
        plot_ly () %>% 
          layout(title = paste(str_sub(input$data_source_genetic,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
          add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
      } else {
        disease_column <- data.frame(table(inputdata$Disease),stringsAsFactors = FALSE)
        colnames(disease_column) <-c("cancer","cancer_count")
        
        data_source_column <- data.frame(table(inputdata$data_source),stringsAsFactors = FALSE)
        colnames(data_source_column) <-c("data_source","data_source_count")
        
        treatment_column <- data.frame(table(inputdata$treatment),stringsAsFactors = FALSE)
        colnames(treatment_column) <- c("treatment","treatment_count")
        
        description_column <- data.frame(table(inputdata$description),stringsAsFactors = FALSE)
        colnames(description_column) <- c("description","description_count")
        
        PFS_status_column <- data.frame(table(inputdata$PFS.status),stringsAsFactors = FALSE)
        if (nrow(PFS_status_column)==0) {
          PFS_status_column<-data.frame(PFS_status="unknown",PFS_status_count=nrow(inputdata))
        }
        colnames(PFS_status_column) <- c("PFS_status","PFS_status_count")
        
        OS_status_column <- data.frame(table(inputdata$OS.status),stringsAsFactors = FALSE)
        if (nrow(OS_status_column)==0) {
          OS_status_column<-data.frame(OS_status="unknown",OS_status_count=nrow(inputdata))
        }
        colnames(OS_status_column) <- c("OS_status","OS_status_count")
        
        description_column[,"description"] <- as.character(description_column[,"description"])
        description_column[description_column[,"description"]=="Pre","description"]<-"Pre-treatment"
        description_column[description_column[,"description"]=="On","description"]<-"On-treatment"
        
        all.column <- rbind.fill(disease_column,data_source_column,treatment_column,description_column,PFS_status_column,OS_status_column)
        
        incProgress(100, detail = "Doing part %")
        plot_ly(all.column) %>%  
          add_pie(all.column, labels= ~cancer,value= ~cancer_count,type="pie",
                  textposition='inside',textinfo='label',
                  domain=list(row=0,column=0),hole = 0.2) %>%
          add_pie(all.column, labels= ~data_source,value= ~data_source_count,type="pie",
                  textposition='inside',textinfo='label',
                  domain=list(row=0,column=1),hole = 0.2) %>%
          add_pie(all.column, labels= ~treatment,value= ~treatment_count,type="pie",
                  textposition='inside',textinfo='label',
                  domain=list(row=0,column=2),hole = 0.2) %>%
          add_pie(all.column, labels= ~description,value= ~description_count,type="pie",
                  textposition='inside',textinfo='label',
                  domain=list(row=1,column=0),hole = 0.2) %>%
          add_pie(all.column, labels= ~PFS_status,value= ~PFS_status_count,type="pie",
                  textposition='inside',textinfo='label',
                  domain=list(row=1,column=1),hole = 0.2) %>%
          add_pie(all.column, labels= ~OS_status,value= ~OS_status_count,type="pie",
                  textposition='inside',textinfo='label',
                  domain=list(row=1,column=2),hole = 0.2) %>%
          layout(title="Statistical summary of selected studies",titlefont=list(size=18,color="black",family= "Aril"),
                 showlegend = F,
                 grid=list(rows=2,columns=3)) 
        
      }
      
    }) 
  }
  
  summary_plot <- function(inputdata,feature){
    withProgress(message = 'Making plot', value = 0, {
    if (nrow(inputdata)==0) {
      incProgress(100, detail = "Doing part %")
      plot_ly () %>% 
        layout(title = paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
        add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
    } else {
    disease_column <- data.frame(table(inputdata$Disease),stringsAsFactors = FALSE)
    colnames(disease_column) <-c("cancer","cancer_count")
  
    data_source_column <- data.frame(table(inputdata$data_source),stringsAsFactors = FALSE)
    colnames(data_source_column) <-c("data_source","data_source_count")
    
    treatment_column <- data.frame(table(inputdata$treatment),stringsAsFactors = FALSE)
    colnames(treatment_column) <- c("treatment","treatment_count")
    
    description_column <- data.frame(table(inputdata$description),stringsAsFactors = FALSE)
    colnames(description_column) <- c("description","description_count")
  
    PFS_status_column <- data.frame(table(inputdata$PFS.status),stringsAsFactors = FALSE)
    if (nrow(PFS_status_column)==0) {
      PFS_status_column<-data.frame(PFS_status="unknown",PFS_status_count=nrow(inputdata))
    }
    colnames(PFS_status_column) <- c("PFS_status","PFS_status_count")
    
    OS_status_column <- data.frame(table(inputdata$OS.status),stringsAsFactors = FALSE)
    if (nrow(OS_status_column)==0) {
      OS_status_column<-data.frame(OS_status="unknown",OS_status_count=nrow(inputdata))
    }
    colnames(OS_status_column) <- c("OS_status","OS_status_count")
    
    description_column[,"description"] <- as.character(description_column[,"description"])
    description_column[description_column[,"description"]=="Pre","description"]<-"Pre-treatment"
    description_column[description_column[,"description"]=="On","description"]<-"On-treatment"
    
    if (feature=="mutation"| feature=="multi_dimensions" | feature=="single_cell_cell" |
        feature=="mutation_loads" | feature=="mutation_signature" | feature=="expression" | 
        feature=="expression sum" | feature=="expression pairs" | feature=="cibersort") {
      category_column <- data.frame(table(inputdata$category),stringsAsFactors = FALSE)
      colnames(category_column) <- c("category","category_count")
      all.column <- rbind.fill(disease_column,data_source_column,treatment_column,description_column,PFS_status_column,OS_status_column,category_column)
      incProgress(100, detail = "Doing part %")
      plot_ly(all.column) %>%  
        add_pie(all.column, labels= ~cancer,value= ~cancer_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~data_source,value= ~data_source_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=1),hole = 0.2) %>%
        add_pie(all.column, labels= ~treatment,value= ~treatment_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=2),hole = 0.2) %>%
        add_pie(all.column, labels= ~description,value= ~description_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~PFS_status,value= ~PFS_status_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=1),hole = 0.2) %>%
        add_pie(all.column, labels= ~OS_status,value= ~OS_status_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=2),hole = 0.2) %>%
        add_pie(all.column, labels= ~category,value= ~category_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=2,column=0),hole = 0.2) %>%
        layout(title="Statistical summary of selected studies",titlefont=list(size=18,color="black",family= "Aril"),
               showlegend = F,
               grid=list(rows=3,columns=3)) 
    } else if (feature=="single_cell_gene") {
      all.column <- rbind.fill(disease_column,data_source_column,treatment_column,description_column,PFS_status_column,OS_status_column)
      incProgress(100, detail = "Doing part %")
      plot_ly(all.column) %>%  
        add_pie(all.column, labels= ~cancer,value= ~cancer_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~data_source,value= ~data_source_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=1),hole = 0.2) %>%
        add_pie(all.column, labels= ~treatment,value= ~treatment_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=2),hole = 0.2) %>%
        add_pie(all.column, labels= ~description,value= ~description_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~PFS_status,value= ~PFS_status_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=1),hole = 0.2) %>%
        add_pie(all.column, labels= ~OS_status,value= ~OS_status_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=2),hole = 0.2) %>%
        layout(title="Statistical summary of selected studies",titlefont=list(size=18,color="black",family= "Aril"),
               showlegend = F,
               grid=list(rows=2,columns=3))
    } else {
      category_OS_column <- data.frame(table(inputdata$category.OS),stringsAsFactors = FALSE)
      colnames(category_OS_column) <- c("category_OS","category_OS_count")
      category_PFS_column <- data.frame(table(inputdata$category.PFS),stringsAsFactors = FALSE)
      colnames(category_PFS_column) <- c("category_PFS","category_PFS_count")
      all.column <- rbind.fill(disease_column,data_source_column,treatment_column,description_column,PFS_status_column,OS_status_column,category_OS_column,category_PFS_column)
      incProgress(100, detail = "Doing part %")
      plot_ly(all.column) %>%  
        add_pie(all.column, labels= ~cancer,value= ~cancer_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~data_source,value= ~data_source_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=1),hole = 0.2) %>%
        add_pie(all.column, labels= ~treatment,value= ~treatment_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=0,column=2),hole = 0.2) %>%
        add_pie(all.column, labels= ~description,value= ~description_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~PFS_status,value= ~PFS_status_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=1),hole = 0.2) %>%
        add_pie(all.column, labels= ~OS_status,value= ~OS_status_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=1,column=2),hole = 0.2) %>%
        add_pie(all.column, labels= ~category_OS,value= ~category_OS_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=2,column=0),hole = 0.2) %>%
        add_pie(all.column, labels= ~category_PFS,value= ~category_PFS_count,type="pie",
                textposition='inside',textinfo='label',
                domain=list(row=2,column=1),hole = 0.2) %>%
        layout(title="Statistical summary of selected studies",titlefont=list(size=20,color="black",family= "Aril"),
               showlegend = F,
               grid=list(rows=3,columns=3)) 
    }
  
    
    }
      
    }) 
  }
  
  clinical_ui = c("treatment type","Pre- or On-treatment","RECIST","Age","Gender")
  clinical_server =c("treatment","description","RECIST","Age","Gender")
  names(clinical_server) <- clinical_ui
  
#  shinyjs::hide(id = "loading-expression-summary", anim = TRUE, animType = "fade")
  
  names_in_code =c("mutation","mutation_loads","mutation_signature","expression",
                   "expression_sum","expression_pairs","cibersort",
                   "multi_dimensions")
  names_in_code_pan_cancer <- c("pan_cancer_assessment_individual_mutation",
                                "pan_cancer_assessment_individual_mutation_loads",
                                "pan_cancer_assessment_individual_mutation_signature",
                                "pan_cancer_assessment_individual_expression",
                                "pan_cancer_assessment_individual_expression_sum",
                                "pan_cancer_assessment_individual_expression_pairs",
                                "pan_cancer_assessment_individual_cibersort")
  names_in_main_function=c("selectedmutation_mutation","selecteddatabase_mutation_loads","selecteddatabase_mutation_signature",
                           "selecteddatabase_expression","selecteddatabase_expression_sum","selecteddatabase_expression_pairs",
                           "selecteddatabase_cibersort","selecteddatabase_multi_dimensions_final")
  names_in_main_function_pan_cancer=c("selectedmutation_pan_cancer_assessment_mutation",
                                      "selecteddatabase_pan_cancer_assessment_mutation_loads",
                                      "selecteddatabase_pan_cancer_assessment_mutation_signature",
                                      "selecteddatabase_pan_cancer_assessment_expression",
                                      "selecteddatabase_pan_cancer_assessment_expression_sum",
                                      "selecteddatabase_pan_cancer_assessment_expression_pairs",
                                      "selecteddatabase_pan_cancer_assessment_cibersort")
#  names_in_code_browse <- c("genetic","transcriptomic","single-cell")
  


  ###################################
  #### all summary clinical plot UI
  ###################################
#  selected.database<-NULL
    for (l in 1:length(names_in_code)) {
    local({
      my_i <- l
      output[[paste0(names_in_code[my_i],"_summary_clinical_plots")]] <- renderUI({
        input[[paste0("do_",names_in_code[my_i])]]
        selected.database.tmp <- names_in_main_function[my_i]
        selected.database <- get(selected.database.tmp)()
        #assign("selected.database",names_in_main_function[1])
        #selected.database <- get("selected.database")
        if (nrow(selected.database)==0 ) {
          tagList(plotlyOutput(paste0(names_in_code[my_i],"_summary_clinical_plots0")),height="auto")
        } else {
          lapply(1:length(input[[paste0("data_source_",names_in_code[my_i])]]),function(i) {
            column(width=6,wellPanel(plotlyOutput(paste0(names_in_code[my_i],"_summary_clinical_plots",i),height="auto")),style="padding-bottom:20px")
            
          })
        }
      })
      # ###################################
      # #### all plot plot UI
      # ###################################
      
      
          output[[paste0(names_in_code[my_i],"_plots_plots")]] <- renderUI({
            input[[paste0("do_",names_in_code[my_i])]]
            if(my_i==1) {
              selected.database <- selecteddatabase_mutation()
              selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
              selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input$gene_mutation,]
              if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
                tagList(plotlyOutput("mutation_plots_plots0",height="auto"))
              } else {
                lapply(1:length(input$data_source_mutation),function(i) {
                  column(width=12,wellPanel(plotlyOutput(paste0("mutation_plots_plots",i),height="auto")),style="padding-bottom:20px")
                })
              }
            } else {
              selected.database.tmp <- names_in_main_function[my_i]
              selected.database <- get(selected.database.tmp)()
              if (nrow(selected.database)==0 | ncol(selected.database)<13) {
                tagList(plotlyOutput(paste0(names_in_code[my_i],"_plots_plots0")),height="auto")
             } else {
               lapply(1:length(input[[paste0("data_source_",names_in_code[my_i])]]),function(i) {
                 column(width=6,wellPanel(plotlyOutput(paste0(names_in_code[my_i],"_plots_plots",i),height="auto")),style="padding-bottom:20px")
                })
              }
           }
          })
      
      output[[paste0(names_in_code_pan_cancer[my_i],"_plots_plots")]] <- renderUI({
        input[[paste0("do_",names_in_code_pan_cancer[my_i])]]
        if(my_i==1) {
          selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
          selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
          selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input$gene_pan_cancer_assessment_mutation,]
          if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
            tagList(plotlyOutput("pan_cancer_assessment_individual_mutation_plots_plots0",height="auto"))
          } else {
            lapply(1:length(input$data_source_pan_cancer_assessment_mutation),function(i) {
              column(width=12,wellPanel(plotlyOutput(paste0("pan_cancer_assessment_individual_mutation_plots_plots",i),height="auto")),style="padding-bottom:20px")
            })
          }
        } else {
          selected.database.tmp <- names_in_main_function_pan_cancer[my_i]
          selected.database <- get(selected.database.tmp)()
          if (nrow(selected.database)==0 | ncol(selected.database)<13) {
            tagList(plotlyOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_plots_plots0")),height="auto")
          } else {
            lapply(1:length(input[[paste0("data_source_pan_cancer_assessment_",names_in_code[my_i])]]),function(i) {
              column(width=6,wellPanel(plotlyOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_plots_plots",i),height="auto")),style="padding-bottom:20px")
            })
          }
        }
      })
      
      # ###################################
      # #### all survival plot UI
      # ###################################
      
          output[[paste0(names_in_code[my_i],"_survival_plots")]] <- renderUI({
            input[[paste0("do_",names_in_code[my_i])]]
            if(my_i==1) {
              selected.database <- selecteddatabase_mutation()
              selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
              selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input$gene_mutation,]
              if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
                tagList(plotlyOutput("mutation_survival_plots0",height="auto"))
              } else {
                lapply(1:length(input$data_source_mutation),function(i) {
                  column(width=6,wellPanel(plotlyOutput(paste0("mutation_survival_plots",i),height="auto"),
                                           div(DT::dataTableOutput(paste0("mutation_survival_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                           style="height:800px"),
                         style="padding-bottom:20px")
                })
              }
            } else {
              selected.database.tmp <- names_in_main_function[my_i]
              selected.database <- get(selected.database.tmp)()
              if (nrow(selected.database)==0 | ncol(selected.database)<13) {
                tagList(plotlyOutput(paste0(names_in_code[my_i],"_survival_plots0")),height="auto")
              } else {
                lapply(1:length(input[[paste0("data_source_",names_in_code[my_i])]]),function(i) {
                  column(width=6,wellPanel(plotlyOutput(paste0(names_in_code[my_i],"_survival_plots",i),height="auto"),
                                           div(DT::dataTableOutput(paste0(names_in_code[my_i],"_survival_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                           style="height:800px"),
                         
                         
                         style="padding-bottom:20px")
                })
              }
            }
          })
      
      output[[paste0(names_in_code_pan_cancer[my_i],"_survival_plots")]] <- renderUI({
        input[[paste0("do_",names_in_code_pan_cancer[my_i])]]
        if(my_i==1) {
          selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
          selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
          selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input$gene_pan_cancer_assessment_mutation,]
          if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
            tagList(plotlyOutput("pan_cancer_assessment_individual_mutation_survival_plots0",height="auto"))
          } else {
            lapply(1:length(input$data_source_pan_cancer_assessment_mutation),function(i) {
              column(width=6,wellPanel(plotlyOutput(paste0("pan_cancer_assessment_individual_mutation_survival_plots",i),height="auto"),
                                       div(DT::dataTableOutput(paste0("pan_cancer_assessment_individual_mutation_survival_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                       style="height:800px"),
                     style="padding-bottom:20px")
            })
          }
        } else {
          selected.database.tmp <- names_in_main_function_pan_cancer[my_i]
          selected.database <- get(selected.database.tmp)()
          if (nrow(selected.database)==0 | ncol(selected.database)<13) {
            tagList(plotlyOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_survival_plots0")),height="auto")
          } else {
            lapply(1:length(input[[paste0("data_source_pan_cancer_assessment_",names_in_code[my_i])]]),function(i) {
              column(width=6,wellPanel(plotlyOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_survival_plots",i),height="auto"),
                                       div(DT::dataTableOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_survival_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                       style="height:800px"),
                     
                     
                     style="padding-bottom:20px")
            })
          }
        }
      })
          
      # ###################################
      # #### all PFS plot UI
      # ###################################
      
          output[[paste0(names_in_code[my_i],"_PFS_plots")]] <- renderUI({
            input[[paste0("do_",names_in_code[my_i])]]
            if(my_i==1) {
              selected.database <- selecteddatabase_mutation()
              selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
              selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input$gene_mutation,]
              if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
                tagList(plotlyOutput("mutation_PFS_plots0",height="auto"))
              } else {
                lapply(1:length(input$data_source_mutation),function(i) {
                  column(width=6,wellPanel(plotlyOutput(paste0("mutation_PFS_plots",i),height="auto"),
                                           div(DT::dataTableOutput(paste0("mutation_PFS_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                           style="height:800px"),
                         style="padding-bottom:20px")
                })
              }
            } else {
              selected.database.tmp <- names_in_main_function[my_i]
              selected.database <- get(selected.database.tmp)()
              if (nrow(selected.database)==0 | ncol(selected.database)<13) {
                tagList(plotlyOutput(paste0(names_in_code[my_i],"_PFS_plots0")),height="auto")
              } else {
                lapply(1:length(input[[paste0("data_source_",names_in_code[my_i])]]),function(i) {
                  column(width=6,wellPanel(plotlyOutput(paste0(names_in_code[my_i],"_PFS_plots",i),height="auto"),
                                           div(DT::dataTableOutput(paste0(names_in_code[my_i],"_PFS_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                           style="height:800px"),
                         style="padding-bottom:20px")
                })
              }
            }
          })
      
      output[[paste0(names_in_code_pan_cancer[my_i],"_PFS_plots")]] <- renderUI({
        input[[paste0("do_",names_in_code_pan_cancer[my_i])]]
        if(my_i==1) {
          selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
          selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
          selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input$gene_pan_cancer_assessment_mutation,]
          if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
            tagList(plotlyOutput("pan_cancer_assessment_individual_mutation_PFS_plots0",height="auto"))
          } else {
            lapply(1:length(input$data_source_pan_cancer_assessment_mutation),function(i) {
              column(width=6,wellPanel(plotlyOutput(paste0("pan_cancer_assessment_individual_mutation_PFS_plots",i),height="auto"),
                                       div(DT::dataTableOutput(paste0("pan_cancer_assessment_individual_mutation_PFS_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                       style="height:800px"),
                     style="padding-bottom:20px")
            })
          }
        } else {
          selected.database.tmp <- names_in_main_function_pan_cancer[my_i]
          selected.database <- get(selected.database.tmp)()
          if (nrow(selected.database)==0 | ncol(selected.database)<13) {
            tagList(plotlyOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_PFS_plots0")),height="auto")
          } else {
            lapply(1:length(input[[paste0("data_source_pan_cancer_assessment_",names_in_code[my_i])]]),function(i) {
              column(width=6,wellPanel(plotlyOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_PFS_plots",i),height="auto"),
                                       div(DT::dataTableOutput(paste0("pan_cancer_assessment_individual_",names_in_code[my_i],"_PFS_summary_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                       style="height:800px"),
                     style="padding-bottom:20px")
            })
          }
        }
      })
      # ###################################
      # #### all evaluation table
      # ###################################
      output[[paste0(names_in_code[my_i],"_evaluation_table")]] <- DT::renderDataTable({
        selected.database.tmp <- names_in_main_function[my_i]
        selected.database <- get(selected.database.tmp)()
        if (nrow(selected.database)==0 | ncol(selected.database) >500) {
          datatable(data.frame("NA"="no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
          splitted_list <- splitted_list[ str_sub(input[[paste0("data_source_",names_in_code[my_i])]],end=-19) ]
          final.table <- NULL
          for (k in 1:length(input[[paste0("data_source_",names_in_code[my_i])]])) {
            my_i <- k
            d<-splitted_list[[my_i]]
            if (is.na(names(splitted_list)[my_i])) {
              next
            } else {
              d$category <- as.character(d$category)
              d$PFS.status <- as.character(d$PFS.status)
              unit.table <- as.data.frame.matrix(table(d[,c("category","PFS.status")]))
              unit.table <- data.frame(category=rownames(unit.table),unit.table)
              final.table <- rbind.fill(final.table,data.frame(data_source=unique(as.character(d$data_source)),unit.table,hiddenColumn=c(rep(0,(nrow(unit.table)-1)),1)))
            }
          }
          DT::datatable(final.table,options = list(columnDefs=list(list(visible=FALSE,targets=4)),searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE,selection = "single") %>%
            formatStyle(0:4,valueColumns="hiddenColumn",`border-bottom`=styleEqual(1,'1px solid black')) #%>%
          
        }
        
      })
      
      
      
      
      
    })
    }
  #######################################
  #### signatures prioritization
  #######################################
  # input <- list(data_source_prio=c( "Melanoma-Hugo (040 samples/-+-)","Melanoma-Riaz (139 samples/-+-)",
  #                                   "Melanoma-Samstein (320 samples/+++)",
  #                                   "Melanoma-Snyder-Nathanson (064 samples/+--)",
  #                                   "Melanoma-Van Allen (112 samples/+--)",
  #                                   "Melanoma-Jerby-Arnon-ValCo2 (112 samples/-+-)",
  #                                   "Melanoma-Auslander (037 samples/+++)",
  #                                   "Melanoma-Chen (104 samples/++-)",
  #                                   "Melanoma-Gide (091 samples/-++)",
  #                                   "Melanoma-Prat (025 samples/-+-)",
  #                                   "Non-Small Cell Lung Cancer-Hira Rizvi (240 samples/-++)",
  #                                   "Non-Small Cell Lung Cancer-Naiyer Rizvi (035 samples/-+-)",
  #                                   "Non-Small Cell Lung Cancer-Samstein (350 samples/-++)",
  #                                   "Metastatic Urothelial Cancer-Mariathasan (348 samples/-+-)",
  #                                   "Glioblastoma-Cloughesy (029 samples/-+-)",
  #                                   "Renal Cell Carcinoma-Miao (098 samples/-++)",
  #                                   "Hepatocellular Carcinoma-Harding (031 samples/-+-)",
  #                                   "Bladder Cancer-Samstein (215 samples/-++)",
  #                                   "Breast Cancer-Samstein (044 samples/+++)",
  #                                   "Cancer of Unknown Primary-Samstein (088 samples/+++)",
  #                                   "Colorectal Cancer-Samstein (110 samples/+++)",
  #                                   "Esophagogastric Cancer-Samstein (126 samples/+++)",
  #                                   "Glioma-Samstein (117 samples/-++)",
  #                                   "Head and Neck Cancer-Samstein (139 samples/-++)",
  #                                   "Renal Cell Carcinoma-Samstein (151 samples/-++)",
  #                                   "Prostate Cancer-Kwek (022 samples/+--)",
  #                                   "Gastric Cancer-Kim (061 samples/-+-)",
  #                                   "Basal cell carcinoma-Yost (015 samples/-+-)"),
  #               treatment_prio=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
  #               description_prio=c("Pre","On","Unknown"))
  
  # input <- list(data_source_prio=c( "Melanoma-Hugo (040 samples/-+-)","Melanoma-Riaz (139 samples/-+-)",
  #                                   "Melanoma-Samstein (320 samples/+++)",
  #                                   "Melanoma-Snyder-Nathanson (064 samples/+--)",
  #                                   "Melanoma-Van Allen (112 samples/+--)",
  #                                   "Non-Small Cell Lung Cancer-Hira Rizvi (240 samples/-++)",
  #                                   "Metastatic Urothelial Cancer-Mariathasan (348 samples/-+-)",
  #                                   "Head and Neck Cancer-Samstein (139 samples/-++)"
  # ),
  # treatment_prio=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
  # description_prio=c("Pre","On"))
  

  observe({
    ### when all samples
    if (input$sample_prio=="custom selection") {
      enable("data_source_prio")
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           server=T)
      enable("treatment_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                           choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           #selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           server = T)
      enable("description_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           #selected=c("Pre","On","Unknown"),
                           server = T)
    }
    if (input$sample_prio=="all samples") {
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           selected = meta_analysis_prio_choices,
                           server=T)
      disable("data_source_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              server = T)
      disable("treatment_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           selected=c("Pre","On","Unknown"),
                           server = T)
      disable("description_prio")
    }
    ### when all sample with anti-CTLA-4
    if (input$sample_prio=="samples with anti-CTLA-4 treatment") {
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           selected = meta_analysis_prio_selected_anti_CTLA_4,
                           server=T)
      disable("data_source_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                           choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           selected="anti-CTLA-4",
                           server = T)
      disable("treatment_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           selected=c("Pre","On","Unknown"),
                           server = T)
      disable("description_prio")
    }
    ### when all samples with anti-PD-1
    if (input$sample_prio=="samples with anti-PD-1/L1 treatment") {
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           selected = meta_analysis_prio_selected_anti_PD_1,
                           server=T)
      disable("data_source_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                           choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           selected="anti-PD-1/L1",
                           server = T)
      disable("treatment_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           selected=c("Pre","On","Unknown"),
                           server = T)
      disable("description_prio")
    }
    # ### when all samples with anti-PD-1+anti-CTLA-4
    # if (input$sample_prio=="samples with anti-PD-1+anti-CTLA-4 treatment") {
    #   updateSelectizeInput(session=session,inputId="data_source_prio",
    #                        choices = c( "Melanoma-Hugo (040 samples/-+-)","Melanoma-Riaz (139 samples/-+-)",
    #                                     "Melanoma-Samstein (320 samples/+++)",
    #                                     "Melanoma-Snyder-Nathanson (064 samples/+--)",
    #                                     "Melanoma-Van Allen (112 samples/+--)",
    #                                     "Melanoma-Jerby-Arnon-ValCo2 (112 samples/-+-)",
    #                                     "Melanoma-Auslander (037 samples/+++)",
    #                                     "Melanoma-Chen (104 samples/++-)",
    #                                     "Melanoma-Gide (091 samples/-++)",
    #                                     "Melanoma-Prat (025 samples/-+-)",
    #                                     "Non-Small Cell Lung Cancer-Hira Rizvi (240 samples/-++)",
    #                                     "Non-Small Cell Lung Cancer-Naiyer Rizvi (035 samples/-+-)",
    #                                     "Non-Small Cell Lung Cancer-Samstein (350 samples/-++)",
    #                                     "Metastatic Urothelial Cancer-Mariathasan (348 samples/-+-)",
    #                                     "Glioblastoma-Cloughesy (029 samples/-+-)",
    #                                     "Renal Cell Carcinoma-Miao (098 samples/-++)",
    #                                     "Hepatocellular Carcinoma-Harding (031 samples/-+-)",
    #                                     "Bladder Cancer-Samstein (215 samples/-++)",
    #                                     "Breast Cancer-Samstein (044 samples/+++)",
    #                                     "Cancer of Unknown Primary-Samstein (088 samples/+++)",
    #                                     "Colorectal Cancer-Samstein (110 samples/+++)",
    #                                     "Esophagogastric Cancer-Samstein (126 samples/+++)",
    #                                     "Glioma-Samstein (117 samples/-++)",
    #                                     "Head and Neck Cancer-Samstein (139 samples/-++)",
    #                                     "Renal Cell Carcinoma-Samstein (151 samples/-++)",
    #                                     "Prostate Cancer-Kwek (022 samples/+--)",
    #                                     "Gastric Cancer-Kim (061 samples/-+-)",
    #                                     "Basal cell carcinoma-Yost (015 samples/-+-)"),
    #                        selected = c( "Melanoma-Samstein (320 samples/+++)",
    #                                      "Melanoma-Auslander (037 samples/+++)",
    #                                      "Melanoma-Gide (091 samples/-++)",
    #                                      "Non-Small Cell Lung Cancer-Hira Rizvi (240 samples/-++)",
    #                                      "Non-Small Cell Lung Cancer-Samstein (350 samples/-++)",
    #                                      "Renal Cell Carcinoma-Miao (098 samples/-++)",
    #                                      "Bladder Cancer-Samstein (215 samples/-++)",
    #                                      "Breast Cancer-Samstein (044 samples/+++)",
    #                                      "Cancer of Unknown Primary-Samstein (088 samples/+++)",
    #                                      "Colorectal Cancer-Samstein (110 samples/+++)",
    #                                      "Esophagogastric Cancer-Samstein (126 samples/+++)",
    #                                      "Glioma-Samstein (117 samples/-++)",
    #                                      "Head and Neck Cancer-Samstein (139 samples/-++)",
    #                                      "Renal Cell Carcinoma-Samstein (151 samples/-++)"),
    #                        server=T)
    #   disable("data_source_prio")
    #   updateSelectizeInput(session=session,inputId="treatment_prio",
    #                        choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
    #                        selected="anti-PD-1/L1+anti-CTLA-4",
    #                        server = T)
    #   disable("treatment_prio")
    #   updateSelectizeInput(session=session,inputId="description_prio",
    #                        choices=c("Pre","On","Unknown"),
    #                        selected=c("Pre","On","Unknown"),
    #                        server = T)
    #   disable("description_prio")
    # }
    #### when samples before treatment
    if (input$sample_prio=="samples before treatment") {
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           selected = meta_analysis_prio_selected_before_treatment,
                           server=T)
      disable("data_source_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                           choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           server = T)
      disable("treatment_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           selected="Pre",
                           server = T)
      disable("description_prio")
    }
    #### when sample after treatment 
    if (input$sample_prio=="samples after treatment") {
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           selected = meta_analysis_prio_selected_after_treatment,
                           server=T)
      disable("data_source_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                           choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           server = T)
      disable("treatment_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           selected="On",
                           server = T)
      disable("description_prio")
    }
    ### when case study 1
    if (input$sample_prio=="case study 1") {
      updateSelectizeInput(session=session,inputId="data_source_prio",
                           choices = meta_analysis_prio_choices,
                           selected = meta_analysis_prio_selected_case_study_1,
                           server=T)
      disable("data_source_prio")
      updateSelectizeInput(session=session,inputId="treatment_prio",
                           choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                           server = T)
      disable("treatment_prio")
      updateSelectizeInput(session=session,inputId="description_prio",
                           choices=c("Pre","On","Unknown"),
                           selected=c("Pre","On","Unknown"),
                           server = T)
      disable("description_prio")
    }
  })
  
  observe({
    if (input$data_source_prio=="" || is.null(input$data_source_prio) ||
        input$treatment_prio=="" || is.null(input$treatment_prio) ||
        input$description_prio=="" || is.null(input$description_prio)){
      disable("do_prio")
    } else {
      enable("do_prio")
    }
  })
  
  selecteddatabase_prio <- eventReactive(input$do_prio,{
    data_source1 <- str_sub(input$data_source_prio,end=-19)
    # treatment1 <- input$treatment_prio
    treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_prio])
    description1 <- input$description_prio
    genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
    a<-data.table(database.test.process())
    m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
    m<-m[,colnames(m) %in% genes1,with=FALSE]
    as.data.frame(m)
    
  })
  ############################
  #### summary
  ############################
  output$prio_summary_disease_number <- renderText({
    input$do_prio ##keep these two rows
    isolate({ ##keep these two rows
      length(unique(selecteddatabase_prio()$Disease))})})
  output$prio_summary_sample_number <- renderText({
    input$do_prio ##keep these two rows
    isolate({ ##keep these two rows
      nrow(selecteddatabase_prio())})})
  output$prio_summary_plots <- renderPlotly({
    summary_plot_browse(inputdata=selecteddatabase_prio())
  })
  #############################
  #### signatures prio mutation
  #############################
  
  ### meta-analysis ###
  # observeEvent(input$do_prio,{
  #   isolate({
  #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
  #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
  #     subset.browse_prio_mutation_meta_p <- browse_prio_mutation_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
  #     if (length(subset.browse_prio_mutation_meta_p)==0) {
  #       #output$prio_mutation_plots_meta <- renderPlotly({
  #       prio_mutation_plots_meta_p <- plot_ly () %>% 
  #         layout(title = "meta-anlaysis across selected cohorts",font=list(size=15,family="Aril")) %>%
  #         add_annotations(x=2,y=2,text="warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
  #       #})
  #       prio_mutation_tables_meta_data <- datatable(data.frame("NA"="warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
  #     } else {
  #       fit <- density(na.omit(subset.browse_prio_mutation_meta_p))
  #       vline <- function(x = 0, color = "red") {
  #         list(
  #           type = "line", 
  #           y0 = 0, 
  #           y1 = 1, 
  #           yref = "paper",
  #           x0 = x, 
  #           x1 = x, 
  #           line = list(color = color)
  #         )}
  #       prio_mutation_plots_meta_p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
  #         layout(title="meta-anlaysis across selected cohorts",
  #                height = 480,
  #                xaxis = list(title="-log10(p.value)",
  #                             titlefont=list(size=20,color="black",family= "Aril black"),
  #                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
  #                yaxis=list(title="Density",
  #                           titlefont=list(size=20,color="black",family= "Aril black"),
  #                           showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
  #                shapes=list(vline(1.3))) %>%
  #         add_annotations(x = 1.3,
  #                         y = 1,
  #                         text = "p.value = 0.05",
  #                         xref = paste0("x", 1),  # add this
  #                         yref = paste0("y", 1),  # add this
  #                         showarrow = FALSE,
  #                         ax = 20,
  #                         ay = -40,
  #                         font=list(size=20,color="black",family= "Aril black"))
  #       
  #       
  #       prio_mutation_meta_table <- data.frame("p.value"=subset.browse_prio_mutation_meta_p,
  #                                              "FDR"=p.adjust(subset.browse_prio_mutation_meta_p))
  #       prio_mutation_meta_table_sig <- prio_mutation_meta_table[prio_mutation_meta_table$p.value<=0.05 & !is.na(prio_mutation_meta_table$p.value),]
  #       prio_mutation_meta_table_sig <- prio_mutation_meta_table_sig[order(prio_mutation_meta_table_sig$p.value),]
  #       shinyInput <- function(FUN, len, id, ...) {
  #         inputs <- character(len)
  #         for (i in seq_len(len)) {
  #           inputs[i] <- as.character(FUN(paste0(id, rownames(prio_mutation_meta_table_sig)[i],"_",i), ...))
  #         }
  #         inputs
  #       }
  #       d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
  #         # Gene=rownames(d.nonna.t.p0.05),
  #         # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
  #         # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
  #         signif(prio_mutation_meta_table_sig,2),
  #         View=shinyInput(actionButton,nrow(prio_mutation_meta_table_sig),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_mutation_meta\",  this.id)'),
  #         stringsAsFactors = FALSE#,
  #         #row.names = 1:nrow(d.nonna.t.p0.05)
  #       ))
  #       prio_mutation_tables_meta_data <- d.nonna.t.p0.05.test$data
  #       
  #     }
  #     output$prio_mutation_plots_meta <- renderPlotly({
  #       prio_mutation_plots_meta_p     
  #     })
  #     output$prio_mutation_tables_meta <- DT::renderDataTable(
  #       prio_mutation_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none'  
  #     )
  #   })  
  # })
  
  #### meta-analysis ### for table only
  
  observeEvent(input$do_prio,{
    isolate({
      withProgress(message = 'Preparing data', value = 0, {
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
      browse_prio_mutation_meta_selected <- browse_prio_mutation_meta[
        browse_prio_mutation_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
      if (nrow(browse_prio_mutation_meta_selected)==0 | 
          all(is.na(browse_prio_mutation_meta_selected$TE)) |
          all(is.na(browse_prio_mutation_meta_selected$seTE))  ){
        prio_mutation_tables_meta_data <- datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        # meta.p <- NA
        # meta.fdr <- NA
        fit.t <- NULL
      } else {
        incProgress(20, detail = "Doing part %")
        if (input$sample_prio=="custom selection") {
          splitted <- browse_prio_mutation_meta_selected %>% split(.$genes)
          fit <- lapply(splitted,
                      function(x) {
                        if (all(is.na(x$TE)) | all(is.na(x$seTE)) | nrow(x)==0 | is.null(x)) {
                          c(NA,NA,NA)
                        } else {
                          a <- metagen(TE,seTE,data=x,sm="OR")
                          #c(a$pval.fixed,a$pval.random,a$pval.Q)})
                          c(a$pval.random,a$pval.fixed,a$pval.Q)
                        }
                      })
          
          fit.t <- t(as.data.frame(fit))
          fit.t.fdr.random <- p.adjust(fit.t[,1],"fdr")
          fit.t.fdr.fixed <- p.adjust(fit.t[,2],"fdr")
          # fit.t <- data.frame(pval.fixed=fit.t[,1],
          #                   fdr.fixed=fit.t.fdr,
          #                   pval.random=fit.t[,2],
          #                   pval.Q=fit.t[,3])
          # fit.t <- fit.t[order(fit.t$pval.fixed),]
          fit.t <- data.frame(Pval.random=fit.t[,1],
                              FDR.random=fit.t.fdr.random,
                              Pval.fixed=fit.t[,2],
                              FDR.fixed=fit.t.fdr.fixed)
          fit.t <- fit.t[order(fit.t$Pval.random),]
          
        } else {
          fit.t <- meta_analysis_prio_precalculated_mutation[[input$sample_prio]][,1:4]
          #fit.t <- fit.t[order(fit.t$Pval.random),]
        }
        #c("pval.fixed","pval.random","pval.Q")
        
        shinyInput <- function(FUN, len, id, ...) {
          inputs <- character(len)
          for (i in seq_len(len)) {
            inputs[i] <- as.character(FUN(paste0(id, rownames(fit.t)[i],"_",i), ...))
          }
          inputs
        }
        d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
          # Gene=rownames(d.nonna.t.p0.05),
          # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
          # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
          signif(fit.t,2),
          View=shinyInput(actionButton,nrow(fit.t),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_mutation_meta\",  this.id)'),
          stringsAsFactors = FALSE#,
          #row.names = 1:nrow(d.nonna.t.p0.05)
        ))
        prio_mutation_tables_meta_data <- d.nonna.t.p0.05.test$data
      }
        subset.browse_prio_mutation.t_fisher_part_pre <- as.data.frame(t(browse_prio_mutation[rownames(browse_prio_mutation) %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
        incProgress(80, detail = "Doing part %")
        if(nrow(subset.browse_prio_mutation.t_fisher_part_pre)>0) {
          if (input$sample_prio=="custom selection") {
            colnames(subset.browse_prio_mutation.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_mutation.t_fisher_part_pre),"_",2))[1,])
            subset.browse_prio_mutation.t_meta_part <- fit.t
            combined.p <- fisher.method(subset.browse_prio_mutation.t_fisher_part_pre,na.rm=TRUE)$p.value
            subset.browse_prio_mutation.t_fisher_part <- data.frame("combined p-value"=combined.p,
                                                                  "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
                                                                  subset.browse_prio_mutation.t_fisher_part_pre,stringsAsFactors = F )
            subset.browse_prio_mutation.t <- merge(subset.browse_prio_mutation.t_meta_part,subset.browse_prio_mutation.t_fisher_part,by="row.names",all=T)
            rownames(subset.browse_prio_mutation.t) <- subset.browse_prio_mutation.t[,1]
            subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[,-1]
          
            subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[order(subset.browse_prio_mutation.t[,1]),]
          # subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[,-which(apply(subset.browse_prio_mutation.t,2,function(x) all(is.na(x))))]
            subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[,apply(subset.browse_prio_mutation.t,2,function(x) !all(is.na(x)))]
            # brks <- quantile(subset.browse_prio_mutation.t, probs = seq(.05, .95, .05), na.rm = TRUE)
            # clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}
          } else {
            subset.browse_prio_mutation.t <- meta_analysis_prio_precalculated_mutation[[input$sample_prio]]
          }
          brks <- quantile(subset.browse_prio_mutation.t, probs = seq(.05, .95, .05), na.rm = TRUE)
          clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}
          
        } else {
          subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t_fisher_part_pre
          
        }
      #}
      incProgress(100, detail = "Doing part %")
      
      
      
      output$prio_mutation_tables_meta <- DT::renderDataTable(
        
        prio_mutation_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none',caption = 'Pval.random: p-value for test of overall treatment effect (random effects model). FDR.random: adjusted p-value of random effects model using false discovery rate method. Pval.fixed: p-value for test of overall treatment effect (fixed effect model). FDR.fixed: adjusted p-value of fixed effect model using false discovery rate method.'
      )
      
      output$prio_mutation_download <- downloadHandler(
        filename = function() {
          if (input$prio_mutation_type == "Excel (CSV)") {
            "newfile.csv"
          } else if (input$prio_mutation_type == "Text (TSV)") {
            "newfile.txt"
          } else if (input$prio_mutation_type == "JSON") {
            "newfile.json"
          }
        },
        content = function(file) {
          if (input$prio_mutation_type == "Excel (CSV)") {
            fwrite (subset.browse_prio_mutation.t,row.names=T,file)
          } else if (input$prio_mutation_type == "Text (TSV)") {
            fwrite (subset.browse_prio_mutation.t,row.names=T,file)
            #write.table (selectedmutation_mutation(),file,quote=F, row.names = FALSE, sep="\t")
          } else if (input$prio_mutation_type == "JSON") {
            write(toJSON(subset.browse_prio_mutation.t),row.names=T,file)
          }
        }
      )
      output$prio_mutation_table <- DT::renderDataTable({
        datatable(subset.browse_prio_mutation.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
          formatStyle(names(subset.browse_prio_mutation.t), backgroundColor = styleInterval(brks, clrs)) %>%
          formatSignif(columns = c(1:ncol(subset.browse_prio_mutation.t)), digits = 2,getOption("scipen"))
      })
      })
    })
  })
  
  observeEvent(input$select_button_prio_mutation_meta,{
    s <- as.numeric(strsplit(input$select_button_prio_mutation_meta, "_")[[1]][3])
    Viewname <- as.character(strsplit(input$select_button_prio_mutation_meta, "_")[[1]][2])
    output$prio_mutation_tables_meta_modals <- renderUI({
      tagList(
        bsModal(paste0("prio_mutation_tables_meta_bsModal",s), Viewname, "select_button_prio_mutation_meta", size = "large",
                plotOutput("browse_prio_mutation_meta_bsModal_plot")
        )
      )
    })
    toggleModal(session,paste0("prio_mutation_tables_meta_bsModal",s), toggle = Viewname)
    ##Reset the select_button
    session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_mutation_meta")
  })
  observeEvent(input$select_button_prio_mutation_meta,{
    input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
    input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
    browse_prio_mutation_meta_selected.singlegene <- browse_prio_mutation_meta[
      browse_prio_mutation_meta$genes==as.character(strsplit(input$select_button_prio_mutation_meta, "_")[[1]][2]) & 
        browse_prio_mutation_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
    output$browse_prio_mutation_meta_bsModal_plot <- renderPlot({
      # forest(metabin(event.res,n.res,event.nonres,n.nonres,
      #                 data=browse_prio_mutation_meta_selected,
      #                 sm="OR",
      #                 label.e = "response",label.c = "nonresponse",
      #                 studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
      #        test.overall = TRUE,ff.study.label = "bold")
      if (nrow(browse_prio_mutation_meta_selected.singlegene)==0 |
          is.null(browse_prio_mutation_meta_selected.singlegene) |
          all(is.na(browse_prio_mutation_meta_selected.singlegene$TE)) |
          all(is.na(browse_prio_mutation_meta_selected.singlegene$seTE))) {
        plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_prio,end=-19),collapse = ","),cex.main=1.5,bty="l")
        abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
        text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
      } else {
        
        m1 <- metagen(TE,seTE,data=browse_prio_mutation_meta_selected.singlegene,
                sm="OR",
                #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
                studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1]))
          m1$upper[exp(m1$upper)>1e+50] <- 1000
          m1$upper.fixed[exp(m1$upper.fixed)>1e+50] <- 1000
          m1$upper.random[exp(m1$upper.random)>1e+50] <- 1000
          m1$lower.fixed[exp(m1$lower.fixed)>1e+50] <- 1000
          m1$lower.random[exp(m1$lower.random)>1e+50] <- 1000
          
          forest(m1,layout="RevMan5",   
              pooled.total=TRUE,
               hetstat=TRUE,
               plotwidth="6inch",test.overall = TRUE,
               xlim=c(0.1,10),backtransfer=T,
               label.test.overall.random="random effects model: ",
               label.test.overall.fixed="fixed effects model: ",
               fs.test.overall = 15,
               scientific.pval = TRUE,ff.study.label = "bold")
               # 
               # leftcols=c("effect","w.fixed","w.random"))
               # #rightcols=c("effect","w.fixed","w.random"))
      }
    })
    
  })
  ### each dataset ####
  output$prio_mutation_plots <- renderUI({
    input$do_prio
    isolate({
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
      subset.browse_prio_mutation <- browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
      
      #selected.database_genetic_mutation <- selecteddatabase_genetic_mutation()
      if (nrow(subset.browse_prio_mutation)==0 ) {
        tagList(plotlyOutput("prio_mutation_plots0",height="auto"),dataTableOutput("prio_mutation_tables0"))
      } else {
        lapply(1:length(input$data_source_prio),function(i) {
          column(width=6,wellPanel(plotlyOutput(paste0("prio_mutation_plots",i),height="auto"),
                                   #h4(paste0("Significant genes in ",str_sub(input$data_source_prio,end=-19)[i]),style="margin-top:20px;margin-bottom:10px;font-size:20px;font-weight:bold;color:#DCB239"),
                                   h4("Significant genes:",style="margin-top:20px;margin-bottom:10px;font-size:20px;font-weight:bold;color:#DCB239"),
                                   # h4("Select signature to check (click row in the table above)",
                                   #    style="margin-top:10px;margin-bottom:10px;font-size:20px;font-weight:bold;color:black"),
                                   #
                                   # disabled(actionButton(paste0("prio_mutation_check",i),"View",style="background-color: lightblue;margin-left:275px;margin-bottom:20px;")),

                                   div(DT::dataTableOutput(paste0("prio_mutation_tables",i)),style="margin-top:30px;margin-bottom:10px"),
                                   #bsModal(paste0("prio_mutation_modal",i),"Plot", "", size = "large", uiOutput("test_ui")),
                                   #uiOutput("prio_mutation_modals",height = "auto"),

                                   style="height:1300px"),

                 style="padding-bottom:20px")



          #column(width=6,wellPanel(DT::dataTableOutput(paste0("genetic_mutation_tables",i))),style="padding-bottom:20px")
        })
        # lapply(1:length(input$data_source_prio),function(i) {
        # uiOutput("prio_mutation_modals",height = "auto")
        # })

      }
    })
  })
  
  # 
  observeEvent(input$do_prio,{
    input$do_prio
    isolate({
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
      subset.browse_prio_mutation <- browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
      
      #subset.browse_prio_mutation <- browse_prio_mutation[rownames(browse_prio_mutation) %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
      if (nrow(subset.browse_prio_mutation)==0) {
        output$prio_mutation_plots0 <- renderPlotly({
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_prio,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
        })

        output$prio_mutation_tables0 <- DT::renderDataTable({
          datatable(data.frame("NA"="no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        })

      } else {
        # splitted_list <- selected.database_prio_mutation %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        # splitted_list <- splitted_list[input$data_source_prio]
        isolate({
          for (k in 1:length(input$data_source_prio)) {
            local({
              my_i <- k
              browse_prio_mutation_plot_name <- paste0("prio_mutation_plots", my_i)
              browse_prio_mutation_table_name <- paste0("prio_mutation_tables", my_i)
              #browse_prio_mutation_modal_name <- paste0("prio_mutation_modals", my_i)
              #             d<-splitted_list[[my_i]]

              output[[browse_prio_mutation_plot_name]] <- renderPlotly({
                isolate({
                  withProgress(message = 'Making plot', value = 0, {
                    if (strsplit(rownames(subset.browse_prio_mutation)[my_i],"\\.")[[1]][1]=="NA") {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>%
                        layout(title = str_sub(input$data_source_prio,end=-19)[my_i],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    } else {
                      if (length(na.omit(as.numeric(browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))==0) {
                        incProgress(100, detail = "Doing part %")
                        plot_ly () %>%
                          layout(title = unique(str_sub(input$data_source_prio,end=-19)[my_i]),font=list(size=15,family="Aril")) %>%
                          add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                      } else {
                        vline <- function(x = 0, color = "red") {
                          list(
                            type = "line",
                            y0 = 0,
                            y1 = 1,
                            yref = "paper",
                            x0 = x,
                            x1 = x,
                            line = list(color = color)
                          )
                        }
                        d <- browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                        d.nonna <- d[,!is.na(d[1,])]
                        d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))))
                        rownames(d.nonna.t) <- colnames(d.nonna)
                        colnames(d.nonna.t) <- c(paste0("p-value in ",rownames(d.nonna)),paste0("-log10(p-value) in ",rownames(d.nonna)))
                        fit <- density(d.nonna.t[,2])
                        # p<-plot_ly(x=d.nonna.t[,2],
                        #            #y=as.numeric(d[-1]),
                        #            type="histogram",#type="box",
                        #            #histnorm="probability",
                        #            name="Histogram"
                        # ) %>%
                        #   add_trace(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",yaxis="y2",name="Density") %>%
                        p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
                          layout(title=str_sub(input$data_source_prio,end=-19)[my_i],
                                 height = 480,
                                 xaxis = list(title="-log10(p.value)",
                                              titlefont=list(size=20,color="black",family= "Aril black"),
                                              showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title="Density",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 shapes=list(vline(1.3))) %>%
                          add_annotations(x = 1.3,
                                          y = 1,
                                          text = "p.value = 0.05",
                                          xref = paste0("x", 1),  # add this
                                          yref = paste0("y", 1),  # add this
                                          showarrow = FALSE,
                                          ax = 20,
                                          ay = -40,
                                          font=list(size=20,color="black",family= "Aril black"))
                        incProgress(100, detail = "Doing part %")
                        p
                      }
                    }
                  })
                })
              })


                isolate({
                  if (strsplit(rownames(subset.browse_prio_mutation)[my_i],"\\.")[[1]][1]=="NA") {
                    prio_mutation_tables_data <-
                    datatable(data.frame("NA"="no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    if (length(na.omit(as.numeric(browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))==0) {
                      prio_mutation_tables_data <-
                      datatable(data.frame("NA"="no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {

                      d <- browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                      d.nonna <- d[,!is.na(d[1,])]
                      if (ncol(d.nonna)==0) {
                        prio_mutation_tables_data <-
                        datatable(data.frame("NA"="no clinical data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                      } else {
                        #d.nonna.t <- data.frame(signif(data.frame(t(d.nonna)),3),signif(-log10(data.frame(as.numeric(t(d.nonna)))),3),signif(p.adjust(as.numeric(t(d.nonna)),"fdr"),3))
                        #d.nonna.t <- data.frame(data.frame(t(signif(d.nonna,3))),signif(-log10(data.frame(as.numeric(t(d.nonna)))),3),signif(p.adjust(as.numeric(t(d.nonna)),"fdr"),3))
                        d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))),p.adjust(as.numeric(t(d.nonna)),"fdr"))

                        rownames(d.nonna.t) <- colnames(d.nonna)
                        #colnames(d.nonna.t) <- c(paste0("p-value in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("-log10(p-value) in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("FDR in ",str_split(rownames(d.nonna),"_",2)[[1]][1]))
                        colnames(d.nonna.t) <- c("p-value","-log10(p-value)","FDR")
                        d.nonna.t.p0.05 <- d.nonna.t[d.nonna.t[,1]<=0.05,]
                        #if(nrow(d.nonna.t.p0.05) > 0) {
                          d.nonna.t.p0.05 <- d.nonna.t.p0.05[order(d.nonna.t.p0.05[,1]),]
                        #   d.nonna.t.p0.05 <- apply(d.nonna.t.p0.05,2,function(x) formatC(x,format="E",digits=3))
                        # }
                        #d.nonna.t.p0.05[is.na(d.nonna.t.p0.05)] <- "NA"
                        #datatable(d.nonna.t.p0.05[order(d.nonna.t.p0.05[,1]),])
                          shinyInput <- function(FUN, len, id, ...) {
                            inputs <- character(len)
                            for (i in seq_len(len)) {
                              inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
                            }
                            inputs
                          }
                          d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
                                                        # Gene=rownames(d.nonna.t.p0.05),
                                                        # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
                                                        # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
                                                        signif(d.nonna.t.p0.05[,c(1,3)],2),
                                                        View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_"),label="View",onclick='Shiny.onInputChange(\"select_button_prio_mutation\",  this.id)'),
                                                        stringsAsFactors = FALSE#,
                                                        #row.names = 1:nrow(d.nonna.t.p0.05)
                                                        ))
                          prio_mutation_tables_data <- d.nonna.t.p0.05.test$data
                        #datatable(d.nonna.t.p0.05.test$data,selection = "none"
                                  # options=list(
                                  #   rowCallback=JS(
                                  #     "function(row, data) {",
                                  #     "for (i = 1; i < data.length; i++) {",
                                  #     "if (data[i]<0.001 | data[i]>1000){",
                                  #     "$('td:eq('+i+')', row).html(data[i].toExponential(3));",
                                  #     "}",
                                  #
                                  #     "}",
                                  #     "}"
                                  #   )
                                  # )
                        #) %>% formatSignif(columns = c(2:ncol(d.nonna.t.p0.05.test$data)), digits = 2,getOption("scipen"))
                        #datatable(d.nonna.t.p0.05,options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                  }
                })
                output[[browse_prio_mutation_table_name]] <- DT::renderDataTable(
                  prio_mutation_tables_data,server = FALSE, escape = FALSE, selection = 'none'
                  #d.nonna.t.p0.05.test$data,server = FALSE, escape = FALSE, selection = 'none'
              )



            })
          }
        })
      }
    })

  })

  observeEvent(input$select_button_prio_mutation,{
    s <- as.numeric(strsplit(input$select_button_prio_mutation, "_")[[1]][3])
    output$prio_mutation_modals <- renderUI({
    tagList(
      bsModal(paste('prio_mutation_model', s ,sep=''), "View", "select_button_prio_mutation", size = "large",

              plotlyOutput("browse_prio_mutation_bsModal_plot")
              #textOutput("browse_prio_mutation_bsModal_plot")
              )
    )
    })
    toggleModal(session,paste('prio_mutation_model', s ,sep=''), toggle = "View")
    ##Reset the select_button
    #session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_mutation")
  })

   observeEvent(input$select_button_prio_mutation,{

    # output$browse_prio_mutation_bsModal_plot <- renderText({
    #   input$select_button_prio_mutation
    # })

    output$browse_prio_mutation_bsModal_plot <- renderPlotly({
      #input$select_button
      #strsplit(input$select_button, "_")[[1]][2]

       isolate({
        input_gene_mutation <- as.character(strsplit(input$select_button_prio_mutation, "_")[[1]][2])

        data_source1 <- as.character(strsplit(input$select_button_prio_mutation, "_")[[1]][1])
        treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_prio])
        description1 <- input$description_prio
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
        a<-data.table(database.test)
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        selected.database<- as.data.frame(m)


        selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]


        if (nrow(selected.database)==0) {
          selected.mutations.function <- selected.database
        } else {
          gene1 <- input_gene_mutation
          selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% gene1,]
          if (nrow(selected.mutations.1)>0 ) {
            #### important: resharp mutation file to database.test format
            b<-aggregate(variant_class~(PatientID+gene),data=selected.mutations.1,FUN=paste0)
            c <- b %>% spread(gene,variant_class,fill="no mutation")
            c[c=="NULL"]<-"no mutation"

            if ((ncol(c)-1) < length(gene1)) {
              c.add <- matrix("no mutation",nrow(c),length(gene1[gene1 %ni% colnames(c)[2:ncol(c)]]))
              colnames(c.add) <- gene1[gene1 %ni% colnames(c)[2:ncol(c)]]
              c <- cbind(c, c.add)
            }

            c.rest <- data.frame(PatientID=unique(selected.mutations[selected.mutations[,"PatientID"] %ni% selected.mutations.1[,"PatientID"],"PatientID"]))
            c <- rbind.fill(c,c.rest)
            c[c=="NULL"] <- "no mutation"
            c[is.na(c)] <- "no mutation"

            #c <- data.frame(c,category=apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") }))
            c <- data.frame(c,category="no mutation")
            c[,"category"] <- as.character(c[,"category"])
            c[!apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") == "no mutation" }),"category"]<- "mutated"

            selected.database <- merge(selected.database,c,by="PatientID",all.x=T)
            selected.database[is.na(selected.database[,"category"]),"category"] <- "unknown"
            selected.mutations.function <- as.data.frame(selected.database)
          } else {
            selected.database <- data.frame(selected.database,category="unknown")
            selected.mutations.function <- selected.database
          }
        }




        #selected.mutations.function <- selectedmutation_mutation()
        #selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_mutation,]
        if (nrow(selected.database)==0 | nrow(selected.mutations)==0) {

            plot_ly () %>%
              layout(title = "No data",font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))




        } else {
          selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_mutation,]

          if (nrow(selected.mutations.1)==0) {

              plot_ly () %>%
                layout(title = "No data",font=list(size=15,family="Aril")) %>%
                add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple"))





          } else {



            selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
            splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
            #splitted_list <- splitted_list[input$data_source_mutation]
            #      splitted_list_mutation <- selected.mutations.1 %>% as.matrix %>% as.data.frame %>% split(.$data_source)
            splitted_list_mutation <- selected.mutations #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
            #splitted_list_mutation <- splitted_list_mutation[input$data_source_mutation]
            splitted_list_mutation.function <- selected.mutations.function #%>% split(.$data_source)
            #splitted_list_mutation.function <- splitted_list_mutation.function[input$data_source_mutation]

            #for (k in 1:length(input$data_source_mutation)) {
              #local({
                # <- k
                #mutation_plots_plots_name <- paste0("mutation_plots_plots", my_i)
                d<-splitted_list#[[my_i]]
                d.mutations <- splitted_list_mutation#[[my_i]]
                d.mutations.function <- splitted_list_mutation.function#[[my_i]]
                #output[[mutation_plots_plots_name]] <- renderPlotly({
                  #withProgress(message = 'Making plot', value = 0, {
                    isolate({
                      # if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_mutation)[my_i])) {
                      #   incProgress(100, detail = "Doing part %")
                      #   plot_ly () %>%
                      #     layout(title = input$data_source_mutation[[my_i]],font=list(size=15,family="Aril")) %>%
                      #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                      # } else {
                        if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d.mutations)==0) {
                          #incProgress(100, detail = "Doing part %")
                          plot_ly () %>%
                            layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                            add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                        } else {
                          #                    withProgress(message = 'Making plot', value = 0, {
                          d.mutations.function <- d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & d.mutations.function[,"category"]!="unknown",]

                          d.table <- unlist(table(d.mutations.function[,c("PFS.status","category")]))
                          d.table <- d.table[rownames(d.table) %in% c("nonresponse","response"),colnames(d.table) %in% c("mutated","no mutation")]
                          try.test <- try(p.v <- fisher.test(d.table)$p.value)
                          #try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")]))[1:2,])$p.value)
                          if (class(try.test)=="try-error" | is.na(try.test)) {
                            p.v <- " = NA"
                          } else {

                            p.v<-paste0(" = ",signif(p.v,2))


                          }



                          ind <- sort(unique(d[,"Groups"]))
                          d.mutations.res <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[2] & !is.na(d[,"Groups"])),"PatientID"] ,]
                          d.mutations.nonres <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[1] & !is.na(d[,"Groups"])),"PatientID"] ,]

                          d.mutations.res<-d.mutations.res[,-1]
                          #d.mutations.res<-d.mutations.res[,c("PatientID","gene","variant_class","Protein_change","Chromosome","data_source")]
                          d.mutations.res<-d.mutations.res[,c(6,2:5,7)]
                          colnames(d.mutations.res)[1] <- "sample"
                          if (nrow(d.mutations.res[d.mutations.res[,"gene"] %in% input_gene_mutation,])==0) {
                            p1 <- ggplot(d.mutations.res, aes(x=sample,y="NA")) +
                              xlab(paste0("response samples (n=",length(unique(d.mutations.res$sample)),")")) +
                              theme(axis.text.x=element_text(angle=50, hjust=1,size=4),
                                    axis.text.y=element_text(size=8,colour='black'),
                                    axis.title.x=element_text(size=11),
                                    axis.title.y=element_blank())
                            p.bar.res <- data.frame(PFS.status="response",genes=input_gene_mutation,Number=0,n=length(unique(d.mutations.res$sample)))
                            p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                            # inputDat <- waterfall_qual(d.mutations.res, clinData, mutBurden, file_type="Custom",
                            #                            label_col=NULL)
                            # data_frame <- inputDat[[1]]
                            # p2 <- ggplot(na.omit(data_frame),aes_string(x=NA, y=0, fill='trv_type')) +
                            #   geom_bar(position='stack', alpha=.75, width=1, stat='identity') +
                            #   theme_bw() + coord_flip() +
                            #   theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.title.y=element_blank(),legend.position=('none'),
                            #         axis.text.x=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank()) +
                            #   ylab("% mutated samples") + ylim(100,0)+ geom_blank()
                            # p3 <- subplot(style(p2,showlegend=T),style(p1,showlegend=T),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)

                            # p3 <- plot_ly () %>%
                            #   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple"))
                          } else {
                            p1<-#ggplotly(
                              waterfall(d.mutations.res, fileType = "Custom",
                                        variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))), ### here all.mutation is important, if not, different data source will have different color enve in the same mutation type
                                        plotGenes=unlist(str_split(input_gene_mutation,",")),
                                        mainXlabel = TRUE,my_x_label=ind[2],
                                        plotMutBurden = FALSE,returnp="p1",
                                        mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                      '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                      '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                      '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                      '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                      '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                      '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                      '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                            Number.genes.res <- NULL
                            for (j in 1:length(input_gene_mutation)) {
                              Number.genes.res <- c(Number.genes.res,length(unique(d.mutations.res[d.mutations.res[,"gene"]==input_gene_mutation[j],"sample"])))
                            }
                            p.bar.res <- data.frame(PFS.status="response",genes=input_gene_mutation,Number=Number.genes.res,n=length(unique(d.mutations.res$sample)))
                            p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                          }
                          d.mutations.nonres<-d.mutations.nonres[,-1]
                          d.mutations.nonres<-d.mutations.nonres[,c(6,2:5,7)]
                          colnames(d.mutations.nonres)[1] <- "sample"
                          if (nrow(d.mutations.nonres[d.mutations.nonres[,"gene"] %in% input_gene_mutation,])==0) {
                            q1 <- ggplot(d.mutations.nonres, aes(x=sample,y="NA")) +
                              xlab(paste0("nonresponse samples (n=",length(unique(d.mutations.nonres$sample)),")")) +
                              theme(axis.text.x=element_text(angle=50, hjust=1,size=4),
                                    axis.text.y=element_text(size=8,colour='black'),
                                    axis.title.x=element_text(size=11),
                                    axis.title.y=element_blank())
                            p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_mutation,Number=0,n=length(unique(d.mutations.nonres$sample)))
                            p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                            # inputDat <- waterfall_qual(d.mutations.nonres, clinData, mutBurden, file_type="Custom",
                            #                            label_col=NULL)
                            # data_frame <- inputDat[[1]]
                            # q2 <- ggplot(na.omit(data_frame),aes_string(x=NA, y=0, fill='trv_type')) +
                            #   geom_bar(position='stack', alpha=.75, width=1, stat='identity') +
                            #   theme_bw() + coord_flip() +
                            #   theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.title.y=element_blank(),legend.position=('none')) +
                            #
                            #   ylab("% mutated samples") + ylim(100,0)+ geom_blank()
                            # q3 <- subplot(style(q2,showlegend=F),style(q1,showlegend=F),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                            #
                            # q3 <- plot_ly () %>%
                            #   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple"))
                          } else {
                            q1<-#ggplotly(
                              waterfall(d.mutations.nonres, fileType = "Custom",
                                        variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                                        plotGenes=unlist(str_split(input_gene_mutation,",")),
                                        mainXlabel = TRUE,my_x_label=ind[1],
                                        plotMutBurden = FALSE,returnp="p1",
                                        mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                      '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                      '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                      '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                      '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                      '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                      '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                      '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                            Number.genes.nonres <- NULL
                            for (j in 1:length(input_gene_mutation)) {
                              Number.genes.nonres <- c(Number.genes.nonres,length(unique(d.mutations.nonres[d.mutations.nonres[,"gene"]==input_gene_mutation[j],"sample"])))
                            }
                            p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_mutation,Number=Number.genes.nonres,n=length(unique(d.mutations.nonres$sample)))
                            p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                          }

                          p.bar.res <- data.frame(p.bar.res,perc=signif(p.bar.res$Number/p.bar.res$n*100,3))
                          p.bar.nonres <- data.frame(p.bar.nonres,perc=signif(p.bar.nonres$Number/p.bar.nonres$n*100,3))

                          if (sum(p.bar.res$perc)==0 & sum(p.bar.nonres$perc)!=0 ) {
                            p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                            p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                          } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)==0) {
                            p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                            p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                          } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)!=0) {
                            p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                            p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                          } else {
                            p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                            p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                          }
                          #p.bar <- rbind(p.bar.res,p.bar.nonres)
                          p.bar <- rbind(p.bar.nonres,p.bar.res)
                          p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
                          bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
                            theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
                            geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                            #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                            geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))


                          #incProgress(100, detail = "Doing part %")
                          subplot(style(bar.plot,showlegend=F),
                                  subplot(p1,q1,nrows=2,margin=0.1,titleX=TRUE,titleY=TRUE) %>%
                                    layout(#title = as.character(unique(d$data_source)),
                                      titlefont=list(size=20,color="black",family= "Aril"),
                                      legend=list(font=list(size=10))),
                                  #xaxis=list(title=lable_x,side=side_x,titlefont=list(size=15))),
                                  widths=c(1/3,2/3),margin=0.05,titleX=TRUE,titleY=TRUE) %>%
                            layout(title=as.character(unique(d$data_source)))
                          #                  })
                        }

                      #}
                    })
                  #})
                #})
              #})
            #}

          }
        }


      })


    })
   })
  # #### download ####
   # observeEvent(input$do_prio,{
   #   input$do_prio
   #   isolate({
   #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
   #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
   #     subset.browse_prio_mutation.t_fisher_part_pre <- as.data.frame(t(browse_prio_mutation[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
   #     if(nrow(subset.browse_prio_mutation.t_fisher_part_pre)>0) {
   #       colnames(subset.browse_prio_mutation.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_mutation.t_fisher_part_pre),"_",2))[1,])
   #       #paste0("p-value in ",str_sub(input$data_source_prio,end=-19))
   #       meta.p <- browse_prio_mutation_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
   #       meta.fdr <- p.adjust(meta.p)
   #       subset.browse_prio_mutation.t_meta_part <- data.frame("meta-analysis p-value"=meta.p,
   #                                                             "FDR of meta-analysis p-value"=meta.fdr)
   #       combined.p <- fisher.method(subset.browse_prio_mutation.t_fisher_part_pre,na.rm=TRUE)$p.value
   #       subset.browse_prio_mutation.t_fisher_part <- data.frame("combined p-value"=combined.p,
   #                                                               "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
   #                                                               subset.browse_prio_mutation.t_fisher_part_pre,stringsAsFactors = F )
   #       subset.browse_prio_mutation.t <- merge(subset.browse_prio_mutation.t_meta_part,subset.browse_prio_mutation.t_fisher_part,by="row.names",all=T)
   #       rownames(subset.browse_prio_mutation.t) <- subset.browse_prio_mutation.t[,1]
   #       subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[,-1]
   #       subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[order(subset.browse_prio_mutation.t[,"meta.analysis.p.value"]),]
   #       subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[,-which(apply(subset.browse_prio_mutation.t,2,function(x)all(is.na(x))))]
   #       
   #       #subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t[order(subset.browse_prio_mutation.t[,"combined.p.value"],subset.browse_prio_mutation.t[,names(apply(subset.browse_prio_mutation.t,2,function(x) length(na.omit(x)))[2])]),]
   #       brks <- quantile(subset.browse_prio_mutation.t, probs = seq(.05, .95, .05), na.rm = TRUE)
   #       clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}
   #     } else {
   #       subset.browse_prio_mutation.t <- subset.browse_prio_mutation.t_fisher_part_pre
   #     }
   #   })
   #   output$prio_mutation_download <- downloadHandler(
   #     filename = function() {
   #       if (input$prio_mutation_type == "Excel (CSV)") {
   #         "newfile.csv"
   #       } else if (input$prio_mutation_type == "Text (TSV)") {
   #         "newfile.txt"
   #       } else if (input$prio_mutation_type == "JSON") {
   #         "newfile.json"
   #       }
   #     },
   #     content = function(file) {
   #       if (input$prio_mutation_type == "Excel (CSV)") {
   #         fwrite (subset.browse_prio_mutation.t,row.names=T,file)
   #       } else if (input$prio_mutation_type == "Text (TSV)") {
   #         fwrite (subset.browse_prio_mutation.t,row.names=T,file)
   #         #write.table (selectedmutation_mutation(),file,quote=F, row.names = FALSE, sep="\t")
   #       } else if (input$prio_mutation_type == "JSON") {
   #         write(toJSON(subset.browse_prio_mutation.t),row.names=T,file)
   #       }
   #     }
   #   )
   #   output$prio_mutation_table <- DT::renderDataTable({
   #     datatable(subset.browse_prio_mutation.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>% 
   #       formatStyle(names(subset.browse_prio_mutation.t), backgroundColor = styleInterval(brks, clrs)) %>%
   #       formatSignif(columns = c(1:ncol(subset.browse_prio_mutation.t)), digits = 2,getOption("scipen"))
   #   })
   # })

  ###############################
  #### signatures prio mutational signature
  #################################

   #### meta-analysis ####
   # observeEvent(input$do_prio,{
   #   isolate({
   #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
   #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
   #     subset.browse_prio_mutation_signature_meta_p <- browse_prio_mutation_signature_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
   #     if (length(subset.browse_prio_mutation_signature_meta_p)==0) {
   #       #output$prio_mutation_signature_plots_meta <- renderPlotly({
   #       prio_mutation_signature_plots_meta_p <- plot_ly () %>%
   #         layout(title = "meta-anlaysis across selected cohorts",font=list(size=15,family="Aril")) %>%
   #         add_annotations(x=2,y=2,text="Warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
   #       #})
   #       prio_mutation_signature_tables_meta_data <- datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
   #     } else {
   #       fit <- density(na.omit(subset.browse_prio_mutation_signature_meta_p))
   #       vline <- function(x = 0, color = "red") {
   #         list(
   #           type = "line",
   #           y0 = 0,
   #           y1 = 1,
   #           yref = "paper",
   #           x0 = x,
   #           x1 = x,
   #           line = list(color = color)
   #         )}
   #       prio_mutation_signature_plots_meta_p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
   #         layout(title="meta-anlaysis across all cohorts",
   #                height = 480,
   #                xaxis = list(title="-log10(p.value)",
   #                             titlefont=list(size=20,color="black",family= "Aril black"),
   #                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
   #                yaxis=list(title="Density",
   #                           titlefont=list(size=20,color="black",family= "Aril black"),
   #                           showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
   #                shapes=list(vline(1.3))) %>%
   #         add_annotations(x = 1.3,
   #                         y = 1,
   #                         text = "p.value = 0.05",
   #                         xref = paste0("x", 1),  # add this
   #                         yref = paste0("y", 1),  # add this
   #                         showarrow = FALSE,
   #                         ax = 20,
   #                         ay = -40,
   #                         font=list(size=20,color="black",family= "Aril black"))
   #
   #
   #       prio_mutation_signature_meta_table <- data.frame("p.value"=subset.browse_prio_mutation_signature_meta_p,
   #                                                        "FDR"=p.adjust(subset.browse_prio_mutation_signature_meta_p))
   #       prio_mutation_signature_meta_table_sig <- prio_mutation_signature_meta_table# keep all mutation sigantures [prio_mutation_signature_meta_table$p.value<=0.05 & !is.na(prio_mutation_signature_meta_table$p.value),]
   #       prio_mutation_signature_meta_table_sig <- prio_mutation_signature_meta_table_sig[order(prio_mutation_signature_meta_table_sig$p.value),]
   #       shinyInput <- function(FUN, len, id, ...) {
   #         inputs <- character(len)
   #         for (i in seq_len(len)) {
   #           inputs[i] <- as.character(FUN(paste0(id, rownames(prio_mutation_signature_meta_table_sig)[i],"_",i), ...))
   #         }
   #         inputs
   #       }
   #       d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
   #         # Gene=rownames(d.nonna.t.p0.05),
   #         # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
   #         # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
   #         signif(prio_mutation_signature_meta_table_sig,2),
   #         View=shinyInput(actionButton,nrow(prio_mutation_signature_meta_table_sig),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_mutation_signature_meta\",  this.id)'),
   #         stringsAsFactors = FALSE#,
   #         #row.names = 1:nrow(d.nonna.t.p0.05)
   #       ))
   #       prio_mutation_signature_tables_meta_data <- d.nonna.t.p0.05.test$data
   #
   #     }
   #     output$prio_mutation_signature_plots_meta <- renderPlotly({
   #       prio_mutation_signature_plots_meta_p
   #     })
   #     output$prio_mutation_signature_tables_meta <- DT::renderDataTable(
   #       prio_mutation_signature_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none'
   #     )
   #   })
   # })

   observeEvent(input$do_prio,{
     isolate({
       withProgress(message = 'Preparing data', value = 0, {
       input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
       input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
       browse_prio_mutation_signature_meta_selected <- browse_prio_mutation_signature_meta[
         browse_prio_mutation_signature_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
       if (nrow(browse_prio_mutation_signature_meta_selected)==0 |
           all(is.na(browse_prio_mutation_signature_meta_selected$TE)) |
           all(is.na(browse_prio_mutation_signature_meta_selected$seTE)) ) {
         prio_mutation_signature_tables_meta_data <- datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
         # meta.p <- NA
         # meta.fdr <- NA
         fit.t <- NULL
       } else {
         incProgress(20, detail = "Doing part %")
         if (input$sample_prio=="custom selection") {
          splitted <- browse_prio_mutation_signature_meta_selected %>% split(.$genes)
          fit <- lapply(splitted,
                       function(x) {
                         if (all(is.na(x$TE)) | all(is.na(x$seTE)) | nrow(x)==0 | is.null(x)) {
                           c(NA,NA,NA)
                         } else {
                           a <- metagen(TE,seTE,data=x,sm="OR")
                           #c(a$pval.fixed,a$pval.random,a$pval.Q)})
                           c(a$pval.random,a$pval.fixed,a$pval.Q)
                         }
                       })
          fit.t <- t(as.data.frame(fit))
          fit.t.fdr.random <- p.adjust(fit.t[,1],"fdr")
          fit.t.fdr.fixed <- p.adjust(fit.t[,2],"fdr")
          # fit.t <- data.frame(pval.fixed=fit.t[,1],
          #                   fdr.fixed=fit.t.fdr,
          #                   pval.random=fit.t[,2],
          #                   pval.Q=fit.t[,3])
          # fit.t <- fit.t[order(fit.t$pval.fixed),]
          fit.t <- data.frame(Pval.random=fit.t[,1],
                              FDR.random=fit.t.fdr.random,
                              Pval.fixed=fit.t[,2],
                              FDR.fixed=fit.t.fdr.fixed)
          fit.t <- fit.t[order(fit.t$Pval.random),]
         #c("pval.fixed","pval.random","pval.Q")
         } else {
           fit.t <- meta_analysis_prio_precalculated_mutation_signature[[input$sample_prio]][,1:4]
           #fit.t <- fit.t[order(fit.t$Pval.random),]
         }
         shinyInput <- function(FUN, len, id, ...) {
           inputs <- character(len)
           for (i in seq_len(len)) {
             inputs[i] <- as.character(FUN(paste0(id, rownames(fit.t)[i],"_",i), ...))
           }
           inputs
         }
         d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
           # Gene=rownames(d.nonna.t.p0.05),
           # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
           # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
           signif(fit.t,2),
           View=shinyInput(actionButton,nrow(fit.t),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_mutation_signature_meta\",  this.id)'),
           stringsAsFactors = FALSE#,
           #row.names = 1:nrow(d.nonna.t.p0.05)
         ))
         prio_mutation_signature_tables_meta_data <- d.nonna.t.p0.05.test$data
       }
         subset.browse_prio_mutation_signature.t_fisher_part_pre <- as.data.frame(t(browse_prio_mutation_signature[rownames(browse_prio_mutation_signature) %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
         incProgress(80, detail = "Doing part %")
         if(nrow(subset.browse_prio_mutation_signature.t_fisher_part_pre)>0) {
           if (input$sample_prio=="custom selection") {
            colnames(subset.browse_prio_mutation_signature.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_mutation_signature.t_fisher_part_pre),"_",2))[1,])
            subset.browse_prio_mutation_signature.t_meta_part <- fit.t
            combined.p <- fisher.method(subset.browse_prio_mutation_signature.t_fisher_part_pre,na.rm=TRUE)$p.value
            subset.browse_prio_mutation_signature.t_fisher_part <- data.frame("combined p-value"=combined.p,
                                                                             "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
                                                                             subset.browse_prio_mutation_signature.t_fisher_part_pre,stringsAsFactors = F )
            subset.browse_prio_mutation_signature.t <- merge(subset.browse_prio_mutation_signature.t_meta_part,subset.browse_prio_mutation_signature.t_fisher_part,by="row.names",all=T)
            rownames(subset.browse_prio_mutation_signature.t) <- subset.browse_prio_mutation_signature.t[,1]
            subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[,-1]
            subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[order(subset.browse_prio_mutation_signature.t[,1]),]
           # subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[,-which(apply(subset.browse_prio_mutation_signature.t,2,function(x)all(is.na(x))))]
            subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[,apply(subset.browse_prio_mutation_signature.t,2,function(x) !all(is.na(x)))]
           } else {
             subset.browse_prio_mutation_signature.t <- meta_analysis_prio_precalculated_mutation_signature[[input$sample_prio]]
             
           }
           brks <- quantile(subset.browse_prio_mutation_signature.t, probs = seq(.05, .95, .05), na.rm = TRUE)
           clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}

         } else {
           subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t_fisher_part_pre

         }
       #}
       incProgress(100, detail = "Doing part %")



       output$prio_mutation_signature_tables_meta <- DT::renderDataTable(
         prio_mutation_signature_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none',caption = 'Pval.random: p-value for test of overall treatment effect (random effects model). FDR.random: adjusted p-value of random effects model using false discovery rate method. Pval.fixed: p-value for test of overall treatment effect (fixed effect model). FDR.fixed: adjusted p-value of fixed effect model using false discovery rate method.'
       )

       output$prio_mutation_signature_download <- downloadHandler(
         filename = function() {
           if (input$prio_mutation_signature_type == "Excel (CSV)") {
             "newfile.csv"
           } else if (input$prio_mutation_signature_type == "Text (TSV)") {
             "newfile.txt"
           } else if (input$prio_mutation_signature_type == "JSON") {
             "newfile.json"
           }
         },
         content = function(file) {
           if (input$prio_mutation_signature_type == "Excel (CSV)") {
             fwrite (subset.browse_prio_mutation_signature.t,row.names=T,file)
           } else if (input$prio_mutation_signature_type == "Text (TSV)") {
             fwrite (subset.browse_prio_mutation_signature.t,row.names=T,file)
             #write.table (selectedmutation_signature_mutation_signature(),file,quote=F, row.names = FALSE, sep="\t")
           } else if (input$prio_mutation_signature_type == "JSON") {
             write(toJSON(subset.browse_prio_mutation_signature.t),row.names=T,file)
           }
         }
       )
       output$prio_mutation_signature_table <- DT::renderDataTable({
         datatable(subset.browse_prio_mutation_signature.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
           formatStyle(names(subset.browse_prio_mutation_signature.t), backgroundColor = styleInterval(brks, clrs)) %>%
           formatSignif(columns = c(1:ncol(subset.browse_prio_mutation_signature.t)), digits = 2,getOption("scipen"))
       })
       })
     })
   })

   observeEvent(input$select_button_prio_mutation_signature_meta,{
     s <- as.numeric(strsplit(input$select_button_prio_mutation_signature_meta, "_")[[1]][3])
     Viewname <- as.character(strsplit(input$select_button_prio_mutation_signature_meta, "_")[[1]][2])
     output$prio_mutation_signature_tables_meta_modals <- renderUI({
       tagList(
         bsModal(paste0("prio_mutation_signature_tables_meta_bsModal",s), Viewname, "select_button_prio_mutation_signature_meta", size = "large",
                 plotOutput("browse_prio_mutation_signature_meta_bsModal_plot")
         )
       )
     })
     toggleModal(session,paste0("prio_mutation_signature_tables_meta_bsModal",s), toggle = Viewname)
     ##Reset the select_button
     session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_mutation_signature_meta")
   })
   observeEvent(input$select_button_prio_mutation_signature_meta,{
     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
     browse_prio_mutation_signature_meta_selected <- browse_prio_mutation_signature_meta[
       browse_prio_mutation_signature_meta$genes==as.character(strsplit(input$select_button_prio_mutation_signature_meta, "_")[[1]][2]) &
         browse_prio_mutation_signature_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
     output$browse_prio_mutation_signature_meta_bsModal_plot <- renderPlot({
       # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
       #                 data=browse_prio_mutation_signature_meta_selected,
       #                 sm="MD",
       #                 label.e = "response",label.c = "nonresponse",
       #                 studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
       #        test.overall = TRUE,ff.study.label = "bold")
       if (nrow(browse_prio_mutation_signature_meta_selected)==0 |
           is.null(browse_prio_mutation_signature_meta_selected) |
           all(is.na(browse_prio_mutation_signature_meta_selected$TE)) |
           all(is.na(browse_prio_mutation_signature_meta_selected$seTE))) {
         plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_prio,end=-19),collapse = ","),cex.main=1.5,bty="l")
         abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
         text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
       } else {
         m1 <- metagen(TE,seTE,data=browse_prio_mutation_signature_meta_selected,
                       sm="OR",
                       #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
                       studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1]))
         
         m1$upper[exp(m1$upper)>1e+50] <- 1000
         m1$upper.fixed[exp(m1$upper.fixed)>1e+50] <- 1000
         m1$upper.random[exp(m1$upper.random)>1e+50] <- 1000
         m1$lower.fixed[exp(m1$lower.fixed)>1e+50] <- 1000
         m1$lower.random[exp(m1$lower.random)>1e+50] <- 1000
         
         forest(m1,layout="RevMan5",    
                hetstat=TRUE,
                plotwidth="6inch",test.overall = TRUE,
                xlim=c(0.1,10),backtransfer=T,
                label.test.overall.random="random effects model: ",
                label.test.overall.fixed="fixed effects model: ",
                fs.test.overall = 15,
                scientific.pval = TRUE,ff.study.label = "bold")
         # forest(metagen(TE,seTE,data=browse_prio_mutation_signature_meta_selected,
         #                sm="OR",
         #                #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
         #                studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1])),
         #        plotwidth="6inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
       }
       
     })

   })
   #### each dataset####
   output$prio_mutation_signature_plots <- renderUI({
     input$do_prio
     isolate({
       input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
       input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

       #      browse_prio_mutation_signature <- browse_prio_mutation_signature_function()
       subset.browse_prio_mutation_signature <- browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
       #selected.database_prio_mutation_signature <- selecteddatabase_prio_mutation_signature()
       if (nrow(subset.browse_prio_mutation_signature)==0 ) {
         tagList(plotlyOutput("prio_mutation_signature_plots0",height="auto"),dataTableOutput("prio_mutation_signature_tables0"))
       } else {
         lapply(1:length(input$data_source_prio),function(i) {
           column(width=6,wellPanel(plotlyOutput(paste0("prio_mutation_signature_plots",i),height="auto"),
                                    h4("All mutation signatures:",style="margin-top:20px;margin-bottom:10px;font-size:20px;font-weight:bold;color:#DCB239"),
                                    div(DT::dataTableOutput(paste0("prio_mutation_signature_tables",i)),style="margin-top:30px;margin-bottom:10px"),style="height:1300px"),style="padding-bottom:20px")
           #column(width=6,wellPanel(DT::dataTableOutput(paste0("prio_mutation_signature_tables",i))),style="padding-bottom:20px")
         })
       }
     })
   })

   observeEvent(input$do_prio,{
     input$do_prio

     isolate({
       #        browse_prio_mutation_signature <- browse_prio_mutation_signature_function()
       input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
       input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

       #      browse_prio_mutation_signature <- browse_prio_mutation_signature_function()
       subset.browse_prio_mutation_signature <- browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]

       if (nrow(subset.browse_prio_mutation_signature)==0) {
         output$prio_mutation_signature_plots0 <- renderPlotly({
           plot_ly () %>%
             layout(title = paste(str_sub(input$data_source_prio,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="Warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
         })

         output$prio_mutation_signature_tables0 <- DT::renderDataTable({
           datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
         })

       } else {
         # splitted_list <- selected.database_prio_mutation_signature %>% as.matrix %>% as.data.frame %>% split(.$data_source)
         # splitted_list <- splitted_list[input$data_source_prio]
         isolate({
           for (k in 1:length(input$data_source_prio)) {
             local({
               my_i <- k
               browse_prio_mutation_signature_plot_name <- paste0("prio_mutation_signature_plots", my_i)
               browse_prio_mutation_signature_table_name <- paste0("prio_mutation_signature_tables", my_i)
               #             d<-splitted_list[[my_i]]

               output[[browse_prio_mutation_signature_plot_name]] <- renderPlotly({
                 isolate({
                   withProgress(message = 'Making plot', value = 0, {
                     if (strsplit(rownames(subset.browse_prio_mutation_signature)[my_i],"\\.")[[1]][1]=="NA") {
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>%
                         layout(title = str_sub(input$data_source_prio,end=-19)[my_i],font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="Warning 2: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     } else {
                       if (length(na.omit(as.numeric(subset.browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))==0) {
                         incProgress(100, detail = "Doing part %")
                         plot_ly () %>%
                           layout(title = unique(str_sub(input$data_source_prio,end=-19)[my_i]),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="Warning 3: no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                       } else {
                         vline <- function(x = 0, color = "red") {
                           list(
                             type = "line",
                             y0 = 0,
                             y1 = 1,
                             yref = "paper",
                             x0 = x,
                             x1 = x,
                             line = list(color = color)
                           )
                         }
                         d <- subset.browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                         d.nonna <- d[,!is.na(d[1,])]
                         d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))))
                         rownames(d.nonna.t) <- colnames(d.nonna)
                         colnames(d.nonna.t) <- c(paste0("p-value in ",rownames(d.nonna)),paste0("-log10(p-value) in ",rownames(d.nonna)))
                         fit <- density(d.nonna.t[,2])
                         # p<-plot_ly(x=d.nonna.t[,2],
                         #            #y=as.numeric(d[-1]),
                         #            type="histogram",#type="box",
                         #            #histnorm="probability",
                         #            name="Histogram"
                         # ) %>%
                         #   add_trace(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",yaxis="y2",name="Density") %>%
                         p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
                           layout(title=str_sub(input$data_source_prio[my_i],end=-19),
                                  height = 480,
                                  xaxis = list(title="-log10(p.value)",
                                               titlefont=list(size=20,color="black",family= "Aril black"),
                                               showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                  yaxis=list(title="Density",
                                             titlefont=list(size=20,color="black",family= "Aril black"),
                                             showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                  shapes=list(vline(1.3))) %>%
                           add_annotations(x = 1.3,
                                           y = 1,
                                           text = "p.value = 0.05",
                                           xref = paste0("x", 1),  # add this
                                           yref = paste0("y", 1),  # add this
                                           showarrow = FALSE,
                                           ax = 20,
                                           ay = -40,
                                           font=list(size=20,color="black",family= "Aril black"))
                         incProgress(100, detail = "Doing part %")
                         p
                       }
                     }
                   })
                 })

               })

               #output[[browse_prio_mutation_signature_table_name]] <- DT::renderDataTable({
               isolate({
                 if (rownames(subset.browse_prio_mutation_signature)[my_i]=="NA") {
                   prio_mutation_signature_tables_data <-
                     datatable(data.frame("NA"="Warning 2: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (length(na.omit(as.numeric(subset.browse_prio_mutation_signature[str_sub(input$data_source_prio,end=-19)[my_i],])))==0) {
                     prio_mutation_signature_tables_data <-
                       datatable(data.frame("NA"="Warning 3: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {

                     d <- subset.browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                     d.nonna <- d[,!is.na(d[1,])]
                     if (ncol(d.nonna)==0) {
                       prio_mutation_signature_tables_data <-
                         datatable(data.frame("NA"="no clinical data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     } else {
                       d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))),p.adjust(as.numeric(t(d.nonna)),"fdr"))
                       rownames(d.nonna.t) <- colnames(d.nonna)
                       colnames(d.nonna.t) <- c(paste0("p-value in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("-log10(p-value) in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("FDR in ",str_split(rownames(d.nonna),"_",2)[[1]][1]))
                       colnames(d.nonna.t) <- c("p-value","-log10(p-value)","FDR")
                       d.nonna.t.p0.05 <- d.nonna.t# keep all mutation signatures[d.nonna.t[,1]<=0.05,]
                       d.nonna.t.p0.05 <- d.nonna.t.p0.05[order(d.nonna.t.p0.05[,1]),]

                       shinyInput <- function(FUN, len, id, ...) {
                         inputs <- character(len)
                         for (i in seq_len(len)) {
                           inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
                         }
                         inputs
                       }

                       d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
                         # Gene=rownames(d.nonna.t.p0.05),
                         # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
                         # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
                         signif(d.nonna.t.p0.05[,c(1,3)],2),
                         View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_"),label="View",onclick='Shiny.onInputChange(\"select_button_prio_mutation_signature\",  this.id)'),
                         stringsAsFactors = FALSE#,
                         #row.names = 1:nrow(d.nonna.t.p0.05)
                       ))

                       prio_mutation_signature_tables_data <- d.nonna.t.p0.05.test$data

                       # datatable(d.nonna.t.p0.05) %>%
                       #   formatSignif(columns = c(1:ncol(d.nonna.t.p0.05)), digits = 2,getOption("scipen"))
                       #datatable(d.nonna.t.p0.05,options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                     }
                   }
                 }
               })
               output[[browse_prio_mutation_signature_table_name]] <- DT::renderDataTable(
                 prio_mutation_signature_tables_data,server = FALSE, escape = FALSE, selection = 'none'
               )

               #})

             })
           }
         })
       }
     })

   })

   observeEvent(input$select_button_prio_mutation_signature,{
     s <- as.numeric(strsplit(input$select_button_prio_mutation_signature, "_")[[1]][3])
     output$prio_mutation_signature_modals <- renderUI({
       tagList(
         bsModal(paste('prio_mutation_signature_model', s ,sep=''), "View", "select_button_prio_mutation_signature", size = "large",

                 plotlyOutput("browse_prio_mutation_signature_bsModal_plot")
                 #textOutput("browse_prio_mutation_signature_bsModal_plot")
         )
       )
     })
     toggleModal(session,paste('prio_mutation_signature_model', s ,sep=''), toggle = "View")
     ##Reset the select_button
     session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_mutation_signature")
   })

   observeEvent(input$select_button_prio_mutation_signature,{

     data_source1 <- as.character(strsplit(input$select_button_prio_mutation_signature, "_")[[1]][1])
     gene <- as.character(strsplit(input$select_button_prio_mutation_signature, "_")[[1]][2])
     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

     subset.browse_prio_mutation_signature_plot <- browse_prio_mutation_signature_plot[paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),gene]
     d <- rbind(data.frame(PFS.status="nonresponse",Gene=as.numeric(strsplit(strsplit(as.character(subset.browse_prio_mutation_signature_plot),"_")[[1]][2],",")[[1]])),
                data.frame(PFS.status="response",Gene=as.numeric(strsplit(strsplit(as.character(subset.browse_prio_mutation_signature_plot),"_")[[1]][1],",")[[1]])))
     p<-plot_ly(data=d,x=~PFS.status,
                y=~Gene,
                type="violin",#type="box",
                points="all",#boxpoints="all",
                jitter=0.3,
                pointpos=-1.8,
                #mode="markders",
                color= ~PFS.status,
                colors=c("#7CDBD5","#DCB239"),
                #hoveron="points+boxes",
                legendgroup=~PFS.status,
                showlegend=TRUE
     ) %>%
       layout(title = unique(data_source1),titlefont=list(size=20,color="black",family= "Aril"),
              height = 480,
              xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
              yaxis=list(title=gene,
                         titlefont=list(size=20,color="black",family= "Aril"),
                         showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
              )) %>%
       add_annotations(x=0.5,
                       y=d[which.max(d[,"Gene"]),"Gene"],

                       text=paste0("p-value = ",signif(as.numeric(browse_prio_mutation_signature[paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),gene]),2)),
                       xref = paste0("x", 1),  # add this
                       yref = paste0("y", 1),  # add this
                       showarrow = FALSE,
                       ax = 20,
                       ay = -40,
                       font=list(size=20,color="black",family= "Aril black"))
     output$browse_prio_mutation_signature_bsModal_plot <- renderPlotly(
       p
     )
     # output$browse_prio_mutation_signature_bsModal_plot <- renderText(
     #   # input$select_button_prio_mutation_signature
     #   # nrow(browse_prio_mutation_signature_plot)
     #   # paste0(paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),"_____",gene)
     #   paste0(data_source1,";",gene,";",paste(input_treatment_prio,collapse = "_"),";",
     #          paste(input_description_prio,collapse = "_"),";",subset.browse_prio_mutation_signature_plot,";",
     #          browse_prio_mutation_signature_plot[1,1],";",rownames(browse_prio_mutation_signature_plot)[1],";",
     #          colnames(browse_prio_mutation_signature_plot)[1])
     # )


   })

   #### download ####
   # observeEvent(input$do_prio,{
   #   input$do_prio
   #   isolate({
   #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
   #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
   #
   #     #        browse_prio_mutation_signature <- browse_prio_mutation_signature_function()
   #     #subset.browse_prio_mutation_signature.t <- unfactor(as.data.frame(t(browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))
   #     subset.browse_prio_mutation_signature.t_fisher_part_pre <- as.data.frame(t(browse_prio_mutation_signature[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
   #
   #     #colnames(subset.browse_prio_mutation_signature.t) <- paste0("p-value in ",sapply(str_split(colnames(subset.browse_prio_mutation_signature.t),"_",2),function(x) x[1]))
   #     if(nrow(subset.browse_prio_mutation_signature.t_fisher_part_pre)>0) {
   #       colnames(subset.browse_prio_mutation_signature.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_mutation_signature.t_fisher_part_pre),"_",2))[1,])
   #       #paste0("p-value in ",str_sub(input$data_source_prio,end=-19))
   #       meta.p <- browse_prio_mutation_signature_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
   #       meta.fdr <- p.adjust(meta.p)
   #       subset.browse_prio_mutation_signature.t_meta_part <- data.frame("meta-analysis p-value"=meta.p,
   #                                                                       "FDR of meta-analysis p-value"=meta.fdr)
   #       combined.p <- fisher.method(subset.browse_prio_mutation_signature.t_fisher_part_pre,na.rm=TRUE)$p.value
   #       subset.browse_prio_mutation_signature.t_fisher_part <- data.frame("combined p-value"=combined.p,
   #                                                                         "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
   #                                                                         subset.browse_prio_mutation_signature.t_fisher_part_pre,stringsAsFactors = F )
   #       subset.browse_prio_mutation_signature.t <- merge(subset.browse_prio_mutation_signature.t_meta_part,subset.browse_prio_mutation_signature.t_fisher_part,by="row.names",all=T)
   #       rownames(subset.browse_prio_mutation_signature.t) <- subset.browse_prio_mutation_signature.t[,1]
   #       subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[,-1]
   #       subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[order(subset.browse_prio_mutation_signature.t[,"meta.analysis.p.value"]),]
   #       subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[,-which(apply(subset.browse_prio_mutation_signature.t,2,function(x)all(is.na(x))))]
   #
   #       #subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t[order(subset.browse_prio_mutation_signature.t[,"combined.p.value"],subset.browse_prio_mutation_signature.t[,names(apply(subset.browse_prio_mutation_signature.t,2,function(x) length(na.omit(x)))[2])]),]
   #       brks <- quantile(subset.browse_prio_mutation_signature.t, probs = seq(.05, .95, .05), na.rm = TRUE)
   #       clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}
   #     } else {
   #       subset.browse_prio_mutation_signature.t <- subset.browse_prio_mutation_signature.t_fisher_part_pre
   #     }
   #   })
   #   output$prio_mutation_signature_download <- downloadHandler(
   #     filename = function() {
   #       if (input$prio_mutation_signature_type == "Excel (CSV)") {
   #         "newfile.csv"
   #       } else if (input$prio_mutation_signature_type == "Text (TSV)") {
   #         "newfile.txt"
   #       } else if (input$prio_mutation_signature_type == "JSON") {
   #         "newfile.json"
   #       }
   #     },
   #     content = function(file) {
   #       if (input$prio_mutation_signature_type == "Excel (CSV)") {
   #         fwrite (subset.browse_prio_mutation_signature.t,row.names=T,file)
   #       } else if (input$prio_mutation_signature_type == "Text (TSV)") {
   #         fwrite (subset.browse_prio_mutation_signature.t,row.names=T,file)
   #         #write.table (selectedmutation_signature_mutation_signature(),file,quote=F, row.names = FALSE, sep="\t")
   #       } else if (input$prio_mutation_signature_type == "JSON") {
   #         write(toJSON(subset.browse_prio_mutation_signature.t),row.names=T,file)
   #       }
   #     }
   #   )
   #   output$prio_mutation_signature_table <- DT::renderDataTable({
   #     datatable(subset.browse_prio_mutation_signature.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
   #       formatStyle(names(subset.browse_prio_mutation_signature.t), backgroundColor = styleInterval(brks, clrs)) %>%
   #       formatSignif(columns = c(1:ncol(subset.browse_prio_mutation_signature.t)), digits = 2,getOption("scipen"))
   #   })
   # })

  ##########################################
  #### signature prio expression
  ##########################################

  #### meta-analysis #### ### for plot and table
  # observeEvent(input$do_prio,{
  #   isolate({
  #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
  #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
  #     subset.browse_prio_expression_meta_p <- browse_prio_expression_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
  #     if (length(subset.browse_prio_expression_meta_p)==0) {
  #       #output$prio_expression_plots_meta <- renderPlotly({
  #         prio_expression_plots_meta_p <- plot_ly () %>%
  #           layout(title = "meta-anlaysis across selected cohorts",font=list(size=15,family="Aril")) %>%
  #           add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
  #       #})
  #         prio_expression_tables_meta_data <- datatable(data.frame("NA"="no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
  #     } else {
  #       fit <- density(na.omit(subset.browse_prio_expression_meta_p))
  #       vline <- function(x = 0, color = "red") {
  #         list(
  #           type = "line",
  #           y0 = 0,
  #           y1 = 1,
  #           yref = "paper",
  #           x0 = x,
  #           x1 = x,
  #           line = list(color = color)
  #         )}
  #       prio_expression_plots_meta_p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
  #         layout(title="meta-anlaysis across selected cohorts",
  #                height = 480,
  #                xaxis = list(title="-log10(p.value)",
  #                             titlefont=list(size=20,color="black",family= "Aril black"),
  #                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
  #                yaxis=list(title="Density",
  #                           titlefont=list(size=20,color="black",family= "Aril black"),
  #                           showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
  #                shapes=list(vline(1.3))) %>%
  #         add_annotations(x = 1.3,
  #                         y = 1,
  #                         text = "p.value = 0.05",
  #                         xref = paste0("x", 1),  # add this
  #                         yref = paste0("y", 1),  # add this
  #                         showarrow = FALSE,
  #                         ax = 20,
  #                         ay = -40,
  #                         font=list(size=20,color="black",family= "Aril black"))
  #
  #
  #       prio_expression_meta_table <- data.frame("p.value"=subset.browse_prio_expression_meta_p,
  #                                                "FDR"=p.adjust(subset.browse_prio_expression_meta_p))
  #       prio_expression_meta_table_sig <- prio_expression_meta_table[prio_expression_meta_table$p.value<=0.05 & !is.na(prio_expression_meta_table$p.value),]
  #       prio_expression_meta_table_sig <- prio_expression_meta_table_sig[order(prio_expression_meta_table_sig$p.value),]
  #       shinyInput <- function(FUN, len, id, ...) {
  #         inputs <- character(len)
  #         for (i in seq_len(len)) {
  #           inputs[i] <- as.character(FUN(paste0(id, rownames(prio_expression_meta_table_sig)[i],"_",i), ...))
  #         }
  #         inputs
  #       }
  #       d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
  #         # Gene=rownames(d.nonna.t.p0.05),
  #         # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
  #         # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
  #         signif(prio_expression_meta_table_sig,2),
  #         View=shinyInput(actionButton,nrow(prio_expression_meta_table_sig),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_expression_meta\",  this.id)'),
  #         stringsAsFactors = FALSE#,
  #         #row.names = 1:nrow(d.nonna.t.p0.05)
  #       ))
  #       prio_expression_tables_meta_data <- d.nonna.t.p0.05.test$data
  #
  #     }
  #     output$prio_expression_plots_meta <- renderPlotly({
  #       prio_expression_plots_meta_p
  #     })
  #     output$prio_expression_tables_meta <- DT::renderDataTable(
  #       prio_expression_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none'
  #     )
  #   })
  # })

  #### meta-analysis ### for table only
  observeEvent(input$do_prio,{
    isolate({
      withProgress(message = 'Preparing data', value = 0, {
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
      browse_prio_expression_meta_selected <- browse_prio_expression_meta[
        browse_prio_expression_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
      if (nrow(browse_prio_expression_meta_selected)==0 |
          all(is.na(browse_prio_expression_meta_selected$TE)) |
          all(is.na(browse_prio_expression_meta_selected$seTE)) ) {
        prio_expression_tables_meta_data <- datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        # meta.p <- NA
        # meta.fdr <- NA
        fit.t <- NULL
      } else {
        incProgress(20, detail = "Doing part %")
        if (input$sample_prio=="custom selection") {
          splitted <- browse_prio_expression_meta_selected %>% split(.$genes)
          fit <- lapply(splitted,
                        function(x) {
                          if (all(is.na(x$TE)) | all(is.na(x$seTE)) | nrow(x)==0 | is.null(x)) {
                            c(NA,NA,NA)
                          } else {
                            a <- metagen(TE,seTE,data=x,sm="OR")
                            #c(a$pval.fixed,a$pval.random,a$pval.Q)})
                            c(a$pval.random,a$pval.fixed,a$pval.Q)
                          }
                        })
          
          fit.t <- t(as.data.frame(fit))
          fit.t.fdr.random <- p.adjust(fit.t[,1],"fdr")
          fit.t.fdr.fixed <- p.adjust(fit.t[,2],"fdr")
          # fit.t <- data.frame(pval.fixed=fit.t[,1],
          #                   fdr.fixed=fit.t.fdr,
          #                   pval.random=fit.t[,2],
          #                   pval.Q=fit.t[,3])
          # fit.t <- fit.t[order(fit.t$pval.fixed),]
          fit.t <- data.frame(Pval.random=fit.t[,1],
                              FDR.random=fit.t.fdr.random,
                              Pval.fixed=fit.t[,2],
                              FDR.fixed=fit.t.fdr.fixed)
          fit.t <- fit.t[order(fit.t$Pval.random),]
        } else {
          fit.t <- meta_analysis_prio_precalculated_expression[[input$sample_prio]][,1:4]
          # fit.t <- fit.t[order(fit.t$pval.fixed),]
        }
        #c("pval.fixed","pval.random","pval.Q")

        shinyInput <- function(FUN, len, id, ...) {
          inputs <- character(len)
          for (i in seq_len(len)) {
            inputs[i] <- as.character(FUN(paste0(id, rownames(fit.t)[i],"_",i), ...))
          }
          inputs
        }
        d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
          # Gene=rownames(d.nonna.t.p0.05),
          # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
          # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
          signif(fit.t,2),
          View=shinyInput(actionButton,nrow(fit.t),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_expression_meta\",  this.id)'),
          stringsAsFactors = FALSE#,
          #row.names = 1:nrow(d.nonna.t.p0.05)
        ))
        prio_expression_tables_meta_data <- d.nonna.t.p0.05.test$data
      }
        subset.browse_prio_expression.t_fisher_part_pre <- as.data.frame(t(browse_prio_expression[rownames(browse_prio_expression) %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))

        if(nrow(subset.browse_prio_expression.t_fisher_part_pre)>0) {
          if (input$sample_prio=="custom selection") {
            colnames(subset.browse_prio_expression.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_expression.t_fisher_part_pre),"_",2))[1,])
            subset.browse_prio_expression.t_meta_part <- fit.t
            combined.p <- fisher.method(subset.browse_prio_expression.t_fisher_part_pre,na.rm=TRUE)$p.value
            subset.browse_prio_expression.t_fisher_part <- data.frame("combined p-value"=combined.p,
                                                                    "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
                                                                    subset.browse_prio_expression.t_fisher_part_pre,stringsAsFactors = F )
            subset.browse_prio_expression.t <- merge(subset.browse_prio_expression.t_meta_part,subset.browse_prio_expression.t_fisher_part,by="row.names",all=T)
            rownames(subset.browse_prio_expression.t) <- subset.browse_prio_expression.t[,1]
            subset.browse_prio_expression.t <- subset.browse_prio_expression.t[,-1]
            subset.browse_prio_expression.t <- subset.browse_prio_expression.t[order(subset.browse_prio_expression.t[,1]),]
          # subset.browse_prio_expression.t <- subset.browse_prio_expression.t[,-which(apply(subset.browse_prio_expression.t,2,function(x)all(is.na(x))))]
            subset.browse_prio_expression.t <- subset.browse_prio_expression.t[,apply(subset.browse_prio_expression.t,2,function(x) !all(is.na(x)))]
          } else {
            subset.browse_prio_expression.t <- meta_analysis_prio_precalculated_expression[[input$sample_prio]]
            
          }
          brks <- quantile(subset.browse_prio_expression.t, probs = seq(.05, .95, .05), na.rm = TRUE)
          clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}

        } else {
          subset.browse_prio_expression.t <- subset.browse_prio_expression.t_fisher_part_pre

        }
      #}
      incProgress(100, detail = "Doing part %")



      output$prio_expression_tables_meta <- DT::renderDataTable(
        prio_expression_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none',caption = 'Pval.random: p-value for test of overall treatment effect (random effects model). FDR.random: adjusted p-value of random effects model using false discovery rate method. Pval.fixed: p-value for test of overall treatment effect (fixed effect model). FDR.fixed: adjusted p-value of fixed effect model using false discovery rate method.'
      )

          output$prio_expression_download <- downloadHandler(
            filename = function() {
              if (input$prio_expression_type == "Excel (CSV)") {
                "newfile.csv"
              } else if (input$prio_expression_type == "Text (TSV)") {
                "newfile.txt"
              } else if (input$prio_expression_type == "JSON") {
                "newfile.json"
              }
            },
            content = function(file) {
              if (input$prio_expression_type == "Excel (CSV)") {
                fwrite (subset.browse_prio_expression.t,row.names=T,file)
              } else if (input$prio_expression_type == "Text (TSV)") {
                fwrite (subset.browse_prio_expression.t,row.names=T,file)
                #write.table (selectedexpression_expression(),file,quote=F, row.names = FALSE, sep="\t")
              } else if (input$prio_expression_type == "JSON") {
                write(toJSON(subset.browse_prio_expression.t),row.names=T,file)
              }
            }
          )
          output$prio_expression_table <- DT::renderDataTable({
            datatable(subset.browse_prio_expression.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
              formatStyle(names(subset.browse_prio_expression.t), backgroundColor = styleInterval(brks, clrs)) %>%
              formatSignif(columns = c(1:ncol(subset.browse_prio_expression.t)), digits = 2,getOption("scipen"))
          })
      })
     })
  })

  observeEvent(input$select_button_prio_expression_meta,{
    s <- as.numeric(strsplit(input$select_button_prio_expression_meta, "_")[[1]][3])
    Viewname <- as.character(strsplit(input$select_button_prio_expression_meta, "_")[[1]][2])
    output$prio_expression_tables_meta_modals <- renderUI({
      tagList(
        bsModal(paste0("prio_expression_tables_meta_bsModal",s), Viewname, "select_button_prio_expression_meta", size = "large",
                plotOutput("browse_prio_expression_meta_bsModal_plot")
        )
      )
    })
    toggleModal(session,paste0("prio_expression_tables_meta_bsModal",s), toggle = Viewname)
    ##Reset the select_button
    session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_expression_meta")
  })

  observeEvent(input$select_button_prio_expression_meta,{
    input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
    input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
    browse_prio_expression_meta_selected <- browse_prio_expression_meta[
      browse_prio_expression_meta$genes==as.character(strsplit(input$select_button_prio_expression_meta, "_")[[1]][2]) &
        browse_prio_expression_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
    output$browse_prio_expression_meta_bsModal_plot <- renderPlot({
      # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
      #                 data=browse_prio_expression_meta_selected,
      #                 sm="MD",
      #                 label.e = "response",label.c = "nonresponse",
      #                 studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
             # test.overall = TRUE,ff.study.label = "bold")
      if (nrow(browse_prio_expression_meta_selected)==0 |
          is.null(browse_prio_expression_meta_selected) |
          all(is.na(browse_prio_expression_meta_selected$TE)) |
          all(is.na(browse_prio_expression_meta_selected$seTE))) {
        plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_prio,end=-19),collapse = ","),cex.main=1.5,bty="l")
        abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
        text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
      } else {
        m1 <- metagen(TE,seTE,data=browse_prio_expression_meta_selected,
                      sm="OR",
                      #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
                      studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1]))
        
        m1$upper[exp(m1$upper)>1e+50] <- 1000
        m1$upper.fixed[exp(m1$upper.fixed)>1e+50] <- 1000
        m1$upper.random[exp(m1$upper.random)>1e+50] <- 1000
        m1$lower.fixed[exp(m1$lower.fixed)>1e+50] <- 1000
        m1$lower.random[exp(m1$lower.random)>1e+50] <- 1000
        
        forest(m1,layout="RevMan5",    
               hetstat=TRUE,
               plotwidth="6inch",test.overall = TRUE,
               xlim=c(0.1,10),backtransfer=T,
               label.test.overall.random="random effects model: ",
               label.test.overall.fixed="fixed effects model: ",
               fs.test.overall = 15,
               scientific.pval = TRUE,ff.study.label = "bold")
        # forest(metagen(TE,seTE,data=browse_prio_expression_meta_selected,
        #                sm="OR",
        #                #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
        #                studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1])),
        #        plotwidth="6inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
      }
    })

  })
  #### each dataset####
  output$prio_expression_plots <- renderUI({
    input$do_prio
    isolate({
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

      #      browse_prio_expression <- browse_prio_expression_function()
      subset.browse_prio_expression <- browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
      #selected.database_prio_expression <- selecteddatabase_prio_expression()
      if (nrow(subset.browse_prio_expression)==0 ) {
        tagList(plotlyOutput("prio_expression_plots0",height="auto"),dataTableOutput("prio_expression_tables0"))
      } else {
        lapply(1:length(input$data_source_prio),function(i) {
          column(width=6,wellPanel(plotlyOutput(paste0("prio_expression_plots",i),height="auto"),
                                   h4("Significant genes:",style="margin-top:20px;margin-bottom:10px;font-size:20px;font-weight:bold;color:#DCB239"),
                                   div(DT::dataTableOutput(paste0("prio_expression_tables",i)),style="margin-top:30px;margin-bottom:10px"),style="height:1300px"),style="padding-bottom:20px")
          #column(width=6,wellPanel(DT::dataTableOutput(paste0("prio_expression_tables",i))),style="padding-bottom:20px")
        })
      }
    })
  })

  observeEvent(input$do_prio,{
    input$do_prio

    isolate({
      #        browse_prio_expression <- browse_prio_expression_function()
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

      #      browse_prio_expression <- browse_prio_expression_function()
      subset.browse_prio_expression <- browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]

      if (nrow(subset.browse_prio_expression)==0) {
        output$prio_expression_plots0 <- renderPlotly({
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_prio,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="Warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
        })

        output$prio_expression_tables0 <- DT::renderDataTable({
          datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        })

      } else {
        # splitted_list <- selected.database_prio_expression %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        # splitted_list <- splitted_list[input$data_source_prio]
        isolate({
          for (k in 1:length(input$data_source_prio)) {
            local({
              my_i <- k
              browse_prio_expression_plot_name <- paste0("prio_expression_plots", my_i)
              browse_prio_expression_table_name <- paste0("prio_expression_tables", my_i)
              #             d<-splitted_list[[my_i]]

              output[[browse_prio_expression_plot_name]] <- renderPlotly({
                isolate({
                  withProgress(message = 'Making plot', value = 0, {
                    if (strsplit(rownames(subset.browse_prio_expression)[my_i],"\\.")[[1]][1]=="NA") {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>%
                        layout(title = str_sub(input$data_source_prio,end=-19)[my_i],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="Warning 2: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    } else {
                      if (length(na.omit(as.numeric(subset.browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))==0) {
                        incProgress(100, detail = "Doing part %")
                        plot_ly () %>%
                          layout(title = unique(str_sub(input$data_source_prio,end=-19)[my_i]),font=list(size=15,family="Aril")) %>%
                          add_annotations(x=2,y=2,text="Warning 3: no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                      } else {
                        vline <- function(x = 0, color = "red") {
                          list(
                            type = "line",
                            y0 = 0,
                            y1 = 1,
                            yref = "paper",
                            x0 = x,
                            x1 = x,
                            line = list(color = color)
                          )
                        }
                        d <- subset.browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                        d.nonna <- d[,!is.na(d[1,])]
                        d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))))
                        rownames(d.nonna.t) <- colnames(d.nonna)
                        colnames(d.nonna.t) <- c(paste0("p-value in ",rownames(d.nonna)),paste0("-log10(p-value) in ",rownames(d.nonna)))
                        fit <- density(d.nonna.t[,2])
                        # p<-plot_ly(x=d.nonna.t[,2],
                        #            #y=as.numeric(d[-1]),
                        #            type="histogram",#type="box",
                        #            #histnorm="probability",
                        #            name="Histogram"
                        # ) %>%
                        #   add_trace(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",yaxis="y2",name="Density") %>%
                        p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
                          layout(title=str_sub(input$data_source_prio[my_i],end=-19),
                                 height = 480,
                                 xaxis = list(title="-log10(p.value)",
                                              titlefont=list(size=20,color="black",family= "Aril black"),
                                              showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title="Density",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 shapes=list(vline(1.3))) %>%
                          add_annotations(x = 1.3,
                                          y = 1,
                                          text = "p.value = 0.05",
                                          xref = paste0("x", 1),  # add this
                                          yref = paste0("y", 1),  # add this
                                          showarrow = FALSE,
                                          ax = 20,
                                          ay = -40,
                                          font=list(size=20,color="black",family= "Aril black"))
                        incProgress(100, detail = "Doing part %")
                        p
                      }
                    }
                  })
                })

              })

              #output[[browse_prio_expression_table_name]] <- DT::renderDataTable({
                isolate({
                  if (rownames(subset.browse_prio_expression)[my_i]=="NA") {
                    prio_expression_tables_data <-
                    datatable(data.frame("NA"="Warning 2: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    if (length(na.omit(as.numeric(subset.browse_prio_expression[str_sub(input$data_source_prio,end=-19)[my_i],])))==0) {
                      prio_expression_tables_data <-
                      datatable(data.frame("NA"="Warning 3: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {

                      d <- subset.browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                      d.nonna <- d[,!is.na(d[1,])]
                      if (ncol(d.nonna)==0) {
                        prio_expression_tables_data <-
                        datatable(data.frame("NA"="no clinical data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                      } else {
                        d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))),p.adjust(as.numeric(t(d.nonna)),"fdr"))
                        rownames(d.nonna.t) <- colnames(d.nonna)
                        colnames(d.nonna.t) <- c(paste0("p-value in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("-log10(p-value) in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("FDR in ",str_split(rownames(d.nonna),"_",2)[[1]][1]))
                        colnames(d.nonna.t) <- c("p-value","-log10(p-value)","FDR")
                        d.nonna.t.p0.05 <- d.nonna.t[d.nonna.t[,1]<=0.05,]
                        d.nonna.t.p0.05 <- d.nonna.t.p0.05[order(d.nonna.t.p0.05[,1]),]

                        shinyInput <- function(FUN, len, id, ...) {
                          inputs <- character(len)
                          for (i in seq_len(len)) {
                            inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
                          }
                          inputs
                        }

                        d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
                          # Gene=rownames(d.nonna.t.p0.05),
                          # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
                          # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
                          signif(d.nonna.t.p0.05[,c(1,3)],2),
                          View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_"),label="View",onclick='Shiny.onInputChange(\"select_button_prio_expression\",  this.id)'),
                          stringsAsFactors = FALSE#,
                          #row.names = 1:nrow(d.nonna.t.p0.05)
                        ))

                        prio_expression_tables_data <- d.nonna.t.p0.05.test$data

                        # datatable(d.nonna.t.p0.05) %>%
                        #   formatSignif(columns = c(1:ncol(d.nonna.t.p0.05)), digits = 2,getOption("scipen"))
                        #datatable(d.nonna.t.p0.05,options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                  }
                })
                output[[browse_prio_expression_table_name]] <- DT::renderDataTable(
                  prio_expression_tables_data,server = FALSE, escape = FALSE, selection = 'none'
                )

              #})

            })
          }
        })
      }
    })

  })

  observeEvent(input$select_button_prio_expression,{
    s <- as.numeric(strsplit(input$select_button_prio_expression, "_")[[1]][3])
    output$prio_expression_modals <- renderUI({
      tagList(
        bsModal(paste('prio_expression_model', s ,sep=''), "View", "select_button_prio_expression", size = "large",

                plotlyOutput("browse_prio_expression_bsModal_plot")
                #textOutput("browse_prio_expression_bsModal_plot")
        )
      )
    })
    toggleModal(session,paste('prio_expression_model', s ,sep=''), toggle = "View")
    ##Reset the select_button
    session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_expression")
  })

  observeEvent(input$select_button_prio_expression,{

    data_source1 <- as.character(strsplit(input$select_button_prio_expression, "_")[[1]][1])
    gene <- as.character(strsplit(input$select_button_prio_expression, "_")[[1]][2])
    input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
    input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

    subset.browse_prio_expression_plot <- browse_prio_expression_plot[paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),gene]
    d <- rbind(data.frame(PFS.status="nonresponse",Gene=as.numeric(strsplit(strsplit(as.character(subset.browse_prio_expression_plot),"_")[[1]][2],",")[[1]])),
               data.frame(PFS.status="response",Gene=as.numeric(strsplit(strsplit(as.character(subset.browse_prio_expression_plot),"_")[[1]][1],",")[[1]])))
    p<-plot_ly(data=d,x=~PFS.status,
               y=~Gene,
               type="violin",#type="box",
               points="all",#boxpoints="all",
               jitter=0.3,
               pointpos=-1.8,
               #mode="markders",
               color= ~PFS.status,
               colors=c("#7CDBD5","#DCB239"),
               #hoveron="points+boxes",
               legendgroup=~PFS.status,
               showlegend=TRUE
    ) %>%
      layout(title = unique(data_source1),titlefont=list(size=20,color="black",family= "Aril"),
             height = 480,
             xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                        showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
             yaxis=list(title=gene,
                        titlefont=list(size=20,color="black",family= "Aril"),
                        showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
             )) %>%
      add_annotations(x=0.5,
                      y=d[which.max(d[,"Gene"]),"Gene"],

                      text=paste0("p-value = ",signif(as.numeric(browse_prio_expression[paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),gene]),2)),
                      xref = paste0("x", 1),  # add this
                      yref = paste0("y", 1),  # add this
                      showarrow = FALSE,
                      ax = 20,
                      ay = -40,
                      font=list(size=20,color="black",family= "Aril black"))
    output$browse_prio_expression_bsModal_plot <- renderPlotly(
      p
    )
    # output$browse_prio_expression_bsModal_plot <- renderText(
    #   # input$select_button_prio_expression
    #   # nrow(browse_prio_expression_plot)
    #   # paste0(paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),"_____",gene)
    #   paste0(data_source1,";",gene,";",paste(input_treatment_prio,collapse = "_"),";",
    #          paste(input_description_prio,collapse = "_"),";",subset.browse_prio_expression_plot,";",
    #          browse_prio_expression_plot[1,1],";",rownames(browse_prio_expression_plot)[1],";",
    #          colnames(browse_prio_expression_plot)[1])
    # )


  })

  #### download ####
  # observeEvent(input$do_prio,{
  #   input$do_prio
  #   isolate({
  #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
  #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
  #
  #     #        browse_prio_expression <- browse_prio_expression_function()
  #     #subset.browse_prio_expression.t <- unfactor(as.data.frame(t(browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))
  #     subset.browse_prio_expression.t_fisher_part_pre <- as.data.frame(t(browse_prio_expression[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
  #
  #     #colnames(subset.browse_prio_expression.t) <- paste0("p-value in ",sapply(str_split(colnames(subset.browse_prio_expression.t),"_",2),function(x) x[1]))
  #     if(nrow(subset.browse_prio_expression.t_fisher_part_pre)>0) {
  #       colnames(subset.browse_prio_expression.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_expression.t_fisher_part_pre),"_",2))[1,])
  #       #paste0("p-value in ",str_sub(input$data_source_prio,end=-19))
  #       meta.p <- browse_prio_expression_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
  #       meta.fdr <- p.adjust(meta.p)
  #       subset.browse_prio_expression.t_meta_part <- data.frame("meta-analysis p-value"=meta.p,
  #                                                               "FDR of meta-analysis p-value"=meta.fdr)
  #       combined.p <- fisher.method(subset.browse_prio_expression.t_fisher_part_pre,na.rm=TRUE)$p.value
  #       subset.browse_prio_expression.t_fisher_part <- data.frame("combined p-value"=combined.p,
  #                                                                 "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
  #                                                                 subset.browse_prio_expression.t_fisher_part_pre,stringsAsFactors = F )
  #       subset.browse_prio_expression.t <- merge(subset.browse_prio_expression.t_meta_part,subset.browse_prio_expression.t_fisher_part,by="row.names",all=T)
  #       rownames(subset.browse_prio_expression.t) <- subset.browse_prio_expression.t[,1]
  #       subset.browse_prio_expression.t <- subset.browse_prio_expression.t[,-1]
  #       subset.browse_prio_expression.t <- subset.browse_prio_expression.t[order(subset.browse_prio_expression.t[,"meta.analysis.p.value"]),]
  #       subset.browse_prio_expression.t <- subset.browse_prio_expression.t[,-which(apply(subset.browse_prio_expression.t,2,function(x)all(is.na(x))))]
  #
  #       #subset.browse_prio_expression.t <- subset.browse_prio_expression.t[order(subset.browse_prio_expression.t[,"combined.p.value"],subset.browse_prio_expression.t[,names(apply(subset.browse_prio_expression.t,2,function(x) length(na.omit(x)))[2])]),]
  #       brks <- quantile(subset.browse_prio_expression.t, probs = seq(.05, .95, .05), na.rm = TRUE)
  #       clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}
  #     } else {
  #       subset.browse_prio_expression.t <- subset.browse_prio_expression.t_fisher_part_pre
  #     }
  #   })
  #   output$prio_expression_download <- downloadHandler(
  #     filename = function() {
  #       if (input$prio_expression_type == "Excel (CSV)") {
  #         "newfile.csv"
  #       } else if (input$prio_expression_type == "Text (TSV)") {
  #         "newfile.txt"
  #       } else if (input$prio_expression_type == "JSON") {
  #         "newfile.json"
  #       }
  #     },
  #     content = function(file) {
  #       if (input$prio_expression_type == "Excel (CSV)") {
  #         fwrite (subset.browse_prio_expression.t,row.names=T,file)
  #       } else if (input$prio_expression_type == "Text (TSV)") {
  #         fwrite (subset.browse_prio_expression.t,row.names=T,file)
  #         #write.table (selectedexpression_expression(),file,quote=F, row.names = FALSE, sep="\t")
  #       } else if (input$prio_expression_type == "JSON") {
  #         write(toJSON(subset.browse_prio_expression.t),row.names=T,file)
  #       }
  #     }
  #   )
  #   output$prio_expression_table <- DT::renderDataTable({
  #     datatable(subset.browse_prio_expression.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
  #       formatStyle(names(subset.browse_prio_expression.t), backgroundColor = styleInterval(brks, clrs)) %>%
  #       formatSignif(columns = c(1:ncol(subset.browse_prio_expression.t)), digits = 2,getOption("scipen"))
  #   })
  # })


  ############################
  #### signatures prio cibersort
  ############################
  #input <- list(data_source_prio=c("Melanoma-Van Allen (112 samples/+--)","Melanoma-Hugo (040 samples/-+-)","Melanoma-Snyder-Nathanson (064 samples/+--)"),treatment_prio=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),description_prio=c("Pre","On"))

  #### meta-analysis ####
  # observeEvent(input$do_prio,{
  #   isolate({
  #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
  #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
  #     subset.browse_prio_cibersort_meta_p <- browse_prio_cibersort_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
  #     if (length(subset.browse_prio_cibersort_meta_p)==0) {
  #       #output$prio_cibersort_plots_meta <- renderPlotly({
  #       prio_cibersort_plots_meta_p <- plot_ly () %>%
  #         layout(title = "meta-anlaysis across selected cohorts",font=list(size=15,family="Aril")) %>%
  #         add_annotations(x=2,y=2,text="Warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
  #       #})
  #       prio_cibersort_tables_meta_data <- datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
  #     } else {
  #       fit <- density(na.omit(subset.browse_prio_cibersort_meta_p))
  #       vline <- function(x = 0, color = "red") {
  #         list(
  #           type = "line",
  #           y0 = 0,
  #           y1 = 1,
  #           yref = "paper",
  #           x0 = x,
  #           x1 = x,
  #           line = list(color = color)
  #         )}
  #       prio_cibersort_plots_meta_p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
  #         layout(title="meta-anlaysis across all cohorts",
  #                height = 480,
  #                xaxis = list(title="-log10(p.value)",
  #                             titlefont=list(size=20,color="black",family= "Aril black"),
  #                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
  #                yaxis=list(title="Density",
  #                           titlefont=list(size=20,color="black",family= "Aril black"),
  #                           showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
  #                shapes=list(vline(1.3))) %>%
  #         add_annotations(x = 1.3,
  #                         y = 1,
  #                         text = "p.value = 0.05",
  #                         xref = paste0("x", 1),  # add this
  #                         yref = paste0("y", 1),  # add this
  #                         showarrow = FALSE,
  #                         ax = 20,
  #                         ay = -40,
  #                         font=list(size=20,color="black",family= "Aril black"))
  #
  #
  #       prio_cibersort_meta_table <- data.frame("p.value"=subset.browse_prio_cibersort_meta_p,
  #                                               "FDR"=p.adjust(subset.browse_prio_cibersort_meta_p))
  #       prio_cibersort_meta_table_sig <- prio_cibersort_meta_table# keep all mutation sigantures [prio_cibersort_meta_table$p.value<=0.05 & !is.na(prio_cibersort_meta_table$p.value),]
  #       prio_cibersort_meta_table_sig <- prio_cibersort_meta_table_sig[order(prio_cibersort_meta_table_sig$p.value),]
  #       shinyInput <- function(FUN, len, id, ...) {
  #         inputs <- character(len)
  #         for (i in seq_len(len)) {
  #           inputs[i] <- as.character(FUN(paste0(id, rownames(prio_cibersort_meta_table_sig)[i],"_",i), ...))
  #         }
  #         inputs
  #       }
  #       d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
  #         # Gene=rownames(d.nonna.t.p0.05),
  #         # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
  #         # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
  #         signif(prio_cibersort_meta_table_sig,2),
  #         View=shinyInput(actionButton,nrow(prio_cibersort_meta_table_sig),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_cibersort_meta\",  this.id)'),
  #         stringsAsFactors = FALSE#,
  #         #row.names = 1:nrow(d.nonna.t.p0.05)
  #       ))
  #       prio_cibersort_tables_meta_data <- d.nonna.t.p0.05.test$data
  #
  #     }
  #     output$prio_cibersort_plots_meta <- renderPlotly({
  #       prio_cibersort_plots_meta_p
  #     })
  #     output$prio_cibersort_tables_meta <- DT::renderDataTable(
  #       prio_cibersort_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none'
  #     )
  #   })
  # })
  # observeEvent(input$select_button_prio_cibersort_meta,{
  #   s <- as.numeric(strsplit(input$select_button_prio_cibersort_meta, "_")[[1]][3])
  #   output$prio_cibersort_tables_meta_modals <- renderUI({
  #     tagList(
  #       bsModal(paste0("prio_cibersort_tables_meta_bsModal",s), "View", "select_button_prio_cibersort_meta", size = "large",
  #               plotOutput("browse_prio_cibersort_meta_bsModal_plot")
  #       )
  #     )
  #   })
  #   toggleModal(session,paste0("prio_cibersort_tables_meta_bsModal",s), toggle = "View")
  #   ##Reset the select_button
  #   session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_cibersort_meta")
  # })
  # observeEvent(input$select_button_prio_cibersort_meta,{
  #   input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
  #   input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
  #   browse_prio_cibersort_meta_selected <- browse_prio_cibersort_meta[
  #     browse_prio_cibersort_meta$genes==as.character(strsplit(input$select_button_prio_cibersort_meta, "_")[[1]][2]) &
  #       browse_prio_cibersort_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
  #   output$browse_prio_cibersort_meta_bsModal_plot <- renderPlot({
  #     forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
  #                     data=browse_prio_cibersort_meta_selected,
  #                     label.e = "response",label.c = "nonresponse",
  #                     studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
  #            test.overall = TRUE,ff.study.label = "bold")
  #   })
  #
  # })

  ### meta-analysis ### table only

  observeEvent(input$do_prio,{
    isolate({
      withProgress(message = 'Preparing data', value = 0, {
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
      browse_prio_cibersort_meta_selected <- browse_prio_cibersort_meta[
        browse_prio_cibersort_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
      if (nrow(browse_prio_cibersort_meta_selected)==0 |
          all(is.na(browse_prio_cibersort_meta_selected$TE)) |
          all(is.na(browse_prio_cibersort_meta_selected$seTE)) ) {
        prio_cibersort_tables_meta_data <- datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        # meta.p <- NA
        # meta.fdr <- NA
        fit.t <- NULL
      } else {
        incProgress(20, detail = "Doing part %")
        if (input$sample_prio=="custom selection") {
          splitted <- browse_prio_cibersort_meta_selected %>% split(.$genes)
          fit <- lapply(splitted,
                      function(x) {
                        if (all(is.na(x$TE)) | all(is.na(x$seTE)) | nrow(x)==0 | is.null(x)) {
                          c(NA,NA,NA)
                        } else {
                          a <- metagen(TE,seTE,data=x,sm="OR")
                          #c(a$pval.fixed,a$pval.random,a$pval.Q)})
                          c(a$pval.random,a$pval.fixed,a$pval.Q)
                        }
                      })
        
          fit.t <- t(as.data.frame(fit))
          fit.t.fdr.random <- p.adjust(fit.t[,1],"fdr")
          fit.t.fdr.fixed <- p.adjust(fit.t[,2],"fdr")
          # fit.t <- data.frame(pval.fixed=fit.t[,1],
          #                   fdr.fixed=fit.t.fdr,
          #                   pval.random=fit.t[,2],
          #                   pval.Q=fit.t[,3])
          # fit.t <- fit.t[order(fit.t$pval.fixed),]
          fit.t <- data.frame(Pval.random=fit.t[,1],
                              FDR.random=fit.t.fdr.random,
                              Pval.fixed=fit.t[,2],
                              FDR.fixed=fit.t.fdr.fixed)
          fit.t <- fit.t[order(fit.t$Pval.random),]
        #c("pval.fixed","pval.random","pval.Q")
        } else {
          fit.t <- meta_analysis_prio_precalculated_cibersort[[input$sample_prio]][,1:4]
          
        }
        shinyInput <- function(FUN, len, id, ...) {
          inputs <- character(len)
          for (i in seq_len(len)) {
            inputs[i] <- as.character(FUN(paste0(id, rownames(fit.t)[i],"_",i), ...))
          }
          inputs
        }
        d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
          # Gene=rownames(d.nonna.t.p0.05),
          # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
          # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
          signif(fit.t,2),
          View=shinyInput(actionButton,nrow(fit.t),"alldatasource_",label="View",onclick='Shiny.onInputChange(\"select_button_prio_cibersort_meta\",  this.id)'),
          stringsAsFactors = FALSE#,
          #row.names = 1:nrow(d.nonna.t.p0.05)
        ))
        prio_cibersort_tables_meta_data <- d.nonna.t.p0.05.test$data
      }
        subset.browse_prio_cibersort.t_fisher_part_pre <- as.data.frame(t(browse_prio_cibersort[rownames(browse_prio_cibersort) %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
        if(nrow(subset.browse_prio_cibersort.t_fisher_part_pre)>0) {
          if (input$sample_prio=="custom selection") {
            colnames(subset.browse_prio_cibersort.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_cibersort.t_fisher_part_pre),"_",2))[1,])
            subset.browse_prio_cibersort.t_meta_part <- fit.t
            combined.p <- fisher.method(subset.browse_prio_cibersort.t_fisher_part_pre,na.rm=TRUE)$p.value
            subset.browse_prio_cibersort.t_fisher_part <- data.frame("combined p-value"=combined.p,
                                                                   "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
                                                                   subset.browse_prio_cibersort.t_fisher_part_pre,stringsAsFactors = F )
            subset.browse_prio_cibersort.t <- merge(subset.browse_prio_cibersort.t_meta_part,subset.browse_prio_cibersort.t_fisher_part,by="row.names",all=T)
            rownames(subset.browse_prio_cibersort.t) <- subset.browse_prio_cibersort.t[,1]
            subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[,-1]
            subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[order(subset.browse_prio_cibersort.t[,1]),]
          # subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[,-which(apply(subset.browse_prio_cibersort.t,2,function(x)all(is.na(x))))]
            subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[,apply(subset.browse_prio_cibersort.t,2,function(x) !all(is.na(x)))]
          } else {
            subset.browse_prio_cibersort.t <- meta_analysis_prio_precalculated_cibersort[[input$sample_prio]]
            
          }
          brks <- quantile(subset.browse_prio_cibersort.t, probs = seq(.05, .95, .05), na.rm = TRUE)
          clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}

        } else {
          subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t_fisher_part_pre

        }
      #}
      incProgress(100, detail = "Doing part %")



      output$prio_cibersort_tables_meta <- DT::renderDataTable(
        prio_cibersort_tables_meta_data,server = FALSE, escape = FALSE, selection = 'none',caption = 'Pval.random: p-value for test of overall treatment effect (random effects model). FDR.random: adjusted p-value of random effects model using false discovery rate method. Pval.fixed: p-value for test of overall treatment effect (fixed effect model). FDR.fixed: adjusted p-value of fixed effect model using false discovery rate method.'
      )

      output$prio_cibersort_download <- downloadHandler(
        filename = function() {
          if (input$prio_cibersort_type == "Excel (CSV)") {
            "newfile.csv"
          } else if (input$prio_cibersort_type == "Text (TSV)") {
            "newfile.txt"
          } else if (input$prio_cibersort_type == "JSON") {
            "newfile.json"
          }
        },
        content = function(file) {
          if (input$prio_cibersort_type == "Excel (CSV)") {
            fwrite (subset.browse_prio_cibersort.t,row.names=T,file)
          } else if (input$prio_cibersort_type == "Text (TSV)") {
            fwrite (subset.browse_prio_cibersort.t,row.names=T,file)
            #write.table (selectedcibersort_cibersort(),file,quote=F, row.names = FALSE, sep="\t")
          } else if (input$prio_cibersort_type == "JSON") {
            write(toJSON(subset.browse_prio_cibersort.t),row.names=T,file)
          }
        }
      )
      output$prio_cibersort_table <- DT::renderDataTable({
        datatable(subset.browse_prio_cibersort.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
          formatStyle(names(subset.browse_prio_cibersort.t), backgroundColor = styleInterval(brks, clrs)) %>%
          formatSignif(columns = c(1:ncol(subset.browse_prio_cibersort.t)), digits = 2,getOption("scipen"))
      })
      })
    })
  })

  observeEvent(input$select_button_prio_cibersort_meta,{
    s <- as.numeric(strsplit(input$select_button_prio_cibersort_meta, "_")[[1]][3])
    Viewname <- as.character(strsplit(input$select_button_prio_cibersort_meta, "_")[[1]][2])
    output$prio_cibersort_tables_meta_modals <- renderUI({
      tagList(
        bsModal(paste0("prio_cibersort_tables_meta_bsModal",s), Viewname, "select_button_prio_cibersort_meta", size = "large",
                plotOutput("browse_prio_cibersort_meta_bsModal_plot")
        )
      )
    })
    toggleModal(session,paste0("prio_cibersort_tables_meta_bsModal",s), toggle = Viewname)
    ##Reset the select_button
    session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_cibersort_meta")
  })

  observeEvent(input$select_button_prio_cibersort_meta,{
    input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
    input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
    browse_prio_cibersort_meta_selected <- browse_prio_cibersort_meta[
      browse_prio_cibersort_meta$genes==as.character(strsplit(input$select_button_prio_cibersort_meta, "_")[[1]][2]) &
        browse_prio_cibersort_meta$data_source %in% paste0(str_sub(input$data_source_prio,end=-19),"_",paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))),]
    output$browse_prio_cibersort_meta_bsModal_plot <- renderPlot({
      
      # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
      #                 data=browse_prio_cibersort_meta_selected,
      #                 sm="MD",
      #                 label.e = "response",label.c = "nonresponse",
      #                 studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
      #        test.overall = TRUE,ff.study.label = "bold")
      if (nrow(browse_prio_cibersort_meta_selected)==0 |
          is.null(browse_prio_cibersort_meta_selected) |
          all(is.na(browse_prio_cibersort_meta_selected$TE)) |
          all(is.na(browse_prio_cibersort_meta_selected$seTE))) {
        plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_prio,end=-19),collapse = ","),cex.main=1.5,bty="l")
        abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
        text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
      } else {
        m1 <- metagen(TE,seTE,data=browse_prio_cibersort_meta_selected,
                      sm="OR",
                      #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
                      studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1]))
        m1$upper[exp(m1$upper)>1e+50] <- 1000
        m1$upper.fixed[exp(m1$upper.fixed)>1e+50] <- 1000
        m1$upper.random[exp(m1$upper.random)>1e+50] <- 1000
        m1$lower.fixed[exp(m1$lower.fixed)>1e+50] <- 1000
        m1$lower.random[exp(m1$lower.random)>1e+50] <- 1000
        
        forest(m1,layout="RevMan5",    
               hetstat=TRUE,
               plotwidth="6inch",test.overall = TRUE,
               xlim=c(0.1,10),backtransfer=T,
               label.test.overall.random="random effects model: ",
               label.test.overall.fixed="fixed effects model: ",
               fs.test.overall = 15,
               scientific.pval = TRUE,ff.study.label = "bold")
        # forest(metagen(TE,seTE,data=browse_prio_cibersort_meta_selected,
        #                sm="OR",
        #                #studlab = as.character(as.data.frame(str_split(data_source,"_",2))[1,])),
        #                studlab = as.character(t(as.data.frame(str_split(data_source,"_",2)))[,1])),
        #        plotwidth="6inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
      }
    })

  })
  #### each dataset####
  output$prio_cibersort_plots <- renderUI({
    input$do_prio
    isolate({
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

      #      browse_prio_cibersort <- browse_prio_cibersort_function()
      subset.browse_prio_cibersort <- browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
      #selected.database_prio_cibersort <- selecteddatabase_prio_cibersort()
      if (nrow(subset.browse_prio_cibersort)==0 ) {
        tagList(plotlyOutput("prio_cibersort_plots0",height="auto"),dataTableOutput("prio_cibersort_tables0"))
      } else {
        lapply(1:length(input$data_source_prio),function(i) {
          column(width=6,wellPanel(plotlyOutput(paste0("prio_cibersort_plots",i),height="auto"),
                                   h4("All mutation signatures:",style="margin-top:20px;margin-bottom:10px;font-size:20px;font-weight:bold;color:#DCB239"),
                                   div(DT::dataTableOutput(paste0("prio_cibersort_tables",i)),style="margin-top:30px;margin-bottom:10px"),style="height:1300px"),style="padding-bottom:20px")
          #column(width=6,wellPanel(DT::dataTableOutput(paste0("prio_cibersort_tables",i))),style="padding-bottom:20px")
        })
      }
    })
  })

  observeEvent(input$do_prio,{
    input$do_prio

    isolate({
      #        browse_prio_cibersort <- browse_prio_cibersort_function()
      input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
      input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

      #      browse_prio_cibersort <- browse_prio_cibersort_function()
      subset.browse_prio_cibersort <- browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]

      if (nrow(subset.browse_prio_cibersort)==0) {
        output$prio_cibersort_plots0 <- renderPlotly({
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_prio,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="Warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
        })

        output$prio_cibersort_tables0 <- DT::renderDataTable({
          datatable(data.frame("NA"="Warning 1: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        })

      } else {
        # splitted_list <- selected.database_prio_cibersort %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        # splitted_list <- splitted_list[input$data_source_prio]
        isolate({
          for (k in 1:length(input$data_source_prio)) {
            local({
              my_i <- k
              browse_prio_cibersort_plot_name <- paste0("prio_cibersort_plots", my_i)
              browse_prio_cibersort_table_name <- paste0("prio_cibersort_tables", my_i)
              #             d<-splitted_list[[my_i]]

              output[[browse_prio_cibersort_plot_name]] <- renderPlotly({
                isolate({
                  withProgress(message = 'Making plot', value = 0, {
                    if (strsplit(rownames(subset.browse_prio_cibersort)[my_i],"\\.")[[1]][1]=="NA") {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>%
                        layout(title = str_sub(input$data_source_prio,end=-19)[my_i],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="Warning 2: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    } else {
                      if (length(na.omit(as.numeric(subset.browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))==0) {
                        incProgress(100, detail = "Doing part %")
                        plot_ly () %>%
                          layout(title = unique(str_sub(input$data_source_prio,end=-19)[my_i]),font=list(size=15,family="Aril")) %>%
                          add_annotations(x=2,y=2,text="Warning 3: no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                      } else {
                        vline <- function(x = 0, color = "red") {
                          list(
                            type = "line",
                            y0 = 0,
                            y1 = 1,
                            yref = "paper",
                            x0 = x,
                            x1 = x,
                            line = list(color = color)
                          )
                        }
                        d <- subset.browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                        d.nonna <- d[,!is.na(d[1,])]
                        d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))))
                        rownames(d.nonna.t) <- colnames(d.nonna)
                        colnames(d.nonna.t) <- c(paste0("p-value in ",rownames(d.nonna)),paste0("-log10(p-value) in ",rownames(d.nonna)))
                        fit <- density(d.nonna.t[,2])
                        # p<-plot_ly(x=d.nonna.t[,2],
                        #            #y=as.numeric(d[-1]),
                        #            type="histogram",#type="box",
                        #            #histnorm="probability",
                        #            name="Histogram"
                        # ) %>%
                        #   add_trace(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",yaxis="y2",name="Density") %>%
                        p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
                          layout(title=str_sub(input$data_source_prio[my_i],end=-19),
                                 height = 480,
                                 xaxis = list(title="-log10(p.value)",
                                              titlefont=list(size=20,color="black",family= "Aril black"),
                                              showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title="Density",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 shapes=list(vline(1.3))) %>%
                          add_annotations(x = 1.3,
                                          y = 1,
                                          text = "p.value = 0.05",
                                          xref = paste0("x", 1),  # add this
                                          yref = paste0("y", 1),  # add this
                                          showarrow = FALSE,
                                          ax = 20,
                                          ay = -40,
                                          font=list(size=20,color="black",family= "Aril black"))
                        incProgress(100, detail = "Doing part %")
                        p
                      }
                    }
                  })
                })

              })

              #output[[browse_prio_cibersort_table_name]] <- DT::renderDataTable({
              isolate({
                if (rownames(subset.browse_prio_cibersort)[my_i]=="NA") {
                  prio_cibersort_tables_data <-
                    datatable(data.frame("NA"="Warning 2: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                } else {
                  if (length(na.omit(as.numeric(subset.browse_prio_cibersort[str_sub(input$data_source_prio,end=-19)[my_i],])))==0) {
                    prio_cibersort_tables_data <-
                      datatable(data.frame("NA"="Warning 3: no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {

                    d <- subset.browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]
                    d.nonna <- d[,!is.na(d[1,])]
                    if (ncol(d.nonna)==0) {
                      prio_cibersort_tables_data <-
                        datatable(data.frame("NA"="no clinical data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      d.nonna.t <- data.frame(data.frame(as.numeric(t(d.nonna))),-log10(data.frame(as.numeric(t(d.nonna)))),p.adjust(as.numeric(t(d.nonna)),"fdr"))
                      rownames(d.nonna.t) <- colnames(d.nonna)
                      colnames(d.nonna.t) <- c(paste0("p-value in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("-log10(p-value) in ",str_split(rownames(d.nonna),"_",2)[[1]][1]),paste0("FDR in ",str_split(rownames(d.nonna),"_",2)[[1]][1]))
                      colnames(d.nonna.t) <- c("p-value","-log10(p-value)","FDR")
                      d.nonna.t.p0.05 <- d.nonna.t# keep all mutation signatures[d.nonna.t[,1]<=0.05,]
                      d.nonna.t.p0.05 <- d.nonna.t.p0.05[order(d.nonna.t.p0.05[,1]),]

                      shinyInput <- function(FUN, len, id, ...) {
                        inputs <- character(len)
                        for (i in seq_len(len)) {
                          inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
                        }
                        inputs
                      }

                      d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
                        # Gene=rownames(d.nonna.t.p0.05),
                        # p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
                        # FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
                        signif(d.nonna.t.p0.05[,c(1,3)],2),
                        View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),paste0(str_sub(input$data_source_prio,end=-19)[my_i],"_"),label="View",onclick='Shiny.onInputChange(\"select_button_prio_cibersort\",  this.id)'),
                        stringsAsFactors = FALSE#,
                        #row.names = 1:nrow(d.nonna.t.p0.05)
                      ))

                      prio_cibersort_tables_data <- d.nonna.t.p0.05.test$data

                      # datatable(d.nonna.t.p0.05) %>%
                      #   formatSignif(columns = c(1:ncol(d.nonna.t.p0.05)), digits = 2,getOption("scipen"))
                      #datatable(d.nonna.t.p0.05,options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    }
                  }
                }
              })
              output[[browse_prio_cibersort_table_name]] <- DT::renderDataTable(
                prio_cibersort_tables_data,server = FALSE, escape = FALSE, selection = 'none'
              )

              #})

            })
          }
        })
      }
    })

  })

  observeEvent(input$select_button_prio_cibersort,{
    s <- as.numeric(strsplit(input$select_button_prio_cibersort, "_")[[1]][3])
    output$prio_cibersort_modals <- renderUI({
      tagList(
        bsModal(paste('prio_cibersort_model', s ,sep=''), "View", "select_button_prio_cibersort", size = "large",

                plotlyOutput("browse_prio_cibersort_bsModal_plot")
                #textOutput("browse_prio_cibersort_bsModal_plot")
        )
      )
    })
    toggleModal(session,paste('prio_cibersort_model', s ,sep=''), toggle = "View")
    ##Reset the select_button
    session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_prio_cibersort")
  })

  observeEvent(input$select_button_prio_cibersort,{

    data_source1 <- as.character(strsplit(input$select_button_prio_cibersort, "_")[[1]][1])
    gene <- as.character(strsplit(input$select_button_prio_cibersort, "_")[[1]][2])
    input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
    input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]

    subset.browse_prio_cibersort_plot <- browse_prio_cibersort_plot[paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),gene]
    d <- rbind(data.frame(PFS.status="nonresponse",Gene=as.numeric(strsplit(strsplit(as.character(subset.browse_prio_cibersort_plot),"_")[[1]][2],",")[[1]])),
               data.frame(PFS.status="response",Gene=as.numeric(strsplit(strsplit(as.character(subset.browse_prio_cibersort_plot),"_")[[1]][1],",")[[1]])))
    p<-plot_ly(data=d,x=~PFS.status,
               y=~Gene,
               type="violin",#type="box",
               points="all",#boxpoints="all",
               jitter=0.3,
               pointpos=-1.8,
               #mode="markders",
               color= ~PFS.status,
               colors=c("#7CDBD5","#DCB239"),
               #hoveron="points+boxes",
               legendgroup=~PFS.status,
               showlegend=TRUE
    ) %>%
      layout(title = unique(data_source1),titlefont=list(size=20,color="black",family= "Aril"),
             height = 480,
             xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                        showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
             yaxis=list(title=gene,
                        titlefont=list(size=20,color="black",family= "Aril"),
                        showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
             )) %>%
      add_annotations(x=0.5,
                      y=d[which.max(d[,"Gene"]),"Gene"],

                      text=paste0("p-value = ",signif(as.numeric(browse_prio_cibersort[paste0(data_source1,"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),gene]),2)),
                      xref = paste0("x", 1),  # add this
                      yref = paste0("y", 1),  # add this
                      showarrow = FALSE,
                      ax = 20,
                      ay = -40,
                      font=list(size=20,color="black",family= "Aril black"))
    output$browse_prio_cibersort_bsModal_plot <- renderPlotly(
      p
    )


  })

  #### download ####
  # observeEvent(input$do_prio,{
  #   input$do_prio
  #   isolate({
  #     input_treatment_prio <- input$treatment_prio[c(which(input$treatment_prio=="anti-PD-1/L1"),which(input$treatment_prio=="anti-CTLA-4"),which(input$treatment_prio=="anti-PD-1/L1+anti-CTLA-4"))]
  #     input_description_prio <- input$description_prio[c(which(input$description_prio=="Pre"),which(input$description_prio=="On"),which(input$description_prio=="Unknown"))]
  #
  #     #        browse_prio_cibersort <- browse_prio_cibersort_function()
  #     #subset.browse_prio_cibersort.t <- unfactor(as.data.frame(t(browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),])))
  #     subset.browse_prio_cibersort.t_fisher_part_pre <- as.data.frame(t(browse_prio_cibersort[paste0(str_sub(input$data_source_prio,end=-19),"_",paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_")),]))
  #
  #     #colnames(subset.browse_prio_cibersort.t) <- paste0("p-value in ",sapply(str_split(colnames(subset.browse_prio_cibersort.t),"_",2),function(x) x[1]))
  #     if(nrow(subset.browse_prio_cibersort.t_fisher_part_pre)>0) {
  #       colnames(subset.browse_prio_cibersort.t_fisher_part_pre) <- as.character(as.data.frame(str_split(colnames(subset.browse_prio_cibersort.t_fisher_part_pre),"_",2))[1,])
  #       #paste0("p-value in ",str_sub(input$data_source_prio,end=-19))
  #       meta.p <- browse_prio_cibersort_meta_p[,paste0(paste(input_treatment_prio,collapse = "_"),"_",paste(input_description_prio,collapse = "_"))]
  #       meta.fdr <- p.adjust(meta.p)
  #       subset.browse_prio_cibersort.t_meta_part <- data.frame("meta-analysis p-value"=meta.p,
  #                                                              "FDR of meta-analysis p-value"=meta.fdr)
  #       combined.p <- fisher.method(subset.browse_prio_cibersort.t_fisher_part_pre,na.rm=TRUE)$p.value
  #       subset.browse_prio_cibersort.t_fisher_part <- data.frame("combined p-value"=combined.p,
  #                                                                "FDR of combined p-value"=p.adjust(combined.p,"fdr"),
  #                                                                subset.browse_prio_cibersort.t_fisher_part_pre,stringsAsFactors = F )
  #       subset.browse_prio_cibersort.t <- merge(subset.browse_prio_cibersort.t_meta_part,subset.browse_prio_cibersort.t_fisher_part,by="row.names",all=T)
  #       rownames(subset.browse_prio_cibersort.t) <- subset.browse_prio_cibersort.t[,1]
  #       subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[,-1]
  #       subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[order(subset.browse_prio_cibersort.t[,"meta.analysis.p.value"]),]
  #       subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[,-which(apply(subset.browse_prio_cibersort.t,2,function(x)all(is.na(x))))]
  #
  #       #subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t[order(subset.browse_prio_cibersort.t[,"combined.p.value"],subset.browse_prio_cibersort.t[,names(apply(subset.browse_prio_cibersort.t,2,function(x) length(na.omit(x)))[2])]),]
  #       brks <- quantile(subset.browse_prio_cibersort.t, probs = seq(.05, .95, .05), na.rm = TRUE)
  #       clrs <- round(seq(40, 255, length.out = length(brks)+1), 0) %>% {paste0("rgb(255,", ., ",", ., ")")}
  #     } else {
  #       subset.browse_prio_cibersort.t <- subset.browse_prio_cibersort.t_fisher_part_pre
  #     }
  #   })
  #   output$prio_cibersort_download <- downloadHandler(
  #     filename = function() {
  #       if (input$prio_cibersort_type == "Excel (CSV)") {
  #         "newfile.csv"
  #       } else if (input$prio_cibersort_type == "Text (TSV)") {
  #         "newfile.txt"
  #       } else if (input$prio_cibersort_type == "JSON") {
  #         "newfile.json"
  #       }
  #     },
  #     content = function(file) {
  #       if (input$prio_cibersort_type == "Excel (CSV)") {
  #         fwrite (subset.browse_prio_cibersort.t,row.names=T,file)
  #       } else if (input$prio_cibersort_type == "Text (TSV)") {
  #         fwrite (subset.browse_prio_cibersort.t,row.names=T,file)
  #         #write.table (selectedcibersort_cibersort(),file,quote=F, row.names = FALSE, sep="\t")
  #       } else if (input$prio_cibersort_type == "JSON") {
  #         write(toJSON(subset.browse_prio_cibersort.t),row.names=T,file)
  #       }
  #     }
  #   )
  #   output$prio_cibersort_table <- DT::renderDataTable({
  #     datatable(subset.browse_prio_cibersort.t,options = list(searching = TRUE,paging = TRUE,info=FALSE)) %>%
  #       formatStyle(names(subset.browse_prio_cibersort.t), backgroundColor = styleInterval(brks, clrs)) %>%
  #       formatSignif(columns = c(1:ncol(subset.browse_prio_cibersort.t)), digits = 2,getOption("scipen"))
  #   })
  # })

   

  ###########################################  
  #### mutation
  ###########################################
  
  
  # input <- list(data_source_mutation=c("Melanoma-Samstein (320 samples/+++)",
  # "Renal Cell Carcinoma-Miao (098 samples/-++"),
  # treatment_mutation="anti-PD-1/L1+anti-CTLA-4",
  # description_mutation=c("Pre","On","Unknown"),
  # gene_mutation=c("BRAF","KRAS","HRAS","NRAS","NF1"),
  # clinical_mutation="Pre- or On-treatment")
  
  
  observe({
    if (input$defined_gene_mutation=="user-defined list") {
      updateSelectizeInput(session,"gene_mutation",
                       choices=a.colnames,selected=c("BRAF","KRAS","HRAS","NRAS","NF1"),
                       server=TRUE)
    } else {
      selected_mutations <- switch (input$defined_gene_mutation,
                                    "DNA_damage_response" = DNA_damage_response,
                                    "PDL1_expression_and_PD1_checkpoint_pathway_in_cancer"=PDL1_expression_and_PD1_checkpoint_pathway_in_cancer,
                                    "T_cell_receptor_signaling_pathway"=T_cell_receptor_signaling_pathway,
                                    "B_cell_receptor_signaling_pathway"=B_cell_receptor_signaling_pathway,
                                    "Cell_cycle_control"=Cell_cycle_control,
                                    "Survival_cell_death_regulation_signaling"=Survival_cell_death_regulation_signaling)
      updateSelectizeInput(session,"gene_mutation",
                           choices=a.colnames,selected=selected_mutations,
                           server=TRUE)
    }
    
  })
  observe({
    if (input$data_source_mutation=="" || is.null(input$data_source_mutation) ||
        input$treatment_mutation=="" || is.null(input$treatment_mutation) ||
        input$description_mutation=="" || is.null(input$description_mutation) ||
        input$gene_mutation=="" || is.null(input$gene_mutation)){
      disable("do_mutation")
    } else {
      enable("do_mutation")
    }
  })
  
  selecteddatabase_mutation <- eventReactive(input$do_mutation,{
    data_source1 <- str_sub(input$data_source_mutation,end=-19)
    #treatment1 <- input$treatment_mutation
    treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_mutation])
    description1 <- input$description_mutation
    genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
    a<-data.table(database.test.process())
    m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
    m<-m[,colnames(m) %in% genes1,with=FALSE]
    as.data.frame(m)
  })
  
  selectedmutation_mutation <- function() {
    all.mutations <- all.mutations.process()
    input_gene_mutation <- input$gene_mutation
    selected.database<-selecteddatabase_mutation()
    if (nrow(selected.database)==0) {
      selected.database
    } else {
      gene1 <- input_gene_mutation
      selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"] ,]
      selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% gene1,]
      if (nrow(selected.mutations.1)>0 ) {
      #### important: resharp mutation file to database.test format
      b<-aggregate(variant_class~(PatientID+gene),data=selected.mutations.1,FUN=paste0)
      c <- b %>% spread(gene,variant_class,fill="no mutation")
      c[c=="NULL"]<-"no mutation"
      
      if ((ncol(c)-1) < length(gene1)) {
        c.add <- matrix("no mutation",nrow(c),length(gene1[gene1 %ni% colnames(c)[2:ncol(c)]]))
        colnames(c.add) <- gene1[gene1 %ni% colnames(c)[2:ncol(c)]]
        c <- cbind(c, c.add)
      }
      
      c.rest <- data.frame(PatientID=unique(selected.mutations[selected.mutations[,"PatientID"] %ni% selected.mutations.1[,"PatientID"],"PatientID"]))
      c <- rbind.fill(c,c.rest)
      c[c=="NULL"] <- "no mutation"
      c[is.na(c)] <- "no mutation"
      
      #c <- data.frame(c,category=apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") }))
      c <- data.frame(c,category="no mutation")
      c[,"category"] <- as.character(c[,"category"])
      c[!apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") == "no mutation" }),"category"]<- "mutated"
      
      selected.database <- merge(selected.database,c,by="PatientID",all.x=T)
      selected.database[is.na(selected.database[,"category"]),"category"] <- "unknown"
      as.data.frame(selected.database)
      } else {
        selected.database <- data.frame(selected.database,category="unknown")
        selected.database
      }
    }
  }
  
  observeEvent(input$do_mutation,{
    selected.database.pre <- selecteddatabase_mutation()
    selected.mutation.pre <- selectedmutation_mutation()
    ##########
    #### Summary
    ##########
    output$mutation_summary_disease_number <- renderText({
      input$do_mutation ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selected.database.pre$Disease))})})
    output$mutation_summary_sample_number <- renderText({
      input$do_mutation ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selected.database.pre)})})
    output$mutation_summary_plots <- renderPlotly({
      summary_plot(inputdata=selected.mutation.pre,feature="mutation")
    })
    

     observeEvent(input$do_mutation,{
       input$do_mutation
      isolate({
        # selected.database <- selectedmutation_mutation()
        selected.mutation.pre$treatment <- as.factor(selected.mutation.pre$treatment)
        selected.mutation.pre$description <- as.factor(selected.mutation.pre$description)
        selected.mutation.pre$RECIST <- as.factor(selected.mutation.pre$RECIST)
        selected.mutation.pre$Gender <- as.factor(selected.mutation.pre$Gender)
        if (nrow(selected.mutation.pre)==0 ) {
          output[["mutation_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="Warning 1: no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.mutation <- data.frame(selected.mutation.pre,Groups=selected.mutation.pre$PFS.status)
          
          splitted_list <- selected.mutation %>% split(.$data_source) #selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_mutation,end=-19)]
          for (k in 1:length(input$data_source_mutation)) {
            local({
              my_i <- k
              mutation_summary_clinical_plots_name <- paste0("mutation_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[mutation_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_mutation]]) & d[,"category"]!="unknown",])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_mutation]]) & d[,"category"]!="unknown",]
                      if(input$clinical_mutation=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      plot.data <- as.data.frame(table(d[,c(clinical_server[input$clinical_mutation],"category")]))
                      colnames(plot.data) <- c("clinical","category","numbers")
                      p<-plot_ly(plot.data,type="bar",x=~clinical,y=~numbers,
                                 color=~category,
                                 colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                 legendgroup=~category) %>% 
                        layout(legend=list(font=list(size=15)),
                               title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title="Clinical Attribute",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title="Sample number",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               ))
                      incProgress(100, detail = "Doing part %")
                      p
                      
                    } 
                  }
                })
              })
            })
          }
        }
      }) 
     })
    
    
    
    
    output$mutation_summary_download <- downloadHandler(
      filename = function() {
        if (input$mutation_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$mutation_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$mutation_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        if (input$mutation_summary_type == "Excel (CSV)") {
          fwrite (selected.mutation.pre,file)
        } else if (input$mutation_summary_type == "Text (TSV)") {
          fwrite (selected.mutation.pre,file)
          #write.table (selectedmutation_mutation(),file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$mutation_summary_type == "JSON") {
          write(toJSON(selected.mutation.pre),file)
        }
      }
    )
    output$mutation_summary_table <- DT::renderDataTable({
      input$do_mutation
      isolate({
        #selected.database <- selectedmutation_mutation()
        selected.mutation.pre$category <- as.character(selected.mutation.pre$category)
        selected.mutation.pre
      })
    })

  ############
  #### mutation plot
  ############

  observeEvent(input$do_mutation,{
    input$do_mutation
      isolate({
      input_gene_mutation <- input$gene_mutation
      selected.database <- selected.database.pre#selecteddatabase_mutation()
      all.mutations <- all.mutations.process()
      selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
      selected.mutations.function <- selected.mutation.pre#selectedmutation_mutation()
      #selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_mutation,]
      if (nrow(selected.database)==0 | nrow(selected.mutations)==0) {
        
        output$mutation_plots_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })
        
        output[["mutation_plots_plots0"]] <- renderPlotly({
          plot_ly () %>% 
            layout(title = paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
        })
        
        # output$mutation_plots_plots_combined_p <- renderText({
        #   "NA"
        # })
        # output$mutation_plots_plots_single_p <- renderText({
        #   "NA"
        # })
        
      } else {
        selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_mutation,]
        
        if (nrow(selected.mutations.1)==0) {
          
          output$mutation_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          
          output[["mutation_plots_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$mutation_plots_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_plots_plots_single_p <- renderText({
          #   "NA"
          # })
          
          
        } else {
        selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
        splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list <- splitted_list[str_sub(input$data_source_mutation,end=-19)]
        #      splitted_list_mutation <- selected.mutations.1 %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list_mutation <- selected.mutations %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list_mutation <- splitted_list_mutation[str_sub(input$data_source_mutation,end=-19)]
        splitted_list_mutation.function <- selected.mutations.function %>% split(.$data_source)
        splitted_list_mutation.function <- splitted_list_mutation.function[str_sub(input$data_source_mutation,end=-19)]

        #mutation_combined_p <- NULL
        mutation_plots_plots_metadata <- NULL
         for (k in 1:length(input$data_source_mutation)) {
        #   #local({ ## can't use local
           my_i <- k
           d<-splitted_list_mutation.function[[my_i]]
           
           if (is.na(names(splitted_list_mutation.function)[my_i])) {
             tmp <- data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
           } else {
             if (nrow(d)==0) {
               tmp = data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
             } else {
               #d <- d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown" & !is.na(d[,"category"]),]
               if (nrow(d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown" & !is.na(d[,"category"]),])==0) {
                 tmp = data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
               } else {
                 d$category_binary <- NA
                 d$response_binary <- NA
                 d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                 d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                 d[d$category!="unknown" & d$category=="no mutation","category_binary"] <- 0
                 d[d$category!="unknown" & d$category=="mutated","category_binary"] <- 1
                 d$scaled <- scale(as.numeric(d$category_binary))
                 if (all(is.na(d$scaled))) {
                   tmp = data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
                 } else {
                   tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                   if (nrow(tmp.pre$coefficients)>1) {
                     tmp <- data.frame(data_source=input$data_source_mutation[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                   } else {
                     tmp = data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
                   }
                 }
                 
                 
               }
             }
           }
           mutation_plots_plots_metadata <- rbind.fill(mutation_plots_plots_metadata,tmp)
           
           
           # my_i <- k
           # d<-splitted_list[[my_i]]
           # d.mutations <- splitted_list_mutation[[my_i]]
           # d.mutations.function <- splitted_list_mutation.function[[my_i]]
           # d.mutations.function.res <- d.mutations.function[d.mutations.function[,"PFS.status"]=="response" & !is.na(d.mutations.function[,"PFS.status"]),]
           # d.mutations.function.nonres <- d.mutations.function[d.mutations.function[,"PFS.status"]=="nonresponse" & !is.na(d.mutations.function[,"PFS.status"]),]
           # if (nrow(d.mutations.function.res)>1 & nrow(d.mutations.function.nonres)>1) {
           #   tmp <- data.frame(data_source = unique(d.mutations.function$data_source),
           #                     event.res=nrow(d.mutations.function.res[d.mutations.function.res$category=="mutated",]),
           #                     n.res=nrow(d.mutations.function.res),
           #                     event.nonres=nrow(d.mutations.function.nonres[d.mutations.function.nonres$category=="mutated",]),
           #                     n.nonres=nrow(d.mutations.function.nonres))
           #   mutation_plots_plots_metadata <- rbind(mutation_plots_plots_metadata,tmp)
           # } else {
           #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
           #                     event.res=NA,
           #                     n.res=NA,
           #                     event.nonres=NA,
           #                     n.nonres=NA)
           #   mutation_plots_plots_metadata <- rbind(mutation_plots_plots_metadata,tmp)
           # }
         }
        
        if (all(is.na(mutation_plots_plots_metadata))) {
          output$mutation_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
        } else {
          output$mutation_plots_plots_meta <- renderPlot({
            
            # forest(metabin(event.res,n.res,event.nonres,n.nonres,
            #                 data=mutation_plots_plots_metadata,
            #                 label.e = "response",label.c = "nonresponse",
            #                 studlab = data_source),
            #        test.overall = TRUE,ff.study.label = "bold")
            m1 <- metagen(TE,seTE,
                          data=mutation_plots_plots_metadata,
                          sm="OR",
                          #label.e = "response",label.c = "nonresponse",
                          studlab = str_sub(data_source,end=-19))
            m1$upper[exp(m1$upper)>1e+50] <- 1000
            m1$upper.fixed[exp(m1$upper.fixed)>1e+50] <- 1000
            m1$upper.random[exp(m1$upper.random)>1e+50] <- 1000
            m1$lower.fixed[exp(m1$lower.fixed)>1e+50] <- 1000
            m1$lower.random[exp(m1$lower.random)>1e+50] <- 1000
            
            forest(m1,layout="RevMan5",    
                   hetstat=TRUE,
                   plotwidth="3inch",test.overall = TRUE,
                   xlim=c(0.1,10),backtransfer=T,
                   label.test.overall.random="random effects model: ",
                   label.test.overall.fixed="fixed effects model: ",
                   fs.test.overall = 15,
                   scientific.pval = TRUE,ff.study.label = "bold")
            # forest(metagen(TE,seTE,
            #                data=mutation_plots_plots_metadata,
            #                sm="OR",
            #                #label.e = "response",label.c = "nonresponse",
            #                studlab = data_source),
            #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
          })
        }
        #   isolate({
        #   if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_mutation)[my_i])) {
        #     mutation_combined_p <- c(mutation_combined_p,NA)
        #     
        #   } else {
        #     if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d.mutations)==0) {
        #       mutation_combined_p <- c(mutation_combined_p,NA)
        #       
        #     } else {
        #       
        #       d.mutations.function <- d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & d.mutations.function[,"category"]!="unknown",]
        #       d.table <- unlist(table(d.mutations.function[,c("PFS.status","category")]))
        #       d.table <- d.table[rownames(d.table) %in% c("nonresponse","response"),colnames(d.table) %in% c("mutated","no mutation")]
        #       try.test <- try(p.v <- fisher.test(d.table)$p.value)
        #       if (class(try.test)=="try-error" | is.na(try.test)) {
        #         mutation_combined_p <- c(mutation_combined_p,NA)
        #         
        #       } else {
        #         mutation_combined_p <- c(mutation_combined_p,p.v)
        #       }
        #     }
        #   }
        #   })
        # }
        # 
        # if(all(is.na(mutation_combined_p))) {
        #   output$mutation_plots_plots_combined_p <- renderText({
        #     "NA"
        #   })
        #   output$mutation_plots_plots_single_p <- renderText({
        #     "NA"
        #   })
        # } else {
        #   output$mutation_plots_plots_combined_p <- renderText({
        #     #paste0(mutation_combined_p,collapse = ",")
        #     signif(fisher.method(rbind(a=na.omit(mutation_combined_p)))$p.value,2)
        #   })
        #   output$mutation_plots_plots_single_p <- renderText({
        #     paste0(signif(mutation_combined_p,2),collapse = ", ")
        #     #signif(fisher.method(rbind(a=mutation_combined_p),na.rm=TRUE)$p.value,3)
        #   })
        # }
              
              
              
              
              
        for (k in 1:length(input$data_source_mutation)) {
          local({
            my_i <- k
            mutation_plots_plots_name <- paste0("mutation_plots_plots", my_i)
            d<-splitted_list[[my_i]]
            d.mutations <- splitted_list_mutation[[my_i]]
            d.mutations.function <- splitted_list_mutation.function[[my_i]]
            output[[mutation_plots_plots_name]] <- renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
              isolate({
                if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_mutation)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>% 
                    layout(title = str_sub(input$data_source_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  #if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d[d$PFS.status!="Unknown",])==0 | nrow(d.mutations)==0) {
                  if (nrow(d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & 
                                                d.mutations.function[,"PFS.status"]!="Unknown" & 
                                                d.mutations.function[,"category"]!="unknown" &
                                                !is.na(d.mutations.function[,"category"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
#                    withProgress(message = 'Making plot', value = 0, {
                    d.mutations.function <- d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & 
                                                                   d.mutations.function[,"PFS.status"]!="Unknown" & 
                                                                   d.mutations.function[,"category"]!="unknown" &
                                                                   !is.na(d.mutations.function[,"category"]),]
                    # d<-data.frame(d,IND="no mutation")
                    # d[,"IND"]<-as.character(d[,"IND"])
                    # for (j in 1:length(input_gene_mutation)) {
                    #   d[d[,"PatientID"] %in% d.mutations[d.mutations[,"gene"]==input_gene_mutation[j],"PatientID"],"IND"]<- paste0("mutation in ",input_gene_mutation[j])
                    # }
                    d.table <- unlist(table(d.mutations.function[,c("PFS.status","category")]))
                    d.table <- d.table[rownames(d.table) %in% c("nonresponse","response"),colnames(d.table) %in% c("mutated","no mutation")]
                    try.test <- try(p.v <- fisher.test(d.table)$p.value)
                    #try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")]))[1:2,])$p.value)
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      p.v <- " = NA"
                    } else {
                      # p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")])))$p.value
                      # if (p.v < 0.001) {
                      #   p.v <- " < 0.001"
                      # } else {
                        p.v<-paste0(" = ",signif(p.v,2))
                        
                      # } 
                    }
                    
                    #p.hist <- plot_ly(d.mutations.function,x=~PFS.status,y=~category,histfunc="percent",type="histogram")
                    # plot.data <- as.data.frame(table(d.mutations.function[,c("PFS.status","category")]))
                    # colnames(plot.data) <- c("clinical","category","numbers")
                    # p.bar<-plot_ly(plot.data,type="bar",x=~clinical,y=~numbers,
                    #            color=~category,
                    #            colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                    #            legendgroup=~category) %>% 
                    #   layout(legend=list(font=list(size=15)),
                    #          title = unique(d.mutations.function$data_source),titlefont=list(size=20,color="black",family= "Aril"),
                    #          height = 480,
                    #          xaxis=list(title=" ",
                    #                     titlefont=list(size=20,color="black",family= "Aril black"),
                    #                     showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    #          yaxis=list(title="Sample number",
                    #                     titlefont=list(size=20,color="black",family= "Aril black"),
                    #                     showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                    #          ))
                    
                    
                    ind <- sort(unique(d[,"Groups"]))
                    d.mutations.res <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[2] & !is.na(d[,"Groups"])),"PatientID"] ,]
                    d.mutations.nonres <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[1] & !is.na(d[,"Groups"])),"PatientID"] ,]
                    
                    d.mutations.res<-d.mutations.res[,-1]
                    #d.mutations.res<-d.mutations.res[,c("PatientID","gene","variant_class","Protein_change","Chromosome","data_source")]
                    d.mutations.res<-d.mutations.res[,c(6,2:5,7)]
                    colnames(d.mutations.res)[1] <- "sample"
                    if (nrow(d.mutations.res[d.mutations.res[,"gene"] %in% input_gene_mutation,])==0) {
                      p1 <- ggplot(d.mutations.res, aes(x=sample,y="NA")) + 
                        xlab(paste0("response samples (n=",length(unique(d.mutations.res$sample)),")")) + 
                        theme(axis.text.x=element_text(angle=50, hjust=1,size=4), 
                              axis.text.y=element_text(size=8,colour='black'),
                              axis.title.x=element_text(size=11),
                              axis.title.y=element_blank())
                      p.bar.res <- data.frame(PFS.status="response",genes=input_gene_mutation,Number=0,n=length(unique(d.mutations.res$sample)))
                      p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                      # inputDat <- waterfall_qual(d.mutations.res, clinData, mutBurden, file_type="Custom",
                      #                            label_col=NULL)
                      # data_frame <- inputDat[[1]]
                      # p2 <- ggplot(na.omit(data_frame),aes_string(x=NA, y=0, fill='trv_type')) +
                      #   geom_bar(position='stack', alpha=.75, width=1, stat='identity') + 
                      #   theme_bw() + coord_flip() + 
                      #   theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.title.y=element_blank(),legend.position=('none'),
                      #         axis.text.x=element_blank(),axis.ticks.x=element_blank(),axis.title.x=element_blank()) + 
                      #   ylab("% mutated samples") + ylim(100,0)+ geom_blank()
                      # p3 <- subplot(style(p2,showlegend=T),style(p1,showlegend=T),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                      
                      # p3 <- plot_ly () %>% 
                      #   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      p1<-#ggplotly(
                        waterfall(d.mutations.res, fileType = "Custom",
                                  variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))), ### here all.mutation is important, if not, different data source will have different color enve in the same mutation type
                                  plotGenes=unlist(str_split(input_gene_mutation,",")),
                                  mainXlabel = TRUE,my_x_label=ind[2],
                                  plotMutBurden = FALSE,returnp="p1",
                                  mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                      Number.genes.res <- NULL
                      for (j in 1:length(input_gene_mutation)) {
                        Number.genes.res <- c(Number.genes.res,length(unique(d.mutations.res[d.mutations.res[,"gene"]==input_gene_mutation[j],"sample"])))
                      }
                      p.bar.res <- data.frame(PFS.status="response",genes=input_gene_mutation,Number=Number.genes.res,n=length(unique(d.mutations.res$sample)))
                      p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                      #height =  480 )
                      # p2<-#ggplotly(
                      #   waterfall(d.mutations.res, fileType = "Custom",
                      #             variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                      #             plotGenes=unlist(str_split(input_gene_mutation,",")),
                      #             mainXlabel = TRUE,my_x_label=ind[2],
                      #             plotMutBurden = FALSE,returnp="p2",xticks=FALSE,
                      #             mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                      #                           '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                      #                           '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                      #                           '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                      #                           '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                      #                           '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                      #                           '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                      #                           '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                      # #height =  480 )
                      # #                 tooltip = c("x","y","text"),legendgroup=~variant_class)
                      # p3 <- subplot(style(p2,showlegend=T),style(p1,showlegend=T),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                    }
                    d.mutations.nonres<-d.mutations.nonres[,-1]
                    d.mutations.nonres<-d.mutations.nonres[,c(6,2:5,7)]
                    colnames(d.mutations.nonres)[1] <- "sample"
                    if (nrow(d.mutations.nonres[d.mutations.nonres[,"gene"] %in% input_gene_mutation,])==0) {
                      q1 <- ggplot(d.mutations.nonres, aes(x=sample,y="NA")) + 
                        xlab(paste0("nonresponse samples (n=",length(unique(d.mutations.nonres$sample)),")")) + 
                        theme(axis.text.x=element_text(angle=50, hjust=1,size=4), 
                              axis.text.y=element_text(size=8,colour='black'),
                              axis.title.x=element_text(size=11),
                              axis.title.y=element_blank())
                      p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_mutation,Number=0,n=length(unique(d.mutations.nonres$sample)))
                      p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                      # inputDat <- waterfall_qual(d.mutations.nonres, clinData, mutBurden, file_type="Custom",
                      #                            label_col=NULL)
                      # data_frame <- inputDat[[1]]
                      # q2 <- ggplot(na.omit(data_frame),aes_string(x=NA, y=0, fill='trv_type')) +
                      #   geom_bar(position='stack', alpha=.75, width=1, stat='identity') + 
                      #   theme_bw() + coord_flip() + 
                      #   theme(axis.text.y=element_blank(),axis.ticks.y=element_blank(),axis.title.y=element_blank(),legend.position=('none')) +
                      #   
                      #   ylab("% mutated samples") + ylim(100,0)+ geom_blank()
                      # q3 <- subplot(style(q2,showlegend=F),style(q1,showlegend=F),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                      # 
                      # q3 <- plot_ly () %>% 
                      #   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      q1<-#ggplotly(
                        waterfall(d.mutations.nonres, fileType = "Custom",
                                  variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                                  plotGenes=unlist(str_split(input_gene_mutation,",")),
                                  mainXlabel = TRUE,my_x_label=ind[1],
                                  plotMutBurden = FALSE,returnp="p1",
                                  mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                      Number.genes.nonres <- NULL
                      for (j in 1:length(input_gene_mutation)) {
                        Number.genes.nonres <- c(Number.genes.nonres,length(unique(d.mutations.nonres[d.mutations.nonres[,"gene"]==input_gene_mutation[j],"sample"])))
                      }
                      p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_mutation,Number=Number.genes.nonres,n=length(unique(d.mutations.nonres$sample)))
                      p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                      #height =  480 )
                      # q2<-#ggplotly(
                      #   waterfall(d.mutations.nonres, fileType = "Custom",
                      #             variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                      #             plotGenes=unlist(str_split(input_gene_mutation,",")),
                      #             mainXlabel = TRUE,my_x_label=ind[1],
                      #             plotMutBurden = FALSE,returnp="p2",xticks=TRUE,
                      #             mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                      #                           '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                      #                           '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                      #                           '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                      #                           '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                      #                           '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                      #                           '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                      #                           '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                      # #height =  480 )
                      # #                     tooltip = c("x","y","text"),legendgroup=~variant_class)
                      # q3 <- subplot(style(q2,showlegend=F),style(q1,showlegend=F),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                    }
                    
                    # if (nrow(d.mutations.res)==0 | nrow(d.mutations.nonres)==0) {
                    #   lable_x <- NULL
                    #   side_x <- "bottom"
                    # } else {
                    #   lable_x <- paste0("p-value",p.v)
                    #   side_x <- "top"                
                    # }
                    p.bar.res <- data.frame(p.bar.res,perc=signif(p.bar.res$Number/p.bar.res$n*100,3))
                    p.bar.nonres <- data.frame(p.bar.nonres,perc=signif(p.bar.nonres$Number/p.bar.nonres$n*100,3))
                    
                    if (sum(p.bar.res$perc)==0 & sum(p.bar.nonres$perc)!=0 ) {
                      p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                      p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                    } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)==0) {
                      p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                      p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                    } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)!=0) {
                      p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                      p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                    } else {
                      p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                      p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                    }
                    #p.bar <- rbind(p.bar.res,p.bar.nonres)
                    p.bar <- rbind(p.bar.nonres,p.bar.res)
                    p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
                    bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
                      theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
                      geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                      #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                      geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))
                    
                    
                    
                    # p.bar <- rbind(p.bar.res,p.bar.nonres)
                    # p.bar <- data.frame(p.bar,perc=signif(p.bar$Number/p.bar$n*100,3))
                    # 
                    # if (sum(p.bar$perc)==0) {
                    #   p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
                    #   bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
                    #     theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
                    #     #geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                    #     #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                    #     geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))
                    # } else {
                    #   p.bar <- p.bar[p.bar[,"perc"]!=0,]
                    #   p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
                    #   
                    #   bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
                    #     theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
                    #     geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                    #     #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                    #     geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))
                    # }
                    # bar.plot <- plot_ly(p.bar,x = p.bar$PFS.status, y = p.bar$perc, name=p.bar$genes,type = 'bar',
                    #         color = p.bar$genes,text=paste0(p.bar$genes, " (",signif(p.bar$perc,3),")"),textposition="middle",hoverinfo = 'text') %>% 
                    #   layout(yaxis = list(title = 'Percentage (%)'),barmode = "stack") %>% 
                    #   add_annotations(text=paste0(p.bar$genes, " (",signif(p.bar$perc,3),")"),x=p.bar$PFS.status,y=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2),showarrow=FALSE) %>%
                    #   add_annotations(text=paste0("p-value",p.v),x=0.5,y=p.bar[which.max(p.bar[,"perc"]),"perc"],
                    #                   font=list(size=25,color="black",family= "Aril black"),showarrow=FALSE)
                    # 
                    # bar.plot <- plot_ly(p.bar,x = ~PFS.status, y = ~perc*100, type = 'bar',color = ~genes,text=~Number,hoverinfo = 'text') %>% 
                    #   layout(yaxis = list(title = 'Percentage (%)'),barmode = "stack") %>%
                    #   
                    incProgress(100, detail = "Doing part %")
                    subplot(style(bar.plot,showlegend=F),
                            subplot(p1,q1,nrows=2,margin=0.1,titleX=TRUE,titleY=TRUE) %>% 
                                layout(#title = as.character(unique(d$data_source)), 
                                       titlefont=list(size=20,color="black",family= "Aril"),
                                       legend=list(font=list(size=10))),
                                       #xaxis=list(title=lable_x,side=side_x,titlefont=list(size=15))),
                            widths=c(1/3,2/3),margin=0.05,titleX=TRUE,titleY=TRUE) %>% 
                      layout(title=as.character(unique(d$data_source)))
#                  })
                    } 
                  
                }
              })
            })
          })
          })
        }
        
      }
      }
        

    })
    
  })
  ##############
  #### mutation survival
  ##############
  
  observeEvent(input$do_mutation,{
    input$do_mutation
    isolate({
      input_gene_mutation <- input$gene_mutation
      selected.database <- selected.mutation.pre#selectedmutation_mutation()
      input_adjust_OS_mutation <- input$adjust_OS_mutation
      # selected.database <- selecteddatabase_mutation()
      # selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
      # selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_mutation,]
#      if (nrow(selected.database)==0 | nrow(selected.mutations.1)==0) {
      if (nrow(selected.database)==0 | nrow(selected.database[selected.database[,"category"]!="unknown",])==0) {

        output$mutation_survival_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })

        output[["mutation_survival_plots0"]] <- renderPlotly({
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
        })

        # output$mutation_survival_plots_combined_p <- renderText({
        #   "NA"
        # })
        # output$mutation_survival_plots_single_p <- renderText({
        #   "NA"
        # })

      } else {
        selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
        splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
        splitted_list <- splitted_list[str_sub(input$data_source_mutation,end=-19)]
        # splitted_list_mutation <- selected.mutations.1 %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        # splitted_list_mutation <- splitted_list_mutation[input$data_source_mutation]


        # mutation_survival_combined_p <- NULL
        mutation_survival_plots_metadata <- NULL
        for (k in 1:length(input$data_source_mutation)) {
          isolate({
            my_i <- k
            d<-splitted_list[[my_i]]
            if (is.na(names(splitted_list)[my_i]) ) {
              #mutation_survival_combined_p <- c(mutation_survival_combined_p,NA)
              tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                event.top=NA,
                                n.top=NA,
                                event.bottom=NA,
                                n.bottom=NA)
              mutation_survival_plots_metadata <- rbind(mutation_survival_plots_metadata,tmp)

            } else {
              if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                # mutation_survival_combined_p <- c(mutation_survival_combined_p,NA)
                tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                  event.top=NA,
                                  n.top=NA,
                                  event.bottom=NA,
                                  n.bottom=NA)
                mutation_survival_plots_metadata <- rbind(mutation_survival_plots_metadata,tmp)

              } else {
                d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                if (length(unique(d$category))==1) {
                  # mutation_survival_combined_p <- c(mutation_survival_combined_p,NA)
                  tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                    event.top=NA,
                                    n.top=NA,
                                    event.bottom=NA,
                                    n.bottom=NA)
                  mutation_survival_plots_metadata <- rbind(mutation_survival_plots_metadata,tmp)

                } else {
                  d<-data.frame(d,dead=NA)
                  d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                  d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                  d <- d[!is.na(d[,"dead"]),]
                  d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                  d <- data.frame(data_source=as.character(d$data_source),
                                  dead=as.numeric(d$dead),
                                  OS=as.numeric(d$OS),
                                  Age=as.character(d$Age),
                                  Gender=as.character(d$Gender),
                                  category=as.character(d[,"category"]))
                  #d<-data.frame(d[,c(1:6)],unfactor(d[,c("dead","OS","PFS","Age","Gender")]),category=as.character(d[,"category"]))
                  #d$OS<-as.numeric(d$OS)
                  d<-data.frame(d,IND=paste0("no mutation (",
                                             nrow(d[d[,"category"]=="no mutation",]),")"))
                  d[,"IND"]<-as.character(d[,"IND"])
                  d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                              nrow(d[d[,"category"]=="mutated",]),")")
                  d<-d[order(d[,"IND"]),]

                  tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                    event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                    n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                    event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                    n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                  mutation_survival_plots_metadata <- rbind(mutation_survival_plots_metadata,tmp)

                  # if (input_adjust_OS_mutation=="None") {
                  #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                  #   if (class(diff.result.try)=="try-error") {
                  #     p.value <- NA
                  #   } else {
                  #     p.value <- signif(summary(diff.result)$logtest[3],2)
                  #   }
                  # } else {
                  #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_mutation),data=d) )
                  #   if (class(diff.result.adjust.try)=="try-error") {
                  #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                  #
                  #   } else {
                  #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                  #   }
                  # }
                  # # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                  # # plot.result <- ggsurv(fit.result,surv.col = c("#7CDBD5","#DCB239"))
                  # # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                  # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                  # mutation_survival_combined_p <- c(mutation_survival_combined_p,p.value)
                }
              }
            }
          })
        }
        if(all(is.na(mutation_survival_plots_metadata))) {
           output$mutation_survival_plots_meta <- renderPlot({
             plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
             abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
             text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
           })
        #   output$mutation_survival_plots_combined_p <- renderText({
        #     "NA"
        #   })
        #   output$mutation_survival_plots_single_p <- renderText({
        #     "NA"
        #   })
        } else {

          output$mutation_survival_plots_meta <- renderPlot({
            meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                             data=mutation_survival_plots_metadata,sm="OR",
                             studlab = data_source)
            # meta2 <- metagen(meta1$TE,meta1$seTE,sm="OR",
            #                  studlab=mutation_survival_plots_metadata$data_source)
            # forest(meta2,test.overall= T,ff.study.label = "bold")
            meta2 <- data.frame(data_source=meta1$studlab,
                                TE=meta1$TE,
                                seTE=meta1$seTE)
            m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source)
            forest(m1,layout="RevMan5",    
                   hetstat=TRUE,
                   plotwidth="3inch",test.overall = TRUE,
                   xlim=c(0.1,10),backtransfer=T,
                   label.test.overall.random="random effects model: ",
                   label.test.overall.fixed="fixed effects model: ",
                   fs.test.overall = 15,
                   scientific.pval = TRUE,ff.study.label = "bold")
            # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
            #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
          })

          # output$mutation_survival_plots_combined_p <- renderText({
          #   #paste0(mutation_survival_combined_p,collapse = ",")
          #   signif(fisher.method(rbind(a=na.omit(mutation_survival_combined_p)))$p.value,2)
          #   #paste0(signif(fisher.method(rbind(a=mutation_survival_combined_p),na.rm=TRUE)$p.value,2)," (",paste0(signif(mutation_survival_combined_p,2),collapse = ","),").")
          # })
          # output$mutation_survival_plots_single_p <- renderText({
          #   paste0(signif(mutation_survival_combined_p,2),collapse = ", ")
          #   #signif(fisher.method(rbind(a=mutation_survival_combined_p,b=mutation_survival_combined_p),na.rm=TRUE)$p.value[1],3)
          #   #paste0(signif(fisher.method(rbind(a=mutation_survival_combined_p),na.rm=TRUE)$p.value,2)," (",paste0(signif(mutation_survival_combined_p,2),collapse = ","),").")
          # })
         }






        for (k in 1:length(input$data_source_mutation)) {
          local({
            my_i <- k
            mutation_survival_plots_name <- paste0("mutation_survival_plots", my_i)
            mutation_survival_summary_tables_name <- paste0("mutation_survival_summary_tables", my_i)
            d<-splitted_list[[my_i]]
            # d.mutations <- splitted_list_mutation[[my_i]]
             #output[[mutation_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
              isolate({
#                if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_mutation)[my_i])) {
               if (is.na(names(splitted_list)[my_i]) ) {
                 incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                } else {
                  if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
#                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) ,]
                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                    if (length(unique(d$category))==1) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="There is only 1 group under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                    d<-data.frame(d,dead=NA)
                    d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                    d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                    d <- d[!is.na(d[,"dead"]),]
                    d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                    d <- data.frame(data_source=as.character(d$data_source),
                                    dead=as.numeric(d$dead),
                                    OS=as.numeric(d$OS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    category=as.character(d[,"category"]))
                    #d<-data.frame(d[,c(1:6)],unfactor(d[,c("dead","OS","PFS","Age","Gender")]),category=as.character(d[,"category"]))
                    #d$OS<-as.numeric(d$OS)
                    d<-data.frame(d,IND=paste0("no mutation (",
                                               nrow(d[d[,"category"]=="no mutation",]),")"))
                    # d<-data.frame(d,IND=paste0("samples without mutation (",
                    #                            nrow(d[d[,"category"]=="no mutation",]),")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                nrow(d[d[,"category"]=="mutated",]),")")
                    # d[d[,"category"]=="mutated","IND"]<- paste0("samples with mutation (",
                    #                                                 nrow(d[d[,"category"]=="mutated",]),")")
                    d<-d[order(d[,"IND"]),]

                    fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                      if (input_adjust_OS_mutation=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_mutation),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_mutation," was observed)")

                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_mutation
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }


#                     diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
# #                    fit.result <- survfit(Surv(OS,dead)~category,data=d)
# #                    plot.result <- ggsurv(fit.result,surv.col = c("lightgreen","plum"))
# #                    diff.result <- survdiff(Surv(OS,dead)~category,data=d)
#                     p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                             #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=median(d$OS),y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                    incProgress(100, detail = "Doing part %")
                    p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                    #ggplotly(plot.result,height = 480 ) %>%
                      layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                             yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                             legend=list(orientation="v",x=0.6,y=1))
                    }

                    }

                  }

                }
              })
              })
            #})
               output[[mutation_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[mutation_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
          })
        }
      }
    })
  })
  
  ##############
  #### mutation PFS
  ##############
  observeEvent(input$do_mutation,{
    input$do_mutation
    isolate({
      input_gene_mutation <- input$gene_mutation
      selected.database <- selected.mutation.pre#selectedmutation_mutation()
      # selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
      # selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_mutation,]
      input_adjust_OS_mutation <- input$adjust_OS_mutation
      if (nrow(selected.database)==0 | nrow(selected.database[selected.database[,"category"]!="no data available",])==0) {
        output$mutation_PFS_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })
        output[["mutation_PFS_plots0"]] <- renderPlotly({
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
        })
        # output$mutation_PFS_plots_combined_p <- renderText({
        #   "NA"
        # })
        # output$mutation_PFS_plots_single_p <- renderText({
        #   "NA"
        # })

      } else {
        selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
        splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
        splitted_list <- splitted_list[str_sub(input$data_source_mutation,end=-19)]

        # mutation_PFS_combined_p <- NULL
        mutation_PFS_plots_metadata <- NULL
        for (k in 1:length(input$data_source_mutation)) {
          isolate({
            my_i <- k
            d<-splitted_list[[my_i]]
                if (is.na(names(splitted_list)[my_i])) {
                  # mutation_PFS_combined_p <- c(mutation_PFS_combined_p,NA)
                  tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                    event.top=NA,
                                    n.top=NA,
                                    event.bottom=NA,
                                    n.bottom=NA)
                  mutation_PFS_plots_metadata <- rbind(mutation_PFS_plots_metadata,tmp)

                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                    # mutation_PFS_combined_p <- c(mutation_PFS_combined_p,NA)
                    tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                      event.top=NA,
                                      n.top=NA,
                                      event.bottom=NA,
                                      n.bottom=NA)
                    mutation_PFS_plots_metadata <- rbind(mutation_PFS_plots_metadata,tmp)

                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                    if (length(unique(d$category))==1) {
                      # mutation_PFS_combined_p <- c(mutation_PFS_combined_p,NA)
                      tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      mutation_PFS_plots_metadata <- rbind(mutation_PFS_plots_metadata,tmp)

                    } else {
                      d <- data.frame(data_source=as.character(d$data_source),
                                      PFS.status=as.character(d$PFS.status),
                                      PFS=as.numeric(d$PFS),
                                      Age=as.character(d$Age),
                                      Gender=as.character(d$Gender),
                                      category=as.character(d[,"category"]))
                      d<-data.frame(d,IND=paste0("no mutation (",
                                                 nrow(d[d[,"category"]=="no mutation",]),")"))
                      d[,"IND"]<-as.character(d[,"IND"])
                      d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                  nrow(d[d[,"category"]=="mutated",]),")")
                      d <- data.frame(d,group.IND=0)
                      d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                      d <- d[order(d[,"IND"]),]

                      tmp <- data.frame(data_source =input$data_source_mutation[my_i],
                                        event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                        n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                        event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                        n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                      mutation_PFS_plots_metadata <- rbind(mutation_PFS_plots_metadata,tmp)

                      # if (input_adjust_OS_mutation=="None") {
                      #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                      #   if (class(diff.result.try)=="try-error") {
                      #     p.value <- NA
                      #   } else {
                      #     p.value <- signif(summary(diff.result)$logtest[3],2)
                      #   }
                      # } else {
                      #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_mutation),data=d) )
                      #   if (class(diff.result.adjust.try)=="try-error") {
                      #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                      #
                      #   } else {
                      #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                      #   }
                      # }
                      #
                      #
                      # # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                      # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                      # # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                      # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                      # mutation_PFS_combined_p <- c(mutation_PFS_combined_p,p.value)
                    }

                  }

                }
          })
        }
        if (all(is.na(mutation_PFS_plots_metadata))) {
          output$mutation_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          # output$mutation_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_PFS_plots_combinedsingle_p <- renderText({
          #   "NA"
          # })
        } else {
          output$mutation_PFS_plots_meta <- renderPlot({
            meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                             data=mutation_PFS_plots_metadata,sm="OR",
                             studlab = data_source)
            # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
            #                  studlab=mutation_PFS_plots_metadata$data_source)
            # forest(meta2,test.overall= T,ff.study.label = "bold")
            meta2 <- data.frame(data_source=meta1$studlab,
                                TE=meta1$TE,
                                seTE=meta1$seTE)
            m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source)
            forest(m1,layout="RevMan5",    
                   hetstat=TRUE,
                   plotwidth="3inch",test.overall = TRUE,
                   xlim=c(0.1,10),backtransfer=T,
                   label.test.overall.random="random effects model: ",
                   label.test.overall.fixed="fixed effects model: ",
                   fs.test.overall = 15,
                   scientific.pval = TRUE,ff.study.label = "bold")
            # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
            #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
          })
          # output$mutation_PFS_plots_combined_p <- renderText({
          #   #paste0(mutation_survival_combined_p,collapse = ",")
          #   signif(fisher.method(rbind(a=na.omit(mutation_PFS_combined_p)))$p.value,2)
          #   #paste0(signif(fisher.method(rbind(a=mutation_survival_combined_p),na.rm=TRUE)$p.value,2)," (",paste0(signif(mutation_survival_combined_p,2),collapse = ","),").")
          # })
          # output$mutation_PFS_plots_single_p <- renderText({
          #   paste0(signif(mutation_PFS_combined_p,2),collapse = ", ")
          #   #signif(fisher.method(rbind(a=mutation_survival_combined_p,b=mutation_survival_combined_p),na.rm=TRUE)$p.value[1],3)
          #   #paste0(signif(fisher.method(rbind(a=mutation_survival_combined_p),na.rm=TRUE)$p.value,2)," (",paste0(signif(mutation_survival_combined_p,2),collapse = ","),").")
          # })
        }



        for (k in 1:length(input$data_source_mutation)) {
          local({
            my_i <- k
            mutation_PFS_plots_name <- paste0("mutation_PFS_plots", my_i)
            mutation_PFS_summary_tables_name <- paste0("mutation_PFS_summary_tables", my_i)
            d<-splitted_list[[my_i]]
            #output[[mutation_PFS_plots_name]] <- renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
              isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no PFS data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                    if (length(unique(d$category))==1) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="There is only 1 group under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                    } else {
                    d <- data.frame(data_source=as.character(d$data_source),
                                    PFS.status=as.character(d$PFS.status),
                                    PFS=as.numeric(d$PFS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    category=as.character(d[,"category"]))
                    d<-data.frame(d,IND=paste0("no mutation (",
                                               nrow(d[d[,"category"]=="no mutation",]),")"))
                    # d<-data.frame(d,IND=paste0("samples without mutation (",
                    #                            nrow(d[d[,"category"]=="no mutation",]),")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                nrow(d[d[,"category"]=="mutated",]),")")
                    # d[d[,"category"]=="mutated","IND"]<- paste0("samples with mutation (",
                    #                                             nrow(d[d[,"category"]=="mutated",]),")")
                    d <- data.frame(d,group.IND=0)
                    d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                    d <- d[order(d[,"IND"]),]
                    fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                    plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))


                      if (input_adjust_OS_mutation=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_mutation),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_mutation," was observed)")

                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_mutation
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }

                    # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                    incProgress(100, detail = "Doing part %")
                    p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                      layout(title=as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                             yaxis=list(title="Preogression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                             legend=list(orientation="v",x=0.6,y=1))
                    }
                    }

                  }

                }
              })
            })
          #})
          output[[mutation_PFS_plots_name]] <- renderPlotly({
            p1
          })
          output[[mutation_PFS_summary_tables_name]] <- renderDataTable(
            table1
          )
          })
        }
      }
    })
  })
  

  })

  
  ###########################################
  #### mutation_loads
  ###########################################
  # input <- list(data_source_mutation_loads=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)" ),
  # treatment_mutation_loads=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
  # description_mutation_loads=c("Pre","On"),slider_mutation_loads_OS=0.5)
  
  observe({
    if (input$data_source_mutation_loads=="" || is.null(input$data_source_mutation_loads) ||
        input$treatment_mutation_loads=="" || is.null(input$treatment_mutation_loads) ||
        input$description_mutation_loads=="" || is.null(input$description_mutation_loads) ){
      disable("do_mutation_loads")
    } else {
      enable("do_mutation_loads")
    }
  })
  selecteddatabase_mutation_loads <- eventReactive(input$do_mutation_loads,{
    data_source1 <- str_sub(input$data_source_mutation_loads,end=-19)
    #treatment1 <- input$treatment_mutation_loads
    treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_mutation_loads])
    description1 <- input$description_mutation_loads
    genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","Mutation.Load")
    a<-data.table(database.test.process())
    m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1 ]
    m<-m[,colnames(m) %in% genes1,with=FALSE]
    m<-m[,Mutation.Load:=as.numeric(Mutation.Load)]
    m<-as.data.frame(m)
    if (nrow(m)>0 & ncol(m)>12) {
#      m <- data.frame(m,category.OS="no data available",category.PFS="no data available")
      m <- data.frame(m,category="unknown")
      m1 <- m %>% split(.$data_source) %>% 
        lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_expression," expression",sep="")
          #d[,"category.OS"] <- as.character(d[,"category.OS"])
          d[,"category"] <- as.character(d[,"category"])
          d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<= quantile(na.omit(d[,"Mutation.Load"]),input$slider_mutation_loads_OS),"category"] <- 
            paste0("bottom ",
                   input$slider_mutation_loads_OS*100,
                   "%")
          d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_mutation_loads_OS),"category"] <- 
            paste0("top ",
                   (1-input$slider_mutation_loads_OS)*100,
                   "%")
          d})
      m2<-ldply(m1,data.frame)
      m2<-m2[,-1]
      m2
    } else {
      m2<-m
      m2
    }
  })

  #########
  #### mutation loads Summary
  #########
  
  output$mutation_loads_summary_disease_number <- renderText({
    input$do_mutation_loads ##keep these two rows
    isolate({ ##keep these two rows
      length(unique(selecteddatabase_mutation_loads()$Disease))})})
  output$mutation_loads_summary_sample_number <- renderText({
    input$do_mutation_loads ##keep these two rows
    isolate({ ##keep these two rows
      nrow(selecteddatabase_mutation_loads())})})
  output$mutation_loads_summary_plots <- renderPlotly({
    summary_plot(inputdata=selecteddatabase_mutation_loads(),feature="mutation_loads")
  })
  

  observeEvent(input$do_mutation_loads,{
    input$do_mutation_loads
    isolate({
      selected.database <- selecteddatabase_mutation_loads()
      selected.database$treatment <- as.factor(selected.database$treatment)
      selected.database$description <- as.factor(selected.database$description)
      selected.database$RECIST <- as.factor(selected.database$RECIST)
      selected.database$Gender <- as.factor(selected.database$Gender)
      if (nrow(selected.database)==0 ) {
        output[["mutation_loads_summary_clincial_plots0"]] <- renderPlotly({
          plot_ly () %>% 
            layout(title = paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
        })
      } else {
        selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
        splitted_list <- selected.database %>% split(.$data_source) #as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list <- splitted_list[str_sub(input$data_source_mutation_loads,end=-19)]
        for (k in 1:length(input$data_source_mutation_loads)) {
          local({
            my_i <- k
            mutation_loads_summary_clinical_plots_name <- paste0("mutation_loads_summary_clinical_plots", my_i)
            d<-splitted_list[[my_i]]
            
            output[[mutation_loads_summary_clinical_plots_name]] <- renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>% 
                    layout(title = str_sub(input$data_source_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  if (nrow(d[!is.na(d[,clinical_server[input$clinical_mutation_loads]]) & !is.na(d[,"Mutation.Load"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,clinical_server[input$clinical_mutation_loads]]) & !is.na(d[,"Mutation.Load"]),]
                    if(input$clinical_mutation_loads=="Age") {
                      d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                      #d[d[,"Age"]<=10,"Age"] <- "0-10"
                      d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                      d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                    }
                   p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_mutation_loads]],
                               y=~as.numeric(as.character(Mutation.Load)), 
                               type="violin",#type="box",
                               points="all",#boxpoints="all",
                               jitter=0.3,
                               pointpos=-1.8,
                               #mode="markders",
                               color=d[,clinical_server[input$clinical_mutation_loads]],showlegend=TRUE,
                               colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                               hoveron="points+boxes",
                               legendgroup=d[,clinical_server[input$clinical_mutation_loads]], 
                               text= ~paste0("Sample ID: ", PatientID,
                                             "<br>mutation_loads: ",Mutation.Load,
                                             "<br>Treatment: ",treatment,
                                             "<br>Description: ",description)) %>%
                      layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             height = 480,
                             xaxis=list(title="Clinical Attribute",
                                        titlefont=list(size=20,color="black",family= "Aril black"),
                                        showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                             yaxis=list(title="Mutation Loads",
                                        titlefont=list(size=20,color="black",family= "Aril black"),
                                        showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                             ))
                    incProgress(100, detail = "Doing part %")
                    p
                    
                  } 
                }
              })
            })
          })
        }
      }
    }) 
  })
  
  
  output$mutation_loads_summary_download <- downloadHandler(
    filename = function() {
      if (input$mutation_loads_summary_type == "Excel (CSV)") {
        "newfile.csv"
      } else if (input$mutation_loads_summary_type == "Text (TSV)") {
        "newfile.txt"
      } else if (input$mutation_loads_summary_type == "JSON") {
        "newfile.json"
      }
    },
    content = function(file) {
      if (input$mutation_loads_summary_type == "Excel (CSV)") {
        write.csv (selecteddatabase_mutation_loads(),file,row.names = FALSE)
      } else if (input$mutation_loads_summary_type == "Text (TSV)") {
        write.table (selecteddatabase_mutation_loads(),file,quote=F, row.names = FALSE, sep="\t")
      } else if (input$mutation_loads_summary_type == "JSON") {
        write(toJSON(selecteddatabase_mutation_loads()),file)
      }
    }
  )

  output$mutation_loads_summary_table <- DT::renderDataTable({
    selecteddatabase_mutation_loads()
  })
    

  ##########
  #### mutation loads plots
  ##########

  
  observeEvent(input$do_mutation_loads,{
    input$do_mutation_loads
    
    isolate({
      selected.database <- selecteddatabase_mutation_loads()
      if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Mutation.Load"]),])==0) {
        output$mutation_loads_plots_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })
        output[["mutation_loads_plots_plots0"]] <- renderPlotly({
          plot_ly () %>% 
            layout(title = paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
        })

      } else {
        selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
        splitted_list <- selected.database %>% split(as.character(.$data_source)) #%>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
        splitted_list <- splitted_list[str_sub(input$data_source_mutation_loads,end=-19)]
        #      if (!is.null(length(splitted_list))) {
        
        
        mutation_loads_plots_plots_metadata <- NULL
        for (k in 1:length(input$data_source_mutation_loads)) {
            my_i <- k
            d<-splitted_list[[my_i]]
            
            if (is.na(names(splitted_list)[my_i])) {
              tmp <- data.frame(data_source=input$data_source_mutation_loads[my_i],TE=NA,seTE=NA)
            } else {
              if (nrow(d)==0) {
                tmp = data.frame(data_source=input$data_source_mutation_loads[my_i],TE=NA,seTE=NA)
              } else {
                #d <- d[!is.na(d$PFS.status) & !is.na(d$Mutation.Load) & d$Mutation.Load!="unknown",]
                if (nrow(d[!is.na(d$PFS.status) & !is.na(d$Mutation.Load) & d$Mutation.Load!="unknown",])==0) {
                  tmp = data.frame(data_source=input$data_source_mutation_loads[my_i],TE=NA,seTE=NA)
                } else {
                  #d$Mutation.Load_binary <- NA
                  d$response_binary <- NA
                  d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                  d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                  # d[d$Mutation.Load!="unknown" & d$Mutation.Load=="no mutation","category_binary"] <- 0
                  # d[d$category!="unknown" & d$category=="mutated","category_binary"] <- 1
                  d$scaled <- scale(as.numeric(d$Mutation.Load))
                  if (all(is.na(d$scaled))) {
                    tmp = data.frame(data_source=input$data_source_mutation_loads[my_i],TE=NA,seTE=NA)
                  } else {
                    tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                    if (nrow(tmp.pre$coefficients)>1) {
                      tmp <- data.frame(data_source=input$data_source_mutation_loads[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                    } else {
                      tmp = data.frame(data_source=input$data_source_mutation_loads[my_i],TE=NA,seTE=NA)
                    }
                  }
                  
                  
                }
              }
            }
            mutation_loads_plots_plots_metadata <- rbind.fill(mutation_loads_plots_plots_metadata,tmp)

        }
        if (all(is.na(mutation_loads_plots_plots_metadata))) {
          output$mutation_loads_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
        } else {
          output$mutation_loads_plots_plots_meta <- renderPlot({
            m1 <- metagen(TE,seTE,
                          data=mutation_loads_plots_plots_metadata,
                          sm="OR",
                          #label.e = "response",label.c = "nonresponse",
                          studlab = str_sub(data_source,end=-19))
            forest(m1,layout="RevMan5",    
                   hetstat=TRUE,
                   plotwidth="3inch",test.overall = TRUE,
                   xlim=c(0.1,10),backtransfer=T,
                   label.test.overall.random="random effects model: ",
                   label.test.overall.fixed="fixed effects model: ",
                   fs.test.overall = 15,
                   scientific.pval = TRUE,ff.study.label = "bold")
            # forest(metagen(TE,seTE,
            #                data=mutation_loads_plots_plots_metadata,
            #                sm="OR",
            #                #label.e = "response",label.c = "nonresponse",
            #                studlab = data_source),
            #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
            
          })
        }
        
        
        
        for (k in 1:length(input$data_source_mutation_loads)) {
          local({
            my_i <- k
            mutation_loads_plots_plots_name <- paste0("mutation_loads_plots_plots", my_i)
            
            d<-splitted_list[[my_i]]
            output[[mutation_loads_plots_plots_name]] <- renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
#              isolate({
              if (is.na(names(splitted_list)[my_i])) {
                incProgress(100, detail = "Doing part %")
                plot_ly () %>% 
                  layout(title = str_sub(input$data_source_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no data available in this study under this selection",showarrow=F,font=list(size=16,color="purple")) 
              } else {
                #d <- splitted_list[[my_i]]
                if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Mutation.Load"]),])==0) {
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>% 
                    layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  #d<-splitted_list[[my_i]]
                  d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Mutation.Load"]),]
                  #d <- d[!is.na(d[,"Mutation.Load"]),]
                  d[,"Groups"] <- as.character(d[,"Groups"])
                  
                  d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                  d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                  ind <- sort(unique(d[,"Groups"]))
                  try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Mutation.Load"]),as.numeric(d[d[,"Groups"]==ind[2],"Mutation.Load"]))$p.value)
                  if (class(try.test)=="try-error" | is.na(try.test)) {
                    p.v <- " = NA"
                  } else {
                    # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Mutation.Load"]),as.numeric(d[d[,"Groups"]==ind[2],"Mutation.Load"]))$p.value
                    # if (p.v < 0.001) {
                    #   p.v <- "< 0.001"
                    # } else {
                      p.v<-paste0(" = ",signif(p.v,2))
                   # } 
                  }
                  
                  if (as.character(unique(d$data_source))=="Melanoma-Samstein") {
                    lable_y <- "Number of mutations per Mb"
                    text_y <- "no immunotherapy outcome for this data"
                    color_y <- "purple"
                    size_y <- 16
                  } else {
                    lable_y <- "Number of mutations"
                    text_y <- paste0("p-value",p.v)
                    color_y <- "black"
                    size_y <- 15
                  }
                  p<-plot_ly(data=d,x=~Groups,
                             y=~Mutation.Load,
                             type="violin",#type="box",
                             points="all",#boxpoints="all",
                             jitter=0.3,
                             pointpos=-1.8,
                             #mode="markders",
                             color= ~PFS.status, showlegend=TRUE,
                             colors=c("#7CDBD5","#DCB239"),#"gray"),
                             #hoveron="points+boxes",
                             legendgroup=~PFS.status,
                             text= ~paste0("Sample ID: ", PatientID,
                                           "<br>Mutation loads: ",Mutation.Load,
                                           "<br>Treatment: ",treatment,
                                           "<br>Description: ",description)) %>%
                    layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                           height = 480 ,
                           xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril black"),
                                      showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                           yaxis=list(title=lable_y,
                                      titlefont=list(size=20,color="black",family= "Aril black"),
                                      showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15))) %>% 
                    add_annotations(x=0.5,
                                    y=d[which.max(d[,"Mutation.Load"]),"Mutation.Load"],
                                    text=text_y,
                                    font=list(size=size_y,color=color_y),
                                    xref = paste0("x", 1),  # add this
                                    yref = paste0("y", 1),  # add this
                                    showarrow = FALSE,
                                    ax = 20,
                                    ay = -40,
                                    font=list(size=25,color="black",family= "Aril black"))
                  incProgress(100, detail = "Doing part %")
                  p
                } 
                
              }
            })
            })
            #          }
          })
        }
      }
    })
  })
  
  
  
  #############
  #### mutation loads survival
  #############


  observeEvent(input$do_mutation_loads,{
    input$do_mutation_loads
    isolate({
      selected.database <- selecteddatabase_mutation_loads()
      input_adjust_OS_mutation_loads <- input$adjust_OS_mutation_loads
      if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
        
        output$mutation_loads_survival_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })
        
        output[["mutation_loads_survival_plots0"]] <- renderPlotly({
          plot_ly () %>% 
            layout(title = paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
        })
        
        # output$mutation_loads_survival_plots_combined_p <- renderText({
        #   "NA"
        # })
        # output$mutation_loads_survival_plots_single_p <- renderText({
        #   "NA"
        # })
      } else {
        splitted_list <- selected.database %>% split(.$data_source)# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list <- splitted_list[str_sub(input$data_source_mutation_loads,end=-19)]
        
        mutation_loads_survival_plots_metadata <- NULL
        for (k in 1:length(input$data_source_mutation_loads)) {
            my_i <- k
            d<-splitted_list[[my_i]]
                isolate({
                  if (is.na(names(splitted_list)[my_i])) {
                    tmp <- data.frame(data_source =input$data_source_mutation_loads[my_i],
                                      event.top=NA,
                                      n.top=NA,
                                      event.bottom=NA,
                                      n.bottom=NA)
                    mutation_loads_survival_plots_metadata <- rbind(mutation_loads_survival_plots_metadata,tmp)
                  } else {
                    if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),])==0) {
                      tmp <- data.frame(data_source =input$data_source_mutation_loads[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      mutation_loads_survival_plots_metadata <- rbind(mutation_loads_survival_plots_metadata,tmp)
                    } else {
                      d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),]
                      d<-data.frame(d,dead=NA)
                      d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                      d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                      d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"dead"]),]
                      d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                      d <- data.frame(data_source=as.character(d$data_source),
                                      dead=as.numeric(d$dead),
                                      OS=as.numeric(d$OS),
                                      Age=as.character(d$Age),
                                      Gender=as.character(d$Gender),
                                      Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                      # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Mutation.Load")]))
                      d<-data.frame(d,IND=paste0("bottom ",
                                                 input$slider_mutation_loads_OS*100, 
                                                 "% (",
                                                 nrow(d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<= quantile(na.omit(d[,"Mutation.Load"]),input$slider_mutation_loads_OS),]),
                                                 ")"))
                      d[,"IND"]<-as.character(d[,"IND"])
                      d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_mutation_loads_OS),"IND"]<- 
                        paste0("top ",
                               (1-input$slider_mutation_loads_OS)*100,
                               "% (",
                               nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),])
                               ,")")
                      d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                      d<-d[order(d[,"IND"]),]
                      
                      tmp <- data.frame(data_source =input$data_source_mutation_loads[my_i],
                                        event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                        n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                        event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                        n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                      mutation_loads_survival_plots_metadata <- rbind(mutation_loads_survival_plots_metadata,tmp)
                      
                      
                      # if (input_adjust_OS_mutation_loads=="None") {
                      #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                      #   if (class(diff.result.try)=="try-error") {
                      #     p.value <- NA
                      #   } else {
                      #     p.value <- signif(summary(diff.result)$logtest[3],2)
                      #   }
                      # } else {
                      #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_mutation_loads),data=d) )
                      #   if (class(diff.result.adjust.try)=="try-error") {
                      #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                      #     
                      #   } else {
                      #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                      #   }
                      # }
                      
                      # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                      # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                      # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                      # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                      #mutation_loads_survival_plots_combined_p <- c(mutation_loads_survival_plots_combined_p,p.value)
                    }
                  }
                })
        }
        if (all(is.na(mutation_loads_survival_plots_metadata))) {
          output$mutation_loads_survival_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          # output$mutation_loads_survival_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_loads_survival_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          output$mutation_loads_survival_plots_meta <- renderPlot({
            # meta1 <- metabin(event.top,n.top,event.bottom,n.bottom,
            #                  data=mutation_loads_survival_plots_metadata,sm="OR",
            #                  studlab = data_source)
            meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                             data=mutation_loads_survival_plots_metadata,sm="OR",
                             studlab = data_source)
            meta2 <- data.frame(data_source=meta1$studlab,
                                TE=meta1$TE,
                                seTE=meta1$seTE)
            m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
            forest(m1,layout="RevMan5",    
                   hetstat=TRUE,
                   plotwidth="3inch",test.overall = TRUE,
                   xlim=c(0.1,10),backtransfer=T,
                   label.test.overall.random="random effects model: ",
                   label.test.overall.fixed="fixed effects model: ",
                   fs.test.overall = 15,
                   scientific.pval = TRUE,ff.study.label = "bold")
            # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
            #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
            
            # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
            #                  studlab=mutation_loads_survival_plots_metadata$data_source)
            # forest(meta2,test.overall= T,ff.study.label = "bold")
            
            
          })
          
          # output$mutation_loads_survival_plots_combined_p <- renderText({
          #   signif(fisher.method(rbind(a=na.omit(mutation_loads_survival_plots_combined_p)))$p.value,2)
          # })
          # output$mutation_loads_survival_plots_single_p <- renderText({
          #   paste0(signif(mutation_loads_survival_plots_combined_p,2),collapse = ", ")
          # })
        }
        
        
        
        
        for (k in 1:length(input$data_source_mutation_loads)) {
          local({
            my_i <- k
            mutation_loads_survival_plots_name <- paste0("mutation_loads_survival_plots", my_i)
            mutation_loads_survival_summary_tables_name <- paste0("mutation_loads_survival_summary_tables", my_i)
            d<-splitted_list[[my_i]]
            #output[[mutation_loads_survival_plots_name]] <- renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
              isolate({
              if (is.na(names(splitted_list)[my_i])) {
                incProgress(100, detail = "Doing part %")
                p1 <- plot_ly () %>% 
                  layout(title = str_sub(input$data_source_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no data available in this study under this selection",showarrow=F,font=list(size=16,color="purple")) 
                table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                
              } else {
                
                if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),])==0 ) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>% 
                    layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  
                } else {
                  d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),]
                  d<-data.frame(d,dead=NA)
                  d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                  d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                  d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"dead"]),]
                  d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                  #d <- data.frame(data_source=as.character(d$data_source),dead=as.numeric(d$dead),OS=as.numeric(d$OS),category=as.character(d[,"category.OS"]))
                  
                  if (as.character(unique(d$data_source))=="Samstein") {
                    lable_y <- "number of mutations per Mb"
                  } else {
                    lable_y <- "number of mutations"
                  }
                  d <- data.frame(data_source=as.character(d$data_source),
                                  dead=as.numeric(d$dead),
                                  OS=as.numeric(d$OS),
                                  Age=as.character(d$Age),
                                  Gender=as.character(d$Gender),
                                  Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                  # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Mutation.Load")]))
                  d<-data.frame(d,IND=paste0("bottom ",
                                             input$slider_mutation_loads_OS*100, 
                                             "% (",
                                             nrow(d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<= quantile(na.omit(d[,"Mutation.Load"]),input$slider_mutation_loads_OS),]),
                                             ")"))
                  d[,"IND"]<-as.character(d[,"IND"])
                  d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_mutation_loads_OS),"IND"]<- 
                    paste0("top ",
                           (1-input$slider_mutation_loads_OS)*100,
                           "% (",
                           nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),])
                           ,")")
                  d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                  d<-d[order(d[,"IND"]),]
                  fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                  
                  try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                  if (class(try.test)=="try-error" | is.na(try.test)) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                  
                    if (input_adjust_OS_mutation_loads=="None") {
                      diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                      if (class(diff.result.try)=="try-error") {
                        p.value <- NA
                        table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                      } else {
                        p.value <- signif(summary(diff.result)$logtest[3],2)
                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                          caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                 signif(summary(diff.result)$logtest[1],2),
                                                                                 " on ",
                                                                                 signif(summary(diff.result)$logtest[2],2),
                                                                                 " df, p-value = ",
                                                                                 signif(summary(diff.result)$logtest[3],2)),
                                                                          br(),
                                                                          paste0("n = ",
                                                                                 signif(summary(diff.result)$n,2),
                                                                                 ", number of events = ",
                                                                                 signif(summary(diff.result)$nevent,2)),
                                                                          style="text-align:left;color:black;font-size:15px"),
                                          options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      
                    }
                  } else {
                    diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_mutation_loads),data=d) )
                    if (class(diff.result.adjust.try)=="try-error") {
                      diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                      p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                        " (No ",input_adjust_OS_mutation_loads," was observed)")
                      
                      table1 <- datatable(signif(coef(summary(diff.result)),2),
                                          caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                 signif(summary(diff.result)$logtest[1],2),
                                                                                 " on ",
                                                                                 signif(summary(diff.result)$logtest[2],2),
                                                                                 " df, p-value = ",
                                                                                 signif(summary(diff.result)$logtest[3],2)),
                                                                          br(),
                                                                          paste0("n = ",
                                                                                 signif(summary(diff.result)$n,2),
                                                                                 ", number of events = ",
                                                                                 signif(summary(diff.result)$nevent,2)),
                                                                          style="text-align:left;color:black;font-size:15px"),
                                          options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    } else {
                      p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                      table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                      rownames(table1.pre)[2] <- input_adjust_OS_mutation_loads
                      table1 <- datatable(table1.pre,
                                          caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                 signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                 " on ",
                                                                                 signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                 " df, p-value = ",
                                                                                 signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                          br(),
                                                                          paste0("n = ",
                                                                                 signif(summary(diff.result.adjust)$n,2),
                                                                                 ", number of events = ",
                                                                                 signif(summary(diff.result.adjust)$nevent,2)),
                                                                          style="text-align:left;color:black;font-size:15px"),
                                          options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    }
                  }
                  
                  
                  # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                  # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                  plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                    ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                    #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                             paste0("p-value = ",p.value),x=300,y=0.05)) +
                    ggplot2::guides(color=FALSE,linetype=FALSE)
                  p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                    layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                           xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                           yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                           legend=list(orientation="v",x=0.6,y=1)) #%>%
                  incProgress(100, detail = "Doing part %")
                  p1
                  }
                }
              }
            })
            })
            #})
              output[[mutation_loads_survival_plots_name]] <- renderPlotly({
                p1
              })
              output[[mutation_loads_survival_summary_tables_name]] <- renderDataTable(
                table1
              )
          })
        }
      }
    })
  })
  
  #############
  #### mutation loads PFS
  #############
  observeEvent(input$do_mutation_loads,{
    input$do_mutation_loads
    isolate({
      selected.database <- selecteddatabase_mutation_loads()
      input_adjust_OS_mutation_loads <- input$adjust_OS_mutation_loads
      if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
        output$mutation_loads_PFS_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })
        
        output[["mutation_loads_PFS_plots0"]] <- renderPlotly({
          plot_ly () %>% 
            layout(title = paste(str_sub(input$data_source_mutation_loads,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
        })
        
        
        # output$mutation_loads_PFS_plots_combined_p <- renderText({
        #   "NA"
        # })
        # output$mutation_loads_PFS_plots_single_p <- renderText({
        #   "NA"
        # })
      } else {
        splitted_list <- selected.database %>% split(.$data_source)#%>% as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list <- splitted_list[str_sub(input$data_source_mutation_loads,end=-19)]
        
        mutation_loads_PFS_plots_metadata <- NULL
        for (k in 1:length(input$data_source_mutation_loads)) {
            my_i <- k
            d<-splitted_list[[my_i]]
                #isolate({
                  if (is.na(names(splitted_list)[my_i])) {
                    tmp <- data.frame(data_source =input$data_source_mutation_loads[my_i],
                                      event.top=NA,
                                      n.top=NA,
                                      event.bottom=NA,
                                      n.bottom=NA)
                    mutation_loads_PFS_plots_metadata <- rbind(mutation_loads_PFS_plots_metadata,tmp)
                  } else {
                    if (nrow(d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                      tmp <- data.frame(data_source =input$data_source_mutation_loads[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      mutation_loads_PFS_plots_metadata <- rbind(mutation_loads_PFS_plots_metadata,tmp)
                    } else {
                      d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                      d <- data.frame(data_source=as.character(d$data_source),
                                      PFS.status=as.character(d$PFS.status),
                                      PFS=as.numeric(d$PFS),
                                      Age=as.character(d$Age),
                                      Gender=as.character(d$Gender),
                                      Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                      # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Mutation.Load")]))
                      d<-data.frame(d,IND=paste0("bottom ",input$slider_mutation_loads_OS*100, "% (",
                                                 nrow(d[d[,"Mutation.Load"]<= quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),]),
                                                 ")"))
                      d[,"IND"]<-as.character(d[,"IND"])
                      d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),"IND"]<- 
                        paste0("top ",(1-input$slider_mutation_loads_OS)*100,"% (",
                               nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),])
                               ,")")
                      d <- data.frame(d,group.IND=0)
                      d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                      #        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                      d<-d[order(d[,"IND"]),]
                      
                      tmp <- data.frame(data_source =input$data_source_mutation_loads[my_i],
                                        event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                        n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                        event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                        n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                      mutation_loads_PFS_plots_metadata <- rbind(mutation_loads_PFS_plots_metadata,tmp)
                      
                      
                      # if (input_adjust_OS_mutation_loads=="None") {
                      #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                      #   if (class(diff.result.try)=="try-error") {
                      #     p.value <- NA
                      #   } else {
                      #     p.value <- signif(summary(diff.result)$logtest[3],2)
                      #   }
                      # } else {
                      #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_mutation_loads),data=d) )
                      #   if (class(diff.result.adjust.try)=="try-error") {
                      #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                      #     
                      #   } else {
                      #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                      #   }
                      # }
                      # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                      # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                      # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                      # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                      # mutation_loads_PFS_plots_combined_p <- c(mutation_loads_PFS_plots_combined_p,p.value)
                    }
                  }
                #})
        }
        if (all(is.na(mutation_loads_PFS_plots_metadata))) {
          output$mutation_loads_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          # output$mutation_loads_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_loads_PFS_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          output$mutation_loads_PFS_plots_meta <- renderPlot({
            meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                             data=mutation_loads_PFS_plots_metadata,sm="OR",
                             studlab = data_source)
            meta2 <- data.frame(data_source=meta1$studlab,
                                TE=meta1$TE,
                                seTE=meta1$seTE)
            m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
            forest(m1,layout="RevMan5",    
                   hetstat=TRUE,
                   plotwidth="3inch",test.overall = TRUE,
                   xlim=c(0.1,10),backtransfer=T,
                   label.test.overall.random="random effects model: ",
                   label.test.overall.fixed="fixed effects model: ",
                   fs.test.overall = 15,
                   scientific.pval = TRUE,ff.study.label = "bold")
            # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
            #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
            
            # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
            #                  studlab=mutation_loads_PFS_plots_metadata$data_source)
            # forest(meta2,test.overall= T,ff.study.label = "bold")
          })
          # output$mutation_loads_PFS_plots_combined_p <- renderText({
          #   signif(fisher.method(rbind(a=na.omit(mutation_loads_PFS_plots_combined_p)))$p.value,2)
          # })
          # output$mutation_loads_PFS_plots_single_p <- renderText({
          #   paste0(signif(mutation_loads_PFS_plots_combined_p,2),collapse = ", ")
          # })
        }
        
        for (k in 1:length(input$data_source_mutation_loads)) {
          local({
            my_i <- k
            mutation_loads_PFS_plots_name <- paste0("mutation_loads_PFS_plots", my_i)
            mutation_loads_PFS_summary_tables_name <- paste0("mutation_loads_PFS_summary_tables", my_i)
            d<-splitted_list[[my_i]]
           
            #output[[mutation_loads_PFS_plots_name]] <- renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
              isolate({
              if (is.na(names(splitted_list)[my_i])) {
                incProgress(100, detail = "Doing part %")
                p1 <- plot_ly () %>% 
                  layout(title = str_sub(input$data_source_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                
              } else {
                if (nrow(d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>% 
                    layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  
                } else {
                  d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                  
                  if (as.character(unique(d$data_source))=="Samstein") {
                    lable_y <- "number of mutations per Mb"
                  } else {
                    lable_y <- "number of mutations"
                  }
                  d <- data.frame(data_source=as.character(d$data_source),
                                  PFS.status=as.character(d$PFS.status),
                                  PFS=as.numeric(d$PFS),
                                  Age=as.character(d$Age),
                                  Gender=as.character(d$Gender),
                                  Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                  # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Mutation.Load")]))
                  d<-data.frame(d,IND=paste0("bottom ",input$slider_mutation_loads_OS*100, "% (",
                                             nrow(d[d[,"Mutation.Load"]<= quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),]),
                                             ")"))
                  d[,"IND"]<-as.character(d[,"IND"])
                  d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),"IND"]<- 
                    paste0("top ",(1-input$slider_mutation_loads_OS)*100,"% (",
                           nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_mutation_loads_OS),])
                           ,")")
                  d <- data.frame(d,group.IND=0)
                  d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                  #        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                  d<-d[order(d[,"IND"]),]
                  fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                  
                  try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                  if (class(try.test)=="try-error" | is.na(try.test)) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                  
                  if (input_adjust_OS_mutation_loads=="None") {
                    diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                    if (class(diff.result.try)=="try-error") {
                      p.value <- NA
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                    } else {
                      p.value <- signif(summary(diff.result)$logtest[3],2)
                      table1 <- datatable(signif(coef(summary(diff.result)),2),
                                          caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                 signif(summary(diff.result)$logtest[1],2),
                                                                                 " on ",
                                                                                 signif(summary(diff.result)$logtest[2],2),
                                                                                 " df, p-value = ",
                                                                                 signif(summary(diff.result)$logtest[3],2)),
                                                                          br(),
                                                                          paste0("n = ",
                                                                                 signif(summary(diff.result)$n,2),
                                                                                 ", number of events = ",
                                                                                 signif(summary(diff.result)$nevent,2)),
                                                                          style="text-align:left;color:black;font-size:15px"),
                                          options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      
                    }
                  } else {
                    diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_mutation_loads),data=d) )
                    if (class(diff.result.adjust.try)=="try-error") {
                      diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                      p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                        " (No ",input_adjust_OS_mutation_loads," was observed)")
                      
                      table1 <- datatable(signif(coef(summary(diff.result)),2),
                                          caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                 signif(summary(diff.result)$logtest[1],2),
                                                                                 " on ",
                                                                                 signif(summary(diff.result)$logtest[2],2),
                                                                                 " df, p-value = ",
                                                                                 signif(summary(diff.result)$logtest[3],2)),
                                                                          br(),
                                                                          paste0("n = ",
                                                                                 signif(summary(diff.result)$n,2),
                                                                                 ", number of events = ",
                                                                                 signif(summary(diff.result)$nevent,2)),
                                                                          style="text-align:left;color:black;font-size:15px"),
                                          options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    } else {
                      p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                      table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                      rownames(table1.pre)[2] <- input_adjust_OS_mutation_loads
                      table1 <- datatable(table1.pre,
                                          caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                 signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                 " on ",
                                                                                 signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                 " df, p-value = ",
                                                                                 signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                          br(),
                                                                          paste0("n = ",
                                                                                 signif(summary(diff.result.adjust)$n,2),
                                                                                 ", number of events = ",
                                                                                 signif(summary(diff.result.adjust)$nevent,2)),
                                                                          style="text-align:left;color:black;font-size:15px"),
                                          options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    }
                  }
                  
                  
                  
                  # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                  # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                  plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                    ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                    #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                             paste0("p-value = ",p.value),x=300,y=0.05)) +
                    ggplot2::guides(color=FALSE,linetype=FALSE)
                  p1<-ggplotly(plot.result.ann, height = 450  ) %>%
                    layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                           xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                           yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                           legend=list(orientation="v",x=0.6,y=1)) 
                  incProgress(100, detail = "Doing part %")
                  p1
                  }
                }
              }
            })
            })
          #})
              output[[mutation_loads_PFS_plots_name]] <- renderPlotly({
                p1
              })
              output[[mutation_loads_PFS_summary_tables_name]] <- renderDataTable(
                table1
              )
          })
        }
      }
    })
  })
  

  
    ###########################################
    #### mutational signatures
    ###########################################
  # input <- list(data_source_mutation_signature=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),
  #               treatment_mutation_signature=c("anti-PD-1/L1","anti-CTLA-4"),
  #               description_mutation_signature=c("Pre","On"),
  #               signature_mutation_signature="Signature.1 (age correlated)",
  #               slider_mutation_signature_OS=0.5)
  
  updateSelectizeInput(session,"signature_mutation_signature",
                       choices= sort(c("Signature.1 (age correlated)","Signature.2","Signature.3 (BRCA1 and BRCA2 associated)","Signature.4 (tobacco associated)","Signature.5","Signature.6 (DNA mismatch repair associated)","Signature.7 (ultraviolet associated)","Signature.8","Signature.9","Signature.10","Signature.11","Signature.12","Signature.13","Signature.14 (high tumor mutation burden associated)","Signature.15 (DNA mismatch repair associated)","Signature.16","Signature.17","Signature.18","Signature.19","Signature.20 (DNA mismatch repair associated)","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25 (Hodgkin's cell lines only)","Signature.26 (DNA mismatch repair associated)","Signature.27","Signature.28","Signature.29","Signature.30","unknown")),
                       selected="Signature.1 (age correlated)",
                       server=TRUE)
  observe({
    if (input$data_source_mutation_signature=="" || is.null(input$data_source_mutation_signature) ||
        input$treatment_mutation_signature=="" || is.null(input$treatment_mutation_signature) ||
        input$description_mutation_signature=="" || is.null(input$description_mutation_signature) ||
        input$signature_mutation_signature=="" || is.null(input$signature_mutation_signature)){
      disable("do_mutation_signature")
    } else {
      enable("do_mutation_signature")
    }
  })

    selecteddatabase_mutation_signature <- eventReactive(input$do_mutation_signature,{
      data_source1 <- str_sub(input$data_source_mutation_signature,end=-19)
      #treatment1 <- input$treatment_mutation_signature
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_mutation_signature])
      description1 <- input$description_mutation_signature
      genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",str_split(input$signature_mutation_signature," ",2)[[1]][1])
      a<-data.table(database.test.process())
      m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
      m<-m[,colnames(m) %in% genes1,with=FALSE]
      colnames(m)[13] <- "Signature"
      m<-m[,Signature:=as.numeric(Signature)]
      m <- as.data.frame(m)
      if (nrow(m)>0) {
#        m<-data.frame(m,category.OS="no data available",category.PFS="no data available")
        m<-data.frame(m,category="unknown")
        m[,"category"] <- as.character(m[,"category"])
        m1 <- m %>% split(.$data_source) %>% 
          lapply(function(d) {
            d[!is.na(d[,"Signature"]) & d[,"Signature"]> quantile(na.omit(d[,"Signature"]),input$slider_mutation_signature_OS),"category"] <- paste0("top ",
                                                                                                                                                        (1-input$slider_mutation_signature_OS)*100,
                                                                                                                                                        "% ",sep="")
            d[!is.na(d[,"Signature"]) & d[,"Signature"]<=quantile(na.omit(d[,"Signature"]),input$slider_mutation_signature_OS),"category"] <- paste0("bottom ",
                                                                                                                                                        input$slider_mutation_signature_OS*100,
                                                                                                                                                        "% ",sep="")
            
            d })
        
        m2 <- ldply(m1,data.frame)
        m2 <- m2[,-1]
        m2
      } else {
        m2<-m
        m2
      } 
    })
    
    ##########
    #### mutational signature summary
    #########
    output$mutation_signature_summary_disease_number <- renderText({
      input$do_mutation_signature ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selecteddatabase_mutation_signature()$Disease))})})
    output$mutation_signature_summary_sample_number <- renderText({
      input$do_mutation_signature ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selecteddatabase_mutation_signature())})})
    output$mutation_signature_summary_plots <- renderPlotly({
      summary_plot(inputdata=selecteddatabase_mutation_signature(),feature="mutation_signature")
    })
    

    observeEvent(input$do_mutation_signature,{
      input$do_mutation_signature
      isolate({
        input_signature_mutation_signature <- input$signature_mutation_signature
        selected.database <- selecteddatabase_mutation_signature()
        if (nrow(selected.database)==0 ) {
          output[["mutation_signature_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_mutation_signature,end=-19)]
          for (k in 1:length(input$data_source_mutation_signature)) {
            local({
              my_i <- k
              mutation_signature_summary_clinical_plots_name <- paste0("mutation_signature_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[mutation_signature_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_mutation_signature]]) & !is.na(d[,"Signature"]),])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_mutation_signature]]) & !is.na(d[,"Signature"]),]
                      if(input$clinical_mutation_signature=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        #d[d[,"Age"]<=10,"Age"] <- "0-10"
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      
                      p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_mutation_signature]],
                                 y=~as.numeric(as.character(Signature)), 
                                 type="violin",#type="box",
                                 points="all",#boxpoints="all",
                                 jitter=0.3,
                                 pointpos=-1.8,
                                 #mode="markders",
                                 color=d[,clinical_server[input$clinical_mutation_signature]],showlegend=TRUE,
                                 colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                 hoveron="points+boxes",
                                 legendgroup=d[,clinical_server[input$clinical_mutation_signature]], 
                                 text= ~paste0("Sample ID: ", PatientID,
                                               "<br>Signature: ",str_split(input$signature_mutation_signature," ",2)[[1]][1]," :",Signature,
                                               "<br>Treatment: ",treatment,
                                               "<br>Description: ",description)) %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title="Clinical Attribute",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title=paste(str_split(input_signature_mutation_signature," ",2)[[1]][1], " weight",sep=""),
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               ))
                      incProgress(100, detail = "Doing part %")
                      p
                      
                    } 
                  }
                })
              })
            })
          }
        }
      }) 
    })
    
    
    output$mutation_signature_summary_download <- downloadHandler(
      filename = function() {
        if (input$mutation_signature_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$mutation_signature_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$mutation_signature_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        if (input$mutation_signature_summary_type == "Excel (CSV)") {
          write.csv (selecteddatabase_mutation_signature(),file,row.names = FALSE)
        } else if (input$mutation_signature_summary_type == "Text (TSV)") {
          write.table (selecteddatabase_mutation_signature(),file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$mutation_signature_summary_type == "JSON") {
          write(toJSON(selecteddatabase_mutation_signature()),file)
        }
      }
    )
    
    output$mutation_signature_summary_table <- DT::renderDataTable({
      selecteddatabase_mutation_signature()
    })
    
    ##########
    #### mutational signature plots
    ##########
    observeEvent(input$do_mutation_signature,{
      input$do_mutation_signature
      isolate({
        input_signature_mutation_signature <- input$signature_mutation_signature
        selected.database <- selecteddatabase_mutation_signature()
        if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Signature"]),])==0) {
          output$mutation_signature_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["mutation_signature_plots_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$mutation_signature_plots_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_signature_plots_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_mutation_signature,end=-19)]
          
          #mutation_signature_plots_plots_combined_p <- NULL
          mutation_signature_plots_plots_metadata <- NULL
          for (k in 1:length(input$data_source_mutation_signature)) {
            my_i <- k
            d<-splitted_list[[my_i]]
            
            if (is.na(names(splitted_list)[my_i])) {
              tmp <- data.frame(data_source=input$data_source_mutation_signature[my_i],TE=NA,seTE=NA)
            } else {
              if (nrow(d)==0) {
                tmp = data.frame(data_source=input$data_source_mutation_signature[my_i],TE=NA,seTE=NA)
              } else {
                #d <- d[!is.na(d$PFS.status) & !is.na(d$Signature) & d$Signature!="unknown",]
                if (nrow(d[!is.na(d$PFS.status) & !is.na(d$Signature) & d$Signature!="unknown",])==0) {
                  tmp = data.frame(data_source=input$data_source_mutation_signature[my_i],TE=NA,seTE=NA)
                } else {
                  # d$category_binary <- NA
                  d$response_binary <- NA
                  d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                  d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                  # d[d$category!="unknown" & d$category=="no mutation_signature","category_binary"] <- 0
                  # d[d$category!="unknown" & d$category=="mutated","category_binary"] <- 1
                  d$scaled <- scale(as.numeric(d$Signature))
                  if (all(is.na(d$scaled))) {
                    tmp = data.frame(data_source=input$data_source_mutation_signature[my_i],TE=NA,seTE=NA)
                  } else {
                    tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                    if (nrow(tmp.pre$coefficients)>1) {
                      tmp <- data.frame(data_source=input$data_source_mutation_signature[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                    } else {
                      tmp <- data.frame(data_source=input$data_source_mutation_signature[my_i],TE=NA,seTE=NA)
                    }
                  }
                  
                  
                }
              }
            }
            mutation_signature_plots_plots_metadata <- rbind.fill(mutation_signature_plots_plots_metadata,tmp)
            
            # d.res <- d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),]
            # d.nonres <- d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),]
            # if (nrow(d.res)>1 & nrow(d.nonres)>1) {
            #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
            #                     n.res=length(na.omit(d.res[,"Signature"])),
            #                     mean.res=mean(na.omit(d.res[,"Signature"])),
            #                     sd.res=sd(na.omit(d.res[,"Signature"])),
            #                     n.nonres=length(na.omit(d.nonres[,"Signature"])),
            #                     mean.nonres=mean(na.omit(d.nonres[,"Signature"])),
            #                     sd.nonres=sd(na.omit(d.nonres[,"Signature"])))
            #   mutation_signature_plots_plots_metadata <- rbind(mutation_signature_plots_plots_metadata,tmp)
            # } else {
            #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
            #                     n.res=NA,
            #                     mean.res=NA,
            #                     sd.res=NA,
            #                     n.nonres=NA,
            #                     mean.nonres=NA,
            #                     sd.nonres=NA)
            #   mutation_signature_plots_plots_metadata <- rbind(mutation_signature_plots_plots_metadata,tmp)
            # }   
          }
          if (all(is.na(mutation_signature_plots_plots_metadata))) {
            output$mutation_signature_plots_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$mutation_signature_plots_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$mutation_signature_plots_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            # output$mutation_signature_plots_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(mutation_signature_plots_plots_combined_p)))$p.value,2)
            # })
            # output$mutation_signature_plots_plots_single_p <- renderText({
            #   paste0(signif(mutation_signature_plots_plots_combined_p,2),collapse = ", ")
            # })
            output$mutation_signature_plots_plots_meta <- renderPlot({
              
              # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
              #                 data=mutation_signature_plots_plots_metadata,
              #                 label.e = "response",label.c = "nonresponse",
              #                 studlab = data_source),
              #        test.overall = TRUE,ff.study.label = "bold")
              m1 <- metagen(TE,seTE,
                            data=mutation_signature_plots_plots_metadata,
                            sm="OR",
                            #label.e = "response",label.c = "nonresponse",
                            studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,
              #                data=mutation_signature_plots_plots_metadata,
              #                sm="OR",
              #                #label.e = "response",label.c = "nonresponse",
              #                studlab = data_source),
              #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
            })
          }
          
          
          for (k in 1:length(input$data_source_mutation_signature)) {
            local({
              my_i <- k
              mutation_signature_plots_plots_name <- paste0("mutation_signature_plots_plots", my_i)
              d<-splitted_list[[my_i]]
              output[[mutation_signature_plots_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>% 
                    layout(title = str_sub(input$data_source_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  #d <- splitted_list[[my_i]]
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),]
                    d[,"Groups"] <- as.character(d[,"Groups"])
                    d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    ind <- sort(unique(d[,"Groups"]))
                    try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Signature"]),as.numeric(d[d[,"Groups"]==ind[2],"Signature"]))$p.value)
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      p.v <- "NA"
                    } else {
                      # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Signature"]),as.numeric(d[d[,"Groups"]==ind[2],"Signature"]))$p.value
                      # if (is.na(p.v)) {
                      #   p.v <- " = NA"
                      # } else {
                      #   if (p.v < 0.001) {
                      #     p.v <- " < 0.001"
                      #   } else {
                          p.v<-paste0(" = ",signif(p.v,2))
                      #   } 
                      # }
                    }
                    p<-plot_ly(data=d,x=~Groups,
                            y=~Signature,
                            type="violin",#type="box",
                            points="all",#boxpoints="all",
                            jitter=0.3,
                            pointpos=-1.8,
                            #mode="markders",
                            color= ~PFS.status,
                            colors=c("#7CDBD5","#DCB239","gray"),
                            #hoveron="points+boxes",
                            legendgroup=~PFS.status,
                            showlegend=TRUE,
                            text= ~paste0("Sample ID: ", PatientID,
                                          "<br>Signature: ",str_split(input$signature_mutation_signature," ",2)[[1]][1]," :",Signature,
                                          "<br>Treatment: ",treatment,
                                          "<br>Description: ",description)) %>%
                      layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             height = 480,
                             xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                        showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                             yaxis=list(title=paste(str_split(input_signature_mutation_signature," ",2)[[1]][1], " weight",sep=""),
                                        titlefont=list(size=20,color="black",family= "Aril"),
                                        showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                             )) %>%
                             add_annotations(x=0.5,
                                              y=d[which.max(d[,"Signature"]),"Signature"],
                                              text=paste0("p-value",p.v),
                                              xref = paste0("x", 1),  # add this
                                              yref = paste0("y", 1),  # add this
                                              showarrow = FALSE,
                                              ax = 20,
                                              ay = -40,
                                              font=list(size=20,color="black",family= "Aril black"))
                    incProgress(100, detail = "Doing part %")
                    p
                    } 
                  
                }
              })
              })
            })
          }
        }
      })
    })
    
    #############
    #### mutational signature survival
    #############
    observeEvent(input$do_mutation_signature,{
      input$do_mutation_signature
      isolate({
        input_signature_mutation_signature <- input$signature_mutation_signature
        selected.database <- selecteddatabase_mutation_signature()
        input_adjust_OS_mutation_signature <- input$adjust_OS_mutation_signature
        if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
          output$mutation_signature_survival_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["mutation_signature_survival_plots0"]] <- renderPlotly({
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })
          # output$mutation_signature_survival_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_signature_survival_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          #    selected.database[selected.database[,"OS.status"]=="living" & !is.na(selected.database[,"OS.status"]),"dead"] <- 0
          #    selected.database[selected.database[,"OS.status"]=="deceased" & !is.na(selected.database[,"OS.status"]),"dead"] <- 1
          #    selected.database <- selected.database[!is.na(selected.database[,"Signature"]) & !is.na(selected.database[,"dead"]),]
          #    selected.database <- selected.database[selected.database[,"dead"]==1 | selected.database[,"dead"]==0,]

          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_mutation_signature,end=-19)]

          # mutation_signature_survival_plots_combined_p <- NULL
          mutation_signature_survival_plots_metadata <- NULL
          for (k in 1:length(input$data_source_mutation_signature)) {
              my_i <- k
              d<-splitted_list[[my_i]]
              d[,"Signature"] <- as.numeric(as.character(d[,"Signature"]))
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # mutation_signature_survival_plots_combined_p <- c(mutation_signature_survival_plots_combined_p,NA)
                      tmp <- data.frame(data_source =input$data_source_mutation_signature[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      mutation_signature_survival_plots_metadata <- rbind(mutation_signature_survival_plots_metadata,tmp)

                    } else {

                      if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),])==0) {
                        # mutation_signature_survival_plots_combined_p <- c(mutation_signature_survival_plots_combined_p,NA)
                        tmp <- data.frame(data_source =input$data_source_mutation_signature[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        mutation_signature_survival_plots_metadata <- rbind(mutation_signature_survival_plots_metadata,tmp)

                      } else {
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),]
                        d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                        d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                        d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"dead"]),]
                        d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        dead=as.numeric(d$dead),
                                        OS=as.numeric(d$OS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        Signature=as.numeric(d[,"Signature"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Signature")]))
                        d<-data.frame(d,IND=paste0("bottom ",
                                                   input$slider_mutation_signature_OS*100,
                                                   "% (",
                                                   nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_mutation_signature_OS),]),
                                                   ")"))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),"IND"]<-
                          paste0("top ",
                                 (1-input$slider_mutation_signature_OS)*100,
                                 "% (",
                                 nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),])
                                 ,")")
                        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                        d<-d[order(d[,"IND"]),]

                        tmp <- data.frame(data_source =input$data_source_mutation_signature[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        mutation_signature_survival_plots_metadata <- rbind(mutation_signature_survival_plots_metadata,tmp)


                        # if (input_adjust_OS_mutation_signature=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_mutation_signature),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                        #
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        # # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                        # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                        # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # mutation_signature_survival_plots_combined_p <- c(mutation_signature_survival_plots_combined_p,p.value)
                      }
                    }
                  })
          }
          if (all(is.na(mutation_signature_survival_plots_metadata))) {
            output$mutation_signature_survival_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$mutation_signature_survival_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$mutation_signature_survival_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$mutation_signature_survival_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=mutation_signature_survival_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=mutation_signature_survival_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")

            })
            # output$mutation_signature_survival_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(mutation_signature_survival_plots_combined_p)))$p.value,2)
            # })
            # output$mutation_signature_survival_plots_single_p <- renderText({
            #   paste0(signif(mutation_signature_survival_plots_combined_p,2),collapse = ", ")
            # })
          }



          for (k in 1:length(input$data_source_mutation_signature)) {
            local({
              my_i <- k
              mutation_signature_survival_plots_name <- paste0("mutation_signature_survival_plots", my_i)
              mutation_signature_survival_summary_tables_name <- paste0("mutation_signature_survival_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              d[,"Signature"] <- as.numeric(as.character(d[,"Signature"]))
              #output[[mutation_signature_survival_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                } else {

                  if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),]
                    d<-data.frame(d,dead=NA)
                    d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                    d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                    d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"dead"]),]
                    d <- d[d[,"dead"]==1 | d[,"dead"]==0,]

                    d <- data.frame(data_source=as.character(d$data_source),
                                    dead=as.numeric(d$dead),
                                    OS=as.numeric(d$OS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    Signature=as.numeric(d[,"Signature"]))
                    #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Signature")]))
                    d<-data.frame(d,IND=paste0("bottom ",
                                               input$slider_mutation_signature_OS*100,
                                               "% (",
                                               nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_mutation_signature_OS),]),
                                               ")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),"IND"]<-
                      paste0("top ",
                             (1-input$slider_mutation_signature_OS)*100,
                             "% (",
                             nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),])
                             ,")")
                    d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                    d<-d[order(d[,"IND"]),]
                    fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                      if (input_adjust_OS_mutation_signature=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_mutation_signature),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_mutation_signature," was observed)")

                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_mutation_signature
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }

                    # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                    p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                      layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                             yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                             legend=list(orientation="v",x=0.6,y=1)) #%>%
                    incProgress(100, detail = "Doing part %")
                    p1
                    }
                  }
                }
              })
              })
                #})
                output[[mutation_signature_survival_plots_name]] <- renderPlotly({
                  p1
                })
                output[[mutation_signature_survival_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })
    
    #############
    #### mutational signature PFS
    #############

    
    observeEvent(input$do_mutation_signature,{
      input$do_mutation_signature
      isolate({
        input_signature_mutation_signature <- input$signature_mutation_signature
        selected.database <- selecteddatabase_mutation_signature()
        input_adjust_OS_mutation_signature <- input$adjust_OS_mutation_signature
        if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Signature"]),])==0) {
          output$mutation_signature_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["mutation_signature_PFS_plots0"]] <- renderPlotly({
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })

          # output$mutation_signature_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$mutation_signature_PFS_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_mutation_signature,end=-19)]

          mutation_signature_PFS_plots_metadata <- NULL
          for (k in 1:length(input$data_source_mutation_signature)) {
            my_i <- k
            d<-splitted_list[[my_i]]
            #isolate({
            if (is.na(names(splitted_list)[my_i])) {
              tmp <- data.frame(data_source =input$data_source_mutation_signature[my_i],
                                event.top=NA,
                                n.top=NA,
                                event.bottom=NA,
                                n.bottom=NA)
              mutation_signature_PFS_plots_metadata <- rbind(mutation_signature_PFS_plots_metadata,tmp)
            } else {
              if (nrow(d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                tmp <- data.frame(data_source =input$data_source_mutation_signature[my_i],
                                  event.top=NA,
                                  n.top=NA,
                                  event.bottom=NA,
                                  n.bottom=NA)
                mutation_signature_PFS_plots_metadata <- rbind(mutation_signature_PFS_plots_metadata,tmp)
              } else {
                d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                d <- data.frame(data_source=as.character(d$data_source),
                                PFS.status=as.character(d$PFS.status),
                                PFS=as.numeric(d$PFS),
                                Age=as.character(d$Age),
                                Gender=as.character(d$Gender),
                                Signature=as.numeric(d[,"Signature"]))
                # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Signature")]))
                d<-data.frame(d,IND=paste0("bottom ",input$slider_mutation_signature_OS*100, "% (",
                                           nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_mutation_signature_OS),]),
                                           ")"))
                d[,"IND"]<-as.character(d[,"IND"])
                d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),"IND"]<-
                  paste0("top ",(1-input$slider_mutation_signature_OS)*100,"% (",
                         nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),])
                         ,")")
                d <- data.frame(d,group.IND=0)
                d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                #        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                d<-d[order(d[,"IND"]),]

                tmp <- data.frame(data_source =input$data_source_mutation_signature[my_i],
                                  event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                  n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                  event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                  n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                mutation_signature_PFS_plots_metadata <- rbind(mutation_signature_PFS_plots_metadata,tmp)


                # if (input_adjust_OS_mutation_signature=="None") {
                #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                #   if (class(diff.result.try)=="try-error") {
                #     p.value <- NA
                #   } else {
                #     p.value <- signif(summary(diff.result)$logtest[3],2)
                #   }
                # } else {
                #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_mutation_signature),data=d) )
                #   if (class(diff.result.adjust.try)=="try-error") {
                #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                #
                #   } else {
                #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                #   }
                # }
                # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                # mutation_signature_PFS_plots_combined_p <- c(mutation_signature_PFS_plots_combined_p,p.value)
              }
            }
            #})
          }
          if (all(is.na(mutation_signature_PFS_plots_metadata))) {
            output$mutation_signature_PFS_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$mutation_signature_PFS_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$mutation_signature_PFS_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$mutation_signature_PFS_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=mutation_signature_PFS_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=mutation_signature_PFS_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
            })
            # output$mutation_signature_PFS_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(mutation_signature_PFS_plots_combined_p)))$p.value,2)
            # })
            # output$mutation_signature_PFS_plots_single_p <- renderText({
            #   paste0(signif(mutation_signature_PFS_plots_combined_p,2),collapse = ", ")
            # })
          }


          for (k in 1:length(input$data_source_mutation_signature)) {
            local({
              my_i <- k
              mutation_signature_PFS_plots_name <- paste0("mutation_signature_PFS_plots", my_i)
              mutation_signature_PFS_summary_tables_name <- paste0("mutation_signature_PFS_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              d[,"Signature"] <- as.numeric(as.character(d[,"Signature"]))
              #output[[mutation_signature_PFS_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                } else {

                  if (nrow(d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {
                    d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                    d <- data.frame(data_source=as.character(d$data_source),
                                    PFS.status=as.character(d$PFS.status),
                                    PFS=as.numeric(d$PFS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    Signature=as.numeric(d[,"Signature"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Signature")]))
                    d<-data.frame(d,IND=paste0("bottom ",
                                               input$slider_mutation_signature_OS*100,
                                               "% (",
                                               nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_mutation_signature_OS),]),
                                               ")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),"IND"]<-
                      paste0("top ",
                             (1-input$slider_mutation_signature_OS)*100,
                             "% (",
                             nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_mutation_signature_OS),])
                             ,")")
                    d <- data.frame(d,group.IND=0)
                    d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1

                    d<-d[order(d[,"IND"]),]
                    fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                      if (input_adjust_OS_mutation_signature=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_mutation_signature),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_mutation_signature," was observed)")

                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_mutation_signature
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                    # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                    p1<-ggplotly(plot.result.ann,height = 480 ) %>%
                      layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                             yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                             legend=list(orientation="v",x=0.6,y=1)) #%>%
                    incProgress(100, detail = "Doing part %")
                    p1
                    }
                  }
                }
              })
              })
              #})
                output[[mutation_signature_PFS_plots_name]] <- renderPlotly({
                  p1
                })
                output[[mutation_signature_PFS_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })
    

    
    ###########################################
    #### expression
    ###########################################
    #input <- list(data_source_expression=c("Melanoma-Chen-Pre-On"),treatment_expression=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression=c("Pre","On"),gene_expression="CD8A",slider_expression_OS=0.5,slider_expression_PFS=0.5,clinical_expression="treatment type",dynamic_check_expression=TRUE)
    #input <- list(data_source_expression=c("Melanoma-Auslander","Melanoma-Chen","Melanoma-Hugo","Melanoma-Prat","Melanoma-Riaz","Melanoma-Van Allen"),treatment_expression=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression=c("Pre","on"),gene_expression="CD19",slider_expression_OS=0.5,slider_expression_PFS=0.5,clinical_expression="treatment type")
    #input <- list(data_source_expression=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)","Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)"),treatment_expression=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression="Dynamic: (On-Pre)",gene_expression="CD19",slider_expression_OS=0.5,slider_expression_PFS=0.5,clinical_expression="treatment type")
    #input <- list(data_source_expression=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)","Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)"),treatment_expression=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),dynamic_check_expression=TRUE,description_expression="Dynamic: (On-Pre)",gene_expression="CD19",slider_expression_OS=0.5,slider_expression_PFS=0.5,clinical_expression="treatment type")
    
    ## due to dynamic feature, please keep two observe({}), and keep if(input$dynamic_check_expression) {update...description_expression} at the top of this file
    observe({
    updateSelectizeInput(session,"gene_expression",
                         choices=a.colnames,selected="CD19",
                         server=TRUE)
    })
    observe({
      if (input$dynamic_check_expression) {

        updateSelectizeInput(session=session,inputId="data_source_expression",
                             choices=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                             selected=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                             server=T)

      }
    })

    observe({
      if (input$data_source_expression=="" || is.null(input$data_source_expression) ||
          input$treatment_expression=="" || is.null(input$treatment_expression) ||
          input$description_expression=="" || is.null(input$description_expression) ||
          input$gene_expression=="" || is.null(input$gene_expression)){
        disable("do_expression")
      } else {
        enable("do_expression")
      }
    })
#     observeEvent(input$do_expression,{
#       if (input$gene_expression=="ALL") {
#         hideTab(inputId="expression_tabs",target="Immunotherapy response")
#         hideTab(inputId="expression_tabs",target="Overall survival")
#         hideTab(inputId="expression_tabs",target="Progression free survival")
#         hideTab(inputId="expression_tabs",target="Evaluation")
#         shinyjs::hideElement(id="expression_clinical_attribute")
# #        shinyjs::hideElement(id="clinical_expression")
# #        tabsetPanel(tabPanel("Summary",h1("test")))
#       } else {
#         showTab(inputId="expression_tabs",target="Immunotherapy response")
#         showTab(inputId="expression_tabs",target="Overall survival")
#         showTab(inputId="expression_tabs",target="Progression free survival")
#         showTab(inputId="expression_tabs",target="Evaluation")
#         shinyjs::showElement(id="expression_clinical_attribute")
#       }
#     })
    
    selecteddatabase_expression <- eventReactive(input$do_expression,{
      if (input$dynamic_check_expression) {
        user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
        
        #data_source1 <- input$data_source_expression[input$data_source_expression %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
        data_source1 <- str_sub(input$data_source_expression,end=-19)[str_sub(input$data_source_expression,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
        #treatment1 <- input$treatment_expression
        treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_expression])
        description1 <- c("Pre","On")
        a<-data.table(database.test.process())
        #a<-data.table(database.test)
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_expression)
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        colnames(m)[ncol(m)] <- "UserGene"
        m <- as.data.frame(m)
        if (nrow(m) >0) {
          samples <- m$PatientID
          samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
          samples_paired <- table(samples)
          samples_paired <- samples_paired[samples_paired==2]
          samples_paired_pre <- paste0(names(samples_paired),"Pre")
          #samples_paired_on <- paste0(names(samples_paired),"On")
          samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
          m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],UserGene=m[m[,"PatientID"] %in% samples_paired_on,"UserGene"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserGene"]),
                      data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],UserGene=m[m[,"PatientID"] %in% samples_paired_on,"UserGene"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserGene"]))
          m<-data.frame(m,category="unknown")
          m[,"category"] <- as.character(m[,"category"])
          m1 <- m %>% split(.$data_source) %>% 
            lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_expression," expression",sep="")
              d[!is.na(d[,"UserGene"]) & d[,"UserGene"]>=quantile(na.omit(d[,"UserGene"]),input$slider_expression_OS),"category"] <- paste("top ",(1-input$slider_expression_OS)*100,"% ",sep="")
              d[!is.na(d[,"UserGene"]) & d[,"UserGene"]<quantile(na.omit(d[,"UserGene"]),input$slider_expression_OS),"category"] <- paste("bottom ",input$slider_expression_OS*100,"% ",sep="")
              d})
          m2<-ldply(m1,data.frame)
          m2<-m2[,-1]
          m2$UserGene <- as.numeric(m2$UserGene)
          m2
        } else {
          m
        }
        
        
        
      } else {
      
      data_source1 <- str_sub(input$data_source_expression,end=-19)
      #treatment1 <- input$treatment_expression
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_expression])
      description1 <- input$description_expression
      a<-data.table(database.test.process())
      #a<-data.table(database.test)
      m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
      genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_expression)
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        colnames(m)[ncol(m)] <- "UserGene"
        m <- as.data.frame(m)
        if (nrow(m)>0) {
          m<-data.frame(m,category="unknown")
          m[,"category"] <- as.character(m[,"category"])
          m1 <- m %>% split(.$data_source) %>% 
            lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_expression," expression",sep="")
              d[!is.na(d[,"UserGene"]) & d[,"UserGene"]>=quantile(na.omit(d[,"UserGene"]),input$slider_expression_OS),"category"] <- paste("top ",(1-input$slider_expression_OS)*100,"% ",sep="")
              d[!is.na(d[,"UserGene"]) & d[,"UserGene"]<quantile(na.omit(d[,"UserGene"]),input$slider_expression_OS),"category"] <- paste("bottom ",input$slider_expression_OS*100,"% ",sep="")
              d})
          m2<-ldply(m1,data.frame)
          m2<-m2[,-1]
          m2$UserGene <- as.numeric(m2$UserGene)
          m2
        } else {
          m2<-m
          m2
        }
      }
    })
    
    # selecteddatabase_expression_ALL <- eventReactive(input$do_expression,{
    #   data_source1 <- input$data_source_expression
    #   treatment1 <- input$treatment_expression
    #   description1 <- input$description_expression
    #   a<-data.table(database.test.process())
    #   m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
    #   if (input$gene_expression=="ALL") {
    #     m<-as.data.frame(m)
    #    
    #   m.data_source <- as.character(unique(m[,"data_source"]))
    #   m.result <- NULL
    #   
    #   for (k in 1:length(m.data_source)) {
    #     d <- m[m[,"data_source"]==m.data_source[k],]
    #     d.res <- d[d[,"PFS.status"]=="response",]
    #     d.nonres <- d[d[,"PFS.status"]=="nonresponse",]
    #     d.p.value <- NULL
    #     for (j in 13:ncol(d)) {
    #       try.test <- try(p.v <- wilcox.test(as.numeric(na.omit(d.res[,j])),as.numeric(na.omit(d.nonres[,j])))$p.value)
    #       if (class(try.test)=="try-error") {
    #         p.v <- "NA"
    #       } else {
    #         p.v <- round(wilcox.test(as.numeric(na.omit(d.res[,j])),as.numeric(na.omit(d.nonres[,j])))$p.value,3)
    #       }
    #       d.p.value <- c(d.p.value,p.v)
    #     }
    #     m.result <- rbind(m.result,d.p.value)
    #   }
    #   colnames(m.result) <- colnames(m)[13:ncol(m)]
    #   m.result <- data.frame(data_source=m.data_source,m.result)
    #   m.result
    #   }
    # })
    
    ################
    #### expression summary
    ################
    output$expression_summary_disease_number <- renderText({
      input$do_expression ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selecteddatabase_expression()$Disease))})})
    output$expression_summary_sample_number <- renderText({
      input$do_expression ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selecteddatabase_expression())})})
    output$expression_summary_plots <- renderPlotly({
      summary_plot(inputdata=selecteddatabase_expression(),feature="expression")
    })
    

    
    observeEvent(input$do_expression,{
      input$do_expression
      isolate({
        input_gene_expression <- input$gene_expression
        # if (input_gene_expression=="ALL") { 
        #   selected.database <- selecteddatabase_expression_ALL()
        # } else {
          selected.database <- selecteddatabase_expression()
        # }
        if (nrow(selected.database)==0 ) {
          output[["expression_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          # if (input_gene_expression=="ALL") {
          #   # m<-selected.database
          #   # m.data_source <- as.character(unique(m[,"data_source"]))
          #   # m.result <- NULL
          #   # 
          #   # for (k in 1:length(m.data_source)) {
          #   #   d <- m[m[,"data_source"]==m.data_source[k],]
          #   #   d.res <- d[d[,"PFS.status"]=="response",]
          #   #   d.nonres <- d[d[,"PFS.status"]=="nonresponse",]
          #   #   d.p.value <- NULL
          #   #   for (j in 13:ncol(d)) {
          #   #     try.test <- try(p.v <- wilcox.test(as.numeric(na.omit(d.res[,j])),as.numeric(na.omit(d.nonres[,j])))$p.value)
          #   #     if (class(try.test)=="try-error") {
          #   #       p.v <- "NA"
          #   #     } else {
          #   #       p.v <- round(wilcox.test(as.numeric(na.omit(d.res[,j])),as.numeric(na.omit(d.nonres[,j])))$p.value,3)
          #   #     }
          #   #     d.p.value <- c(d.p.value,p.v)
          #   #   }
          #   #   m.result <- rbind(m.result,d.p.value)
          #   # }
          #   # colnames(m.result) <- colnames(m)[13:ncol(m)]
          #   # m.result <- data.frame(data_source=m.data_source,m.result)
          #     splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          #    splitted_list <- splitted_list[input$data_source_expression]
          #    for (k in 1:length(input$data_source_expression)) {
          #   #for (k in 1:nrow(selected.database)) {
          #     local({
          #       my_i <- k
          #       expression_summary_clinical_plots_name <- paste0("expression_summary_clinical_plots", my_i)
          #       d<-splitted_list[[my_i]]
          #       
          #       output[[expression_summary_clinical_plots_name]] <- renderPlotly({
          #         withProgress(message = 'Making plot', value = 0, {
          #           if (is.na(names(splitted_list)[my_i])) {
          #             incProgress(100, detail = "Doing part %")
          #             plot_ly () %>% 
          #               layout(title = input$data_source_expression[[my_i]],font=list(size=15,family="Aril")) %>%
          #               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          #           } else {
          #             if (length(na.omit(d[2:length(d)]) )==0) {
          #               incProgress(100, detail = "Doing part %")
          #               plot_ly () %>% 
          #                 layout(title = unique(d$data_source),font=list(size=15,family="Aril")) %>%
          #                 add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          #             } else {
          #               # d <- d[!is.na(d[,clinical_server[input$clinical_expression]]) & !is.na(d[,"UserGene"]),]
          #               # if(input$clinical_expression=="Age") {
          #               #   d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
          #               #   #d[d[,"Age"]<=10,"Age"] <- "0-10"
          #               #   d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
          #               #   d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
          #               # }
          #               
          #               #d <- data.frame(Genes=names(d)[-1],p.values=d[-1])
          #               vline <- function(x = 0, color = "red") {
          #                 list(
          #                   type = "line", 
          #                   y0 = 0, 
          #                   y1 = 1, 
          #                   yref = "paper",
          #                   x0 = x, 
          #                   x1 = x, 
          #                   line = list(color = color)
          #                 )
          #               }
          #               d <- d[d!="NA"]
          #               d.nonna <- -log10(as.numeric(d[-1]))
          #               fit <- density(d.nonna)
          #               p<-plot_ly(x=d.nonna,
          #                          #y=as.numeric(d[-1]), 
          #                          type="histogram",#type="box",
          #                          #histnorm="probability",
          #                          name="Histogram"
          #                          ) %>%
          #                 add_trace(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",yaxis="y2",name="Density") %>%
          #                 layout(title=unique(d[1]),
          #                        height = 480,
          #                        xaxis = list(title="-log10(p.value)",
          #                                     titlefont=list(size=20,color="black",family= "Aril black"),
          #                                     showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
          #                        yaxis=list(title="Histogram number",
          #                                   titlefont=list(size=20,color="black",family= "Aril black"),
          #                                   showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
          #                        shapes=list(vline(1.3)),
          #                        yaxis2=list(title="Density",overlaying="y",side="right")) %>%
          #                 add_annotations(x = 1.3,
          #                                 y2 = 1,
          #                                 text = "p.value = 0.05",
          #                                 xref = paste0("x", 1),  # add this
          #                                 yref = paste0("y", 1),  # add this
          #                                 showarrow = FALSE,
          #                                 ax = 20,
          #                                 ay = -40,
          #                                 font=list(size=20,color="black",family= "Aril black"))
          #                 # layout(title = unique(d$data_source),titlefont=list(size=20,color="black",family= "Aril"),
          #                 #        height = 480,
          #                 #        xaxis=list(title="Clinical Attribute",
          #                 #                   titlefont=list(size=20,color="black",family= "Aril black"),
          #                 #                   showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
          #                 #        yaxis=list(title=paste(input_gene_expression, " expression value",sep=""),
          #                 #                   titlefont=list(size=20,color="black",family= "Aril black"),
          #                 #                   showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
          #                 #        ))
          #               incProgress(100, detail = "Doing part %")
          #               p
          #               
          #             } 
          #           }
          #         })
          #       })
          #     })
          #   }
          # } else {
          
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression,end=-19)]
          for (k in 1:length(input$data_source_expression)) {
            local({
              my_i <- k
              expression_summary_clinical_plots_name <- paste0("expression_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[expression_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_expression]]) & !is.na(d[,"UserGene"]),])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_expression]]) & !is.na(d[,"UserGene"]),]
                      if (input$dynamic_check_expression) {
                        d <- subset(d,description=="Pre")
                      }
                      if(input$clinical_expression=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        #d[d[,"Age"]<=10,"Age"] <- "0-10"
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      
                      if (input$dynamic_check_expression) {
                        p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_expression]],
                                   y=~signif(as.numeric(as.character(UserGene)),3),  
                                   type="violin",#type="box",
                                   points="all",#boxpoints="all",
                                   jitter=0.3,
                                   pointpos=-1.8,
                                   #mode="markders",
                                   color=d[,clinical_server[input$clinical_expression]],showlegend=TRUE,
                                   colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                   hoveron="points+boxes",
                                   legendgroup=d[,clinical_server[input$clinical_expression]], 
                                   text= ~paste0("Sample ID: ", PatientID,
                                                 "<br>Changes of expression: ",input$gene_expression," :",signif(as.numeric(as.character(UserGene)),3), 
                                                 "<br>Treatment: ",treatment,
                                                 "<br>Description: ",description)) %>%
                          layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                 height = 480,
                                 xaxis=list(title="Clinical Attribute",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title=paste("Changes of ",input_gene_expression, " expression value",sep=""),
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                 ))
                        
                      } else {
                          p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_expression]],
                                      y=~signif(as.numeric(as.character(UserGene)),3), 
                                      type="violin",#type="box",
                                      points="all",#boxpoints="all",
                                      jitter=0.3,
                                      pointpos=-1.8,
                                      #mode="markders",
                                      color=d[,clinical_server[input$clinical_expression]],showlegend=TRUE,
                                      colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                      hoveron="points+boxes",
                                      legendgroup=d[,clinical_server[input$clinical_expression]], 
                                      text= ~paste0("Sample ID: ", PatientID,
                                                    "<br>Expression: ",input$gene_expression," :",signif(as.numeric(as.character(UserGene)),3), 
                                                    "<br>Treatment: ",treatment,
                                                    "<br>Description: ",description)) %>%
                              layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                 height = 480,
                                 xaxis=list(title="Clinical Attribute",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title=paste(input_gene_expression, " expression value",sep=""),
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                 ))
                      }
                      incProgress(100, detail = "Doing part %")
                      p

                    } 
                  }
                })
              })
            })
          }
         # }
        }
      }) 
    })
    
    output$expression_summary_download <- downloadHandler(
      filename = function() {
        if (input$expression_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$expression_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$expression_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        # if (input$gene_expression=="ALL") {
        #   
        #   d <- t(selecteddatabase_expression_ALL())
        #   colnames(d) <- d[1,]
        #   #colnames(d)[1] <- "Genes"
        #   d <- d[-1,]
        #   
        #   if (input$expression_summary_type == "Excel (CSV)") {
        #     write.csv (d,file,row.names = FALSE)
        #   } else if (input$expression_summary_type == "Text (TSV)") {
        #     write.table (d,file,quote=F, row.names = FALSE, sep="\t")
        #   } else if (input$expression_summary_type == "JSON") {
        #     write(toJSON(d),file)
        #   }
        # } else {
        selected.database <- selecteddatabase_expression()
        if (input$dynamic_check_expression) {
          colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- paste0("Changes of ",input$gene_expression)
        } else {
          colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- input$gene_expression
        }
          if (input$expression_summary_type == "Excel (CSV)") {
            write.csv (selected.database,file,row.names = FALSE)
          } else if (input$expression_summary_type == "Text (TSV)") {
            write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
          } else if (input$expression_summary_type == "JSON") {
            write(toJSON(selected.database),file)
          }
        #}
      }
    )
    
    output$expression_summary_table <- DT::renderDataTable({
      # if (input$gene_expression=="ALL") {
      #   d <- t(selecteddatabase_expression_ALL())
      #   colnames(d) <- d[1,]
      #   #colnames(d)[1] <- "Genes"
      #   d <- d[-1,]
      #   d
      # } else {
      selected.database <- selecteddatabase_expression()
      if (input$dynamic_check_expression) {
        colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- paste0("Changes of ",input$gene_expression)
      } else {
        colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- input$gene_expression
      }
      
      selected.database
      #}
    })
    
    ################
    #### expression plots
    ################
    
    observeEvent(input$do_expression,{
      input$do_expression

      isolate({
        input_gene_expression <- input$gene_expression
        selected.database <- selecteddatabase_expression()
        if (nrow(selected.database)==0 ) {
          output$expression_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["expression_plots_plots0"]] <- renderPlotly({
            #show("loading-c")
            #hide("loading-d")
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })
          # output$expression_plots_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_plots_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression,end=-19)]

          # expression_plots_plots_combined_p <- NULL
          expression_plots_plots_metadata <- NULL#data.frame(TE=NA,seTE=NA)
          for (k in 1:length(input$data_source_expression)) {
            my_i <- k
            d <- splitted_list[[my_i]]
            if (is.na(names(splitted_list)[my_i])) {
              tmp <- data.frame(data_source=input$data_source_expression[my_i],TE=NA,seTE=NA)
            } else {
              if (nrow(d)==0) {
                tmp = data.frame(data_source=input$data_source_expression[my_i],TE=NA,seTE=NA)
              } else {
                #d <- d[!is.na(d$PFS.status) & !is.na(d$UserGene),]
                if (nrow(d[!is.na(d$PFS.status) & !is.na(d$UserGene),])==0) {
                  tmp = data.frame(data_source=input$data_source_expression[my_i],TE=NA,seTE=NA)
                } else {
                  d$response_binary <- NA
                  d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                  d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                  d$scaled <- scale(as.numeric(d$UserGene))
                  if (all(is.na(d$scaled))) {
                    tmp = data.frame(data_source=input$data_source_expression[my_i],TE=NA,seTE=NA)
                  } else {
                    tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                    if (nrow(tmp.pre$coefficients)>1) {
                      tmp <- data.frame(data_source=input$data_source_expression[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                    } else {
                      tmp = data.frame(data_source=input$data_source_expression[my_i],TE=NA,seTE=NA)
                      
                    }
                  }
                }
              }
            }
            expression_plots_plots_metadata <- rbind.fill(expression_plots_plots_metadata,tmp)

          }
          # expression_plots_plots_metadata$TE <- as.numeric(expression_plots_plots_metadata$TE)
          # expression_plots_plots_metadata$seTE <- as.numeric(expression_plots_plots_metadata$seTE)
          #
          # for (k in 1:length(input$data_source_expression)) {
          #     my_i <- k
          #     d<-splitted_list[[my_i]]
          #     d$UserGene <- as.numeric(d$UserGene)
          #     d.res <- d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),]
          #     d.nonres <- d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),]
          #         #show("loading-c")
          #
          #     if (nrow(d.res)>1 & nrow(d.nonres)>1) {
          #       tmp <- data.frame(data_source = as.character(unique(d$data_source)),
          #                         n.res=length(na.omit(d.res[,"UserGene"])),
          #                         mean.res=mean(na.omit(d.res[,"UserGene"])),
          #                         sd.res=sd(na.omit(d.res[,"UserGene"])),
          #                         n.nonres=length(na.omit(d.nonres[,"UserGene"])),
          #                         mean.nonres=mean(na.omit(d.nonres[,"UserGene"])),
          #                         sd.nonres=sd(na.omit(d.nonres[,"UserGene"])))
          #       expression_plots_plots_metadata <- rbind(expression_plots_plots_metadata,tmp)
          #     } else {
          #       tmp <- data.frame(data_source = as.character(unique(d$data_source)),
          #                         n.res=NA,
          #                         mean.res=NA,
          #                         sd.res=NA,
          #                         n.nonres=NA,
          #                         mean.nonres=NA,
          #                         sd.nonres=NA)
          #       expression_plots_plots_metadata <- rbind(expression_plots_plots_metadata,tmp)
          #     }
          #
          #         # if (is.na(names(splitted_list)[my_i])) {
          #         #   #hide("loading-d")
          #         #   expression_plots_plots_combined_p <- c(expression_plots_plots_combined_p,NA)
          #         # } else {
          #         #   if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),])==0 ) {
          #         #     expression_plots_plots_combined_p <- c(expression_plots_plots_combined_p,NA)
          #         #   } else {
          #         #     d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),]
          #         #     if (input$dynamic_check_expression) {
          #         #       d <- subset(d,description=="Pre")
          #         #     }
          #         #     d[,"Groups"] <- as.character(d[,"Groups"])
          #         #     d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
          #         #     d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
          #         #     ind <- sort(unique(d[,"Groups"]))
          #         #     try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserGene"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserGene"])))$p.value)
          #         #     if (class(try.test)=="try-error" | is.na(try.test)) {
          #         #       expression_plots_plots_combined_p <- c(expression_plots_plots_combined_p,NA)
          #         #     } else {
          #         #       expression_plots_plots_combined_p <- c(expression_plots_plots_combined_p,p.v)                      }
          #         #
          #         #
          #         #   }
          #         #
          #         # }
          # }
          if(all(is.na(expression_plots_plots_metadata))) {
            output$expression_plots_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_plots_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_plots_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_plots_plots_meta <- renderPlot({
              m1 <- metagen(TE,seTE,
                            data=expression_plots_plots_metadata,
                            sm="OR",
                            #label.e = "response",label.c = "nonresponse",
                            studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,
              #                 data=expression_plots_plots_metadata,
              #                sm="OR",
              #                 #label.e = "response",label.c = "nonresponse",
              #                 studlab = data_source),
              #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))

              # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
              #                 data=expression_plots_plots_metadata,
              #                 label.e = "response",label.c = "nonresponse",
              #                 studlab = data_source),
              #        test.overall = TRUE,ff.study.label = "bold")
            })
            # output$expression_plots_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_plots_plots_combined_p)))$p.value,2)
            # })
            # output$expression_plots_plots_single_p <- renderText({
            #   paste0(signif(expression_plots_plots_combined_p,2),collapse = ", ")
            # })
          }



          for (k in 1:length(input$data_source_expression)) {
            local({
              my_i <- k
              expression_plots_plots_name <- paste0("expression_plots_plots", my_i)
              d<-splitted_list[[my_i]]

              output[[expression_plots_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                #show("loading-c")
                if (is.na(names(splitted_list)[my_i])) {
                  #hide("loading-d")
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>%
                    layout(title = str_sub(input$data_source_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),])==0 ) {
                    #hide("loading-d")
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),]
                    if (input$dynamic_check_expression) {
                      d <- subset(d,description=="Pre")
                    }
                    #if (input$dynamic_check_expression) {

                    # "Gide_ipiPD1_1_Pre","Gide_ipiPD1_13_Pre","Gide_ipiPD1_19_Pre","Gide_ipiPD1_27_Pre",
                    # "Gide_ipiPD1_33_Pre","Gide_ipiPD1_35_Pre","Gide_ipiPD1_36_Pre","Gide_PD1_10_Pre","Gide_PD1_13_Pre","Gide_PD1_17_Pre","Gide_PD1_25_Pre",
                    # "Gide_PD1_29_Pre","Gide_PD1_41_Pre","Gide_PD1_47_Pre","Gide_PD1_48_Pre","Gide_PD1_8_Pre","Gide_ipiPD1_1_On","Gide_ipiPD1_13_On",
                    # "Gide_ipiPD1_19_On","Gide_ipiPD1_27_On","Gide_ipiPD1_33_On","Gide_ipiPD1_35_On","Gide_ipiPD1_36_On","Gide_PD1_10_On","Gide_PD1_13_On",
                    # "Gide_PD1_17_On","Gide_PD1_25_On","Gide_PD1_29_On","
                    #   Gide_PD1_41_On","Gide_PD1_47_On","Gide_PD1_48_On","Gide_PD1_8_On",

#                       preandon_paired <- c("Riaz_Pt1_Pre","Riaz_Pt10_Pre","Riaz_Pt101_Pre",
# "Riaz_Pt103_Pre","Riaz_Pt106_Pre","Riaz_Pt11_Pre","Riaz_Pt17_Pre","Riaz_Pt18_Pre","Riaz_Pt2_Pre","Riaz_Pt23_Pre","Riaz_Pt26_Pre",
# "Riaz_Pt27_Pre","Riaz_Pt28_Pre","Riaz_Pt3_Pre","Riaz_Pt3.2_Pre","Riaz_Pt30_Pre","Riaz_Pt31_Pre","Riaz_Pt34_Pre","Riaz_Pt36_Pre",
# "Riaz_Pt37_Pre","Riaz_Pt37.2_Pre","Riaz_Pt38_Pre","Riaz_Pt4_Pre","Riaz_Pt44_Pre","Riaz_Pt46_Pre","Riaz_Pt47_Pre",
# "Riaz_Pt48_Pre","Riaz_Pt49_Pre","Riaz_Pt5_Pre","Riaz_Pt5.2_Pre","Riaz_Pt52_Pre",
# "Riaz_Pt62_Pre","Riaz_Pt65_Pre","Riaz_Pt67_Pre","Riaz_Pt77_Pre","Riaz_Pt78_Pre","Riaz_Pt79_Pre","Riaz_Pt8_Pre","Riaz_Pt82_Pre",
# "Riaz_Pt84_Pre","Riaz_Pt85_Pre","Riaz_Pt89_Pre","Riaz_Pt9_Pre","Riaz_Pt92_Pre","Riaz_Pt92.2_Pre","Riaz_Pt94_Pre","Riaz_Pt98_Pre",
# "Riaz_Pt1_On","Riaz_Pt10_On","Riaz_Pt101_On","Riaz_Pt103_On","Riaz_Pt106_On","Riaz_Pt11_On","Riaz_Pt17_On","Riaz_Pt18_On",
# "Riaz_Pt2_On","Riaz_Pt23_On","Riaz_Pt26_On","Riaz_Pt27_On",
# "Riaz_Pt28_On","Riaz_Pt3_On","Riaz_Pt3.2_On","Riaz_Pt30_On","Riaz_Pt31_On","Riaz_Pt34_On","Riaz_Pt36_On",
# "Riaz_Pt37_On","Riaz_Pt37.2_On","Riaz_Pt38_On","Riaz_Pt4_On","Riaz_Pt44_On","Riaz_Pt46_On","Riaz_Pt47_On",
# "Riaz_Pt48_On","Riaz_Pt49_On","Riaz_Pt5_On","Riaz_Pt5.2_On","Riaz_Pt52_On","Riaz_Pt62_On","Riaz_Pt65_On",
# "Riaz_Pt67_On","Riaz_Pt77_On","Riaz_Pt78_On","Riaz_Pt79_On","Riaz_Pt8_On","Riaz_Pt82_On","Riaz_Pt84_On",
# "Riaz_Pt85_On","Riaz_Pt89_On","Riaz_Pt9_On","Riaz_Pt92_On","Riaz_Pt92.2_On","Riaz_Pt94_On","Riaz_Pt98_On",
# "Riaz_Pt89_On","Riaz_Pt9_On","Riaz_Pt92_On","Riaz_Pt92.2_On","Riaz_Pt94_On","Riaz_Pt98_On")
#
#
#
#
#                       d <- d[d[,"PatientID"] %in% preandon_paired,]
                    #}
                    d[,"Groups"] <- as.character(d[,"Groups"])
                    d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    ind <- sort(unique(d[,"Groups"]))
                    if (as.character(unique(d$data_source)) == "Melanoma-Van Allen" | as.character(unique(d$data_source)) == "Melanoma-Riaz" | as.character(unique(d$data_source)) == "Melanoma-Auslander") {
                      lable_y <- " (RSEM)"
                    } else if (as.character(unique(d$data_source)) == "Melanoma-Chen" | as.character(unique(d$data_source)) == "Melanoma-Prat") {
                      lable_y <- " (NanoString signal)"
                    } else if (as.character(unique(d$data_source)) == "Melanoma-Hugo" | as.character(unique(d$data_source)) == "Melanoma-Gide" | as.character(unique(d$data_source)) == "Melanoma-Snyder-Nathanson") {
                      lable_y <- " (FPKM)"
                    } else {
                      lable_y <- NULL
                    }


                    try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserGene"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserGene"])))$p.value)
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      p.v <- " = NA"
                    } else {
                      # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserGene"]),as.numeric(d[d[,"Groups"]==ind[2],"UserGene"]))$p.value
                      # if (p.v < 0.001) {
                      #   p.v <- "< 0.001"
                      # } else {
                        p.v<-paste0(" = ",signif(p.v,2))
                      # }
                    }

                    if (input$dynamic_check_expression) {
                      p<-plot_ly(data=d,x=~Groups,
                                 y=~signif(as.numeric(as.character(UserGene)),3),
                                 type="violin",#type="box",
                                 points="all",#boxpoints="all",
                                 jitter=0.3,
                                 pointpos=-1.8,
                                 #mode="markders",
                                 color=~PFS.status,showlegend=TRUE,
                                 colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                 #hoveron="points+boxes",
                                 legendgroup=~PFS.status,
                                 text= ~paste0("Sample ID: ", PatientID,
                                               "<br>Change of ",input_gene_expression," expression:",signif(as.numeric(as.character(UserGene)),3),
                                               "<br>Treatment: ",treatment))  %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title=" ",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title=paste("Change of ",input_gene_expression, " expression value",lable_y,sep=""),
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               )) %>%
                        add_annotations(x = 0.5,
                                        y = d[which.max(d[,"UserGene"]),"UserGene"],
                                        text = paste0("p",p.v),
                                        xref = paste0("x", 1),  # add this
                                        yref = paste0("y", 1),  # add this
                                        showarrow = FALSE,
                                        ax = 20,
                                        ay = -40,
                                        font=list(size=20,color="black",family= "Aril black"))
                    } else {
                        p<-plot_ly(data=d,x=~Groups,
                               y=~signif(as.numeric(as.character(UserGene)),3),
                               type="violin",#type="box",
                               points="all",#boxpoints="all",
                               jitter=0.3,
                               pointpos=-1.8,
                               #mode="markders",
                               color=~PFS.status,showlegend=TRUE,
                               colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                               #hoveron="points+boxes",
                               legendgroup=~PFS.status,
                               text= ~paste0("Sample ID: ", PatientID,
                                             "<br>Expression: ",input_gene_expression," :",signif(as.numeric(as.character(UserGene)),3),
                                             "<br>Treatment: ",treatment))  %>%
                          layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             height = 480,
                             xaxis=list(title=" ",
                                        titlefont=list(size=20,color="black",family= "Aril black"),
                                        showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                             yaxis=list(title=paste(input_gene_expression, " expression value",lable_y,sep=""),
                                        titlefont=list(size=20,color="black",family= "Aril black"),
                                        showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                             )) %>%
                          add_annotations(x = 0.5,
                                      y = d[which.max(d[,"UserGene"]),"UserGene"],
                                      text = paste0("p",p.v),
                                      xref = paste0("x", 1),  # add this
                                      yref = paste0("y", 1),  # add this
                                      showarrow = FALSE,
                                      ax = 20,
                                      ay = -40,
                                      font=list(size=20,color="black",family= "Aril black"))
                    }
                    #hide("loading-d")
                    incProgress(100, detail = "Doing part %")
                    p
                  }

                }
              })
              })
            })
          }
        }
      })

    })
#    hide("loading-d")

    ################
    #### expression survival
    ################

    observeEvent(input$do_expression,{
      input$do_expression
      #if (input$gene_expression!="ALL") {
      isolate({
        input_gene_expression <- input$gene_expression
        selected.database <- selecteddatabase_expression()
        input_adjust_OS_expression <- input$adjust_OS_expression
        if (nrow(selected.database)==0 | ncol(selected.database)<13) {
          output$expression_survival_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["expression_survival_plots0"]] <- renderPlotly({
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression,end=-19)]

          # expression_survival_plots_combined_p <- NULL
          expression_survival_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression)) {
              my_i <- k
              expression_survival_plots_name <- paste0("expression_survival_plots", my_i)

              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # expression_survival_plots_combined_p <- c(expression_survival_plots_combined_p,NA)
                      # tmp <- data.frame(data_source =as.character(unique(d$data_source)),
                      tmp <- data.frame(data_source =input$data_source_expression[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      expression_survival_plots_metadata <- rbind(expression_survival_plots_metadata,tmp)

                    } else {
                      if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                        # expression_survival_plots_combined_p <- c(expression_survival_plots_combined_p,NA)
                        # tmp <- data.frame(data_source =as.character(unique(d$data_source)),
                        tmp <- data.frame(data_source =input$data_source_expression[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        expression_survival_plots_metadata <- rbind(expression_survival_plots_metadata,tmp)

                      } else {
                        if (input$dynamic_check_expression) {
                          d <- subset(d,description=="Pre")
                        }
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),]
                        d<-data.frame(d,dead=NA)
                        d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                        d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                        d <- d[!is.na(d[,"UserGene"]) & !is.na(d[,"dead"]),]
                        d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                        #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserGene")]))
                        d <- data.frame(data_source=as.character(d$data_source),
                        
                                        dead=as.numeric(d$dead),
                                        OS=as.numeric(d$OS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        UserGene=as.numeric(d[,"UserGene"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserGene")]))
                        d<-data.frame(d,IND=paste("bottom ",input$slider_expression_OS*100,"% (",
                                                  nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                                                  sep=""))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),"IND"]<-
                          paste("top ",(1-input$slider_expression_OS)*100,"% (",
                                nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                                sep="")
                        d<-d[order(d[,"IND"]),]
                        d[,"dead"] <- as.numeric(d[,"dead"])

                        tmp <- data.frame(data_source =input$data_source_expression[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        expression_survival_plots_metadata <- rbind(expression_survival_plots_metadata,tmp)

                        # if (input_adjust_OS_expression=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_expression),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                        #
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        #
                        #
                        #
                        #
                        # # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                        # # #tt=function(x,t,...) {pspline(x + t/365.25)}
                        # # #fit.result <- survfit(Surv(OS,dead)~IND + Age,data=d)
                        # # try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                        # # if (class(try.test)=="try-error" | is.na(try.test)) {
                        # #   p.value <- NA
                        # # } else {
                        # #   plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # #   diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                        # #   p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # # }
                        #
                        # expression_survival_plots_combined_p <- c(expression_survival_plots_combined_p,p.value)

                      }
                    }
                  })
          }
          if(all(is.na(expression_survival_plots_metadata))) {
            output$expression_survival_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_survival_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_survival_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_survival_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=expression_survival_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))

            })
            # output$expression_survival_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_survival_plots_combined_p)))$p.value,2)
            # })
            # output$expression_survival_plots_single_p <- renderText({
            #   paste0(signif(expression_survival_plots_combined_p,2),collapse = ", ")
            # })
          }


          for (k in 1:length(input$data_source_expression)) {
            local({
              my_i <- k
              expression_survival_plots_name <- paste0("expression_survival_plots", my_i)
              expression_survival_summary_tables_name <- paste0("expression_survival_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[expression_survival_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                } else {
                  if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    if (input$dynamic_check_expression) {
                      d <- subset(d,description=="Pre")
                    }
                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),]
                    #if (input$dynamic_check_expression) {
                      #d <- d[as.numeric(as.character(d[,"UserGene"]))!=0,]

                    # "Gide_ipiPD1_1_Pre","Gide_ipiPD1_13_Pre","Gide_ipiPD1_19_Pre","Gide_ipiPD1_27_Pre",
                    # "Gide_ipiPD1_33_Pre","Gide_ipiPD1_35_Pre","Gide_ipiPD1_36_Pre","Gide_PD1_10_Pre","Gide_PD1_13_Pre","Gide_PD1_17_Pre","Gide_PD1_25_Pre",
                    # "Gide_PD1_29_Pre","Gide_PD1_41_Pre","Gide_PD1_47_Pre","Gide_PD1_48_Pre","Gide_PD1_8_Pre","Gide_ipiPD1_1_On","Gide_ipiPD1_13_On",
                    # "Gide_ipiPD1_19_On","Gide_ipiPD1_27_On","Gide_ipiPD1_33_On","Gide_ipiPD1_35_On","Gide_ipiPD1_36_On","Gide_PD1_10_On","Gide_PD1_13_On",
                    # "Gide_PD1_17_On","Gide_PD1_25_On","Gide_PD1_29_On","
                    # Gide_PD1_41_On","Gide_PD1_47_On","Gide_PD1_48_On","Gide_PD1_8_On",


                    # preandon_paired <- c("Riaz_Pt1_Pre","Riaz_Pt10_Pre","Riaz_Pt101_Pre",
                    #                      "Riaz_Pt103_Pre","Riaz_Pt106_Pre","Riaz_Pt11_Pre","Riaz_Pt17_Pre","Riaz_Pt18_Pre","Riaz_Pt2_Pre","Riaz_Pt23_Pre","Riaz_Pt26_Pre",
                    #                      "Riaz_Pt27_Pre","Riaz_Pt28_Pre","Riaz_Pt3_Pre","Riaz_Pt3.2_Pre","Riaz_Pt30_Pre","Riaz_Pt31_Pre","Riaz_Pt34_Pre","Riaz_Pt36_Pre",
                    #                      "Riaz_Pt37_Pre","Riaz_Pt37.2_Pre","Riaz_Pt38_Pre","Riaz_Pt4_Pre","Riaz_Pt44_Pre","Riaz_Pt46_Pre","Riaz_Pt47_Pre",
                    #                      "Riaz_Pt48_Pre","Riaz_Pt49_Pre","Riaz_Pt5_Pre","Riaz_Pt5.2_Pre","Riaz_Pt52_Pre",
                    #                      "Riaz_Pt62_Pre","Riaz_Pt65_Pre","Riaz_Pt67_Pre","Riaz_Pt77_Pre","Riaz_Pt78_Pre","Riaz_Pt79_Pre","Riaz_Pt8_Pre","Riaz_Pt82_Pre",
                    #                      "Riaz_Pt84_Pre","Riaz_Pt85_Pre","Riaz_Pt89_Pre","Riaz_Pt9_Pre","Riaz_Pt92_Pre","Riaz_Pt92.2_Pre","Riaz_Pt94_Pre","Riaz_Pt98_Pre",
                    #                      "Riaz_Pt1_On","Riaz_Pt10_On","Riaz_Pt101_On","Riaz_Pt103_On","Riaz_Pt106_On","Riaz_Pt11_On","Riaz_Pt17_On","Riaz_Pt18_On",
                    #                      "Riaz_Pt2_On","Riaz_Pt23_On","Riaz_Pt26_On","Riaz_Pt27_On",
                    #                      "Riaz_Pt28_On","Riaz_Pt3_On","Riaz_Pt3.2_On","Riaz_Pt30_On","Riaz_Pt31_On","Riaz_Pt34_On","Riaz_Pt36_On",
                    #                      "Riaz_Pt37_On","Riaz_Pt37.2_On","Riaz_Pt38_On","Riaz_Pt4_On","Riaz_Pt44_On","Riaz_Pt46_On","Riaz_Pt47_On",
                    #                      "Riaz_Pt48_On","Riaz_Pt49_On","Riaz_Pt5_On","Riaz_Pt5.2_On","Riaz_Pt52_On","Riaz_Pt62_On","Riaz_Pt65_On",
                    #                      "Riaz_Pt67_On","Riaz_Pt77_On","Riaz_Pt78_On","Riaz_Pt79_On","Riaz_Pt8_On","Riaz_Pt82_On","Riaz_Pt84_On",
                    #                      "Riaz_Pt85_On","Riaz_Pt89_On","Riaz_Pt9_On","Riaz_Pt92_On","Riaz_Pt92.2_On","Riaz_Pt94_On","Riaz_Pt98_On",
                    #                      "Riaz_Pt89_On","Riaz_Pt9_On","Riaz_Pt92_On","Riaz_Pt92.2_On","Riaz_Pt94_On","Riaz_Pt98_On")
                    #
                    #
                    #
                    #     d <- d[d[,"PatientID"] %in% preandon_paired,]

                    #}
                    d<-data.frame(d,dead=NA)
                    d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                    d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                    d <- d[!is.na(d[,"UserGene"]) & !is.na(d[,"dead"]),]
                    d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                    #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserGene")]))
                    d <- data.frame(data_source=as.character(d$data_source),
                                    dead=as.numeric(d$dead),
                                    OS=as.numeric(d$OS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    UserGene=as.numeric(d[,"UserGene"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserGene")]))
                    d<-data.frame(d,IND=paste("bottom ",input$slider_expression_OS*100,"% (",
                                              nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                                              sep=""))
                    # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_expression_OS*100,"% ",
                    #                           input_gene_expression," expression (",
                    #                           nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                    #                           sep=""))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),"IND"]<-
                      paste("top ",(1-input$slider_expression_OS)*100,"% (",
                            nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                            sep="")
                    # d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),"IND"]<-
                    #   paste("samples with top ",(1-input$slider_expression_OS)*100,"% ", input_gene_expression," expression (",
                    #         nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                    #         sep="")
                    d<-d[order(d[,"IND"]),]
                    d[,"dead"] <- as.numeric(d[,"dead"])

                    fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                    #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))


                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                      #plot.result <- ggsurv(fit.result)

                      if (input_adjust_OS_expression=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                              caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                     signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                              br(),
                                                                              paste0("n = ",
                                                                                     signif(summary(diff.result)$n,2),
                                                                                     ", number of events = ",
                                                                                     signif(summary(diff.result)$nevent,2)),
                                                                              style="text-align:left;color:black;font-size:15px"),
                                              options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                        }
                      } else {
                        diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_expression),data=d) )
                        if (class(diff.result.adjust.try)=="try-error") {
                          diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                          p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                            " (No ",input_adjust_OS_expression," was observed)")

                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                              caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                     signif(summary(diff.result)$logtest[1],2),
                                                                                     " on ",
                                                                                     signif(summary(diff.result)$logtest[2],2),
                                                                                     " df, p-value = ",
                                                                                     signif(summary(diff.result)$logtest[3],2)),
                                                                              br(),
                                                                              paste0("n = ",
                                                                                     signif(summary(diff.result)$n,2),
                                                                                     ", number of events = ",
                                                                                     signif(summary(diff.result)$nevent,2)),
                                                                              style="text-align:left;color:black;font-size:15px"),
                                              options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        } else {
                          p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                          table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                          rownames(table1.pre)[2] <- input_adjust_OS_expression
                          table1 <- datatable(table1.pre,
                                              caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                     signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                     " on ",
                                                                                     signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                     " df, p-value = ",
                                                                                     signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                              br(),
                                                                              paste0("n = ",
                                                                                     signif(summary(diff.result.adjust)$n,2),
                                                                                     ", number of events = ",
                                                                                     signif(summary(diff.result.adjust)$nevent,2)),
                                                                              style="text-align:left;color:black;font-size:15px"),
                                              options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        }
                      }




                      # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                      # #diff.result <- survdiff(coxph(Surv(OS,dead)~IND+Age,data=d))
                      # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                      plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                        ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                               #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                        ggplot2::guides(color=FALSE,linetype=FALSE)
                      p1<-ggplotly(plot.result.ann,
                                   height = 450 ) %>%
                        layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                               yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                               legend=list(orientation="v",x=0.6,y=1))
                      incProgress(100, detail = "Doing part %")
                      p1


                    }
                  }
                }
              })
              })
              #})
                output[[expression_survival_plots_name]] <- renderPlotly({
                  p1
                })
                output[[expression_survival_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
      #}
    })
    

    ###############
    #### expression PFS
    ###############

    observeEvent(input$do_expression,{
      input$do_expression
      #if (input$gene_expression!="ALL") {
      isolate({
        input_gene_expression <- input$gene_expression
        selected.database <- selecteddatabase_expression()
        input_adjust_OS_expression <- input$adjust_OS_expression
        if (nrow(selected.database)==0 | ncol(selected.database)<13) {
          output$expression_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["expression_PFS_plots0"]] <- renderPlotly({
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })

          # output$expression_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_PFS_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression,end=-19)]

          # expression_PFS_plots_combined_p <- NULL
          expression_PFS_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression)) {
              my_i <- k
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # expression_PFS_plots_combined_p <- c(expression_PFS_plots_combined_p,NA)
                      tmp <- data.frame(data_source =input$data_source_expression[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      expression_PFS_plots_metadata <- rbind(expression_PFS_plots_metadata,tmp)

                    } else {
                      if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                        # expression_PFS_plots_combined_p <- c(expression_PFS_plots_combined_p,NA)
                        tmp <- data.frame(data_source =input$data_source_expression[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        expression_PFS_plots_metadata <- rbind(expression_PFS_plots_metadata,tmp)

                      } else {
                        if (input$dynamic_check_expression) {
                          d <- subset(d,description=="Pre")
                        }
                        d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),]
                        #d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","UserGene")]))
                        d <- data.frame(data_source=as.character(d$data_source),
                                        PFS.status=as.character(d$PFS.status),
                                        PFS=as.numeric(d$PFS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        UserGene=as.numeric(d[,"UserGene"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Gender","UserGene")]))
                        d<-data.frame(d,IND=paste("bottom ",input$slider_expression_OS*100,"% (",

                                                  nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                                                  sep=""))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),"IND"]<-
                          paste("top ",(1-input$slider_expression_OS)*100,"% (",
                                nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                                sep="")
                        d <- data.frame(d,group.IND=0)
                        d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                        d<-d[order(d[,"IND"]),]

                        tmp <- data.frame(data_source =input$data_source_expression[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        expression_PFS_plots_metadata <- rbind(expression_PFS_plots_metadata,tmp)

                        # if (input_adjust_OS_expression=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_expression),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                        #
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        #
                        #
                        #
                        #
                        # # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                        # #
                        # # try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                        # # if (class(try.test)=="try-error" | is.na(try.test)) {
                        # #   p.value <- NA
                        # # } else {
                        # #   plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # #   diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                        # #   p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # # }
                        # expression_PFS_plots_combined_p <- c(expression_PFS_plots_combined_p,p.value)
                      }
                    }
                  })
          }
          if(all(is.na(expression_PFS_plots_metadata))) {
            output$expression_PFS_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_PFS_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_PFS_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_PFS_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=expression_PFS_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=expression_PFS_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
            })
            # output$expression_PFS_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_PFS_plots_combined_p)))$p.value,2)
            # })
            # output$expression_PFS_plots_single_p <- renderText({
            #   paste0(signif(expression_PFS_plots_combined_p,2),collapse = ", ")
            # })
          }


          for (k in 1:length(input$data_source_expression)) {
            local({
              my_i <- k
              expression_PFS_plots_name <- paste0("expression_PFS_plots", my_i)
              expression_PFS_summary_tables_name <- paste0("expression_PFS_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[expression_PFS_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {
                    if (input$dynamic_check_expression) {
                      d <- subset(d,description=="Pre")
                    }
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),]
                   # if (input$dynamic_check_expression) {
                      #d <- d[as.numeric(as.character(d[,"UserGene"]))!=0,]
                    # "Gide_ipiPD1_1_Pre","Gide_ipiPD1_13_Pre","Gide_ipiPD1_19_Pre","Gide_ipiPD1_27_Pre",
                    # "Gide_ipiPD1_33_Pre","Gide_ipiPD1_35_Pre","Gide_ipiPD1_36_Pre","Gide_PD1_10_Pre","Gide_PD1_13_Pre","Gide_PD1_17_Pre","Gide_PD1_25_Pre",
                    # "Gide_PD1_29_Pre","Gide_PD1_41_Pre","Gide_PD1_47_Pre","Gide_PD1_48_Pre","Gide_PD1_8_Pre","Gide_ipiPD1_1_On","Gide_ipiPD1_13_On",
                    # "Gide_ipiPD1_19_On","Gide_ipiPD1_27_On","Gide_ipiPD1_33_On","Gide_ipiPD1_35_On","Gide_ipiPD1_36_On","Gide_PD1_10_On","Gide_PD1_13_On",
                    # "Gide_PD1_17_On","Gide_PD1_25_On","Gide_PD1_29_On","
                    # Gide_PD1_41_On","Gide_PD1_47_On","Gide_PD1_48_On","Gide_PD1_8_On",
                    # preandon_paired <- c("Riaz_Pt1_Pre","Riaz_Pt10_Pre","Riaz_Pt101_Pre",
                    #                      "Riaz_Pt103_Pre","Riaz_Pt106_Pre","Riaz_Pt11_Pre","Riaz_Pt17_Pre","Riaz_Pt18_Pre","Riaz_Pt2_Pre","Riaz_Pt23_Pre","Riaz_Pt26_Pre",
                    #                      "Riaz_Pt27_Pre","Riaz_Pt28_Pre","Riaz_Pt3_Pre","Riaz_Pt3.2_Pre","Riaz_Pt30_Pre","Riaz_Pt31_Pre","Riaz_Pt34_Pre","Riaz_Pt36_Pre",
                    #                      "Riaz_Pt37_Pre","Riaz_Pt37.2_Pre","Riaz_Pt38_Pre","Riaz_Pt4_Pre","Riaz_Pt44_Pre","Riaz_Pt46_Pre","Riaz_Pt47_Pre",
                    #                      "Riaz_Pt48_Pre","Riaz_Pt49_Pre","Riaz_Pt5_Pre","Riaz_Pt5.2_Pre","Riaz_Pt52_Pre",
                    #                      "Riaz_Pt62_Pre","Riaz_Pt65_Pre","Riaz_Pt67_Pre","Riaz_Pt77_Pre","Riaz_Pt78_Pre","Riaz_Pt79_Pre","Riaz_Pt8_Pre","Riaz_Pt82_Pre",
                    #                      "Riaz_Pt84_Pre","Riaz_Pt85_Pre","Riaz_Pt89_Pre","Riaz_Pt9_Pre","Riaz_Pt92_Pre","Riaz_Pt92.2_Pre","Riaz_Pt94_Pre","Riaz_Pt98_Pre",
                    #                      "Riaz_Pt1_On","Riaz_Pt10_On","Riaz_Pt101_On","Riaz_Pt103_On","Riaz_Pt106_On","Riaz_Pt11_On","Riaz_Pt17_On","Riaz_Pt18_On",
                    #                      "Riaz_Pt2_On","Riaz_Pt23_On","Riaz_Pt26_On","Riaz_Pt27_On",
                    #                      "Riaz_Pt28_On","Riaz_Pt3_On","Riaz_Pt3.2_On","Riaz_Pt30_On","Riaz_Pt31_On","Riaz_Pt34_On","Riaz_Pt36_On",
                    #                      "Riaz_Pt37_On","Riaz_Pt37.2_On","Riaz_Pt38_On","Riaz_Pt4_On","Riaz_Pt44_On","Riaz_Pt46_On","Riaz_Pt47_On",
                    #                      "Riaz_Pt48_On","Riaz_Pt49_On","Riaz_Pt5_On","Riaz_Pt5.2_On","Riaz_Pt52_On","Riaz_Pt62_On","Riaz_Pt65_On",
                    #                      "Riaz_Pt67_On","Riaz_Pt77_On","Riaz_Pt78_On","Riaz_Pt79_On","Riaz_Pt8_On","Riaz_Pt82_On","Riaz_Pt84_On",
                    #                      "Riaz_Pt85_On","Riaz_Pt89_On","Riaz_Pt9_On","Riaz_Pt92_On","Riaz_Pt92.2_On","Riaz_Pt94_On","Riaz_Pt98_On",
                    #                      "Riaz_Pt89_On","Riaz_Pt9_On","Riaz_Pt92_On","Riaz_Pt92.2_On","Riaz_Pt94_On","Riaz_Pt98_On")
                    #
                    #
                    #
                    #  d <- d[d[,"PatientID"] %in% preandon_paired,]
                    #}
                    d <- data.frame(data_source=as.character(d$data_source),
                                    PFS.status=as.character(d$PFS.status),
                                    PFS=as.numeric(d$PFS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    UserGene=as.numeric(d[,"UserGene"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","UserGene")]))
                    d<-data.frame(d,IND=paste("bottom ",input$slider_expression_OS*100,"% (",

                                              nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                                              sep=""))
                    # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_expression_OS*100,"% ",
                    #                           input_gene_expression," expression (",
                    #                           nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                    #                           sep=""))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),"IND"]<-
                      paste("top ",(1-input$slider_expression_OS)*100,"% (",
                            nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                            sep="")
                    # d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),"IND"]<-
                    #   paste("samples with top ",(1-input$slider_expression_OS)*100,"% ", input_gene_expression," expression (",
                    #         nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_expression_OS),]),")",
                    #         sep="")
                    d <- data.frame(d,group.IND=0)
                    d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                    d<-d[order(d[,"IND"]),]

                    fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)

                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = str_sub(input$data_source_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                    } else {

                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))


                      if (input_adjust_OS_expression=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                              caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                     signif(summary(diff.result)$logtest[1],2),
                                                                                     " on ",
                                                                                     signif(summary(diff.result)$logtest[2],2),
                                                                                     " df, p-value = ",
                                                                                     signif(summary(diff.result)$logtest[3],2)),
                                                                              br(),
                                                                              paste0("n = ",
                                                                                     signif(summary(diff.result)$n,2),
                                                                                     ", number of events = ",
                                                                                     signif(summary(diff.result)$nevent,2)),
                                                                              style="text-align:left;color:black;font-size:15px"),
                                              options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                        }
                      } else {
                        diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_expression),data=d) )
                        if (class(diff.result.adjust.try)=="try-error") {
                          diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                          p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                            " (No ",input_adjust_OS_expression," was observed)")

                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                              caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                     signif(summary(diff.result)$logtest[1],2),
                                                                                     " on ",
                                                                                     signif(summary(diff.result)$logtest[2],2),
                                                                                     " df, p-value = ",
                                                                                     signif(summary(diff.result)$logtest[3],2)),
                                                                              br(),
                                                                              paste0("n = ",
                                                                                     signif(summary(diff.result)$n,2),
                                                                                     ", number of events = ",
                                                                                     signif(summary(diff.result)$nevent,2)),
                                                                              style="text-align:left;color:black;font-size:15px"),
                                              options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        } else {
                          p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                          table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                          rownames(table1.pre)[2] <- input_adjust_OS_expression
                          table1 <- datatable(table1.pre,
                                              caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                     signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                     " on ",
                                                                                     signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                     " df, p-value = ",
                                                                                     signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                              br(),
                                                                              paste0("n = ",
                                                                                     signif(summary(diff.result.adjust)$n,2),
                                                                                     ", number of events = ",
                                                                                     signif(summary(diff.result.adjust)$nevent,2)),
                                                                              style="text-align:left;color:black;font-size:15px"),
                                              options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        }
                      }



                      # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                      # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                      plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                             #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                      p1<-ggplotly(plot.result.ann,
                                 height = 450 ) %>%
                        layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                             yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                             legend=list(orientation="v",x=0.6,y=1))
                      incProgress(100, detail = "Doing part %")
                      p1
                    }
                  }
                }
              })
              })
              #})
                output[[expression_PFS_plots_name]] <- renderPlotly({
                  p1
                })
                output[[expression_PFS_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
      #}
    })
    

    
    
    ###########################################
    #### expression sum
    ###########################################
    
    observe({
    #   database.tmp <- database.test.process()[,1:12]
    #   updateSelectizeInput(session=session,inputId="treatment_expression_sum",
    #                        choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "treatment"])),
    #                        selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "treatment"])),
    #                        server=T)
    #   if (!input$dynamic_check_expression_sum) {
    #     updateSelectizeInput(session=session,inputId="description_expression_sum",
    #                          choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "description"])),
    #                          selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_expression_sum,end=-19), "description"])),
    #                          server=T)
    #     
    #   }
      if (input$defined_gene_expression_sum=="user-defined list") {
        updateSelectizeInput(session,"gene_expression_sum",
                             choices=a.colnames,selected=c("CD4","CD6","CD8A","CD8B"),
                             server=TRUE)
      } else {
        selected_expression_sum <- switch (input$defined_gene_expression_sum,
                                           "DNA_damage_response" = DNA_damage_response,
                                           "PDL1_expression_and_PD1_checkpoint_pathway_in_cancer"=PDL1_expression_and_PD1_checkpoint_pathway_in_cancer,
                                           "T_cell_receptor_signaling_pathway"=T_cell_receptor_signaling_pathway,
                                           "B_cell_receptor_signaling_pathway"=B_cell_receptor_signaling_pathway,
                                           "Cell_cycle_control"=Cell_cycle_control,
                                           "Survival_cell_death_regulation_signaling"=Survival_cell_death_regulation_signaling)
        updateSelectizeInput(session,"gene_expression_sum",
                             choices=a.colnames,selected=selected_expression_sum,
                             server=TRUE)
      }
    })

    observe({
      if (input$dynamic_check_expression_sum) {
        
        updateSelectizeInput(session=session,inputId="data_source_expression_sum",
                             choices=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                             selected=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                             server=T)
        
      }
    })
    observe({
      if (input$data_source_expression_sum=="" || is.null(input$data_source_expression_sum) ||
          input$treatment_expression_sum=="" || is.null(input$treatment_expression_sum) ||
          input$description_expression_sum=="" || is.null(input$description_expression_sum) ||
          input$gene_expression_sum=="" || is.null(input$gene_expression_sum)){
        disable("do_expression_sum")
      } else {
        enable("do_expression_sum")
      }
    })
    
    selecteddatabase_expression_sum <- eventReactive(input$do_expression_sum,{
      
      if(input$dynamic_check_expression_sum) {
        user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
        
        data_source1 <- str_sub(input$data_source_expression_sum,end=-19)[str_sub(input$data_source_expression_sum,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
        #treatment1 <- input$treatment_expression_sum
        treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_expression_sum])
        description1 <- c("Pre","On")
        genessum <- input$gene_expression_sum
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",genessum)
        a<-data.table(database.test.process())
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        m<-as.data.frame(m)
        
        if (nrow(m)>0 ) {
          m <- data.frame(m,sum=apply(as.data.frame(m[,13:ncol(m)]),1,function(x) sum(na.omit(x))))
          m[m[,"sum"]==0,"sum"]<-NA
          
          samples <- m$PatientID
          samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
          samples_paired <- table(samples)
          samples_paired <- samples_paired[samples_paired==2]
          samples_paired_pre <- paste0(names(samples_paired),"Pre")
          #samples_paired_on <- paste0(names(samples_paired),"On")
          samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
          m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],sum=m[m[,"PatientID"] %in% samples_paired_on,"sum"]-m[m[,"PatientID"] %in% samples_paired_pre,"sum"]),
                     data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],sum=m[m[,"PatientID"] %in% samples_paired_on,"sum"]-m[m[,"PatientID"] %in% samples_paired_pre,"sum"]))
          
          
          m<-data.frame(m,category="unknown")
          m[,"category"] <- as.character(m[,"category"])
          m1 <- m %>% split(.$data_source) %>% 
            lapply(function(d) {
              d[!is.na(d[,"sum"]) & d[,"sum"]>=quantile(na.omit(d[,"sum"]),input$slider_expression_sum_OS),"category"] <- paste0("top ",(1-input$slider_expression_sum_OS)*100,"%")
              d[!is.na(d[,"sum"]) & d[,"sum"]<quantile(na.omit(d[,"sum"]),input$slider_expression_sum_OS),"category"] <- paste0("bottom ",input$slider_expression_sum_OS*100,"%")
              d})
          m2<-ldply(m1,data.frame)
          m2<-m2[,-1]
          m2
        } else {
          m
        }
        
      } else {
      
      data_source1 <- str_sub(input$data_source_expression_sum,end=-19)
      #treatment1 <- input$treatment_expression_sum
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_expression_sum])
      description1 <- input$description_expression_sum
      genessum <- input$gene_expression_sum
      genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",genessum)
      a<-data.table(database.test.process())
      m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
      m<-m[,colnames(m) %in% genes1,with=FALSE]
      m<-as.data.frame(m)
      
      if (nrow(m)>0 ) {
        m <- data.frame(m,sum=apply(as.data.frame(m[,13:ncol(m)]),1,function(x) sum(na.omit(x))))
        m[m[,"sum"]==0,"sum"]<-NA
        m<-data.frame(m,category="unknown")
        m[,"category"] <- as.character(m[,"category"])
        m1 <- m %>% split(.$data_source) %>% 
          lapply(function(d) {
            d[!is.na(d[,"sum"]) & d[,"sum"]>=quantile(na.omit(d[,"sum"]),input$slider_expression_sum_OS),"category"] <- paste0("top ",(1-input$slider_expression_sum_OS)*100,"%")
            d[!is.na(d[,"sum"]) & d[,"sum"]<quantile(na.omit(d[,"sum"]),input$slider_expression_sum_OS),"category"] <- paste0("bottom ",input$slider_expression_sum_OS*100,"%")
            d})
        m2<-ldply(m1,data.frame)
        m2<-m2[,-1]
        m2
      } else {
        m
      }
      
      
      }
    })
    ############
    #### expression sum summary
    ############
    #input <- list(data_source_expression_sum=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),treatment_expression_sum=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression_sum=c("Pre","on"),gene_expression_sum=c("CD19","CD8A"),slider_expression_sum_OS=0.5,slider_expression_sum_PFS=0.5,clinical_expression_sum="treatment type")
    
    output$expression_sum_summary_disease_number <- renderText({
      input$do_expression_sum ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selecteddatabase_expression_sum()$Disease))})})
    output$expression_sum_summary_sample_number <- renderText({
      input$do_expression_sum ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selecteddatabase_expression_sum())})})
    output$expression_sum_summary_plots <- renderPlotly({
      summary_plot(inputdata=selecteddatabase_expression_sum(),feature="expression sum")
    })
    
    observeEvent(input$do_expression_sum,{
      input$do_expression_sum
      isolate({
        input_gene_expression_sum <- input$gene_expression_sum
        selected.database <- selecteddatabase_expression_sum()
        if (nrow(selected.database)==0 ) {
          output[["expression_sum_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_sum,end=-19)]
          for (k in 1:length(input$data_source_expression_sum)) {
            local({
              my_i <- k
              expression_sum_summary_clinical_plots_name <- paste0("expression_sum_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[expression_sum_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_expression_sum]]) & !is.na(d[,"sum"]),])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_expression_sum]]) & !is.na(d[,"sum"]),]
                      if (input$dynamic_check_expression_sum) {
                        d <- subset(d,description=="Pre")
                      }
                      if(input$clinical_expression_sum=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        #d[d[,"Age"]<=10,"Age"] <- "0-10"
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      if (input$dynamic_check_expression_sum) {
                        p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_expression_sum]],
                                   y=~signif(as.numeric(as.character(sum)),3), 
                                   type="violin",#type="box",
                                   points="all",#boxpoints="all",
                                   jitter=0.3,
                                   pointpos=-1.8,
                                   #mode="markders",
                                   color=d[,clinical_server[input$clinical_expression_sum]],showlegend=TRUE,
                                   colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                   hoveron="points+boxes",
                                   legendgroup=d[,clinical_server[input$clinical_expression_sum]], 
                                   text= ~paste0("Sample ID: ", PatientID,
                                                 "<br>Changes of expression sum: ",signif(as.numeric(as.character(sum)),3),
                                                 "<br>Treatment: ",treatment,
                                                 "<br>Description: On-Pre")) %>%
                          layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                 height = 480,
                                 xaxis=list(title="Clinical Attribute",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title="Changes of expression sum of selected genes",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                 ))
                      } else {
                      p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_expression_sum]],
                                 y=~signif(as.numeric(as.character(sum)),3), 
                                 type="violin",#type="box",
                                 points="all",#boxpoints="all",
                                 jitter=0.3,
                                 pointpos=-1.8,
                                 #mode="markders",
                                 color=d[,clinical_server[input$clinical_expression_sum]],showlegend=TRUE,
                                 colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                 hoveron="points+boxes",
                                 legendgroup=d[,clinical_server[input$clinical_expression_sum]], 
                                 text= ~paste0("Sample ID: ", PatientID,
                                               "<br>Expression: "," :",signif(as.numeric(as.character(sum)),3), 
                                               "<br>Treatment: ",treatment,
                                               "<br>Description: ",description)) %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title="Clinical Attribute",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title="Expression sum of selected genes",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               ))
                      }
                      incProgress(100, detail = "Doing part %")
                      p
                      
                    } 
                  }
                })
              })
            })
          }
        }
      }) 
    })
    
    
    output$expression_sum_summary_download <- downloadHandler(
      filename = function() {
        if (input$expression_sum_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$expression_sum_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$expression_sum_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        selected.database <- selecteddatabase_expression_sum()
        if (input$dynamic_check_expression_sum) {
          colnames(selected.database)[which(colnames(selected.database)=="sum")] <- "Changes of expression sum"
        } else {
          colnames(selected.database)[which(colnames(selected.database)=="sum")] <- "Expression sum"
        }
        if (input$expression_sum_summary_type == "Excel (CSV)") {
          write.csv (selected.database,file,row.names = FALSE)
        } else if (input$expression_sum_summary_type == "Text (TSV)") {
          write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$expression_sum_summary_type == "JSON") {
          write(toJSON(selected.database),file)
        }
      }
    )
    
    output$expression_sum_summary_table <- DT::renderDataTable({
      selected.database <- selecteddatabase_expression_sum()
      if (input$dynamic_check_expression_sum) {
        colnames(selected.database)[which(colnames(selected.database)=="sum")] <- "Changes of expression sum"
      } else {
        colnames(selected.database)[which(colnames(selected.database)=="sum")] <- "Expression sum"
      }
      selected.database
    })
    
    #############
    #### expression sum plot
    #############
    observeEvent(input$do_expression_sum,{
      input$do_expression_sum
      isolate({
        input_gene_expression_sum <- input$gene_expression_sum
        selected.database <- selecteddatabase_expression_sum()
        if (nrow(selected.database)==0 ) {
          output$expression_sum_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          
          output[["expression_sum_plots_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$expression_sum_plots_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_sum_plots_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_sum,end=-19)]
          
          # expression_sum_plots_plots_combined_p <- NULL
          expression_sum_plots_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression_sum)) {
              my_i <- k
              isolate({
                d<-splitted_list[[my_i]]
                
                if (is.na(names(splitted_list)[my_i])) {
                  tmp <- data.frame(data_source=input$data_source_expression_sum[my_i],TE=NA,seTE=NA)
                } else {
                  if (nrow(d)==0) {
                    tmp = data.frame(data_source=input$data_source_expression_sum[my_i],TE=NA,seTE=NA)
                  } else {
                    #d <- d[!is.na(d$PFS.status) & !is.na(d$sum) & d$sum!="unknown",]
                    if (nrow(d[!is.na(d$PFS.status) & !is.na(d$sum) & d$sum!="unknown",])==0) {
                      tmp = data.frame(data_source=input$data_source_expression_sum[my_i],TE=NA,seTE=NA)
                    } else {
                      # d$category_binary <- NA
                      d$response_binary <- NA
                      d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                      d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                      # d[d$category!="unknown" & d$category=="no mutation","category_binary"] <- 0
                      # d[d$category!="unknown" & d$category=="mutated","category_binary"] <- 1
                      d$scaled <- scale(as.numeric(d$sum))
                      if (all(is.na(d$scaled))) {
                        tmp = data.frame(data_source=input$data_source_expression_sum[my_i],TE=NA,seTE=NA)
                      } else {
                        tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                        if (nrow(tmp.pre$coefficients)>1) {
                          tmp <- data.frame(data_source=input$data_source_expression_sum[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                        } else {
                          tmp <- data.frame(data_source=input$data_source_expression_sum[my_i],TE=NA,seTE=NA)
                        }
                      }
                      
                      
                    }
                  }
                }
               expression_sum_plots_plots_metadata <- rbind.fill(expression_sum_plots_plots_metadata,tmp)
                
                # d$sum <- as.numeric(d$sum)
                # d.res <- d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),]
                # d.nonres <- d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),]
                # 
                # if (nrow(d.res)>1 & nrow(d.nonres)>1) {
                #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
                #                   n.res=length(na.omit(d.res[,"sum"])),
                #                   mean.res=mean(na.omit(d.res[,"sum"])),
                #                   sd.res=sd(na.omit(d.res[,"sum"])),
                #                   n.nonres=length(na.omit(d.nonres[,"sum"])),
                #                   mean.nonres=mean(na.omit(d.nonres[,"sum"])),
                #                   sd.nonres=sd(na.omit(d.nonres[,"sum"])))
                #   expression_sum_plots_plots_metadata <- rbind(expression_sum_plots_plots_metadata,tmp)
                # } else {
                #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
                #                   n.res=NA,
                #                   mean.res=NA,
                #                   sd.res=NA,
                #                   n.nonres=NA,
                #                   mean.nonres=NA,
                #                   sd.nonres=NA)
                #   expression_sum_plots_plots_metadata <- rbind(expression_sum_plots_plots_metadata,tmp)
                # }
              
                  
                    # if (is.na(names(splitted_list)[my_i])) {
                    #   expression_sum_plots_plots_combined_p <- c(expression_sum_plots_plots_combined_p,NA)
                    # } else {
                    #   #d[,11:ncol(d)] <- unfactor(d[,11:ncol(d)])
                    #   if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),])==0 ) {
                    #     expression_sum_plots_plots_combined_p <- c(expression_sum_plots_plots_combined_p,NA)
                    #   } else {
                    #     d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),]
                    #     if (input$dynamic_check_expression_sum) {
                    #       d <- subset(d,description=="Pre")
                    #     }
                    #     d[,"Groups"] <- as.character(d[,"Groups"])
                    #     d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    #     d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    #     ind <- sort(unique(d[,"Groups"]))
                    #     try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"sum"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"sum"])))$p.value)
                    #     if (class(try.test)=="try-error" | is.na(try.test)) {
                    #       expression_sum_plots_plots_combined_p <- c(expression_sum_plots_plots_combined_p,NA)
                    #     } else {
                    #       expression_sum_plots_plots_combined_p <- c(expression_sum_plots_plots_combined_p,p.v)
                    #     }
                    #   }
                    # }
                  })
          }
          if(all(is.na(expression_sum_plots_plots_metadata))) {
            output$expression_sum_plots_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_sum_plots_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_sum_plots_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_sum_plots_plots_meta <- renderPlot({
              
              # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
              #                 data=expression_sum_plots_plots_metadata,
              #                 label.e = "response",label.c = "nonresponse",
              #                 studlab = data_source),
              #        test.overall = TRUE,ff.study.label = "bold")
              m1 <- metagen(TE,seTE,
                            data=expression_sum_plots_plots_metadata,
                            sm="OR",
                            #label.e = "response",label.c = "nonresponse",
                            studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,
              #                data=expression_sum_plots_plots_metadata,
              #                sm="OR",
              #                #label.e = "response",label.c = "nonresponse",
              #                studlab = data_source),
              #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
            })
            # output$expression_sum_plots_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_sum_plots_plots_combined_p)))$p.value,2)
            # })
            # output$expression_sum_plots_plots_single_p <- renderText({
            #   paste0(signif(expression_sum_plots_plots_combined_p,2),collapse = ", ")
            # })
          }
          
          
          for (k in 1:length(input$data_source_expression_sum)) {
            local({
              my_i <- k
              expression_sum_plots_plots_name <- paste0("expression_sum_plots_plots", my_i)
              d<-splitted_list[[my_i]]
              output[[expression_sum_plots_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>% 
                    layout(title = str_sub(input$data_source_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  #d[,11:ncol(d)] <- unfactor(d[,11:ncol(d)])
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),]
                    if (input$dynamic_check_expression_sum) {
                      d <- subset(d,description=="Pre")
                    }
                    d[,"Groups"] <- as.character(d[,"Groups"])
                    d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    ind <- sort(unique(d[,"Groups"]))
                    if (as.character(unique(d$data_source)) == "Melanoma-Van Allen" | as.character(unique(d$data_source)) == "Melanoma-Riaz" | as.character(unique(d$data_source)) == "Melanoma-Auslander") {
                      lable_y <- " (RSEM)"
                    } else if (as.character(unique(d$data_source)) == "Melanoma-Chen" | as.character(unique(d$data_source)) == "Melanoma-Prat") {
                      lable_y <- " (NanoString signal)"
                    } else if (as.character(unique(d$data_source)) == "Melanoma-Hugo" | as.character(unique(d$data_source)) == "Melanoma-Gide" | as.character(unique(d$data_source)) == "Melanoma-Snyder-Nathanson") {
                      lable_y <- " (FPKM)"
                    } else {
                      lable_y <- NULL
                    }
                    try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"sum"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"sum"])))$p.value)
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      p.v <- " = NA"
                    } else {
                      # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"sum"]),as.numeric(d[d[,"Groups"]==ind[2],"sum"]))$p.value
                      # if (p.v < 0.001) {
                      #   p.v <- " < 0.001"
                      # } else {
                        p.v<-paste0(" = ",signif(p.v,2))
                      # } 
                    }
                    
                    if (input$dynamic_check_expression_sum) {
                      p<-plot_ly(data=d,x= ~Groups,
                                 y= ~signif(as.numeric(as.character(sum)),3), 
                                 type="violin",#type="box",
                                 points="all",#boxpoints="all",
                                 jitter=0.3,
                                 pointpos=-1.8,
                                 #mode="markders",
                                 color= ~PFS.status,
                                 colors=c("#7CDBD5","#DCB239","gray"),
                                 #hoveron="points+boxes",
                                 legendgroup=~PFS.status,showlegend=TRUE,
                                 text= ~paste0("Sample ID: ", PatientID,
                                               "<br>Expression: ",signif(as.numeric(as.character(sum)),3), 
                                               "<br>Treatment: ",treatment,
                                               "<br>Description: On-Pre")
                                 
                      )  %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title=paste0("Changes of expression sum of selected expression",
                                                       lable_y),
                                          titlefont=list(size=20,color="black",family= "Aril"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               )) %>% 
                        add_annotations(x = 0.5,
                                        y = d[which.max(d[,"sum"]),"sum"],
                                        text = paste0("p-value",p.v),
                                        xref = paste0("x", 1),  # add this
                                        yref = paste0("y", 1),  # add this
                                        showarrow = FALSE,
                                        ax = 20,
                                        ay = -40,
                                        font=list(size=20,color="black",family= "Aril black"))
                    } else {
                    
                    p<-plot_ly(data=d,x= ~Groups,
                               y= ~signif(as.numeric(as.character(sum)),3), 
                               type="violin",#type="box",
                               points="all",#boxpoints="all",
                               jitter=0.3,
                               pointpos=-1.8,
                               #mode="markders",
                               color= ~PFS.status,
                               colors=c("#7CDBD5","#DCB239","gray"),
                               #hoveron="points+boxes",
                               legendgroup=~PFS.status,showlegend=TRUE,
                               text= ~paste0("Sample ID: ", PatientID,
                                             "<br>Expression: ",signif(as.numeric(as.character(sum)),3), 
                                             "<br>Treatment: ",treatment,
                                             "<br>Description: ",description)
                               
                    )  %>%
                      layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             height = 480,
                             xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                        showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                             yaxis=list(title=paste0("Sum of selected expression",
                                                     lable_y),
                                        titlefont=list(size=20,color="black",family= "Aril"),
                                        showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                             )) %>% 
                      add_annotations(x = 0.5,
                                      y = d[which.max(d[,"sum"]),"sum"],
                                      text = paste0("p-value",p.v),
                                      xref = paste0("x", 1),  # add this
                                      yref = paste0("y", 1),  # add this
                                      showarrow = FALSE,
                                      ax = 20,
                                      ay = -40,
                                      font=list(size=20,color="black",family= "Aril black"))
                    }
                    incProgress(100, detail = "Doing part %")
                    p
                  } 
                  
                }
              })
              })
              })
            })
          }
        }
      })
    })
    

    ############
    #### expression sum survival
    ############
    observeEvent(input$do_expression_sum,{
      input$do_expression_sum
      isolate({
        input_gene_expression_sum <- input$gene_expression_sum
        selected.database <- selecteddatabase_expression_sum()
        input_adjust_OS_expression_sum <- input$adjust_OS_expression_sum
        if (nrow(selected.database)==0 ) {
          output$expression_sum_survival_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          
          output[["expression_sum_survival_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$expression_sum_survival_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_sum_survival_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_sum,end=-19)]
          
          # expression_sum_survival_plots_combined_p <- NULL
          expression_sum_survival_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression_sum)) {
              my_i <- k
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # expression_sum_survival_plots_combined_p <- c(expression_sum_survival_plots_combined_p,NA) 
                      tmp <- data.frame(data_source =input$data_source_expression_sum[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      expression_sum_survival_plots_metadata <- rbind(expression_sum_survival_plots_metadata,tmp)
                      
                    } else {
                      if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0 ) {
                        # expression_sum_survival_plots_combined_p <- c(expression_sum_survival_plots_combined_p,NA)
                        tmp <- data.frame(data_source =input$data_source_expression_sum[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        expression_sum_survival_plots_metadata <- rbind(expression_sum_survival_plots_metadata,tmp)
                        
                      } else {
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                        if (input$dynamic_check_expression_sum) {
                          d <- subset(d,description=="Pre")
                        }
                        d<-data.frame(d,dead=NA)
                        d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                        d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                        d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                        d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        dead=as.numeric(d$dead),
                                        OS=as.numeric(d$OS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        sum=as.numeric(d[,"sum"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))
                        d<-data.frame(d,IND=paste0("bottom ",input$slider_expression_sum_OS*100,"% (",
                                                   nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_expression_sum_OS),]),")"))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"sum"]>=quantile(d[,"sum"],input$slider_expression_sum_OS),"IND"] <- 
                          paste0("top ",(1-input$slider_expression_sum_OS)*100,"% (",
                                 nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_expression_sum_OS),]),")")
                        d<-d[order(d[,"IND"]),]
                        d[,"dead"] <- as.numeric(d[,"dead"])
                        
                        tmp <- data.frame(data_source =input$data_source_expression_sum[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        expression_sum_survival_plots_metadata <- rbind(expression_sum_survival_plots_metadata,tmp)
                        
                        # if (input_adjust_OS_expression_sum=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_expression_sum),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                        #     
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        # 
                        # # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                        # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                        # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # expression_sum_survival_plots_combined_p <- c(expression_sum_survival_plots_combined_p,p.value)                      
                        }
                    }
                  })
          }
          if(all(is.na(expression_sum_survival_plots_metadata))) {
            output$expression_sum_survival_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_sum_survival_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_sum_survival_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_sum_survival_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=expression_sum_survival_plots_metadata,sm="OR",
                               studlab = data_source)
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=expression_sum_survival_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
            })
            # output$expression_sum_survival_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_sum_survival_plots_combined_p)))$p.value,2)
            # })
            # output$expression_sum_survival_plots_single_p <- renderText({
            #   paste0(signif(expression_sum_survival_plots_combined_p,2),collapse = ", ")
            # })
          }
          
          
          for (k in 1:length(input$data_source_expression_sum)) {
            local({
              my_i <- k
              expression_sum_survival_plots_name <- paste0("expression_sum_survival_plots", my_i)
              expression_sum_survival_summary_tables_name <- paste0("expression_sum_survival_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[expression_sum_survival_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>% 
                    layout(title = str_sub(input$data_source_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  
                } else {
                  if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    
                  } else {
                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                    if (input$dynamic_check_expression_sum) {
                      d <- subset(d,description=="Pre")
                    }
                    d<-data.frame(d,dead=NA)
                    d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                    d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                    d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                    d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                    d <- data.frame(data_source=as.character(d$data_source),
                                    dead=as.numeric(d$dead),
                                    OS=as.numeric(d$OS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    sum=as.numeric(d[,"sum"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))
                    d<-data.frame(d,IND=paste0("bottom ",input$slider_expression_sum_OS*100,"% (",
                                               nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_expression_sum_OS),]),")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"sum"]>=quantile(d[,"sum"],input$slider_expression_sum_OS),"IND"] <- 
                      paste0("top ",(1-input$slider_expression_sum_OS)*100,"% (",
                             nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_expression_sum_OS),]),")")
                    d<-d[order(d[,"IND"]),]
                    d[,"dead"] <- as.numeric(d[,"dead"])
                    
                    fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                    
                      if (input_adjust_OS_expression_sum=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        
                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_expression_sum),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_expression_sum," was observed)")
                        
                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_expression_sum
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                    # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                    p1<-ggplotly(plot.result.ann,
                                 height = 450 ) %>%
                      layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                             yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                             legend=list(orientation="v",x=0.6,y=1))
                    incProgress(100, detail = "Doing part %")
                    p1
                    }
                  }
                }
              })
              })
              #})
                output[[expression_sum_survival_plots_name]] <- renderPlotly({
                  p1
                })
                output[[expression_sum_survival_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })
    

    ##########
    #### expression sum PFS
    ##########
    observeEvent(input$do_expression_sum,{
      input$do_expression_sum
      isolate({
        input_gene_expression_sum <- input$gene_expression_sum
        selected.database <- selecteddatabase_expression_sum()
        input_adjust_OS_expression_sum <- input$adjust_OS_expression_sum
        if (nrow(selected.database)==0 ) {
          output$expression_sum_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["expression_sum_PFS_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$expression_sum_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_sum_PFS_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_sum,end=-19)]
          
          
          # expression_sum_PFS_plots_combined_p <- NULL
          expression_sum_PFS_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression_sum)) {
              my_i <- k
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # expression_sum_PFS_plots_combined_p <- c(expression_sum_PFS_plots_combined_p,NA) 
                      tmp <- data.frame(data_source =input$data_source_expression_sum[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      expression_sum_PFS_plots_metadata <- rbind(expression_sum_PFS_plots_metadata,tmp)
                      
                    } else {
                      if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0 ) {
                        # expression_sum_PFS_plots_combined_p <- c(expression_sum_PFS_plots_combined_p,NA) 
                        tmp <- data.frame(data_source =input$data_source_expression_sum[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        expression_sum_PFS_plots_metadata <- rbind(expression_sum_PFS_plots_metadata,tmp)
                        
                      } else {
                        d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                        if (input$dynamic_check_expression_sum) {
                          d <- subset(d,description=="Pre")
                        }
                        d <- data.frame(data_source=as.character(d$data_source),
                                        PFS.status=as.character(d$PFS.status),
                                        PFS=as.numeric(d$PFS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        sum=as.numeric(d[,"sum"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","sum")]))
                        d<-data.frame(d,IND=paste("bottom ",input$slider_expression_sum_OS*100,"% (", 
                                                  nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_expression_sum_OS),]),")",
                                                  sep=""))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"sum"]> quantile(d[,"sum"],input$slider_expression_sum_OS),"IND"]<- 
                          paste("top ",(1-input$slider_expression_sum_OS)*100,"% (",
                                nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_expression_sum_OS),]),")",
                                sep="")
                        d <- data.frame(d,group.IND=0)
                        d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                        d<-d[order(d[,"IND"]),]
                        
                        tmp <- data.frame(data_source =input$data_source_expression_sum[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        expression_sum_PFS_plots_metadata <- rbind(expression_sum_PFS_plots_metadata,tmp)
                        
                        
                        # if (input_adjust_OS_expression_sum=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_expression_sum),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                        #     
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        # # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                        # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                        # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # expression_sum_PFS_plots_combined_p <- c(expression_sum_PFS_plots_combined_p,p.value)
                      }
                    }
                  })
          }
          if(all(is.na(expression_sum_PFS_plots_metadata))) {
            output$expression_sum_PFS_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_sum_PFS_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_sum_PFS_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_sum_PFS_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=expression_sum_PFS_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=expression_sum_PFS_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
            })
            # output$expression_sum_PFS_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_sum_PFS_plots_combined_p)))$p.value,2)
            # })
            # output$expression_sum_PFS_plots_single_p <- renderText({
            #   paste0(signif(expression_sum_PFS_plots_combined_p,2),collapse = ", ")
            # })
          }
          
          
          for (k in 1:length(input$data_source_expression_sum)) {
            local({
              my_i <- k
              expression_sum_PFS_plots_name <- paste0("expression_sum_PFS_plots", my_i)
              expression_sum_PFS_summary_tables_name <- paste0("expression_sum_PFS_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[expression_sum_PFS_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>% 
                    layout(title = str_sub(input$data_source_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  
                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                    if (input$dynamic_check_expression_sum) {
                      d <- subset(d,description=="Pre")
                    }
                    d <- data.frame(data_source=as.character(d$data_source),
                                    PFS.status=as.character(d$PFS.status),
                                    PFS=as.numeric(d$PFS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    sum=as.numeric(d[,"sum"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","sum")]))
                    d<-data.frame(d,IND=paste("bottom ",input$slider_expression_sum_OS*100,"% (", 
                                              nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_expression_sum_OS),]),")",
                                              sep=""))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"sum"]> quantile(d[,"sum"],input$slider_expression_sum_OS),"IND"]<- 
                      paste("top ",(1-input$slider_expression_sum_OS)*100,"% (",
                            nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_expression_sum_OS),]),")",
                            sep="")
                    d <- data.frame(d,group.IND=0)
                    d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                    d<-d[order(d[,"IND"]),]
                    fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                    
                      if (input_adjust_OS_expression_sum=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        
                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_expression_sum),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_expression_sum," was observed)")
                        
                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_expression_sum
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                    # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                    p1 <- ggplotly(plot.result.ann,
                                   height = 480 ) %>%
                      layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                             yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                             legend=list(orientation="v",x=0.6,y=1))
                    incProgress(100, detail = "Doing part %")
                    p1
                    }
                  }
                }
              })
              })
              #})
                output[[expression_sum_PFS_plots_name]] <- renderPlotly({
                  p1
                })
                output[[expression_sum_PFS_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })
    

    
    ###############################################
    #### expression relation pairs
    ###############################################
    #input <- list(gene_expression_pairs_1=c("CD28","CD276"),gene_expression_pairs_2=c("PDCD1","TNFRSF4"),gene_expression_pairs_3=c("CTLA4","TNFRSF4"),data_source_expression_pairs=c("Glioblastoma-Cloughesy (029 samples/-+-)"),treatment_expression_pairs=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_expression_pairs=c("Pre","On"),slider_expression_pairs_OS=1,slider_expression_pairs_PFS=1)
    
    updateSelectizeInput(session,"gene_expression_pairs_1",choices=a.colnames,selected = c("CD28","CD276"),server=TRUE)
    updateSelectizeInput(session,"gene_expression_pairs_2",choices=a.colnames,selected = c("PDCD1","TNFRSF4"),server=TRUE)
    #updateSelectizeInput(session,"gene_expression_pairs_3",choices=a.colnames,selected = c("CTLA4","TNFRSF4"),server=TRUE)
    
    ids<<-c(1,2)
    #ids<<-c(1,2,3)
    #  ids <- NULL
    observeEvent(input$add_expression_pairs,{
      if (is.null(ids)) {
        ids<<-1
      } else {
        ids <<- c(ids,max(ids)+1)
      }
      output$gene_expression_pairs <- renderUI({
        tagList(
          lapply(3:length(ids),function(i) {
            selectizeInput(paste("gene_expression_pairs_",ids[i],sep=""),sprintf("Gene pairs #%d",ids[i]),choices=NULL,multiple=TRUE,selected = NULL,options=list(maxItems=2))
          })
        )
      })
      lapply(3:length(ids),function(i) {
        updateSelectizeInput(session,paste("gene_expression_pairs_",ids[i],sep=""),choices=a.colnames,selected=c("CTLA4","PDCD1"),server=TRUE)
      })
      
      updateSliderInput(session=session,
                        "slider_expression_pairs_OS",
                        "Seperate samples by sum of gene relation pairs (F): ",
                        min=0,max=length(ids),value=0)
    })
    
    # output$slider_expression_pairs_OS <- renderUI({
    #   sliderInput("slider_expression_pairs_OS",
    #               "Seperate samples by sum of gene relation pairs (F): ",
    #               min=0,max=2,value=0) })
    
    #length_ids <- eventReactive(input$add_expression_pairs,{
    #  length(ids)
    #})
    #output$slider_expression_pairs_OS <- renderUI({
    # observeEvent(input$add_expression_pairs,{
    #   updateSliderInput(session=session,
    #                     "slider_expression_pairs_OS",
    #                     "Seperate samples by sum of gene relation pairs (F): ",
    #                     min=0,max=length(ids),value=0)
    #   # sliderInput("slider_expression_pairs_OS",
    #   #           "Seperate samples by sum of gene relation pairs (F): ",
    #   #           min=0,max=length_ids(),value=0) 
    #   #})
    # })
    # output$slider_expression_pairs_PFS <- renderUI({
    #   sliderInput("slider_expression_pairs_PFS",
    #             "Seperate samples for progression free survival analysis by sum of gene relation pairs: ",
    #             min=0,max=length_ids(),value=0) })
    observe({
      if (input$data_source_expression_pairs=="" || is.null(input$data_source_expression_pairs) ||
          input$treatment_expression_pairs=="" || is.null(input$treatment_expression_pairs) ||
          input$description_expression_pairs=="" || is.null(input$description_expression_pairs) ) {# ||
#          input$add_expression_pairs==FALSE){
        disable("do_expression_pairs")
      } else {
        enable("do_expression_pairs")
      }
    })
    
    
    selecteddatabase_expression_pairs <- eventReactive(input$do_expression_pairs,{
      data_source1 <- str_sub(input$data_source_expression_pairs,end=-19)
      # treatment1 <- input$treatment_expression_pairs
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_expression_pairs])
      description1 <- input$description_expression_pairs
      if (is.null(ids)) {
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
      } else {
        genessum<-list()
        txt_ids <- sapply(1:length(ids), function(i) paste("gene_expression_pairs_",ids[i],sep=""))
        for (i in 1:length(txt_ids)) {
          genessum[[i]] <- input[[ txt_ids[i] ]] 
        }
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",unlist(genessum))
      }
      a<-data.table(database.test.process())
      m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
      m<-m[,colnames(m) %in% genes1,with=FALSE]
      selected.database<-as.data.frame(m)
      
      if (nrow(selected.database)>0) {
        n <- ncol(selected.database)
        genessum<-list()
        txt_ids <- sapply(1:length(ids), function(i) paste("gene_expression_pairs_",ids[i],sep=""))
        isolate({
          for (i in 1:length(txt_ids)) {
            genessum[[i]] <- input[[ txt_ids[i] ]]
            if (length(input[[ txt_ids[i] ]]) == 1) {
              selected.database <- data.frame(selected.database,NA)
            } else {
              selected.database <- data.frame(selected.database,selected.database[,input[[ txt_ids[i] ]][1]] > selected.database[,input[[ txt_ids[i] ]][2]])
            }
          }
        })
        colnames(selected.database)[(n+1):ncol(selected.database)] <- unlist(lapply(genessum,function(x) paste(x,collapse=".")))
        cols <- sapply(selected.database, is.logical)
        for (j in (n+1):ncol(selected.database)) {
          selected.database[,j] <- as.numeric(selected.database[,j])
        }
        selected.database <- data.frame(selected.database,sum=NA)
        for (i in 1:nrow(selected.database)) {
          #selected.database[i,"sum"] <- sum(na.omit(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)])))
          selected.database[i,"sum"] <- sum(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)]))
        }
        m<-data.frame(selected.database,category="unknown")
        m[,"category"] <- as.character(m[,"category"])
        m1 <- m %>% split(.$data_source) %>% 
          lapply(function(d) {
            d[!is.na(d[,"sum"]) & d[,"sum"]>input$slider_expression_pairs_OS,"category"] <- paste0("higher than ",input$slider_expression_pairs_OS," pairs")
            d[!is.na(d[,"sum"]) & d[,"sum"]<=input$slider_expression_pairs_OS,"category"] <- paste0("lower than or equal to ",input$slider_expression_pairs_OS," pairs")
            d
          })
        m2 <- ldply(m1,data.frame)
        m2 <- m2[,-1]
        m2
      } else {
        selected.database
      }
    })
    ##############
    #### expression pairs summary
    ##############
    
    output$expression_pairs_summary_disease_number <- renderText({
      input$do_expression_pairs ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selecteddatabase_expression_pairs()$Disease))})})
    output$expression_pairs_summary_sample_number <- renderText({
      input$do_expression_pairs ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selecteddatabase_expression_pairs())})})
    output$expression_pairs_summary_plots <- renderPlotly({
      summary_plot(inputdata=selecteddatabase_expression_pairs(),feature="expression pairs")
    })

    observeEvent(input$do_expression_pairs,{
      input$do_expression_pairs
      isolate({
        selected.database <- selecteddatabase_expression_pairs()
        if (nrow(selected.database)==0 ) {
          output[["expression_pairs_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_pairs,end=-19)]
          for (k in 1:length(input$data_source_expression_pairs)) {
            local({
              my_i <- k
              expression_pairs_summary_clinical_plots_name <- paste0("expression_pairs_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[expression_pairs_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_expression_pairs]]) & !is.na(d[,"sum"]),])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_expression_pairs]]) & !is.na(d[,"sum"]),]
                      if(input$clinical_expression_pairs=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        #d[d[,"Age"]<=10,"Age"] <- "0-10"
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      
                      p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_expression_pairs]],
                                 y=~as.numeric(as.character(sum)), 
                                 type="violin",#type="box",
                                 points="all",#boxpoints="all",
                                 jitter=0.3,
                                 pointpos=-1.8,
                                 #mode="markders",
                                 color=d[,clinical_server[input$clinical_expression_pairs]],showlegend=TRUE,
                                 colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                 hoveron="points+boxes",
                                 legendgroup=d[,clinical_server[input$clinical_expression_pairs]], 
                                 text= ~paste0("Sample ID: ", PatientID,
                                               "<br>expression_pairs: ",sum,
                                               "<br>Treatment: ",treatment,
                                               "<br>Description: ",description)) %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title="Clinical Attribute",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title="Number of positive gene pairs",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               ))
                      incProgress(100, detail = "Doing part %")
                      p
                      
                    } 
                  }
                })
              })
            })
          }
        }
      }) 
    })
    
    
    
    
    
    output$expression_pairs_summary_download <- downloadHandler(
      filename = function() {
        if (input$expression_pairs_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$expression_pairs_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$expression_pairs_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        if (input$expression_pairs_summary_type == "Excel (CSV)") {
          write.csv (selecteddatabase_expression_pairs(),file,row.names = FALSE)
        } else if (input$expression_pairs_summary_type == "Text (TSV)") {
          write.table (selecteddatabase_expression_pairs(),file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$expression_pairs_summary_type == "JSON") {
          write(toJSON(selecteddatabase_expression_pairs()),file)
        }
      }
    )
    
    output$expression_pairs_summary_table <- DT::renderDataTable({
      selecteddatabase_expression_pairs()
    })
    
    #############
    #### expression pairs plot
    #############
observeEvent(input$do_expression_pairs,{
  input$do_expression_pairs
  isolate({
    selected.database <- selecteddatabase_expression_pairs()
    if (nrow(selected.database)==0 ) {
      output$expression_pairs_plots_plots_meta <- renderPlot({
        plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),cex.main=1.5,bty="l")
        abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
        text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
      })

      output[["expression_pairs_plots_plots0"]] <- renderPlotly({
        plot_ly () %>%
          layout(title = paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
          add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
      })

      # output$expression_pairs_plots_plots_combined_p <- renderText({
      #   "NA"
      # })
      # output$expression_pairs_plots_plots_single_p <- renderText({
      #   "NA"
      # })
    } else {

      splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
      splitted_list <- splitted_list[str_sub(input$data_source_expression_pairs,end=-19)]

      # expression_pairs_plots_plots_combined_p <- NULL
      expression_pairs_plots_plots_metadata <- NULL
      for (k in 1:length(input$data_source_expression_pairs)) {
          my_i <- k
          isolate({
          d<-splitted_list[[my_i]]

          if (is.na(names(splitted_list)[my_i])) {
            tmp <- data.frame(data_source=input$data_source_expression_pairs[my_i],TE=NA,seTE=NA)
          } else {
            if (nrow(d)==0) {
              tmp = data.frame(data_source=input$data_source_expression_pairs[my_i],TE=NA,seTE=NA)
            } else {
              #d <- d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",]
              if (nrow(d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",])==0 |
                length(sort(unique(d$category)))<2) {
                tmp = data.frame(data_source=input$data_source_expression_pairs[my_i],TE=NA,seTE=NA)
              } else {
                category_index <- sort(unique(d$category))
                d$category_binary <- NA
                d$response_binary <- NA
                d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                d[d$category!="unknown" & d$category==category_index[2],"category_binary"] <- 0
                d[d$category!="unknown" & d$category==category_index[1],"category_binary"] <- 1
                d$scaled <- scale(as.numeric(d$category_binary))
                if (all(is.na(d$scaled))) {
                  tmp = data.frame(data_source=input$data_source_expression_pairs[my_i],TE=NA,seTE=NA)
                } else {
                  tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                  if (nrow(tmp.pre$coefficients)>1) {
                    tmp <- data.frame(data_source=input$data_source_expression_pairs[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                  } else {
                    tmp <- data.frame(data_source=input$data_source_expression_pairs[my_i],TE=NA,seTE=NA)
                  }
                }


              }
            }
          }
          expression_pairs_plots_plots_metadata <- rbind.fill(expression_pairs_plots_plots_metadata,tmp)

          # d.res <- d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),]
          # d.nonres <- d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),]
          #
          # if (nrow(d.res)>1 & nrow(d.nonres)>1) {
          #   tmp <- data.frame(data_source = unique(d$data_source),
          #                     event.res=nrow(d.res[d.res$category==sort(unique(d$category))[1],]),
          #                     n.res=nrow(d.res),
          #                     event.nonres=nrow(d.nonres[d.nonres$category==sort(unique(d$category))[1],]),
          #                     n.nonres=nrow(d.nonres))
          #   expression_pairs_plots_plots_metadata <- rbind(expression_pairs_plots_plots_metadata,tmp)
          # } else {
          #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
          #                     event.res=NA,
          #                     n.res=NA,
          #                     event.nonres=NA,
          #                     n.nonres=NA)
          #   expression_pairs_plots_plots_metadata <- rbind(expression_pairs_plots_plots_metadata,tmp)
          # }


                # if (is.na(names(splitted_list)[my_i])) {
                #   expression_pairs_plots_plots_combined_p <- c(expression_pairs_plots_plots_combined_p,NA)
                # } else {
                #   if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d[!is.na(d[,"sum"]),])==0) {
                #     expression_pairs_plots_plots_combined_p <- c(expression_pairs_plots_plots_combined_p,NA)
                #   } else {
                #     d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),]
                #     d <- data.frame(d,Groups=d$PFS.status)
                #     ind <- sort(unique(d[,"Groups"]))
                #     d <- d[!is.na(d[,"sum"]),]
                #     d[,"sum"]<-paste0("F = ",d[,"sum"])
                #     d[,"PFS.status"] <- unfactor(d[,"PFS.status"])
                #     d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),"PFS.status"] <- paste0("response (n=",nrow(d[d[,"PFS.status"]=="response",]),")")
                #     d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),"PFS.status"] <- paste0("nonresponse (n=",nrow(d[d[,"PFS.status"]=="nonresponse",]),")")
                #     d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","sum")])))
                #     colnames(d.plot)[2] <- "Relation.pairs"
                #     try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","sum")])))$p.value)
                #     if (class(try.test)=="try-error" | is.na(try.test)) {
                #       expression_pairs_plots_plots_combined_p <- c(expression_pairs_plots_plots_combined_p,NA)
                #     } else {
                #       expression_pairs_plots_plots_combined_p <- c(expression_pairs_plots_plots_combined_p,p.v)
                #     }
                #
                #   }
                #
                # }
              })
      }
      if(all(is.na(expression_pairs_plots_plots_metadata))) {
        output$expression_pairs_plots_plots_meta <- renderPlot({
          plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
          abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
          text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
        })
        # output$expression_pairs_plots_plots_combined_p <- renderText({
        #   "NA"
        # })
        # output$expression_pairs_plots_plots_single_p <- renderText({
        #   "NA"
        # })
      } else {
        output$expression_pairs_plots_plots_meta <- renderPlot({

          # forest(metabin(event.res,n.res,event.nonres,n.nonres,
          #                data=expression_pairs_plots_plots_metadata,
          #                label.e = "response",label.c = "nonresponse",
          #                studlab = data_source),
          #        test.overall = TRUE,ff.study.label = "bold")
          m1 <- metagen(TE,seTE,
                        data=expression_pairs_plots_plots_metadata,
                        sm="OR",
                        #label.e = "response",label.c = "nonresponse",
                        studlab = str_sub(data_source,end=-19))
          forest(m1,layout="RevMan5",
                 hetstat=TRUE,
                 plotwidth="3inch",test.overall = TRUE,
                 xlim=c(0.1,10),backtransfer=T,
                 label.test.overall.random="random effects model: ",
                 label.test.overall.fixed="fixed effects model: ",
                 fs.test.overall = 15,
                 scientific.pval = TRUE,ff.study.label = "bold")
          # forest(metagen(TE,seTE,
          #                data=expression_pairs_plots_plots_metadata,
          #                sm="OR",
          #                #label.e = "response",label.c = "nonresponse",
          #                studlab = data_source),
          #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))

        })
        # output$expression_pairs_plots_plots_combined_p <- renderText({
        #   signif(fisher.method(rbind(a=na.omit(expression_pairs_plots_plots_combined_p)))$p.value,2)
        # })
        # output$expression_pairs_plots_plots_single_p <- renderText({
        #   paste0(signif(expression_pairs_plots_plots_combined_p,2),collapse = ", ")
        # })
      }


      for (k in 1:length(input$data_source_expression_pairs)) {
        local({
          my_i <- k
          expression_pairs_plots_plots_name <- paste0("expression_pairs_plots_plots", my_i)
          d<-splitted_list[[my_i]]
          output[[expression_pairs_plots_plots_name]] <- renderPlotly({
            withProgress(message = 'Making plot', value = 0, {
            isolate({
              if (is.na(names(splitted_list)[my_i])) {
                incProgress(100, detail = "Doing part %")
                plot_ly () %>%
                  layout(title = str_sub(input$data_source_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
              } else {
                 if (nrow(d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",])==0 | 
                     length(sort(unique(d$category)))<2) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>%
                    layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                } else {
                  # d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),]
                  # d <- data.frame(d,Groups=d$PFS.status)
                  # ind <- sort(unique(d[,"Groups"]))
                  # d <- d[!is.na(d[,"sum"]),]
                  # d[,"sum"]<-paste0("F = ",d[,"sum"])
                  # #d[,"PFS.status"] <- unfactor(d[,"PFS.status"])
                  # d[,"PFS.status"] <- as.character(d[,"PFS.status"])
                  # d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),"PFS.status"] <- paste0("response (n=",nrow(d[d[,"PFS.status"]=="response",]),")")
                  # d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),"PFS.status"] <- paste0("nonresponse (n=",nrow(d[d[,"PFS.status"]=="nonresponse",]),")")
                  # d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","sum")])))
                  # colnames(d.plot)[2] <- "Relation.pairs"
                  # try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","sum")])))$p.value)
                  d <- d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",]
                  d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","category")])))
                  colnames(d.plot)[2] <- "Groups"
                  d.plot[,"PFS.status"] <- as.character(d.plot[,"PFS.status"])
                  d.plot[,"Groups"] <- as.character(d.plot[,"Groups"])
                  d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="nonresponse","Freq"]),")",sep="" )
                  d.plot[d.plot[,"PFS.status"]=="response","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="response","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="response","Freq"]),")",sep="" )
                  d.plot[,"Groups"] <- paste(d.plot[,"Groups"]," group",sep="")
                  try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","category")])))$p.value)

                  if (class(try.test)=="try-error" | is.na(try.test)) {
                    p.v <- " = NA"
                  } else {
                    # p.v <- fisher.test(unlist(table(d[,c("PFS.status","sum")])))$p.value
                    # if (p.v < 0.001) {
                    #   p.v <- " < 0.001"
                    # } else {
                      p.v<-paste0(" = ",signif(p.v,2))
                    # }
                  }
                  # p<-ggplotly(ggplot(d.plot, aes(fill=Relation.pairs, y=Freq, x=PFS.status)) + ggtitle(as.character(unique(d$data_source))) +
                  #               theme(plot.title = element_text(size=15,family="Aril"),axis.title.y = element_text(size=15,family="Aril"),axis.text.x=element_text(angle=-45)) +
                  #               scale_fill_discrete(name=" ") +
                  #               geom_bar( stat="identity", position="fill") +ylab("Sum of gene relations frequency") + xlab(" ") +
                  #            annotate("text",
                  #                     x=1.5,
                  #                     y=1.2,
                  #                     label=paste0("p-value",p.v),
                  #                     size=4))
                  p <- ggplotly(ggplot(d.plot, aes(fill=Groups, y=Freq, x=PFS.status)) + ggtitle(as.character(unique(d$data_source))) +
                                  theme(plot.title = element_text(size=15,family="Aril"),axis.title.y = element_text(size=15,family="Aril"),axis.text.x=element_text(angle=-45)) +
                                  #scale_fill_manual(values=define.color[na.omit(unique(as.character(d.plot[,"Groups"])))]) +
                                  scale_fill_discrete(name=" ") +
                                  geom_bar( stat="identity", position="fill") +ylab("Frequency of groups") + xlab(" ") +
                                  annotate("text",x=1.5,y=1.2,label=paste0("p-value",p.v),size=4))
                  incProgress(100, detail = "Doing part %")
                  p
                  }

              }
            })
          })
          })
        })
      }

    }
  })
})
# })
    
    

    ############
    #### expression pairs survival
    ############
    observeEvent(input$do_expression_pairs,{
      input$do_expression_pairs
      isolate({
        selected.database <- selecteddatabase_expression_pairs()
        input_adjust_OS_expression_pairs <- input$adjust_OS_expression_pairs
        if (nrow(selected.database)==0 | nrow(selected.database[selected.database[,"category"]!="unknown",])==0) {
          output$expression_pairs_survival_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })

          output[["expression_pairs_survival_plots0"]] <- renderPlotly({
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })
          # output$expression_pairs_survival_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_pairs_survival_plots_single_p <- renderText({
          #   "NA"
          # })

        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_pairs,end=-19)]

          # expression_pairs_survival_plots_combined_p <- NULL
          expression_pairs_survival_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression_pairs)) {
              my_i <- k
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # expression_pairs_survival_plots_combined_p <- c(expression_pairs_survival_plots_combined_p,NA)
                      tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      expression_pairs_survival_plots_metadata <- rbind(expression_pairs_survival_plots_metadata,tmp)

                    } else {

                      if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]),])==0 | nrow(d[!is.na(d[,"sum"]),])==0) {
                        # expression_pairs_survival_plots_combined_p <- c(expression_pairs_survival_plots_combined_p,NA)
                        tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        expression_pairs_survival_plots_metadata <- rbind(expression_pairs_survival_plots_metadata,tmp)

                      } else {
                        d <- d[!is.na(d[,"sum"]),]
                        if (length(unique(d[,"sum"]))<2) {
                          expression_pairs_survival_plots_combined_p <- c(expression_pairs_survival_plots_combined_p,NA)
                        } else {
                          d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                          d<-data.frame(d,dead=NA)
                          d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                          d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                          d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                          d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                          d <- data.frame(data_source=as.character(d$data_source),
                                          dead=as.numeric(d$dead),
                                          OS=as.numeric(d$OS),
                                          Age=as.character(d$Age),
                                          Gender=as.character(d$Gender),
                                          sum=as.numeric(d[,"sum"]))
                          # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))


                          d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_expression_pairs_OS," pairs (",
                                                     nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,]),")"))
                          d[,"IND"]<-as.character(d[,"IND"])
                          d[d[,"sum"]>input$slider_expression_pairs_OS,"IND"] <-
                            paste0("higher than ",input$slider_expression_pairs_OS," pairs (",
                                   nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,]),")")
                          if (nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,])==0 | nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,])==0) {
                            # expression_pairs_survival_plots_combined_p <- c(expression_pairs_survival_plots_combined_p,NA)
                            tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                              event.top=NA,
                                              n.top=NA,
                                              event.bottom=NA,
                                              n.bottom=NA)
                            expression_pairs_survival_plots_metadata <- rbind(expression_pairs_survival_plots_metadata,tmp)

                          } else {
                            d<-d[order(d[,"IND"]),]
                            #d[,"dead"] <- as.numeric(d[,"dead"])
                            tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                              event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                              n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                              event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                              n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                            expression_pairs_survival_plots_metadata <- rbind(expression_pairs_survival_plots_metadata,tmp)

                            # if (input_adjust_OS_expression_pairs=="None") {
                            #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                            #   if (class(diff.result.try)=="try-error") {
                            #     p.value <- NA
                            #   } else {
                            #     p.value <- signif(summary(diff.result)$logtest[3],2)
                            #   }
                            # } else {
                            #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_expression_pairs),data=d) )
                            #   if (class(diff.result.adjust.try)=="try-error") {
                            #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                            #
                            #   } else {
                            #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                            #   }
                            # }
                            #
                            # # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                            # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                            # # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                            # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                            # expression_pairs_survival_plots_combined_p <- c(expression_pairs_survival_plots_combined_p,p.value)                          }
                        }
                      }
                    }
                    }
                  })
          }
          if(all(is.na(expression_pairs_survival_plots_metadata))) {
            output$expression_pairs_survival_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_pairs_survival_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_pairs_survival_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_pairs_survival_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=expression_pairs_survival_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))

              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=expression_pairs_survival_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
            })
            # output$expression_pairs_survival_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_pairs_survival_plots_combined_p)))$p.value,2)
            # })
            # output$expression_pairs_survival_plots_single_p <- renderText({
            #   paste0(signif(expression_pairs_survival_plots_combined_p,2),collapse = ", ")
            # })
          }


          for (k in 1:length(input$data_source_expression_pairs)) {
            local({
              my_i <- k
              expression_pairs_survival_plots_name <- paste0("expression_pairs_survival_plots", my_i)
              expression_pairs_survival_summary_tables_name <- paste0("expression_pairs_survival_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[expression_pairs_survival_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  isolate({
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = str_sub(input$data_source_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {

                    if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                    } else {
                      d <- d[!is.na(d[,"sum"]),]
                      if (length(unique(d[,"sum"]))<2) {
                        p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples because of same expression relations, please select more samples",showarrow=F,font=list(size=16,color="purple"))
                        table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                      } else {
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                        d<-data.frame(d,dead=NA)
                        d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                        d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                        d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                        d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        dead=as.numeric(d$dead),
                                        OS=as.numeric(d$OS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        sum=as.numeric(d[,"sum"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))


                        d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_expression_pairs_OS," pairs (",
                                                   nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,]),")"))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"sum"]>input$slider_expression_pairs_OS,"IND"] <-
                          paste0("higher than ",input$slider_expression_pairs_OS," pairs (",
                                 nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,]),")")
                        if (nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,])==0 | nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,])==0) {
                          p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples under this cutoff",showarrow=F,font=list(size=16,color="purple"))
                          table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                        } else {
                          d<-d[order(d[,"IND"]),]
                          #d[,"dead"] <- as.numeric(d[,"dead"])

                          fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                          
                          try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                          if (class(try.test)=="try-error" | is.na(try.test)) {
                            incProgress(100, detail = "Doing part %")
                            p1 <- plot_ly () %>%
                              layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                              add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                            table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                          } else {
                            plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                            if (input_adjust_OS_expression_pairs=="None") {
                              diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                              if (class(diff.result.try)=="try-error") {
                                p.value <- NA
                                table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                              } else {
                                p.value <- signif(summary(diff.result)$logtest[3],2)
                                table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                  caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                         signif(summary(diff.result)$logtest[1],2),
                                                                                         " on ",
                                                                                         signif(summary(diff.result)$logtest[2],2),
                                                                                         " df, p-value = ",
                                                                                         signif(summary(diff.result)$logtest[3],2)),
                                                                                  br(),
                                                                                  paste0("n = ",
                                                                                         signif(summary(diff.result)$n,2),
                                                                                         ", number of events = ",
                                                                                         signif(summary(diff.result)$nevent,2)),
                                                                                  style="text-align:left;color:black;font-size:15px"),
                                                  options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                            }
                          } else {
                            diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_expression_pairs),data=d) )
                            if (class(diff.result.adjust.try)=="try-error") {
                              diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                              p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                                " (No ",input_adjust_OS_expression_pairs," was observed)")

                              table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                  caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                         signif(summary(diff.result)$logtest[1],2),
                                                                                         " on ",
                                                                                         signif(summary(diff.result)$logtest[2],2),
                                                                                         " df, p-value = ",
                                                                                         signif(summary(diff.result)$logtest[3],2)),
                                                                                  br(),
                                                                                  paste0("n = ",
                                                                                         signif(summary(diff.result)$n,2),
                                                                                         ", number of events = ",
                                                                                         signif(summary(diff.result)$nevent,2)),
                                                                                  style="text-align:left;color:black;font-size:15px"),
                                                  options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                            } else {
                              p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                              table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                              rownames(table1.pre)[2] <- input_adjust_OS_expression_pairs
                              table1 <- datatable(table1.pre,
                                                  caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                         signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                         " on ",
                                                                                         signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                         " df, p-value = ",
                                                                                         signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                  br(),
                                                                                  paste0("n = ",
                                                                                         signif(summary(diff.result.adjust)$n,2),
                                                                                         ", number of events = ",
                                                                                         signif(summary(diff.result.adjust)$nevent,2)),
                                                                                  style="text-align:left;color:black;font-size:15px"),
                                                  options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                            }
                          }
                          # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                          # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                          plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                            ggplot2::geom_text(aes(label = #sprintf(" p-value = %0.2g",
                                                            #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                     paste0("p-value = ",p.value),x=300,y=0.05)) +
                            ggplot2::guides(color=FALSE,linetype=FALSE)
                          p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                            layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                   xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                   yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                   legend=list(orientation="v",x=0.6,y=1))
                          incProgress(100, detail = "Doing part %")
                          p1
                          }
                          }
                      }
                    }

                  }
                })
              })
              #})
                output[[expression_pairs_survival_plots_name]] <- renderPlotly({
                  p1
                })
                output[[expression_pairs_survival_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })
    

    ############
    #### expression pairs PFS
    ############
    observeEvent(input$do_expression_pairs,{
      input$do_expression_pairs
      isolate({
        selected.database <- selecteddatabase_expression_pairs()
        input_adjust_OS_expression_pairs <- input$adjust_OS_expression_pairs
        if (nrow(selected.database)==0 ) {
          output$expression_pairs_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["expression_pairs_PFS_plots0"]] <- renderPlotly({
            plot_ly () %>%
              layout(title = paste(str_sub(input$data_source_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })

          # output$expression_pairs_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$expression_pairs_PFS_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_expression_pairs,end=-19)]

          # expression_pairs_PFS_plots_combined_p <- NULL
          expression_pairs_PFS_plots_metadata <- NULL
          for (k in 1:length(input$data_source_expression_pairs)) {
              my_i <- k
              expression_pairs_PFS_plots_name <- paste0("expression_pairs_PFS_plots", my_i)
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # expression_pairs_PFS_plots_combined_p <- c(expression_pairs_PFS_plots_combined_p,NA)
                      tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      expression_pairs_PFS_plots_metadata <- rbind(expression_pairs_PFS_plots_metadata,tmp)

                    } else {

                      if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]),])==0 | nrow(d[!is.na(d[,"sum"]),])==0) {
                        # expression_pairs_PFS_plots_combined_p <- c(expression_pairs_PFS_plots_combined_p,NA)
                        tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        expression_pairs_PFS_plots_metadata <- rbind(expression_pairs_PFS_plots_metadata,tmp)

                      } else {
                        d <- d[!is.na(d[,"sum"]),]
                        if (length(unique(d[,"sum"]))<2) {
                          # expression_pairs_PFS_plots_combined_p <- c(expression_pairs_PFS_plots_combined_p,NA)
                          tmp <- data.frame(data_source =as.character(unique(d$data_source)),
                                            event.top=NA,
                                            n.top=NA,
                                            event.bottom=NA,
                                            n.bottom=NA)
                          expression_pairs_PFS_plots_metadata <- rbind(expression_pairs_PFS_plots_metadata,tmp)

                        } else {
                          d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                          d <- data.frame(data_source=as.character(d$data_source),
                                          PFS.status=as.character(d$PFS.status),
                                          PFS=as.numeric(d$PFS),
                                          Age=as.character(d$Age),
                                          Gender=as.character(d$Gender),
                                          sum=as.numeric(d[,"sum"]))
                          # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","sum")]))
                          d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_expression_pairs_OS," pairs (",
                                                     nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,]),")"))
                          d[,"IND"]<-as.character(d[,"IND"])
                          d[d[,"sum"]>input$slider_expression_pairs_OS,"IND"] <-
                            paste0("larger than ",input$slider_expression_pairs_OS," pairs (",
                                   nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,]),")")
                          if (nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,])==0 | nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,])==0) {
                            # expression_pairs_PFS_plots_combined_p <- c(expression_pairs_PFS_plots_combined_p,NA)
                            tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                              event.top=NA,
                                              n.top=NA,
                                              event.bottom=NA,
                                              n.bottom=NA)
                            expression_pairs_PFS_plots_metadata <- rbind(expression_pairs_PFS_plots_metadata,tmp)

                          } else {
                            d <- data.frame(d,group.IND=0)
                            d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                            d<-d[order(d[,"IND"]),]

                            tmp <- data.frame(data_source =input$data_source_expression_pairs[my_i],
                                              event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                              n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                              event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                              n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                            expression_pairs_PFS_plots_metadata <- rbind(expression_pairs_PFS_plots_metadata,tmp)

                            # if (input_adjust_OS_expression_pairs=="None") {
                            #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                            #   if (class(diff.result.try)=="try-error") {
                            #     p.value <- NA
                            #   } else {
                            #     p.value <- signif(summary(diff.result)$logtest[3],2)
                            #   }
                            # } else {
                            #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_expression_pairs),data=d) )
                            #   if (class(diff.result.adjust.try)=="try-error") {
                            #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                            #
                            #   } else {
                            #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                            #   }
                            # }
                            # # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                            # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                            # # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                            # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                            # expression_pairs_PFS_plots_combined_p <- c(expression_pairs_PFS_plots_combined_p,p.value)
                          }
                        }
                      }

                    }
                  })
          }
          if(all(is.na(expression_pairs_PFS_plots_metadata))) {
            output$expressioin_pairs_PFS_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$expression_pairs_PFS_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$expression_pairs_PFS_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$expression_pairs_PFS_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=expression_pairs_PFS_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))

              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=expression_pairs_PFS_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
            })
            # output$expression_pairs_PFS_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(expression_pairs_PFS_plots_combined_p)))$p.value,2)
            # })
            # output$expression_pairs_PFS_plots_single_p <- renderText({
            #   paste0(signif(expression_pairs_PFS_plots_combined_p,2),collapse = ", ")
            # })
          }

          for (k in 1:length(input$data_source_expression_pairs)) {
            local({
              my_i <- k
              expression_pairs_PFS_plots_name <- paste0("expression_pairs_PFS_plots", my_i)
              expression_pairs_PFS_summary_tables_name <- paste0("expression_pairs_PFS_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[expression_pairs_PFS_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = str_sub(input$data_source_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {

                    if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no PFS data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                    } else {
                      d <- d[!is.na(d[,"sum"]),]
                      if (length(unique(d[,"sum"]))<2) {
                        p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples because of same expression relations, please select more samples",showarrow=F,font=list(size=16,color="purple"))
                        table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                      } else {
                        d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        PFS.status=as.character(d$PFS.status),
                                        PFS=as.numeric(d$PFS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        sum=as.numeric(d[,"sum"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","sum")]))
                        d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_expression_pairs_OS," pairs (",
                                                   nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,]),")"))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"sum"]>input$slider_expression_pairs_OS,"IND"] <-
                          paste0("larger than ",input$slider_expression_pairs_OS," pairs (",
                                 nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,]),")")
                        if (nrow(d[d[,"sum"]<=input$slider_expression_pairs_OS,])==0 | nrow(d[d[,"sum"]>input$slider_expression_pairs_OS,])==0) {
                          p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples under this cutoff",showarrow=F,font=list(size=16,color="purple"))
                          table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                        } else {
                          d <- data.frame(d,group.IND=0)
                          d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                          d<-d[order(d[,"IND"]),]
                          fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                          
                          try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                          if (class(try.test)=="try-error" | is.na(try.test)) {
                            incProgress(100, detail = "Doing part %")
                            p1 <- plot_ly () %>%
                              layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                              add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                            table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                          } else {
                            plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                            if (input_adjust_OS_expression_pairs=="None") {
                              diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                              if (class(diff.result.try)=="try-error") {
                                p.value <- NA
                                table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                              } else {
                                p.value <- signif(summary(diff.result)$logtest[3],2)
                                table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                  caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                         signif(summary(diff.result)$logtest[1],2),
                                                                                         " on ",
                                                                                         signif(summary(diff.result)$logtest[2],2),
                                                                                         " df, p-value = ",
                                                                                         signif(summary(diff.result)$logtest[3],2)),
                                                                                  br(),
                                                                                  paste0("n = ",
                                                                                         signif(summary(diff.result)$n,2),
                                                                                         ", number of events = ",
                                                                                         signif(summary(diff.result)$nevent,2)),
                                                                                  style="text-align:left;color:black;font-size:15px"),
                                                  options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                            }
                          } else {
                            diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_expression_pairs),data=d) )
                            if (class(diff.result.adjust.try)=="try-error") {
                              diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                              p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                                " (No ",input_adjust_OS_expression_pairs," was observed)")

                              table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                  caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                         signif(summary(diff.result)$logtest[1],2),
                                                                                         " on ",
                                                                                         signif(summary(diff.result)$logtest[2],2),
                                                                                         " df, p-value = ",
                                                                                         signif(summary(diff.result)$logtest[3],2)),
                                                                                  br(),
                                                                                  paste0("n = ",
                                                                                         signif(summary(diff.result)$n,2),
                                                                                         ", number of events = ",
                                                                                         signif(summary(diff.result)$nevent,2)),
                                                                                  style="text-align:left;color:black;font-size:15px"),
                                                  options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                            } else {
                              p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                              table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                              rownames(table1.pre)[2] <- input_adjust_OS_expression_pairs
                              table1 <- datatable(table1.pre,
                                                  caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                         signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                         " on ",
                                                                                         signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                         " df, p-value = ",
                                                                                         signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                  br(),
                                                                                  paste0("n = ",
                                                                                         signif(summary(diff.result.adjust)$n,2),
                                                                                         ", number of events = ",
                                                                                         signif(summary(diff.result.adjust)$nevent,2)),
                                                                                  style="text-align:left;color:black;font-size:15px"),
                                                  options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                            }
                          }
                          # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                          # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                          plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                            ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                            #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                     paste0("p-value = ",p.value),x=300,y=0.05)) +
                            ggplot2::guides(color=FALSE,linetype=FALSE)
                          p1 <- ggplotly(plot.result.ann,height = 480 ) %>%
                            layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                   xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                   yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                   legend=list(orientation="v",x=0.6,y=1))
                          incProgress(100, detail = "Doing part %")
                          p1
                          }
                          }
                      }
                    }

                  }
                })
              })
              #})
                output[[expression_pairs_PFS_plots_name]] <- renderPlotly({
                  p1
                })
                output[[expression_pairs_PFS_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })

    

    ###############################################
    #### cibersort
    ###############################################
    
    # observe({
    #   database.tmp <- database.test.process()[,1:12]
    #   updateSelectizeInput(session=session,inputId="treatment_cibersort",
    #                        choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "treatment"])),
    #                        selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "treatment"])),
    #                        server=T)
    #   if (!input$dynamic_check_cibersort) {
    #     updateSelectizeInput(session=session,inputId="description_cibersort",
    #                          choices=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "description"])),
    #                          selected=as.character(unique(database.tmp[database.tmp[,"data_source"] %in% str_sub(input$data_source_cibersort,end=-19), "description"])),
    #                          server=T)
    #     
    #   }
    # 
    # })
    observe({
      if (input$dynamic_check_cibersort) {
        
        updateSelectizeInput(session=session,inputId="data_source_cibersort",
                             choices=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                             selected=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                             server=T)
        
      }
    })
    
    observe({
      if (input$data_source_cibersort=="" || is.null(input$data_source_cibersort) ||
          input$treatment_cibersort=="" || is.null(input$treatment_cibersort) ||
          input$description_cibersort=="" || is.null(input$description_cibersort) ||
          input$immucell_cibersort=="" || is.null(input$immucell_cibersort)){
        disable("do_cibersort")
      } else {
        enable("do_cibersort")
      }
    })
    selecteddatabase_cibersort <- eventReactive(input$do_cibersort,{
      
      if(input$dynamic_check_cibersort) {
      user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
      
      data_source1 <- str_sub(input$data_source_cibersort,end=-19)[str_sub(input$data_source_cibersort,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
      # treatment1 <- input$treatment_cibersort
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_cibersort])
      description1 <- c("Pre","On")
      a<-data.table(database.test.process())
      m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
      genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$immucell_cibersort)
      m<-m[,colnames(m) %in% genes1,with=FALSE]
      colnames(m)[ncol(m)] <- "UserImmu"
      m <- as.data.frame(m)
      if (nrow(m) >0) {
        samples <- m$PatientID
        samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
        samples_paired <- table(samples)
        samples_paired <- samples_paired[samples_paired==2]
        samples_paired_pre <- paste0(names(samples_paired),"Pre")
        #samples_paired_on <- paste0(names(samples_paired),"On")
        samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
        m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],UserImmu=m[m[,"PatientID"] %in% samples_paired_on,"UserImmu"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserImmu"]),
                   data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],UserImmu=m[m[,"PatientID"] %in% samples_paired_on,"UserImmu"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserImmu"]))
        m<-data.frame(m,category="unknown")
        m[,"category"] <- as.character(m[,"category"])
        m1 <- m %>% split(.$data_source) %>% 
          lapply(function(d) {#d[d[,"UserImmu"]>=median(na.omit(d[,"UserImmu"])),"IND"] <- paste("top 50% ", input$gene_cibersort," expression",sep="")
            d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]>=quantile(na.omit(d[,"UserImmu"]),input$slider_cibersort_OS),"category"] <- paste("top ",(1-input$slider_cibersort_OS)*100,"% ",sep="")
            d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]<quantile(na.omit(d[,"UserImmu"]),input$slider_cibersort_OS),"category"] <- paste("bottom ",input$slider_cibersort_OS*100,"% ",sep="")
            d})
        m2<-ldply(m1,data.frame)
        m2<-m2[,-1]
        m2$UserImmu <- as.numeric(m2$UserImmu)
        m2
      } else {
        m
      }
      
      } else {
      

      data_source1 <- str_sub(input$data_source_cibersort,end=-19)
      # treatment1 <- input$treatment_cibersort
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_cibersort])
      description1 <- input$description_cibersort
      genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$immucell_cibersort)
      a<-data.table(database.test.process())
      m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
      m<-m[,colnames(m) %in% genes1,with=FALSE]
      colnames(m)[ncol(m)] <- "UserImmu"
      m <- as.data.frame(m)
      
      if (nrow(m)>0 ) {
        m<-data.frame(m,category="unknown")
        m[,"category"] <- as.character(m[,"category"])
        #m<-m[!is.na(m[,"UserImmu"]),]
        m1 <- m %>% split(.$data_source) %>% 
          lapply(function(d) {
            d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]> quantile(na.omit(d[,"UserImmu"]),input$slider_cibersort_OS),"category"] <- paste0("top ",
                                                                                                                                   (1-input$slider_cibersort_OS)*100,
                                                                                                                                   "% ")
            d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]<=quantile(na.omit(d[,"UserImmu"]),input$slider_cibersort_OS),"category"] <- paste0("bottom ",
                                                                                                                                   input$slider_cibersort_OS*100,
                                                                                                                                   "% ")
            

            d })
        m2 <- ldply(m1,data.frame)
        m2 <- m2[,-1]
        m2$UserImmu <- as.numeric(m2$UserImmu)
#        colnames(m2)[13] <- input$immucell_cibersort
        m2
      } else {
        m2<-m
#        colnames(m2)[13] <- input$immucell_cibersort
        m2
      }
      
      }
    })
    
    ##########
    #### cibersort summary
    ##########
    
    output$cibersort_summary_disease_number <- renderText({
      input$do_cibersort ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selecteddatabase_cibersort()$Disease))})})
    output$cibersort_summary_sample_number <- renderText({
      input$do_cibersort ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selecteddatabase_cibersort())})})
    output$cibersort_summary_plots <- renderPlotly({
      summary_plot(inputdata=selecteddatabase_cibersort(),feature="cibersort")
    })
    

    observeEvent(input$do_cibersort,{
      input$do_cibersort
      isolate({
        input_immucell_cibersort <- input$immucell_cibersort
        selected.database <- selecteddatabase_cibersort()
        if (nrow(selected.database)==0 ) {
          output[["cibersort_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_cibersort,end=-19)]
          for (k in 1:length(input$data_source_cibersort)) {
            local({
              my_i <- k
              cibersort_summary_clinical_plots_name <- paste0("cibersort_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[cibersort_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_cibersort]]) & !is.na(d[,"UserImmu"]),])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_cibersort]]) & !is.na(d[,"UserImmu"]),]
                      if (input$dynamic_check_cibersort) {
                        d <- subset(d,description=="Pre")
                      }
                      if(input$clinical_cibersort=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        #d[d[,"Age"]<=10,"Age"] <- "0-10"
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      if (input$dynamic_check_cibersort) {
                        p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_cibersort]],
                                   y=~signif(as.numeric(as.character(UserImmu)),3),
                                   type="violin",#type="box",
                                   points="all",#boxpoints="all",
                                   jitter=0.3,
                                   pointpos=-1.8,
                                   #mode="markders",
                                   color=d[,clinical_server[input$clinical_cibersort]],showlegend=TRUE,
                                   colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                   hoveron="points+boxes",
                                   legendgroup=d[,clinical_server[input$clinical_cibersort]], 
                                   text= ~paste0("Sample ID: ", PatientID,
                                                 "<br>Changes of ",input_immucell_cibersort," components: ",signif(as.numeric(as.character(UserImmu)),3),
                                                 "<br>Treatment: ",treatment,
                                                 "<br>Description: ",description)) %>%
                          layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                 height = 480,
                                 xaxis=list(title="Clinical Attribute",
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                 yaxis=list(title=paste0("Changes of ",input_immucell_cibersort," component"),
                                            titlefont=list(size=20,color="black",family= "Aril black"),
                                            showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                 ))
                      } else {
                          p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_cibersort]],
                                 y=~signif(as.numeric(as.character(UserImmu)),3),
                                 type="violin",#type="box",
                                 points="all",#boxpoints="all",
                                 jitter=0.3,
                                 pointpos=-1.8,
                                 #mode="markders",
                                 color=d[,clinical_server[input$clinical_cibersort]],showlegend=TRUE,
                                 colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                 hoveron="points+boxes",
                                 legendgroup=d[,clinical_server[input$clinical_cibersort]], 
                                 text= ~paste0("Sample ID: ", PatientID,
                                               "<br>Immune cell: ",input_immucell_cibersort," :",signif(as.numeric(as.character(UserImmu)),3),
                                               "<br>Treatment: ",treatment,
                                               "<br>Description: ",description)) %>%
                            layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title="Clinical Attribute",
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title=paste0(input_immucell_cibersort," component"),
                                          titlefont=list(size=20,color="black",family= "Aril black"),
                                          showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               ))
                      }
                      incProgress(100, detail = "Doing part %")
                      p
                      
                    } 
                  }
                })
              })
            })
          }
        }
      }) 
    })
    output$cibersort_summary_download <- downloadHandler(
      filename = function() {
        if (input$cibersort_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$cibersort_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$cibersort_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        selected.database <- selecteddatabase_cibersort()
        if (input$dynamic_check_cibersort) {
          colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- paste0("Changes of ",input$immucell_cibersort)
        } else {
          colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- input$immucell_cibersort
        }
        
        if (input$cibersort_summary_type == "Excel (CSV)") {
          write.csv (selected.database,file,row.names = FALSE)
        } else if (input$cibersort_summary_type == "Text (TSV)") {
          write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$cibersort_summary_type == "JSON") {
          write(toJSON(selected.database),file)
        }
      }
    )
    
    output$cibersort_summary_table <- DT::renderDataTable({
      selected.database <- selecteddatabase_cibersort()
      if (input$dynamic_check_cibersort) {
        colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- paste0("Changes of ",input$immucell_cibersort)
      } else {
        colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- input$immucell_cibersort
      }
      selected.database
    })
    
    
    ##########
    #### cibersort plots
    ##########
    observeEvent(input$do_cibersort,{
      input$do_cibersort
      isolate({
        input_immucell_cibersort <- input$immucell_cibersort
        selected.database <- selecteddatabase_cibersort()
        if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
          output$cibersort_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["cibersort_plots_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$cibersort_plots_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$cibersort_plots_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_cibersort,end=-19)]
          
          # cibersort_plots_plots_combined_p <- NULL
          cibersort_plots_plots_metadata <- NULL
          for (k in 1:length(input$data_source_cibersort)) {
              my_i <- k
              d<-splitted_list[[my_i]]
              
              if (is.na(names(splitted_list)[my_i])) {
                tmp <- data.frame(data_source=input$data_source_cibersort[my_i],TE=NA,seTE=NA)
              } else {
                if (nrow(d)==0) {
                  tmp = data.frame(data_source=input$data_source_cibersort[my_i],TE=NA,seTE=NA)
                } else {
                  #d <- d[!is.na(d$PFS.status) & !is.na(d$UserImmu),]
                  if (nrow(d[!is.na(d$PFS.status) & !is.na(d$UserImmu),])==0) {
                    tmp = data.frame(data_source=input$data_source_cibersort[my_i],TE=NA,seTE=NA)
                  } else {
                    d$response_binary <- NA
                    d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                    d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                    d$scaled <- scale(as.numeric(d$UserImmu))
                    if (all(is.na(d$scaled))) {
                      tmp = data.frame(data_source=input$data_source_cibersort[my_i],TE=NA,seTE=NA)
                    } else {
                      tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                      if (nrow(tmp.pre$coefficients)>1) {
                        tmp <- data.frame(data_source=input$data_source_cibersort[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                      } else {
                        tmp = data.frame(data_source=input$data_source_cibersort[my_i],TE=NA,seTE=NA)
                        
                      }
                    }
                  }
                }
              }
              cibersort_plots_plots_metadata <- rbind.fill(cibersort_plots_plots_metadata,tmp)
              
              # d$UserImmu <- as.numeric(d$UserImmu)
              # d.res <- d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),]
              # d.nonres <- d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),]
              # if (nrow(d.res)>1 & nrow(d.nonres)>1) {
              #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
              #                     n.res=length(na.omit(d.res[,"UserImmu"])),
              #                     mean.res=mean(na.omit(d.res[,"UserImmu"])),
              #                     sd.res=sd(na.omit(d.res[,"UserImmu"])),
              #                     n.nonres=length(na.omit(d.nonres[,"UserImmu"])),
              #                     mean.nonres=mean(na.omit(d.nonres[,"UserImmu"])),
              #                     sd.nonres=sd(na.omit(d.nonres[,"UserImmu"])))
              #   cibersort_plots_plots_metadata <- rbind(cibersort_plots_plots_metadata,tmp)
              # } else {
              #   tmp <- data.frame(data_source = as.character(unique(d$data_source)),
              #                     n.res=NA,
              #                     mean.res=NA,
              #                     sd.res=NA,
              #                     n.nonres=NA,
              #                     mean.nonres=NA,
              #                     sd.nonres=NA)
              #   cibersort_plots_plots_metadata <- rbind(cibersort_plots_plots_metadata,tmp)
              # }   
          }
          #         if (is.na(names(splitted_list)[my_i])) {
          #           cibersort_plots_plots_combined_p <- c(cibersort_plots_plots_combined_p,NA) 
          #         } else {
          #           #d <- splitted_list[[my_i]]
          #           if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,13]),])==0) {
          #             cibersort_plots_plots_combined_p <- c(cibersort_plots_plots_combined_p,NA) 
          #           } else {
          #             d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),]
          #             if (input$dynamic_check_cibersort) {
          #               d <- subset(d,description=="Pre")
          #             }
          #             d[,"Groups"] <- as.character(d[,"Groups"])
          #             d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
          #             d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
          #             ind <- sort(unique(d[,"Groups"]))
          #             try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserImmu"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserImmu"])))$p.value)
          #             if (class(try.test)=="try-error" | is.na(try.test)) {
          #               cibersort_plots_plots_combined_p <- c(cibersort_plots_plots_combined_p,NA)
          #             } else {
          #               cibersort_plots_plots_combined_p <- c(cibersort_plots_plots_combined_p,p.v) 
          #             }
          #             
          # 
          #           } 
          #           
          #         }
          # }
          if(all(is.na(cibersort_plots_plots_metadata))) {
            output$cibersort_plots_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$cibersort_plots_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$excibersort_plots_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$cibersort_plots_plots_meta <- renderPlot({
              
              # forest(metacont(n.res,mean.res,sd.res,n.nonres,mean.nonres,sd.nonres,
              #                 data=cibersort_plots_plots_metadata,
              #                 label.e = "response",label.c = "nonresponse",
              #                 studlab = data_source),
              #        test.overall = TRUE,ff.study.label = "bold")
              m1 <- metagen(TE,seTE,
                            data=cibersort_plots_plots_metadata,
                            sm="OR",
                            #label.e = "response",label.c = "nonresponse",
                            studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,
              #                data=cibersort_plots_plots_metadata,
              #                sm="OR",
              #                #label.e = "response",label.c = "nonresponse",
              #                studlab = data_source),
              #        plotwidth="3inch",test.overall = TRUE,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
            })
            # output$cibersort_plots_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(cibersort_plots_plots_combined_p)))$p.value,2)
            # })
            # output$cibersort_plots_plots_single_p <- renderText({
            #   paste0(signif(cibersort_plots_plots_combined_p,2),collapse = ", ")
            # })
          }
          
          
          for (k in 1:length(input$data_source_cibersort)) {
            local({
              my_i <- k
              cibersort_plots_plots_name <- paste0("cibersort_plots_plots", my_i)
              d<-splitted_list[[my_i]]
              output[[cibersort_plots_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                if (is.na(names(splitted_list)[my_i])) {
                  plot_ly () %>% 
                    layout(title = str_sub(input$data_source_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  #d <- splitted_list[[my_i]]
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),])==0) {
                    plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),]
                    if (input$dynamic_check_cibersort) {
                      d <- subset(d,description=="Pre")
                    }
                    d[,"Groups"] <- as.character(d[,"Groups"])
                    d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    ind <- sort(unique(d[,"Groups"]))
                    try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserImmu"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserImmu"])))$p.value)
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      p.v <- " = NA"
                    } else {
                      # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserImmu"]),as.numeric(d[d[,"Groups"]==ind[2],"UserImmu"]))$p.value
                      # if (p.v < 0.001) {
                      #   p.v <- " < 0.001"
                      # } else {
                        p.v<-paste0(" = ",signif(p.v,2))
                      # } 
                    }
                    
                    if (input$dynamic_check_cibersort) {
                      p <- plot_ly(data=d,x=~Groups,
                                   y=~signif(as.numeric(as.character(UserImmu)),3),
                                   type="violin",#type="box",
                                   points="all",#boxpoints="all",
                                   jitter=0.3,
                                   pointpos=-1.8,
                                   #mode="markders",
                                   color= ~PFS.status,
                                   colors=c("#7CDBD5","#DCB239","gray"),
                                   #hoveron="points+boxes",
                                   legendgroup=~PFS.status,
                                   showlegend=TRUE,
                                   text= ~paste0("Sample ID: ", PatientID,
                                                 "<br>Changes of ",input_immucell_cibersort," component:",signif(as.numeric(as.character(UserImmu)),3),
                                                 "<br>Treatment: ",treatment,
                                                 "<br>Description: ",description)) %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title=paste0("Changaes of ",input_immucell_cibersort," component"),
                                          titlefont=list(size=20,color="black",family= "Aril"),
                                          showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               )) %>% 
                        add_annotations(x = 0.5,
                                        y = d[which.max(d[,"UserImmu"]),"UserImmu"],
                                        text = paste0("p-value",p.v),
                                        xref = paste0("x", 1),  # add this
                                        yref = paste0("y", 1),  # add this
                                        showarrow = FALSE,
                                        ax = 20,
                                        ay = -40,
                                        font=list(size=20,color="black",family= "Aril black"))
                    } else {
                      p <- plot_ly(data=d,x=~Groups,
                                   y=~signif(as.numeric(as.character(UserImmu)),3),
                                   type="violin",#type="box",
                                   points="all",#boxpoints="all",
                                   jitter=0.3,
                                   pointpos=-1.8,
                                   #mode="markders",
                                   color= ~PFS.status,
                                   colors=c("#7CDBD5","#DCB239","gray"),
                                   #hoveron="points+boxes",
                                   legendgroup=~PFS.status,
                                   showlegend=TRUE,
                                   text= ~paste0("Sample ID: ", PatientID,
                                                 "<br>Immune cell: ",input_immucell_cibersort," :",signif(as.numeric(as.character(UserImmu)),3),
                                                 "<br>Treatment: ",treatment,
                                                 "<br>Description: ",description)) %>%
                        layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                               height = 480,
                               xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                          showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                               yaxis=list(title=paste0(input_immucell_cibersort," component"),
                                          titlefont=list(size=20,color="black",family= "Aril"),
                                          showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                               )) %>% 
                        add_annotations(x = 0.5,
                                        y = d[which.max(d[,"UserImmu"]),"UserImmu"],
                                        text = paste0("p-value",p.v),
                                        xref = paste0("x", 1),  # add this
                                        yref = paste0("y", 1),  # add this
                                        showarrow = FALSE,
                                        ax = 20,
                                        ay = -40,
                                        font=list(size=20,color="black",family= "Aril black"))
                    }
                    
                    incProgress(100, detail = "Doing part %")
                    p
                  } 
                  
                }
              })
              })
            })
          }
        }
      })
    })
    
  
    #############
    #### cibersort survival
    #############
   
    observeEvent(input$do_cibersort,{
      input$do_cibersort
      isolate({
        input_immucell_cibersort <- input$immucell_cibersort
        selected.database <- selecteddatabase_cibersort()
        input_adjust_OS_cibersort <- input$adjust_OS_cibersort
        if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
          output$cibersort_survival_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["cibersort_survival_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$cibersort_survival_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$cibersort_survival_plots_single_p <- renderText({
          #   "NA"
          # })
          
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_cibersort,end=-19)]
          
          # cibersort_survival_plots_combined_p <- NULL
          cibersort_survival_plots_metadata <- NULL
          for (k in 1:length(input$data_source_cibersort)) {
              my_i <- k
              d<-splitted_list[[my_i]]
              d[,"UserImmu"] <- as.numeric(as.character(d[,"UserImmu"]))
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # cibersort_survival_plots_combined_p <- c(cibersort_survival_plots_combined_p,NA) 
                      tmp <- data.frame(data_source =input$data_source_cibersort[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                     cibersort_survival_plots_metadata <- rbind(cibersort_survival_plots_metadata,tmp)
                      
                    } else {
                      
                      #if (nrow(d[!is.na(d[,"OS.status"]),])==0 | nrow(d[!is.na(d[,"OS"]),])==0 | nrow(d[!is.na(d[,"UserImmu"]),])==0 | sum(d[!is.na(d[,"UserImmu"]),"UserImmu"])==0) {
                      if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),])==0) {
                        # cibersort_survival_plots_combined_p <- c(cibersort_survival_plots_combined_p,NA)
                        tmp <- data.frame(data_source =input$data_source_cibersort[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        cibersort_survival_plots_metadata <- rbind(cibersort_survival_plots_metadata,tmp)
                        
                      } else {
                        if (input$dynamic_check_cibersort) {
                          d <- subset(d,description=="Pre")
                        }
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),]
                        d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                        d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                        d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"dead"]),]
                        d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        dead=as.numeric(d$dead),
                                        OS=as.numeric(d$OS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        UserImmu=as.numeric(d[,"UserImmu"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserImmu")]))
                        d<-data.frame(d,IND=paste0("bottom ",
                                                   input$slider_cibersort_OS*100, 
                                                   "% (",
                                                   nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_cibersort_OS),]),
                                                   ")"))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),"IND"]<- 
                          paste0("top ",
                                 (1-input$slider_cibersort_OS)*100,
                                 "% (",
                                 nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),])
                                 ,")")
                        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                        d<-d[order(d[,"IND"]),]
                        
                        tmp <- data.frame(data_source =input$data_source_cibersort[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$dead==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$dead==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        cibersort_survival_plots_metadata <- rbind(cibersort_survival_plots_metadata,tmp)
                        
                        
                        # if (input_adjust_OS_cibersort=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_cibersort),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)
                        #     
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        # # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                        # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                        # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # cibersort_survival_plots_combined_p <- c(cibersort_survival_plots_combined_p,p.value)
                      }
                    }
                  })
          }
          if(all(is.na(cibersort_survival_plots_metadata))) {
            output$cibersort_survival_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$cibersort_survival_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$excibersort_survival_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$cibersort_survival_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=cibersort_survival_plots_metadata,sm="OR",
                               studlab = data_source)
              
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=cibersort_survival_plots_metadata$data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              # forest(meta2,test.overall= T,ff.study.label = "bold")
              
            })
            # output$cibersort_survival_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(cibersort_survival_plots_combined_p)))$p.value,2)
            # })
            # output$cibersort_survival_plots_single_p <- renderText({
            #   paste0(signif(cibersort_survival_plots_combined_p,2),collapse = ", ")
            # })
          }
          
          
          for (k in 1:length(input$data_source_cibersort)) {
            local({
              my_i <- k
              cibersort_survival_plots_name <- paste0("cibersort_survival_plots", my_i)
              cibersort_survival_summary_tables_name <- paste0("cibersort_survival_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              d[,"UserImmu"] <- as.numeric(as.character(d[,"UserImmu"]))
              #output[[cibersort_survival_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  p1 <- plot_ly () %>% 
                    layout(title = str_sub(input$data_source_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  
                } else {
                  
                  if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),])==0) {
                    p1 <- plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    
                  } else {
                    if (input$dynamic_check_cibersort) {
                      d <- subset(d,description=="Pre")
                    }
                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),]
                    d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                    d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                    d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"dead"]),]
                    d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                    d <- data.frame(data_source=as.character(d$data_source),
                                    dead=as.numeric(d$dead),
                                    OS=as.numeric(d$OS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    UserImmu=as.numeric(d[,"UserImmu"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserImmu")]))
                    d<-data.frame(d,IND=paste0("bottom ",
                                               input$slider_cibersort_OS*100, 
                                               "% (",
                                               nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_cibersort_OS),]),
                                               ")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),"IND"]<- 
                      paste0("top ",
                             (1-input$slider_cibersort_OS)*100,
                             "% (",
                             nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),])
                             ,")")
                    d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                    d<-d[order(d[,"IND"]),]
                    fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                    
                      if (input_adjust_OS_cibersort=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                       } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        
                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_cibersort),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_cibersort," was observed)")
                        
                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_cibersort
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                    # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                    p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                      layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                             yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                             legend=list(orientation="v",x=0.6,y=1)) #%>%
                    incProgress(100, detail = "Doing part %")
                    p1
                    }
                  }
                }
              })
              })
              #})
                output[[cibersort_survival_plots_name]] <- renderPlotly({
                  p1
                })
                output[[cibersort_survival_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })
    

    #############
    #### cibersort PFS
    #############
    observeEvent(input$do_cibersort,{
      input$do_cibersort
      isolate({
        input_immucell_cibersort <- input$immucell_cibersort
        selected.database <- selecteddatabase_cibersort()
        input_adjust_OS_cibersort <- input$adjust_OS_cibersort
        if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
          output$cibersort_PFS_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["cibersort_PFS_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
          
          # output$cibersort_PFS_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$cibersort_PFS_plots_single_p <- renderText({
          #   "NA"
          # })
          
        } else {
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_cibersort,end=-19)]
          
          
          # cibersort_PFS_plots_combined_p <- NULL
          cibersort_PFS_plots_metadata <- NULL
          for (k in 1:length(input$data_source_cibersort)) {
              my_i <- k
              d<-splitted_list[[my_i]]
              d[,"UserImmu"] <- as.numeric(as.character(d[,"UserImmu"]))
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      # cibersort_PFS_plots_combined_p <- c(cibersort_PFS_plots_combined_p,NA) 
                      tmp <- data.frame(data_source =input$data_source_cibersort[my_i],
                                        event.top=NA,
                                        n.top=NA,
                                        event.bottom=NA,
                                        n.bottom=NA)
                      cibersort_PFS_plots_metadata <- rbind(cibersort_PFS_plots_metadata,tmp)
                      
                    } else {
                      
                      if (nrow(d[!is.na(d[,"UserImmu"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                        # cibersort_PFS_plots_combined_p <- c(cibersort_PFS_plots_combined_p,NA) 
                        tmp <- data.frame(data_source =input$data_source_cibersort[my_i],
                                          event.top=NA,
                                          n.top=NA,
                                          event.bottom=NA,
                                          n.bottom=NA)
                        cibersort_PFS_plots_metadata <- rbind(cibersort_PFS_plots_metadata,tmp)
                        
                      } else {
                        d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                        if (input$dynamic_check_cibersort) {
                          d <- subset(d,description=="Pre")
                        }
                        d <- data.frame(data_source=as.character(d$data_source),
                                        PFS.status=as.character(d$PFS.status),
                                        PFS=as.numeric(d$PFS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        UserImmu=as.numeric(d[,"UserImmu"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","UserImmu")]))
                        d<-data.frame(d,IND=paste0("bottom ",
                                                   input$slider_cibersort_OS*100, 
                                                   "% (",
                                                   nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_cibersort_OS),]),
                                                   ")"))
                        d[,"IND"]<-as.character(d[,"IND"])
                        d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),"IND"]<- 
                          paste0("top ",
                                 (1-input$slider_cibersort_OS)*100,
                                 "% (",
                                 nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),])
                                 ,")")
                        d <- data.frame(d,group.IND=0)
                        d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                        
                        d<-d[order(d[,"IND"]),]
                        
                        tmp <- data.frame(data_source =input$data_source_cibersort[my_i],
                                          event.top=nrow(d[d$IND==unique(d$IND)[2] & d$group.IND==1,]),
                                          n.top=nrow(d[d$IND==unique(d$IND)[2],]),
                                          event.bottom=nrow(d[d$IND==unique(d$IND)[1] & d$group.IND==1,]),
                                          n.bottom=nrow(d[d$IND==unique(d$IND)[1],]))
                        cibersort_PFS_plots_metadata <- rbind(cibersort_PFS_plots_metadata,tmp)
                        
                        # if (input_adjust_OS_cibersort=="None") {
                        #   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        #   if (class(diff.result.try)=="try-error") {
                        #     p.value <- NA
                        #   } else {
                        #     p.value <- signif(summary(diff.result)$logtest[3],2)
                        #   }
                        # } else {
                        #   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_cibersort),data=d) )
                        #   if (class(diff.result.adjust.try)=="try-error") {
                        #     p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)
                        #     
                        #   } else {
                        #     p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                        #   }
                        # }
                        # # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                        # # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                        # # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                        # # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        # cibersort_PFS_plots_combined_p <- c(cibersort_PFS_plots_combined_p,p.value)
                      }
                    }
                  })
          }
          if(all(is.na(cibersort_PFS_plots_metadata))) {
            output$cibersort_PFS_plots_meta <- renderPlot({
              plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main="meta-analysis",cex.main=1.5,bty="l")
              abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
              text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            })
            # output$cibersort_PFS_plots_combined_p <- renderText({
            #   "NA"
            # })
            # output$excibersort_PFS_plots_single_p <- renderText({
            #   "NA"
            # })
          } else {
            output$cibersort_PFS_plots_meta <- renderPlot({
              meta1 <- metabin(event.bottom,n.bottom,event.top,n.top,
                               data=cibersort_PFS_plots_metadata,sm="OR",
                               studlab = data_source)
              meta2 <- data.frame(data_source=meta1$studlab,
                                  TE=meta1$TE,
                                  seTE=meta1$seTE)
              m1 <- metagen(TE,seTE,data=meta2,sm="OR",studlab = str_sub(data_source,end=-19))
              forest(m1,layout="RevMan5",    
                     hetstat=TRUE,
                     plotwidth="3inch",test.overall = TRUE,
                     xlim=c(0.1,10),backtransfer=T,
                     label.test.overall.random="random effects model: ",
                     label.test.overall.fixed="fixed effects model: ",
                     fs.test.overall = 15,
                     scientific.pval = TRUE,ff.study.label = "bold")
              # forest(metagen(TE,seTE,data=meta2,sm="OR",studlab = data_source),
              #        plotwidth="3inch",test.overall= T,scientific.pval = TRUE,ff.study.label = "bold",rightcols=c("effect","w.fixed","w.random"))
              
              # meta2 <- metagen(meta1$TE,meta1$seTE,sm="RR",
              #                  studlab=cibersort_PFS_plots_metadata$data_source)
              # forest(meta2,test.overall= T,ff.study.label = "bold")
            })
            # output$cibersort_PFS_plots_combined_p <- renderText({
            #   signif(fisher.method(rbind(a=na.omit(cibersort_PFS_plots_combined_p)))$p.value,2)
            # })
            # output$cibersort_PFS_plots_single_p <- renderText({
            #   paste0(signif(cibersort_PFS_plots_combined_p,2),collapse = ", ")
            # })
          }
          
          
          for (k in 1:length(input$data_source_cibersort)) {
            local({
              my_i <- k
              cibersort_PFS_plots_name <- paste0("cibersort_PFS_plots", my_i)
              cibersort_PFS_summary_tables_name <- paste0("cibersort_PFS_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              d[,"UserImmu"] <- as.numeric(as.character(d[,"UserImmu"]))
              #output[[cibersort_PFS_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  p1 <- plot_ly () %>% 
                    layout(title = str_sub(input$data_source_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  
                } else {
                  
                  #if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d[!is.na(d[,"PFS"]),])==0 | nrow(d[!is.na(d[,"UserImmu"]),])==0 | sum(d[!is.na(d[,"UserImmu"]),"UserImmu"])==0) {
                  if (nrow(d[!is.na(d[,"UserImmu"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                    p1 <- plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    
                  } else {
                    d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                    if (input$dynamic_check_cibersort) {
                      d <- subset(d,description=="Pre")
                    }
                    d <- data.frame(data_source=as.character(d$data_source),
                                    PFS.status=as.character(d$PFS.status),
                                    PFS=as.numeric(d$PFS),
                                    Age=as.character(d$Age),
                                    Gender=as.character(d$Gender),
                                    UserImmu=as.numeric(d[,"UserImmu"]))
                    # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","UserImmu")]))
                    d<-data.frame(d,IND=paste0("bottom ",
                                               input$slider_cibersort_OS*100, 
                                               "% (",
                                               nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_cibersort_OS),]),
                                               ")"))
                    d[,"IND"]<-as.character(d[,"IND"])
                    d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),"IND"]<- 
                      paste0("top ",
                             (1-input$slider_cibersort_OS)*100,
                             "% (",
                             nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_cibersort_OS),])
                             ,")")
                    d <- data.frame(d,group.IND=0)
                    d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                    
                    d<-d[order(d[,"IND"]),]
                    fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                    
                    try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      incProgress(100, detail = "Doing part %")
                      p1 <- plot_ly () %>%
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                    } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                    
                      if (input_adjust_OS_cibersort=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                          p.value <- NA
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                        } else {
                          p.value <- signif(summary(diff.result)$logtest[3],2)
                          table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                        
                      }
                    } else {
                      diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_cibersort),data=d) )
                      if (class(diff.result.adjust.try)=="try-error") {
                        diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                        p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                          " (No ",input_adjust_OS_cibersort," was observed)")
                        
                        table1 <- datatable(signif(coef(summary(diff.result)),2),
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      } else {
                        p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                        table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                        rownames(table1.pre)[2] <- input_adjust_OS_cibersort
                        table1 <- datatable(table1.pre,
                                            caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                   " on ",
                                                                                   signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                   " df, p-value = ",
                                                                                   signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                            br(),
                                                                            paste0("n = ",
                                                                                   signif(summary(diff.result.adjust)$n,2),
                                                                                   ", number of events = ",
                                                                                   signif(summary(diff.result.adjust)$nevent,2)),
                                                                            style="text-align:left;color:black;font-size:15px"),
                                            options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                      }
                    }
                    # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                    # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                    plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +  
                      ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                               paste0("p-value = ",p.value),x=300,y=0.05)) +
                      ggplot2::guides(color=FALSE,linetype=FALSE)
                    p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                      layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                             xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                             yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                             legend=list(orientation="v",x=0.6,y=1)) #%>%
                    incProgress(100, detail = "Doing part %")
                    p1
                    }
                  }
                }
              })
              })
              #})
                output[[cibersort_PFS_plots_name]] <- renderPlotly({
                  p1
                })
                output[[cibersort_PFS_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })
          }
        }
      })
    })


    ############################################
    #### multi-dimensions
    #############################################
    
    #input <- list(data_source_multi_dimensions=c("Melanoma-Auslander (037 samples/+++)","Melanoma-Chen (104 samples/++-)","Melanoma-Hugo (040 samples/-+-)","Melanoma-Prat (025 samples/-+-)","Melanoma-Riaz (139 samples/-+-)","Melanoma-Van Allen (112 samples/+--)"),treatment_multi_dimensions=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),description_multi_dimensions=c("Pre","On"), xaxis_multi_dimensions="mutational loads",yaxis_multi_dimensions="mutation",xaxis_gene_multi_dimensions="Mutation.Load",yaxis_gene_multi_dimensions=c("TAP1","TAP2","TAPBP","CIITA","RFX5","RFXANK"),slider_multi_dimensions_x=0.5,slider_multi_dimensions_y=0.5,do_multi_dimensions=TRUE,clinical_multi_dimensions="RECIST")
    
    
    observeEvent(input$xaxis_multi_dimensions,{
      output$xaxis_gene_multi_dimensions <- renderUI({
        if (input$xaxis_multi_dimensions=="expression" ) {
          selectizeInput("xaxis_gene_multi_dimensions",
                         "Insert a gene, for example: CD19",
                         choices=NULL)
        } else if (input$xaxis_multi_dimensions=="expression sum" ) {
          selectizeInput("xaxis_gene_multi_dimensions",
                         "Insert a gene set, for example: PDCD1 CTLA4",
                         choices=NULL,multiple=TRUE)
        } else if (input$xaxis_multi_dimensions=="mutation") {
          selectizeInput("xaxis_gene_multi_dimensions",
                         "Insert a gene(s), for example: BRAF KRAS",
                         choices=NULL,multiple=TRUE)
        } else if (input$xaxis_multi_dimensions=="immune cell components") {
          selectInput("xaxis_gene_multi_dimensions",
                      "Immune cell type: ",
                      choices = sort(c("B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive","T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta","NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting","Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")),
                      selected = "T.cells.CD8")
        } else if (input$xaxis_multi_dimensions=="mutational loads") {
          selectInput("xaxis_gene_multi_dimensions",
                      "Mutational loads: ",
                      choices = "Mutation.Load",
                      selected = "Mutation.Load")
        } else if (input$xaxis_multi_dimensions=="mutational signatures") {
          selectInput("xaxis_gene_multi_dimensions",
                      "Mutational signature type: ",
                      choices = sort(c("Signature.1 (age correlated)","Signature.2","Signature.3 (BRCA1 and BRCA2 associated)","Signature.4 (tobacco associated)","Signature.5","Signature.6 (DNA mismatch repair associated)","Signature.7 (ultraviolet associated)","Signature.8","Signature.9","Signature.10","Signature.11","Signature.12","Signature.13","Signature.14 (high tumor mutation burden associated)","Signature.15 (DNA mismatch repair associated)","Signature.16","Signature.17","Signature.18","Signature.19","Signature.20 (DNA mismatch repair associated)","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25 (Hodgkin's cell lines only)","Signature.26 (DNA mismatch repair associated)","Signature.27","Signature.28","Signature.29","Signature.30","unknown")),
                      selected = "Signature.1 (age correlated)")
        }
      })
      if (input$xaxis_multi_dimensions=="expression" ) {
        updateSelectizeInput(session,"xaxis_gene_multi_dimensions",
                             choices=a.colnames,
                             selected = "CD19",server=TRUE)
      } else if (input$xaxis_multi_dimensions=="expression sum" ) {
        updateSelectizeInput(session,"xaxis_gene_multi_dimensions",
                             choices=a.colnames,
                             selected = c("PDCD1","CTLA4"),
                             server=TRUE)
      } else if (input$xaxis_multi_dimensions=="mutation") {
        updateSelectizeInput(session,"xaxis_gene_multi_dimensions",
                             choices=a.colnames,
                             selected=c("BRAF","KRAS","HRAS","NRAS","NF1"),server=TRUE)
      } })
    xids<<-NULL
    observeEvent(input$xaxis_add_multi_dimensions,{
      
      if (is.null(xids)) {
        xids <<- 1
      } else {
        xids <<- c(xids,max(xids)+1)
      }
      output$xaxis_gene_multi_dimensions_x <- renderUI({
        tagList(
          lapply(1:length(xids),function(i) {
            #      insertUI(selector="#xaxis_add_multi_dimensions",where="afterEnd",ui=selectizeInput(paste("xaxis_gene_multi_dimensions_",xids[i],sep=""),sprintf("Gene pairs #%d",xids[i]),choices=a.colnames,multiple=TRUE,selected = c("CD2","CD5"),options=list(maxItems=2)))
            selectizeInput(paste("xaxis_gene_multi_dimensions_",xids[i],sep=""),sprintf("Gene pairs #%d",xids[i]),choices=NULL,multiple=TRUE,selected = c("CD2","CD5"),options=list(maxItems=2))
          })
        )
      })
      lapply(1:length(xids),function(i) {
        #      insertUI(selector="#xaxis_add_multi_dimensions",where="afterEnd",ui=selectizeInput(paste("xaxis_gene_multi_dimensions_",xids[i],sep=""),sprintf("Gene pairs #%d",xids[i]),choices=a.colnames,multiple=TRUE,selected = c("CD2","CD5"),options=list(maxItems=2)))
        updateSelectizeInput(session,paste("xaxis_gene_multi_dimensions_",xids[i],sep=""),choices=a.colnames,selected = c("CD2","CD5"),server=TRUE)
      })
      
    })
    
    observeEvent(input$yaxis_multi_dimensions,{
      output$yaxis_gene_multi_dimensions <- renderUI({
        if (input$yaxis_multi_dimensions=="expression" ) {
          selectizeInput("yaxis_gene_multi_dimensions",
                         "Insert a gene, for example: PDCD1",
                         choices=NULL)
        } else if (input$yaxis_multi_dimensions=="expression sum" ) {
          selectizeInput("yaxis_gene_multi_dimensions",
                         "Insert a gene set, for example: CD4,CD6 and CD8",
                         choices=NULL,multiple=TRUE)
        } else if (input$yaxis_multi_dimensions=="mutation") {
          selectizeInput("yaxis_gene_multi_dimensions",
                         "Insert a gene(s), for example: HRAS NRAS",
                         choices=NULL,multiple=TRUE)
        } else if (input$yaxis_multi_dimensions=="immune cell components") {
          selectInput("yaxis_gene_multi_dimensions",
                      "Immune cell type: ",
                      choices = sort(c("B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive","T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta","NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting","Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")),
                      selected = "T.cells.CD8")
        } else if (input$yaxis_multi_dimensions=="mutational loads") {
          selectInput("yaxis_gene_multi_dimensions",
                      "Mutational loads: ",
                      choices = "Mutation.Load",
                      selected = "Mutation.Load")
        } else if (input$yaxis_multi_dimensions=="mutational signatures") {
          selectInput("yaxis_gene_multi_dimensions",
                      "Mutational signature type: ",
                      choices = sort(c("Signature.1 (age correlated)","Signature.2","Signature.3 (BRCA1 and BRCA2 associated)","Signature.4 (tobacco associated)","Signature.5","Signature.6 (DNA mismatch repair associated)","Signature.7 (ultraviolet associated)","Signature.8","Signature.9","Signature.10","Signature.11","Signature.12","Signature.13","Signature.14 (high tumor mutation burden associated)","Signature.15 (DNA mismatch repair associated)","Signature.16","Signature.17","Signature.18","Signature.19","Signature.20 (DNA mismatch repair associated)","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25 (Hodgkin's cell lines only)","Signature.26 (DNA mismatch repair associated)","Signature.27","Signature.28","Signature.29","Signature.30","unknown")),
                      selected = "Signature.1 (age correlated)")
        }})
      if (input$yaxis_multi_dimensions=="expression" ) {
        updateSelectizeInput(session,"yaxis_gene_multi_dimensions",
                             choices=a.colnames,
                             selected = "PDCD1",server=TRUE)
      } else if (input$yaxis_multi_dimensions=="expression sum" ) {
        updateSelectizeInput(session,"yaxis_gene_multi_dimensions",
                             choices=a.colnames,
                             selected = c("CD4","CD6","CD8A","CD8B"),
                             server=TRUE)
      } else if (input$yaxis_multi_dimensions=="mutation") {
        updateSelectizeInput(session,"yaxis_gene_multi_dimensions",
                             choices=a.colnames,
                             selected=c("HRAS","NRAS"),server=TRUE)
      }
    })
    yids<<-NULL
    observeEvent(input$yaxis_add_multi_dimensions,{
      
      if (is.null(yids)) {
        yids <<- 1
      } else {
        yids <<- c(yids,max(yids)+1)
      }
      output$yaxis_gene_multi_dimensions_y <- renderUI ({
        tagList(
          lapply(1:length(yids),function(i) {
            #      insertUI(selector="#yaxis_add_multi_dimensions",where="afterEnd",ui=selectizeInput(paste("yaxis_gene_multi_dimensions_",yids[i],sep=""),sprintf("Gene pairs #%d",yids[i]),choices=a.colnames,multiple=TRUE,selected = c("CD4","CD6"),options=list(maxItems=2)))
            selectizeInput(paste("yaxis_gene_multi_dimensions_",yids[i],sep=""),sprintf("Gene pairs #%d",yids[i]),choices=NULL,multiple=TRUE,selected = c("CD4","CD6"),options=list(maxItems=2))
          })
        )
      })
      lapply(1:length(yids),function(i) {
        #      insertUI(selector="#yaxis_add_multi_dimensions",where="afterEnd",ui=selectizeInput(paste("yaxis_gene_multi_dimensions_",yids[i],sep=""),sprintf("Gene pairs #%d",yids[i]),choices=a.colnames,multiple=TRUE,selected = c("CD4","CD6"),options=list(maxItems=2)))
        updateSelectizeInput(session,paste("yaxis_gene_multi_dimensions_",yids[i],sep=""),choices=a.colnames,selected = c("CD4","CD6"),server=TRUE)
      })
      
    })
    length_ids_multi_dimensions_x <- eventReactive(input$xaxis_add_multi_dimensions,{
      length(xids)
    })
    length_ids_multi_dimensions_y <- eventReactive(input$yaxis_add_multi_dimensions,{
      length(yids)
    })
    observe({
      if (input$xaxis_multi_dimensions=="expression relation pairs" & input$yaxis_multi_dimensions=="expression relation pairs") {
        if (input$data_source_multi_dimensions=="" || is.null(input$data_source_multi_dimensions) ||
            input$treatment_multi_dimensions=="" || is.null(input$treatment_multi_dimensions) ||
            input$description_multi_dimensions=="" || is.null(input$description_multi_dimensions) ||
            input$xaxis_add_multi_dimensions==FALSE || input$yaxis_add_multi_dimensions==FALSE){
          disable("do_multi_dimensions")
        } else {
          enable("do_multi_dimensions")
        }
      } else if (input$xaxis_multi_dimensions=="expression relation pairs" & input$yaxis_multi_dimensions!="expression relation pairs") {
        if (input$data_source_multi_dimensions=="" || is.null(input$data_source_multi_dimensions) ||
            input$treatment_multi_dimensions=="" || is.null(input$treatment_multi_dimensions) ||
            input$description_multi_dimensions=="" || is.null(input$description_multi_dimensions) ||
            input$xaxis_add_multi_dimensions==FALSE){
          disable("do_multi_dimensions")
        } else {
          enable("do_multi_dimensions")
        }
      } else if (input$xaxis_multi_dimensions=="expression relation pairs" & input$yaxis_multi_dimensions!="expression relation pairs") {
        if (input$data_source_multi_dimensions=="" || is.null(input$data_source_multi_dimensions) ||
            input$treatment_multi_dimensions=="" || is.null(input$treatment_multi_dimensions) ||
            input$description_multi_dimensions=="" || is.null(input$description_multi_dimensions) ||
            input$yaxis_add_multi_dimensions==FALSE){
          disable("do_multi_dimensions")
        } else {
          enable("do_multi_dimensions")
        }
      } else {
        if (input$data_source_multi_dimensions=="" || is.null(input$data_source_multi_dimensions) ||
            input$treatment_multi_dimensions=="" || is.null(input$treatment_multi_dimensions) ||
            input$description_multi_dimensions=="" || is.null(input$description_multi_dimensions)){
          disable("do_multi_dimensions")
        } else {
          enable("do_multi_dimensions")
        }
      }
      
    })
    
    observe({
      if (input$xaxis_multi_dimensions=="mutation") {
        output$slider_multi_dimensions_x <- renderUI({
          sliderInput("slider_multi_dimensions_x",
                      "Seperate samples into X-low and X-high groups by setting a cutoff of X variable (0 denotes X-low samples without mutation; 1 denotes X-high samples with mutations): ",
                      min=0,max=1,value=0)  })
      } else if (input$xaxis_multi_dimensions=="expression relation pairs") {
        output$slider_multi_dimensions_x <- renderUI({
          sliderInput("slider_multi_dimensions_x",
                      "Seperate samples into X-low and X-high groups by setting a cutoff of X variable: ",
                      min=0,max=length_ids_multi_dimensions_x(),
                      value=0 ) })
                      # min=min(selecteddatabase_multi_dimensions()[,"sumx"]),max=max(selecteddatabase_multi_dimensions()[,"sumx"]),
                      # value=min(selecteddatabase_multi_dimensions()[,"sumx"]))  })
      } else {
        output$slider_multi_dimensions_x <- renderUI({
          sliderInput("slider_multi_dimensions_x",
                      "Seperate samples into X-low and X-high groups by setting a cutoff of X variable: ",
                      min=0,max=1,value=0.5)  })
      }
      if (input$yaxis_multi_dimensions=="mutation") {
        output$slider_multi_dimensions_y <- renderUI({
          sliderInput("slider_multi_dimensions_y",
                      "Seperate samples into Y-low and Y-high groups by setting a cutoff of Y variable (0 denotes X-low samples without mutation; 1 denotes X-high samples with mutations): ",
                      min=0,max=1,value=0)  })
      } else if (input$yaxis_multi_dimensions=="expression relation pairs") {
        output$slider_multi_dimensions_y <- renderUI({
          sliderInput("slider_multi_dimensions_y",
                      "Seperate samples into Y-low and Y-high groups by setting a cutoff of Y variable: ",
                      min=0,max=length_ids_multi_dimensions_y(),
                      value=0)  })
      } else {
        output$slider_multi_dimensions_y <- renderUI({
          sliderInput("slider_multi_dimensions_y",
                      "Seperate samples into Y-low and Y-high groups by setting a cutoff of Y variable: ",
                      min=0,max=1,value=0.5)  })
      }
    })
    
    selecteddatabase_multi_dimensions <- eventReactive(input$do_multi_dimensions,{
      
      multi_dimensions_expression_pairs <- function(data_source_multi_dimensions,
                                                    treatment_multi_dimensions,
                                                    description_multi_dimensions,
                                                    ids,x_or_y="xaxis_gene_multi_dimensions") {
        # Due to dplyr issue #318, we need temp variables for input values
        data_source1 <- data_source_multi_dimensions
        treatment1 <- treatment_multi_dimensions
        description1 <- description_multi_dimensions
        if (is.null(ids)) {
          genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
        } else {
          genessum<-list()
          #        for (i in 1:length(ids)) {
          txt_ids <- sapply(1:length(ids), function(i) paste(x_or_y,ids[i],sep="_"))
          #        }
          for (i in 1:length(txt_ids)) {
            genessum[[i]] <- input[[ txt_ids[i] ]] 
          }
          genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",unlist(genessum))
        }
        a<-data.table(database.test.process())
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        #    m<-m[,colnames(m) %in% genes1,with=FALSE]
        selected.database<-as.data.frame(m)
        
        if (nrow(selected.database)>0) {
          n <- ncol(selected.database)
          for (i in 1:length(txt_ids)) {
            if (length(input[[ txt_ids[i] ]])==1) {
              selected.database <- data.frame(selected.database,NA)
            } else {
              selected.database <- data.frame(selected.database,selected.database[,input[[ txt_ids[i] ]][1]] > selected.database[,input[[ txt_ids[i] ]][2]])
            }
          }
          colnames(selected.database)[(n+1):ncol(selected.database)] <- unlist(lapply(genessum,function(x) paste(x,collapse=".")))
          cols <- sapply(selected.database, is.logical)
          for (j in (n+1):ncol(selected.database)) {
            selected.database[,j] <- as.numeric(selected.database[,j])
          }
          selected.database <- data.frame(selected.database,sum=NA)
          for (i in 1:nrow(selected.database)) {
            selected.database[i,"sum"] <- sum(na.omit(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)])))
          }
          selected.database
        }
      }
      data_source1 <- str_sub(input$data_source_multi_dimensions,end=-19)
      # treatment1 <- input$treatment_multi_dimensions
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_multi_dimensions])
      description1 <- input$description_multi_dimensions
      
      if (input$xaxis_multi_dimensions!="mutation" & input$yaxis_multi_dimensions!="mutation") {
        
        if (input$xaxis_multi_dimensions=="mutational signatures" ) {
          xgenessum <- str_split(input$xaxis_gene_multi_dimensions," ",2)[[1]][1]
        } else if (input$xaxis_multi_dimensions=="expression relation pairs") {
          if (is.null(xids)) {
            xgenessum<-NULL
          } else {
            xgenessum<-list()
            txt_xids <- sapply(1:length(xids), function(i) paste("xaxis_gene_multi_dimensions_",xids[i],sep=""))
            for (i in 1:length(txt_xids)) {
              xgenessum[[i]] <- input[[ txt_xids[i] ]] 
            }}
        } else {
          xgenessum <- input$xaxis_gene_multi_dimensions
        }
        if (input$yaxis_multi_dimensions=="mutational signatures" ) {
          ygenessum <- str_split(input$yaxis_gene_multi_dimensions," ",2)[[1]][1]
        } else if  (input$yaxis_multi_dimensions=="expression relation pairs") {
          if (is.null(yids)) {
            ygenessum<-NULL
          } else {
            ygenessum<-list()
            #          for (i in 1:length(yids)) {
            txt_yids <- sapply(1:length(yids), function(i) paste("yaxis_gene_multi_dimensions_",yids[i],sep=""))
            #          }
            for (i in 1:length(txt_yids)) {
              ygenessum[[i]] <- input[[ txt_yids[i] ]] 
            }}
        } else {
          ygenessum <- input$yaxis_gene_multi_dimensions
        }
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",xgenessum,ygenessum)
        a<-data.table(database.test.process())
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        m<-as.data.frame(m)
        if (input$xaxis_multi_dimensions=="expression relation pairs" & input$yaxis_multi_dimensions=="expression relation pairs") {
          a1 <- multi_dimensions_expression_pairs(data_source1,#input$data_source_multi_dimensions,
                                                  treatment1,#input$treatment_multi_dimensions,
                                                  description1,#input$description_multi_dimensions,
                                                  ids=xids,x_or_y="xaxis_gene_multi_dimensions")
          colnames(a1)[ncol(a1)] <- "sumx"
          a2 <- multi_dimensions_expression_pairs(data_source1,#input$data_source_multi_dimensions,
                                                  treatment1,#input$treatment_multi_dimensions,
                                                  description1,#input$description_multi_dimensions,
                                                  ids=yids,x_or_y="yaxis_gene_multi_dimensions")
          colnames(a2)[ncol(a2)] <- "sumy"
          a<- cbind(a1,a2)
          a<-a[,!duplicated(colnames(a))]
          as.data.frame(a)
        } else if (input$xaxis_multi_dimensions=="expression relation pairs" & input$yaxis_multi_dimensions!="expression relation pairs") {
          a1 <- multi_dimensions_expression_pairs(data_source1,#input$data_source_multi_dimensions,
                                                  treatment1,#input$treatment_multi_dimensions,
                                                  description1,#input$description_multi_dimensions,
                                                  ids=xids,x_or_y="xaxis_gene_multi_dimensions")
          colnames(a1)[ncol(a1)] <- "sumx"
          m<-data.frame(m,sumy=apply(as.matrix(m[,ygenessum]),1,function(x) sum(na.omit(x))))
          if (nrow(m)==0) {
            m<-as.data.frame(m)
          } else {
            m[apply(as.matrix(m[,ygenessum]),1,function(x) length(na.omit(x))==0),"sumy"]<-NA
            m<-as.data.frame(m)}
          a<- cbind(a1,m)
          a<-a[,!duplicated(colnames(a))]
          as.data.frame(a)
        } else if (input$xaxis_multi_dimensions!="expression relation pairs" & input$yaxis_multi_dimensions=="expression relation pairs") {
          a1 <- multi_dimensions_expression_pairs(data_source1,#input$data_source_multi_dimensions,
                                                  treatment1,#input$treatment_multi_dimensions,
                                                  description1,#input$description_multi_dimensions,
                                                  ids=yids,x_or_y="yaxis_gene_multi_dimensions")
          colnames(a1)[ncol(a1)] <- "sumy"
          m<-data.frame(m,sumx=apply(as.matrix(m[,xgenessum]),1,function(x) sum(na.omit(x))))
          if (nrow(m)==0) {
            m<-as.data.frame(m)
          } else {
            m[apply(as.matrix(m[,xgenessum]),1,function(x) length(na.omit(x))==0),"sumx"]<-NA
            m<-as.data.frame(m)}
          a<- cbind(a1,m)
          a<-a[,!duplicated(colnames(a))]
          as.data.frame(a)
        } else {
          m<-data.frame(m,sumx=apply(as.matrix(m[,xgenessum]),1,function(x) sum(na.omit(x))),sumy=apply(as.matrix(m[,ygenessum]),1,function(x) sum(na.omit(x))))
          if (nrow(m)==0) {
            as.data.frame(m)
          } else {
            m[apply(as.matrix(m[,xgenessum]),1,function(x) length(na.omit(x))==0),"sumx"]<-NA
            m[apply(as.matrix(m[,ygenessum]),1,function(x) length(na.omit(x))==0),"sumy"]<-NA
            as.data.frame(m)} }
      } else if (input$xaxis_multi_dimensions=="mutation" & input$yaxis_multi_dimensions!="mutation"){
        if (input$yaxis_multi_dimensions=="mutational signatures" ) {
          ygenessum <- str_split(input$yaxis_gene_multi_dimensions," ",2)[[1]][1]
        } else if  (input$yaxis_multi_dimensions=="expression relation pairs") {
          if (is.null(yids)) {
            ygenessum<-NULL
          } else {
            ygenessum<-list()
            #            for (i in 1:length(yids)) {
            txt_yids <- sapply(1:length(yids), function(i) paste("yaxis_gene_multi_dimensions_",yids[i],sep=""))
            #            }
            for (i in 1:length(txt_yids)) {
              ygenessum[[i]] <- input[[ txt_yids[i] ]] 
            }}
        } else {
          ygenessum <- input$yaxis_gene_multi_dimensions
        }
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",ygenessum)
        a<-data.table(database.test.process())
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        m<-as.data.frame(m)
        if (input$yaxis_multi_dimensions=="expression relation pairs") {
          a1 <- multi_dimensions_expression_pairs(data_source1,#input$data_source_multi_dimensions,
                                                  treatment1,#input$treatment_multi_dimensions,
                                                  description1,#input$description_multi_dimensions,
                                                  ids=yids,x_or_y="yaxis_gene_multi_dimensions")
          colnames(a1)[ncol(a1)] <- "sumy"
          a1
        } else {
          m<-data.frame(m,sumy=apply(as.matrix(m[,ygenessum]),1,function(x) sum(na.omit(x))))
          if (nrow(m)==0) {
            as.data.frame(m)
          } else {
            m[apply(as.matrix(m[,ygenessum]),1,function(x) length(na.omit(x))==0),"sumy"]<-NA
            as.data.frame(m)}}
      } else if (input$xaxis_multi_dimensions!="mutation" & input$yaxis_multi_dimensions=="mutation"){
        if (input$xaxis_multi_dimensions=="mutational signatures" ) {
          xgenessum <- str_split(input$xaxis_gene_multi_dimensions," ",2)[[1]][1]
        } else if (input$xaxis_multi_dimensions=="expression relation pairs") {
          if (is.null(xids)) {
            xgenessum<-NULL
          } else {
            xgenessum<-list()
            #          for (i in 1:length(xids)) {
            txt_xids <- sapply(1:length(xids), function(i) paste("xaxis_gene_multi_dimensions_",xids[i],sep=""))
            #          }
            for (i in 1:length(txt_xids)) {
              xgenessum[[i]] <- input[[ txt_xids[i] ]] 
            }}
        } else {
          xgenessum <- input$xaxis_gene_multi_dimensions
        }
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",xgenessum)
        a<-data.table(database.test.process())
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        m<-as.data.frame(m)
        if (input$xaxis_multi_dimensions=="expression relation pairs") {
          a1 <- multi_dimensions_expression_pairs(data_source1,#input$data_source_multi_dimensions,
                                                  treatment1,#input$treatment_multi_dimensions,
                                                  description1,#input$description_multi_dimensions,
                                                  ids=xids,x_or_y="xaxis_gene_multi_dimensions")
          colnames(a1)[ncol(a1)] <- "sumx"
          a1
        } else {
          m<-data.frame(m,sumx=apply(as.matrix(m[,xgenessum]),1,function(x) sum(na.omit(x))))
          if (nrow(m)==0) {
            as.data.frame(m)
          } else {
            m[apply(as.matrix(m[,xgenessum]),1,function(x) length(na.omit(x))==0),"sumx"]<-NA
            as.data.frame(m)}}
      } else if (input$xaxis_multi_dimensions=="mutation" & input$yaxis_multi_dimensions=="mutation") {
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
        a<-data.table(database.test.process())
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        as.data.frame(m)
      }
      
    })
    multi_dimensions_seperate_groups <- function (d.my.i, slider.x, slider.y) {
      if (input$xaxis_multi_dimensions=="mutation" & input$yaxis_multi_dimensions=="mutation") {
        ############# mutation mutation##########
        selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% d.my.i[,"PatientID"],]
        if (nrow(selected.mutations)==0) {
          d.my.i <- data.frame(d.my.i,IND="no data available")
          d.my.i
          # plot_ly () %>% 
          #   layout(title = unique(d.my.i$data_source),font=list(size=15,family="Aril")) %>%
          #   add_annotations(x=2,y=2,text="no mutation data available in this dataset",showarrow=F,font=list(size=16,color="purple"))
        } else {
          d.my.i <- data.frame(d.my.i,IND.x="X-low",IND.y="Y-low",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %in% input$xaxis_gene_multi_dimensions,"PatientID"]),"IND.x"] <- "X-high"
          #                     d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %ni% input$xaxis_gene_multi_dimensions,"PatientID"]),"IND.x"] <- "X-low"
          d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %in% input$yaxis_gene_multi_dimensions,"PatientID"]),"IND.y"] <- "Y-high"
          #                     d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %ni% input$yaxis_gene_multi_dimensions,"PatientID"]),"IND.y"] <- "Y-low"
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          d.my.i
        }
      } else if (input$xaxis_multi_dimensions=="mutation" & input$yaxis_multi_dimensions!="mutation") {
        ############## mutation non-mutation ################
        selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% d.my.i[,"PatientID"],]
        d.y <- d.my.i[!is.na(d.my.i[,"sumy"]),]
        if (nrow(selected.mutations)==0 & nrow(d.y)==0 ) {
          ## no mutation data, no d.my.i data ##
          d.my.i <- data.frame(d.my.i,IND="no data available")
          d.my.i
          # plot_ly () %>% 
          #   layout(title = unique(d.my.i$data_source),font=list(size=15,family="Aril")) %>%
          #   add_annotations(x=2,y=2,text="no data available in this dataset",showarrow=F,font=list(size=16,color="purple"))
          
        } else if (nrow(selected.mutations)==0 & nrow(d.y)!=0 ) {
          ## no mutation data, have d.my.i data ##
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumy"]),]
          d.y.na <- d.my.i[is.na(d.my.i[,"sumy"]),]
          d.my.i <- data.frame(d.my.i,IND.x="NA",IND.y="Y-low",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          #d.my.i[,"sumy"] <- unfactor(d.my.i[,"sumy"])
          d.my.i[,"sumy"] <- as.numeric(d.my.i[,"sumy"])
          if (input$yaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumy"] > slider.y,"IND.y"] <- "Y-high"
          } else {
            d.my.i[d.my.i[,"sumy"] > quantile(d.my.i[,"sumy"],slider.y),"IND.y"] <- "Y-high"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          rbind.fill(d.my.i,d.y.na)
        } else if (nrow(selected.mutations)!=0 & nrow(d.y)==0 ) {
          ## have mutation data, no d.my.i data ##
          d.my.i <- data.frame(d.my.i,IND.x="X-low",IND.y="NA",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %in% input$xaxis_gene_multi_dimensions,"PatientID"]),"IND.x"] <- "X-high"
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          d.my.i
        } else if (nrow(selected.mutations)!=0 & nrow(d.y)!=0) {
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumy"]),]
          d.y.na <- d.my.i[is.na(d.my.i[,"sumy"]),]
          d.my.i <- data.frame(d.my.i,IND.x="X-low",IND.y="Y-low",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          # d.my.i[,"sumy"] <- unfactor(d.my.i[,"sumy"])
          d.my.i[,"sumy"] <- as.numeric(d.my.i[,"sumy"])
          d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %in% input$xaxis_gene_multi_dimensions,"PatientID"]),"IND.x"] <- "X-high"
         if (input$yaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumy"] > slider.y,"IND.y"] <- "Y-high"
          } else {
            d.my.i[d.my.i[,"sumy"] > quantile(d.my.i[,"sumy"],slider.y),"IND.y"] <- "Y-high"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          rbind(d.my.i,d.y.na)
        }
        
        
      } else if (input$xaxis_multi_dimensions!="mutation" & input$yaxis_multi_dimensions=="mutation") {
        ###################### non mutation, mutation ####################################
        selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% d.my.i[,"PatientID"],]
        d.x <- d.my.i[!is.na(d.my.i[,"sumx"]),]
        if (nrow(d.x)==0 & nrow(selected.mutations)==0) {
          ## no d.my.i, and.my.i no mutation ##
          d.my.i <- data.frame(d.my.i,IND="no data available")
          d.my.i
          # plot_ly () %>% 
          #   layout(title = unique(d.my.i$data_source),font=list(size=15,family="Aril")) %>%
          #   add_annotations(x=2,y=2,text="no data available in this dataset",showarrow=F,font=list(size=16,color="purple"))
        } else if (nrow(d.x)!=0 & nrow(selected.mutations)==0) {
          ## have d.my.i, no mutation##
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumx"]),]
          d.x.na <- d.my.i[is.na(d.my.i[,"sumx"]),]
          d.my.i <- data.frame(d.my.i,IND.x="X-low",IND.y="NA",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          # d.my.i[,"sumx"] <- unfactor(d.my.i[,"sumx"])
          d.my.i[,"sumx"] <- as.numeric(d.my.i[,"sumx"])
          if (input$xaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumx"] > slider.x,"IND.x"] <- "X-high"
          } else {
            d.my.i[d.my.i[,"sumx"] > quantile(d.my.i[,"sumx"],slider.x),"IND.x"] <- "X-high"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          rbind.fill(d.my.i,d.x.na)
        } else if (nrow(d.x)==0 & nrow(selected.mutations)!=0) {
          ## no d.my.i, have mutation##
          d.my.i <- data.frame(d.my.i,IND.x="NA",IND.y="Y-low",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %in% input$yaxis_gene_multi_dimensions,"PatientID"]),"IND.y"] <- "Y-high"
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          d.my.i
        } else if (nrow(d.x)!=0 & nrow(selected.mutations)!=0) {
          ## have d.my.i, have mutation ##
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumx"]),]
          d.x.na <- d.my.i[is.na(d.my.i[,"sumx"]),]
          d.my.i <- data.frame(d.my.i,IND.x="X-low",IND.y="Y-low",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          
          d.my.i[d.my.i[,"PatientID"] %in% unique(selected.mutations[selected.mutations[,"gene"] %in% input$yaxis_gene_multi_dimensions,"PatientID"]),"IND.y"] <- "Y-high"
          #d.x <- d.my.i[!is.na(d.my.i[,"sumx"]),]
          
          # d.my.i[,"sumx"] <- unfactor(d.my.i[,"sumx"])
          d.my.i[,"sumx"] <- as.numeric(d.my.i[,"sumx"])
          if (input$xaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumx"] > slider.x,"IND.x"] <- "X-high"
          } else {
            d.my.i[d.my.i[,"sumx"] > quantile(d.my.i[,"sumx"],slider.x),"IND.x"] <- "X-high"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          rbind.fill(d.my.i,d.x.na)
        }
        
      } else if (input$xaxis_multi_dimensions!="mutation" & input$yaxis_multi_dimensions!="mutation") {
        ############## non mutation , non mutation ############
        d.x <- d.my.i[!is.na(d.my.i[,"sumx"]) ,]
        d.x.na <- d.my.i[is.na(d.my.i[,"sumx"]) ,]
        d.y <- d.my.i[!is.na(d.my.i[,"sumy"]) ,]
        d.y.na <- d.my.i[is.na(d.my.i[,"sumy"]) ,]
        if (nrow(d.x)==0 & nrow(d.y)==0) {
          ## no d.x, no d.y ##
          d.my.i <- data.frame(d.my.i,IND="no data available")
          #d.my.i
          # plot_ly () %>% 
          #   layout(title = unique(d.my.i$data_source),font=list(size=15,family="Aril")) %>%
          #   add_annotations(x=2,y=2,text="no data available in this dataset",showarrow=F,font=list(size=16,color="purple"))
        } else if (nrow(d.x)==0 & nrow(d.y)!=0) {
          ## no d.x, have d.y ##
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumy"]) ,]
          d.my.i <- data.frame(d.my.i,IND.x="NA",IND.y="Y-low",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          # d.my.i[,"sumy"] <- unfactor(d.my.i[,"sumy"])
          d.my.i[,"sumy"] <- as.numeric(d.my.i[,"sumy"])
          if (input$yaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumy"] > slider.y,"IND.y"] <- "Y-high"
          } else {
            d.my.i[d.my.i[,"sumy"] > quantile(d.my.i[,"sumy"],slider.y),"IND.y"] <- "Y-high"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          #d.my.i
        } else if (nrow(d.x)!=0 & nrow(d.y)==0) {
          ## no d.x, have d.y ##
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumx"]) ,]
          d.my.i <- data.frame(d.my.i,IND.x="X-low",IND.y="NA",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          # d.my.i[,"sumx"] <- unfactor(d.my.i[,"sumx"])
          d.my.i[,"sumx"] <- as.numeric(d.my.i[,"sumx"])
          if (input$xaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumx"] > slider.x,"IND.x"] <- "X-high"
          } else {
            d.my.i[d.my.i[,"sumx"] > quantile(d.my.i[,"sumx"],slider.x),"IND.x"] <- "X-high"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          #d.my.i
        } else if (nrow(d.x)!=0 & nrow(d.y)!=0) {
          ## have d.x, have d.y ##
          d.my.i <- d.my.i[!is.na(d.my.i[,"sumx"]) & !is.na(d.my.i[,"sumy"]),]
          d.my.i <- data.frame(d.my.i,IND.x="NA",IND.y="NA",IND="NA")
          d.my.i[,"IND.x"] <- as.character(d.my.i[,"IND.x"])
          d.my.i[,"IND.y"] <- as.character(d.my.i[,"IND.y"])
          d.my.i[,"IND"] <- as.character(d.my.i[,"IND"])
          # d.my.i[,"sumx"] <- unfactor(d.my.i[,"sumx"])
          # d.my.i[,"sumy"] <- unfactor(d.my.i[,"sumy"])
          d.my.i[,"sumx"] <- as.numeric(d.my.i[,"sumx"])
          d.my.i[,"sumy"] <- as.numeric(d.my.i[,"sumy"])
          if (input$xaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumx"] > slider.x,"IND.x"] <- "X-high"
            d.my.i[d.my.i[,"sumx"] <= slider.x,"IND.x"] <- "X-low"
          } else {
            d.my.i[d.my.i[,"sumx"] > quantile(d.my.i[,"sumx"],slider.x),"IND.x"] <- "X-high"
            d.my.i[d.my.i[,"sumx"] <= quantile(d.my.i[,"sumx"],slider.x),"IND.x"] <- "X-low"
          }
          if (input$yaxis_multi_dimensions=="expression relation pairs") {
            d.my.i[d.my.i[,"sumy"] > slider.y,"IND.y"] <- "Y-high"
            d.my.i[d.my.i[,"sumy"] <= slider.y,"IND.y"] <- "Y-low"
          } else {
            d.my.i[d.my.i[,"sumy"] > quantile(d.my.i[,"sumy"],slider.y),"IND.y"] <- "Y-high"
            d.my.i[d.my.i[,"sumy"] <= quantile(d.my.i[,"sumy"],slider.y),"IND.y"] <- "Y-low"
          }
          d.my.i[,"IND"] <- paste(d.my.i[,"IND.x"],d.my.i[,"IND.y"],sep="/")
          #d.my.i
        }
        rbind.fill(d.my.i,rbind(d.x.na,d.y.na))
      }
    }
    selecteddatabase_multi_dimensions_final <- function() {
      selected.database <- selecteddatabase_multi_dimensions()
       m1 <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source) %>% 
          lapply(function(d) { 
            multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)
          })
        m2 <- ldply(m1,data.frame)
        m2 <- m2[,-1]
        # cc <- c("IND.x" %in% colnames(m2),"IND.y" %in% colnames(m2),"IND" %in% colnames(m2))
        # names(cc) <- c("IND.x","IND.y","IND")
        # colnames(m2[,names(cc[cc])]) <- c("category.X","category.Y","category")[cc]
        dd <- c("category.X","category.Y","category")
        names(dd)<-c("IND.x","IND.y","IND")
        if (length(na.omit(dd[colnames(m2)[(ncol(m2)-2):ncol(m2)]]))==1) {
          colnames(m2)[ncol(m2)] <- dd[colnames(m2)[ncol(m2)]]
        } else if (length(na.omit(dd[colnames(m2)[(ncol(m2)-2):ncol(m2)]]))==2) {
          colnames(m2)[(ncol(m2)-1):ncol(m2)] <- dd[colnames(m2)[(ncol(m2)-1):ncol(m2)]]
        } else {
          colnames(m2)[(ncol(m2)-2):ncol(m2)] <- dd[colnames(m2)[(ncol(m2)-2):ncol(m2)]]
        }
        m2
    }
    ##############################
    #### multi-dimensions summary
    ##############################
    output$multi_dimensions_summary_disease_number <- renderText({
      input$do_multi_dimensions ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selecteddatabase_multi_dimensions()$Disease))})})
    output$multi_dimensions_summary_sample_number <- renderText({
      input$do_multi_dimensions ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selecteddatabase_multi_dimensions())})})
    output$multi_dimensions_summary_plots <- renderPlotly({
#      isolate({
      m2 <- selecteddatabase_multi_dimensions_final()
      summary_plot(inputdata=m2,feature="multi_dimensions")
#      })
    })
    

    observeEvent(input$do_multi_dimensions,{
      input$do_multi_dimensions
      isolate({
#        selected.database <- selecteddatabase_multi_dimensions()
        selected.database <- selecteddatabase_multi_dimensions_final()       

        if (nrow(selected.database)==0 ) {
          output[["multi_dimensions_summary_clincial_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_multi_dimensions,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_multi_dimensions,end=-19)]
          for (k in 1:length(input$data_source_multi_dimensions)) {
            local({
              my_i <- k
              multi_dimensions_summary_clinical_plots_name <- paste0("multi_dimensions_summary_clinical_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[multi_dimensions_summary_clinical_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                  if (is.na(names(splitted_list)[my_i])) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    if (nrow(d[!is.na(d[,clinical_server[input$clinical_multi_dimensions]]) & d[,"category"]!="unknown" & !is.na(d[,"category"]),])==0 ) {
                      incProgress(100, detail = "Doing part %")
                      plot_ly () %>% 
                        layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                    } else {
                      d <- d[!is.na(d[,clinical_server[input$clinical_multi_dimensions]]) & d[,"category"]!="unknown" & !is.na(d[,"category"]),]
                      if(input$clinical_multi_dimensions=="Age") {
                        d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                        #d[d[,"Age"]<=10,"Age"] <- "0-10"
                        d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                        d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                      }
                      plot.data <- as.data.frame(unlist(table(d[,c(clinical_server[input$clinical_multi_dimensions],"category")])))
                      colnames(plot.data) <- c("clinical","category","numbers")
                      
                      p<-ggplotly(ggplot(plot.data, aes(fill=category, y=numbers, x=clinical)) + ggtitle(as.character(unique(d$data_source))) + 
                                 geom_bar( stat="identity", position="fill") +ylab("Frequency") + xlab("Clinical Attribute") )
                      incProgress(100, detail = "Doing part %")
                      p
                      
                    } 
                  
                    }
                })
              })
            })
          }
        }
      }) 
    })
    
    
    
    
    
    output$multi_dimensions_summary_download <- downloadHandler(
      filename = function() {
        if (input$multi_dimensions_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$multi_dimensions_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$multi_dimensions_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        selected.database <- unique(selecteddatabase_multi_dimensions_final())
        colnames(selected.database)[which(colnames(selected.database)=="sumx")] <- "Xvariable"
        colnames(selected.database)[which(colnames(selected.database)=="sumy")] <- "Yvariable"
        
        if (input$multi_dimensions_summary_type == "Excel (CSV)") {
          write.csv (selected.database,file,row.names = FALSE)
        } else if (input$multi_dimensions_summary_type == "Text (TSV)") {
          write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$multi_dimensions_summary_type == "JSON") {
          write(toJSON(selected.database),file)
        }
      }
    )
    
    output$multi_dimensions_summary_table <- DT::renderDataTable({
      selected.database <- unique(selecteddatabase_multi_dimensions_final())
      colnames(selected.database)[which(colnames(selected.database)=="sumx")] <- "Xvariable"
      colnames(selected.database)[which(colnames(selected.database)=="sumy")] <- "Yvariable"
      selected.database
    })
    

    ###############################
    #### multi-dimensions plots
    ###############################

    isolate({
    observeEvent(input$do_multi_dimensions,{
      input$do_multi_dimensions
      isolate({
        input_data_source_multi_dimensions <- str_sub(input$data_source_multi_dimensions,end=-19)
        selected.database <- selecteddatabase_multi_dimensions()
        if (nrow(selected.database)==0) {
          output$multi_dimensions_plots_plots_meta <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=paste(str_sub(input$data_source_multi_dimensions,end=-19),collapse = ","),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
          })
          output[["multi_dimensions_plots_plots0"]] <- renderPlotly({
            plot_ly () %>% 
              layout(title = paste(str_sub(input$data_source_multi_dimensions,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          })
          # output$multi_dimensions_plots_plots_combined_p <- renderText({
          #   "NA"
          # })
          # output$multi_dimensions_plots_plots_single_p <- renderText({
          #   "NA"
          # })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_multi_dimensions,end=-19)]
          
          multi_dimensions_plots_plots_combined_p <- NULL
          #multi_dimensions_plots_plots_metadata <- NULL
          for (k in 1:length(input$data_source_multi_dimensions)) {
              my_i <- k
              d<-splitted_list[[my_i]]
              # d.res <- d[d[,"PFS.status"]=="response" & !is.na(d[,"PFS.status"]),]
              # d.nonres <- d[d[,"PFS.status"]=="nonresponse" & !is.na(d[,"PFS.status"]),]
              
                  isolate({
                    
                    # if (is.na(names(splitted_list)[my_i])) {
                    #   tmp <- data.frame(data_source=input$data_source_multi_dimensions[my_i],TE=NA,seTE=NA)
                    # } else {
                    #   if (nrow(d)==0) {
                    #     tmp = data.frame(data_source=input$data_source_multi_dimensions[my_i],TE=NA,seTE=NA)
                    #   } else {
                    #     d <- d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",]
                    #     if (nrow(d)==0) {
                    #       tmp = data.frame(data_source=input$data_source_multi_dimensions[my_i],TE=NA,seTE=NA)
                    #     } else {
                    #       d$category_binary <- NA
                    #       d$response_binary <- NA
                    #       d[d$PFS.status=="nonresponse" & !is.na(d$PFS.status),"response_binary"] <- 0
                    #       d[d$PFS.status=="response" & !is.na(d$PFS.status),"response_binary"] <- 1
                    #       d[d$category!="unknown" & d$category=="no mutation","category_binary"] <- 0
                    #       d[d$category!="unknown" & d$category=="mutated","category_binary"] <- 1
                    #       d$scaled <- scale(as.numeric(d$category_binary))
                    #       if (all(is.na(d$scaled))) {
                    #         tmp = data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
                    #       } else {
                    #         tmp.pre <- summary(glm(response_binary~scaled,data=d,family = binomial))
                    #         if (nrow(tmp.pre$coefficients)>1) {
                    #           tmp <- data.frame(data_source=input$data_source_mutation[my_i],TE=tmp.pre$coefficients[2,1],seTE=tmp.pre$coefficients[2,2])
                    #         } else {
                    #           tmp = data.frame(data_source=input$data_source_mutation[my_i],TE=NA,seTE=NA)
                    #         }
                    #       }
                    #       
                    #       
                    #     }
                    #   }
                    # }
                    # mutation_plots_plots_metadata <- rbind.fill(mutation_plots_plots_metadata,tmp)
                    if (is.na(names(splitted_list)[my_i])) {
                      multi_dimensions_plots_plots_combined_p <- c(multi_dimensions_plots_plots_combined_p,NA)
                    } else {
                      if (nrow(d[!is.na(d[,"PFS.status"]),])==0) {
                        multi_dimensions_plots_plots_combined_p <- c(multi_dimensions_plots_plots_combined_p,NA)
                      } else {
                        d <- multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)

                        if (unique(d[,"IND"])=="unknown") {
                          multi_dimensions_plots_plots_combined_p <- c(multi_dimensions_plots_plots_combined_p,NA)
                        } else {
                          d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","IND")])))
                          colnames(d.plot)[2] <- "Groups"
                          d.plot[,"PFS.status"] <- as.character(d.plot[,"PFS.status"])
                          d.plot[,"Groups"] <- as.character(d.plot[,"Groups"])
                          d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="nonresponse","Freq"]),")",sep="" )
                          d.plot[d.plot[,"PFS.status"]=="response","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="response","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="response","Freq"]),")",sep="" )
                          d.plot[,"Groups"] <- paste(d.plot[,"Groups"]," group",sep="")
                          try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")])))$p.value)

                          if (class(try.test)=="try-error" | is.na(try.test)) {
                            multi_dimensions_plots_plots_combined_p <- c(multi_dimensions_plots_plots_combined_p,NA)
                          } else {
                            multi_dimensions_plots_plots_combined_p <- c(multi_dimensions_plots_plots_combined_p,p.v)
                          }

                        }
                      }

                    }
                  })

          }
          if(all(is.na(multi_dimensions_plots_plots_combined_p))) {
            output$multi_dimensions_plots_plots_combined_p <- renderText({
              "NA"
            })
            output$multi_dimensions_plots_plots_single_p <- renderText({
              "NA"
            })
          } else {
            output$multi_dimensions_plots_plots_combined_p <- renderText({
              signif(fisher.method(rbind(a=na.omit(multi_dimensions_plots_plots_combined_p)))$p.value,2)
            })
            output$multi_dimensions_plots_plots_single_p <- renderText({
              paste0(signif(multi_dimensions_plots_plots_combined_p,2),collapse = ", ")
            })
          }
          
          
          for (k in 1:length(input$data_source_multi_dimensions)) {
            local({
              my_i <- k
              multi_dimensions_plots_plots_name <- paste0("multi_dimensions_plots_plots", my_i)
              d<-splitted_list[[my_i]]
              output[[multi_dimensions_plots_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot_ly () %>% 
                    layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    plot_ly () %>% 
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)
                    
                    if (unique(d[,"IND"])=="unknown") {
                      plot_ly () %>%
                        layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no X/Y available in this dataset",showarrow=F,font=list(size=16,color="purple"))
                    } else {
                      d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","IND")])))
                      colnames(d.plot)[2] <- "Groups"
                      d.plot[,"PFS.status"] <- as.character(d.plot[,"PFS.status"])
                      d.plot[,"Groups"] <- as.character(d.plot[,"Groups"])
                      d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="nonresponse","Freq"]),")",sep="" )
                      d.plot[d.plot[,"PFS.status"]=="response","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="response","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="response","Freq"]),")",sep="" )
                      d.plot[,"Groups"] <- paste(d.plot[,"Groups"]," group",sep="")
                      try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")])))$p.value)
                      
                      if (class(try.test)=="try-error" | is.na(try.test)) {
                        p.v <- " = NA"
                      } else {
                        # p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")])))$p.value
                        # if (p.v < 0.001) {
                        #   p.v <- " < 0.001"
                        # } else {
                          p.v<-paste0(" = ",signif(p.v,2))
                        # } 
                      }
                      define.color <- c('#EE82EE', '#48D1CC', '#9ACD32', '#F08080', '#FF7F50',
                                        '#CF59AE', '#3CB371', '#006666', '#FF1493', '#C0C0C0')
                      names(define.color) <- c("X-low/Y-low group","X-low/Y-high group","X-high/Y-low group","X-high/Y-high group",
                                               "NA/Y-low group","NA/Y-high group","X-low/NA group","X-high/NA group","NA/NA group","no data available group")
                      
                      incProgress(100, detail = "Doing part %")
                      ggplotly(ggplot(d.plot, aes(fill=Groups, y=Freq, x=PFS.status)) + ggtitle(as.character(unique(d$data_source))) + 
                                 theme(plot.title = element_text(size=15,family="Aril"),axis.title.y = element_text(size=15,family="Aril"),axis.text.x=element_text(angle=-45)) +
                                 scale_fill_manual(values=define.color[na.omit(unique(as.character(d.plot[,"Groups"])))]) + 
                                 scale_fill_discrete(name=" ") + 
                                 geom_bar( stat="identity", position="fill") +ylab("Frequency of groups") + xlab(" ") + 
                                 annotate("text",x=1.5,y=1.2,label=paste0("p-value",p.v),size=4))
                      #color=define.color[na.omit(unique(as.character(d.plot[,"Groups"])))]
                      
                    }
                  } 
                  
                }
              })
              })
              })
            })
            
          }
          
        }
      })
    })
    })
    

    #####################################################
    #### multi dimensions survival
    #####################################################

    
    isolate({
    observeEvent(input$do_multi_dimensions,{
      input$do_multi_dimensions
      isolate({
        input_data_source_multi_dimensions <- str_sub(input$data_source_multi_dimensions,end=-19)
        selected.database <- selecteddatabase_multi_dimensions()
        input_adjust_OS_multi_dimensions <- input$adjust_OS_multi_dimensions
        #      if (nrow(selected.database)==0 | nrow(selected.database[is.na(selected.database[,"sumx"]) & is.na(selected.database[,"sumy"]),])==nrow(selected.database)) {
        if (nrow(selected.database)==0) {
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_multi_dimensions,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))

          output$multi_dimensions_survival_plots_combined_p <- renderText({
            "NA"
          })
          output$multi_dimensions_survival_plots_single_p <- renderText({
            "NA"
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_multi_dimensions,end=-19)]

          multi_dimensions_survival_plots_combined_p <- NULL
          for (k in 1:length(input$data_source_multi_dimensions)) {
              my_i <- k
              multi_dimensions_survival_plots_name <- paste0("multi_dimensions_survival_plots", my_i)
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      multi_dimensions_survival_plots_combined_p <- c(multi_dimensions_survival_plots_combined_p,NA)
                    } else {
                      if (nrow(d[!is.na(d[,"PFS.status"]),])==0) {
                        multi_dimensions_survival_plots_combined_p <- c(multi_dimensions_survival_plots_combined_p,NA)
                      } else {
                        d <- multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)
                        if (unique(d[,"IND"])=="unknown") {
                          multi_dimensions_survival_plots_combined_p <- c(multi_dimensions_survival_plots_combined_p,NA)
                        } else {
                          if (nrow(d[!is.na(d[,"OS.status"]),])==0 | nrow(d[!is.na(d[,"OS"]),])==0 ) {
                            multi_dimensions_survival_plots_combined_p <- c(multi_dimensions_survival_plots_combined_p,NA)
                          } else {
                            d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]),]
                            d<-data.frame(d,dead=NA)
                            d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                            d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                            d <- d[!is.na(d[,"IND"]) & !is.na(d[,"dead"]),]
                            d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                            d <- data.frame(data_source=as.character(d$data_source),
                                            dead=as.numeric(d$dead),
                                            OS=as.numeric(d$OS),
                                            Age=as.character(d$Age),
                                            Gender=as.character(d$Gender),
                                            IND=as.character(d[,"IND"]))

                            # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","IND")]))
                            d[,"dead"] <- as.numeric(d[,"dead"])
                            d<-d[order(d[,"IND"]),]
                            define.color <- c('#EE82EE', '#48D1CC', '#9ACD32', '#F08080', '#FF7F50',
                                              '#CF59AE', '#3CB371', '#006666', '#FF1493')
                            names(define.color) <- c("X-low/Y-low","X-low/Y-high","X-high/Y-low","X-high/Y-high",
                                                     "NA/Y-low","NA/Y-high","X-low/NA","X-high/NA","NA/NA")

                            if (input_adjust_OS_multi_dimensions=="None") {
                              diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                              if (class(diff.result.try)=="try-error") {
                                p.value <- NA
                              } else {
                                p.value <- signif(summary(diff.result)$logtest[3],2)
                              }
                            } else {
                              diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_multi_dimensions),data=d) )
                              if (class(diff.result.adjust.try)=="try-error") {
                                p.value <- signif(summary(coxph(Surv(OS,dead)~IND,data=d))$logtest[3],2)

                              } else {
                                p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                              }
                            }
                            # fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                            # plot.result <- ggsurv(fit.result,surv.col = define.color[na.omit(unique(as.character(d[,"IND"])))])
                            # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                            # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                            multi_dimensions_survival_plots_combined_p <- c(multi_dimensions_survival_plots_combined_p,p.value)

                          }
                        }
                      }

                    }
                  })
          }
          if(all(is.na(multi_dimensions_survival_plots_combined_p))) {
            output$multi_dimensions_survival_plots_combined_p <- renderText({
              "NA"
            })
            output$multi_dimensions_survival_plots_single_p <- renderText({
              "NA"
            })
          } else {
            output$multi_dimensions_survival_plots_combined_p <- renderText({
              signif(fisher.method(rbind(a=na.omit(multi_dimensions_survival_plots_combined_p)))$p.value,2)
            })
            output$multi_dimensions_survival_plots_single_p <- renderText({
              paste0(signif(multi_dimensions_survival_plots_combined_p,2),collapse = ", ")
            })
          }


          for (k in 1:length(input$data_source_multi_dimensions)) {
            local({
              my_i <- k
              multi_dimensions_survival_plots_name <- paste0("multi_dimensions_survival_plots", my_i)
              multi_dimensions_survival_summary_tables_name <- paste0("multi_dimensions_survival_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[multi_dimensions_survival_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {
                    d <- multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)
                    if (unique(d[,"IND"])=="unknown") {
                      p1 <- plot_ly () %>%
                        layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no X/Y available in this dataset",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                    } else {
                      if (nrow(d[!is.na(d[,"OS.status"]),])==0 | nrow(d[!is.na(d[,"OS"]),])==0 ) {
                        p1 <- plot_ly () %>%
                          layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                          add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                        table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                      } else {
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]),]
                        d<-data.frame(d,dead=NA)
                        d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                        d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                        d <- d[!is.na(d[,"IND"]) & !is.na(d[,"dead"]),]
                        d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        dead=as.numeric(d$dead),
                                        OS=as.numeric(d$OS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        IND=as.character(d[,"IND"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","IND")]))
                        d[,"dead"] <- as.numeric(d[,"dead"])
                        d<-d[order(d[,"IND"]),]
                        define.color <- c('#EE82EE', '#48D1CC', '#9ACD32', '#F08080', '#FF7F50',
                          '#CF59AE', '#3CB371', '#006666', '#FF1493')
                        names(define.color) <- c("X-low/Y-low","X-low/Y-high","X-high/Y-low","X-high/Y-high",
                                                 "NA/Y-low","NA/Y-high","X-low/NA","X-high/NA","NA/NA")
                        fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                        
                        try.test <- try(plot.result <- ggsurv(fit.result,surv.col = define.color[na.omit(unique(as.character(d[,"IND"])))]))
                        if (class(try.test)=="try-error" | is.na(try.test)) {
                          incProgress(100, detail = "Doing part %")
                          p1 <- plot_ly () %>%
                            layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                            add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                          plot.result <- ggsurv(fit.result,surv.col = define.color[na.omit(unique(as.character(d[,"IND"])))])

                          if (input_adjust_OS_multi_dimensions=="None") {
                            diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                            if (class(diff.result.try)=="try-error") {
                              p.value <- NA
                              table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                            } else {
                              p.value <- signif(summary(diff.result)$logtest[3],2)
                              table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                       signif(summary(diff.result)$logtest[1],2),
                                                                                       " on ",
                                                                                       signif(summary(diff.result)$logtest[2],2),
                                                                                       " df, p-value = ",
                                                                                       signif(summary(diff.result)$logtest[3],2)),
                                                                                br(),
                                                                                paste0("n = ",
                                                                                       signif(summary(diff.result)$n,2),
                                                                                       ", number of events = ",
                                                                                       signif(summary(diff.result)$nevent,2)),
                                                                                style="text-align:left;color:black;font-size:15px"),
                                                options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                          }
                        } else {
                          diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_multi_dimensions),data=d) )
                          if (class(diff.result.adjust.try)=="try-error") {
                            diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                            p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                              " (No ",input_adjust_OS_multi_dimensions," was observed)")

                            table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                       signif(summary(diff.result)$logtest[1],2),
                                                                                       " on ",
                                                                                       signif(summary(diff.result)$logtest[2],2),
                                                                                       " df, p-value = ",
                                                                                       signif(summary(diff.result)$logtest[3],2)),
                                                                                br(),
                                                                                paste0("n = ",
                                                                                       signif(summary(diff.result)$n,2),
                                                                                       ", number of events = ",
                                                                                       signif(summary(diff.result)$nevent,2)),
                                                                                style="text-align:left;color:black;font-size:15px"),
                                                options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                          } else {
                            p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                            table1.pre <- signif(coef(summary(diff.result.adjust))[1:3,],2)
                            rownames(table1.pre)[2] <- input_adjust_OS_multi_dimensions
                            table1 <- datatable(table1.pre,
                                                caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                       signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                       " on ",
                                                                                       signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                       " df, p-value = ",
                                                                                       signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                br(),
                                                                                paste0("n = ",
                                                                                       signif(summary(diff.result.adjust)$n,2),
                                                                                       ", number of events = ",
                                                                                       signif(summary(diff.result.adjust)$nevent,2)),
                                                                                style="text-align:left;color:black;font-size:15px"),
                                                options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                          }
                        }
                        # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                        # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                          ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                          #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                   paste0("p-value = ",p.value),x=300,y=0.05)) +
                          ggplot2::guides(color=FALSE,linetype=FALSE)
                        p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                          layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                 xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                 yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black"))
                                 ) #%>%
                        incProgress(100, detail = "Doing part %")
                        p1
                        }
                      }
                    }
                  }

                }
              })
              })
              #})
                output[[multi_dimensions_survival_plots_name]] <- renderPlotly({
                  p1
                })
                output[[multi_dimensions_survival_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })

          }

        }
      })
    })
    })
    
 
    ##########################################################
    #### multi dimensions PFS
    #########################################################

    
    isolate({
    observeEvent(input$do_multi_dimensions,{
      input$do_multi_dimensions
      isolate({
        input_data_source_multi_dimensions <- str_sub(input$data_source_multi_dimensions,end=-19)
        selected.database <- selecteddatabase_multi_dimensions()
        input_adjust_OS_multi_dimensions <- input$adjust_OS_multi_dimensions
        if (nrow(selected.database)==0) {
          plot_ly () %>%
            layout(title = paste(str_sub(input$data_source_multi_dimensions,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))

          output$multi_dimensions_PFS_plots_combined_p <- renderText({
            "NA"
          })
          output$multi_dimensions_PFS_plots_single_p <- renderText({
            "NA"
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_multi_dimensions,end=-19)]

          multi_dimensions_PFS_plots_combined_p <- NULL
          for (k in 1:length(input$data_source_multi_dimensions)) {
              my_i <- k
              d<-splitted_list[[my_i]]
                  isolate({
                    if (is.na(names(splitted_list)[my_i])) {
                      multi_dimensions_PFS_plots_combined_p <- c(multi_dimensions_PFS_plots_combined_p,NA)
                    } else {
                      if (nrow(d[!is.na(d[,"PFS.status"]),])==0) {
                        multi_dimensions_PFS_plots_combined_p <- c(multi_dimensions_PFS_plots_combined_p,NA)
                      } else {
                        d <- multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)

                        if (unique(d[,"IND"])=="unknown") {
                          multi_dimensions_PFS_plots_combined_p <- c(multi_dimensions_PFS_plots_combined_p,NA)
                        } else {
                          if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d[!is.na(d[,"PFS"]),])==0 ) {
                            multi_dimensions_PFS_plots_combined_p <- c(multi_dimensions_PFS_plots_combined_p,NA)
                          } else {
                            d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]),]
                            d <- data.frame(data_source=as.character(d$data_source),
                                            PFS.status=as.character(d$PFS.status),
                                            PFS=as.numeric(d$PFS),
                                            Age=as.character(d$Age),
                                            Gender=as.character(d$Gender),
                                            IND=as.character(d[,"IND"]))
                            # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","IND")]))
                            d <- data.frame(d,group.IND=0)
                            d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                            d<-d[order(d[,"IND"]),]
                            define.color <- c('#EE82EE', '#48D1CC', '#9ACD32', '#F08080', '#FF7F50',
                                              '#CF59AE', '#3CB371', '#006666', '#FF1493')
                            names(define.color) <- c("X-low/Y-low","X-low/Y-high","X-high/Y-low","X-high/Y-high",
                                                     "NA/Y-low","NA/Y-high","X-low/NA","X-high/NA","NA/NA")

                            if (input_adjust_OS_multi_dimensions=="None") {
                              diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                              if (class(diff.result.try)=="try-error") {
                                p.value <- NA
                              } else {
                                p.value <- signif(summary(diff.result)$logtest[3],2)
                              }
                            } else {
                              diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_multi_dimensions),data=d) )
                              if (class(diff.result.adjust.try)=="try-error") {
                                p.value <- signif(summary(coxph(Surv(PFS,group.IND)~IND,data=d))$logtest[3],2)

                              } else {
                                p.value <- signif(summary(diff.result.adjust)$logtest[3],2)
                              }
                            }
                            # fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                            # plot.result <- ggsurv(fit.result,surv.col = define.color[na.omit(unique(as.character(d[,"IND"])))])
                            # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                            # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                            multi_dimensions_PFS_plots_combined_p <- c(multi_dimensions_PFS_plots_combined_p,p.value)

                          }
                        }
                      }

                    }
                  })
          }
          if(all(is.na(multi_dimensions_PFS_plots_combined_p))) {
            output$multi_dimensions_PFS_plots_combined_p <- renderText({
              "NA"
            })
            output$multi_dimensions_PFS_plots_single_p <- renderText({
              "NA"
            })
          } else {
            output$multi_dimensions_PFS_plots_combined_p <- renderText({
              signif(fisher.method(rbind(a=na.omit(multi_dimensions_PFS_plots_combined_p)))$p.value,2)
            })
            output$multi_dimensions_PFS_plots_single_p <- renderText({
              paste0(signif(multi_dimensions_PFS_plots_combined_p,2),collapse = ", ")
            })
          }


          for (k in 1:length(input$data_source_multi_dimensions)) {
            local({
              my_i <- k
              multi_dimensions_PFS_plots_name <- paste0("multi_dimensions_PFS_plots", my_i)
              multi_dimensions_PFS_summary_tables_name <- paste0("multi_dimensions_PFS_summary_tables", my_i)
              d<-splitted_list[[my_i]]
              #output[[multi_dimensions_PFS_plots_name]] <- renderPlotly({
                withProgress(message = 'Making plot', value = 0, {
                isolate({
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  p1 <- plot_ly () %>%
                    layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                    add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                  table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]),])==0) {
                    incProgress(100, detail = "Doing part %")
                    p1 <- plot_ly () %>%
                      layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                      add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                    table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                  } else {
                    d <- multi_dimensions_seperate_groups(d.my.i=d, slider.x=input$slider_multi_dimensions_x, slider.y=input$slider_multi_dimensions_y)

                    if (unique(d[,"IND"])=="unknown") {
                      p1 <- plot_ly () %>%
                        layout(title = str_sub(input$data_source_multi_dimensions,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                        add_annotations(x=2,y=2,text="no X/Y available in this dataset",showarrow=F,font=list(size=16,color="purple"))
                      table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                    } else {
                      if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d[!is.na(d[,"PFS"]),])==0 ) {
                        p1 <- plot_ly () %>%
                          layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                          add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                        table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                      } else {
                        d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]),]
                        d <- data.frame(data_source=as.character(d$data_source),
                                        PFS.status=as.character(d$PFS.status),
                                        PFS=as.numeric(d$PFS),
                                        Age=as.character(d$Age),
                                        Gender=as.character(d$Gender),
                                        IND=as.character(d[,"IND"]))
                        # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","IND")]))
                        d <- data.frame(d,group.IND=0)
                        d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                        d<-d[order(d[,"IND"]),]
                        define.color <- c('#EE82EE', '#48D1CC', '#9ACD32', '#F08080', '#FF7F50',
                                          '#CF59AE', '#3CB371', '#006666', '#FF1493')
                        names(define.color) <- c("X-low/Y-low","X-low/Y-high","X-high/Y-low","X-high/Y-high",
                                                 "NA/Y-low","NA/Y-high","X-low/NA","X-high/NA","NA/NA")
                        fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                        
                        try.test <- try(plot.result <- ggsurv(fit.result,surv.col = define.color[na.omit(unique(as.character(d[,"IND"])))]))
                        if (class(try.test)=="try-error" | is.na(try.test)) {
                          incProgress(100, detail = "Doing part %")
                          p1 <- plot_ly () %>%
                            layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                            add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                          table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        } else {
                        
                          plot.result <- ggsurv(fit.result,surv.col = define.color[na.omit(unique(as.character(d[,"IND"])))])

                          if (input_adjust_OS_multi_dimensions=="None") {
                            diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                            if (class(diff.result.try)=="try-error") {
                              p.value <- NA
                              table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                            } else {
                              p.value <- signif(summary(diff.result)$logtest[3],2)
                              table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                       signif(summary(diff.result)$logtest[1],2),
                                                                                       " on ",
                                                                                       signif(summary(diff.result)$logtest[2],2),
                                                                                       " df, p-value = ",
                                                                                       signif(summary(diff.result)$logtest[3],2)),
                                                                                br(),
                                                                                paste0("n = ",
                                                                                       signif(summary(diff.result)$n,2),
                                                                                       ", number of events = ",
                                                                                       signif(summary(diff.result)$nevent,2)),
                                                                                style="text-align:left;color:black;font-size:15px"),
                                                options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                          }
                        } else {
                          diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_multi_dimensions),data=d) )
                          if (class(diff.result.adjust.try)=="try-error") {
                            diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                            p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                              " (No ",input_adjust_OS_multi_dimensions," was observed)")

                            table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                       signif(summary(diff.result)$logtest[1],2),
                                                                                       " on ",
                                                                                       signif(summary(diff.result)$logtest[2],2),
                                                                                       " df, p-value = ",
                                                                                       signif(summary(diff.result)$logtest[3],2)),
                                                                                br(),
                                                                                paste0("n = ",
                                                                                       signif(summary(diff.result)$n,2),
                                                                                       ", number of events = ",
                                                                                       signif(summary(diff.result)$nevent,2)),
                                                                                style="text-align:left;color:black;font-size:15px"),
                                                options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                          } else {
                            p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                            table1.pre <- signif(coef(summary(diff.result.adjust))[1:3,],2)
                            rownames(table1.pre)[2] <- input_adjust_OS_multi_dimensions
                            table1 <- datatable(table1.pre,
                                                caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                       signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                       " on ",
                                                                                       signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                       " df, p-value = ",
                                                                                       signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                br(),
                                                                                paste0("n = ",
                                                                                       signif(summary(diff.result.adjust)$n,2),
                                                                                       ", number of events = ",
                                                                                       signif(summary(diff.result.adjust)$nevent,2)),
                                                                                style="text-align:left;color:black;font-size:15px"),
                                                options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                          }
                        }
                        # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                        # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                        plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                          ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                         #        pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                   paste0("p-value = ",p.value),x=300,y=0.05)) +
                          ggplot2::guides(color=FALSE,linetype=FALSE)
                        p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                          layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                 xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                 yaxis=list(title="Progression survival proportion",titlefont=list(size=20,color="black",family= "Aril black"))
                                 ) #%>%
                        incProgress(100, detail = "Doing part %")
                        p1
                        }
                      }
                    }
                  }

                }
              })
              })
              #})
                output[[multi_dimensions_PFS_plots_name]] <- renderPlotly({
                  p1
                })
                output[[multi_dimensions_PFS_summary_tables_name]] <- renderDataTable(
                  table1
                )
            })

          }

        }
      })
    })
    })
    
 

    ######################################################
    #### single cell gene expression
    ######################################################
    
    ####################### For single-cell data
    ####
    #### need create 1. single_cell_gene_gene; 2. single_cell_gene_cell_matrix
    #### 3. single_cell_cell_marker_gene; 4. Melanoma_Sade_Feldman_datatable
    #### 5. single_cell_cell_tSNEname; 6. single_cell_cell_markername
    #### 7. single_cell_cell_cell
    ####
    cal_gene_cell <- eventReactive (input$do_user_data, {
      if (is.null(input$file_gene_cell)) {
        NULL
      } else {
        gene_cell_matrix <- user_gene_cell()
        Seurat.new <- CreateSeuratObject(counts=gene_cell_matrix,project="user",min.cells=3,min.features = 200)
        Seurat.new <- NormalizeData(Seurat.new,normalization.method = "LogNormalize",scale.factor=10000)
        Seurat.new <- FindVariableFeatures(object = Seurat.new, selection.method = "vst", nfeatures = 1000)
        Seurat.new <- ScaleData(object=Seurat.new)
        Seurat.new <- RunPCA(object=Seurat.new,features=VariableFeatures(object=Seurat.new))
        Seurat.new <- FindNeighbors(object=Seurat.new,dims=1:10)
        Seurat.new <- FindClusters(object=Seurat.new,resolution=0.5)
        Seurat.new <- RunTSNE(object=Seurat.new,dims=1:10)
        return(Seurat.new)
        #Seurat.new <- FindAllMarkers(Seurat.new,only.pos=FALSE,min.pct=0.25,logfc.threshold = 0.25)
      }
    })
    
    # observeEvent(input$do_user_data,{
    #   if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
    #     user_c <- user_clinical()
    #     user_gene_cell <- cal_gene_cell()
    #     user_patient_cell <- user_patient_cell()
    #     colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
    #     colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
    #     ### 1.single_cell_gene_gene
    #     single_cell_gene_gene[[length(single_cell_gene_gene)+1]] <- as.character(unique(rownames(user_gene_cell)))
    #     names(single_cell_gene_gene)[length(single_cell_gene_gene)] <- unique(as.character(user_c$data_source))[1]
    #     ### 2. single_cell_gene_cell_matrix
    #     user_c_patient_cell <- merge(user_c,user_patient_cell,by="PatientID",all=T)
    #     user_c_patient_cell_ident <- merge(user_c_patient_cell,data.frame(ident=user_gene_cell@active.ident),
    #                                        by.x="cells",by.y="row.names",all=T)
    #     user_c_patient_cell_ident_tsne <- merge(user_c_patient_cell_ident,user_gene_cell@reductions$tsne@cell.embeddings,
    #                                             by.x="cells",by.y="row.names")#,all=T)
    #     user_c_patient_cell_ident_tsne_value <- merge(user_c_patient_cell_ident_tsne,t(as.matrix(user_gene_cell@assays$RNA@data)),
    #                                                   by.x="cells",by.y="row.names")#,all=T)   
    #     user_c_patient_cell_ident_tsne_value.pre16 <- user_c_patient_cell_ident_tsne_value[,c("Disease","cells","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","tSNE_1","tSNE_2","ident")]
    #     user_c_patient_cell_ident_tsne_value <- cbind(user_c_patient_cell_ident_tsne_value.pre16,
    #                                                   user_c_patient_cell_ident_tsne_value[,17:ncol(user_c_patient_cell_ident_tsne_value)])
    #     
    #     
    #     single_cell_gene_cell_matrix[[length(single_cell_gene_cell_matrix)+1]] <- user_c_patient_cell_ident_tsne_value
    #     names(single_cell_gene_cell_matrix)[length(single_cell_gene_cell_matrix)] <- unique(as.character(user_c$data_source))[1]
    #     ### 3. single_cell_cell_marker_gene
    #     user_gene_cell_markers <- FindAllMarkers(user_gene_cell,only.pos=FALSE,min.pct=0.25,logfc.threshold = 0.25)
    #     single_cell_cell_marker_gene[[length(single_cell_cell_marker_gene)+1]] <- user_gene_cell_markers
    #     names(single_cell_cell_marker_gene)[length(single_cell_cell_marker_gene)] <- unique(as.character(user_c$data_source))[1]
    #     ### 4. Melanoma_Sade_Feldman_datatable
    #     user_c_patient_cell_ident_datatable <- as.data.frame.matrix(table(user_c_patient_cell_ident[,c("PatientID","ident")]))
    #     user_c_patient_cell_ident_datatable <- data.frame(user_c_patient_cell_ident_datatable,all_cells=apply(user_c_patient_cell_ident_datatable,1,sum))
    #     single_cell_cell_markertablename[[length(single_cell_cell_markertablename)+1]] <- user_c_patient_cell_ident_datatable
    #     names(single_cell_cell_markertablename)[length(single_cell_cell_markertablename)] <- unique(as.character(user_c$data_source))[1]
    #     ### 5. single_cell_cell_tSNEname
    #     png(paste0("./www/",unique(as.character(user_c$data_source))[1],"_tSNE.png"))
    #     DimPlot(user_gene_cell,reduction="tsne")
    #     dev.off()
    #     single_cell_cell_tSNEname[[length(single_cell_cell_tSNEname)+1]] <- paste0(unique(as.character(user_c$data_source))[1],"_tSNE.png")
    #     names(single_cell_cell_tSNEname)[length(single_cell_cell_tSNEname)] <- unique(as.character(user_c$data_source))[1]
    #     ### 6. single_cell_cell_markername
    #     top10 <- user_gene_cell_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
    #     png(paste0("./www/",unique(as.character(user_c$data_source))[1],"_marker_genes.png"))
    #     DoHeatmap(user_gene_cell, features = top10$gene) + NoLegend()
    #     dev.off()
    #     single_cell_cell_markername[[length(single_cell_cell_markername)+1]] <- paste0(unique(as.character(user_c$data_source))[1],"_marker_genes.png")
    #     names(single_cell_cell_markername)[length(single_cell_cell_markername)] <- unique(as.character(user_c$data_source))[1]
    #     ### 7. single_cell_cell_cell
    #     single_cell_cell_cell[[length(single_cell_cell_cell)+1]] <- sort(as.character(unique(user_gene_cell@active.ident)))
    #     names(single_cell_cell_cell)[length(single_cell_cell_cell)] <- unique(as.character(user_c$data_source))[1]
    #     
    #   }
    # })
    
    
    
    all.single.cell.single_cell_gene_gene <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 1.single_cell_gene_gene
        single_cell_gene_gene[[length(single_cell_gene_gene)+1]] <- as.character(unique(rownames(user_gene_cell)))
        names(single_cell_gene_gene)[length(single_cell_gene_gene)] <- unique(as.character(user_c$data_source))[1]
      }
      return(single_cell_gene_gene)
    })
    
    all.single.cell.single_cell_gene_cell_matrix <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 2. single_cell_gene_cell_matrix
        user_c_patient_cell <- merge(user_c,user_patient_cell,by="PatientID",all=T)
        user_c_patient_cell_ident <- merge(user_c_patient_cell,data.frame(ident=user_gene_cell@active.ident),
                                           by.x="cells",by.y="row.names",all=T)
        user_c_patient_cell_ident_tsne <- merge(user_c_patient_cell_ident,user_gene_cell@reductions$tsne@cell.embeddings,
                                                by.x="cells",by.y="row.names")#,all=T)
        user_c_patient_cell_ident_tsne_value <- merge(user_c_patient_cell_ident_tsne,t(as.matrix(user_gene_cell@assays$RNA@data)),
                                                      by.x="cells",by.y="row.names")#,all=T)
        user_c_patient_cell_ident_tsne_value.pre16 <- user_c_patient_cell_ident_tsne_value[,c("Disease","cells","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","tSNE_1","tSNE_2","ident")]
        user_c_patient_cell_ident_tsne_value <- cbind(user_c_patient_cell_ident_tsne_value.pre16,
                                                      user_c_patient_cell_ident_tsne_value[,17:ncol(user_c_patient_cell_ident_tsne_value)])
        
        
        single_cell_gene_cell_matrix[[length(single_cell_gene_cell_matrix)+1]] <- user_c_patient_cell_ident_tsne_value
        names(single_cell_gene_cell_matrix)[length(single_cell_gene_cell_matrix)] <- unique(as.character(user_c$data_source))[1]
        
      }
      return(single_cell_gene_cell_matrix)
    })
    
    all.single.cell.single_cell_cell_marker_gene <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 3. single_cell_cell_marker_gene
        user_gene_cell_markers <- FindAllMarkers(user_gene_cell,only.pos=FALSE,min.pct=0.25,logfc.threshold = 0.25)
        single_cell_cell_marker_gene[[length(single_cell_cell_marker_gene)+1]] <- user_gene_cell_markers
        names(single_cell_cell_marker_gene)[length(single_cell_cell_marker_gene)] <- unique(as.character(user_c$data_source))[1]
      }
      return(single_cell_cell_marker_gene)
    })
    
    all.single.cell.single_cell_cell_markertablename <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 4. Melanoma_Sade_Feldman_datatable
        user_c_patient_cell <- merge(user_c,user_patient_cell,by="PatientID",all=T)
        user_c_patient_cell_ident <- merge(user_c_patient_cell,data.frame(ident=user_gene_cell@active.ident),
                                           by.x="cells",by.y="row.names",all=T)
        user_c_patient_cell_ident_datatable <- as.data.frame.matrix(table(user_c_patient_cell_ident[,c("PatientID","ident")]))
        user_c_patient_cell_ident_datatable <- data.frame(user_c_patient_cell_ident_datatable,all_cells=apply(user_c_patient_cell_ident_datatable,1,sum))
        single_cell_cell_markertablename[[length(single_cell_cell_markertablename)+1]] <- user_c_patient_cell_ident_datatable
        names(single_cell_cell_markertablename)[length(single_cell_cell_markertablename)] <- unique(as.character(user_c$data_source))[1]
      }
      return(single_cell_cell_markertablename)
    })
    
    all.single.cell.single_cell_cell_tSNEname <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 5. single_cell_cell_tSNEname
        png(paste0("./www/",unique(as.character(user_c$data_source))[1],"_tSNE.png"))
        DimPlot(user_gene_cell,reduction="tsne")
        dev.off()
        single_cell_cell_tSNEname[[length(single_cell_cell_tSNEname)+1]] <- paste0(unique(as.character(user_c$data_source))[1],"_tSNE.png")
        names(single_cell_cell_tSNEname)[length(single_cell_cell_tSNEname)] <- unique(as.character(user_c$data_source))[1]
        
      }
      return(single_cell_cell_tSNEname)
    })
    
    all.single.cell.single_cell_cell_markername <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 6. single_cell_cell_markername
        user_gene_cell_markers <- FindAllMarkers(user_gene_cell,only.pos=FALSE,min.pct=0.25,logfc.threshold = 0.25)
        top10 <- user_gene_cell_markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
        png(paste0("./www/",unique(as.character(user_c$data_source))[1],"_marker_genes.png"))
        DoHeatmap(user_gene_cell, features = top10$gene) + NoLegend()
        dev.off()
        single_cell_cell_markername[[length(single_cell_cell_markername)+1]] <- paste0(unique(as.character(user_c$data_source))[1],"_marker_genes.png")
        names(single_cell_cell_markername)[length(single_cell_cell_markername)] <- unique(as.character(user_c$data_source))[1]
      }
      return(single_cell_cell_markername)
    })
    
    all.single.cell.single_cell_cell_cell <- reactive({#})
      #observeEvent(input$do_user_data,{
      if (!is.null(input$file_gene_cell) & !is.null(input$file_patient_cell) & !is.null(input$file_clinical) ) {
        user_c <- user_clinical()
        user_gene_cell <- cal_gene_cell()
        user_patient_cell <- user_patient_cell()
        colnames(user_c)[which(colnames(user_c)=="treatment type")] <- "treatment"
        colnames(user_c)[which(colnames(user_c)=="treatment status")] <- "description"
        ### 8. single_cell_cell_cell
        single_cell_cell_cell[[length(single_cell_cell_cell)+1]] <- sort(as.character(unique(user_gene_cell@active.ident)))
        names(single_cell_cell_cell)[length(single_cell_cell_cell)] <- unique(as.character(user_c$data_source))[1]
      }
      return(single_cell_cell_cell)
    })
    
    observeEvent(input$data_source_single_cell_gene,{
      single_cell_gene_gene <- all.single.cell.single_cell_gene_gene()
      if (length(input$data_source_single_cell_gene)==1) {
        tmp_single_cell_gene_gene <- single_cell_gene_gene[[str_sub(input$data_source_single_cell_gene,end=-19)]]
      } else {
        tmp_single_cell_gene_gene <- NULL
        for (i in 1:length(input$data_source_single_cell_gene)) {
          tmp_single_cell_gene_gene <- unique(c(tmp_single_cell_gene_gene,single_cell_gene_gene[[str_sub(input$data_source_single_cell_gene[i],end=-19)]]))
        }
      }
      updateSelectizeInput(session,"gene_single_cell_gene",
                           choices=tmp_single_cell_gene_gene,#Krieg_panel_1_genes, #rownames(Krieg_panel_1@assays$RNA@counts),
                           selected="CD19",
                           server=TRUE)
    })
    # observeEvent(input$data_source_single_cell_gene,{
    #   if (str_sub(input$data_source_single_cell_gene,end=-19)=="Melanoma-Krieg_panel_1") {
    #     updateSelectizeInput(session,"gene_single_cell_gene",
    #                          choices=Krieg_panel_1_genes, #rownames(Krieg_panel_1@assays$RNA@counts),
    #                          selected="CTLA4",
    #                          server=TRUE)
    #   } else if (str_sub(input$data_source_single_cell_gene,end=-19)=="Melanoma-Krieg_panel_2") {
    #     updateSelectizeInput(session,"gene_single_cell_gene",
    #                          choices=Krieg_panel_2_genes, #rownames(Krieg_panel_2@assays$RNA@counts),
    #                          selected="CTLA4",
    #                          server=TRUE)
    #   } else if (str_sub(input$data_source_single_cell_gene,end=-19)=="Melanoma-Krieg_panel_3") {
    #     updateSelectizeInput(session,"gene_single_cell_gene",
    #                          choices=Krieg_panel_3_genes, #rownames(Krieg_panel_3@assays$RNA@counts),
    #                          selected="CD16",
    #                          server=TRUE)
    #   } else if (str_sub(input$data_source_single_cell_gene,end=-19)=="Melanoma-Sade_Feldman") {
    #     updateSelectizeInput(session,"gene_single_cell_gene",
    #                          choices=Sade_Feldman_genes, #rownames(pbmc@assays$RNA@counts),
    #                          selected="CD19",
    #                          server=TRUE)
    #   } else {
    #     updateSelectizeInput(session,"gene_single_cell_gene",
    #                          choices=unique(c(Sade_Feldman_genes,Krieg_panel_1_genes,Krieg_panel_2_genes,Krieg_panel_3_genes)), #rownames(pbmc@assays$RNA@counts),
    #                          selected="CD19",
    #                          server=TRUE)
    #   }
    # })
    observe({
      if (input$data_source_single_cell_gene=="" || is.null(input$data_source_single_cell_gene) ||
          input$treatment_single_cell_gene=="" || is.null(input$treatment_single_cell_gene) ||
          input$description_single_cell_gene=="" || is.null(input$description_single_cell_gene) ||
          input$gene_single_cell_gene=="" || is.null(input$gene_single_cell_gene)){
        disable("do_single_cell_gene")
      } else {
        enable("do_single_cell_gene")
      }
    })
    selecteddatabase_single_cell_gene <- eventReactive(input$do_single_cell_gene,{
      genes1 <- c("Disease","cells","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","tSNE_1","tSNE_2","ident",input$gene_single_cell_gene)
      m1 <-NULL
      single_cell_gene_cell_matrix <- all.single.cell.single_cell_gene_cell_matrix()
      
      for (i in 1:length(input$data_source_single_cell_gene)) {
        seurat_data_source <- switch (str_sub(input$data_source_single_cell_gene,end=-19)[i],
                                      "Melanoma-Sade_Feldman"=Melanoma_Sade_Feldman,
                                      "Melanoma-Krieg_panel_1"=Melanoma_Krieg_panel_1,
                                      "Melanoma-Krieg_panel_2"=Melanoma_Krieg_panel_2,
                                      "Melanoma-Krieg_panel_3"=Melanoma_Krieg_panel_3)
        m<-seurat_data_source[,colnames(seurat_data_source) %in% genes1]
        m1 <- rbind.fill(m1,m)
        
      }
      data_source1 <- str_sub(input$data_source_single_cell_gene,end=-19)
      # treatment1 <- input$treatment_single_cell_gene
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_single_cell_gene])
      description1 <- input$description_single_cell_gene
      #genes1 <- c("Disease","cell","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","tSNE_1","tSNE_2","ident",input$gene_single_cell_gene)
      #a<-data.table(seurat_data_source)
      m1<-m1 %>% filter(data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1)
      #m<-m[,colnames(m) %in% genes1,with=FALSE]
      if (ncol(m1)>16) {
        colnames(m1)[ncol(m1)] <- "UserGene"
      } 
      m1 <- as.data.frame(m1)
      m1
    })
    
    observeEvent(input$do_single_cell_gene,{
      selected.database <- selecteddatabase_single_cell_gene()
  #####################
  #### single cell gene summary
  #####################
    output$single_cell_gene_summary_disease_number <- renderText({
      input$do_single_cell_gene ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selected.database$Disease))})})
    
    output$single_cell_gene_summary_sample_number <- renderText({
      input$do_single_cell_gene ##keep these two rows
      isolate({ ##keep these two rows
        length(unique(selected.database$PatientID))})})
    
    output$single_cell_gene_summary_cell_number <- renderText({
      input$do_single_cell_gene ##keep these two rows
      isolate({ ##keep these two rows
        nrow(selected.database)})})
    output$single_cell_gene_summary_clinical_plots <- renderUI({
      input$do_single_cell_gene
      isolate({
        if (nrow(selected.database)==0 | ncol(selected.database)<17) {
          tagList(plotlyOutput("single_cell_gene_summary_clinical_plots0"),height="auto")
        } else {
          lapply(1:length(input$data_source_single_cell_gene),function(i) {
            column(width=6,wellPanel(plotOutput(paste0("single_cell_gene_summary_clinical_plots",i),height=480)),style="padding-bottom:20px")
            #column(width=6,wellPanel(plotlyOutput(paste0("single_cell_gene_summary_clinical_plots",i),height="auto")),style="padding-bottom:20px")
#            div(plotlyOutput(paste0("single_cell_gene_summary_clinical_plots",i),height="auto"),style="margin-bottom:20px;")
          })
        }
      })
    })
    
    isolate({
      input_gene_single_cell_gene <- input$gene_single_cell_gene
      if (nrow(selected.database)==0 ) {
        output[["single_cell_gene_summary_clincial_plots0"]] <- renderPlotly({
          plot_ly () %>% 
            layout(title = paste(str_sub(input$data_source_single_cell_gene,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
        })
      } else {
        selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
        splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
        splitted_list <- splitted_list[str_sub(input$data_source_single_cell_gene,end=-19)]
        for (k in 1:length(input$data_source_single_cell_gene)) {
          local({
            my_i <- k
            single_cell_gene_summary_clinical_plots_name <- paste0("single_cell_gene_summary_clinical_plots", my_i)
            d<-splitted_list[[my_i]]
            
            output[[single_cell_gene_summary_clinical_plots_name]] <- renderPlot({  #renderPlotly({
              withProgress(message = 'Making plot', value = 0, {
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
                  abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
                  text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
                  # plot_ly () %>% 
                  #   layout(title = input$data_source_single_cell_gene[[my_i]],font=list(size=15,family="Aril")) %>%
                  #   add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  if (nrow(d[!is.na(d[,clinical_server[input$clinical_single_cell_gene]]) & !is.na(d[,"UserGene"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
                    abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
                    text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
                    # plot_ly () %>% 
                    #   layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    #   add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,clinical_server[input$clinical_single_cell_gene]]) & !is.na(d[,"UserGene"]),]
                    if(input$clinical_single_cell_gene=="Age") {
                      d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                      #d[d[,"Age"]<=10,"Age"] <- "0-10"
                      d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                      d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                    }
                    
                    p <- ggplot(d,aes(x=d[,clinical_server[input$clinical_single_cell_gene]],y=as.numeric(as.character(UserGene)),fill=Groups))+scale_fill_manual(values=c("#DDA0DD", "#90EE90"))+
                      geom_violin()+geom_boxplot(width=0.05,fill="white")+ggtitle(as.character(unique(d$data_source))) +
                      ylab(input_gene_single_cell_gene) + xlab(" ") +
                      theme(plot.title = element_text(size=20,family = "serif"),axis.text=element_text(size=17),axis.title.y = element_text(size=20,family="serif"),axis.text.x=element_text(angle=-45))#+
                      #annotate("text",x=1.5,y=max(as.numeric(d[,"UserGene"])),label=paste0("p-value",p.v),size=6,family="Roboto")
                    
                    # p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_single_cell_gene]],
                    #            y=~as.numeric(as.character(UserGene)), 
                    #            type="violin",#type="box",
                    #            points="all",#boxpoints="all",
                    #            jitter=0.3,
                    #            pointpos=-1.8,
                    #            #mode="markders",
                    #            color=d[,clinical_server[input$clinical_single_cell_gene]],showlegend=TRUE,
                    #            colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                    #            hoveron="points+boxes",
                    #            legendgroup=d[,clinical_server[input$clinical_single_cell_gene]], 
                    #            text= ~paste0("Sample ID: ", PatientID,
                    #                          "<br>Expression: ",UserGene,
                    #                          "<br>Treatment: ",treatment,
                    #                          "<br>Description: ",description)) %>%
                    #   layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                    #          height = 480,
                    #          xaxis=list(title="Clinical Attribute",
                    #                     titlefont=list(size=20,color="black",family= "Aril black"),
                    #                     showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    #          yaxis=list(title=input_gene_single_cell_gene,
                    #                     titlefont=list(size=20,color="black",family= "Aril black"),
                    #                     showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                    #          ))
                    incProgress(100, detail = "Doing part %")
                    p
                    
                  } 
                }
              })
            })
          })
        }
      }
    }) 
    
    
    output$single_cell_gene_summary_plots <- renderPlotly({
      summary_plot(inputdata=selected.database,feature="single_cell_gene")
    })
    output$single_cell_gene_summary_download <- downloadHandler(
      filename = function() {
        if (input$single_cell_gene_summary_type == "Excel (CSV)") {
          "newfile.csv"
        } else if (input$single_cell_gene_summary_type == "Text (TSV)") {
          "newfile.txt"
        } else if (input$single_cell_gene_summary_type == "JSON") {
          "newfile.json"
        }
      },
      content = function(file) {
        if (input$single_cell_gene_summary_type == "Excel (CSV)") {
          write.csv (selected.database,file,row.names = FALSE)
        } else if (input$single_cell_gene_summary_type == "Text (TSV)") {
          write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
        } else if (input$single_cell_gene_summary_type == "JSON") {
          write(toJSON(selected.database),file)
        }
      }
    )
    
    output$single_cell_gene_summary_table <- DT::renderDataTable({
      selected.database
    })
      #################
      #### single cell gene expression plot
      ###################
      output$single_cell_gene_violin_plots <- renderUI({
        input$do_single_cell_gene
        isolate({
          
          if (nrow(selected.database)==0 | ncol(selected.database)<17) {
            tagList(plotOutput("single_cell_gene_violin_plots0",height=480))
          } else {
            lapply(1:length(input$data_source_single_cell_gene),function(i) {
              column(width=6,wellPanel(plotOutput(paste0("single_cell_gene_violin_plots",i),height=480)),style="padding-bottom:20px")
              #div(plotOutput(paste0("single_cell_gene_violin_plots",i),height=480),style="margin-bottom:20px;")
            })
          }
        })
      })
      
    
#      input$do_single_cell_gene
      isolate({
        input_gene_single_cell_gene <- input$gene_single_cell_gene
        
        if (nrow(selected.database)==0 & ncol(selected.database)<=16) {
          output[["single_cell_gene_violin_plots0"]] <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            # plot_ly () %>% 
            #   layout(title = paste(input$data_source_single_cell_gene,collapse = ","),font=list(size=15,family="Aril")) %>%
            #   add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_single_cell_gene,end=-19)]
          for (k in 1:length(input$data_source_single_cell_gene)) {
            local({
              my_i <- k
              single_cell_gene_violin_plots_name <- paste0("single_cell_gene_violin_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[single_cell_gene_violin_plots_name]] <- renderPlot({
                withProgress(message = 'Making plot', value = 0, {
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
                  abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
                  text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
                  # plot_ly () %>% 
                  #   layout(title = input$data_source_single_cell_gene[[my_i]],font=list(size=15,family="Aril")) %>%
                  #   add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
                    abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
                    text(x=2,y=2,"no data available for this gene",col="#800080",cex=1.5)
                    # plot_ly () %>% 
                    #   layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    #   add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),]
                    d[,"Groups"] <- as.character(d[,"Groups"])
                    d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    ind <- sort(unique(d[,"Groups"]))
                    if (as.character(unique(d$data_source)) == "Melanoma-Sade_Feldman") {
                      label_y <- paste(input_gene_single_cell_gene, " expression value (TPM)",sep="")
                    } else {
                      label_y <- paste(input_gene_single_cell_gene, " expression value (CyTOF)",sep="")
                    }
                    try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserGene"]),as.numeric(d[d[,"Groups"]==ind[2],"UserGene"]))$p.value)
                    if (class(try.test)=="try-error" | is.na(try.test)) {
                      p.v <- " = NA"
                    } else {
                      # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserGene"]),as.numeric(d[d[,"Groups"]==ind[2],"UserGene"]))$p.value
                      # if (p.v < 0.001) {
                      #   p.v <- " < 0.001"
                      # } else {
                        p.v<-paste0(" = ",signif(p.v,2))
                      # } 
                    }
                    p <- ggplot(d,aes(x=Groups,y=as.numeric(UserGene),fill=Groups))+scale_fill_manual(values=c("#DDA0DD", "#90EE90"))+
                                geom_violin()+geom_boxplot(width=0.05,fill="white")+ggtitle(as.character(unique(d$data_source))) +
                                ylab(label_y) + xlab(" ") +
                                theme(plot.title = element_text(size=20,family = "serif"),axis.text=element_text(size=17),axis.title.y = element_text(size=20,family="serif"),axis.text.x=element_text(angle=-45))+
                                annotate("text",x=1.5,y=max(as.numeric(d[,"UserGene"])),label=paste0("p-value",p.v),size=6)#,family="Roboto")
                    incProgress(100, detail = "Doing part %")
                    p

                  } 
                }
              })
              })
            })
          }
        }
      }) 
      
      output$single_cell_gene_tSNE_plots <- renderUI({
        input$do_single_cell_gene
        isolate({
          if (nrow(selected.database)==0 | ncol(selected.database)<17) {
            tagList(plotOutput("single_cell_gene_tSNE_plots0",height=480))
          } else {
            lapply(1:length(input$data_source_single_cell_gene),function(i) {
              div(plotOutput(paste0("single_cell_gene_tSNE_plots",i),height=480),style="margin-bottom:20px;")
            })
          }
        })
      })
      
      isolate({
        input_gene_single_cell_gene <- input$gene_single_cell_gene
        
        if (nrow(selected.database)==0 ) {
          output[["single_cell_gene_tSNE_plots0"]] <- renderPlot({
            plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
            abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
            text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
            # plot_ly () %>% 
            #   layout(title = paste(input$data_source_single_cell_gene,collapse = ","),font=list(size=15,family="Aril")) %>%
            #   add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
          })
        } else {
          selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
          splitted_list <- selected.database %>% split(.$data_source) #as.matrix %>% as.data.frame %>% split(.$data_source)
          splitted_list <- splitted_list[str_sub(input$data_source_single_cell_gene,end=-19)]
          for (k in 1:length(input$data_source_single_cell_gene)) {
            local({
              my_i <- k
              single_cell_gene_tSNE_plots_name <- paste0("single_cell_gene_tSNE_plots", my_i)
              d<-splitted_list[[my_i]]
              
              output[[single_cell_gene_tSNE_plots_name]] <- renderPlot({
                withProgress(message = 'Making plot', value = 0, {
                if (is.na(names(splitted_list)[my_i])) {
                  incProgress(100, detail = "Doing part %")
                  plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
                  abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
                  text(x=2,y=2,"no data available under this selection",col="#800080",cex=1.5)
                  # plot_ly () %>% 
                  #   layout(title = input$data_source_single_cell_gene[[my_i]],font=list(size=15,family="Aril")) %>%
                  #   add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                } else {
                  if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),])==0 ) {
                    incProgress(100, detail = "Doing part %")
                    plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=as.character(unique(d$data_source)),cex.main=1.5,bty="l")
                    abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
                    text(x=2,y=2,"no data available for this gene",col="#800080",cex=1.5)
                    # plot_ly () %>% 
                    #   layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                    #   add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                  } else {
                    d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),]
                    d[,"Groups"] <- as.character(d[,"Groups"])
                    d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                    d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                    ind <- sort(unique(d[,"Groups"]))
                    
                    d.response <- d[d[,"PFS.status"]=="response",]
                    d.nonresponse <- d[d[,"PFS.status"]=="nonresponse",]
                   # ggplot(d,aes(x=Groups,y=as.numeric(UserGene),fill=Groups))+scale_fill_manual(values=c("#DDA0DD", "#90EE90"))+
                    #   geom_tSNE()+geom_boxplot(width=0.05,fill="white")+ggtitle(as.character(as.character(unique(d$data_source)))) +
                    #   ylab(label_y) + xlab("Immunotherapy response") +
                    #   theme(plot.title = element_text(size=20,family = "Aril"),axis.text=element_text(size=12),axis.title = element_text(size=17))+
                    #   annotate("text",x=1.5,y=max(as.numeric(d[,"UserGene"])),label=paste0("p-value: ",p.v),size=6)
                    nonres.plot <- ggplot(d.nonresponse) + geom_point(aes(x=tSNE_1, y=tSNE_2,color=UserGene),size=0.3) + scale_color_gradient(low="grey",high = "blue") + 
                      theme(panel.background = element_blank(),axis.title.y = element_text(size=20,family="serif"),
                            axis.title.x = element_text(size=20,family="serif")) + labs(color=input_gene_single_cell_gene)#+ xlab(unique(d.nonresponse$Groups))
                    res.plot <- ggplot(d.response) + geom_point(aes(x=tSNE_1, y=tSNE_2,color=UserGene),size=0.3) + scale_color_gradient(low="grey",high = "blue") + 
                      theme(panel.background = element_blank(),axis.title.y = element_text(size=20,family="serif"),
                            axis.title.x = element_text(size=20,family="serif")) + labs(color=input_gene_single_cell_gene)#+ xlab(unique(d.response$Groups))
                    incProgress(100, detail = "Doing part %")
                    ggarrange.plot <- ggarrange(nonres.plot, res.plot ,labels=c(unique(d.nonresponse$Groups),unique(d.response$Groups)),
                              common.legend=TRUE,legend="right",ncol=2,nrow=1)
                    annotate_figure(ggarrange.plot,
                                    top=text_grob(as.character(unique(d$data_source)),size=20,family="serif"))
                  } 
                }
              })
              })
            })
          }
        }
      }) 
    })


    ######################################################
    #### single cell cell population
    ######################################################
    

     
    observeEvent(input$data_source_single_cell_cell,{
      single_cell_cell_cell <- all.single.cell.single_cell_cell_cell()
      updateSelectizeInput(session,"cluster_single_cell_cell",
                           choices=single_cell_cell_cell[[str_sub(input$data_source_single_cell_cell,end=-19)]], #rownames(Krieg_panel_1@assays$RNA@counts),
                           selected="C04 (IGHD+,VPREB3+,CD19+,PAX5+)",
                           server=TRUE)
    })
    
    
    
    
    output$ui_bsModal_check_first_single_cell_cell <- renderUI({
      single_cell_cell_tSNEname <- all.single.cell.single_cell_cell_tSNEname()
      single_cell_cell_markertablename <- all.single.cell.single_cell_cell_markertablename()
      single_cell_cell_markername <- all.single.cell.single_cell_cell_markername()
      
      tSNEname <- single_cell_cell_tSNEname[[str_sub(input$data_source_single_cell_cell,end=-19)]]
      markername <- single_cell_cell_markername[[str_sub(input$data_source_single_cell_cell,end=-19)]]
      markertablename <- single_cell_cell_markertablename[[str_sub(input$data_source_single_cell_cell,end=-19)]]
      #   switch (str_sub(input$data_source_single_cell_cell,end=-19),
      #                               "Melanoma-Sade_Feldman"="Sade_Feldman_tSNE.png",
      #                               "Melanoma-Krieg_panel_1"="Krieg_panel_1_tSNE.png",
      #                               "Melanoma-Krieg_panel_2"="Krieg_panel_2_tSNE.png",
      #                               "Melanoma-Krieg_panel_3"="Krieg_panel_3_tSNE.png")
      # markername <- switch (str_sub(input$data_source_single_cell_cell,end=-19),
      #                     "Melanoma-Sade_Feldman"="Sade_Feldman_marker_genes.png",
      #                     "Melanoma-Krieg_panel_1"="Krieg_panel_1_marker_genes.png",
      #                     "Melanoma-Krieg_panel_2"="Krieg_panel_2_marker_genes.png",
      #                     "Melanoma-Krieg_panel_3"="Krieg_panel_3_marker_genes.png")
      # markertablename <- switch (str_sub(input$data_source_single_cell_cell,end=-19),
      #                       "Melanoma-Sade_Feldman"=Melanoma_Sade_Feldman_datatable,
      #                       "Melanoma-Krieg_panel_1"=Melanoma_Krieg_panel_1_datatable,
      #                       "Melanoma-Krieg_panel_2"=Melanoma_Krieg_panel_2_datatable,
      #                       "Melanoma-Krieg_panel_3"=Melanoma_Krieg_panel_3_datatable)
      tagList(
        h1("tSNE plot for all cell populations",
           style="height:40px;font-size:23px;color:black;background-color:#DCB239;text-align:center;
                                                          "),
        div(img(src=tSNEname),style="text-align: center;
                                          margin-top: 20px;
                                          margin-bottom: 20px;
                                          height: inherit;"),
        
        h1("Marker genes for all cell populations",
           style="height:40px;font-size:23px;color:black;background-color:#DCB239;text-align:center;
                                                          "),
        div(img(src=markername),style="text-align: center;
                                          margin-top: 20px;
                                          margin-bottom: 20px;
                                          height: inherit;"),
        h1("Statistics for all cell populations",
           style="height:40px;font-size:23px;color:black;background-color:#DCB239;text-align:center;
           "),
        #div(datatable(markertablename,options=list(scrollX=TRUE)),style="text-align:center")
        datatable(markertablename)
      )
      
    })
    
    
    
    observe({
      if (input$data_source_single_cell_cell=="" || is.null(input$data_source_single_cell_cell) ||
          input$treatment_single_cell_cell=="" || is.null(input$treatment_single_cell_cell) ||
          input$description_single_cell_cell=="" || is.null(input$description_single_cell_cell) ||
          input$cluster_single_cell_cell=="" || is.null(input$cluster_single_cell_cell)){
        disable("do_single_cell_cell")
      } else {
        enable("do_single_cell_cell")
      }
    })
    selecteddatabase_single_cell_cell_pre <- eventReactive(input$do_single_cell_cell,{
      input$do_single_cell_cell
      #      Melanoma_Krieg_panel_3 <- Melanomaq_Krieg_panel_3
      single_cell_gene_cell_matrix <- all.single.cell.single_cell_gene_cell_matrix()
      seurat_data_source <- single_cell_gene_cell_matrix[[str_sub(input$data_source_single_cell_cell,end=-19)]]
      # seurat_data_source <- switch (str_sub(input$data_source_single_cell_cell,end=-19),
      #                               "Melanoma-Sade_Feldman"=Melanoma_Sade_Feldman,
      #                               "Melanoma-Krieg_panel_1"=Melanoma_Krieg_panel_1,
      #                               "Melanoma-Krieg_panel_2"=Melanoma_Krieg_panel_2,
      #                               "Melanoma-Krieg_panel_3"=Melanoma_Krieg_panel_3)
      
      # Due to dplyr issue #318, we need temp variables for input values
      data_source1 <- str_sub(input$data_source_single_cell_cell,end=-19)
      treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_single_cell_cell])
      description1 <- input$description_single_cell_cell
      genes1 <- c("Disease","cells","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","tSNE_1","tSNE_2","ident")
      #a<-data.table(each.cell)
      #a<-merge(a,data.frame(cells=names(seurat_data_source@active.ident),cluster=seurat_data_source@active.ident),by="cells",all.x=T)
      m<-seurat_data_source[seurat_data_source$data_source %in% data_source1 & seurat_data_source$treatment %in% treatment1 & seurat_data_source$description %in% description1,]
      #m<-subset(seurat_data_source,data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1)
      
      #m<-m[,colnames(m) %in% genes1,with=FALSE]
      m<-m[,colnames(m) %in% genes1]#,with=FALSE]
      #m<-m[,colnames(m) %in% genes1]
      #as.data.frame(m)
      m
    })
    selecteddatabase_single_cell_cell <- eventReactive(input$do_single_cell_cell,{  
      m <- selecteddatabase_single_cell_cell_pre()
      if (nrow(m)>0 ){
        m.byPatientID <- table(m[,c("PatientID","ident")])
        m.byPatientID.Freq <- matrix(as.numeric(m.byPatientID),nrow(m.byPatientID),ncol(m.byPatientID))
        rownames(m.byPatientID.Freq) <- rownames(m.byPatientID)
        colnames(m.byPatientID.Freq) <- colnames(m.byPatientID)
        # m.byPatientID.Freq <- data.frame(m.byPatientID.Freq,
        #                                  Freq=round(apply(as.data.frame(m.byPatientID.Freq[,as.numeric(substr(input$cluster_single_cell_cell,1,3))+1]),1,sum)/apply(m.byPatientID.Freq[,1:ncol(m.byPatientID.Freq)],1,sum),3))
        m.byPatientID.Freq <- data.frame(m.byPatientID.Freq,
                                         Freq=round(apply(as.data.frame(m.byPatientID.Freq[,input$cluster_single_cell_cell]),1,sum)/apply(m.byPatientID.Freq[,1:ncol(m.byPatientID.Freq)],1,sum),3))
        
        m.byPatientID.Freq.clinical <- merge(unique(m[,c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")]),
                                             m.byPatientID.Freq,
                                             by.x="PatientID",by.y="row.names",all.x=T)
        ### insert category, refer to expression
        m <- m.byPatientID.Freq.clinical 
        m<-data.frame(m,category="unknown")
        m[,"category"] <- as.character(m[,"category"])
        m[!is.na(m[,"Freq"]) & m[,"Freq"]>=quantile(na.omit(m[,"Freq"]),input$slider_single_cell_cell_OS,na.rm = T),"category"] <- paste0("top ",(1-input$slider_single_cell_cell_OS)*100,"% ")
        m[!is.na(m[,"Freq"]) & m[,"Freq"]<quantile(na.omit(m[,"Freq"]),input$slider_single_cell_cell_OS,na.rm = T),"category"] <- paste0("bottom ",input$slider_single_cell_cell_OS*100,"% ")
        m
      } else {
        m
      }
    })
    selecteddatabase_single_cell_cell_marker_gene <- eventReactive(input$do_single_cell_cell,{
      single_cell_cell_marker_gene <- all.single.cell.single_cell_cell_marker_gene()
      single_cell_cell_marker_gene[[str_sub(input$data_source_single_cell_cell,end=-19)]]
      
      # marker_gene_data_source <- switch (str_sub(input$data_source_single_cell_cell,end=-19),
      #                               "Melanoma-Sade_Feldman"=Melanoma_marker_Sade_Feldman,
      #                               "Melanoma-Krieg_panel_1"=Melanoma_marker_Krieg_panel_1,
      #                               "Melanoma-Krieg_panel_2"=Melanoma_marker_Krieg_panel_2,
      #                               "Melanoma-Krieg_panel_3"=Melanoma_marker_Krieg_panel_3)
      # marker_gene_data_source
    })
    
    observeEvent(input$do_single_cell_cell,{
      selected.database <- selecteddatabase_single_cell_cell()
      #################
      #### single cell cell summary
      ###################
      output$single_cell_cell_summary_disease_number <- renderText({
        input$do_single_cell_cell ##keep these two rows
        isolate({ ##keep these two rows
          length(unique(selected.database$Disease))})})
      
      output$single_cell_cell_summary_sample_number <- renderText({
        input$do_single_cell_cell ##keep these two rows
        isolate({ ##keep these two rows
          length(unique(selected.database$PatientID))})})
      
      output$single_cell_cell_summary_cell_number <- renderText({
        input$do_single_cell_cell ##keep these two rows
        isolate({ ##keep these two rows
          nrow(selecteddatabase_single_cell_cell_pre())})})
      
      output$single_cell_cell_summary_plots <- renderPlotly({
        summary_plot(inputdata=selected.database,feature="single_cell_cell")
      })
      output$single_cell_cell_summary_clinical_plots <- renderPlotly({
        input$do_single_cell_cell
        d <- selected.database
        #         isolate({
        if (nrow(selected.database)==0 | length(input$cluster_single_cell_cell)==0) {
          plot_ly () %>% add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
        } else {
          withProgress(message = 'Making plot', value = 0, {
            if (nrow(d[!is.na(d[,clinical_server[input$clinical_single_cell_cell]]) & !is.na(d[,"Freq"]),])==0 ) {
              plot_ly () %>% 
                layout(title = unique(d$data_source),font=list(size=15,family="Aril")) %>%
                add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
            } else {
              d <- d[!is.na(d[,clinical_server[input$clinical_single_cell_cell]]) & !is.na(d[,"Freq"]),]
              if(input$clinical_single_cell_cell=="Age") {
                d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                #d[d[,"Age"]<=10,"Age"] <- "0-10"
                d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
              }
              p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_single_cell_cell]],
                         y=~as.numeric(as.character(Freq)), 
                         type="violin",#type="box",
                         points="all",#boxpoints="all",
                         jitter=0.3,
                         pointpos=-1.8,
                         #mode="markders",
                         color=d[,clinical_server[input$clinical_single_cell_cell]],showlegend=TRUE,
                         colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                         hoveron="points+boxes",
                         legendgroup=d[,clinical_server[input$clinical_single_cell_cell]], 
                         text= ~paste0("Sample ID: ", PatientID,
                                       "<br>Fraction of selected cluster: ",Freq,
                                       "<br>Treatment: ",treatment,
                                       "<br>Description: ",description)) %>%
                layout(title = unique(d$data_source),titlefont=list(size=20,color="black",family= "Aril"),
                       height = 480,
                       xaxis=list(title="Clinical Attribute",
                                  titlefont=list(size=20,color="black",family= "Aril black"),
                                  showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                       yaxis=list(title=paste("Fraction of cluster ",paste(input$cluster_single_cell_cell,collapse = ","),sep=""),
                                  titlefont=list(size=20,color="black",family= "Aril black"),
                                  showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                       ))
              incProgress(100, detail = "Doing part %")
              p
            }
            
          })
        }
        #         })
        
      })
      
      
      
      
      
      output$single_cell_cell_cluster_1 <- renderText({
        input$do_single_cell_cell
        isolate({
          paste(input$cluster_single_cell_cell,collapse = ",")
        })
      })
      output$single_cell_cell_cluster_2 <- renderText({
        input$do_single_cell_cell
        isolate({
          paste(input$cluster_single_cell_cell,collapse = ",")
        })
      })
      output$single_cell_cell_summary_cell_download <- downloadHandler(
        filename = function() {
          if (input$single_cell_cell_summary_cell_type == "Excel (CSV)") {
            "newfile.csv"
          } else if (input$single_cell_cell_summary_cell_type == "Text (TSV)") {
            "newfile.txt"
          } else if (input$single_cell_cell_summary_cell_type == "JSON") {
            "newfile.json"
          }
        },
        content = function(file) {
          if (input$single_cell_cell_summary_cell_type == "Excel (CSV)") {
            write.csv (selecteddatabase_single_cell_cell(),file,row.names = FALSE)
          } else if (input$single_cell_cell_summary_cell_type == "Text (TSV)") {
            write.table (selecteddatabase_single_cell_cell(),file,quote=F, row.names = FALSE, sep="\t")
          } else if (input$single_cell_cell_summary_cell_type == "JSON") {
            write(toJSON(selecteddatabase_single_cell_cell()),file)
          }
        }
      )
      output$single_cell_cell_summary_cell_table <- DT::renderDataTable({
        selecteddatabase_single_cell_cell()
      })
      output$single_cell_cell_summary_marker_gene_download <- downloadHandler(
        filename = function() {
          if (input$single_cell_cell_summary_marker_gene_type == "Excel (CSV)") {
            "newfile.csv"
          } else if (input$single_cell_cell_summary_marker_gene_type == "Text (TSV)") {
            "newfile.txt"
          } else if (input$single_cell_cell_summary_marker_gene_type == "JSON") {
            "newfile.json"
          }
        },
        content = function(file) {
          if (input$single_cell_cell_summary_marker_gene_type == "Excel (CSV)") {
            write.csv (selecteddatabase_single_cell_cell_marker_gene(),file,row.names = FALSE)
          } else if (input$single_cell_cell_summary_marker_gene_type == "Text (TSV)") {
            write.table (selecteddatabase_single_cell_cell_marker_gene(),file,quote=F, row.names = FALSE, sep="\t")
          } else if (input$single_cell_cell_summary_marker_gene_type == "JSON") {
            write(toJSON(selecteddatabase_single_cell_cell_marker_gene()),file)
          }
        }
      )
      output$single_cell_cell_summary_marker_gene_table <- DT::renderDataTable({
        selecteddatabase_single_cell_cell_marker_gene()
      })
      
      #############################################
      ####  single cell cell population plot
      #############################################
      
      output$single_cell_cell_plots_boxplot_cluster <- renderText({
        input$do_single_cell_cell
        isolate({
          paste(input$cluster_single_cell_cell,collapse = ",")
        })
      })
      output$single_cell_cell_plots_tSNE_cluster <- renderText({
        input$do_single_cell_cell
        isolate({
          paste(input$cluster_single_cell_cell,collapse = ",")
        })
      })
      output$single_cell_cell_plots_heatmap_cluster <- renderText({
        input$do_single_cell_cell
        isolate({
          paste(input$cluster_single_cell_cell,collapse = ",")
        })
      })
      
      output$single_cell_cell_plots_boxplot <- renderPlotly({
        #    selecteddatabase_single_cell_cell() :
        #  PatientID      cells data_source treatment description PFS.status RECIST OS.status  OS PFS Age cluster
        #  1 Sade_Feldman_Post_P1 A10_P1_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       1
        #  2 Sade_Feldman_Post_P1 A10_P4_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       2
        #  3 Sade_Feldman_Post_P1 A10_P5_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       1
        input$do_single_cell_cell
        
        isolate({
          #        selected.database <- selecteddatabase_single_cell_cell()
          
          if (nrow(selected.database)==0 | length(input$cluster_single_cell_cell)==0) {
            plot_ly () %>% add_annotations(x=2,y=2,text="no data available",showarrow=F,font=list(size=16,color="purple"))
          } else {
            withProgress(message = 'Making plot', value = 0, {
              d<-data.frame(selected.database,Groups=selected.database$PFS.status)
              if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Freq"]),])==0 ) {
                plot_ly () %>% 
                  layout(title = unique(d$data_source),font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
              } else {
                d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Freq"]),]
                d[,"Groups"] <- as.character(d[,"Groups"])
                d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response",]),")")
                d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse",]),")")
                ind <- unique(d[,"Groups"])
                
                try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Freq"]),as.numeric(d[d[,"Groups"]==ind[2],"Freq"]))$p.value)
                if (class(try.test)=="try-error") {
                  p.v <- " = NA"
                } else {
                  p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Freq"]),as.numeric(d[d[,"Groups"]==ind[2],"Freq"]))$p.value
                  # if (p.v < 0.001) {
                  #   p.v <- " < 0.001"
                  # } else {
                  p.v<-paste0(" = ",signif(p.v,3))
                  # } 
                }
                
                p<-plot_ly(data=d,x= ~Groups,
                           y= ~as.numeric(as.character(Freq)), 
                           type="violin",#type="box",
                           points="all",#boxpoints="all",
                           jitter=0.3,
                           pointpos=-1.8,
                           #mode="markders",
                           color= ~PFS.status,
                           colors=c("plum","lightgreen"),
                           #hoveron="points+boxes",
                           legendgroup=~PFS.status, showlegend=TRUE,
                           text= ~paste0("Sample ID: ", PatientID,
                                         "<br>Fraction: selected cluster(s)"," :",Freq,
                                         "<br>Treatment: ",treatment,
                                         "<br>Description: ",description)
                           
                )  %>%
                  layout(title = unique(d$data_source),titlefont=list(size=20,color="black",family= "Aril"),
                         height = 480,
                         xaxis=list(title=" ",
                                    titlefont=list(size=20,color="black",family= "Aril"),
                                    showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                         yaxis=list(title=paste("Fraction of cluster ",paste(input$cluster_single_cell_cell,collapse = ","),sep=""),
                                    titlefont=list(size=20,color="black",family= "Aril"),
                                    showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                         )) %>% 
                  add_annotations(x = 0.5,
                                  y = d[which.max(d[,"Freq"]),"Freq"],
                                  text = paste0("p-value",p.v),
                                  xref = paste0("x", 1),  # add this
                                  yref = paste0("y", 1),  # add this
                                  showarrow = FALSE,
                                  ax = 20,
                                  ay = -40,
                                  font=list(size=20,color="black",family= "Aril black"))
                incProgress(100, detail = "Doing part %")
                p
              }
              
              
            })
          }
        })
      })
      
          output$single_cell_cell_plots_tSNE <- renderPlot({
            input$do_single_cell_cell
            isolate({
              withProgress(message = 'Making plot', value = 0, {
      #          selected.database <- selecteddatabase_single_cell_cell()
                color.grey <- rep("Grey",20)
                mainPalette=c('#4f00A8', '#CF59AE', '#ace7b9', '#F37812', '#CD5C5C',
                              '#FF8C00', '#7CFC00', '#00FFFF',  '#A80100',  '#000000',  '#cdf0d5',
                              '#F2B079',  '#E9967A',  '#FFFFE0',  '#228B22',  '#AFEEEE','#CF5A59',
                              '#006666',  '#59CF74',  '#888811',  '#FF0000',  '#FFEFD5',  '#ADFF2F',
                              '#00CED1',  '#A80079',  '#00A8A8',  '#002AA8',  '#FDF31C',  '#FF7F50',
                              '#FFDAB9',  '#8FBC8F',  '#008080',  '#BC2D94',   '#009933',   '#5977CF',
                              '#8C8C8C',   '#FFD700',   '#EEE8AA',   '#2E8B57',   '#B0C4DE')
                # for (i in 1:length(input$cluster_single_cell_cell)) { color.grey[as.numeric(substr(input$cluster_single_cell_cell[i],1,3))+1]<-mainPalette[as.numeric(substr(input$cluster_single_cell_cell[i],1,3))+1]}
                # for (i in 1:length(input$cluster_single_cell_cell)) { color.grey[i]<-mainPalette[i]}
                color.grey[as.numeric(str_sub(input$cluster_single_cell_cell,2,3))+1] <- mainPalette[as.numeric(str_sub(input$cluster_single_cell_cell,2,3))+1]
                color.character<-color.grey
                m <- selecteddatabase_single_cell_cell_pre()
                d <- data.frame(m,Groups=m$PFS.status)
                d <- d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"ident"]),]
                d[,"Groups"] <- as.character(d[,"Groups"])
                d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response",]),")")
                d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse",]),")")
                ind <- unique(d[,"Groups"])

                d.response <- d[d[,"PFS.status"]=="response",]
                d.nonresponse <- d[d[,"PFS.status"]=="nonresponse",]

                res.plot <- ggplot(d.response) + geom_point(aes(x=tSNE_1, y=tSNE_2,color=ident),size=0.3) + scale_color_manual(values=color.character) +
                  theme(panel.background = element_blank(),axis.title.y = element_text(size=20,family="serif"),
                        axis.title.x = element_text(size=20,family="serif")) + labs(color="cell populations")#+ xlab(unique(d.nonresponse$Groups))
                nonres.plot <- ggplot(d.nonresponse) + geom_point(aes(x=tSNE_1, y=tSNE_2,color=ident),size=0.3) + scale_color_manual(values=color.character) +
                  theme(panel.background = element_blank(),axis.title.y = element_text(size=20,family="serif"),
                        axis.title.x = element_text(size=20,family="serif")) + labs(color="cell populations")#+ xlab(unique(d.nonresponse$Groups))
                incProgress(100, detail = "Doing part %")
                ggarrange.plot <- ggarrange(nonres.plot, res.plot ,labels=c(unique(d.nonresponse$Groups),unique(d.response$Groups)),
                                            common.legend=TRUE,legend="right",ncol=2,nrow=1)
                annotate_figure(ggarrange.plot,
                                top=text_grob(unique(d$data_source),size=20,family="serif"))
      #          incProgress(100, detail = "Doing part %")
                #    subplot(p3,subplot(style(p1,showlegend=FALSE),style(p2,showlegend=TRUE),margin=0.05,titleX=TRUE,titleY=TRUE),nrows=2,heights=c(1/10,9/10),titleX=TRUE,titleY=TRUE) %>%
                #      layout(title = paste("<br><b>Cluster ",paste(input$cluster_single_cell_cell,collapse=","),"<b>",sep=""),
                #             titlefont=list(size=20,color="black",family= "Aril"))
                #grid.arrange(p1,p2,ncol=2,top=textGrob(paste("Cluster ",paste(input$cluster_single_cell_cell,collapse=","),sep=""),gp=gpar(fontsize=20,font=3)))
              })
            })
          })



      #     output$single_cell_cell_plots_heatmap <- renderPlot({
      #       input$do_single_cell_cell
      #       isolate({
      #         withProgress(message = 'Making plot', value = 0, {
      # #          selected.database <- selecteddatabase_single_cell_cell()
      #           top10 <- selecteddatabase_single_cell_cell_marker_gene() %>% group_by(cluster) %>% top_n(4, wt=avg_logFC)
      #           top10 <- as.data.frame(top10)
      #           a<-top10[top10[,"cluster"] %in% input$cluster_single_cell_cell,]
      #           if (nrow(a)==0) {
      #             incProgress(100, detail = "Doing part %")
      #             plot(1, type="n", xlab="", ylab="", xlim=c(1, 3), ylim=c(1, 3),main=str_sub(input$data_source_single_cell_cell,end=-19),cex.main=1.5,bty="l")
      #             abline(h=1:2.5,v=1:2.5,lty=3,col="lightgray")
      #             text(x=2,y=2,"no signifcant maker genes under this selection",col="#800080",cex=1.5)
      #           } else {
      #           #        b <- selected.database[!is.na(selected.database[,"cluster"]),]
      # 
      #           #        DoHeatmap(object = pbmc, genes.use = top10$gene, slim.col.label = TRUE, remove.key = TRUE,cex.row = 3, use.scaled = FALSE)
      #           #        DoHeatmap(object = pbmc, features = a$gene, cells = as.character(b[b[,"cluster"] %in% input$cluster_single_cell_cell,"cells"]), slot="counts")
      # #            Melanoma_Krieg_panel_3 <- Melanomaq_Krieg_panel_3
      #             seurat_data_source <- switch (str_sub(input$data_source_single_cell_cell,end=-19),
      #                                         "Melanoma-Sade_Feldman"=Melanoma_Sade_Feldman,
      #                                         "Melanoma-Krieg_panel_1"=Melanoma_Krieg_panel_1,
      #                                         "Melanoma-Krieg_panel_2"=Melanoma_Krieg_panel_2,
      #                                         "Melanoma-Krieg_panel_3"=Melanoma_Krieg_panel_3)
      #           data_source1 <- str_sub(input$data_source_single_cell_cell,end=-19)
      #           treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_single_cell_cell])
      #           description1 <- input$description_single_cell_cell
      #           genes1 <- c("Disease","cells","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","tSNE_1","tSNE_2","ident",a$gene)
      #           m<-seurat_data_source[seurat_data_source$data_source %in% data_source1 & seurat_data_source$treatment %in% treatment1 & seurat_data_source$description %in% description1,]
      #           m<-m[,colnames(m) %in% genes1]
      #           m<-as.data.frame(m)
      #           m<- m[!is.na(m[,"ident"]),]
      #           m<-melt(m[,c(2,16:ncol(m))])
      #           m<-m[!is.na(m[,"value"]),]
      #           incProgress(100, detail = "Doing part %")
      #           data.use <- m
      #           genes.use <- a$gene
      # 
      # 
      #           col.low = "#FF00FF"
      #           col.mid = "#000000"
      #           col.high = "#FFFF00"
      #           remove.key = FALSE
      #           rotate.key = FALSE
      #           group.spacing = 0.15
      #           cex.col = 15
      #           cex.row = 20
      #           group.cex = 20
      #           group.by = "ident"
      #           group.label.loc = "top"
      #           group.label.rot = TRUE
      #           draw.line = TRUE
      # 
      # 
      #           if (rotate.key) {
      #             key.direction <- "horizontal"
      #             key.title.pos <- "top"
      #           } else {
      #             key.direction <- "vertical"
      #             key.title.pos <- "top"
      #           }
      # 
      #           heatmap <- ggplot(
      #             data = data.use,
      #             mapping = aes(x = cells, y = variable, fill = value)
      #           ) +
      #             geom_tile() +
      #             scale_fill_gradient2(
      #               low = col.low,
      #               mid = col.mid,
      #               high = col.high,
      #               name = "Expression",
      #               guide = guide_colorbar(
      #                 direction = key.direction,
      #                 title.position = key.title.pos
      #               )
      #             ) +
      #             scale_y_discrete(position = "top", labels = rev(genes.use)) +
      #             theme(
      #               axis.title.x = element_blank(),
      #               strip.text.x = element_text(size = group.cex),
      #               axis.text.x = element_text(size = cex.col),
      #               axis.text.y = element_text(size = cex.row),
      #               #axis.text.x = element_blank(),
      #               axis.ticks.x = element_blank(),
      #               axis.line = element_blank(),
      #               axis.title.y = element_blank(),
      #               axis.ticks.y = element_blank()
      #               # axis.line = element_blank(),
      #               # axis.title.y = element_blank(),
      #               # axis.ticks.y = element_blank(),
      #               # strip.text.x = element_text(),
      #               # axis.text.y = element_text(),
      #               # axis.text.x = element_text(),
      #               # axis.title.x = element_blank()
      #             )
      # 
      #           #heatmap <- heatmap + theme(axis.text.x = element_text(angle = 60))
      # 
      #           if (!is.null(x = group.by)) {
      #             if (group.label.loc == "top") {
      #               switch <- NULL
      #             } else {
      #               switch <- 'x'
      #             }
      #             heatmap <- heatmap +
      #               facet_grid(
      #                 facets = ~ident,
      #                 drop = TRUE,
      #                 space = "free",
      #                 scales = "free",
      #                 switch = switch
      #               ) +
      #               scale_x_discrete(expand = c(0, 0), drop = TRUE)
      #             if (draw.line) {
      #               panel.spacing <- unit(x = group.spacing, units = 'lines')
      #             } else {
      #               panel.spacing <- unit(x = 0, units = 'lines')
      #             }
      #             heatmap <- heatmap +
      #               theme(strip.background = element_blank(), panel.spacing = panel.spacing)
      #             if (group.label.rot) {
      #               heatmap <- heatmap + theme(strip.text.x = element_text(angle = 60))
      #             }
      #           }
      #           if (remove.key) {
      #             heatmap <- heatmap + theme(legend.position = "none")
      #           }
      #           # if (!is.null(x = title)) {
      #           #   heatmap <- heatmap + labs(title = title)
      #           # }
      #           incProgress(100, detail = "Doing part %")
      #           heatmap
      #           #ggplot(m,mapping=aes(x=cells,y=variable,fill=value)) + geom_tile() + scale_fill_gradient2(low="black",mid="black",high="red") + facet_wrap(~ident)
      # 
      #           # if (length(seurat_data_source@active.ident)>25000) {
      #           #   DoHeatmap(object = seurat_data_source, features = a$gene, cells = sample(names(seurat_data_source@active.ident),25000), slot = "data")
      #           # } else {
      #           #   DoHeatmap(object = seurat_data_source, features = a$gene, slot = "data")
      #           # }
      # #          })
      #           }
      # 
      # #        })
      #       })
      #       })
      #     })
      
      
      #########################################
      #### single cell cell survival
      #########################################
      
      # output$single_cell_cell_survival_plots <- renderPlotly({
        observeEvent(input$do_single_cell_cell,{
        #    selecteddatabase_single_cell_cell() :
        #  PatientID      cells data_source treatment description PFS.status RECIST OS.status  OS PFS Age cluster
        #  1 Sade_Feldman_Post_P1 A10_P1_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       1
        #  2 Sade_Feldman_Post_P1 A10_P4_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       2
        #  3 Sade_Feldman_Post_P1 A10_P5_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       1
        #      input$do_single_cell_cell
        isolate({
          input_adjust_OS_single_cell_cell <- input$adjust_OS_single_cell_cell
          #        selected.database <- selecteddatabase_single_cell_cell()
          if (nrow(selected.database)==0 | length(input$cluster_single_cell_cell)==0 | nrow(selected.database[!is.na(selected.database[,"OS"]),])==0) {
            p1 <- plot_ly () %>%
              layout(title = as.character(unique(selected.database$data_source)),font=list(size=15,family="Aril")) %>%
              add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
            table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
            
          } else {
            withProgress(message = 'Making plot', value = 0, {
              d<-data.frame(selected.database,Groups=selected.database$PFS.status)
              d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Freq"]),]
              d<-data.frame(d,dead=NA)
              d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
              d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
              d <- d[!is.na(d[,"Freq"]) & !is.na(d[,"dead"]),]
              d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
              d <- data.frame(data_source=as.character(d$data_source),
                              dead=as.numeric(d$dead),
                              OS=as.numeric(d$OS),
                              Age=as.character(d$Age),
                              Gender=as.character(d$Gender),
                              Freq=as.numeric(d[,"Freq"]))
              # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Freq")]))
              d<-data.frame(d,IND=paste("bottom ",input$slider_single_cell_cell_OS*100,"% (", 
                                        
                                        nrow(d[d[,"Freq"]<= quantile(d[,"Freq"],input$slider_single_cell_cell_OS),]),")",
                                        sep=""))
              d[,"IND"]<-as.character(d[,"IND"])
              d[d[,"Freq"]> quantile(d[,"Freq"],input$slider_single_cell_cell_OS),"IND"]<- 
                paste("top ",(1-input$slider_single_cell_cell_OS)*100,"% (", 
                      nrow(d[d[,"Freq"]> quantile(d[,"Freq"],input$slider_single_cell_cell_OS),]),")",
                      sep="")
              d<-d[order(d[,"IND"]),]
              d[,"dead"] <- as.numeric(d[,"dead"])
              fit.result <- survfit(Surv(OS,dead)~IND,data=d)
              
              try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
              if (class(try.test)=="try-error" | is.na(try.test)) {
                incProgress(100, detail = "Doing part %")
                p1 <- plot_ly () %>%
                  layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
              } else {
                plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                #plot.result <- ggsurv(fit.result)
                
                if (input_adjust_OS_single_cell_cell=="None") {
                  diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                  if (class(diff.result.try)=="try-error") {
                    p.value <- NA
                    table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    p.value <- signif(summary(diff.result)$logtest[3],2)
                    table1 <- datatable(signif(coef(summary(diff.result)),2),
                                        caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                               signif(summary(diff.result)$logtest[1],2),
                                                                               " on ",
                                                                               signif(summary(diff.result)$logtest[2],2),
                                                                               " df, p-value = ",
                                                                               signif(summary(diff.result)$logtest[3],2)),
                                                                        br(),
                                                                        paste0("n = ",
                                                                               signif(summary(diff.result)$n,2),
                                                                               ", number of events = ",
                                                                               signif(summary(diff.result)$nevent,2)),
                                                                        style="text-align:left;color:black;font-size:15px"),
                                        options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    
                  }
                } else {
                  diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_single_cell_cell),data=d) )
                  if (class(diff.result.adjust.try)=="try-error") {
                    diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                    p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                      " (No ",input_adjust_OS_single_cell_cell," was observed)")
                    
                    table1 <- datatable(signif(coef(summary(diff.result)),2),
                                        caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                               signif(summary(diff.result)$logtest[1],2),
                                                                               " on ",
                                                                               signif(summary(diff.result)$logtest[2],2),
                                                                               " df, p-value = ",
                                                                               signif(summary(diff.result)$logtest[3],2)),
                                                                        br(),
                                                                        paste0("n = ",
                                                                               signif(summary(diff.result)$n,2),
                                                                               ", number of events = ",
                                                                               signif(summary(diff.result)$nevent,2)),
                                                                        style="text-align:left;color:black;font-size:15px"),
                                        options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                  } else {
                    p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                    table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                    rownames(table1.pre)[2] <- input_adjust_OS_single_cell_cell
                    table1 <- datatable(table1.pre,
                                        caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                               signif(summary(diff.result.adjust)$logtest[1],2),
                                                                               " on ",
                                                                               signif(summary(diff.result.adjust)$logtest[2],2),
                                                                               " df, p-value = ",
                                                                               signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                        br(),
                                                                        paste0("n = ",
                                                                               signif(summary(diff.result.adjust)$n,2),
                                                                               ", number of events = ",
                                                                               signif(summary(diff.result.adjust)$nevent,2)),
                                                                        style="text-align:left;color:black;font-size:15px"),
                                        options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                  }
                }
                
                
                
                
                # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                # #diff.result <- survdiff(coxph(Surv(OS,dead)~IND+Age,data=d))
                # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                  ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                           #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                           paste0("p-value = ",p.value),x=300,y=0.05)) +
                  ggplot2::guides(color=FALSE,linetype=FALSE)
                p1<-ggplotly(plot.result.ann,
                             height = 450 ) %>%
                  layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                         xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                         yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                         legend=list(orientation="v",x=0.6,y=1))
                
                
                
              }
              # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
              # diff.result <- survdiff(Surv(OS,dead)~IND,data=d) 
              # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
              # plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
              #   ggplot2::geom_text(aes(label = sprintf("p-value = %0.2g",
              #                                          pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
              #   ggplot2::guides(color=FALSE,linetype=FALSE)
              # p1<-ggplotly(plot.result.ann,
              #              height = 450 ) %>%
              #   layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
              #          xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
              #          yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
              #          legend=list(orientation="v",x=0.6,y=1))
              # 
              # 
              # incProgress(100, detail = "Doing part %") 
              # p1
              
            })
          }
          output$single_cell_cell_survival_plots <- renderPlotly({
            p1
          })
          output$single_cell_cell_survival_tables <- renderDataTable(
            table1
          )
        })
      })
      
      #################################################
      #### single cell cell PFS
      #################################################
      #output$single_cell_cell_PFS_plots <- renderPlotly({
          observeEvent(input$do_single_cell_cell,{
            isolate({
            input_adjust_OS_single_cell_cell <- input$adjust_OS_single_cell_cell
        #    selecteddatabase_single_cell_cell() :
        #  PatientID      cells data_source treatment description PFS.status RECIST OS.status  OS PFS Age cluster
        #  1 Sade_Feldman_Post_P1 A10_P1_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       1
        #  2 Sade_Feldman_Post_P1 A10_P4_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       2
        #  3 Sade_Feldman_Post_P1 A10_P5_M15       Sade_Feldman anti-PD-1          On   response     NA    living 822  NA  49       1
        #      input$do_single_cell_cell
        #      isolate({
        #        selected.database <- selecteddatabase_single_cell_cell()
        #      selected.database <- selecteddatabase_single_cell_cell()
        if (nrow(selected.database)==0 | length(input$cluster_single_cell_cell)==0 | nrow(selected.database[!is.na(selected.database[,"PFS"]),])==0) {
          p1 <- plot_ly () %>%
            layout(title = as.character(unique(selected.database$data_source)),font=list(size=15,family="Aril")) %>%
            add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
          table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        } else {
          withProgress(message = 'Making plot', value = 0, {
            d<-data.frame(selected.database,Groups=selected.database$PFS.status)
            d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"Freq"]),]
            # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Freq")]))
            d <- data.frame(data_source=as.character(d$data_source),
                            PFS.status=as.character(d$PFS.status),
                            PFS=as.numeric(d$PFS),
                            Age=as.character(d$Age),
                            Gender=as.character(d$Gender),
                            Freq=as.numeric(d[,"Freq"]))
            d<-data.frame(d,IND=paste("bottom ",input$slider_single_cell_cell_OS*100,"% (", 
                                      
                                      nrow(d[d[,"Freq"]<= quantile(d[,"Freq"],input$slider_single_cell_cell_OS),]),")",
                                      sep=""))
            d[,"IND"]<-as.character(d[,"IND"])
            d[d[,"Freq"]> quantile(d[,"Freq"],input$slider_single_cell_cell_OS),"IND"]<- 
              paste("top ",(1-input$slider_single_cell_cell_OS)*100,"% (", 
                    nrow(d[d[,"Freq"]> quantile(d[,"Freq"],input$slider_single_cell_cell_OS),]),")",
                    sep="")
            
            d <- data.frame(d,group.IND=0)
            d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
            d<-d[order(d[,"IND"]),]
            d[,"dead"] <- as.numeric(d[,"dead"])
            fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)

            try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
            if (class(try.test)=="try-error" | is.na(try.test)) {
              incProgress(100, detail = "Doing part %")
              p1 <- plot_ly () %>%
                layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
              table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

            } else {

              plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))


              if (input_adjust_OS_single_cell_cell=="None") {
                diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                if (class(diff.result.try)=="try-error") {
                  p.value <- NA
                  table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                } else {
                  p.value <- signif(summary(diff.result)$logtest[3],2)
                  table1 <- datatable(signif(coef(summary(diff.result)),2),
                                      caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                             signif(summary(diff.result)$logtest[1],2),
                                                                             " on ",
                                                                             signif(summary(diff.result)$logtest[2],2),
                                                                             " df, p-value = ",
                                                                             signif(summary(diff.result)$logtest[3],2)),
                                                                      br(),
                                                                      paste0("n = ",
                                                                             signif(summary(diff.result)$n,2),
                                                                             ", number of events = ",
                                                                             signif(summary(diff.result)$nevent,2)),
                                                                      style="text-align:left;color:black;font-size:15px"),
                                      options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                }
              } else {
                diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_single_cell_cell),data=d) )
                if (class(diff.result.adjust.try)=="try-error") {
                  diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                  p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                    " (No ",input_adjust_OS_single_cell_cell," was observed)")

                  table1 <- datatable(signif(coef(summary(diff.result)),2),
                                      caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                             signif(summary(diff.result)$logtest[1],2),
                                                                             " on ",
                                                                             signif(summary(diff.result)$logtest[2],2),
                                                                             " df, p-value = ",
                                                                             signif(summary(diff.result)$logtest[3],2)),
                                                                      br(),
                                                                      paste0("n = ",
                                                                             signif(summary(diff.result)$n,2),
                                                                             ", number of events = ",
                                                                             signif(summary(diff.result)$nevent,2)),
                                                                      style="text-align:left;color:black;font-size:15px"),
                                      options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                } else {
                  p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                  table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                  rownames(table1.pre)[2] <- input_adjust_OS_single_cell_cell
                  table1 <- datatable(table1.pre,
                                      caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                             signif(summary(diff.result.adjust)$logtest[1],2),
                                                                             " on ",
                                                                             signif(summary(diff.result.adjust)$logtest[2],2),
                                                                             " df, p-value = ",
                                                                             signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                      br(),
                                                                      paste0("n = ",
                                                                             signif(summary(diff.result.adjust)$n,2),
                                                                             ", number of events = ",
                                                                             signif(summary(diff.result.adjust)$nevent,2)),
                                                                      style="text-align:left;color:black;font-size:15px"),
                                      options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                }
              }



              # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
              # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
              plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                         #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                         paste0("p-value = ",p.value),x=300,y=0.05)) +
                ggplot2::guides(color=FALSE,linetype=FALSE)
              p1<-ggplotly(plot.result.ann,
                           height = 450 ) %>%
                layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                       xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                       yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                       legend=list(orientation="v",x=0.6,y=1))
              incProgress(100, detail = "Doing part %")

            }
            # plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
            # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
            # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
            # plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
            #   ggplot2::geom_text(aes(label = sprintf("p-value = %0.2g",
            #                                          pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
            #   ggplot2::guides(color=FALSE,linetype=FALSE)
            # p1<-ggplotly(plot.result.ann,
            #              height = 450 ) %>%
            #   layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
            #          xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
            #          yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
            #          legend=list(orientation="v",x=0.6,y=1))
            # incProgress(100, detail = "Doing part %")
            # p1

          })
        }

            output$single_cell_cell_PFS_plots <- renderPlotly({
              p1
            })
            output$single_cell_cell_PFS_tables <- renderDataTable(
              table1
            )
          })
      })
      
      ################
      #### single cell cell evaluation
      #################
      
      output$single_cell_cell_evaluation_table <- DT::renderDataTable({
        if (nrow(selected.database)==0) {
          datatable(data.frame("NA"="no data available"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
        } else {
          
          d <- selected.database
          d$category <- as.character(d$category)
          d$PFS.status <- as.character(d$PFS.status)
          unit.table <- as.data.frame.matrix(table(d[,c("category","PFS.status")]))
          unit.table <- data.frame(category=rownames(unit.table),unit.table)
          final.table <- data.frame(data_source=unique(as.character(d$data_source)),unit.table,hiddenColumn=c(rep(0,(nrow(unit.table)-1)),1))
          
          DT::datatable(final.table,options = list(columnDefs=list(list(visible=FALSE,targets=4)),searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE,selection = "single") %>%
            formatStyle(0:4,valueColumns="hiddenColumn",`border-bottom`=styleEqual(1,'1px solid black')) #%>%
          
        }
        
      })
      
      
      
      
    })   
    
     #####################################################
     #### pan-cancer assessment mutation
     #####################################################
     
     observe({
       if (input$defined_gene_pan_cancer_assessment_mutation=="user-defined list") {
         updateSelectizeInput(session,"gene_pan_cancer_assessment_mutation",
                              choices=a.colnames,selected=c("BRAF","KRAS","HRAS","NRAS","NF1"),
                              server=TRUE)
       } else {
         selected_mutations <- switch (input$defined_gene_pan_cancer_assessment_mutation,
                                       "DNA_damage_response" = DNA_damage_response,
                                       "PDL1_expression_and_PD1_checkpoint_pathway_in_cancer"=PDL1_expression_and_PD1_checkpoint_pathway_in_cancer,
                                       "T_cell_receptor_signaling_pathway"=T_cell_receptor_signaling_pathway,
                                       "B_cell_receptor_signaling_pathway"=B_cell_receptor_signaling_pathway,
                                       "Cell_cycle_control"=Cell_cycle_control,
                                       "Survival_cell_death_regulation_signaling"=Survival_cell_death_regulation_signaling)
         updateSelectizeInput(session,"gene_pan_cancer_assessment_mutation",
                              choices=a.colnames,selected=selected_mutations,
                              server=TRUE)
       }
     })
     
     observe({
       if (input$data_source_pan_cancer_assessment_mutation=="" || is.null(input$data_source_pan_cancer_assessment_mutation) ||
           input$treatment_pan_cancer_assessment_mutation=="" || is.null(input$treatment_pan_cancer_assessment_mutation) ||
           input$description_pan_cancer_assessment_mutation=="" || is.null(input$description_pan_cancer_assessment_mutation)){
         disable("do_pan_cancer_assessment_mutation")
       } else {
         enable("do_pan_cancer_assessment_mutation")
       }
     })
     
     selecteddatabase_pan_cancer_assessment_mutation <- eventReactive(input$do_pan_cancer_assessment_mutation,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_mutation])
       description1 <- input$description_pan_cancer_assessment_mutation
       genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
       a<-data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       as.data.frame(m)
     })
     
     selectedmutation_pan_cancer_assessment_mutation <- function() {
       all.mutations <- all.mutations.process()
       input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
       selected.database<-selecteddatabase_pan_cancer_assessment_mutation()
       if (nrow(selected.database)==0) {
         selected.database
       } else {
         gene1 <- input_gene_pan_cancer_assessment_mutation
         selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"] ,]
         selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% gene1,]
         if (nrow(selected.mutations.1)>0 ) {
           #### important: resharp mutation file to database.test format
           b<-aggregate(variant_class~(PatientID+gene),data=selected.mutations.1,FUN=paste0)
           c <- b %>% spread(gene,variant_class,fill="no mutation")
           c[c=="NULL"]<-"no mutation"
           
           if ((ncol(c)-1) < length(gene1)) {
             c.add <- matrix("no mutation",nrow(c),length(gene1[gene1 %ni% colnames(c)[2:ncol(c)]]))
             colnames(c.add) <- gene1[gene1 %ni% colnames(c)[2:ncol(c)]]
             c <- cbind(c, c.add)
           }
           
           c.rest <- data.frame(PatientID=unique(selected.mutations[selected.mutations[,"PatientID"] %ni% selected.mutations.1[,"PatientID"],"PatientID"]))
           c <- rbind.fill(c,c.rest)
           c[c=="NULL"] <- "no mutation"
           c[is.na(c)] <- "no mutation"
           
           #c <- data.frame(c,category=apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") }))
           c <- data.frame(c,category="no mutation")
           c[,"category"] <- as.character(c[,"category"])
           c[!apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") == "no mutation" }),"category"]<- "mutated"
           
           selected.database <- merge(selected.database,c,by="PatientID",all.x=T)
           selected.database[is.na(selected.database[,"category"]),"category"] <- "unknown"
           as.data.frame(selected.database)
         } else {
           selected.database <- data.frame(selected.database,category="unknown")
           selected.database
         }
       }
     }
     
     observeEvent(input$do_pan_cancer_assessment_mutation,{
       ##########
       #### Summary
       ##########
       output$pan_cancer_assessment_mutation_summary_disease_number <- renderText({
         input$do_pan_cancer_assessment_mutation ##keep these two rows
         isolate({ ##keep these two rows
           length(unique(selecteddatabase_pan_cancer_assessment_mutation()$Disease))})})
       output$pan_cancer_assessment_mutation_summary_sample_number <- renderText({
         input$do_pan_cancer_assessment_mutation ##keep these two rows
         isolate({ ##keep these two rows
           nrow(selecteddatabase_pan_cancer_assessment_mutation())})})
       output$pan_cancer_assessment_mutation_summary_plots <- renderPlotly({
         summary_plot(inputdata=selectedmutation_pan_cancer_assessment_mutation(),feature="mutation")
       })
       
       
       output$pan_cancer_assessment_mutation_summary_clinical_plots <- renderPlotly({
         withProgress(message = 'Making plot', value = 0, {
         selected.database <- selectedmutation_pan_cancer_assessment_mutation()
         selected.database$treatment <- as.factor(selected.database$treatment)
         selected.database$description <- as.factor(selected.database$description)
         selected.database$RECIST <- as.factor(selected.database$RECIST)
         selected.database$Gender <- as.factor(selected.database$Gender)
         if (nrow(selected.database)==0 ) {
           incProgress(100, detail = "Doing part %")
           plot_ly () %>%
             layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database
           d<-splitted_list
           if (nrow(d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_clinical]]) & d[,"category"]!="unknown",])==0 ) {
             incProgress(100, detail = "Doing part %")
             plot_ly () %>%
               layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           } else {
             d <- d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_clinical]]) & d[,"category"]!="unknown",]
             if(input$pan_cancer_assessment_mutation_clinical=="Age") {
               d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
               d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
               d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
             }
             plot.data <- as.data.frame(table(d[,c(clinical_server[input$pan_cancer_assessment_mutation_clinical],"category")]))
             colnames(plot.data) <- c("clinical","category","numbers")
             p<-plot_ly(plot.data,type="bar",x=~clinical,y=~numbers,
                        color=~category,
                        colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                        legendgroup=~category) %>%
               layout(legend=list(font=list(size=15)),
                      title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                      height = 480,
                      xaxis=list(title="Clinical Attribute",
                                 titlefont=list(size=20,color="black",family= "Aril black"),
                                 showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                      yaxis=list(title="Sample number",
                                 titlefont=list(size=20,color="black",family= "Aril black"),
                                 showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                      ))
             incProgress(100, detail = "Doing part %")
             p
           }
         }
         })
       })
       
       
       
       
       
       
       # observeEvent(input$do_pan_cancer_assessment_mutation,{
       #   input$do_pan_cancer_assessment_mutation
       #   isolate({
       #     selected.database <- selectedmutation_pan_cancer_assessment_mutation()
       #     selected.database$treatment <- as.factor(selected.database$treatment)
       #     selected.database$description <- as.factor(selected.database$description)
       #     selected.database$RECIST <- as.factor(selected.database$RECIST)
       #     selected.database$Gender <- as.factor(selected.database$Gender)
       #     if (nrow(selected.database)==0 ) {
       #       output$pan_cancer_assessment_mutation_summary_clincial_plots <- renderPlotly({
       #         plot_ly () %>% 
       #           layout(title = "pan-cancer cohort",font=list(size=15,family="Aril")) %>%
       #           add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
       #       })
       #     } else {
       #       selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
       #       
       #       splitted_list <- selected.database# %>% split(.$data_source) #selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
       #       #splitted_list <- splitted_list[input$data_source_mutation]
       #       #for (k in 1:length(input$data_source_mutation)) {
       #         local({
       #           # my_i <- k
       #           # mutation_summary_clinical_plots_name <- paste0("mutation_summary_clinical_plots", my_i)
       #           d<-splitted_list#[[my_i]]
       #           
       #           output$pan_cancer_assessment_mutation_summary_clincial_plots <- renderPlotly({
       #             withProgress(message = 'Making plot', value = 0, {
       #               # if (is.na(names(splitted_list)[my_i])) {
       #               #   incProgress(100, detail = "Doing part %")
       #               #   plot_ly () %>% 
       #               #     layout(title = input$data_source_mutation[[my_i]],font=list(size=15,family="Aril")) %>%
       #               #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
       #               # } else {
       #                 if (nrow(d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_clinical]]) & d[,"category"]!="no data available",])==0 ) {
       #                   incProgress(100, detail = "Doing part %")
       #                   plot_ly () %>% 
       #                     layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
       #                     add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
       #                 } else {
       #                   d <- d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_clinical]]) & d[,"category"]!="no data available",]
       #                   if(input$pan_cancer_assessment_mutation_clinical=="Age") {
       #                     d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
       #                     d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
       #                     d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
       #                   }
       #                   plot.data <- as.data.frame(table(d[,c(clinical_server[input$pan_cancer_assessment_mutation_clinical],"category")]))
       #                   colnames(plot.data) <- c("clinical","category","numbers")
       #                   p<-plot_ly(plot.data,type="bar",x=~clinical,y=~numbers,
       #                              color=~category,
       #                              colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
       #                              legendgroup=~category) %>% 
       #                     layout(legend=list(font=list(size=15)),
       #                            title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
       #                            height = 480,
       #                            xaxis=list(title="Clinical Attribute",
       #                                       titlefont=list(size=20,color="black",family= "Aril black"),
       #                                       showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
       #                            yaxis=list(title="Sample number",
       #                                       titlefont=list(size=20,color="black",family= "Aril black"),
       #                                       showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
       #                            ))
       #                   incProgress(100, detail = "Doing part %")
       #                   p
       #                   
       #                 } 
       #               #}
       #             })
       #           })
       #         })
       #       #}
       #     }
       #   }) 
       # })
        
       output$pan_cancer_assessment_mutation_summary_download <- downloadHandler(
         filename = function() {
           if (input$pan_cancer_assessment_mutation_summary_type == "Excel (CSV)") {
             "newfile.csv"
           } else if (input$pan_cancer_assessment_mutation_summary_type == "Text (TSV)") {
             "newfile.txt"
           } else if (input$pan_cancer_assessment_mutation_summary_type == "JSON") {
             "newfile.json"
           }
         },
         content = function(file) {
           if (input$pan_cancer_assessment_mutation_summary_type == "Excel (CSV)") {
             fwrite (selectedmutation_pan_cancer_assessment_mutation(),file)
           } else if (input$pan_cancer_assessment_mutation_summary_type == "Text (TSV)") {
             fwrite (selectedmutation_pan_cancer_assessment_mutation(),file)
             #write.table (selectedmutation_mutation(),file,quote=F, row.names = FALSE, sep="\t")
           } else if (input$pan_cancer_assessment_mutation_summary_type == "JSON") {
             write(toJSON(selectedmutation_pan_cancer_assessment_mutation()),file)
           }
         }
       )
       output$pan_cancer_assessment_mutation_summary_table <- DT::renderDataTable({
         input$do_pan_cancer_assessment_mutation
         isolate({
           selected.database <- selectedmutation_pan_cancer_assessment_mutation()
           selected.database$category <- as.character(selected.database$category)
           selected.database
         })
       })   
    ########
    #### mutation plot
    ########
       observeEvent(input$do_pan_cancer_assessment_mutation,{
         input$do_pan_cancer_assessment_mutation
         isolate({
           input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
           selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
           all.mutations <- all.mutations.process()
           selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           selected.database <- selected.database[selected.database[,"PatientID"] %in% selected.mutations[,"PatientID"],]
           selected.mutations.function <- selectedmutation_pan_cancer_assessment_mutation()
           #selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
           if (nrow(selected.database)==0 | nrow(selected.mutations)==0) {
             output[["pan_cancer_assessment_mutation_plots_plots"]] <- renderPlotly({
               plot_ly () %>% 
                 layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                 add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
             })
             
           } else {
             selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
             
             if (nrow(selected.mutations.1)==0) {
               output[["pan_cancer_assessment_mutation_plots_plots"]] <- renderPlotly({
                 plot_ly () %>% 
                   layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                   add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
               })
               
             } else {
               
               
             selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
             splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
             #splitted_list <- splitted_list#[input$data_source_mutation]
             #      splitted_list_mutation <- selected.mutations.1 %>% as.matrix %>% as.data.frame %>% split(.$data_source)
             splitted_list_mutation <- selected.mutations# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
             #splitted_list_mutation <- splitted_list_mutation#[input$data_source_mutation]
             splitted_list_mutation.function <- selected.mutations.function# %>% split(.$data_source)
             #splitted_list_mutation.function <- splitted_list_mutation.function#[input$data_source_mutation]
             
             #for (k in 1:length(input$data_source_mutation)) {
               local({
                 # my_i <- k
                 # mutation_plots_plots_name <- paste0("mutation_plots_plots", my_i)
                 d<-splitted_list#[[my_i]]
                 d.mutations <- splitted_list_mutation#[[my_i]]
                 d.mutations.function <- splitted_list_mutation.function
                 output[["pan_cancer_assessment_mutation_plots_plots"]] <- renderPlotly({
                   withProgress(message = 'Making plot', value = 0, {
                     #              isolate({
                     # if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_mutation)[my_i])) {
                     #   incProgress(100, detail = "Doing part %")
                     #   plot_ly () %>% 
                     #     layout(title = input$data_source_mutation[[my_i]],font=list(size=15,family="Aril")) %>%
                     #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     # } else {
                       if (nrow(d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & 
                                                     d.mutations.function[,"PFS.status"]!="Unknown" &
                                                     d.mutations.function[,"category"]!="unknown" &
                                                     !is.na(d.mutations.function[,"category"]),])==0) {
                         incProgress(100, detail = "Doing part %")
                         plot_ly () %>% 
                           layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       } else {
                         d.mutations.function <- d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & 
                                                                        d.mutations.function[,"PFS.status"]!="Unknown" &
                                                                        d.mutations.function[,"category"]!="unknown" &
                                                                        !is.na(d.mutations.function[,"category"]),]
                         d.table <- unlist(table(d.mutations.function[,c("PFS.status","category")]))
                         d.table <- d.table[rownames(d.table) %in% c("nonresponse","response"),colnames(d.table) %in% c("mutated","no mutation")]
                         try.test <- try(p.v <- fisher.test(d.table)$p.value)
                         if (class(try.test)=="try-error" | is.na(try.test)) {
                           p.v <- " = NA"
                         } else {
                           p.v<-paste0(" = ",signif(p.v,2))
                         }
                         #                    withProgress(message = 'Making plot', value = 0, {
                         # d <- d[!is.na(d[,"PFS.status"]),]
                         # d<-data.frame(d,IND="no mutation")
                         # d[,"IND"]<-as.character(d[,"IND"])
                         # for (j in 1:length(input_gene_pan_cancer_assessment_mutation)) {
                         #   d[d[,"PatientID"] %in% d.mutations[d.mutations[,"gene"]==input_gene_pan_cancer_assessment_mutation[j],"PatientID"],"IND"]<- paste0("mutation in ",input_gene_pan_cancer_assessment_mutation[j])
                         # }
                         # try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")]))[1:2,])$p.value)
                         # if (class(try.test)=="try-error") {
                         #   p.v <- " = NA"
                         # } else {
                         # #   p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")])))$p.value
                         # #   if (p.v < 0.001) {
                         # #     p.v <- " < 0.001"
                         # #   } else {
                         #      p.v<-paste0(" = ",round(p.v,3))
                         # #   } 
                         #  }
                         ind <- sort(unique(d[,"Groups"]))
                         d.mutations.res <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[2] & !is.na(d[,"Groups"])),"PatientID"] ,]
                         d.mutations.nonres <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[1] & !is.na(d[,"Groups"])),"PatientID"] ,]
                         
                         d.mutations.res<-d.mutations.res[,-1]
                         #d.mutations.res<-d.mutations.res[,c("PatientID","gene","variant_class","Protein_change","Chromosome","data_source")]
                         d.mutations.res<-d.mutations.res[,c(6,2:5,7)]
                         colnames(d.mutations.res)[1] <- "sample"
                         if (nrow(d.mutations.res[d.mutations.res[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,])==0) {
                           p1 <- ggplot(d.mutations.res, aes(x=sample,y="NA")) + 
                             xlab(paste0("response samples (n=",length(unique(d.mutations.res$sample)),")")) + 
                             theme(axis.text.x=element_text(angle=50, hjust=1,size=4), 
                                   axis.text.y=element_text(size=8,colour='black'),
                                   axis.title.x=element_text(size=11),
                                   axis.title.y=element_blank())
                           p.bar.res <- data.frame(PFS.status="response",genes=input_gene_pan_cancer_assessment_mutation,Number=0,n=length(unique(d.mutations.res$sample)))
                           p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                           # p3 <- plot_ly () %>% 
                           #   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple")) 
                         } else {
                           p1<-#ggplotly(
                             waterfall(d.mutations.res, fileType = "Custom",
                                       variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))), ### here all.mutation is important, if not, different data source will have different color enve in the same mutation type
                                       plotGenes=unlist(str_split(input_gene_pan_cancer_assessment_mutation,",")),
                                       mainXlabel = TRUE,my_x_label=ind[2],
                                       plotMutBurden = FALSE,returnp="p1",
                                       mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                     '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                     '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                     '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                     '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                     '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                     '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                     '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                           Number.genes.res <- NULL
                           for (j in 1:length(input_gene_pan_cancer_assessment_mutation)) {
                             Number.genes.res <- c(Number.genes.res,length(unique(d.mutations.res[d.mutations.res[,"gene"]==input_gene_pan_cancer_assessment_mutation[j],"sample"])))
                           }
                           p.bar.res <- data.frame(PFS.status="response",genes=input_gene_pan_cancer_assessment_mutation,Number=Number.genes.res,n=length(unique(d.mutations.res$sample)))
                           p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                           # #height =  480 )
                           # p2<-#ggplotly(
                           #   waterfall(d.mutations.res, fileType = "Custom",
                           #             variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                           #             plotGenes=unlist(str_split(input_gene_pan_cancer_assessment_mutation,",")),
                           #             mainXlabel = TRUE,my_x_label=ind[2],
                           #             plotMutBurden = FALSE,returnp="p2",xticks=FALSE,
                           #             mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                           #                           '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                           #                           '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                           #                           '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                           #                           '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                           #                           '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                           #                           '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                           #                           '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                           # #height =  480 )
                           # #                 tooltip = c("x","y","text"),legendgroup=~variant_class)
                           # p3 <- subplot(style(p2,showlegend=T),style(p1,showlegend=T),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                         }
                         d.mutations.nonres<-d.mutations.nonres[,-1]
                         d.mutations.nonres<-d.mutations.nonres[,c(6,2:5,7)]
                         colnames(d.mutations.nonres)[1] <- "sample"
                         if (nrow(d.mutations.nonres[d.mutations.nonres[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,])==0) {
                           q1 <- ggplot(d.mutations.nonres, aes(x=sample,y="NA")) + 
                             xlab(paste0("nonresponse samples (n=",length(unique(d.mutations.nonres$sample)),")")) + 
                             theme(axis.text.x=element_text(angle=50, hjust=1,size=4), 
                                   axis.text.y=element_text(size=8,colour='black'),
                                   axis.title.x=element_text(size=11),
                                   axis.title.y=element_blank())
                           p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_pan_cancer_assessment_mutation,Number=0,n=length(unique(d.mutations.nonres$sample)))
                           p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                           # q3 <- plot_ly () %>% 
                           #   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple")) 
                         } else {
                           q1<-#ggplotly(
                             waterfall(d.mutations.nonres, fileType = "Custom",
                                       variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                                       plotGenes=unlist(str_split(input_gene_pan_cancer_assessment_mutation,",")),
                                       mainXlabel = TRUE,my_x_label=ind[1],
                                       plotMutBurden = FALSE,returnp="p1",
                                       mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                     '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                     '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                     '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                     '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                     '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                     '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                     '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                           Number.genes.nonres <- NULL
                           for (j in 1:length(input_gene_pan_cancer_assessment_mutation)) {
                             Number.genes.nonres <- c(Number.genes.nonres,length(unique(d.mutations.nonres[d.mutations.nonres[,"gene"]==input_gene_pan_cancer_assessment_mutation[j],"sample"])))
                           }
                           p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_pan_cancer_assessment_mutation,Number=Number.genes.nonres,n=length(unique(d.mutations.nonres$sample)))
                           p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                           #height =  480 )
                           # q2<-#ggplotly(
                           #   waterfall(d.mutations.nonres, fileType = "Custom",
                           #             variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                           #             plotGenes=unlist(str_split(input_gene_pan_cancer_assessment_mutation,",")),
                           #             mainXlabel = TRUE,my_x_label=ind[1],
                           #             plotMutBurden = FALSE,returnp="p2",xticks=TRUE,
                           #             mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                           #                           '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                           #                           '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                           #                           '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                           #                           '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                           #                           '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                           #                           '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                           #                           '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                           # #height =  480 )
                           # #                     tooltip = c("x","y","text"),legendgroup=~variant_class)
                           # q3 <- subplot(style(q2,showlegend=F),style(q1,showlegend=F),margin=0.03,widths = c(1/5,4/5),titleX=TRUE,titleY=TRUE)
                         }
                         p.bar.res <- data.frame(p.bar.res,perc=signif(p.bar.res$Number/p.bar.res$n*100,3))
                         p.bar.nonres <- data.frame(p.bar.nonres,perc=signif(p.bar.nonres$Number/p.bar.nonres$n*100,3))
                         
                         if (sum(p.bar.res$perc)==0 & sum(p.bar.nonres$perc)!=0 ) {
                           p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                           p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                         } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)==0) {
                           p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                           p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                         } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)!=0) {
                           p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                           p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                         } else {
                           p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                           p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                         }
                         #p.bar <- rbind(p.bar.res,p.bar.nonres)
                         p.bar <- rbind(p.bar.nonres,p.bar.res)
                         p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
                         bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
                           theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
                           geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                           #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                           geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))
                         
                         incProgress(100, detail = "Doing part %")
                         subplot(style(bar.plot,showlegend=F),
                                 subplot(p1,q1,nrows=2,margin=0.1,titleX=TRUE,titleY=TRUE) %>% 
                                   layout(#title = as.character(unique(d$data_source)), 
                                     titlefont=list(size=20,color="black",family= "Aril"),
                                     legend=list(font=list(size=10))),
                                 #xaxis=list(title=lable_x,side=side_x,titlefont=list(size=15))),
                                 widths=c(1/3,2/3),margin=0.05,titleX=TRUE,titleY=TRUE) %>% 
                           layout(title="the aggregated dataset")
                         #                  })
                       } 
                       
                     #}
                   })
                 })
                 #          })
               })
             #}
           }
               }
         })
         
       })
       
       
       observeEvent(input$do_pan_cancer_assessment_mutation,{
         input$do_pan_cancer_assessment_mutation
         isolate({
           input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
           selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
           all.mutations <- all.mutations.process()
           selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           selected.mutations.function <- selectedmutation_pan_cancer_assessment_mutation()
           #selected.pan_cancer_assessment_mutations.1 <- selected.pan_cancer_assessment_mutations[selected.pan_cancer_assessment_mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
           if (nrow(selected.database)==0 | nrow(selected.mutations)==0) {
             
             output[["pan_cancer_assessment_individual_mutation_plots_plots0"]] <- renderPlotly({
               plot_ly () %>% 
                 layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
                 add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
             })
             
           } else {
             selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
             
             if (nrow(selected.mutations.1)==0) {
               
               output[["pan_cancer_assessment_individual_mutation_plots_plots0"]] <- renderPlotly({
                 plot_ly () %>% 
                   layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
                   add_annotations(x=2,y=2,text="no mutations under this selection",showarrow=F,font=list(size=16,color="purple")) 
               })
               
             } else {
               selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
               splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
               splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
               #      splitted_list_pan_cancer_assessment_mutation <- selected.pan_cancer_assessment_mutations.1 %>% as.matrix %>% as.data.frame %>% split(.$data_source)
               splitted_list_mutation <- selected.mutations %>% as.matrix %>% as.data.frame %>% split(.$data_source)
               splitted_list_mutation <- splitted_list_mutation[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
               splitted_list_mutation.function <- selected.mutations.function %>% split(.$data_source)
               splitted_list_mutation.function <- splitted_list_mutation.function[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
               
               
               for (k in 1:length(input$data_source_pan_cancer_assessment_mutation)) {
                 local({
                   my_i <- k
                   pan_cancer_assessment_individual_mutation_plots_plots_name <- paste0("pan_cancer_assessment_individual_mutation_plots_plots", my_i)
                   d<-splitted_list[[my_i]]
                   d.mutations <- splitted_list_mutation[[my_i]]
                   d.mutations.function <- splitted_list_mutation.function[[my_i]]
                   output[[pan_cancer_assessment_individual_mutation_plots_plots_name]] <- renderPlotly({
                     withProgress(message = 'Making plot', value = 0, {
                       isolate({
                         if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_mutation)[my_i])) {
                           incProgress(100, detail = "Doing part %")
                           plot_ly () %>% 
                             layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                             add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                         } else {
                           if (nrow(d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & 
                                                         d.mutations.function[,"PFS.status"]!="Unknown" &
                                                         d.mutations.function[,"category"]!="unknown" &
                                                         !is.na(d.mutations.function[,"category"]),])==0) {
                             incProgress(100, detail = "Doing part %")
                             plot_ly () %>% 
                               layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                               add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                           } else {
                             #                    withProgress(message = 'Making plot', value = 0, {
                             d.mutations.function <- d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & 
                                                                            d.mutations.function[,"PFS.status"]!="Unknown" &
                                                                            d.mutations.function[,"category"]!="unknown" &
                                                                            !is.na(d.mutations.function[,"category"]),]
                             d.table <- unlist(table(d.mutations.function[,c("PFS.status","category")]))
                             d.table <- d.table[rownames(d.table) %in% c("nonresponse","response"),colnames(d.table) %in% c("mutated","no mutation")]
                             try.test <- try(p.v <- fisher.test(d.table)$p.value)
                             #try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")]))[1:2,])$p.value)
                             if (class(try.test)=="try-error" | is.na(try.test)) {
                               p.v <- " = NA"
                             } else {
                               p.v<-paste0(" = ",signif(p.v,2))
                               
                               # } 
                             }
                             
                             ind <- sort(unique(d[,"Groups"]))
                             d.mutations.res <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[2] & !is.na(d[,"Groups"])),"PatientID"] ,]
                             d.mutations.nonres <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[1] & !is.na(d[,"Groups"])),"PatientID"] ,]
                             
                             d.mutations.res<-d.mutations.res[,-1]
                             #d.pan_cancer_assessment_mutations.res<-d.pan_cancer_assessment_mutations.res[,c("PatientID","gene","variant_class","Protein_change","Chromosome","data_source")]
                             d.mutations.res<-d.mutations.res[,c(6,2:5,7)]
                             colnames(d.mutations.res)[1] <- "sample"
                             if (nrow(d.mutations.res[d.mutations.res[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,])==0) {
                               p1 <- ggplot(d.mutations.res, aes(x=sample,y="NA")) + 
                                 xlab(paste0("response samples (n=",length(unique(d.mutations.res$sample)),")")) + 
                                 theme(axis.text.x=element_text(angle=50, hjust=1,size=4), 
                                       axis.text.y=element_text(size=8,colour='black'),
                                       axis.title.x=element_text(size=11),
                                       axis.title.y=element_blank())
                               p.bar.res <- data.frame(PFS.status="response",genes=input_gene_pan_cancer_assessment_mutation,Number=0,n=length(unique(d.mutations.res$sample)))
                               p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                             } else {
                               p1<-#ggplotly(
                                 waterfall(d.mutations.res, fileType = "Custom",
                                           variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))), ### here all.pan_cancer_assessment_mutation is important, if not, different data source will have different color enve in the same pan_cancer_assessment_mutation type
                                           plotGenes=unlist(str_split(input_gene_pan_cancer_assessment_mutation,",")),
                                           mainXlabel = TRUE,my_x_label=ind[2],
                                           plotMutBurden = FALSE,returnp="p1",
                                           mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                         '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                         '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                         '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                         '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                         '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                         '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                         '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                               Number.genes.res <- NULL
                               for (j in 1:length(input_gene_pan_cancer_assessment_mutation)) {
                                 Number.genes.res <- c(Number.genes.res,length(unique(d.mutations.res[d.mutations.res[,"gene"]==input_gene_pan_cancer_assessment_mutation[j],"sample"])))
                               }
                               p.bar.res <- data.frame(PFS.status="response",genes=input_gene_pan_cancer_assessment_mutation,Number=Number.genes.res,n=length(unique(d.mutations.res$sample)))
                               p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
                             }
                             d.mutations.nonres<-d.mutations.nonres[,-1]
                             d.mutations.nonres<-d.mutations.nonres[,c(6,2:5,7)]
                             colnames(d.mutations.nonres)[1] <- "sample"
                             if (nrow(d.mutations.nonres[d.mutations.nonres[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,])==0) {
                               q1 <- ggplot(d.mutations.nonres, aes(x=sample,y="NA")) + 
                                 xlab(paste0("nonresponse samples (n=",length(unique(d.mutations.nonres$sample)),")")) + 
                                 theme(axis.text.x=element_text(angle=50, hjust=1,size=4), 
                                       axis.text.y=element_text(size=8,colour='black'),
                                       axis.title.x=element_text(size=11),
                                       axis.title.y=element_blank())
                               p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_pan_cancer_assessment_mutation,Number=0,n=length(unique(d.mutations.nonres$sample)))
                               p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                             } else {
                               q1<-#ggplotly(
                                 waterfall(d.mutations.nonres, fileType = "Custom",
                                           variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                                           plotGenes=unlist(str_split(input_gene_pan_cancer_assessment_mutation,",")),
                                           mainXlabel = TRUE,my_x_label=ind[1],
                                           plotMutBurden = FALSE,returnp="p1",
                                           mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                                         '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                                         '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                                         '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                                         '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                                         '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                                         '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                                         '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
                               Number.genes.nonres <- NULL
                               for (j in 1:length(input_gene_pan_cancer_assessment_mutation)) {
                                 Number.genes.nonres <- c(Number.genes.nonres,length(unique(d.mutations.nonres[d.mutations.nonres[,"gene"]==input_gene_pan_cancer_assessment_mutation[j],"sample"])))
                               }
                               p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_pan_cancer_assessment_mutation,Number=Number.genes.nonres,n=length(unique(d.mutations.nonres$sample)))
                               p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
                             }
                             
                             p.bar.res <- data.frame(p.bar.res,perc=signif(p.bar.res$Number/p.bar.res$n*100,3))
                             p.bar.nonres <- data.frame(p.bar.nonres,perc=signif(p.bar.nonres$Number/p.bar.nonres$n*100,3))
                             
                             if (sum(p.bar.res$perc)==0 & sum(p.bar.nonres$perc)!=0 ) {
                               p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                               p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                             } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)==0) {
                               p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                               p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                             } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)!=0) {
                               p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
                               p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
                             } else {
                               p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
                               p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
                             }
                             #p.bar <- rbind(p.bar.res,p.bar.nonres)
                             p.bar <- rbind(p.bar.nonres,p.bar.res)
                             p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
                             bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
                               theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
                               geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                               #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
                               geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))
                             
                             incProgress(100, detail = "Doing part %")
                             subplot(style(bar.plot,showlegend=F),
                                     subplot(p1,q1,nrows=2,margin=0.1,titleX=TRUE,titleY=TRUE) %>% 
                                       layout(#title = as.character(unique(d$data_source)), 
                                         titlefont=list(size=20,color="black",family= "Aril"),
                                         legend=list(font=list(size=10))),
                                     #xaxis=list(title=lable_x,side=side_x,titlefont=list(size=15))),
                                     widths=c(1/3,2/3),margin=0.05,titleX=TRUE,titleY=TRUE) %>% 
                               layout(title=as.character(unique(d$data_source)))
                             #                  })
                           } 
                           
                         }
                       })
                     })
                   })
                 })
               }
               
             }
           }
           
           
         })
         
       })
       ##############
       #### pan cancer assessment mutation survival
       ##############
       
       observeEvent(input$do_pan_cancer_assessment_mutation,{
         input$do_pan_cancer_assessment_mutation
         isolate({
           input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
           selected.database <- selectedmutation_pan_cancer_assessment_mutation()#selectedpan_cancer_assessment_mutation_pan_cancer_assessment_mutation()
           input_adjust_OS_pan_cancer_assessment_mutation <- input$adjust_OS_pan_cancer_assessment_mutation
           # selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
           # selected.pan_cancer_assessment_mutations <- all.pan_cancer_assessment_mutations[all.pan_cancer_assessment_mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           # selected.pan_cancer_assessment_mutations.1 <- selected.pan_cancer_assessment_mutations[selected.pan_cancer_assessment_mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
           #      if (nrow(selected.database)==0 | nrow(selected.pan_cancer_assessment_mutations.1)==0) {
           
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
           # splitted_list_pan_cancer_assessment_mutation <- selected.pan_cancer_assessment_mutations.1 %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           # splitted_list_pan_cancer_assessment_mutation <- splitted_list_pan_cancer_assessment_mutation[input$data_source_pan_cancer_assessment_mutation]
           
           #for (k in 1:length(input$data_source_pan_cancer_assessment_mutation)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_mutation_survival_plots_name <- paste0("pan_cancer_assessment_mutation_survival_plots", my_i)
             # pan_cancer_assessment_mutation_survival_summary_tables_name <- paste0("pan_cancer_assessment_mutation_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             # d.pan_cancer_assessment_mutations <- splitted_list_pan_cancer_assessment_mutation[[my_i]]
             #output[[pan_cancer_assessment_mutation_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 #                if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_pan_cancer_assessment_mutation)[my_i])) {
                 # if (is.na(names(splitted_list)[my_i]) ) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   #                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) ,]
                   d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                   if (length(unique(d$category))==1) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="There is only 1 group under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     d<-data.frame(d,dead=NA)
                     d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                     d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                     d <- d[!is.na(d[,"dead"]),]
                     d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                     d <- data.frame(data_source=as.character(d$data_source),
                                     dead=as.numeric(d$dead),
                                     OS=as.numeric(d$OS),
                                     Age=as.character(d$Age),
                                     Gender=as.character(d$Gender),
                                     category=as.character(d[,"category"]))
                     #d<-data.frame(d[,c(1:6)],unfactor(d[,c("dead","OS","PFS","Age","Gender")]),category=as.character(d[,"category"]))
                     #d$OS<-as.numeric(d$OS)
                     d<-data.frame(d,IND=paste0("no mutation (",
                                                nrow(d[d[,"category"]=="no mutation",]),")"))
                     # d<-data.frame(d,IND=paste0("samples without pan_cancer_assessment_mutation (",
                     #                            nrow(d[d[,"category"]=="no pan_cancer_assessment_mutation",]),")"))
                     d[,"IND"]<-as.character(d[,"IND"])
                     d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                 nrow(d[d[,"category"]=="mutated",]),")")
                     # d[d[,"category"]=="mutated","IND"]<- paste0("samples with pan_cancer_assessment_mutation (",
                     #                                                 nrow(d[d[,"category"]=="mutated",]),")")
                     d<-d[order(d[,"IND"]),]
                     
                     fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                     
                     try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                     if (class(try.test)=="try-error" | is.na(try.test)) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     
                      if (input_adjust_OS_pan_cancer_assessment_mutation=="None") {
                        diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                        if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_mutation," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     #                     diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                     # #                    fit.result <- survfit(Surv(OS,dead)~category,data=d)
                     # #                    plot.result <- ggsurv(fit.result,surv.col = c("lightgreen","plum"))
                     # #                    diff.result <- survdiff(Surv(OS,dead)~category,data=d)
                     #                     p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                     plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=median(d$OS),y=0.05))
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       incProgress(100, detail = "Doing part %")
                     p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                       #ggplotly(plot.result,height = 480 ) %>%
                       layout(title="the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                              legend=list(orientation="v",x=0.6,y=1))
                     }
                     
                   }
                   
                 }
                 
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_mutation_survival_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_mutation_survival_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
           
         })
       })
       
       observeEvent(input$do_pan_cancer_assessment_mutation,{
         input$do_pan_cancer_assessment_mutation
         isolate({
           input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
           selected.database <- selectedmutation_pan_cancer_assessment_mutation()
           input_adjust_OS_pan_cancer_assessment_mutation <- input$adjust_OS_pan_cancer_assessment_mutation
           # selected.database <- selecteddatabase_pan_cancer_assessment_mutation()
           # selected.pan_cancer_assessment_mutations <- all.pan_cancer_assessment_mutations[all.pan_cancer_assessment_mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           # selected.pan_cancer_assessment_mutations.1 <- selected.pan_cancer_assessment_mutations[selected.pan_cancer_assessment_mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
           #      if (nrow(selected.database)==0 | nrow(selected.pan_cancer_assessment_mutations.1)==0) {
           if (nrow(selected.database)==0 | nrow(selected.database[selected.database[,"category"]!="unknown",])==0) {
             
             output[["pan_cancer_assessment_individual_mutation_survival_plots0"]] <- renderPlotly({
               plot_ly () %>%
                 layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
                 add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
             })
             
           } else {
             selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
             splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
             splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
             
             for (k in 1:length(input$data_source_pan_cancer_assessment_mutation)) {
               local({
                 my_i <- k
                 pan_cancer_assessment_individual_mutation_survival_plots_name <- paste0("pan_cancer_assessment_individual_mutation_survival_plots", my_i)
                 pan_cancer_assessment_individual_mutation_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_mutation_survival_summary_tables", my_i)
                 d<-splitted_list[[my_i]]
                 # d.pan_cancer_assessment_mutations <- splitted_list_pan_cancer_assessment_mutation[[my_i]]
                 #output[[pan_cancer_assessment_mutation_survival_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   isolate({
                     #                if (is.na(names(splitted_list)[my_i]) | is.na(names(splitted_list_pan_cancer_assessment_mutation)[my_i])) {
                     if (is.na(names(splitted_list)[my_i]) ) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     } else {
                       if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         #                    d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) ,]
                         d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                         if (length(unique(d$category))==1) {
                           incProgress(100, detail = "Doing part %")
                           p1 <- plot_ly () %>%
                             layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                             add_annotations(x=2,y=2,text="There is only 1 group under this selection",showarrow=F,font=list(size=16,color="purple"))
                           table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         } else {
                           d<-data.frame(d,dead=NA)
                           d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                           d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                           d <- d[!is.na(d[,"dead"]),]
                           d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                           d <- data.frame(data_source=as.character(d$data_source),
                                           dead=as.numeric(d$dead),
                                           OS=as.numeric(d$OS),
                                           Age=as.character(d$Age),
                                           Gender=as.character(d$Gender),
                                           category=as.character(d[,"category"]))
                           #d<-data.frame(d[,c(1:6)],unfactor(d[,c("dead","OS","PFS","Age","Gender")]),category=as.character(d[,"category"]))
                           #d$OS<-as.numeric(d$OS)
                           d<-data.frame(d,IND=paste0("no mutation (",
                                                      nrow(d[d[,"category"]=="no mutation",]),")"))
                           # d<-data.frame(d,IND=paste0("samples without pan_cancer_assessment_mutation (",
                           #                            nrow(d[d[,"category"]=="no pan_cancer_assessment_mutation",]),")"))
                           d[,"IND"]<-as.character(d[,"IND"])
                           d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                       nrow(d[d[,"category"]=="mutated",]),")")
                           # d[d[,"category"]=="mutated","IND"]<- paste0("samples with pan_cancer_assessment_mutation (",
                           #                                                 nrow(d[d[,"category"]=="mutated",]),")")
                           d<-d[order(d[,"IND"]),]
                           
                           fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                           
                           try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                           if (class(try.test)=="try-error" | is.na(try.test)) {
                             incProgress(100, detail = "Doing part %")
                             p1 <- plot_ly () %>%
                               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                               add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                             table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           } else {
                            plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                           
                            if (input_adjust_OS_pan_cancer_assessment_mutation=="None") {
                             diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                             if (class(diff.result.try)=="try-error") {
                               p.value <- NA
                               table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                             } else {
                               p.value <- signif(summary(diff.result)$logtest[3],2)
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                               
                             }
                           } else {
                             diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation),data=d) )
                             if (class(diff.result.adjust.try)=="try-error") {
                               diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                               p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                                 " (No ",input_adjust_OS_pan_cancer_assessment_mutation," was observed)")
                               
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             } else {
                               p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                               table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                               rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation
                               table1 <- datatable(table1.pre,
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result.adjust)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result.adjust)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             }
                           }
                           
                           
                           #                     diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                           # #                    fit.result <- survfit(Surv(OS,dead)~category,data=d)
                           # #                    plot.result <- ggsurv(fit.result,surv.col = c("lightgreen","plum"))
                           # #                    diff.result <- survdiff(Surv(OS,dead)~category,data=d)
                           #                     p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                           plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                             ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=median(d$OS),y=0.05))
                                                      paste0("p-value = ",p.value),x=300,y=0.05)) +
                             incProgress(100, detail = "Doing part %")
                           p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                             #ggplotly(plot.result,height = 480 ) %>%
                             layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                    xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                    yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                    legend=list(orientation="v",x=0.6,y=1))
                           }
                         }
                         
                       }
                       
                     }
                   })
                 })
                 #})
                 output[[pan_cancer_assessment_individual_mutation_survival_plots_name]] <- renderPlotly({
                   p1
                 })
                 output[[pan_cancer_assessment_individual_mutation_survival_summary_tables_name]] <- renderDataTable(
                   table1
                 )
               })
             }
           }
         })
       })
       
       ##############
       #### mutation PFS
       ##############
       observeEvent(input$do_pan_cancer_assessment_mutation,{
         input$do_pan_cancer_assessment_mutation
         isolate({
           input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
           selected.database <- selectedmutation_pan_cancer_assessment_mutation()#selectedpan_cancer_assessment_mutation_pan_cancer_assessment_mutation()
           # selected.pan_cancer_assessment_mutations <- all.pan_cancer_assessment_mutations[all.pan_cancer_assessment_mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           # selected.pan_cancer_assessment_mutations.1 <- selected.pan_cancer_assessment_mutations[selected.pan_cancer_assessment_mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
           input_adjust_OS_pan_cancer_assessment_mutation <- input$adjust_OS_pan_cancer_assessment_mutation
           
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
           
           
           local({
             # my_i <- k
             # pan_cancer_assessment_mutation_PFS_plots_name <- paste0("pan_cancer_assessment_mutation_PFS_plots", my_i)
             # pan_cancer_assessment_mutation_PFS_summary_tables_name <- paste0("pan_cancer_assessment_mutation_PFS_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_mutation_PFS_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 #   
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & d[,"PFS.status"]!="Unknown" & !is.na(d[,"PFS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",])==0) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no PFS data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   
                 } else {
                   d <- d[!is.na(d[,"PFS.status"]) & d[,"PFS.status"]!="Unknown" & !is.na(d[,"PFS"]) & !is.na(d[,"category"]) & d[,"category"]!="unknown",]
                   if (length(unique(d$category))==1) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="There is only 1 group under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="There is only 1 group under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     d <- data.frame(data_source=as.character(d$data_source),
                                     PFS.status=as.character(d$PFS.status),
                                     PFS=as.numeric(d$PFS),
                                     Age=as.character(d$Age),
                                     Gender=as.character(d$Gender),
                                     category=as.character(d[,"category"]))
                     d<-data.frame(d,IND=paste0("no mutation (",
                                                nrow(d[d[,"category"]=="no mutation",]),")"))
                     # d<-data.frame(d,IND=paste0("samples without pan_cancer_assessment_mutation (",
                     #                            nrow(d[d[,"category"]=="no pan_cancer_assessment_mutation",]),")"))
                     d[,"IND"]<-as.character(d[,"IND"])
                     d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                 nrow(d[d[,"category"]=="mutated",]),")")
                     # d[d[,"category"]=="mutated","IND"]<- paste0("samples with pan_cancer_assessment_mutation (",
                     #                                             nrow(d[d[,"category"]=="mutated",]),")")
                     d <- data.frame(d,group.IND=0)
                     d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                     d <- d[order(d[,"IND"]),]
                     fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                     
                     try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                     if (class(try.test)=="try-error" | is.na(try.test)) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     } else {
                      plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     
                     
                      if (input_adjust_OS_pan_cancer_assessment_mutation=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_mutation," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                     # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                     plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       incProgress(100, detail = "Doing part %")
                     p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                       layout(title="the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                              yaxis=list(title="Preogression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                              legend=list(orientation="v",x=0.6,y=1))
                     }
                   }
                   
                 }
                 
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_mutation_PFS_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_mutation_PFS_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           
           
         })
       })
       observeEvent(input$do_pan_cancer_assessment_mutation,{
         input$do_pan_cancer_assessment_mutation
         isolate({
           input_gene_pan_cancer_assessment_mutation <- input$gene_pan_cancer_assessment_mutation
           selected.database <- selectedmutation_pan_cancer_assessment_mutation()
           # selected.pan_cancer_assessment_mutations <- all.pan_cancer_assessment_mutations[all.pan_cancer_assessment_mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           # selected.pan_cancer_assessment_mutations.1 <- selected.pan_cancer_assessment_mutations[selected.pan_cancer_assessment_mutations[,"gene"] %in% input_gene_pan_cancer_assessment_mutation,]
           input_adjust_OS_pan_cancer_assessment_mutation <- input$adjust_OS_pan_cancer_assessment_mutation
           if (nrow(selected.database)==0 | nrow(selected.database[selected.database[,"category"]!="no data available",])==0) {
             
             output[["pan_cancer_assessment_individual_mutation_PFS_plots0"]] <- renderPlotly({
               plot_ly () %>%
                 layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
                 add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
             })
             
           } else {
             selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
             splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
             splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)]
             
             
             for (k in 1:length(input$data_source_pan_cancer_assessment_mutation)) {
               local({
                 my_i <- k
                 pan_cancer_assessment_individual_mutation_PFS_plots_name <- paste0("pan_cancer_assessment_individual_mutation_PFS_plots", my_i)
                 pan_cancer_assessment_individual_mutation_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_mutation_PFS_summary_tables", my_i)
                 d<-splitted_list[[my_i]]
                 #output[[pan_cancer_assessment_mutation_PFS_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   isolate({
                     if (is.na(names(splitted_list)[my_i])) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]),])==0 | nrow(d[d[,"category"]!="unknown",])==0) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no PFS data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         
                       } else {
                         d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & d[,"category"]!="unknown",]
                         if (length(unique(d$category))==1) {
                           incProgress(100, detail = "Doing part %")
                           p1 <- plot_ly () %>%
                             layout(title = unique(as.character(d$data_source)),font=list(size=15,family="Aril")) %>%
                             add_annotations(x=2,y=2,text="There is only 1 group under this selection",showarrow=F,font=list(size=16,color="purple"))
                           table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           
                         } else {
                           d <- data.frame(data_source=as.character(d$data_source),
                                           PFS.status=as.character(d$PFS.status),
                                           PFS=as.numeric(d$PFS),
                                           Age=as.character(d$Age),
                                           Gender=as.character(d$Gender),
                                           category=as.character(d[,"category"]))
                           d<-data.frame(d,IND=paste0("no mutation (",
                                                      nrow(d[d[,"category"]=="no mutation",]),")"))
                           # d<-data.frame(d,IND=paste0("samples without pan_cancer_assessment_mutation (",
                           #                            nrow(d[d[,"category"]=="no mutation",]),")"))
                           d[,"IND"]<-as.character(d[,"IND"])
                           d[d[,"category"]=="mutated","IND"]<- paste0("mutated (",
                                                                       nrow(d[d[,"category"]=="mutated",]),")")
                           # d[d[,"category"]=="mutated","IND"]<- paste0("samples with pan_cancer_assessment_mutation (",
                           #                                             nrow(d[d[,"category"]=="mutated",]),")")
                           d <- data.frame(d,group.IND=0)
                           d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                           d <- d[order(d[,"IND"]),]
                           fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                           
                           try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                           # if (class(try.test)=="try-error" | is.na(try.test)) {
                           #   incProgress(100, detail = "Doing part %")
                           #   p1 <- plot_ly () %>%
                           #     layout(title = as.character(unique(d$data_source),font=list(size=15,family="Aril")) %>%
                           #     add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                           #   table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           # } else {
                            plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                           
                           
                            if (input_adjust_OS_pan_cancer_assessment_mutation=="None") {
                             diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                             if (class(diff.result.try)=="try-error") {
                               p.value <- NA
                               table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                             } else {
                               p.value <- signif(summary(diff.result)$logtest[3],2)
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                               
                             }
                           } else {
                             diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation),data=d) )
                             if (class(diff.result.adjust.try)=="try-error") {
                               diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                               p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                                 " (No ",input_adjust_OS_pan_cancer_assessment_mutation," was observed)")
                               
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             } else {
                               p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                               table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                               rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation
                               table1 <- datatable(table1.pre,
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result.adjust)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result.adjust)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             }
                           }
                           
                           # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                           # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                           plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                             ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                      paste0("p-value = ",p.value),x=300,y=0.05)) +
                             incProgress(100, detail = "Doing part %")
                           p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                             layout(title=as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                    xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                    yaxis=list(title="Preogression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                    legend=list(orientation="v",x=0.6,y=1))
                           
                         }
                         
                       }
                       
                     }
                   })
                 })
                 #})
                 output[[pan_cancer_assessment_individual_mutation_PFS_plots_name]] <- renderPlotly({
                   p1
                 })
                 output[[pan_cancer_assessment_individual_mutation_PFS_summary_tables_name]] <- renderDataTable(
                   table1
                 )
               })
             }
           }
         })
       })
     
     })
     
     ######################################################
     #### pan-cancer assessment mutation loads
     ######################################################
     # input <- list(data_source_pan_cancer_assessment_mutation_loads=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Van Allen (112 samples/+--)" ),
     # treatment_pan_cancer_assessment_mutation_loads=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
     # description_pan_cancer_assessment_mutation_loads=c("Pre","On"),slider_pan_cancer_assessment_mutation_loads_OS=0.5)
 
     # input <- list(data_source_pan_cancer_assessment_mutation_loads=c("Non-Small Cell Lung Cancer-Hira Rizvi (240 samples/-++)"),
     # treatment_pan_cancer_assessment_mutation_loads=c("anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),
     # description_pan_cancer_assessment_mutation_loads=c("Pre","On"),slider_pan_cancer_assessment_mutation_loads_OS=0.5)
     
     
     observe({
       if (input$data_source_pan_cancer_assessment_mutation_loads=="" || is.null(input$data_source_pan_cancer_assessment_mutation_loads) ||
           input$treatment_pan_cancer_assessment_mutation_loads=="" || is.null(input$treatment_pan_cancer_assessment_mutation_loads) ||
           input$description_pan_cancer_assessment_mutation_loads=="" || is.null(input$description_pan_cancer_assessment_mutation_loads) ){
         disable("do_pan_cancer_assessment_mutation_loads")
       } else {
         enable("do_pan_cancer_assessment_mutation_loads")
       }
     })
     
     selecteddatabase_pan_cancer_assessment_mutation_loads <- eventReactive(input$do_pan_cancer_assessment_mutation_loads,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)
       # treatment1 <- input$treatment_pan_cancer_assessment_mutation_loads
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_mutation_loads])
       description1 <- input$description_pan_cancer_assessment_mutation_loads
       genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","Mutation.Load")
       a<-data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       colnames(m)[13] <- "Mutation.Load"
       m<-m[,Mutation.Load:=as.numeric(Mutation.Load)]
       m <- as.data.frame(m)
       if (nrow(m)>0) {
         #        m<-data.frame(m,category.OS="no data available",category.PFS="no data available")
         m<-data.frame(m,category="unknown")
         m[,"category"] <- as.character(m[,"category"])
         d <- m
         # m1 <- m %>% split(.$data_source) %>% 
         #   lapply(function(d) {
             d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),"category"] <- paste0("top ",
                                                                                                                                                            (1-input$slider_pan_cancer_assessment_mutation_loads_OS)*100,
                                                                                                                                                            "% ",sep="")
             d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<=quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),"category"] <- paste0("bottom ",
                                                                                                                                                            input$slider_pan_cancer_assessment_mutation_loads_OS*100,
                                                                                                                                                            "% ",sep="")
             
             d #})
         
         # m2 <- ldply(m1,data.frame)
         # m2 <- m2[,-1]
         # m2
       } else {
         m2<-m
         m2
       } 
     })
     
     selecteddatabase_pan_cancer_assessment_individual_mutation_loads <- eventReactive(input$do_pan_cancer_assessment_mutation_loads,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)
       #treatment1 <- input$treatment_pan_cancer_assessment_mutation_loads
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_mutation_loads])
       description1 <- input$description_pan_cancer_assessment_mutation_loads
       genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender","Mutation.Load")
       a<-data.table(database.test.process())
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1 ]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       m<-m[,Mutation.Load:=as.numeric(Mutation.Load)]
       m<-as.data.frame(m)
       if (nrow(m)>0 & ncol(m)>12) {
         #      m <- data.frame(m,category.OS="no data available",category.PFS="no data available")
         m <- data.frame(m,category="unknown")
         m1 <- m %>% split(.$data_source) %>% 
           lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_expression," expression",sep="")
             #d[,"category.OS"] <- as.character(d[,"category.OS"])
             d[,"category"] <- as.character(d[,"category"])
             d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<= quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),"category"] <- 
               paste0("bottom ",
                      input$slider_pan_cancer_assessment_mutation_loads_OS*100,
                      "%")
             d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),"category"] <- 
               paste0("top ",
                      (1-input$slider_pan_cancer_assessment_mutation_loads_OS)*100,
                      "%")
             d})
         m2<-ldply(m1,data.frame)
         m2<-m2[,-1]
         m2
       } else {
         m2<-m
         m2
       }
     })
     
     ##########
     #### mutational loads summary
     #########
     output$pan_cancer_assessment_mutation_loads_summary_disease_number <- renderText({
       input$do_pan_cancer_assessment_mutation_loads ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_assessment_mutation_loads()$Disease))})})
     output$pan_cancer_assessment_mutation_loads_summary_sample_number <- renderText({
       input$do_pan_cancer_assessment_mutation_loads ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_assessment_mutation_loads())})})
     output$pan_cancer_assessment_mutation_loads_summary_plots <- renderPlotly({
       summary_plot(inputdata=selecteddatabase_pan_cancer_assessment_mutation_loads(),feature="mutation_loads")
     })
     
     
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_loads()
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_mutation_loads_summary_clinical_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[input$data_source_pan_cancer_assessment_mutation_loads]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_loads)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_mutation_loads_summary_clinical_plots_name <- paste0("pan_cancer_assessment_mutation_loads_summary_clinical_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_mutation_loads_summary_clinical_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = input$data_source_pan_cancer_assessment_mutation_loads[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 if (nrow(d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_loads_clinical]]) & !is.na(d[,"Mutation.Load"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   d <- d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_loads_clinical]]) & !is.na(d[,"Mutation.Load"]),]
                   if(input$pan_cancer_assessment_mutation_loads_clinical=="Age") {
                     d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                     #d[d[,"Age"]<=10,"Age"] <- "0-10"
                     d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                     d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                   }
                   
                   p<-plot_ly(data=d,x=d[,clinical_server[input$pan_cancer_assessment_mutation_loads_clinical]],
                              y=~as.numeric(as.character(Mutation.Load)), 
                              type="violin",#type="box",
                              points="all",#boxpoints="all",
                              jitter=0.3,
                              pointpos=-1.8,
                              #mode="markders",
                              color=d[,clinical_server[input$pan_cancer_assessment_mutation_loads_clinical]],showlegend=TRUE,
                              colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                              hoveron="points+boxes",
                              legendgroup=d[,clinical_server[input$pan_cancer_assessment_mutation_loads_clinical]], 
                              text= ~paste0("Sample ID: ", PatientID,
                                            "<br>mutation loads: ",Mutation.Load,
                                            "<br>Treatment: ",treatment,
                                            "<br>Description: ",description)) %>%
                     layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                            height = 480,
                            xaxis=list(title="Clinical Attribute",
                                       titlefont=list(size=20,color="black",family= "Aril black"),
                                       showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                            yaxis=list(title="Mutation loads/burden",
                                       titlefont=list(size=20,color="black",family= "Aril black"),
                                       showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                            ))
                   incProgress(100, detail = "Doing part %")
                   p
                   
                 } 
                 #}
               })
             })
           })
           #}
         }
       }) 
     })
     
     
     output$pan_cancer_assessment_mutation_loads_summary_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_assessment_mutation_loads_summary_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_assessment_mutation_loads_summary_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_assessment_mutation_loads_summary_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         if (input$pan_cancer_assessment_mutation_loads_summary_type == "Excel (CSV)") {
           write.csv (selecteddatabase_pan_cancer_assessment_mutation_loads(),file,row.names = FALSE)
         } else if (input$pan_cancer_assessment_mutation_loads_summary_type == "Text (TSV)") {
           write.table (selecteddatabase_pan_cancer_assessment_mutation_loads(),file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_assessment_mutation_loads_summary_type == "JSON") {
           write(toJSON(selecteddatabase_pan_cancer_assessment_mutation_loads()),file)
         }
       }
     )
     
     output$pan_cancer_assessment_mutation_loads_summary_table <- DT::renderDataTable({
       selecteddatabase_pan_cancer_assessment_mutation_loads()
     })
     
     ##########
     #### mutational loads plots
     ##########
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_loads()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Mutation.Load"]),])==0) {
           output[["pan_cancer_assessment_mutation_loads_plots_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[input$data_source_pan_cancer_assessment_mutation_loads]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_loads)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_mutation_loads_plots_plots_name <- paste0("pan_cancer_assessment_mutation_loads_plots_plots", my_i)
             d<-splitted_list#[[my_i]]
             output[["pan_cancer_assessment_mutation_loads_plots_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = input$data_source_pan_cancer_assessment_mutation_loads[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 #d <- splitted_list[[my_i]]
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Mutation.Load"]),])==0) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Mutation.Load"]),]
                   d[,"Groups"] <- as.character(d[,"Groups"])
                   
                   d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                   d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                   ind <- sort(unique(d[,"Groups"]))
                   try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1] & !is.na(d[,"Groups"]),"Mutation.Load"]),as.numeric(d[d[,"Groups"]==ind[2] & !is.na(d[,"Groups"]),"Mutation.Load"]))$p.value)
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     p.v <- " = NA"
                   } else {
                     # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"loads"]),as.numeric(d[d[,"Groups"]==ind[2],"loads"]))$p.value
                     # if (is.na(p.v)) {
                     #   p.v <- " = NA"
                     # } else {
                     #   if (p.v < 0.001) {
                     #     p.v <- " < 0.001"
                     #   } else {
                     p.v<-paste0(" = ",signif(p.v,2))
                     # } 
                     # }
                   }
                   p<-plot_ly(data=d,x=~Groups,
                              y=~Mutation.Load,
                              type="violin",#type="box",
                              points="all",#boxpoints="all",
                              jitter=0.3,
                              pointpos=-1.8,
                              #mode="markders",
                              color= ~PFS.status,
                              colors=c("#7CDBD5","#DCB239","gray"),
                              #hoveron="points+boxes",
                              legendgroup=~PFS.status,
                              showlegend=TRUE,
                              text= ~paste0("Sample ID: ", PatientID,
                                            "<br>mutation loads: ",Mutation.Load,
                                            "<br>Treatment: ",treatment,
                                            "<br>Description: ",description)) %>%
                     layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                            height = 480,
                            xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                       showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                            yaxis=list(title="Mutation loads/burden",
                                       titlefont=list(size=20,color="black",family= "Aril"),
                                       showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                            )) %>%
                     add_annotations(x=0.5,
                                     y=d[which.max(d[,"Mutation.Load"]),"Mutation.Load"],
                                     text=paste0("p-value",p.v),
                                     xref = paste0("x", 1),  # add this
                                     yref = paste0("y", 1),  # add this
                                     showarrow = FALSE,
                                     ax = 20,
                                     ay = -40,
                                     font=list(size=20,color="black",family= "Aril black"))
                   incProgress(100, detail = "Doing part %")
                   p
                 } 
                 
                 #}
               })
             })
           })
           #}
         }
       })
     })
     
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       
       #isolate({
       selected.database <- selecteddatabase_pan_cancer_assessment_individual_mutation_loads()
       if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Mutation.Load"]),])==0) {
         output[["pan_cancer_assessment_individual_mutation_loads_plots_plots0"]] <- renderPlotly({
           plot_ly () %>% 
             layout(title = "NA",font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
         })
         
       } else {
         selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
         splitted_list <- selected.database %>% split(as.character(.$data_source)) #%>% as.matrix %>% as.data.frame %>% split(as.character(.$data_source))
         splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)]
         #      if (!is.null(length(splitted_list))) {
         
         
         for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_loads)) {
           local({
             my_i <- k
             pan_cancer_assessment_individual_mutation_loads_plots_plots_name <- paste0("pan_cancer_assessment_individual_mutation_loads_plots_plots", my_i)
             
             d<-splitted_list[[my_i]]
             output[[pan_cancer_assessment_individual_mutation_loads_plots_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 #              isolate({
                 if (is.na(names(splitted_list)[my_i])) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no data available in this study under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   #d <- splitted_list[[my_i]]
                   if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Mutation.Load"]),])==0) {
                     incProgress(100, detail = "Doing part %")
                     plot_ly () %>% 
                       layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                   } else {
                     #d<-splitted_list[[my_i]]
                     d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Mutation.Load"]),]
                     d[,"Groups"] <- as.character(d[,"Groups"])
                     
                     d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                     d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                     ind <- sort(unique(d[,"Groups"]))
                     try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Mutation.Load"]),as.numeric(d[d[,"Groups"]==ind[2],"Mutation.Load"]))$p.value)
                     if (class(try.test)=="try-error" | is.na(try.test)) {
                       p.v <- " = NA"
                     } else {
                       # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Mutation.Load"]),as.numeric(d[d[,"Groups"]==ind[2],"Mutation.Load"]))$p.value
                       # if (p.v < 0.001) {
                       #   p.v <- "< 0.001"
                       # } else {
                       p.v<-paste0(" = ",signif(p.v,2))
                       # } 
                     }
                     
                     if (as.character(unique(d$data_source))=="Melanoma-Samstein") {
                       lable_y <- "Number of mutations per Mb"
                       text_y <- "no immunotherapy outcome for this data"
                       color_y <- "purple"
                       size_y <- 16
                     } else {
                       lable_y <- "Number of mutations"
                       text_y <- paste0("p-value",p.v)
                       color_y <- "black"
                       size_y <- 15
                     }
                     p<-plot_ly(data=d,x=~Groups,
                                y=~Mutation.Load,
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color= ~PFS.status, showlegend=TRUE,
                                colors=c("#7CDBD5","#DCB239","gray"),
                                #hoveron="points+boxes",
                                legendgroup=~PFS.status,
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Mutation loads: ",Mutation.Load,
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480 ,
                              xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=lable_y,
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15))) %>% 
                       add_annotations(x=0.5,
                                       y=d[which.max(d[,"Mutation.Load"]),"Mutation.Load"],
                                       text=text_y,
                                       font=list(size=size_y,color=color_y),
                                       xref = paste0("x", 1),  # add this
                                       yref = paste0("y", 1),  # add this
                                       showarrow = FALSE,
                                       ax = 20,
                                       ay = -40,
                                       font=list(size=25,color="black",family= "Aril black"))
                     incProgress(100, detail = "Doing part %")
                     p
                   } 
                   
                 }
               })
             })
             #          }
           })
         }
       }
       #})
     })
     
     #############
     #### mutational loads survival
     #############
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_loads()
         input_adjust_OS_pan_cancer_assessment_mutation_loads <- input$adjust_OS_pan_cancer_assessment_mutation_loads
         
         splitted_list <- selected.database #%>% split(.$data_source)# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
         splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)]
         
         # for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_loads)) {
         local({
           # my_i <- k
           # pan_cancer_assessment_mutation_loads_survival_plots_name <- paste0("pan_cancer_assessment_mutation_loads_survival_plots", my_i)
           # pan_cancer_assessment_mutation_loads_survival_summary_tables_name <- paste0("pan_cancer_assessment_mutation_loads_survival_summary_tables", my_i)
           d<-splitted_list#[[my_i]]
           #output[[pan_cancer_assessment_mutation_loads_survival_plots_name]] <- renderPlotly({
           withProgress(message = 'Making plot', value = 0, {
             isolate({
               
               
               if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),])==0) {
                 incProgress(100, detail = "Doing part %")
                 p1 <- plot_ly () %>% 
                   layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                   add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 
               } else {
                 d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),]
                 d<-data.frame(d,dead=NA)
                 d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                 d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                 d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"dead"]),]
                 d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                 #d <- data.frame(data_source=as.character(d$data_source),dead=as.numeric(d$dead),OS=as.numeric(d$OS),category=as.character(d[,"category.OS"]))
                 
                 if (as.character(unique(d$data_source))=="Melanoma-Samstein") {
                   lable_y <- "TMB"
                 } else {
                   lable_y <- "number of mutations"
                 }
                 d <- data.frame(data_source=as.character(d$data_source),
                                 dead=as.numeric(d$dead),
                                 OS=as.numeric(d$OS),
                                 Age=as.character(d$Age),
                                 Gender=as.character(d$Gender),
                                 Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                 # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Mutation.Load")]))
                 d<-data.frame(d,IND=paste0("bottom ",
                                            input$slider_pan_cancer_assessment_mutation_loads_OS*100, 
                                            "% (",
                                            nrow(d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<= quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),]),
                                            ")"))
                 d[,"IND"]<-as.character(d[,"IND"])
                 d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),"IND"]<- 
                   paste0("top ",
                          (1-input$slider_pan_cancer_assessment_mutation_loads_OS)*100,
                          "% (",
                          nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),])
                          ,")")
                 d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                 d<-d[order(d[,"IND"]),]
                 fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                 
                 try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                 if (class(try.test)=="try-error" | is.na(try.test)) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                  plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                 
                  if (input_adjust_OS_pan_cancer_assessment_mutation_loads=="None") {
                   diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                   if (class(diff.result.try)=="try-error") {
                     p.value <- NA
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                   } else {
                     p.value <- signif(summary(diff.result)$logtest[3],2)
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                     
                   }
                 } else {
                   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_loads),data=d) )
                   if (class(diff.result.adjust.try)=="try-error") {
                     diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                     p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                       " (No ",input_adjust_OS_pan_cancer_assessment_mutation_loads," was observed)")
                     
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   } else {
                     p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                     table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                     rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_loads
                     table1 <- datatable(table1.pre,
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result.adjust)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result.adjust)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   }
                 }
                 
                 
                 # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                 # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                 plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                   ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                            #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                            paste0("p-value = ",p.value),x=300,y=0.05)) +
                   ggplot2::guides(color=FALSE,linetype=FALSE)
                 p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                   layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                          xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                          yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                          legend=list(orientation="v",x=0.6,y=1)) #%>%
                 incProgress(100, detail = "Doing part %")
                 p1
                 }
               }
               
             })
           })
           #})
           output[["pan_cancer_assessment_mutation_loads_survival_plots"]] <- renderPlotly({
             p1
           })
           output[["pan_cancer_assessment_mutation_loads_survival_summary_tables"]] <- renderDataTable(
             table1
           )
         })
         #}
         
       })
     })
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_mutation_loads()
         input_adjust_OS_pan_cancer_assessment_mutation_loads <- input$adjust_OS_pan_cancer_assessment_mutation_loads
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
           output[["pan_cancer_assessment_individual_mutation_loads_survival_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="warning: no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
           
         } else {
           splitted_list <- selected.database %>% split(.$data_source)# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)]
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_loads)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_mutation_loads_survival_plots_name <- paste0("pan_cancer_assessment_individual_mutation_loads_survival_plots", my_i)
               pan_cancer_assessment_individual_mutation_loads_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_mutation_loads_survival_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_mutation_loads_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available in this study under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     
                     if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Mutation.Load"]),]
                       d<-data.frame(d,dead=NA)
                       d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                       d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                       d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"dead"]),]
                       d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                       #d <- data.frame(data_source=as.character(d$data_source),dead=as.numeric(d$dead),OS=as.numeric(d$OS),category=as.character(d[,"category.OS"]))
                       
                       if (as.character(unique(d$data_source))=="Samstein") {
                         lable_y <- "TMB"
                       } else {
                         lable_y <- "number of mutations"
                       }
                       d <- data.frame(data_source=as.character(d$data_source),
                                       dead=as.numeric(d$dead),
                                       OS=as.numeric(d$OS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Mutation.Load")]))
                       d<-data.frame(d,IND=paste0("bottom ",
                                                  input$slider_pan_cancer_assessment_mutation_loads_OS*100, 
                                                  "% (",
                                                  nrow(d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]<= quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),]),
                                                  ")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[!is.na(d[,"Mutation.Load"]) & d[,"Mutation.Load"]> quantile(na.omit(d[,"Mutation.Load"]),input$slider_pan_cancer_assessment_mutation_loads_OS),"IND"]<- 
                         paste0("top ",
                                (1-input$slider_pan_cancer_assessment_mutation_loads_OS)*100,
                                "% (",
                                nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),])
                                ,")")
                       d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                        plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                       
                       if (input_adjust_OS_pan_cancer_assessment_mutation_loads=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           
                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_loads),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_mutation_loads," was observed)")
                           
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_loads
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       
                       
                       # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                         layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                legend=list(orientation="v",x=0.6,y=1)) #%>%
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_mutation_loads_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_mutation_loads_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     #############
     #### mutational loads PFS
     #############
     
     
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_loads()
         input_adjust_OS_pan_cancer_assessment_mutation_loads <- input$adjust_OS_pan_cancer_assessment_mutation_loads
         
         splitted_list <- selected.database #%>% split(.$data_source)#%>% as.matrix %>% as.data.frame %>% split(.$data_source)
         splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)]
         
         
         
         
         local({
           # my_i <- k
           # pan_cancer_assessment_mutation_loads_PFS_plots_name <- paste0("pan_cancer_assessment_mutation_loads_PFS_plots", my_i)
           # pan_cancer_assessment_mutation_loads_PFS_summary_tables_name <- paste0("pan_cancer_assessment_mutation_loads_PFS_summary_tables", my_i)
           d<-splitted_list#[[my_i]]
           
           #output[[pan_cancer_assessment_mutation_loads_PFS_plots_name]] <- renderPlotly({
           withProgress(message = 'Making plot', value = 0, {
             isolate({
               
               if (nrow(d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                 incProgress(100, detail = "Doing part %")
                 p1 <- plot_ly () %>% 
                   layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                   add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 
               } else {
                 d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                 
                 
                 d <- data.frame(data_source=as.character(d$data_source),
                                 PFS.status=as.character(d$PFS.status),
                                 PFS=as.numeric(d$PFS),
                                 Age=as.character(d$Age),
                                 Gender=as.character(d$Gender),
                                 Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                 # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Mutation.Load")]))
                 d<-data.frame(d,IND=paste0("bottom ",input$slider_pan_cancer_assessment_mutation_loads_OS*100, "% (",
                                            nrow(d[d[,"Mutation.Load"]<= quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),]),
                                            ")"))
                 d[,"IND"]<-as.character(d[,"IND"])
                 d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),"IND"]<- 
                   paste0("top ",(1-input$slider_pan_cancer_assessment_mutation_loads_OS)*100,"% (",
                          nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),])
                          ,")")
                 d <- data.frame(d,group.IND=0)
                 d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                 #        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                 d<-d[order(d[,"IND"]),]
                 fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                 
                 try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                 if (class(try.test)=="try-error" | is.na(try.test)) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                  plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                 
                  if (input_adjust_OS_pan_cancer_assessment_mutation_loads=="None") {
                   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                   if (class(diff.result.try)=="try-error") {
                     p.value <- NA
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                   } else {
                     p.value <- signif(summary(diff.result)$logtest[3],2)
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                     
                   }
                 } else {
                   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_loads),data=d) )
                   if (class(diff.result.adjust.try)=="try-error") {
                     diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                     p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                       " (No ",input_adjust_OS_pan_cancer_assessment_mutation_loads," was observed)")
                     
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   } else {
                     p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                     table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                     rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_loads
                     table1 <- datatable(table1.pre,
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result.adjust)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result.adjust)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   }
                 }
                 
                 
                 
                 # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                 # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                 plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                   ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                            #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                            paste0("p-value = ",p.value),x=300,y=0.05)) +
                   ggplot2::guides(color=FALSE,linetype=FALSE)
                 p1<-ggplotly(plot.result.ann, height = 450  ) %>%
                   layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                          xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                          yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                          legend=list(orientation="v",x=0.6,y=1)) 
                 incProgress(100, detail = "Doing part %")
                 p1
                 }
               }
               
             })
           })
           #})
           output[["pan_cancer_assessment_mutation_loads_PFS_plots"]] <- renderPlotly({
             p1
           })
           output[["pan_cancer_assessment_mutation_loads_PFS_summary_tables"]] <- renderDataTable(
             table1
           )
         })
         
         
       })
     })
     observeEvent(input$do_pan_cancer_assessment_mutation_loads,{
       input$do_pan_cancer_assessment_mutation_loads
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_mutation_loads()
         input_adjust_OS_pan_cancer_assessment_mutation_loads <- input$adjust_OS_pan_cancer_assessment_mutation_loads
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
           output[["pan_cancer_assessment_individual_mutation_loads_PFS_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           splitted_list <- selected.database %>% split(.$data_source)#%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)]
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_loads)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_mutation_loads_PFS_plots_name <- paste0("pan_cancer_assessment_individual_mutation_loads_PFS_plots", my_i)
               pan_cancer_assessment_individual_mutation_loads_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_mutation_loads_PFS_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               
               #output[[pan_cancer_assessment_mutation_loads_PFS_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation_loads,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     if (nrow(d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"Mutation.Load"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                       
                       if (as.character(unique(d$data_source))=="Samstein") {
                         lable_y <- "TMB"
                       } else {
                         lable_y <- "number of mutations"
                       }
                       d <- data.frame(data_source=as.character(d$data_source),
                                       PFS.status=as.character(d$PFS.status),
                                       PFS=as.numeric(d$PFS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       Mutation.Load=as.numeric(d[,"Mutation.Load"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Mutation.Load")]))
                       d<-data.frame(d,IND=paste0("bottom ",input$slider_pan_cancer_assessment_mutation_loads_OS*100, "% (",
                                                  nrow(d[d[,"Mutation.Load"]<= quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),]),
                                                  ")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),"IND"]<- 
                         paste0("top ",(1-input$slider_pan_cancer_assessment_mutation_loads_OS)*100,"% (",
                                nrow(d[d[,"Mutation.Load"]> quantile(d[,"Mutation.Load"],input$slider_pan_cancer_assessment_mutation_loads_OS),])
                                ,")")
                       d <- data.frame(d,group.IND=0)
                       d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                       #        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                        plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                       
                       if (input_adjust_OS_pan_cancer_assessment_mutation_loads=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           
                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_loads),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_mutation_loads," was observed)")
                           
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_loads
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       
                       
                       
                       # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann, height = 450  ) %>%
                         layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                legend=list(orientation="v",x=0.6,y=1)) 
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_mutation_loads_PFS_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_mutation_loads_PFS_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     ######################################################
     #### pan-cancer assessment mutation signatures
     ######################################################
     
     
     #input <- list(data_source_pan_cancer_assessment_mutation_signature=c("Melanoma-Van Allen","Melanoma-Hugo","Hepatocellular Carcinoma-Harding"),
     #treatment_pan_cancer_assessment_mutation_signature=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
     #description_pan_cancer_assessment_mutation_signature=c("Pre","On","Unknown"),
     #signature_pan_cancer_assessment_mutation_signature="Signature.1 (age correlated)",
     #slider_pan_cancer_assessment_mutation_signature_OS=0.5)
     
     
     updateSelectizeInput(session,"signature_pan_cancer_assessment_mutation_signature",
                          choices= sort(c("Signature.1 (age correlated)","Signature.2","Signature.3 (BRCA1 and BRCA2 associated)","Signature.4 (tobacco associated)","Signature.5","Signature.6 (DNA mismatch repair associated)","Signature.7 (ultraviolet associated)","Signature.8","Signature.9","Signature.10","Signature.11","Signature.12","Signature.13","Signature.14 (high tumor mutation burden associated)","Signature.15 (DNA mismatch repair associated)","Signature.16","Signature.17","Signature.18","Signature.19","Signature.20 (DNA mismatch repair associated)","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25 (Hodgkin's cell lines only)","Signature.26 (DNA mismatch repair associated)","Signature.27","Signature.28","Signature.29","Signature.30","unknown")),
                          selected="Signature.1 (age correlated)",
                          server=TRUE)
     observe({
       if (input$data_source_pan_cancer_assessment_mutation_signature=="" || is.null(input$data_source_pan_cancer_assessment_mutation_signature) ||
           input$treatment_pan_cancer_assessment_mutation_signature=="" || is.null(input$treatment_pan_cancer_assessment_mutation_signature) ||
           input$description_pan_cancer_assessment_mutation_signature=="" || is.null(input$description_pan_cancer_assessment_mutation_signature) ||
           input$signature_pan_cancer_assessment_mutation_signature=="" || is.null(input$signature_pan_cancer_assessment_mutation_signature)){
         disable("do_pan_cancer_assessment_mutation_signature")
       } else {
         enable("do_pan_cancer_assessment_mutation_signature")
       }
     })
     
     selecteddatabase_pan_cancer_assessment_mutation_signature <- eventReactive(input$do_pan_cancer_assessment_mutation_signature,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)
       # treatment1 <- input$treatment_pan_cancer_assessment_mutation_signature
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_mutation_signature])
       description1 <- input$description_pan_cancer_assessment_mutation_signature
       genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",str_split(input$signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1])
       #a<-data.table(database.test.process())
       a <- data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       colnames(m)[13] <- "Signature"
       m<-m[,Signature:=as.numeric(Signature)]
       m <- as.data.frame(m)
       if (nrow(m)>0) {
         #        m<-data.frame(m,category.OS="no data available",category.PFS="no data available")
         m<-data.frame(m,category="unknown")
         m[,"category"] <- as.character(m[,"category"])
         d <- m
         # m1 <- m %>% split(.$data_source) %>% 
         #   lapply(function(d) {
             d[!is.na(d[,"Signature"]) & d[,"Signature"]> quantile(na.omit(d[,"Signature"]),input$slider_pan_cancer_assessment_mutation_signature_OS),"category"] <- paste0("top ",
                                                                                                                                                                            (1-input$slider_pan_cancer_assessment_mutation_signature_OS)*100,
                                                                                                                                                                            "% ",sep="")
             d[!is.na(d[,"Signature"]) & d[,"Signature"]<=quantile(na.omit(d[,"Signature"]),input$slider_pan_cancer_assessment_mutation_signature_OS),"category"] <- paste0("bottom ",
                                                                                                                                                                            input$slider_pan_cancer_assessment_mutation_signature_OS*100,
                                                                                                                                                                            "% ",sep="")
             
             d# })
         
         # m2 <- ldply(m1,data.frame)
         # m2 <- m2[,-1]
         # m2
       } else {
         m2<-m
         m2
       } 
     })
     
     selecteddatabase_pan_cancer_assessment_individual_mutation_signature <- eventReactive(input$do_pan_cancer_assessment_mutation_signature,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)
       # treatment1 <- input$treatment_pan_cancer_assessment_mutation_signature
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_mutation_signature])
       description1 <- input$description_pan_cancer_assessment_mutation_signature
       genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",str_split(input$signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1])
       a<-data.table(database.test.process())
       #a <- data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       colnames(m)[13] <- "Signature"
       m<-m[,Signature:=as.numeric(Signature)]
       m <- as.data.frame(m)
       if (nrow(m)>0) {
         #        m<-data.frame(m,category.OS="no data available",category.PFS="no data available")
         m<-data.frame(m,category="unknown")
         m[,"category"] <- as.character(m[,"category"])
         #d <- m
         m1 <- m %>% split(.$data_source) %>%
           lapply(function(d) {
         d[!is.na(d[,"Signature"]) & d[,"Signature"]> quantile(na.omit(d[,"Signature"]),input$slider_pan_cancer_assessment_mutation_signature_OS),"category"] <- paste0("top ",
                                                                                                                                                                        (1-input$slider_pan_cancer_assessment_mutation_signature_OS)*100,
                                                                                                                                                                        "% ",sep="")
         d[!is.na(d[,"Signature"]) & d[,"Signature"]<=quantile(na.omit(d[,"Signature"]),input$slider_pan_cancer_assessment_mutation_signature_OS),"category"] <- paste0("bottom ",
                                                                                                                                                                        input$slider_pan_cancer_assessment_mutation_signature_OS*100,
                                                                                                                                                                        "% ",sep="")

         d})

         m2 <- ldply(m1,data.frame)
         m2 <- m2[,-1]
         m2
       } else {
         m2<-m
         m2
       }
     })
     
     ##########
     #### mutational signature summary
     #########
     output$pan_cancer_assessment_mutation_signature_summary_disease_number <- renderText({
       input$do_pan_cancer_assessment_mutation_signature ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_assessment_mutation_signature()$Disease))})})
     output$pan_cancer_assessment_mutation_signature_summary_sample_number <- renderText({
       input$do_pan_cancer_assessment_mutation_signature ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_assessment_mutation_signature())})})
     output$pan_cancer_assessment_mutation_signature_summary_plots <- renderPlotly({
       summary_plot(inputdata=selecteddatabase_pan_cancer_assessment_mutation_signature(),feature="mutation_signature")
     })
     
     
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         input_signature_pan_cancer_assessment_mutation_signature <- input$signature_pan_cancer_assessment_mutation_signature
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_signature()
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_mutation_signature_summary_clinical_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[input$data_source_pan_cancer_assessment_mutation_signature]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_signature)) {
             local({
               # my_i <- k
               # pan_cancer_assessment_mutation_signature_summary_clinical_plots_name <- paste0("pan_cancer_assessment_mutation_signature_summary_clinical_plots", my_i)
               d<-splitted_list#[[my_i]]
               
               output[["pan_cancer_assessment_mutation_signature_summary_clinical_plots"]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   # if (is.na(names(splitted_list)[my_i])) {
                   #   incProgress(100, detail = "Doing part %")
                   #   plot_ly () %>% 
                   #     layout(title = input$data_source_pan_cancer_assessment_mutation_signature[[my_i]],font=list(size=15,family="Aril")) %>%
                   #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                   # } else {
                     if (nrow(d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_signature_clinical]]) & !is.na(d[,"Signature"]),])==0 ) {
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>% 
                         layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     } else {
                       d <- d[!is.na(d[,clinical_server[input$pan_cancer_assessment_mutation_signature_clinical]]) & !is.na(d[,"Signature"]),]
                       if(input$pan_cancer_assessment_mutation_signature_clinical=="Age") {
                         d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                         #d[d[,"Age"]<=10,"Age"] <- "0-10"
                         d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                         d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                       }
                       
                       p<-plot_ly(data=d,x=d[,clinical_server[input$pan_cancer_assessment_mutation_signature_clinical]],
                                  y=~as.numeric(as.character(Signature)), 
                                  type="violin",#type="box",
                                  points="all",#boxpoints="all",
                                  jitter=0.3,
                                  pointpos=-1.8,
                                  #mode="markders",
                                  color=d[,clinical_server[input$pan_cancer_assessment_mutation_signature_clinical]],showlegend=TRUE,
                                  colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                  hoveron="points+boxes",
                                  legendgroup=d[,clinical_server[input$pan_cancer_assessment_mutation_signature_clinical]], 
                                  text= ~paste0("Sample ID: ", PatientID,
                                                "<br>Signature: ",str_split(input$signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1]," :",Signature,
                                                "<br>Treatment: ",treatment,
                                                "<br>Description: ",description)) %>%
                         layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                                height = 480,
                                xaxis=list(title="Clinical Attribute",
                                           titlefont=list(size=20,color="black",family= "Aril black"),
                                           showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                yaxis=list(title=paste(str_split(input_signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1], " weight",sep=""),
                                           titlefont=list(size=20,color="black",family= "Aril black"),
                                           showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                ))
                       incProgress(100, detail = "Doing part %")
                       p
                       
                     } 
                   #}
                 })
               })
             })
           #}
         }
       }) 
     })
     
     
     output$pan_cancer_assessment_mutation_signature_summary_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_assessment_mutation_signature_summary_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_assessment_mutation_signature_summary_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_assessment_mutation_signature_summary_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         if (input$pan_cancer_assessment_mutation_signature_summary_type == "Excel (CSV)") {
           write.csv (selecteddatabase_pan_cancer_assessment_mutation_signature(),file,row.names = FALSE)
         } else if (input$pan_cancer_assessment_mutation_signature_summary_type == "Text (TSV)") {
           write.table (selecteddatabase_pan_cancer_assessment_mutation_signature(),file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_assessment_mutation_signature_summary_type == "JSON") {
           write(toJSON(selecteddatabase_pan_cancer_assessment_mutation_signature()),file)
         }
       }
     )
     
     output$pan_cancer_assessment_mutation_signature_summary_table <- DT::renderDataTable({
       selecteddatabase_pan_cancer_assessment_mutation_signature()
     })
     
     ##########
     #### mutational signature plots
     ##########
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         input_signature_pan_cancer_assessment_mutation_signature <- input$signature_pan_cancer_assessment_mutation_signature
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_signature()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Signature"]),])==0) {
           output[["pan_cancer_assessment_mutation_signature_plots_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[input$data_source_pan_cancer_assessment_mutation_signature]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_signature)) {
             local({
               # my_i <- k
               # pan_cancer_assessment_mutation_signature_plots_plots_name <- paste0("pan_cancer_assessment_mutation_signature_plots_plots", my_i)
               d<-splitted_list#[[my_i]]
               output[["pan_cancer_assessment_mutation_signature_plots_plots"]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   # if (is.na(names(splitted_list)[my_i])) {
                   #   incProgress(100, detail = "Doing part %")
                   #   plot_ly () %>% 
                   #     layout(title = input$data_source_pan_cancer_assessment_mutation_signature[[my_i]],font=list(size=15,family="Aril")) %>%
                   #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                   # } else {
                     #d <- splitted_list[[my_i]]
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>% 
                         layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     } else {
                       d[,"Groups"] <- as.character(d[,"Groups"])
                       #d <- d[!is.na(d[,"Signature"]),]
                       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),]
                       d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                       d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                       ind <- sort(unique(d[,"Groups"]))
                       try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1] & !is.na(d[,"Groups"]),"Signature"]),as.numeric(d[d[,"Groups"]==ind[2] & !is.na(d[,"Groups"]),"Signature"]))$p.value)
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         p.v <- " = NA"
                       } else {
                         # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Signature"]),as.numeric(d[d[,"Groups"]==ind[2],"Signature"]))$p.value
                          # if (is.na(p.v)) {
                          #   p.v <- " = NA"
                          # } else {
                         #   if (p.v < 0.001) {
                         #     p.v <- " < 0.001"
                         #   } else {
                             p.v<-paste0(" = ",signif(p.v,2))
                            # } 
                         # }
                       }
                       p<-plot_ly(data=d,x=~Groups,
                                  y=~Signature,
                                  type="violin",#type="box",
                                  points="all",#boxpoints="all",
                                  jitter=0.3,
                                  pointpos=-1.8,
                                  #mode="markders",
                                  color= ~PFS.status,
                                  colors=c("#7CDBD5","#DCB239","gray"),
                                  #hoveron="points+boxes",
                                  legendgroup=~PFS.status,
                                  showlegend=TRUE,
                                  text= ~paste0("Sample ID: ", PatientID,
                                                "<br>Signature: ",str_split(input$signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1]," :",Signature,
                                                "<br>Treatment: ",treatment,
                                                "<br>Description: ",description)) %>%
                         layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                                height = 480,
                                xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                           showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                yaxis=list(title=paste(str_split(input_signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1], " weight",sep=""),
                                           titlefont=list(size=20,color="black",family= "Aril"),
                                           showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                )) %>%
                         add_annotations(x=0.5,
                                         y=d[which.max(d[,"Signature"]),"Signature"],
                                         text=paste0("p-value",p.v),
                                         xref = paste0("x", 1),  # add this
                                         yref = paste0("y", 1),  # add this
                                         showarrow = FALSE,
                                         ax = 20,
                                         ay = -40,
                                         font=list(size=20,color="black",family= "Aril black"))
                       incProgress(100, detail = "Doing part %")
                       p
                     } 
                     
                   #}
                 })
               })
             })
           #}
         }
       })
     })
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         input_signature_pan_cancer_assessment_mutation_signature <- input$signature_pan_cancer_assessment_mutation_signature
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_mutation_signature()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Signature"]),])==0) {
           output[["pan_cancer_assessment_individual_mutation_signature_plots_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)]


           for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_signature)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_mutation_signature_plots_plots_name <- paste0("pan_cancer_assessment_individual_mutation_signature_plots_plots", my_i)
               d<-splitted_list[[my_i]]
               output[[pan_cancer_assessment_individual_mutation_signature_plots_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                   } else {
                     #d <- splitted_list[[my_i]]
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                     } else {
                       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),]
                       d[,"Groups"] <- as.character(d[,"Groups"])
                       d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                       d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                       ind <- sort(unique(d[,"Groups"]))
                       try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Signature"]),as.numeric(d[d[,"Groups"]==ind[2],"Signature"]))$p.value)
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         p.v <- "NA"
                       } else {
                         # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"Signature"]),as.numeric(d[d[,"Groups"]==ind[2],"Signature"]))$p.value
                         # if (is.na(p.v)) {
                         #   p.v <- " = NA"
                         # } else {
                         #   if (p.v < 0.001) {
                         #     p.v <- " < 0.001"
                         #   } else {
                         p.v<-paste0(" = ",signif(p.v,2))
                         #   }
                         # }
                       }
                       p<-plot_ly(data=d,x=~Groups,
                                  y=~Signature,
                                  type="violin",#type="box",
                                  points="all",#boxpoints="all",
                                  jitter=0.3,
                                  pointpos=-1.8,
                                  #mode="markders",
                                  color= ~PFS.status,
                                  colors=c("#7CDBD5","#DCB239","gray"),
                                  #hoveron="points+boxes",
                                  legendgroup=~PFS.status,
                                  showlegend=TRUE,
                                  text= ~paste0("Sample ID: ", PatientID,
                                                "<br>Signature: ",str_split(input$signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1]," :",Signature,
                                                "<br>Treatment: ",treatment,
                                                "<br>Description: ",description)) %>%
                         layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                height = 480,
                                xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                           showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                yaxis=list(title=paste(str_split(input_signature_pan_cancer_assessment_mutation_signature," ",2)[[1]][1], " weight",sep=""),
                                           titlefont=list(size=20,color="black",family= "Aril"),
                                           showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                )) %>%
                         add_annotations(x=0.5,
                                         y=d[which.max(d[,"Signature"]),"Signature"],
                                         text=paste0("p-value",p.v),
                                         xref = paste0("x", 1),  # add this
                                         yref = paste0("y", 1),  # add this
                                         showarrow = FALSE,
                                         ax = 20,
                                         ay = -40,
                                         font=list(size=20,color="black",family= "Aril black"))
                       incProgress(100, detail = "Doing part %")
                       p
                     }

                   }
                 })
               })
             })
           }
         }
       })
     })
     
     #############
     #### mutational signature survival
     #############
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         input_signature_pan_cancer_assessment_mutation_signature <- input$signature_pan_cancer_assessment_mutation_signature
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_signature()
         input_adjust_OS_pan_cancer_assessment_mutation_signature <- input$adjust_OS_pan_cancer_assessment_mutation_signature
         
         #    selected.database[selected.database[,"OS.status"]=="living" & !is.na(selected.database[,"OS.status"]),"dead"] <- 0
         #    selected.database[selected.database[,"OS.status"]=="deceased" & !is.na(selected.database[,"OS.status"]),"dead"] <- 1
         #    selected.database <- selected.database[!is.na(selected.database[,"Signature"]) & !is.na(selected.database[,"dead"]),]
         #    selected.database <- selected.database[selected.database[,"dead"]==1 | selected.database[,"dead"]==0,]
         
         splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
         splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)]
         
         
         local({
           # my_i <- k
           # pan_cancer_assessment_mutation_signature_survival_plots_name <- paste0("pan_cancer_assessment_mutation_signature_survival_plots", my_i)
           # pan_cancer_assessment_mutation_signature_survival_summary_tables_name <- paste0("pan_cancer_assessment_mutation_signature_survival_summary_tables", my_i)
           d<-splitted_list#[[my_i]]
           d[,"Signature"] <- as.numeric(as.character(d[,"Signature"]))
           #output[[pan_cancer_assessment_mutation_signature_survival_plots_name]] <- renderPlotly({
           withProgress(message = 'Making plot', value = 0, {
             isolate({
               
               
               if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),])==0) {
                 incProgress(100, detail = "Doing part %")
                 p1 <- plot_ly () %>%
                   layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                   add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                 table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
               } else {
                 d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),]
                 d<-data.frame(d,dead=NA)
                 d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                 d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                 d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"dead"]),]
                 d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                 
                 d <- data.frame(data_source=as.character(d$data_source),
                                 dead=as.numeric(d$dead),
                                 OS=as.numeric(d$OS),
                                 Age=as.character(d$Age),
                                 Gender=as.character(d$Gender),
                                 Signature=as.numeric(d[,"Signature"]))
                 #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Signature")]))
                 d<-data.frame(d,IND=paste0("bottom ",
                                            input$slider_pan_cancer_assessment_mutation_signature_OS*100,
                                            "% (",
                                            nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),]),
                                            ")"))
                 d[,"IND"]<-as.character(d[,"IND"])
                 d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),"IND"]<-
                   paste0("top ",
                          (1-input$slider_pan_cancer_assessment_mutation_signature_OS)*100,
                          "% (",
                          nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),])
                          ,")")
                 d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                 d<-d[order(d[,"IND"]),]
                 fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                 
                 try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                 if (class(try.test)=="try-error" | is.na(try.test)) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                  plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                 
                  if (input_adjust_OS_pan_cancer_assessment_mutation_signature=="None") {
                    diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                    if (class(diff.result.try)=="try-error") {
                     p.value <- NA
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     p.value <- signif(summary(diff.result)$logtest[3],2)
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                     
                   }
                 } else {
                   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_signature),data=d) )
                   if (class(diff.result.adjust.try)=="try-error") {
                     diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                     p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                       " (No ",input_adjust_OS_pan_cancer_assessment_mutation_signature," was observed)")
                     
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   } else {
                     p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                     table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                     rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_signature
                     table1 <- datatable(table1.pre,
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result.adjust)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result.adjust)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   }
                 }
                 
                 # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                 # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                 plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                   ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                            #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                            paste0("p-value = ",p.value),x=300,y=0.05)) +
                   ggplot2::guides(color=FALSE,linetype=FALSE)
                 p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                   layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                          xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                          yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                          legend=list(orientation="v",x=0.6,y=1)) #%>%
                 incProgress(100, detail = "Doing part %")
                 p1
                 }
               }
               
             })
           })
           #})
           output[["pan_cancer_assessment_mutation_signature_survival_plots"]] <- renderPlotly({
             p1
           })
           output[["pan_cancer_assessment_mutation_signature_survival_summary_tables"]] <- renderDataTable(
             table1
           )
         })
         
         
       })
     })
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         input_signature_pan_cancer_assessment_mutation_signature <- input$signature_pan_cancer_assessment_mutation_signature
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_mutation_signature()
         input_adjust_OS_pan_cancer_assessment_mutation_signature <- input$adjust_OS_pan_cancer_assessment_mutation_signature
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
           output[["pan_cancer_assessment_individual_mutation_signature_survival_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
         } else {

           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)]


           for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_signature)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_mutation_signature_survival_plots_name <- paste0("pan_cancer_assessment_individual_mutation_signature_survival_plots", my_i)
               pan_cancer_assessment_individual_mutation_signature_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_mutation_signature_survival_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               d[,"Signature"] <- as.numeric(as.character(d[,"Signature"]))
               #output[[pan_cancer_assessment_mutation_signature_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {

                     if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     } else {
                       d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"Signature"]),]
                       d<-data.frame(d,dead=NA)
                       d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                       d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                       d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"dead"]),]
                       d <- d[d[,"dead"]==1 | d[,"dead"]==0,]

                       d <- data.frame(data_source=as.character(d$data_source),
                                       dead=as.numeric(d$dead),
                                       OS=as.numeric(d$OS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       Signature=as.numeric(d[,"Signature"]))
                       #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Signature")]))
                       d<-data.frame(d,IND=paste0("bottom ",
                                                  input$slider_pan_cancer_assessment_mutation_signature_OS*100,
                                                  "% (",
                                                  nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),]),
                                                  ")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),"IND"]<-
                         paste0("top ",
                                (1-input$slider_pan_cancer_assessment_mutation_signature_OS)*100,
                                "% (",
                                nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),])
                                ,")")
                       d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                       plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                       if (input_adjust_OS_pan_cancer_assessment_mutation_signature=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_signature),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_mutation_signature," was observed)")

                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_signature
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }

                       # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                         layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                legend=list(orientation="v",x=0.6,y=1)) #%>%
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_mutation_signature_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_mutation_signature_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     #############
     #### mutational signature PFS
     #############
     
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_mutation_signature()
         input_adjust_OS_pan_cancer_assessment_mutation_signature <- input$adjust_OS_pan_cancer_assessment_mutation_signature
         
         splitted_list <- selected.database #%>% split(.$data_source)#%>% as.matrix %>% as.data.frame %>% split(.$data_source)
         splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)]
         
         
         
         
         local({
           # my_i <- k
           # pan_cancer_assessment_mutation_signature_PFS_plots_name <- paste0("pan_cancer_assessment_mutation_signature_PFS_plots", my_i)
           # pan_cancer_assessment_mutation_signature_PFS_summary_tables_name <- paste0("pan_cancer_assessment_mutation_signature_PFS_summary_tables", my_i)
           d<-splitted_list#[[my_i]]
           
           #output[[pan_cancer_assessment_mutation_signature_PFS_plots_name]] <- renderPlotly({
           withProgress(message = 'Making plot', value = 0, {
             isolate({
               
               if (nrow(d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                 incProgress(100, detail = "Doing part %")
                 p1 <- plot_ly () %>% 
                   layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                   add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 
               } else {
                 d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                 
                 
                 d <- data.frame(data_source=as.character(d$data_source),
                                 PFS.status=as.character(d$PFS.status),
                                 PFS=as.numeric(d$PFS),
                                 Age=as.character(d$Age),
                                 Gender=as.character(d$Gender),
                                 Signature=as.numeric(d[,"Signature"]))
                 # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Signature")]))
                 d<-data.frame(d,IND=paste0("bottom ",input$slider_pan_cancer_assessment_mutation_signature_OS*100, "% (",
                                            nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),]),
                                            ")"))
                 d[,"IND"]<-as.character(d[,"IND"])
                 d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),"IND"]<- 
                   paste0("top ",(1-input$slider_pan_cancer_assessment_mutation_signature_OS)*100,"% (",
                          nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),])
                          ,")")
                 d <- data.frame(d,group.IND=0)
                 d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                 #        d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                 d<-d[order(d[,"IND"]),]
                 fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                 
                 try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                 if (class(try.test)=="try-error" | is.na(try.test)) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                  plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                 
                  if (input_adjust_OS_pan_cancer_assessment_mutation_signature=="None") {
                   diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                   if (class(diff.result.try)=="try-error") {
                     p.value <- NA
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                   } else {
                     p.value <- signif(summary(diff.result)$logtest[3],2)
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                     
                   }
                 } else {
                   diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_signature),data=d) )
                   if (class(diff.result.adjust.try)=="try-error") {
                     diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                     p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                       " (No ",input_adjust_OS_pan_cancer_assessment_mutation_signature," was observed)")
                     
                     table1 <- datatable(signif(coef(summary(diff.result)),2),
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   } else {
                     p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                     table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                     rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_signature
                     table1 <- datatable(table1.pre,
                                         caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                " on ",
                                                                                signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                " df, p-value = ",
                                                                                signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                         br(),
                                                                         paste0("n = ",
                                                                                signif(summary(diff.result.adjust)$n,2),
                                                                                ", number of events = ",
                                                                                signif(summary(diff.result.adjust)$nevent,2)),
                                                                         style="text-align:left;color:black;font-size:15px"),
                                         options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                   }
                 }
                 
                 
                 
                 # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                 # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                 plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                   ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                            #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))#,size=5)
                                            paste0("p-value = ",p.value),x=300,y=0.05)) +
                   ggplot2::guides(color=FALSE,linetype=FALSE)
                 p1<-ggplotly(plot.result.ann, height = 450  ) %>%
                   layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                          xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                          yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                          legend=list(orientation="v",x=0.6,y=1)) 
                 incProgress(100, detail = "Doing part %")
                 p1
                 }
               }
               
             })
           })
           #})
           output[["pan_cancer_assessment_mutation_signature_PFS_plots"]] <- renderPlotly({
             p1
           })
           output[["pan_cancer_assessment_mutation_signature_PFS_summary_tables"]] <- renderDataTable(
             table1
           )
         })
         
         
       })
     })
     
     observeEvent(input$do_pan_cancer_assessment_mutation_signature,{
       input$do_pan_cancer_assessment_mutation_signature
       isolate({
         input_signature_pan_cancer_assessment_mutation_signature <- input$signature_pan_cancer_assessment_mutation_signature
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_mutation_signature()
         input_adjust_OS_pan_cancer_assessment_mutation_signature <- input$adjust_OS_pan_cancer_assessment_mutation_signature
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"Signature"]),])==0) {
           output[["pan_cancer_assessment_individual_mutation_signature_PFS_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)]



           for (k in 1:length(input$data_source_pan_cancer_assessment_mutation_signature)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_mutation_signature_PFS_plots_name <- paste0("pan_cancer_assessment_individual_mutation_signature_PFS_plots", my_i)
               pan_cancer_assessment_individual_mutation_signature_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_mutation_signature_PFS_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               d[,"Signature"] <- as.numeric(as.character(d[,"Signature"]))
               #output[[pan_cancer_assessment_mutation_signature_PFS_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_mutation_signature,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                   } else {

                     if (nrow(d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)

                     } else {
                       d <- d[!is.na(d[,"Signature"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                       d <- data.frame(data_source=as.character(d$data_source),
                                       PFS.status=as.character(d$PFS.status),
                                       PFS=as.numeric(d$PFS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       Signature=as.numeric(d[,"Signature"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","Signature")]))
                       d<-data.frame(d,IND=paste0("bottom ",
                                                  input$slider_pan_cancer_assessment_mutation_signature_OS*100,
                                                  "% (",
                                                  nrow(d[d[,"Signature"]<= quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),]),
                                                  ")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),"IND"]<-
                         paste0("top ",
                                (1-input$slider_pan_cancer_assessment_mutation_signature_OS)*100,
                                "% (",
                                nrow(d[d[,"Signature"]> quantile(d[,"Signature"],input$slider_pan_cancer_assessment_mutation_signature_OS),])
                                ,")")
                       d <- data.frame(d,group.IND=0)
                       d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1

                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                        plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))

                        if (input_adjust_OS_pan_cancer_assessment_mutation_signature=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)

                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_mutation_signature),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_mutation_signature," was observed)")

                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_mutation_signature
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann,height = 480 ) %>%
                         layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                legend=list(orientation="v",x=0.6,y=1)) #%>%
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_mutation_signature_PFS_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_mutation_signature_PFS_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     
     #####################################################
     #### pan-cancer assessment expression
     #####################################################
     observe({
       updateSelectizeInput(session,"gene_pan_cancer_assessment_expression",
                            choices=a.colnames,selected="CD19",
                            server=TRUE)
     })
     observe({
       if (input$dynamic_check_pan_cancer_assessment_expression) {
         
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression",
                              choices=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                              selected=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                              server=T)
         
       }
     })
     
     observe({
       if (input$data_source_pan_cancer_assessment_expression=="" || is.null(input$data_source_pan_cancer_assessment_expression) ||
           input$treatment_pan_cancer_assessment_expression=="" || is.null(input$treatment_pan_cancer_assessment_expression) ||
           input$description_pan_cancer_assessment_expression=="" || is.null(input$description_pan_cancer_assessment_expression) ||
           input$gene_pan_cancer_assessment_expression=="" || is.null(input$gene_pan_cancer_assessment_expression)){
         disable("do_pan_cancer_assessment_expression")
       } else {
         enable("do_pan_cancer_assessment_expression")
       }
     })
     #     observeEvent(input$do_pan_cancer_assessment_expression,{
     #       if (input$gene_pan_cancer_assessment_expression=="ALL") {
     #         hideTab(inputId="pan_cancer_assessment_expression_tabs",target="Immunotherapy response")
     #         hideTab(inputId="pan_cancer_assessment_expression_tabs",target="Overall survival")
     #         hideTab(inputId="pan_cancer_assessment_expression_tabs",target="Progression free survival")
     #         hideTab(inputId="pan_cancer_assessment_expression_tabs",target="Evaluation")
     #         shinyjs::hideElement(id="pan_cancer_assessment_expression_clinical_attribute")
     # #        shinyjs::hideElement(id="clinical_pan_cancer_assessment_expression")
     # #        tabsetPanel(tabPanel("Summary",h1("test")))
     #       } else {
     #         showTab(inputId="pan_cancer_assessment_expression_tabs",target="Immunotherapy response")
     #         showTab(inputId="pan_cancer_assessment_expression_tabs",target="Overall survival")
     #         showTab(inputId="pan_cancer_assessment_expression_tabs",target="Progression free survival")
     #         showTab(inputId="pan_cancer_assessment_expression_tabs",target="Evaluation")
     #         shinyjs::showElement(id="pan_cancer_assessment_expression_clinical_attribute")
     #       }
     #     })
     
     selecteddatabase_pan_cancer_assessment_expression <- eventReactive(input$do_pan_cancer_assessment_expression,{
       if (input$dynamic_check_pan_cancer_assessment_expression) {
         user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
         
         #data_source1 <- input$data_source_pan_cancer_assessment_expression[input$data_source_pan_cancer_assessment_expression %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
         #treatment1 <- input$treatment_pan_cancer_assessment_expression
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression])
         description1 <- c("Pre","On")
         a<-data.table(database.test.batchcorrected)
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_pan_cancer_assessment_expression)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "UserGene"
         m <- as.data.frame(m)
         if (nrow(m) >0) {
           samples <- m$PatientID
           samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
           samples_paired <- table(samples)
           samples_paired <- samples_paired[samples_paired==2]
           samples_paired_pre <- paste0(names(samples_paired),"Pre")
           #samples_paired_on <- paste0(names(samples_paired),"On")
           samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
           m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],UserGene=m[m[,"PatientID"] %in% samples_paired_on,"UserGene"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserGene"]),
                      data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],UserGene=m[m[,"PatientID"] %in% samples_paired_on,"UserGene"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserGene"]))
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           d <- m
           # m1 <- m %>% split(.$data_source) %>% 
           #   lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]>=quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% ",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]<quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% ",sep="")
           # d})
           # m2<-ldply(m1,data.frame)
           # m2<-m2[,-1]
           # m2$UserGene <- as.numeric(m2$UserGene)
           # m2
           d$UserGene <- as.numeric(d$UserGene)
           d
         } else {
           m
         }
         
         
         
       } else {
         
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)
         #treatment1 <- input$treatment_pan_cancer_assessment_expression
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression])
         description1 <- input$description_pan_cancer_assessment_expression
         a<-data.table(database.test.batchcorrected)
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_pan_cancer_assessment_expression)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "UserGene"
         m <- as.data.frame(m)
         if (nrow(m)>0) {
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           d <- m
           # m1 <- m %>% split(.$data_source) %>% 
           #   lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]>=quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% ",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]<quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% ",sep="")
           # d})
           # m2<-ldply(m1,data.frame)
           # m2<-m2[,-1]
           # m2$UserGene <- as.numeric(m2$UserGene)
           # m2
           d$UserGene <- as.numeric(d$UserGene)
           d
         } else {
           m2<-m
           m2
         }
       }
     })
     
     selecteddatabase_pan_cancer_assessment_individual_expression <- eventReactive(input$do_pan_cancer_assessment_expression,{
       if (input$dynamic_check_pan_cancer_assessment_expression) {
         user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
         
         #data_source1 <- input$data_source_pan_cancer_assessment_expression[input$data_source_pan_cancer_assessment_expression %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
         #treatment1 <- input$treatment_pan_cancer_assessment_expression
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression])
         description1 <- c("Pre","On")
         a<-data.table(database.test.process())
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_pan_cancer_assessment_expression)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "UserGene"
         m <- as.data.frame(m)
         if (nrow(m) >0) {
           samples <- m$PatientID
           samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
           samples_paired <- table(samples)
           samples_paired <- samples_paired[samples_paired==2]
           samples_paired_pre <- paste0(names(samples_paired),"Pre")
           #samples_paired_on <- paste0(names(samples_paired),"On")
           samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
           m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],UserGene=m[m[,"PatientID"] %in% samples_paired_on,"UserGene"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserGene"]),
                      data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],UserGene=m[m[,"PatientID"] %in% samples_paired_on,"UserGene"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserGene"]))
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           #d <- m
           m1 <- m %>% split(.$data_source) %>%
             lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]>=quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% ",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]<quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% ",sep="")
           d})
           m2<-ldply(m1,data.frame)
           m2<-m2[,-1]
           m2$UserGene <- as.numeric(m2$UserGene)
           m2
           # d$UserGene <- as.numeric(d$UserGene)
           # d
         } else {
           m
         }
         
         
         
       } else {
         
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)
         #treatment1 <- input$treatment_pan_cancer_assessment_expression
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression])
         description1 <- input$description_pan_cancer_assessment_expression
         a<-data.table(database.test.process())
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_pan_cancer_assessment_expression)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "UserGene"
         m <- as.data.frame(m)
         if (nrow(m)>0) {
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           #d <- m
           m1 <- m %>% split(.$data_source) %>%
             lapply(function(d) {#d[d[,"UserGene"]>=median(na.omit(d[,"UserGene"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]>=quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% ",sep="")
           d[!is.na(d[,"UserGene"]) & d[,"UserGene"]<quantile(na.omit(d[,"UserGene"]),input$slider_pan_cancer_assessment_expression_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% ",sep="")
           d})
           m2<-ldply(m1,data.frame)
           m2<-m2[,-1]
           m2$UserGene <- as.numeric(m2$UserGene)
           m2
           # d$UserGene <- as.numeric(d$UserGene)
           # d
         } else {
           m2<-m
           m2
         }
       }
     })
     
     # selecteddatabase_pan_cancer_assessment_expression_ALL <- eventReactive(input$do_pan_cancer_assessment_expression,{
     #   data_source1 <- input$data_source_pan_cancer_assessment_expression
     #   treatment1 <- input$treatment_pan_cancer_assessment_expression
     #   description1 <- input$description_pan_cancer_assessment_expression
     #   a<-data.table(database.test.process())
     #   m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
     #   if (input$gene_pan_cancer_assessment_expression=="ALL") {
     #     m<-as.data.frame(m)
     #    
     #   m.data_source <- as.character(unique(m[,"data_source"]))
     #   m.result <- NULL
     #   
     #   for (k in 1:length(m.data_source)) {
     #     d <- m[m[,"data_source"]==m.data_source[k],]
     #     d.res <- d[d[,"PFS.status"]=="response",]
     #     d.nonres <- d[d[,"PFS.status"]=="nonresponse",]
     #     d.p.value <- NULL
     #     for (j in 13:ncol(d)) {
     #       try.test <- try(p.v <- wilcox.test(as.numeric(na.omit(d.res[,j])),as.numeric(na.omit(d.nonres[,j])))$p.value)
     #       if (class(try.test)=="try-error") {
     #         p.v <- "NA"
     #       } else {
     #         p.v <- round(wilcox.test(as.numeric(na.omit(d.res[,j])),as.numeric(na.omit(d.nonres[,j])))$p.value,3)
     #       }
     #       d.p.value <- c(d.p.value,p.v)
     #     }
     #     m.result <- rbind(m.result,d.p.value)
     #   }
     #   colnames(m.result) <- colnames(m)[13:ncol(m)]
     #   m.result <- data.frame(data_source=m.data_source,m.result)
     #   m.result
     #   }
     # })
     
     ################
     #### pan_cancer_assessment_expression summary
     ################
     output$pan_cancer_assessment_expression_summary_disease_number <- renderText({
       input$do_pan_cancer_assessment_expression ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_assessment_expression()$Disease))})})
     output$pan_cancer_assessment_expression_summary_sample_number <- renderText({
       input$do_pan_cancer_assessment_expression ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_assessment_expression())})})
     output$pan_cancer_assessment_expression_summary_plots <- renderPlotly({
       summary_plot(inputdata=selecteddatabase_pan_cancer_assessment_expression(),feature="expression")
     })
     
     
     
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         # if (input_gene_pan_cancer_assessment_expression=="ALL") { 
         #   selected.database <- selecteddatabase_pan_cancer_assessment_expression_ALL()
         # } else {
         selected.database <- selecteddatabase_pan_cancer_assessment_expression()
         # }
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_expression_summary_clincial_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           
           
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_summary_clinical_plots_name <- paste0("pan_cancer_assessment_expression_summary_clinical_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_expression_summary_clinical_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 if (nrow(d[!is.na(d[,clinical_server[input$clinical_pan_cancer_assessment_expression]]) & !is.na(d[,"UserGene"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   d <- d[!is.na(d[,clinical_server[input$clinical_pan_cancer_assessment_expression]]) & !is.na(d[,"UserGene"]),]
                   if (input$dynamic_check_pan_cancer_assessment_expression) {
                     d <- subset(d,description=="Pre")
                   }
                   if(input$clinical_pan_cancer_assessment_expression=="Age") {
                     d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                     #d[d[,"Age"]<=10,"Age"] <- "0-10"
                     d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                     d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                   }
                   
                   if (input$dynamic_check_pan_cancer_assessment_expression) {
                     p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_pan_cancer_assessment_expression]],
                                y=~signif(as.numeric(as.character(UserGene)),3),  
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=d[,clinical_server[input$clinical_pan_cancer_assessment_expression]],showlegend=TRUE,
                                colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                hoveron="points+boxes",
                                legendgroup=d[,clinical_server[input$clinical_pan_cancer_assessment_expression]], 
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Changes of expression: ",input$gene_pan_cancer_assessment_expression," :",signif(as.numeric(as.character(UserGene)),3), 
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title="Clinical Attribute",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste("Changes of ",input_gene_pan_cancer_assessment_expression, " expression value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              ))
                     
                   } else {
                     p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_pan_cancer_assessment_expression]],
                                y=~signif(as.numeric(as.character(UserGene)),3), 
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=d[,clinical_server[input$clinical_pan_cancer_assessment_expression]],showlegend=TRUE,
                                colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                hoveron="points+boxes",
                                legendgroup=d[,clinical_server[input$clinical_pan_cancer_assessment_expression]], 
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>expression: ",input$gene_pan_cancer_assessment_expression," :",signif(as.numeric(as.character(UserGene)),3), 
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title="Clinical Attribute",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste(input_gene_pan_cancer_assessment_expression, " expression value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              ))
                   }
                   incProgress(100, detail = "Doing part %")
                   p
                   
                 } 
                 #}
               })
             })
           })
           #}
           # }
         }
       }) 
     })
     
     output$pan_cancer_assessment_expression_summary_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_assessment_expression_summary_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_assessment_expression_summary_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_assessment_expression_summary_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         # if (input$gene_pan_cancer_assessment_expression=="ALL") {
         #   
         #   d <- t(selecteddatabase_pan_cancer_assessment_expression_ALL())
         #   colnames(d) <- d[1,]
         #   #colnames(d)[1] <- "Genes"
         #   d <- d[-1,]
         #   
         #   if (input$pan_cancer_assessment_expression_summary_type == "Excel (CSV)") {
         #     write.csv (d,file,row.names = FALSE)
         #   } else if (input$pan_cancer_assessment_expression_summary_type == "Text (TSV)") {
         #     write.table (d,file,quote=F, row.names = FALSE, sep="\t")
         #   } else if (input$pan_cancer_assessment_expression_summary_type == "JSON") {
         #     write(toJSON(d),file)
         #   }
         # } else {
         selected.database <- selecteddatabase_pan_cancer_assessment_expression()
         if (input$dynamic_check_pan_cancer_assessment_expression) {
           colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- paste0("Changes of ",input$gene_pan_cancer_assessment_expression)
         } else {
           colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- input$gene_pan_cancer_assessment_expression
         }
         if (input$pan_cancer_assessment_expression_summary_type == "Excel (CSV)") {
           write.csv (selected.database,file,row.names = FALSE)
         } else if (input$pan_cancer_assessment_expression_summary_type == "Text (TSV)") {
           write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_assessment_expression_summary_type == "JSON") {
           write(toJSON(selected.database),file)
         }
         #}
       }
     )
     
     output$pan_cancer_assessment_expression_summary_table <- DT::renderDataTable({
       # if (input$gene_pan_cancer_assessment_expression=="ALL") {
       #   d <- t(selecteddatabase_pan_cancer_assessment_expression_ALL())
       #   colnames(d) <- d[1,]
       #   #colnames(d)[1] <- "Genes"
       #   d <- d[-1,]
       #   d
       # } else {
       selected.database <- selecteddatabase_pan_cancer_assessment_expression()
       if (input$dynamic_check_pan_cancer_assessment_expression) {
         colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- paste0("Changes of ",input$gene_pan_cancer_assessment_expression)
       } else {
         colnames(selected.database)[which(colnames(selected.database)=="UserGene")] <- input$gene_pan_cancer_assessment_expression
       }
       
       selected.database
       #}
     })
     
     ################
     #### pan_cancer_assessment_expression plots
     ################
     
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         selected.database <- selecteddatabase_pan_cancer_assessment_expression()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"UserGene"]),])==0) {
           
           output[["pan_cancer_assessment_expression_plots_plots"]] <- renderPlotly({
             #show("loading-c")
             #hide("loading-d")
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_plots_plots_name <- paste0("pan_cancer_assessment_expression_plots_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_expression_plots_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 #show("loading-c")
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   #hide("loading-d")
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),])==0 ) {
                   #hide("loading-d")
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                 } else {
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),]
                   if (input$dynamic_check_pan_cancer_assessment_expression) {
                     d <- subset(d,description=="Pre")
                   }
                   
                   d[,"Groups"] <- as.character(d[,"Groups"])
                   d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                   d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                   ind <- sort(unique(d[,"Groups"]))
                   # if (as.character(unique(d$data_source)) == "Melanoma-Van Allen" | as.character(unique(d$data_source)) == "Melanoma-Riaz" | as.character(unique(d$data_source)) == "Melanoma-Auslander") {
                   #   lable_y <- " (RSEM)"
                   # } else if (as.character(unique(d$data_source)) == "Melanoma-Chen" | as.character(unique(d$data_source)) == "Melanoma-Prat") {
                   #   lable_y <- " (NanoString signal)"
                   # } else if (as.character(unique(d$data_source)) == "Melanoma-Hugo" | as.character(unique(d$data_source)) == "Melanoma-Gide" | as.character(unique(d$data_source)) == "Melanoma-Snyder-Nathanson") {
                   #   lable_y <- " (FPKM)"
                   # } else {
                   #   lable_y <- NULL
                   # }
                   
                   
                   try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserGene"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserGene"])))$p.value)
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     p.v <- " = NA"
                   } else {
                     # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserGene"]),as.numeric(d[d[,"Groups"]==ind[2],"UserGene"]))$p.value
                     # if (p.v < 0.001) {
                     #   p.v <- "< 0.001"
                     # } else {
                     p.v<-paste0(" = ",signif(p.v,2))
                     # }
                   }
                   
                   if (input$dynamic_check_pan_cancer_assessment_expression) {
                     p<-plot_ly(data=d,x=~Groups,
                                y=~signif(as.numeric(as.character(UserGene)),3),
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=~PFS.status,showlegend=TRUE,
                                colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                #hoveron="points+boxes",
                                legendgroup=~PFS.status,
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Change of ",input_gene_pan_cancer_assessment_expression," expression:",signif(as.numeric(as.character(UserGene)),3),
                                              "<br>Treatment: ",treatment))  %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title=" ",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste("Change of ",input_gene_pan_cancer_assessment_expression, " expression value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              )) %>%
                       add_annotations(x = 0.5,
                                       y = d[which.max(d[,"UserGene"]),"UserGene"],
                                       text = paste0("p",p.v),
                                       xref = paste0("x", 1),  # add this
                                       yref = paste0("y", 1),  # add this
                                       showarrow = FALSE,
                                       ax = 20,
                                       ay = -40,
                                       font=list(size=20,color="black",family= "Aril black"))
                   } else {
                     p<-plot_ly(data=d,x=~Groups,
                                y=~signif(as.numeric(as.character(UserGene)),3),
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=~PFS.status,showlegend=TRUE,
                                colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                #hoveron="points+boxes",
                                legendgroup=~PFS.status,
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Expression: ",input_gene_pan_cancer_assessment_expression," :",signif(as.numeric(as.character(UserGene)),3),
                                              "<br>Treatment: ",treatment))  %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title=" ",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste(input_gene_pan_cancer_assessment_expression, " expression value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              )) %>%
                       add_annotations(x = 0.5,
                                       y = d[which.max(d[,"UserGene"]),"UserGene"],
                                       text = paste0("p",p.v),
                                       xref = paste0("x", 1),  # add this
                                       yref = paste0("y", 1),  # add this
                                       showarrow = FALSE,
                                       ax = 20,
                                       ay = -40,
                                       font=list(size=20,color="black",family= "Aril black"))
                   }
                   #hide("loading-d")
                   incProgress(100, detail = "Doing part %")
                   p
                 }
                 
                 #}
               })
             })
           })
           #}
         }
       })
       
     })
     
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression()
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_individual_expression_plots_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_plots_plots_name <- paste0("pan_cancer_assessment_individual_expression_plots_plots", my_i)
               d<-splitted_list[[my_i]]
               
               output[[pan_cancer_assessment_individual_expression_plots_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   #show("loading-c")
                   if (is.na(names(splitted_list)[my_i])) {
                     #hide("loading-d")
                     incProgress(100, detail = "Doing part %")
                     plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                   } else {
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),])==0 ) {
                       #hide("loading-d")
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                     } else {
                       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserGene"]),]
                       if (input$dynamic_check_pan_cancer_assessment_expression) {
                         d <- subset(d,description=="Pre")
                       }
                       
                       d[,"Groups"] <- as.character(d[,"Groups"])
                       d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                       d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                       ind <- sort(unique(d[,"Groups"]))
                       if (as.character(unique(d$data_source)) == "Melanoma-Van Allen" | as.character(unique(d$data_source)) == "Melanoma-Riaz" | as.character(unique(d$data_source)) == "Melanoma-Auslander") {
                         lable_y <- " (RSEM)"
                       } else if (as.character(unique(d$data_source)) == "Melanoma-Chen" | as.character(unique(d$data_source)) == "Melanoma-Prat") {
                         lable_y <- " (NanoString signal)"
                       } else if (as.character(unique(d$data_source)) == "Melanoma-Hugo" | as.character(unique(d$data_source)) == "Melanoma-Gide" | as.character(unique(d$data_source)) == "Melanoma-Snyder-Nathanson") {
                         lable_y <- " (FPKM)"
                       } else {
                         lable_y <- NULL
                       }
                       
                       
                       try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserGene"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserGene"])))$p.value)
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         p.v <- " = NA"
                       } else {
                         # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserGene"]),as.numeric(d[d[,"Groups"]==ind[2],"UserGene"]))$p.value
                         # if (p.v < 0.001) {
                         #   p.v <- "< 0.001"
                         # } else {
                         p.v<-paste0(" = ",signif(p.v,2))
                         # }
                       }
                       
                       if (input$dynamic_check_pan_cancer_assessment_expression) {
                         p<-plot_ly(data=d,x=~Groups,
                                    y=~signif(as.numeric(as.character(UserGene)),3),
                                    type="violin",#type="box",
                                    points="all",#boxpoints="all",
                                    jitter=0.3,
                                    pointpos=-1.8,
                                    #mode="markders",
                                    color=~PFS.status,showlegend=TRUE,
                                    colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                    #hoveron="points+boxes",
                                    legendgroup=~PFS.status,
                                    text= ~paste0("Sample ID: ", PatientID,
                                                  "<br>Change of ",input_gene_pan_cancer_assessment_expression," expression:",signif(as.numeric(as.character(UserGene)),3),
                                                  "<br>Treatment: ",treatment))  %>%
                           layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                  height = 480,
                                  xaxis=list(title=" ",
                                             titlefont=list(size=20,color="black",family= "Aril black"),
                                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                  yaxis=list(title=paste("Change of ",input_gene_pan_cancer_assessment_expression, " expression value",lable_y,sep=""),
                                             titlefont=list(size=20,color="black",family= "Aril black"),
                                             showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                  )) %>%
                           add_annotations(x = 0.5,
                                           y = d[which.max(d[,"UserGene"]),"UserGene"],
                                           text = paste0("p",p.v),
                                           xref = paste0("x", 1),  # add this
                                           yref = paste0("y", 1),  # add this
                                           showarrow = FALSE,
                                           ax = 20,
                                           ay = -40,
                                           font=list(size=20,color="black",family= "Aril black"))
                       } else {
                         p<-plot_ly(data=d,x=~Groups,
                                    y=~signif(as.numeric(as.character(UserGene)),3),
                                    type="violin",#type="box",
                                    points="all",#boxpoints="all",
                                    jitter=0.3,
                                    pointpos=-1.8,
                                    #mode="markders",
                                    color=~PFS.status,showlegend=TRUE,
                                    colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                    #hoveron="points+boxes",
                                    legendgroup=~PFS.status,
                                    text= ~paste0("Sample ID: ", PatientID,
                                                  "<br>expression: ",input_gene_pan_cancer_assessment_expression," :",signif(as.numeric(as.character(UserGene)),3),
                                                  "<br>Treatment: ",treatment))  %>%
                           layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                  height = 480,
                                  xaxis=list(title=" ",
                                             titlefont=list(size=20,color="black",family= "Aril black"),
                                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                  yaxis=list(title=paste(input_gene_pan_cancer_assessment_expression, " expression value",lable_y,sep=""),
                                             titlefont=list(size=20,color="black",family= "Aril black"),
                                             showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                  )) %>%
                           add_annotations(x = 0.5,
                                           y = d[which.max(d[,"UserGene"]),"UserGene"],
                                           text = paste0("p",p.v),
                                           xref = paste0("x", 1),  # add this
                                           yref = paste0("y", 1),  # add this
                                           showarrow = FALSE,
                                           ax = 20,
                                           ay = -40,
                                           font=list(size=20,color="black",family= "Aril black"))
                       }
                       
                       incProgress(100, detail = "Doing part %")
                       p
                     }
                     
                   }
                 })
               })
             })
           }
         }
       })
       
     })
     
     
     ################
     #### pan_cancer_assessment_expression survival
     ################
     
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       #if (input$gene_pan_cancer_assessment_expression!="ALL") {
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         selected.database <- selecteddatabase_pan_cancer_assessment_expression()
         input_adjust_OS_pan_cancer_assessment_expression <- input$adjust_OS_pan_cancer_assessment_expression
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_expression_survival_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
           
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_survival_plots_name <- paste0("pan_cancer_assessment_expression_survival_plots", my_i)
             # pan_cancer_assessment_expression_survival_summary_tables_name <- paste0("pan_cancer_assessment_expression_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_expression_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (input$dynamic_check_pan_cancer_assessment_expression) {
                     d <- subset(d,description=="Pre")
                   }
                   d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),]
                   
                   d<-data.frame(d,dead=NA)
                   d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                   d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                   d <- d[!is.na(d[,"UserGene"]) & !is.na(d[,"dead"]),]
                   d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                   #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserGene")]))
                   d <- data.frame(data_source=as.character(d$data_source),
                                   dead=as.numeric(d$dead),
                                   OS=as.numeric(d$OS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   UserGene=as.numeric(d[,"UserGene"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserGene")]))
                   d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% (",
                                             nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                                             sep=""))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% ",
                   #                           input_gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression (",
                   #                           nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),"IND"]<-
                     paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% (",
                           nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                           sep="")
                   # d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% ", input_gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression (",
                   #         nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                   #         sep="")
                   d<-d[order(d[,"IND"]),]
                   d[,"dead"] <- as.numeric(d[,"dead"])
                   
                   fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_expression=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_expression),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_expression," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_expression_survival_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_expression_survival_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
       #}
     })
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression()
         input_adjust_OS_pan_cancer_assessment_expression <- input$adjust_OS_pan_cancer_assessment_expression
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           output[["pan_cancer_assessment_individual_expression_survival_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_survival_plots_name <- paste0("pan_cancer_assessment_individual_expression_survival_plots", my_i)
               pan_cancer_assessment_individual_expression_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_expression_survival_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_expression_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     } else {
                       if (input$dynamic_check_pan_cancer_assessment_expression) {
                         d <- subset(d,description=="Pre")
                       }
                       d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserGene"]),]
                       
                       d<-data.frame(d,dead=NA)
                       d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                       d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                       d <- d[!is.na(d[,"UserGene"]) & !is.na(d[,"dead"]),]
                       d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                       #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserGene")]))
                       d <- data.frame(data_source=as.character(d$data_source),
                                       dead=as.numeric(d$dead),
                                       OS=as.numeric(d$OS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       UserGene=as.numeric(d[,"UserGene"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserGene")]))
                       d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% (",
                                                 nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                                                 sep=""))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),"IND"]<-
                         paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% (",
                               nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                               sep="")
                       d<-d[order(d[,"IND"]),]
                       d[,"dead"] <- as.numeric(d[,"dead"])
                       
                       fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                       #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                       
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                         #plot.result <- ggsurv(fit.result)
                         
                         if (input_adjust_OS_pan_cancer_assessment_expression=="None") {
                           diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                           if (class(diff.result.try)=="try-error") {
                             p.value <- NA
                             table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           } else {
                             p.value <- signif(summary(diff.result)$logtest[3],2)
                             table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                 caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                        signif(summary(diff.result)$logtest[1],2),
                                                                                        " on ",
                                                                                        signif(summary(diff.result)$logtest[2],2),
                                                                                        " df, p-value = ",
                                                                                        signif(summary(diff.result)$logtest[3],2)),
                                                                                 br(),
                                                                                 paste0("n = ",
                                                                                        signif(summary(diff.result)$n,2),
                                                                                        ", number of events = ",
                                                                                        signif(summary(diff.result)$nevent,2)),
                                                                                 style="text-align:left;color:black;font-size:15px"),
                                                 options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             
                           }
                         } else {
                           diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_expression),data=d) )
                           if (class(diff.result.adjust.try)=="try-error") {
                             diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                             p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                               " (No ",input_adjust_OS_pan_cancer_assessment_expression," was observed)")
                             
                             table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                 caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                        signif(summary(diff.result)$logtest[1],2),
                                                                                        " on ",
                                                                                        signif(summary(diff.result)$logtest[2],2),
                                                                                        " df, p-value = ",
                                                                                        signif(summary(diff.result)$logtest[3],2)),
                                                                                 br(),
                                                                                 paste0("n = ",
                                                                                        signif(summary(diff.result)$n,2),
                                                                                        ", number of events = ",
                                                                                        signif(summary(diff.result)$nevent,2)),
                                                                                 style="text-align:left;color:black;font-size:15px"),
                                                 options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           } else {
                             p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                             table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                             rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression
                             table1 <- datatable(table1.pre,
                                                 caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                        signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                        " on ",
                                                                                        signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                        " df, p-value = ",
                                                                                        signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                 br(),
                                                                                 paste0("n = ",
                                                                                        signif(summary(diff.result.adjust)$n,2),
                                                                                        ", number of events = ",
                                                                                        signif(summary(diff.result.adjust)$nevent,2)),
                                                                                 style="text-align:left;color:black;font-size:15px"),
                                                 options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           }
                         }
                         
                         
                         
                         
                         # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                         # #diff.result <- survdiff(coxph(Surv(OS,dead)~IND+Age,data=d))
                         # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                         plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                           ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                    #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                    paste0("p-value = ",p.value),x=300,y=0.05)) +
                           ggplot2::guides(color=FALSE,linetype=FALSE)
                         p1<-ggplotly(plot.result.ann,
                                      height = 450 ) %>%
                           layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                  xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                  yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                  legend=list(orientation="v",x=0.6,y=1))
                         incProgress(100, detail = "Doing part %")
                         p1
                         
                         
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_expression_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_expression_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
       #}
     })
     
     ###############
     #### pan_cancer_assessment_expression PFS
     ###############
     
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       #if (input$gene_pan_cancer_assessment_expression!="ALL") {
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         selected.database <- selecteddatabase_pan_cancer_assessment_expression()
         input_adjust_OS_pan_cancer_assessment_expression <- input$adjust_OS_pan_cancer_assessment_expression
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_expression_PFS_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
           
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_survival_plots_name <- paste0("pan_cancer_assessment_expression_survival_plots", my_i)
             # pan_cancer_assessment_expression_survival_summary_tables_name <- paste0("pan_cancer_assessment_expression_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_expression_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (input$dynamic_check_pan_cancer_assessment_expression) {
                     d <- subset(d,description=="Pre")
                   }
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),]
                   
                   d <- data.frame(data_source=as.character(d$data_source),
                                   PFS.status=as.character(d$PFS.status),
                                   PFS=as.numeric(d$PFS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   UserGene=as.numeric(d[,"UserGene"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserGene")]))
                   d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% (",
                                             nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                                             sep=""))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% ",
                   #                           input_gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression (",
                   #                           nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),"IND"]<-
                     paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% (",
                           nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                           sep="")
                   # d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% ", input_gene_pan_cancer_assessment_expression," pan_cancer_assessment_expression (",
                   #         nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                   #         sep="")
                   d <- data.frame(d,group.IND=0)
                   d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                   d<-d[order(d[,"IND"]),]
                   
                   fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_expression=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_expression),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_expression," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_expression_PFS_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_expression_PFS_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
       #}
     })
     observeEvent(input$do_pan_cancer_assessment_expression,{
       input$do_pan_cancer_assessment_expression
       isolate({
         input_gene_pan_cancer_assessment_expression <- input$gene_pan_cancer_assessment_expression
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression()
         input_adjust_OS_pan_cancer_assessment_expression <- input$adjust_OS_pan_cancer_assessment_expression
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           output[["pan_cancer_assessment_individual_expression_PFS_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)]
           
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_PFS_plots_name <- paste0("pan_cancer_assessment_individual_expression_PFS_plots", my_i)
               pan_cancer_assessment_individual_expression_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_expression_PFS_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_expression_PFS_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),])==0 ) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       if (input$dynamic_check_pan_cancer_assessment_expression) {
                         d <- subset(d,description=="Pre")
                       }
                       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserGene"]),]
                       
                       d <- data.frame(data_source=as.character(d$data_source),
                                       PFS.status=as.character(d$PFS.status),
                                       PFS=as.numeric(d$PFS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       UserGene=as.numeric(d[,"UserGene"]))
                       d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_OS*100,"% (",
                                                 
                                                 nrow(d[d[,"UserGene"]<= quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                                                 sep=""))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),"IND"]<-
                         paste("top ",(1-input$slider_pan_cancer_assessment_expression_OS)*100,"% (",
                               nrow(d[d[,"UserGene"]> quantile(d[,"UserGene"],input$slider_pan_cancer_assessment_expression_OS),]),")",
                               sep="")
                       d <- data.frame(d,group.IND=0)
                       d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                       d<-d[order(d[,"IND"]),]
                       
                       fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = str_sub(input$data_source_pan_cancer_assessment_expression,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         
                       } else {
                         
                         plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                         
                         
                         if (input_adjust_OS_pan_cancer_assessment_expression=="None") {
                           diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                           if (class(diff.result.try)=="try-error") {
                             p.value <- NA
                             table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           } else {
                             p.value <- signif(summary(diff.result)$logtest[3],2)
                             table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                 caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                        signif(summary(diff.result)$logtest[1],2),
                                                                                        " on ",
                                                                                        signif(summary(diff.result)$logtest[2],2),
                                                                                        " df, p-value = ",
                                                                                        signif(summary(diff.result)$logtest[3],2)),
                                                                                 br(),
                                                                                 paste0("n = ",
                                                                                        signif(summary(diff.result)$n,2),
                                                                                        ", number of events = ",
                                                                                        signif(summary(diff.result)$nevent,2)),
                                                                                 style="text-align:left;color:black;font-size:15px"),
                                                 options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             
                           }
                         } else {
                           diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_expression),data=d) )
                           if (class(diff.result.adjust.try)=="try-error") {
                             diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                             p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                               " (No ",input_adjust_OS_pan_cancer_assessment_expression," was observed)")
                             
                             table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                 caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                        signif(summary(diff.result)$logtest[1],2),
                                                                                        " on ",
                                                                                        signif(summary(diff.result)$logtest[2],2),
                                                                                        " df, p-value = ",
                                                                                        signif(summary(diff.result)$logtest[3],2)),
                                                                                 br(),
                                                                                 paste0("n = ",
                                                                                        signif(summary(diff.result)$n,2),
                                                                                        ", number of events = ",
                                                                                        signif(summary(diff.result)$nevent,2)),
                                                                                 style="text-align:left;color:black;font-size:15px"),
                                                 options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           } else {
                             p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                             table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                             rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression
                             table1 <- datatable(table1.pre,
                                                 caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                        signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                        " on ",
                                                                                        signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                        " df, p-value = ",
                                                                                        signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                 br(),
                                                                                 paste0("n = ",
                                                                                        signif(summary(diff.result.adjust)$n,2),
                                                                                        ", number of events = ",
                                                                                        signif(summary(diff.result.adjust)$nevent,2)),
                                                                                 style="text-align:left;color:black;font-size:15px"),
                                                 options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           }
                         }
                         
                         
                         
                         # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                         # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                         plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                           ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                    #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                    paste0("p-value = ",p.value),x=300,y=0.05)) +
                           ggplot2::guides(color=FALSE,linetype=FALSE)
                         p1<-ggplotly(plot.result.ann,
                                      height = 450 ) %>%
                           layout(title = unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                  xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                  yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                  legend=list(orientation="v",x=0.6,y=1))
                         incProgress(100, detail = "Doing part %")
                         p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_expression_PFS_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_expression_PFS_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
       #}
     })
     
     #####################################################
     #### pan-cancer assessment expression_sum
     #####################################################
     observe({
       if (input$defined_gene_pan_cancer_assessment_expression_sum=="user-defined list") {
         updateSelectizeInput(session,"gene_pan_cancer_assessment_expression_sum",
                              choices=a.colnames,selected=c("CD4","CD6","CD8A","CD8B"),
                              server=TRUE)
       } else {
         selected_pan_cancer_assessment_expression_sum <- switch (input$defined_gene_pan_cancer_assessment_expression_sum,
                                                                  "DNA_damage_response" = DNA_damage_response,
                                                                  "PDL1_pan_cancer_assessment_expression_sum_sum_and_PD1_checkpoint_pathway_in_cancer"=PDL1_expression_and_PD1_checkpoint_pathway_in_cancer,
                                                                  "T_cell_receptor_signaling_pathway"=T_cell_receptor_signaling_pathway,
                                                                  "B_cell_receptor_signaling_pathway"=B_cell_receptor_signaling_pathway,
                                                                  "Cell_cycle_control"=Cell_cycle_control,
                                                                  "Survival_cell_death_regulation_signaling"=Survival_cell_death_regulation_signaling)
         updateSelectizeInput(session,"gene_pan_cancer_assessment_expression_sum",
                              choices=a.colnames,selected=selected_pan_cancer_assessment_expression_sum,
                              server=TRUE)
       }
     })
     observe({
       if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
         
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_expression_sum",
                              choices=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                              selected=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                              server=T)
         
       }
     })
     
     observe({
       if (input$data_source_pan_cancer_assessment_expression_sum=="" || is.null(input$data_source_pan_cancer_assessment_expression_sum) ||
           input$treatment_pan_cancer_assessment_expression_sum=="" || is.null(input$treatment_pan_cancer_assessment_expression_sum) ||
           input$description_pan_cancer_assessment_expression_sum=="" || is.null(input$description_pan_cancer_assessment_expression_sum) ||
           input$gene_pan_cancer_assessment_expression_sum=="" || is.null(input$gene_pan_cancer_assessment_expression_sum)){
         disable("do_pan_cancer_assessment_expression_sum")
       } else {
         enable("do_pan_cancer_assessment_expression_sum")
       }
     })
     
     
     selecteddatabase_pan_cancer_assessment_expression_sum <- eventReactive(input$do_pan_cancer_assessment_expression_sum,{
       if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
         user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
         
         #data_source1 <- input$data_source_pan_cancer_assessment_expression_sum[input$data_source_pan_cancer_assessment_expression_sum %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
         #treatment1 <- input$treatment_pan_cancer_assessment_expression_sum
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression_sum])
         description1 <- c("Pre","On")
         a<-data.table(database.test.batchcorrected)
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genessum <- input$gene_pan_cancer_assessment_expression_sum
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",genessum)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         #colnames(m)[ncol(m)] <- "sum"
         m <- as.data.frame(m)
         if (nrow(m) >0) {
           m <- data.frame(m,sum=apply(as.data.frame(m[,13:ncol(m)]),1,function(x) sum(na.omit(x))))
           m[m[,"sum"]==0,"sum"]<-NA
           
           samples <- m$PatientID
           samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
           samples_paired <- table(samples)
           samples_paired <- samples_paired[samples_paired==2]
           samples_paired_pre <- paste0(names(samples_paired),"Pre")
           #samples_paired_on <- paste0(names(samples_paired),"On")
           samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
           m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],sum=m[m[,"PatientID"] %in% samples_paired_on,"sum"]-m[m[,"PatientID"] %in% samples_paired_pre,"sum"]),
                      data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],sum=m[m[,"PatientID"] %in% samples_paired_on,"sum"]-m[m[,"PatientID"] %in% samples_paired_pre,"sum"]))
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           d <- m
           # m1 <- m %>% split(.$data_source) %>% 
           #   lapply(function(d) {#d[d[,"sum"]>=median(na.omit(d[,"sum"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]>=quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% ",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]<quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% ",sep="")
           # d})
           # m2<-ldply(m1,data.frame)
           # m2<-m2[,-1]
           # m2$sum <- as.numeric(m2$sum)
           # m2
           d$sum <- as.numeric(d$sum)
           d
         } else {
           m
         }
         
         
         
       } else {
         
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)
         #treatment1 <- input$treatment_pan_cancer_assessment_expression_sum
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression_sum])
         description1 <- input$description_pan_cancer_assessment_expression_sum
         a<-data.table(database.test.batchcorrected)
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_pan_cancer_assessment_expression_sum)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "sum"
         m <- as.data.frame(m)
         if (nrow(m)>0) {
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           d <- m
           # m1 <- m %>% split(.$data_source) %>% 
           #   lapply(function(d) {#d[d[,"sum"]>=median(na.omit(d[,"sum"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]>=quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% ",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]<quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% ",sep="")
           # d})
           # m2<-ldply(m1,data.frame)
           # m2<-m2[,-1]
           # m2$sum <- as.numeric(m2$sum)
           # m2
           d$sum <- as.numeric(d$sum)
           d
         } else {
           m2<-m
           m2
         }
       }
     })
     
     selecteddatabase_pan_cancer_assessment_individual_expression_sum <- eventReactive(input$do_pan_cancer_assessment_expression_sum,{
       if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
         user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
         
         #data_source1 <- input$data_source_pan_cancer_assessment_expression_sum[input$data_source_pan_cancer_assessment_expression_sum %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
         #treatment1 <- input$treatment_pan_cancer_assessment_expression_sum
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression_sum])
         description1 <- c("Pre","On")
         a<-data.table(database.test.process())
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genessum <- input$gene_pan_cancer_assessment_expression_sum
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",genessum)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         #colnames(m)[ncol(m)] <- "sum"
         m <- as.data.frame(m)
         if (nrow(m) >0) {
           m <- data.frame(m,sum=apply(as.data.frame(m[,13:ncol(m)]),1,function(x) sum(na.omit(x))))
           m[m[,"sum"]==0,"sum"]<-NA
           
           samples <- m$PatientID
           samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
           samples_paired <- table(samples)
           samples_paired <- samples_paired[samples_paired==2]
           samples_paired_pre <- paste0(names(samples_paired),"Pre")
           #samples_paired_on <- paste0(names(samples_paired),"On")
           samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
           m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],sum=m[m[,"PatientID"] %in% samples_paired_on,"sum"]-m[m[,"PatientID"] %in% samples_paired_pre,"sum"]),
                      data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],sum=m[m[,"PatientID"] %in% samples_paired_on,"sum"]-m[m[,"PatientID"] %in% samples_paired_pre,"sum"]))
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           #d <- m
           m1 <- m %>% split(.$data_source) %>%
             lapply(function(d) {#d[d[,"sum"]>=median(na.omit(d[,"sum"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]>=quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% ",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]<quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% ",sep="")
           d})
           m2<-ldply(m1,data.frame)
           m2<-m2[,-1]
           m2$sum <- as.numeric(m2$sum)
           m2
           # d$sum <- as.numeric(d$sum)
           # d
         } else {
           m
         }
         
         
         
       } else {
         
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)
         #treatment1 <- input$treatment_pan_cancer_assessment_expression_sum
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression_sum])
         description1 <- input$description_pan_cancer_assessment_expression_sum
         a<-data.table(database.test.process())
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$gene_pan_cancer_assessment_expression_sum)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "sum"
         m <- as.data.frame(m)
         if (nrow(m)>0) {
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           #d <- m
           m1 <- m %>% split(.$data_source) %>%
             lapply(function(d) {#d[d[,"sum"]>=median(na.omit(d[,"sum"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]>=quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% ",sep="")
           d[!is.na(d[,"sum"]) & d[,"sum"]<quantile(na.omit(d[,"sum"]),input$slider_pan_cancer_assessment_expression_sum_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% ",sep="")
           d})
           m2<-ldply(m1,data.frame)
           m2<-m2[,-1]
           m2$sum <- as.numeric(m2$sum)
           m2
           # d$sum <- as.numeric(d$sum)
           # d
         } else {
           m2<-m
           m2
         }
       }
     })
     
     ################
     #### pan_cancer_assessment_expression_sum summary
     ################
     output$pan_cancer_assessment_expression_sum_summary_disease_number <- renderText({
       input$do_pan_cancer_assessment_expression_sum ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_assessment_expression_sum()$Disease))})})
     output$pan_cancer_assessment_expression_sum_summary_sample_number <- renderText({
       input$do_pan_cancer_assessment_expression_sum ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_assessment_expression_sum())})})
     output$pan_cancer_assessment_expression_sum_summary_plots <- renderPlotly({
       summary_plot(inputdata=selecteddatabase_pan_cancer_assessment_expression_sum(),feature="expression sum")
     })
     
     
     
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         # if (input_gene_pan_cancer_assessment_expression_sum=="ALL") { 
         #   selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum_ALL()
         # } else {
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum()
         # }
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_expression_sum_summary_clincial_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           
           
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_sum_summary_clinical_plots_name <- paste0("pan_cancer_assessment_expression_sum_summary_clinical_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_expression_sum_summary_clinical_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 if (nrow(d[!is.na(d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]]) & !is.na(d[,"sum"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   d <- d[!is.na(d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]]) & !is.na(d[,"sum"]),]
                   if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                     d <- subset(d,description=="Pre")
                   }
                   if(input$clinical_pan_cancer_assessment_expression_sum=="Age") {
                     d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                     #d[d[,"Age"]<=10,"Age"] <- "0-10"
                     d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                     d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                   }
                   
                   if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                     p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]],
                                y=~signif(as.numeric(as.character(sum)),3),  
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]],showlegend=TRUE,
                                colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                hoveron="points+boxes",
                                legendgroup=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]], 
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Changes of expression_sum: ",input$gene_pan_cancer_assessment_expression_sum," :",signif(as.numeric(as.character(sum)),3), 
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title="Clinical Attribute",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste("Changes of ",input_gene_pan_cancer_assessment_expression_sum, " pan_cancer_assessment_expression_sum value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              ))
                     
                   } else {
                     p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]],
                                y=~signif(as.numeric(as.character(sum)),3), 
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]],showlegend=TRUE,
                                colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                hoveron="points+boxes",
                                legendgroup=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_sum]], 
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>expression_sum: ",input$gene_pan_cancer_assessment_expression_sum," :",signif(as.numeric(as.character(sum)),3), 
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title="Clinical Attribute",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste(input_gene_pan_cancer_assessment_expression_sum, " pan_cancer_assessment_expression_sum value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              ))
                   }
                   incProgress(100, detail = "Doing part %")
                   p
                   
                 } 
                 #}
               })
             })
           })
           #}
           # }
         }
       }) 
     })
     
     output$pan_cancer_assessment_expression_sum_summary_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_assessment_expression_sum_summary_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_assessment_expression_sum_summary_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_assessment_expression_sum_summary_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum()
         if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
           colnames(selected.database)[which(colnames(selected.database)=="sum")] <- paste0("Changes of ",input$gene_pan_cancer_assessment_expression_sum)
         } else {
           colnames(selected.database)[which(colnames(selected.database)=="sum")] <- input$gene_pan_cancer_assessment_expression_sum
         }
         if (input$pan_cancer_assessment_expression_sum_summary_type == "Excel (CSV)") {
           write.csv (selected.database,file,row.names = FALSE)
         } else if (input$pan_cancer_assessment_expression_sum_summary_type == "Text (TSV)") {
           write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_assessment_expression_sum_summary_type == "JSON") {
           write(toJSON(selected.database),file)
         }
         #}
       }
     )
     
     output$pan_cancer_assessment_expression_sum_summary_table <- DT::renderDataTable({
       
       selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum()
       if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
         colnames(selected.database)[which(colnames(selected.database)=="sum")] <- paste0("Changes of ",input$gene_pan_cancer_assessment_expression_sum)
       } else {
         colnames(selected.database)[which(colnames(selected.database)=="sum")] <- input$gene_pan_cancer_assessment_expression_sum
       }
       
       selected.database
       #}
     })
     
     ################
     #### pan_cancer_assessment_expression_sum plots
     ################
     
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"sum"]),])==0) {
           
           output[["pan_cancer_assessment_expression_sum_plots_plots"]] <- renderPlotly({
             #show("loading-c")
             #hide("loading-d")
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_sum_plots_plots_name <- paste0("pan_cancer_assessment_expression_sum_plots_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_expression_sum_plots_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 #show("loading-c")
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   #hide("loading-d")
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),])==0 ) {
                   #hide("loading-d")
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                 } else {
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),]
                   if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                     d <- subset(d,description=="Pre")
                   }
                   
                   d[,"Groups"] <- as.character(d[,"Groups"])
                   d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                   d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                   ind <- sort(unique(d[,"Groups"]))
                   # if (as.character(unique(d$data_source)) == "Melanoma-Van Allen" | as.character(unique(d$data_source)) == "Melanoma-Riaz" | as.character(unique(d$data_source)) == "Melanoma-Auslander") {
                   #   lable_y <- " (RSEM)"
                   # } else if (as.character(unique(d$data_source)) == "Melanoma-Chen" | as.character(unique(d$data_source)) == "Melanoma-Prat") {
                   #   lable_y <- " (NanoString signal)"
                   # } else if (as.character(unique(d$data_source)) == "Melanoma-Hugo" | as.character(unique(d$data_source)) == "Melanoma-Gide" | as.character(unique(d$data_source)) == "Melanoma-Snyder-Nathanson") {
                   #   lable_y <- " (FPKM)"
                   # } else {
                   #   lable_y <- NULL
                   # }
                   
                   
                   try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"sum"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"sum"])))$p.value)
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     p.v <- " = NA"
                   } else {
                     # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"sum"]),as.numeric(d[d[,"Groups"]==ind[2],"sum"]))$p.value
                     # if (p.v < 0.001) {
                     #   p.v <- "< 0.001"
                     # } else {
                     p.v<-paste0(" = ",signif(p.v,2))
                     # }
                   }
                   
                   if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                     p<-plot_ly(data=d,x=~Groups,
                                y=~signif(as.numeric(as.character(sum)),3),
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=~PFS.status,showlegend=TRUE,
                                colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                #hoveron="points+boxes",
                                legendgroup=~PFS.status,
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Change of ",input_gene_pan_cancer_assessment_expression_sum," expression_sum:",signif(as.numeric(as.character(sum)),3),
                                              "<br>Treatment: ",treatment))  %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title=" ",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste("Change of ",input_gene_pan_cancer_assessment_expression_sum, " expression_sum value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              )) %>%
                       add_annotations(x = 0.5,
                                       y = d[which.max(d[,"sum"]),"sum"],
                                       text = paste0("p",p.v),
                                       xref = paste0("x", 1),  # add this
                                       yref = paste0("y", 1),  # add this
                                       showarrow = FALSE,
                                       ax = 20,
                                       ay = -40,
                                       font=list(size=20,color="black",family= "Aril black"))
                   } else {
                     p<-plot_ly(data=d,x=~Groups,
                                y=~signif(as.numeric(as.character(sum)),3),
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=~PFS.status,showlegend=TRUE,
                                colors=c("#7CDBD5","#DCB239"),#c("plum","lightgreen"),
                                #hoveron="points+boxes",
                                legendgroup=~PFS.status,
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Eexpression_sum: ",input_gene_pan_cancer_assessment_expression_sum," :",signif(as.numeric(as.character(sum)),3),
                                              "<br>Treatment: ",treatment))  %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title=" ",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste(input_gene_pan_cancer_assessment_expression_sum, " expression_sum value",sep=""),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              )) %>%
                       add_annotations(x = 0.5,
                                       y = d[which.max(d[,"sum"]),"sum"],
                                       text = paste0("p",p.v),
                                       xref = paste0("x", 1),  # add this
                                       yref = paste0("y", 1),  # add this
                                       showarrow = FALSE,
                                       ax = 20,
                                       ay = -40,
                                       font=list(size=20,color="black",family= "Aril black"))
                   }
                   #hide("loading-d")
                   incProgress(100, detail = "Doing part %")
                   p
                 }
                 
                 #}
               })
             })
           })
           #}
         }
       })
       
     })
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression_sum()
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_individual_expression_sum_plots_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
           
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_sum_plots_plots_name <- paste0("pan_cancer_assessment_individual_expression_sum_plots_plots", my_i)
               d<-splitted_list[[my_i]]
               output[[pan_cancer_assessment_individual_expression_sum_plots_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   isolate({
                     if (is.na(names(splitted_list)[my_i])) {
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>% 
                         layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     } else {
                       #d[,11:ncol(d)] <- unfactor(d[,11:ncol(d)])
                       if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),])==0 ) {
                         incProgress(100, detail = "Doing part %")
                         plot_ly () %>% 
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       } else {
                         d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"sum"]),]
                         if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                           d <- subset(d,description=="Pre")
                         }
                         d[,"Groups"] <- as.character(d[,"Groups"])
                         d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                         d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                         ind <- sort(unique(d[,"Groups"]))
                         if (as.character(unique(d$data_source)) == "Melanoma-Van Allen" | as.character(unique(d$data_source)) == "Melanoma-Riaz" | as.character(unique(d$data_source)) == "Melanoma-Auslander") {
                           lable_y <- " (RSEM)"
                         } else if (as.character(unique(d$data_source)) == "Melanoma-Chen" | as.character(unique(d$data_source)) == "Melanoma-Prat") {
                           lable_y <- " (NanoString signal)"
                         } else if (as.character(unique(d$data_source)) == "Melanoma-Hugo" | as.character(unique(d$data_source)) == "Melanoma-Gide" | as.character(unique(d$data_source)) == "Melanoma-Snyder-Nathanson") {
                           lable_y <- " (FPKM)"
                         } else {
                           lable_y <- NULL
                         }
                         try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"sum"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"sum"])))$p.value)
                         if (class(try.test)=="try-error" | is.na(try.test)) {
                           p.v <- " = NA"
                         } else {
                           # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"sum"]),as.numeric(d[d[,"Groups"]==ind[2],"sum"]))$p.value
                           # if (p.v < 0.001) {
                           #   p.v <- " < 0.001"
                           # } else {
                           p.v<-paste0(" = ",signif(p.v,2))
                           # } 
                         }
                         
                         if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                           p<-plot_ly(data=d,x= ~Groups,
                                      y= ~signif(as.numeric(as.character(sum)),3), 
                                      type="violin",#type="box",
                                      points="all",#boxpoints="all",
                                      jitter=0.3,
                                      pointpos=-1.8,
                                      #mode="markders",
                                      color= ~PFS.status,
                                      colors=c("#7CDBD5","#DCB239","gray"),
                                      #hoveron="points+boxes",
                                      legendgroup=~PFS.status,showlegend=TRUE,
                                      text= ~paste0("Sample ID: ", PatientID,
                                                    "<br>Expression: ",signif(as.numeric(as.character(sum)),3), 
                                                    "<br>Treatment: ",treatment,
                                                    "<br>Description: On-Pre")
                                      
                           )  %>%
                             layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                    height = 480,
                                    xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                               showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                    yaxis=list(title=paste0("Changes of expression sum of selected expression",
                                                            lable_y),
                                               titlefont=list(size=20,color="black",family= "Aril"),
                                               showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                    )) %>% 
                             add_annotations(x = 0.5,
                                             y = d[which.max(d[,"sum"]),"sum"],
                                             text = paste0("p-value",p.v),
                                             xref = paste0("x", 1),  # add this
                                             yref = paste0("y", 1),  # add this
                                             showarrow = FALSE,
                                             ax = 20,
                                             ay = -40,
                                             font=list(size=20,color="black",family= "Aril black"))
                         } else {
                           
                           p<-plot_ly(data=d,x= ~Groups,
                                      y= ~signif(as.numeric(as.character(sum)),3), 
                                      type="violin",#type="box",
                                      points="all",#boxpoints="all",
                                      jitter=0.3,
                                      pointpos=-1.8,
                                      #mode="markders",
                                      color= ~PFS.status,
                                      colors=c("#7CDBD5","#DCB239","gray"),
                                      #hoveron="points+boxes",
                                      legendgroup=~PFS.status,showlegend=TRUE,
                                      text= ~paste0("Sample ID: ", PatientID,
                                                    "<br>Expression: ",signif(as.numeric(as.character(sum)),3), 
                                                    "<br>Treatment: ",treatment,
                                                    "<br>Description: ",description)
                                      
                           )  %>%
                             layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                    height = 480,
                                    xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                               showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                    yaxis=list(title=paste0("Sum of selected expression",
                                                            lable_y),
                                               titlefont=list(size=20,color="black",family= "Aril"),
                                               showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                    )) %>% 
                             add_annotations(x = 0.5,
                                             y = d[which.max(d[,"sum"]),"sum"],
                                             text = paste0("p-value",p.v),
                                             xref = paste0("x", 1),  # add this
                                             yref = paste0("y", 1),  # add this
                                             showarrow = FALSE,
                                             ax = 20,
                                             ay = -40,
                                             font=list(size=20,color="black",family= "Aril black"))
                         }
                         incProgress(100, detail = "Doing part %")
                         p
                       } 
                       
                     }
                   })
                 })
               })
             })
           }
         }
       })
     })
     
     ################
     #### pan_cancer_assessment_expression_sum survival
     ################
     
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       #if (input$gene_pan_cancer_assessment_expression_sum!="ALL") {
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum()
         input_adjust_OS_pan_cancer_assessment_expression_sum <- input$adjust_OS_pan_cancer_assessment_expression_sum
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_expression_sum_survival_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
           
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_sum_survival_plots_name <- paste0("pan_cancer_assessment_expression_sum_survival_plots", my_i)
             # pan_cancer_assessment_expression_sum_survival_summary_tables_name <- paste0("pan_cancer_assessment_expression_sum_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_expression_sum_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                     d <- subset(d,description=="Pre")
                   }
                   d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                   
                   d<-data.frame(d,dead=NA)
                   d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                   d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                   d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                   d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                   #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))
                   d <- data.frame(data_source=as.character(d$data_source),
                                   dead=as.numeric(d$dead),
                                   OS=as.numeric(d$OS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   sum=as.numeric(d[,"sum"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","sum")]))
                   d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% (",
                                             nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                                             sep=""))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% ",
                   #                           input_gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum (",
                   #                           nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),"IND"]<-
                     paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% (",
                           nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                           sep="")
                   # d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% ", input_gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum (",
                   #         nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                   #         sep="")
                   d<-d[order(d[,"IND"]),]
                   d[,"dead"] <- as.numeric(d[,"dead"])
                   
                   fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_expression_sum=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_sum),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_expression_sum," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_sum
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_expression_sum_survival_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_expression_sum_survival_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
       #}
     })
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression_sum()
         input_adjust_OS_pan_cancer_assessment_expression_sum <- input$adjust_OS_pan_cancer_assessment_expression_sum
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_expression_sum_survival_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
           
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_sum_survival_plots_name <- paste0("pan_cancer_assessment_individual_expression_sum_survival_plots", my_i)
               pan_cancer_assessment_individual_expression_sum_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_expression_sum_survival_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_expression_sum_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0 ) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                       if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                         d <- subset(d,description=="Pre")
                       }
                       d<-data.frame(d,dead=NA)
                       d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                       d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                       d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                       d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                       d <- data.frame(data_source=as.character(d$data_source),
                                       dead=as.numeric(d$dead),
                                       OS=as.numeric(d$OS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       sum=as.numeric(d[,"sum"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))
                       d<-data.frame(d,IND=paste0("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% (",
                                                  nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"sum"]>=quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),"IND"] <- 
                         paste0("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% (",
                                nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")")
                       d<-d[order(d[,"IND"]),]
                       d[,"dead"] <- as.numeric(d[,"dead"])
                       
                       fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                       
                        plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                       
                        if (input_adjust_OS_pan_cancer_assessment_expression_sum=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           
                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_sum),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_expression_sum," was observed)")
                           
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_sum
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann,
                                    height = 450 ) %>%
                         layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                legend=list(orientation="v",x=0.6,y=1))
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_expression_sum_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_expression_sum_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     ###############
     #### pan_cancer_assessment_expression_sum PFS
     ###############
     
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       #if (input$gene_pan_cancer_assessment_expression_sum!="ALL") {
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_sum()
         input_adjust_OS_pan_cancer_assessment_expression_sum <- input$adjust_OS_pan_cancer_assessment_expression_sum
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_expression_sum_PFS_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
           
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_sum_survival_plots_name <- paste0("pan_cancer_assessment_expression_sum_survival_plots", my_i)
             # pan_cancer_assessment_expression_sum_survival_summary_tables_name <- paste0("pan_cancer_assessment_expression_sum_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_expression_sum_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                     d <- subset(d,description=="Pre")
                   }
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                   
                   d <- data.frame(data_source=as.character(d$data_source),
                                   PFS.status=as.character(d$PFS.status),
                                   PFS=as.numeric(d$PFS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   sum=as.numeric(d[,"sum"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","sum")]))
                   d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% (",
                                             nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                                             sep=""))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% ",
                   #                           input_gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum (",
                   #                           nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),"IND"]<-
                     paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% (",
                           nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                           sep="")
                   # d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% ", input_gene_pan_cancer_assessment_expression_sum," pan_cancer_assessment_expression_sum (",
                   #         nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                   #         sep="")
                   d <- data.frame(d,group.IND=0)
                   d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                   d<-d[order(d[,"IND"]),]
                   
                   fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_expression_sum=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_sum),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_expression_sum," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_sum
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_expression_sum_PFS_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_expression_sum_PFS_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
       #}
     })
     observeEvent(input$do_pan_cancer_assessment_expression_sum,{
       input$do_pan_cancer_assessment_expression_sum
       isolate({
         input_gene_pan_cancer_assessment_expression_sum <- input$gene_pan_cancer_assessment_expression_sum
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression_sum()
         input_adjust_OS_pan_cancer_assessment_expression_sum <- input$adjust_OS_pan_cancer_assessment_expression_sum
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_individual_expression_sum_PFS_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)]
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression_sum)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_sum_PFS_plots_name <- paste0("pan_cancer_assessment_individual_expression_sum_PFS_plots", my_i)
               pan_cancer_assessment_individual_expression_sum_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_expression_sum_PFS_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_expression_sum_PFS_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_sum,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0 ) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                       if (input$dynamic_check_pan_cancer_assessment_expression_sum) {
                         d <- subset(d,description=="Pre")
                       }
                       d <- data.frame(data_source=as.character(d$data_source),
                                       PFS.status=as.character(d$PFS.status),
                                       PFS=as.numeric(d$PFS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       sum=as.numeric(d[,"sum"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","sum")]))
                       d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_expression_sum_OS*100,"% (", 
                                                 nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                                                 sep=""))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),"IND"]<- 
                         paste("top ",(1-input$slider_pan_cancer_assessment_expression_sum_OS)*100,"% (",
                               nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_sum_OS),]),")",
                               sep="")
                       d <- data.frame(d,group.IND=0)
                       d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                        plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                       
                        if (input_adjust_OS_pan_cancer_assessment_expression_sum=="None") {
                          diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                          if (class(diff.result.try)=="try-error") {
                            p.value <- NA
                            table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           
                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_sum),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_expression_sum," was observed)")
                           
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_sum
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1 <- ggplotly(plot.result.ann,
                                      height = 480 ) %>%
                         layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                legend=list(orientation="v",x=0.6,y=1))
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_expression_sum_PFS_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_expression_sum_PFS_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     #####################################################
     #### pan-cancer assessment expression_pairs
     #####################################################
     
     updateSelectizeInput(session,"gene_pan_cancer_assessment_expression_pairs_1",choices=a.colnames,selected = c("CD28","CD276"),server=TRUE)
     updateSelectizeInput(session,"gene_pan_cancer_assessment_expression_pairs_2",choices=a.colnames,selected = c("PDCD1","TNFRSF4"),server=TRUE)
     #updateSelectizeInput(session,"gene_pan_cancer_assessment_expression_pairs_3",choices=a.colnames,selected = c("CTLA4","TNFRSF4"),server=TRUE)
     
     ids<<-c(1,2)
     #ids<<-c(1,2,3)
     #  ids <- NULL
     observeEvent(input$add_pan_cancer_assessment_expression_pairs,{
       if (is.null(ids)) {
         ids<<-1
       } else {
         ids <<- c(ids,max(ids)+1)
       }
       output$gene_pan_cancer_assessment_expression_pairs <- renderUI({
         tagList(
           lapply(3:length(ids),function(i) {
             selectizeInput(paste("gene_pan_cancer_assessment_expression_pairs_",ids[i],sep=""),sprintf("Gene pairs #%d",ids[i]),choices=NULL,multiple=TRUE,selected = NULL,options=list(maxItems=2))
           })
         )
       })
       lapply(3:length(ids),function(i) {
         updateSelectizeInput(session,paste("gene_pan_cancer_assessment_expression_pairs_",ids[i],sep=""),choices=a.colnames,selected=c("CTLA4","PDCD1"),server=TRUE)
       })
       
       updateSliderInput(session=session,
                         "slider_pan_cancer_assessment_expression_pairs_OS",
                         "Seperate samples by sum of gene relation pairs (F): ",
                         min=0,max=length(ids),value=0)
     })
     
     observe({
       if (input$data_source_pan_cancer_assessment_expression_pairs=="" || is.null(input$data_source_pan_cancer_assessment_expression_pairs) ||
           input$treatment_pan_cancer_assessment_expression_pairs=="" || is.null(input$treatment_pan_cancer_assessment_expression_pairs) ||
           input$description_pan_cancer_assessment_expression_pairs=="" || is.null(input$description_pan_cancer_assessment_expression_pairs) ){
         disable("do_pan_cancer_assessment_expression_pairs")
       } else {
         enable("do_pan_cancer_assessment_expression_pairs")
       }
     })
     
     
     selecteddatabase_pan_cancer_assessment_expression_pairs <- eventReactive(input$do_pan_cancer_assessment_expression_pairs,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)
       # treatment1 <- input$treatment_pan_cancer_assessment_expression_pairs
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression_pairs])
       description1 <- input$description_pan_cancer_assessment_expression_pairs
       if (is.null(ids)) {
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
       } else {
         genessum<-list()
         txt_ids <- sapply(1:length(ids), function(i) paste("gene_pan_cancer_assessment_expression_pairs_",ids[i],sep=""))
         for (i in 1:length(txt_ids)) {
           genessum[[i]] <- input[[ txt_ids[i] ]] 
         }
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",unlist(genessum))
       }
       a<-data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       selected.database<-as.data.frame(m)
       
       if (nrow(selected.database)>0) {
         n <- ncol(selected.database)
         genessum<-list()
         txt_ids <- sapply(1:length(ids), function(i) paste("gene_pan_cancer_assessment_expression_pairs_",ids[i],sep=""))
         isolate({
           for (i in 1:length(txt_ids)) {
             genessum[[i]] <- input[[ txt_ids[i] ]]
             if (length(input[[ txt_ids[i] ]]) == 1) {
               selected.database <- data.frame(selected.database,NA)
             } else {
               selected.database <- data.frame(selected.database,selected.database[,input[[ txt_ids[i] ]][1]] > selected.database[,input[[ txt_ids[i] ]][2]])
             }
           }
         })
         colnames(selected.database)[(n+1):ncol(selected.database)] <- unlist(lapply(genessum,function(x) paste(x,collapse=".")))
         cols <- sapply(selected.database, is.logical)
         for (j in (n+1):ncol(selected.database)) {
           selected.database[,j] <- as.numeric(selected.database[,j])
         }
         selected.database <- data.frame(selected.database,sum=NA)
         for (i in 1:nrow(selected.database)) {
           #selected.database[i,"sum"] <- sum(na.omit(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)])))
           selected.database[i,"sum"] <- sum(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)]))
         }
         m<-data.frame(selected.database,category="unknown")
         m[,"category"] <- as.character(m[,"category"])
         d <- m
         # m1 <- m %>% split(.$data_source) %>% 
         #   lapply(function(d) {
         d[!is.na(d[,"sum"]) & d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,"category"] <- paste0("higher than ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs")
         d[!is.na(d[,"sum"]) & d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,"category"] <- paste0("lower than or equal to ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs")
         #   d
         # })
         d 
         # m2 <- ldply(m1,data.frame)
         # m2 <- m2[,-1]
         # m2
       } else {
         selected.database
       }
     })
     
     selecteddatabase_pan_cancer_assessment_individual_expression_pairs <- eventReactive(input$do_pan_cancer_assessment_expression_pairs,{
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)
       # treatment1 <- input$treatment_pan_cancer_assessment_expression_pairs
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_expression_pairs])
       description1 <- input$description_pan_cancer_assessment_expression_pairs
       if (is.null(ids)) {
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender")
       } else {
         genessum<-list()
         txt_ids <- sapply(1:length(ids), function(i) paste("gene_pan_cancer_assessment_expression_pairs_",ids[i],sep=""))
         for (i in 1:length(txt_ids)) {
           genessum[[i]] <- input[[ txt_ids[i] ]] 
         }
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",unlist(genessum))
       }
       a<-data.table(database.test.process())
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       selected.database<-as.data.frame(m)
       
       if (nrow(selected.database)>0) {
         n <- ncol(selected.database)
         genessum<-list()
         txt_ids <- sapply(1:length(ids), function(i) paste("gene_pan_cancer_assessment_expression_pairs_",ids[i],sep=""))
         isolate({
           for (i in 1:length(txt_ids)) {
             genessum[[i]] <- input[[ txt_ids[i] ]]
             if (length(input[[ txt_ids[i] ]]) == 1) {
               selected.database <- data.frame(selected.database,NA)
             } else {
               selected.database <- data.frame(selected.database,selected.database[,input[[ txt_ids[i] ]][1]] > selected.database[,input[[ txt_ids[i] ]][2]])
             }
           }
         })
         colnames(selected.database)[(n+1):ncol(selected.database)] <- unlist(lapply(genessum,function(x) paste(x,collapse=".")))
         cols <- sapply(selected.database, is.logical)
         for (j in (n+1):ncol(selected.database)) {
           selected.database[,j] <- as.numeric(selected.database[,j])
         }
         selected.database <- data.frame(selected.database,sum=NA)
         for (i in 1:nrow(selected.database)) {
           #selected.database[i,"sum"] <- sum(na.omit(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)])))
           selected.database[i,"sum"] <- sum(as.numeric(selected.database[i,(n+1):(ncol(selected.database)-1)]))
         }
         m<-data.frame(selected.database,category="unknown")
         m[,"category"] <- as.character(m[,"category"])
         #d <- m
         m1 <- m %>% split(.$data_source) %>%
           lapply(function(d) {
         d[!is.na(d[,"sum"]) & d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,"category"] <- paste0("higher than ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs")
         d[!is.na(d[,"sum"]) & d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,"category"] <- paste0("lower than or equal to ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs")
           d
         })
         
         m2 <- ldply(m1,data.frame)
         m2 <- m2[,-1]
         m2
       } else {
         selected.database
       }
     })
     
     ################
     #### pan_cancer_assessment_expression_pairs summary
     ################
     output$pan_cancer_assessment_expression_pairs_summary_disease_number <- renderText({
       input$do_pan_cancer_assessment_expression_pairs ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_assessment_expression_pairs()$Disease))})})
     output$pan_cancer_assessment_expression_pairs_summary_sample_number <- renderText({
       input$do_pan_cancer_assessment_expression_pairs ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_assessment_expression_pairs())})})
     output$pan_cancer_assessment_expression_pairs_summary_plots <- renderPlotly({
       summary_plot(inputdata=selecteddatabase_pan_cancer_assessment_expression_pairs(),feature="expression pairs")
     })
     
     
     
     observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
       input$do_pan_cancer_assessment_expression_pairs
       isolate({
         # if (input_gene_pan_cancer_assessment_expression_pairs=="ALL") { 
         #   selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs_ALL()
         # } else {
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs()
         # }
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_expression_pairs_summary_clincial_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           
           
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_pairs_summary_clinical_plots_name <- paste0("pan_cancer_assessment_expression_pairs_summary_clinical_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_expression_pairs_summary_clinical_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 if (nrow(d[!is.na(d[,clinical_server[input$clinical_pan_cancer_assessment_expression_pairs]]) & !is.na(d[,"sum"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   d <- d[!is.na(d[,clinical_server[input$clinical_pan_cancer_assessment_expression_pairs]]) & !is.na(d[,"sum"]),]
                   
                   if(input$clinical_pan_cancer_assessment_expression_pairs=="Age") {
                     d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                     #d[d[,"Age"]<=10,"Age"] <- "0-10"
                     d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                     d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                   }
                   
                   
                   p<-plot_ly(data=d,x=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_pairs]],
                              y=~signif(as.numeric(as.character(sum)),3), 
                              type="violin",#type="box",
                              points="all",#boxpoints="all",
                              jitter=0.3,
                              pointpos=-1.8,
                              #mode="markders",
                              color=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_pairs]],showlegend=TRUE,
                              colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                              hoveron="points+boxes",
                              legendgroup=d[,clinical_server[input$clinical_pan_cancer_assessment_expression_pairs]], 
                              text= ~paste0("Sample ID: ", PatientID,
                                            "<br>expression_pairs: ",sum,
                                            "<br>Treatment: ",treatment,
                                            "<br>Description: ",description)) %>%
                     layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                            height = 480,
                            xaxis=list(title="Clinical Attribute",
                                       titlefont=list(size=20,color="black",family= "Aril black"),
                                       showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                            yaxis=list(title="Number of positive gene pairs",
                                       titlefont=list(size=20,color="black",family= "Aril black"),
                                       showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                            ))
                   
                   incProgress(100, detail = "Doing part %")
                   p
                   
                 } 
                 #}
               })
             })
           })
           #}
           # }
         }
       }) 
     })
     
     output$pan_cancer_assessment_expression_pairs_summary_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_assessment_expression_pairs_summary_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_assessment_expression_pairs_summary_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_assessment_expression_pairs_summary_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs()
         colnames(selected.database)[which(colnames(selected.database)=="sum")] <- "F value"
         
         if (input$pan_cancer_assessment_expression_pairs_summary_type == "Excel (CSV)") {
           write.csv (selected.database,file,row.names = FALSE)
         } else if (input$pan_cancer_assessment_expression_pairs_summary_type == "Text (TSV)") {
           write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_assessment_expression_pairs_summary_type == "JSON") {
           write(toJSON(selected.database),file)
         }
         #}
       }
     )
     
     output$pan_cancer_assessment_expression_pairs_summary_table <- DT::renderDataTable({
       
       selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs()
       colnames(selected.database)[which(colnames(selected.database)=="sum")] <- "F value"
       selected.database
       #}
     })
     
     ################
     #### pan_cancer_assessment_expression_pairs plots
     ################
     
     observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
       input$do_pan_cancer_assessment_expression_pairs
       
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"sum"]),])==0) {
           
           output[["pan_cancer_assessment_expression_pairs_plots_plots"]] <- renderPlotly({
             #show("loading-c")
             #hide("loading-d")
             plot_ly () %>%
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_pairs_plots_plots_name <- paste0("pan_cancer_assessment_expression_pairs_plots_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_expression_pairs_plots_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 #show("loading-c")
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   #hide("loading-d")
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d$category) & d$category!="unknown",])==0 ) {
                   #hide("loading-d")
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                 } else {
                   d <- d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",]
                   d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","category")])))
                   colnames(d.plot)[2] <- "Groups"
                   d.plot[,"PFS.status"] <- as.character(d.plot[,"PFS.status"])
                   d.plot[,"Groups"] <- as.character(d.plot[,"Groups"])
                   d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="nonresponse","Freq"]),")",sep="" )
                   d.plot[d.plot[,"PFS.status"]=="response","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="response","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="response","Freq"]),")",sep="" )
                   d.plot[,"Groups"] <- paste(d.plot[,"Groups"]," group",sep="")
                   try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","category")])))$p.value)
                   
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     p.v <- " = NA"
                   } else {
                     # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"sum"]),as.numeric(d[d[,"Groups"]==ind[2],"sum"]))$p.value
                     # if (p.v < 0.001) {
                     #   p.v <- "< 0.001"
                     # } else {
                     p.v<-paste0(" = ",signif(p.v,2))
                     # }
                   }
                   
  
                   p <- ggplotly(ggplot(d.plot, aes(fill=Groups, y=Freq, x=PFS.status)) + ggtitle("the aggregated dataset") + 
                                   theme(plot.title = element_text(size=15,family="Aril"),axis.title.y = element_text(size=15,family="Aril"),axis.text.x=element_text(angle=-45)) +
                                   #scale_fill_manual(values=define.color[na.omit(unique(as.character(d.plot[,"Groups"])))]) + 
                                   scale_fill_discrete(name=" ") + 
                                   geom_bar( stat="identity", position="fill") +ylab("Frequency of groups") + xlab(" ") + 
                                   annotate("text",x=1.5,y=1.2,label=paste0("p-value",p.v),size=4))
                   incProgress(100, detail = "Doing part %")
                   p
                 }
                 
                 #}
               })
             })
           })
           #}
         }
       })
       
     })
     observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
       input$do_pan_cancer_assessment_expression_pairs
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression_pairs()
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_individual_expression_pairs_plots_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
         } else {
           
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
           
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_pairs_plots_plots_name <- paste0("pan_cancer_assessment_individual_expression_pairs_plots_plots", my_i)
               d<-splitted_list[[my_i]]
               output[[pan_cancer_assessment_individual_expression_pairs_plots_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   isolate({
                     if (is.na(names(splitted_list)[my_i])) {
                       incProgress(100, detail = "Doing part %")
                       plot_ly () %>%
                         layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     } else {
                       if (nrow(d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",])==0 | 
                           length(sort(unique(d$category)))<2) {
                         incProgress(100, detail = "Doing part %")
                         plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple"))
                       } else {
                         d <- d[!is.na(d$PFS.status) & !is.na(d$category) & d$category!="unknown",]
                         d.plot <- as.data.frame(unlist(table(d[,c("PFS.status","category")])))
                         colnames(d.plot)[2] <- "Groups"
                         d.plot[,"PFS.status"] <- as.character(d.plot[,"PFS.status"])
                         d.plot[,"Groups"] <- as.character(d.plot[,"Groups"])
                         d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="nonresponse","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="nonresponse","Freq"]),")",sep="" )
                         d.plot[d.plot[,"PFS.status"]=="response","PFS.status"] <- paste(d.plot[d.plot[,"PFS.status"]=="response","PFS.status"]," (n=",sum(d.plot[d.plot[,"PFS.status"]=="response","Freq"]),")",sep="" )
                         d.plot[,"Groups"] <- paste(d.plot[,"Groups"]," group",sep="")
                         try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","category")])))$p.value)
                         
                         if (class(try.test)=="try-error" | is.na(try.test)) {
                           p.v <- " = NA"
                         } else {
                           # p.v <- fisher.test(unlist(table(d[,c("PFS.status","sum")])))$p.value
                           # if (p.v < 0.001) {
                           #   p.v <- " < 0.001"
                           # } else {
                           p.v<-paste0(" = ",signif(p.v,2))
                           # }
                         }
                         p <- ggplotly(ggplot(d.plot, aes(fill=Groups, y=Freq, x=PFS.status)) + ggtitle(as.character(unique(d$data_source))) +
                                         theme(plot.title = element_text(size=15,family="Aril"),axis.title.y = element_text(size=15,family="Aril"),axis.text.x=element_text(angle=-45)) +
                                         #scale_fill_manual(values=define.color[na.omit(unique(as.character(d.plot[,"Groups"])))]) +
                                         scale_fill_discrete(name=" ") +
                                         geom_bar( stat="identity", position="fill") +ylab("Frequency of groups") + xlab(" ") +
                                         annotate("text",x=1.5,y=1.2,label=paste0("p-value",p.v),size=4))
                         incProgress(100, detail = "Doing part %")
                         p
                       }
                       
                     }
                   })
                 })
               })
             })
           }
           
         }
       })
     })
     
     ################
     #### pan_cancer_assessment_expression_pairs survival
     ################
     
     observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
       input$do_pan_cancer_assessment_expression_pairs
       #if (input$gene_pan_cancer_assessment_expression_pairs!="ALL") {
       isolate({
         input_gene_pan_cancer_assessment_expression_pairs <- input$gene_pan_cancer_assessment_expression_pairs
         selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs()
         input_adjust_OS_pan_cancer_assessment_expression_pairs <- input$adjust_OS_pan_cancer_assessment_expression_pairs
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_expression_pairs_survival_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
       
       
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_expression_pairs_survival_plots_name <- paste0("pan_cancer_assessment_expression_pairs_survival_plots", my_i)
             # pan_cancer_assessment_expression_pairs_survival_summary_tables_name <- paste0("pan_cancer_assessment_expression_pairs_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_expression_pairs_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   
                   d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                   
                   d<-data.frame(d,dead=NA)
                   d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                   d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                   d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                   d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                   #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))
                   d <- data.frame(data_source=as.character(d$data_source),
                                   dead=as.numeric(d$dead),
                                   OS=as.numeric(d$OS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   sum=as.numeric(d[,"sum"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","sum")]))
                   d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                                              nrow(d[d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,]),")"))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_expression_pairs_OS*100,"% ",
                   #                           input_gene_pan_cancer_assessment_expression_pairs," pan_cancer_assessment_expression_pairs (",
                   #                           nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_pairs_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,"IND"] <- 
                     paste0("higher than ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                            nrow(d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,]),")")
                   # d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_pairs_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_expression_pairs_OS)*100,"% ", input_gene_pan_cancer_assessment_expression_pairs," pan_cancer_assessment_expression_pairs (",
                   #         nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_pairs_OS),]),")",
                   #         sep="")
                   d<-d[order(d[,"IND"]),]
                   d[,"dead"] <- as.numeric(d[,"dead"])
                   
                   fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_expression_pairs=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_pairs),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_expression_pairs," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_pairs
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_expression_pairs_survival_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_expression_pairs_survival_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
     #}
     })
     observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
       input$do_pan_cancer_assessment_expression_pairs
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression_pairs()
         input_adjust_OS_pan_cancer_assessment_expression_pairs <- input$adjust_OS_pan_cancer_assessment_expression_pairs
         if (nrow(selected.database)==0 | nrow(selected.database[selected.database[,"category"]!="unknown",])==0) {
           output[["pan_cancer_assessment_individual_expression_pairs_survival_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
           
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_pairs_survival_plots_name <- paste0("pan_cancer_assessment_individual_expression_pairs_survival_plots", my_i)
               pan_cancer_assessment_individual_expression_pairs_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_expression_pairs_survival_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_expression_pairs_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     
                     if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"sum"]),]
                       if (length(unique(d[,"sum"]))<2) {
                         p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples because of same expression relations, please select more samples",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         
                       } else {
                         d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"sum"]),]
                         d<-data.frame(d,dead=NA)
                         d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                         d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                         d <- d[!is.na(d[,"sum"]) & !is.na(d[,"dead"]),]
                         d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                         d <- data.frame(data_source=as.character(d$data_source),
                                         dead=as.numeric(d$dead),
                                         OS=as.numeric(d$OS),
                                         Age=as.character(d$Age),
                                         Gender=as.character(d$Gender),
                                         sum=as.numeric(d[,"sum"]))
                         # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","sum")]))
                         
                         
                         d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                                                    nrow(d[d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,]),")"))
                         d[,"IND"]<-as.character(d[,"IND"])
                         d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,"IND"] <-
                           paste0("higher than ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                                  nrow(d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,]),")")
                         if (nrow(d[d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,])==0 | nrow(d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,])==0) {
                           p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples under this cutoff",showarrow=F,font=list(size=16,color="purple"))
                           table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           
                         } else {
                           d<-d[order(d[,"IND"]),]
                           #d[,"dead"] <- as.numeric(d[,"dead"])
                           
                           fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                           try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                           if (class(try.test)=="try-error" | is.na(try.test)) {
                             incProgress(100, detail = "Doing part %")
                             p1 <- plot_ly () %>%
                               layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                               add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                             table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           } else {
                            plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                           
                            if (input_adjust_OS_pan_cancer_assessment_expression_pairs=="None") {
                             diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                             if (class(diff.result.try)=="try-error") {
                               p.value <- NA
                               table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                             } else {
                               p.value <- signif(summary(diff.result)$logtest[3],2)
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                               
                             }
                           } else {
                             diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_pairs),data=d) )
                             if (class(diff.result.adjust.try)=="try-error") {
                               diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                               p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                                 " (No ",input_adjust_OS_pan_cancer_assessment_expression_pairs," was observed)")
                               
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             } else {
                               p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                               table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                               rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_pairs
                               table1 <- datatable(table1.pre,
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result.adjust)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result.adjust)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             }
                           }
                           # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                           # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                           plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                             ggplot2::geom_text(aes(label = #sprintf(" p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                      paste0("p-value = ",p.value),x=300,y=0.05)) +
                             ggplot2::guides(color=FALSE,linetype=FALSE)
                           p1 <- ggplotly(plot.result.ann,height = 450 ) %>%
                             layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                    xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                    yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                    legend=list(orientation="v",x=0.6,y=1))
                           incProgress(100, detail = "Doing part %")
                           p1
                           }
                         }
                       }
                     }
                     
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_expression_pairs_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_expression_pairs_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })

###############
#### pan_cancer_assessment_expression_pairs PFS
###############

observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
  input$do_pan_cancer_assessment_expression_pairs
  #if (input$gene_pan_cancer_assessment_expression_pairs!="ALL") {
  isolate({
    selected.database <- selecteddatabase_pan_cancer_assessment_expression_pairs()
    input_adjust_OS_pan_cancer_assessment_expression_pairs <- input$adjust_OS_pan_cancer_assessment_expression_pairs
    if (nrow(selected.database)==0 | ncol(selected.database)<13) {
      
      output[["pan_cancer_assessment_expression_pairs_PFS_plots"]] <- renderPlotly({
        plot_ly () %>%
          layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
          add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
      })
      
      
    } else {
      splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
      splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
      
      
      # for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
      local({
        # my_i <- k
        # pan_cancer_assessment_expression_pairs_survival_plots_name <- paste0("pan_cancer_assessment_expression_pairs_survival_plots", my_i)
        # pan_cancer_assessment_expression_pairs_survival_summary_tables_name <- paste0("pan_cancer_assessment_expression_pairs_survival_summary_tables", my_i)
        d<-splitted_list#[[my_i]]
        #output[[pan_cancer_assessment_expression_pairs_survival_plots_name]] <- renderPlotly({
        withProgress(message = 'Making plot', value = 0, {
          isolate({
            # if (is.na(names(splitted_list)[my_i])) {
            #   incProgress(100, detail = "Doing part %")
            #   p1 <- plot_ly () %>%
            #     layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
            #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
            #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
            # } else {
            if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0 ) {
              incProgress(100, detail = "Doing part %")
              p1 <- plot_ly () %>%
                layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
              table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
            } else {
              
              d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
              
              d <- data.frame(data_source=as.character(d$data_source),
                              PFS.status=as.character(d$PFS.status),
                              PFS=as.numeric(d$PFS),
                              Age=as.character(d$Age),
                              Gender=as.character(d$Gender),
                              sum=as.numeric(d[,"sum"]))
              # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","sum")]))
              d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                                         nrow(d[d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,]),")"))
              # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_expression_pairs_OS*100,"% ",
              #                           input_gene_pan_cancer_assessment_expression_pairs," pan_cancer_assessment_expression_pairs (",
              #                           nrow(d[d[,"sum"]<= quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_pairs_OS),]),")",
              #                           sep=""))
              d[,"IND"]<-as.character(d[,"IND"])
              d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,"IND"] <- 
                paste0("higher than ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                       nrow(d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,]),")")
              
              # d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_pairs_OS),"IND"]<-
              #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_expression_pairs_OS)*100,"% ", input_gene_pan_cancer_assessment_expression_pairs," pan_cancer_assessment_expression_pairs (",
              #         nrow(d[d[,"sum"]> quantile(d[,"sum"],input$slider_pan_cancer_assessment_expression_pairs_OS),]),")",
              #         sep="")
              d <- data.frame(d,group.IND=0)
              d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
              d<-d[order(d[,"IND"]),]
              
              fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
              #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
              
              
              try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
              if (class(try.test)=="try-error" | is.na(try.test)) {
                incProgress(100, detail = "Doing part %")
                p1 <- plot_ly () %>%
                  layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                  add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
              } else {
                plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                #plot.result <- ggsurv(fit.result)
                
                if (input_adjust_OS_pan_cancer_assessment_expression_pairs=="None") {
                  diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                  if (class(diff.result.try)=="try-error") {
                    p.value <- NA
                    table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                  } else {
                    p.value <- signif(summary(diff.result)$logtest[3],2)
                    table1 <- datatable(signif(coef(summary(diff.result)),2),
                                        caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                               signif(summary(diff.result)$logtest[1],2),
                                                                               " on ",
                                                                               signif(summary(diff.result)$logtest[2],2),
                                                                               " df, p-value = ",
                                                                               signif(summary(diff.result)$logtest[3],2)),
                                                                        br(),
                                                                        paste0("n = ",
                                                                               signif(summary(diff.result)$n,2),
                                                                               ", number of events = ",
                                                                               signif(summary(diff.result)$nevent,2)),
                                                                        style="text-align:left;color:black;font-size:15px"),
                                        options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                    
                  }
                } else {
                  diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_pairs),data=d) )
                  if (class(diff.result.adjust.try)=="try-error") {
                    diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                    p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                      " (No ",input_adjust_OS_pan_cancer_assessment_expression_pairs," was observed)")
                    
                    table1 <- datatable(signif(coef(summary(diff.result)),2),
                                        caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                               signif(summary(diff.result)$logtest[1],2),
                                                                               " on ",
                                                                               signif(summary(diff.result)$logtest[2],2),
                                                                               " df, p-value = ",
                                                                               signif(summary(diff.result)$logtest[3],2)),
                                                                        br(),
                                                                        paste0("n = ",
                                                                               signif(summary(diff.result)$n,2),
                                                                               ", number of events = ",
                                                                               signif(summary(diff.result)$nevent,2)),
                                                                        style="text-align:left;color:black;font-size:15px"),
                                        options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                  } else {
                    p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                    table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                    rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_pairs
                    table1 <- datatable(table1.pre,
                                        caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                               signif(summary(diff.result.adjust)$logtest[1],2),
                                                                               " on ",
                                                                               signif(summary(diff.result.adjust)$logtest[2],2),
                                                                               " df, p-value = ",
                                                                               signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                        br(),
                                                                        paste0("n = ",
                                                                               signif(summary(diff.result.adjust)$n,2),
                                                                               ", number of events = ",
                                                                               signif(summary(diff.result.adjust)$nevent,2)),
                                                                        style="text-align:left;color:black;font-size:15px"),
                                        options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                  }
                }
                
                
                plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                  ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                           #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                           paste0("p-value = ",p.value),x=300,y=0.05)) +
                  ggplot2::guides(color=FALSE,linetype=FALSE)
                p1<-ggplotly(plot.result.ann,
                             height = 450 ) %>%
                  layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                         xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                         yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                         legend=list(orientation="v",x=0.6,y=1))
                incProgress(100, detail = "Doing part %")
                p1
                
                
              }
            }
            #}
          })
        })
        #})
        output[["pan_cancer_assessment_expression_pairs_PFS_plots"]] <- renderPlotly({
          p1
        })
        output[["pan_cancer_assessment_expression_pairs_PFS_summary_tables"]] <- renderDataTable(
          table1
        )
      })
      #}
    }
  })
  #}
})
     observeEvent(input$do_pan_cancer_assessment_expression_pairs,{
       input$do_pan_cancer_assessment_expression_pairs
       isolate({
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_expression_pairs()
         input_adjust_OS_pan_cancer_assessment_expression_pairs <- input$adjust_OS_pan_cancer_assessment_expression_pairs
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_individual_expression_pairs_PFS_plots0"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)]
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_expression_pairs)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_expression_pairs_PFS_plots_name <- paste0("pan_cancer_assessment_individual_expression_pairs_PFS_plots", my_i)
               pan_cancer_assessment_individual_expression_pairs_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_expression_pairs_PFS_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               #output[[pan_cancer_assessment_expression_pairs_PFS_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_expression_pairs,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),])==0) {
                       incProgress(100, detail = "Doing part %")
                       p1 <- plot_ly () %>%
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no PFS data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"sum"]),]
                       if (length(unique(d[,"sum"]))<2) {
                         p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples because of same expression relations, please select more samples",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                         
                       } else {
                         d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"sum"]),]
                         d <- data.frame(data_source=as.character(d$data_source),
                                         PFS.status=as.character(d$PFS.status),
                                         PFS=as.numeric(d$PFS),
                                         Age=as.character(d$Age),
                                         Gender=as.character(d$Gender),
                                         sum=as.numeric(d[,"sum"]))
                         # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","sum")]))
                         d<-data.frame(d,IND=paste0("lower than or equal to ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                                                    nrow(d[d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,]),")"))
                         d[,"IND"]<-as.character(d[,"IND"])
                         d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,"IND"] <-
                           paste0("larger than ",input$slider_pan_cancer_assessment_expression_pairs_OS," pairs (",
                                  nrow(d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,]),")")
                         if (nrow(d[d[,"sum"]<=input$slider_pan_cancer_assessment_expression_pairs_OS,])==0 | nrow(d[d[,"sum"]>input$slider_pan_cancer_assessment_expression_pairs_OS,])==0) {
                           p1 <- plot_ly () %>% add_annotations(x=2,y=2,text="can't seperate samples under this cutoff",showarrow=F,font=list(size=16,color="purple"))
                           table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           
                         } else {
                           d <- data.frame(d,group.IND=0)
                           d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                           d<-d[order(d[,"IND"]),]
                           fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                           
                           try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                           if (class(try.test)=="try-error" | is.na(try.test)) {
                             incProgress(100, detail = "Doing part %")
                             p1 <- plot_ly () %>%
                               layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                               add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                             table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                           } else {
                            plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                           
                            if (input_adjust_OS_pan_cancer_assessment_expression_pairs=="None") {
                             diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                             if (class(diff.result.try)=="try-error") {
                               p.value <- NA
                               table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                             } else {
                               p.value <- signif(summary(diff.result)$logtest[3],2)
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                               
                             }
                           } else {
                             diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_expression_pairs),data=d) )
                             if (class(diff.result.adjust.try)=="try-error") {
                               diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                               p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                                 " (No ",input_adjust_OS_pan_cancer_assessment_expression_pairs," was observed)")
                               
                               table1 <- datatable(signif(coef(summary(diff.result)),2),
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             } else {
                               p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                               table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                               rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_expression_pairs
                               table1 <- datatable(table1.pre,
                                                   caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                          " on ",
                                                                                          signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                          " df, p-value = ",
                                                                                          signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                                   br(),
                                                                                   paste0("n = ",
                                                                                          signif(summary(diff.result.adjust)$n,2),
                                                                                          ", number of events = ",
                                                                                          signif(summary(diff.result.adjust)$nevent,2)),
                                                                                   style="text-align:left;color:black;font-size:15px"),
                                                   options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                             }
                           }
                           # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                           # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                           plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +
                             ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                      #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                      paste0("p-value = ",p.value),x=300,y=0.05)) +
                             ggplot2::guides(color=FALSE,linetype=FALSE)
                           p1 <- ggplotly(plot.result.ann,height = 480 ) %>%
                             layout(title=unique(as.character(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                    xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril")),
                                    yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril")),
                                    legend=list(orientation="v",x=0.6,y=1))
                           incProgress(100, detail = "Doing part %")
                           p1
                           }
                         }
                       }
                     }
                     
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_expression_pairs_PFS_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_expression_pairs_PFS_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     ######################################################
     #### pan-cancer assessment cibersort
     ######################################################
     
     
     # input <- list(data_source_pan_cancer_assessment_cibersort=c("Melanoma-Hugo (040 samples/-+-)","Melanoma-Riaz (139 samples/-+-)","Melanoma-Auslander (037 samples/+++)","Melanoma-Snyder-Nathanson (064 samples/+--)","Melanoma-Gide (091 samples/-++)"),
     # treatment_pan_cancer_assessment_cibersort=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
     # pan_cancer_assessment_cibersort_clinical="Gender",
     # description_pan_cancer_assessment_cibersort=c("On"),
     # immucell_pan_cancer_assessment_cibersort="Macrophages.M1",
     # dynamic_check_pan_cancer_assessment_cibersort=FALSE,
     # slider_pan_cancer_assessment_cibersort_OS=0.5,
     # adjust_OS_pan_cancer_assessment_cibersort="None")
     
      observe({
       if (input$dynamic_check_pan_cancer_assessment_cibersort) {
         
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer_assessment_cibersort",
                              choices=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                              selected=c("Melanoma-Riaz (139 samples/-+-)","Melanoma-Gide (091 samples/-++)","Melanoma-Chen-Pre-On (046 samples/+++)"),
                              server=T)
         
       }
     })
     
     observe({
       if (input$data_source_pan_cancer_assessment_cibersort=="" || is.null(input$data_source_pan_cancer_assessment_cibersort) ||
           input$treatment_pan_cancer_assessment_cibersort=="" || is.null(input$treatment_pan_cancer_assessment_cibersort) ||
           input$description_pan_cancer_assessment_cibersort=="" || is.null(input$description_pan_cancer_assessment_cibersort) ||
           input$immucell_pan_cancer_assessment_cibersort=="" || is.null(input$immucell_pan_cancer_assessment_cibersort)){
         disable("do_pan_cancer_assessment_cibersort")
       } else {
         enable("do_pan_cancer_assessment_cibersort")
       }
     })
     
     selecteddatabase_pan_cancer_assessment_cibersort <- eventReactive(input$do_pan_cancer_assessment_cibersort,{
       
      if (input$dynamic_check_pan_cancer_assessment_cibersort) {
        user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
        
        #data_source1 <- input$data_source_pan_cancer_assessment_cibersort[input$data_source_pan_cancer_assessment_cibersort %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
        data_source1 <- str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
        #treatment1 <- input$treatment_pan_cancer_assessment_cibersort
        treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_cibersort])
        description1 <- c("Pre","On")
        a<-data.table(database.test.batchcorrected)
        #a<-data.table(database.test)
        m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
        genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$immucell_pan_cancer_assessment_cibersort)
        m<-m[,colnames(m) %in% genes1,with=FALSE]
        colnames(m)[ncol(m)] <- "UserImmu"
        m <- as.data.frame(m)
        if (nrow(m) >0) {
          samples <- m$PatientID
          samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
          samples_paired <- table(samples)
          samples_paired <- samples_paired[samples_paired==2]
          samples_paired_pre <- paste0(names(samples_paired),"Pre")
          #samples_paired_on <- paste0(names(samples_paired),"On")
          samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
          m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],UserImmu=m[m[,"PatientID"] %in% samples_paired_on,"UserImmu"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserImmu"]),
                     data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],UserImmu=m[m[,"PatientID"] %in% samples_paired_on,"UserImmu"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserImmu"]))
          m<-data.frame(m,category="unknown")
          m[,"category"] <- as.character(m[,"category"])
          d <- m
          # m1 <- m %>% split(.$data_source) %>% 
          #   lapply(function(d) {#d[d[,"UserImmu"]>=median(na.omit(d[,"UserImmu"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_cibersort," pan_cancer_assessment_cibersort",sep="")
          d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]>=quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_cibersort_OS)*100,"% ",sep="")
          d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]<quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_cibersort_OS*100,"% ",sep="")
          # d})
          # m2<-ldply(m1,data.frame)
          # m2<-m2[,-1]
          # m2$UserImmu <- as.numeric(m2$UserImmu)
          # m2
          d$UserImmu <- as.numeric(d$UserImmu)
          d
        } else {
          m
        }
      } else {
       data_source1 <- str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)
       # treatment1 <- input$treatment_pan_cancer_assessment_cibersort
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_cibersort])
       description1 <- input$description_pan_cancer_assessment_cibersort
       genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$immucell_pan_cancer_assessment_cibersort)
       a<-data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       m<-m[,colnames(m) %in% genes1,with=FALSE]
       colnames(m)[ncol(m)] <- "UserImmu"
       #m<-m[,Signature:=as.numeric(Signature)]
       m <- as.data.frame(m)
       if (nrow(m)>0) {
         #        m<-data.frame(m,category.OS="no data available",category.PFS="no data available")
         m<-data.frame(m,category="unknown")
         m[,"category"] <- as.character(m[,"category"])
         d <- m
         # m1 <- m %>% split(.$data_source) %>% 
         #   lapply(function(d) {
             d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]> quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste0("top ",
                                                                                                                                                                   (1-input$slider_pan_cancer_assessment_cibersort_OS)*100,
                                                                                                                                                                   "% ",sep="")
             d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]<=quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste0("bottom ",
                                                                                                                                                                   input$slider_pan_cancer_assessment_cibersort_OS*100,
                                                                                                                                                                   "% ",sep="")
             
             d #})
         
         # m2 <- ldply(m1,data.frame)
         # m2 <- m2[,-1]
         # m2
       } else {
         m2<-m
         m2
       } 
       
      }
     })
     
     selecteddatabase_pan_cancer_assessment_individual_cibersort <- eventReactive(input$do_pan_cancer_assessment_cibersort,{
       
       if (input$dynamic_check_pan_cancer_assessment_cibersort) {
         user_sample <- na.omit(unique(c(as.character(user_clinical()[,"PatientID"]),as.character(user_transcriptome()[,"PatientID"]),as.character(unique(user_mutation()[,"PatientID"])))))
         
         #data_source1 <- input$data_source_pan_cancer_assessment_cibersort[input$data_source_pan_cancer_assessment_cibersort %in% c("Melanoma-Gide","Melanoma-Riaz",user_sample)]
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19) %in% c("Melanoma-Gide","Melanoma-Riaz","Melanoma-Chen-Pre-On",user_sample)]
         #treatment1 <- input$treatment_pan_cancer_assessment_cibersort
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_cibersort])
         description1 <- c("Pre","On")
         a<-data.table(database.test.process())
         #a<-data.table(database.test)
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$immucell_pan_cancer_assessment_cibersort)
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "UserImmu"
         m <- as.data.frame(m)
         if (nrow(m) >0) {
           samples <- m$PatientID
           samples <- c(substr(samples,1,nchar(samples)-2),substr(samples,1,nchar(samples)-3))
           samples_paired <- table(samples)
           samples_paired <- samples_paired[samples_paired==2]
           samples_paired_pre <- paste0(names(samples_paired),"Pre")
           #samples_paired_on <- paste0(names(samples_paired),"On")
           samples_paired_on <- c(paste0(names(samples_paired),"On"),paste0(names(samples_paired),"ON"))
           m <- rbind(data.frame(m[m[,"PatientID"] %in% samples_paired_pre,1:12],UserImmu=m[m[,"PatientID"] %in% samples_paired_on,"UserImmu"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserImmu"]),
                      data.frame(m[m[,"PatientID"] %in% samples_paired_on,1:12],UserImmu=m[m[,"PatientID"] %in% samples_paired_on,"UserImmu"]-m[m[,"PatientID"] %in% samples_paired_pre,"UserImmu"]))
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           #d <- m
           m1 <- m %>% split(.$data_source) %>%
             lapply(function(d) {#d[d[,"UserImmu"]>=median(na.omit(d[,"UserImmu"])),"IND"] <- paste("top 50% ", input$gene_pan_cancer_assessment_cibersort," pan_cancer_assessment_cibersort",sep="")
           d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]>=quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste("top ",(1-input$slider_pan_cancer_assessment_cibersort_OS)*100,"% ",sep="")
           d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]<quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste("bottom ",input$slider_pan_cancer_assessment_cibersort_OS*100,"% ",sep="")
           d})
           m2<-ldply(m1,data.frame)
           m2<-m2[,-1]
           m2$UserImmu <- as.numeric(m2$UserImmu)
           m2
           # d$UserImmu <- as.numeric(d$UserImmu)
           # d
         } else {
           m
         }
       } else {
         data_source1 <- str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)
         # treatment1 <- input$treatment_pan_cancer_assessment_cibersort
         treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer_assessment_cibersort])
         description1 <- input$description_pan_cancer_assessment_cibersort
         genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",input$immucell_pan_cancer_assessment_cibersort)
         a<-data.table(database.test.process())
         m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
         m<-m[,colnames(m) %in% genes1,with=FALSE]
         colnames(m)[ncol(m)] <- "UserImmu"
         #m<-m[,Signature:=as.numeric(Signature)]
         m <- as.data.frame(m)
         if (nrow(m)>0) {
           #        m<-data.frame(m,category.OS="no data available",category.PFS="no data available")
           m<-data.frame(m,category="unknown")
           m[,"category"] <- as.character(m[,"category"])
           #d <- m
           m1 <- m %>% split(.$data_source) %>%
             lapply(function(d) {
           d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]> quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste0("top ",
                                                                                                                                                              (1-input$slider_pan_cancer_assessment_cibersort_OS)*100,
                                                                                                                                                              "% ",sep="")
           d[!is.na(d[,"UserImmu"]) & d[,"UserImmu"]<=quantile(na.omit(d[,"UserImmu"]),input$slider_pan_cancer_assessment_cibersort_OS),"category"] <- paste0("bottom ",
                                                                                                                                                              input$slider_pan_cancer_assessment_cibersort_OS*100,
                                                                                                                                                              "% ",sep="")
           
           d })
           
           m2 <- ldply(m1,data.frame)
           m2 <- m2[,-1]
           m2
         } else {
           m2<-m
           m2
         } 
         
       }
     })
     #############################
     ##### pan cancer assessment cibersort summary
     ###########################################
    
     output$pan_cancer_assessment_cibersort_summary_disease_number <- renderText({
       input$do_pan_cancer_assessment_cibersort ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_assessment_cibersort()$Disease))})})
     output$pan_cancer_assessment_cibersort_summary_sample_number <- renderText({
       input$do_pan_cancer_assessment_cibersort ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_assessment_cibersort())})})
     output$pan_cancer_assessment_cibersort_summary_plots <- renderPlotly({
       summary_plot(inputdata=selecteddatabase_pan_cancer_assessment_cibersort(),feature="cibersort")
     })
     
     
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       isolate({
         input_immucell_pan_cancer_assessment_cibersort <- input$immucell_pan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_cibersort()
         if (nrow(selected.database)==0 ) {
           output[["pan_cancer_assessment_cibersort_summary_clinical_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[input$data_source_pan_cancer_assessment_cibersort]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_cibersort_summary_clinical_plots_name <- paste0("pan_cancer_assessment_cibersort_summary_clinical_plots", my_i)
             d<-splitted_list#[[my_i]]
             
             output[["pan_cancer_assessment_cibersort_summary_clinical_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = input$data_source_pan_cancer_assessment_cibersort[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 if (nrow(d[!is.na(d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]]) & !is.na(d[,"UserImmu"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no clinical data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   d <- d[!is.na(d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]]) & !is.na(d[,"UserImmu"]),]
                   if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                     d <- subset(d,description=="Pre")
                   }
                   if(input$pan_cancer_assessment_cibersort_clinical=="Age") {
                     d[,"Age"] <- as.numeric(as.character(d[,"Age"]))
                     #d[d[,"Age"]<=10,"Age"] <- "0-10"
                     d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] <- d[floor(d[,"Age"]/10)==ceiling(d[,"Age"]/10),"Age"] - 1
                     d[,"Age"] <- paste0(floor(d[,"Age"]/10)*10,"-",(ceiling(d[,"Age"]/10)*10))
                   }
                   
                   if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                     p<-plot_ly(data=d,x=d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]],
                                y=~as.numeric(as.character(UserImmu)), 
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color=d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]],showlegend=TRUE,
                                colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                                hoveron="points+boxes",
                                legendgroup=d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]], 
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Change of infiltration of immune cell: ",input_immucell_pan_cancer_assessment_cibersort," :",UserImmu,
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title="Clinical Attribute",
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste0("Change of ",input_immucell_pan_cancer_assessment_cibersort),
                                         titlefont=list(size=20,color="black",family= "Aril black"),
                                         showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              ))
                   } else {
                    p<-plot_ly(data=d,x=d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]],
                              y=~as.numeric(as.character(UserImmu)), 
                              type="violin",#type="box",
                              points="all",#boxpoints="all",
                              jitter=0.3,
                              pointpos=-1.8,
                              #mode="markders",
                              color=d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]],showlegend=TRUE,
                              colors=c("plum","lightgreen","lightblue","khaki","peru","pink"),
                              hoveron="points+boxes",
                              legendgroup=d[,clinical_server[input$pan_cancer_assessment_cibersort_clinical]], 
                              text= ~paste0("Sample ID: ", PatientID,
                                            "<br>Immune cell: ",input_immucell_pan_cancer_assessment_cibersort," :",UserImmu,
                                            "<br>Treatment: ",treatment,
                                            "<br>Description: ",description)) %>%
                     layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                            height = 480,
                            xaxis=list(title="Clinical Attribute",
                                       titlefont=list(size=20,color="black",family= "Aril black"),
                                       showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                            yaxis=list(title=input_immucell_pan_cancer_assessment_cibersort,
                                       titlefont=list(size=20,color="black",family= "Aril black"),
                                       showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                            ))
                   }
                   incProgress(100, detail = "Doing part %")
                   p
                   
                 } 
                 #}
               })
             })
           })
           #}
         }
       }) 
     })
     
     
     output$pan_cancer_assessment_cibersort_summary_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_assessment_cibersort_summary_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_assessment_cibersort_summary_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_assessment_cibersort_summary_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         selected.database <- selecteddatabase_pan_cancer_assessment_cibersort()
         if (input$dynamic_check_pan_cancer_assessment_cibersort) {
           colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- paste0("Changes of ",input$immucell_pan_cancer_assessment_cibersort)
         } else {
           colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- input$immucell_pan_cancer_assessment_cibersort
         }
         if (input$pan_cancer_assessment_cibersort_summary_type == "Excel (CSV)") {
           write.csv (selected.database,file,row.names = FALSE)
         } else if (input$pan_cancer_assessment_cibersort_summary_type == "Text (TSV)") {
           write.table (selected.database,file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_assessment_cibersort_summary_type == "JSON") {
           write(toJSON(selected.database),file)
         }
       }
     )
     
     output$pan_cancer_assessment_cibersort_summary_table <- DT::renderDataTable({
       selected.database <- selecteddatabase_pan_cancer_assessment_cibersort()
       if (input$dynamic_check_pan_cancer_assessment_cibersort) {
         colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- paste0("Changes of ",input$immucell_pan_cancer_assessment_cibersort)
       } else {
         colnames(selected.database)[which(colnames(selected.database)=="UserImmu")] <- input$immucell_pan_cancer_assessment_cibersort
       }
       selected.database
     })
     
     ##########
     #### cibersort plots
     ##########
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       isolate({
         input_immucell_pan_cancer_assessment_cibersort <- input$immucell_pan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_cibersort()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,"UserImmu"]),])==0) {
           output[["pan_cancer_assessment_cibersort_plots_plots"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database# %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[input$data_source_pan_cancer_assessment_cibersort]
           #for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_cibersort_plots_plots_name <- paste0("pan_cancer_assessment_cibersort_plots_plots", my_i)
             d<-splitted_list#[[my_i]]
             output[["pan_cancer_assessment_cibersort_plots_plots"]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   plot_ly () %>% 
                 #     layout(title = input$data_source_pan_cancer_assessment_cibersort[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 # } else {
                 #d <- splitted_list[[my_i]]
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),])==0) {
                   incProgress(100, detail = "Doing part %")
                   plot_ly () %>% 
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                 } else {
                   d[,"Groups"] <- as.character(d[,"Groups"])
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),]
                   d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                   d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                   ind <- sort(unique(d[,"Groups"]))
                   try.test <- try(p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1] & !is.na(d[,"Groups"]),"UserImmu"]),as.numeric(d[d[,"Groups"]==ind[2] & !is.na(d[,"Groups"]),"UserImmu"]))$p.value)
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     p.v <- " = NA"
                   } else {
                     # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserImmu"]),as.numeric(d[d[,"Groups"]==ind[2],"UserImmu"]))$p.value
                     # if (is.na(p.v)) {
                     #   p.v <- " = NA"
                     # } else {
                       #   if (p.v < 0.001) {
                       #     p.v <- " < 0.001"
                       #   } else {
                       p.v<-paste0(" = ",signif(p.v,2))
                     # } 
                     # }
                   }
                   if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                     p<-plot_ly(data=d,x=~Groups,
                                y=~UserImmu,
                                type="violin",#type="box",
                                points="all",#boxpoints="all",
                                jitter=0.3,
                                pointpos=-1.8,
                                #mode="markders",
                                color= ~PFS.status,
                                colors=c("#7CDBD5","#DCB239","gray"),
                                #hoveron="points+boxes",
                                legendgroup=~PFS.status,
                                showlegend=TRUE,
                                text= ~paste0("Sample ID: ", PatientID,
                                              "<br>Change of: ",input_immucell_pan_cancer_assessment_cibersort," :",UserImmu,
                                              "<br>Treatment: ",treatment,
                                              "<br>Description: ",description)) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              height = 480,
                              xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                         showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                              yaxis=list(title=paste0("Change of ",input_immucell_pan_cancer_assessment_cibersort," component"),
                                         titlefont=list(size=20,color="black",family= "Aril"),
                                         showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                              )) %>%
                       add_annotations(x=0.5,
                                       y=d[which.max(d[,"UserImmu"]),"UserImmu"],
                                       text=paste0("p-value",p.v),
                                       xref = paste0("x", 1),  # add this
                                       yref = paste0("y", 1),  # add this
                                       showarrow = FALSE,
                                       ax = 20,
                                       ay = -40,
                                       font=list(size=20,color="black",family= "Aril black"))
                   } else {
                    p<-plot_ly(data=d,x=~Groups,
                              y=~UserImmu,
                              type="violin",#type="box",
                              points="all",#boxpoints="all",
                              jitter=0.3,
                              pointpos=-1.8,
                              #mode="markders",
                              color= ~PFS.status,
                              colors=c("#7CDBD5","#DCB239","gray"),
                              #hoveron="points+boxes",
                              legendgroup=~PFS.status,
                              showlegend=TRUE,
                              text= ~paste0("Sample ID: ", PatientID,
                                            "<br>Immune cell: ",input_immucell_pan_cancer_assessment_cibersort," :",UserImmu,
                                            "<br>Treatment: ",treatment,
                                            "<br>Description: ",description)) %>%
                     layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                            height = 480,
                            xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                       showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                            yaxis=list(title=paste0(input_immucell_pan_cancer_assessment_cibersort," component"),
                                       titlefont=list(size=20,color="black",family= "Aril"),
                                       showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                            )) %>%
                     add_annotations(x=0.5,
                                     y=d[which.max(d[,"UserImmu"]),"UserImmu"],
                                     text=paste0("p-value",p.v),
                                     xref = paste0("x", 1),  # add this
                                     yref = paste0("y", 1),  # add this
                                     showarrow = FALSE,
                                     ax = 20,
                                     ay = -40,
                                     font=list(size=20,color="black",family= "Aril black"))
                   }
                   incProgress(100, detail = "Doing part %")
                   p
                 } 
                 
                 #}
               })
             })
           })
           #}
         }
       })
     })
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       isolate({
         input_immucell_pan_cancer_assessment_cibersort <- input$immucell_pan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_cibersort()
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
           output[["pan_cancer_assessment_cibersort_plots_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
         } else {
           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)]
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_cibersort_plots_plots_name <- paste0("pan_cancer_assessment_individual_cibersort_plots_plots", my_i)
               d<-splitted_list[[my_i]]
               output[[pan_cancer_assessment_individual_cibersort_plots_plots_name]] <- renderPlotly({
                 withProgress(message = 'Making plot', value = 0, {
                   if (is.na(names(splitted_list)[my_i])) {
                     plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                   } else {
                     #d <- splitted_list[[my_i]]
                     if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),])==0) {
                       plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no immunotherapy response under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     } else {
                       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"UserImmu"]),]
                       if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                         d <- subset(d,description=="Pre")
                       }
                       d[,"Groups"] <- as.character(d[,"Groups"])
                       d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),"Groups"] <- paste("response (n=",nrow(d[d[,"Groups"]=="response" & !is.na(d[,"Groups"]),]),")")
                       d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),"Groups"] <- paste("nonresponse (n=",nrow(d[d[,"Groups"]=="nonresponse" & !is.na(d[,"Groups"]),]),")")
                       ind <- sort(unique(d[,"Groups"]))
                       try.test <- try(p.v <- wilcox.test(as.numeric(as.character(d[d[,"Groups"]==ind[1],"UserImmu"])),as.numeric(as.character(d[d[,"Groups"]==ind[2],"UserImmu"])))$p.value)
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         p.v <- " = NA"
                       } else {
                         # p.v <- wilcox.test(as.numeric(d[d[,"Groups"]==ind[1],"UserImmu"]),as.numeric(d[d[,"Groups"]==ind[2],"UserImmu"]))$p.value
                         # if (p.v < 0.001) {
                         #   p.v <- " < 0.001"
                         # } else {
                         p.v<-paste0(" = ",signif(p.v,2))
                         # } 
                       }
                       
                       if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                         p <- plot_ly(data=d,x=~Groups,
                                      y=~signif(as.numeric(as.character(UserImmu)),3),
                                      type="violin",#type="box",
                                      points="all",#boxpoints="all",
                                      jitter=0.3,
                                      pointpos=-1.8,
                                      #mode="markders",
                                      color= ~PFS.status,
                                      colors=c("#7CDBD5","#DCB239","gray"),
                                      #hoveron="points+boxes",
                                      legendgroup=~PFS.status,
                                      showlegend=TRUE,
                                      text= ~paste0("Sample ID: ", PatientID,
                                                    "<br>Changes of ",input_immucell_pan_cancer_assessment_cibersort," component:",signif(as.numeric(as.character(UserImmu)),3),
                                                    "<br>Treatment: ",treatment,
                                                    "<br>Description: ",description)) %>%
                           layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                  height = 480,
                                  xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                  yaxis=list(title=paste0("Changaes of ",input_immucell_pan_cancer_assessment_cibersort," component"),
                                             titlefont=list(size=20,color="black",family= "Aril"),
                                             showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                  )) %>% 
                           add_annotations(x = 0.5,
                                           y = d[which.max(d[,"UserImmu"]),"UserImmu"],
                                           text = paste0("p-value",p.v),
                                           xref = paste0("x", 1),  # add this
                                           yref = paste0("y", 1),  # add this
                                           showarrow = FALSE,
                                           ax = 20,
                                           ay = -40,
                                           font=list(size=20,color="black",family= "Aril black"))
                       } else {
                         p <- plot_ly(data=d,x=~Groups,
                                      y=~signif(as.numeric(as.character(UserImmu)),3),
                                      type="violin",#type="box",
                                      points="all",#boxpoints="all",
                                      jitter=0.3,
                                      pointpos=-1.8,
                                      #mode="markders",
                                      color= ~PFS.status,
                                      colors=c("#7CDBD5","#DCB239","gray"),
                                      #hoveron="points+boxes",
                                      legendgroup=~PFS.status,
                                      showlegend=TRUE,
                                      text= ~paste0("Sample ID: ", PatientID,
                                                    "<br>Immune cell: ",input_immucell_pan_cancer_assessment_cibersort," :",signif(as.numeric(as.character(UserImmu)),3),
                                                    "<br>Treatment: ",treatment,
                                                    "<br>Description: ",description)) %>%
                           layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                  height = 480,
                                  xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                                             showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                                  yaxis=list(title=paste0(input_immucell_pan_cancer_assessment_cibersort," component"),
                                             titlefont=list(size=20,color="black",family= "Aril"),
                                             showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                                  )) %>% 
                           add_annotations(x = 0.5,
                                           y = d[which.max(d[,"UserImmu"]),"UserImmu"],
                                           text = paste0("p-value",p.v),
                                           xref = paste0("x", 1),  # add this
                                           yref = paste0("y", 1),  # add this
                                           showarrow = FALSE,
                                           ax = 20,
                                           ay = -40,
                                           font=list(size=20,color="black",family= "Aril black"))
                       }
                       
                       incProgress(100, detail = "Doing part %")
                       p
                     } 
                     
                   }
                 })
               })
             })
           }
         }
       })
     })
     #############
     #### cibersort survival
     #############
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       #if (input$immucell_pan_cancer_assessment_cibersort!="ALL") {
       isolate({
         input_immucell_pan_cancer_assessment_cibersort <- input$immucell_pan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_cibersort()
         input_adjust_OS_pan_cancer_assessment_cibersort <- input$adjust_OS_pan_cancer_assessment_cibersort
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_cibersort_survival_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
           
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_cibersort_survival_plots_name <- paste0("pan_cancer_assessment_cibersort_survival_plots", my_i)
             # pan_cancer_assessment_cibersort_survival_summary_tables_name <- paste0("pan_cancer_assessment_cibersort_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_cibersort_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                     d <- subset(d,description=="Pre")
                   }
                   d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),]
                   
                   d<-data.frame(d,dead=NA)
                   d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                   d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                   d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"dead"]),]
                   d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                   #d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserImmu")]))
                   d <- data.frame(data_source=as.character(d$data_source),
                                   dead=as.numeric(d$dead),
                                   OS=as.numeric(d$OS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   UserImmu=as.numeric(d[,"UserImmu"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserImmu")]))
                   d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_cibersort_OS*100,"% (",
                                             nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                                             sep=""))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_cibersort_OS*100,"% ",
                   #                           input_immucell_pan_cancer_assessment_cibersort," pan_cancer_assessment_cibersort (",
                   #                           nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),"IND"]<-
                     paste("top ",(1-input$slider_pan_cancer_assessment_cibersort_OS)*100,"% (",
                           nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                           sep="")
                   # d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_cibersort_OS)*100,"% ", input_immucell_pan_cancer_assessment_cibersort," pan_cancer_assessment_cibersort (",
                   #         nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                   #         sep="")
                   d<-d[order(d[,"IND"]),]
                   d[,"dead"] <- as.numeric(d[,"dead"])
                   
                   fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_cibersort=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_cibersort),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_cibersort," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_cibersort
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_cibersort_survival_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_cibersort_survival_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
       #}
     })
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       isolate({
         input_immucell_pan_cancer_assessment_cibersort <- input$immucell_pan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_cibersort()
         input_adjust_OS_pan_cancer_assessment_cibersort <- input$adjust_OS_pan_cancer_assessment_cibersort
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
           output[["pan_cancer_assessment_individual_cibersort_survival_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
           
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)]
           
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_cibersort_survival_plots_name <- paste0("pan_cancer_assessment_individual_cibersort_survival_plots", my_i)
               pan_cancer_assessment_individual_cibersort_survival_summary_tables_name <- paste0("pan_cancer_assessment_individual_cibersort_survival_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               d[,"UserImmu"] <- as.numeric(as.character(d[,"UserImmu"]))
               #output[[pan_cancer_assessment_cibersort_survival_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     p1 <- plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     
                     if (nrow(d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),])==0) {
                       p1 <- plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no overall survival data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                        if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                          d <- subset(d,description=="Pre")
                        }
                        d <- d[!is.na(d[,"OS.status"]) & !is.na(d[,"OS"]) & !is.na(d[,"UserImmu"]),]
                        # if (length(unique(d$category))<2) {
                        #   p1 <- plot_ly () %>% 
                        #     layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                        #     add_annotations(x=2,y=2,text="can't seperate samples into two groups, please select more samples",showarrow=F,font=list(size=16,color="purple")) 
                        #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                        #   
                        # } else {
                          
                          d[d[,"OS.status"]=="living" & !is.na(d[,"OS.status"]),"dead"] <- 0
                          d[d[,"OS.status"]=="deceased" & !is.na(d[,"OS.status"]),"dead"] <- 1
                          d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"dead"]),]
                          d <- d[d[,"dead"]==1 | d[,"dead"]==0,]
                          d <- data.frame(data_source=as.character(d$data_source),
                                       dead=as.numeric(d$dead),
                                       OS=as.numeric(d$OS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       UserImmu=as.numeric(d[,"UserImmu"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","UserImmu")]))
                       d<-data.frame(d,IND=paste0("bottom ",
                                                  input$slider_pan_cancer_assessment_cibersort_OS*100, 
                                                  "% (",
                                                  nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),
                                                  ")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),"IND"]<- 
                         paste0("top ",
                                (1-input$slider_pan_cancer_assessment_cibersort_OS)*100,
                                "% (",
                                nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),])
                                ,")")
                       d[,"dead"] <- as.numeric(d[,"dead"]) ### I don't know why I need to as.numeric since expression section doesn't need this
                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(OS,dead)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         
                       plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                       
                       if (input_adjust_OS_pan_cancer_assessment_cibersort=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(OS,dead)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           
                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(OS,dead)~IND+get(input_adjust_OS_pan_cancer_assessment_cibersort),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_cibersort," was observed)")
                           
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_cibersort
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       # diff.result <- survdiff(Surv(OS,dead)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) + 
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                         layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                legend=list(orientation="v",x=0.6,y=1)) #%>%
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_cibersort_survival_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_cibersort_survival_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     #############
     #### cibersort PFS
     #############
     
     
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       #if (input$immucellpan_cancer_assessment_cibersort!="ALL") {
       isolate({
         input_immucellpan_cancer_assessment_cibersort <- input$immucellpan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_cibersort()
         input_adjust_OS_pan_cancer_assessment_cibersort <- input$adjust_OS_pan_cancer_assessment_cibersort
         if (nrow(selected.database)==0 | ncol(selected.database)<13) {
           
           output[["pan_cancer_assessment_cibersort_PFS_plots"]] <- renderPlotly({
             plot_ly () %>%
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
           })
           
           
         } else {
           splitted_list <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list#[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)]
           
           
           # for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
           local({
             # my_i <- k
             # pan_cancer_assessment_cibersort_survival_plots_name <- paste0("pan_cancer_assessment_cibersort_survival_plots", my_i)
             # pan_cancer_assessment_cibersort_survival_summary_tables_name <- paste0("pan_cancer_assessment_cibersort_survival_summary_tables", my_i)
             d<-splitted_list#[[my_i]]
             #output[[pan_cancer_assessment_cibersort_survival_plots_name]] <- renderPlotly({
             withProgress(message = 'Making plot', value = 0, {
               isolate({
                 # if (is.na(names(splitted_list)[my_i])) {
                 #   incProgress(100, detail = "Doing part %")
                 #   p1 <- plot_ly () %>%
                 #     layout(title = str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                 #     add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple"))
                 #   table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 # } else {
                 if (nrow(d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserImmu"]),])==0 ) {
                   incProgress(100, detail = "Doing part %")
                   p1 <- plot_ly () %>%
                     layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                     add_annotations(x=2,y=2,text="no overall survival outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                   table1 <- datatable(data.frame("NA"="no overall survival outcome under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                 } else {
                   if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                     d <- subset(d,description=="Pre")
                   }
                   d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"PFS"]) & !is.na(d[,"UserImmu"]),]
                   
                   d <- data.frame(data_source=as.character(d$data_source),
                                   PFS.status=as.character(d$PFS.status),
                                   PFS=as.numeric(d$PFS),
                                   Age=as.character(d$Age),
                                   Gender=as.character(d$Gender),
                                   UserImmu=as.numeric(d[,"UserImmu"]))
                   # d<-data.frame(d[,1:6],unfactor(d[,c("dead","OS","PFS","Age","Gender","UserImmu")]))
                   d<-data.frame(d,IND=paste("bottom ",input$slider_pan_cancer_assessment_cibersort_OS*100,"% (",
                                             nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                                             sep=""))
                   # d<-data.frame(d,IND=paste("samples with bottom ",input$slider_pan_cancer_assessment_cibersort_OS*100,"% ",
                   #                           input_immucellpan_cancer_assessment_cibersort," pan_cancer_assessment_cibersort (",
                   #                           nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                   #                           sep=""))
                   d[,"IND"]<-as.character(d[,"IND"])
                   d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),"IND"]<-
                     paste("top ",(1-input$slider_pan_cancer_assessment_cibersort_OS)*100,"% (",
                           nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                           sep="")
                   # d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),"IND"]<-
                   #   paste("samples with top ",(1-input$slider_pan_cancer_assessment_cibersort_OS)*100,"% ", input_immucellpan_cancer_assessment_cibersort," pan_cancer_assessment_cibersort (",
                   #         nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),")",
                   #         sep="")
                   d <- data.frame(d,group.IND=0)
                   d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                   d<-d[order(d[,"IND"]),]
                   
                   fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                   #fit.result <- survfit(coxph(Surv(OS,dead)~IND+Age,data=d))
                   
                   
                   try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                   if (class(try.test)=="try-error" | is.na(try.test)) {
                     incProgress(100, detail = "Doing part %")
                     p1 <- plot_ly () %>%
                       layout(title = "the aggregated dataset",font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                     table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                   } else {
                     plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                     #plot.result <- ggsurv(fit.result)
                     
                     if (input_adjust_OS_pan_cancer_assessment_cibersort=="None") {
                       diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                       if (class(diff.result.try)=="try-error") {
                         p.value <- NA
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                         p.value <- signif(summary(diff.result)$logtest[3],2)
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         
                       }
                     } else {
                       diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_cibersort),data=d) )
                       if (class(diff.result.adjust.try)=="try-error") {
                         diff.result <- coxph(Surv(OS,dead)~IND,data=d)
                         p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                           " (No ",input_adjust_OS_pan_cancer_assessment_cibersort," was observed)")
                         
                         table1 <- datatable(signif(coef(summary(diff.result)),2),
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       } else {
                         p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                         table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                         rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_cibersort
                         table1 <- datatable(table1.pre,
                                             caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                    " on ",
                                                                                    signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                    " df, p-value = ",
                                                                                    signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                             br(),
                                                                             paste0("n = ",
                                                                                    signif(summary(diff.result.adjust)$n,2),
                                                                                    ", number of events = ",
                                                                                    signif(summary(diff.result.adjust)$nevent,2)),
                                                                             style="text-align:left;color:black;font-size:15px"),
                                             options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                       }
                     }
                     
                     
                     plot.result.ann <- plot.result + theme(legend.position = "bottom", legend.title = element_blank()) +
                       ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                #pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05)) +
                                                paste0("p-value = ",p.value),x=300,y=0.05)) +
                       ggplot2::guides(color=FALSE,linetype=FALSE)
                     p1<-ggplotly(plot.result.ann,
                                  height = 450 ) %>%
                       layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                              xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                              yaxis=list(title="Overall survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                              legend=list(orientation="v",x=0.6,y=1))
                     incProgress(100, detail = "Doing part %")
                     p1
                     
                     
                   }
                 }
                 #}
               })
             })
             #})
             output[["pan_cancer_assessment_cibersort_PFS_plots"]] <- renderPlotly({
               p1
             })
             output[["pan_cancer_assessment_cibersort_PFS_summary_tables"]] <- renderDataTable(
               table1
             )
           })
           #}
         }
       })
       #}
     })
     observeEvent(input$do_pan_cancer_assessment_cibersort,{
       input$do_pan_cancer_assessment_cibersort
       isolate({
         input_immucell_pan_cancer_assessment_cibersort <- input$immucell_pan_cancer_assessment_cibersort
         selected.database <- selecteddatabase_pan_cancer_assessment_individual_cibersort()
         input_adjust_OS_pan_cancer_assessment_cibersort <- input$adjust_OS_pan_cancer_assessment_cibersort
         if (nrow(selected.database)==0 | nrow(selected.database[!is.na(selected.database[,13]),])==0) {
           output[["pan_cancer_assessment_individual_cibersort_PFS_plots0"]] <- renderPlotly({
             plot_ly () %>% 
               layout(title = paste(str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19),collapse = ","),font=list(size=15,family="Aril")) %>%
               add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
           })
           
         } else {
           splitted_list <- selected.database %>% as.matrix %>% as.data.frame %>% split(.$data_source)
           splitted_list <- splitted_list[str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)]
           
           for (k in 1:length(input$data_source_pan_cancer_assessment_cibersort)) {
             local({
               my_i <- k
               pan_cancer_assessment_individual_cibersort_PFS_plots_name <- paste0("pan_cancer_assessment_individual_cibersort_PFS_plots", my_i)
               pan_cancer_assessment_individual_cibersort_PFS_summary_tables_name <- paste0("pan_cancer_assessment_individual_cibersort_PFS_summary_tables", my_i)
               d<-splitted_list[[my_i]]
               d[,"UserImmu"] <- as.numeric(as.character(d[,"UserImmu"]))
               #output[[pan_cancer_assessment_cibersort_PFS_plots_name]] <- renderPlotly({
               withProgress(message = 'Making plot', value = 0, {
                 isolate({
                   if (is.na(names(splitted_list)[my_i])) {
                     p1 <- plot_ly () %>% 
                       layout(title = str_sub(input$data_source_pan_cancer_assessment_cibersort,end=-19)[[my_i]],font=list(size=15,family="Aril")) %>%
                       add_annotations(x=2,y=2,text="no data available under this selection",showarrow=F,font=list(size=16,color="purple")) 
                     table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                     
                   } else {
                     
                     #if (nrow(d[!is.na(d[,"PFS.status"]),])==0 | nrow(d[!is.na(d[,"PFS"]),])==0 | nrow(d[!is.na(d[,"UserImmu"]),])==0 | sum(d[!is.na(d[,"UserImmu"]),"UserImmu"])==0) {
                     if (nrow(d[!is.na(d[,"UserImmu"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),])==0) {
                       p1 <- plot_ly () %>% 
                         layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                         add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
                       table1 <- datatable(data.frame("NA"="no data available under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       
                     } else {
                       d <- d[!is.na(d[,"UserImmu"]) & !is.na(d[,"PFS"]) & !is.na(d[,"PFS.status"]),]
                       if (input$dynamic_check_pan_cancer_assessment_cibersort) {
                         d <- subset(d,description=="Pre")
                       }
                       d <- data.frame(data_source=as.character(d$data_source),
                                       PFS.status=as.character(d$PFS.status),
                                       PFS=as.numeric(d$PFS),
                                       Age=as.character(d$Age),
                                       Gender=as.character(d$Gender),
                                       UserImmu=as.numeric(d[,"UserImmu"]))
                       # d<-data.frame(d[,1:6],unfactor(d[,c("OS.status","OS","PFS","Age","UserImmu")]))
                       d<-data.frame(d,IND=paste0("bottom ",
                                                  input$slider_pan_cancer_assessment_cibersort_OS*100, 
                                                  "% (",
                                                  nrow(d[d[,"UserImmu"]<= quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),]),
                                                  ")"))
                       d[,"IND"]<-as.character(d[,"IND"])
                       d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),"IND"]<- 
                         paste0("top ",
                                (1-input$slider_pan_cancer_assessment_cibersort_OS)*100,
                                "% (",
                                nrow(d[d[,"UserImmu"]> quantile(d[,"UserImmu"],input$slider_pan_cancer_assessment_cibersort_OS),])
                                ,")")
                       d <- data.frame(d,group.IND=0)
                       d[d[,"PFS.status"]=="nonresponse","group.IND"] <- 1
                       
                       d<-d[order(d[,"IND"]),]
                       fit.result <- survfit(Surv(PFS,group.IND)~IND,data=d)
                       
                       try.test <- try(plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5")))
                       if (class(try.test)=="try-error" | is.na(try.test)) {
                         incProgress(100, detail = "Doing part %")
                         p1 <- plot_ly () %>%
                           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
                           add_annotations(x=2,y=2,text="no enough samples under this selection",showarrow=F,font=list(size=16,color="purple"))
                         table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE)
                       } else {
                       plot.result <- ggsurv(fit.result,surv.col = c("#DCB239","#7CDBD5"))
                       
                       if (input_adjust_OS_pan_cancer_assessment_cibersort=="None") {
                         diff.result.try <- try( diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d) )
                         if (class(diff.result.try)=="try-error") {
                           p.value <- NA
                           table1 <- datatable(data.frame("NA"="no enough samples under this selection"),options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=FALSE,escape = FALSE) 
                         } else {
                           p.value <- signif(summary(diff.result)$logtest[3],2)
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                           
                         }
                       } else {
                         diff.result.adjust.try <- try( diff.result.adjust <- coxph(Surv(PFS,group.IND)~IND+get(input_adjust_OS_pan_cancer_assessment_cibersort),data=d) )
                         if (class(diff.result.adjust.try)=="try-error") {
                           diff.result <- coxph(Surv(PFS,group.IND)~IND,data=d)
                           p.value <- paste0(signif(summary(diff.result)$logtest[3],2),
                                             " (No ",input_adjust_OS_pan_cancer_assessment_cibersort," was observed)")
                           
                           table1 <- datatable(signif(coef(summary(diff.result)),2),
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         } else {
                           p.value=signif(summary(diff.result.adjust)$logtest[3],2)
                           table1.pre <- signif(coef(summary(diff.result.adjust))[1:2,],2)
                           rownames(table1.pre)[2] <- input_adjust_OS_pan_cancer_assessment_cibersort
                           table1 <- datatable(table1.pre,
                                               caption=htmltools::tags$caption(paste0("Likelihood ration test = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[1],2),
                                                                                      " on ",
                                                                                      signif(summary(diff.result.adjust)$logtest[2],2),
                                                                                      " df, p-value = ",
                                                                                      signif(summary(diff.result.adjust)$logtest[3],2)),
                                                                               br(),
                                                                               paste0("n = ",
                                                                                      signif(summary(diff.result.adjust)$n,2),
                                                                                      ", number of events = ",
                                                                                      signif(summary(diff.result.adjust)$nevent,2)),
                                                                               style="text-align:left;color:black;font-size:15px"),
                                               options = list(searching = FALSE,paging = FALSE,info=FALSE),rownames=TRUE,escape = FALSE)
                         }
                       }
                       # diff.result <- survdiff(Surv(PFS,group.IND)~IND,data=d)
                       # p.value= 1 - pchisq(diff.result$chisq, length(diff.result$n) - 1)
                       plot.result.ann <- plot.result + theme(legend.position = "right", legend.title = element_blank()) +  
                         ggplot2::geom_text(aes(label = #sprintf("p-value = %0.2g",
                                                  #       pchisq(diff.result$chisq,df=1,lower.tail = F)),x=300,y=0.05))
                                                  paste0("p-value = ",p.value),x=300,y=0.05)) +
                         ggplot2::guides(color=FALSE,linetype=FALSE)
                       p1<-ggplotly(plot.result.ann,height = 450 ) %>%
                         layout(title = as.character(unique(d$data_source)),titlefont=list(size=20,color="black",family= "Aril"),
                                xaxis=list(title="Days",titlefont=list(size=20,color="black",family= "Aril black")),
                                yaxis=list(title="Progression free survival proportion",titlefont=list(size=20,color="black",family= "Aril black")),
                                legend=list(orientation="v",x=0.6,y=1)) #%>%
                       incProgress(100, detail = "Doing part %")
                       p1
                       }
                     }
                   }
                 })
               })
               #})
               output[[pan_cancer_assessment_individual_cibersort_PFS_plots_name]] <- renderPlotly({
                 p1
               })
               output[[pan_cancer_assessment_individual_cibersort_PFS_summary_tables_name]] <- renderDataTable(
                 table1
               )
             })
           }
         }
       })
     })
     
     #######################################################
     #### pan-cancer
     #######################################################
     observe({
       ### when all samples
       if (input$sample_pan_cancer=="custom selection") {
         enable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              server=T)
         enable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              #selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              server = T)
         enable("description_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              #selected=c("Pre","On","Unknown"),
                              server = T)
       }
       if (input$sample_pan_cancer=="all samples") {
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              selected = meta_analysis_prio_choices,
                              server=T)
         disable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              server = T)
         disable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              selected=c("Pre","On","Unknown"),
                              server = T)
         disable("description_pan_cancer")
       }
       ### when all sample with anti-CTLA-4
       if (input$sample_pan_cancer=="samples with anti-CTLA-4 treatment") {
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              selected = meta_analysis_prio_selected_anti_CTLA_4,
                              server=T)
         disable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected="anti-CTLA-4",
                              server = T)
         disable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              selected=c("Pre","On","Unknown"),
                              server = T)
         disable("description_pan_cancer")
       }
       ### when all samples with anti-PD-1
       if (input$sample_pan_cancer=="samples with anti-PD-1/L1 treatment") {
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              selected = meta_analysis_prio_selected_anti_PD_1,
                              server=T)
         disable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected="anti-PD-1/L1",
                              server = T)
         disable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              selected=c("Pre","On","Unknown"),
                              server = T)
         disable("description_pan_cancer")
       }

       #### when samples before treatment
       if (input$sample_pan_cancer=="samples before treatment") {
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              selected = meta_analysis_prio_selected_before_treatment,
                              server=T)
         disable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              server = T)
         disable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              selected="Pre",
                              server = T)
         disable("description_pan_cancer")
       }
       #### when sample after treatment 
       if (input$sample_pan_cancer=="samples after treatment") {
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              selected = meta_analysis_prio_selected_after_treatment,
                              server=T)
         disable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              server = T)
         disable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              selected="On",
                              server = T)
         disable("description_pan_cancer")
       }
       ### when case study 1
       if (input$sample_pan_cancer=="case study 1") {
         updateSelectizeInput(session=session,inputId="data_source_pan_cancer",
                              choices = meta_analysis_prio_choices,
                              selected = meta_analysis_prio_selected_case_study_1,
                              server=T)
         disable("data_source_pan_cancer")
         updateSelectizeInput(session=session,inputId="treatment_pan_cancer",
                              choices=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              selected=c("anti-PD-1/L1","anti-CTLA-4","anti-PD-1/L1+anti-CTLA-4"),
                              server = T)
         disable("treatment_pan_cancer")
         updateSelectizeInput(session=session,inputId="description_pan_cancer",
                              choices=c("Pre","On","Unknown"),
                              selected=c("Pre","On","Unknown"),
                              server = T)
         disable("description_pan_cancer")
       }
     })
     
     observe({
       if (input$data_source_pan_cancer=="" || is.null(input$data_source_pan_cancer) ||
           input$treatment_pan_cancer=="" || is.null(input$treatment_pan_cancer) ||
           input$description_pan_cancer=="" || is.null(input$description_pan_cancer)){
         disable("do_pan_cancer")
       } else {
         enable("do_pan_cancer")
       }
     })
     # selecteddatabase_pan_cancer <- eventReactive(input$do_pan_cancer,{
     #   data_source1 <- str_sub(input$data_source_pan_cancer,end=-19)
     #   # treatment1 <- input$treatment_pan_cancer
     #   treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer])
     #   description1 <- input$description_pan_cancer
     #   genes1 <- c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",
     #               "Signature.1","Signature.2","Signature.3","Signature.4","Signature.5","Signature.6","Signature.7","Signature.8","Signature.9",
     #               "Signature.10","Signature.11","Signature.12","Signature.13","Signature.14","Signature.15","Signature.16","Signature.17","Signature.18",
     #               "Signature.19","Signature.20","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25","Signature.26","Signature.27",
     #               "Signature.28","Signature.29","Signature.30","unknown","B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive",
     #               "T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta",
     #               "NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting",
     #               "Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")
     #   a<-data.table(database.test.process())
     #   m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
     #   m<-m[,colnames(m) %in% genes1,with=FALSE]
     #   as.data.frame(m)
     # 
     # })
     
     selecteddatabase_pan_cancer_batchcorrected <- eventReactive(input$do_pan_cancer,{
       data_source1 <- str_sub(input$data_source_pan_cancer,end=-19)
       # treatment1 <- input$treatment_pan_cancer
       treatment1 <- names(update_treatment_name[update_treatment_name %in% input$treatment_pan_cancer])
       description1 <- input$description_pan_cancer
       # genes1 <- #c("Disease","PatientID","data_source","treatment","description","PFS.status","RECIST","OS.status","OS","PFS","Age","Gender",
       #             c("Mutation.Load","
       #              Signature.1","Signature.2","Signature.3","Signature.4","Signature.5","Signature.6","Signature.7","Signature.8","Signature.9",
       #             "Signature.10","Signature.11","Signature.12","Signature.13","Signature.14","Signature.15","Signature.16","Signature.17","Signature.18",
       #             "Signature.19","Signature.20","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25","Signature.26","Signature.27",
       #             "Signature.28","Signature.29","Signature.30","unknown","B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive",
       #             "T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta",
       #             "NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting",
       #             "Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")
       a<-data.table(database.test.batchcorrected)
       m<-a[data_source %in% data_source1 & treatment %in% treatment1 & description %in% description1]
       # m<-m[,colnames(m) %ni% genes1,with=FALSE]
       as.data.frame(m)
       
     })
     
     ########################
     #### pan-cancer summary
     ########################
     
     output$pan_cancer_summary_disease_number <- renderText({
       input$do_pan_cancer ##keep these two rows
       isolate({ ##keep these two rows
         length(unique(selecteddatabase_pan_cancer_batchcorrected()$Disease))})})
     output$pan_cancer_summary_sample_number <- renderText({
       input$do_pan_cancer ##keep these two rows
       isolate({ ##keep these two rows
         nrow(selecteddatabase_pan_cancer_batchcorrected())})})
     output$pan_cancer_summary_plots <- renderPlotly({
       summary_plot_browse(inputdata=selecteddatabase_pan_cancer_batchcorrected()[,1:12])
     })
     
     #######################
     #### pan-cancer mutation
     #######################

     # input <- list(data_source_pan_cancer=c( "Melanoma-Hugo (040 samples/-+-)",
     #                                           "Melanoma-Riaz (139 samples/-+-)",
     #                                           "Melanoma-Snyder-Nathanson (064 samples/+--)",
     #                                           "Melanoma-Van Allen (112 samples/+--)",
     #                                           #"Melanoma-Jerby-Arnon-ValCo2 (112 samples/-+-)",
     #                                           "Colorectal Cancer-Le (031 samples/-+-)",
     #                                           "Metastatic Urothelial Cancer-Mariathasan (348 samples/-+-)",
     #                                           "Non-Small Cell Lung Cancer-Naiyer Rizvi (035 samples/-+-)"),
     #               treatment_pan_cancer=c("anti-CTLA-4","anti-PD-1/L1","anti-PD-1/L1+anti-CTLA-4"),
     #               description_pan_cancer=c("Pre","On"))

     #selected.mutation <- merge(all.mutations,selecteddatabase_pan_cancer[,1:12],by="PatientID",all.x=T)
     #selected.mutation.res <- selected.mutation[selected.mutation[,"PFS.status"]=="response",]
     #selected.mutation.nonres <- selected.mutation[selected.mutation[,"PFS.status"]=="nonresponse",]
     selecteddatabase_pan_cancer_mutation <- eventReactive(input$do_pan_cancer,{
       withProgress(message = 'Preparing data', value = 0, {

       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       selected.mutations <- all.mutations.pan.cancer[all.mutations.pan.cancer[,"PatientID"] %in% selected.database[,"PatientID"],]

       selected.mutations.res <- selected.mutations[selected.mutations[,"PFS.status"]=="response" & !is.na(selected.mutations[,"PFS.status"]),]
       selected.mutations.nonres <- selected.mutations[selected.mutations[,"PFS.status"]=="nonresponse" & !is.na(selected.mutations[,"PFS.status"]),]

       subset.p.value <- NULL
       d.nonna <- NULL

       if (nrow(selected.mutations.res)!=0 & nrow(selected.mutations.nonres)!=0) {
          subset.genes <- as.character(unique(selected.mutations[,"gene"]))
          for (j in 1:length(subset.genes)) {
            # print(j)
            incProgress(j/length(subset.genes), detail = "Doing part %")
            try.test <- try(p.value <- fisher.test (matrix(c(length(unique(selected.mutations.res[selected.mutations.res[,"gene"]==subset.genes[j],"PatientID"])),
                                                             length(unique(selected.mutations.nonres[selected.mutations.nonres[,"gene"]==subset.genes[j],"PatientID"])),
                                                             length(unique(selected.mutations.res[,"PatientID"]))-length(unique(selected.mutations.res[selected.mutations.res[,"gene"]==subset.genes[j],"PatientID"])),
                                                             length(unique(selected.mutations.nonres[,"PatientID"]))-length(unique(selected.mutations.nonres[selected.mutations.nonres[,"gene"]==subset.genes[j],"PatientID"]))),2,2))$p.value)
            if (class(try.test)=="try-error" | is.na(p.value)) {
              p.value <- "NA"
            } 
            subset.p.value <- c(subset.p.value,p.value)
          }
          d.nonna <- data.frame(p.value=signif(as.numeric(subset.p.value),2),log.p.value= signif(-log10(as.numeric(subset.p.value)),2),FDR=signif(p.adjust(as.numeric(subset.p.value),"fdr"),2))
          rownames(d.nonna) <- subset.genes
          d.nonna <- d.nonna[order(d.nonna[,1]),]
        }
        d.nonna

     })
     })


     output$pan_cancer_mutation_plots <- renderPlotly({
       withProgress(message = 'Making plot', value = 0, {
         if (input$sample_pan_cancer=="custom selection") {
           d.nonna <- selecteddatabase_pan_cancer_mutation()
         } else {
           d.nonna <- pan_cancer_prio_precalculated_mutation[[input$sample_pan_cancer]]
         }
        # d.nonna <- selecteddatabase_pan_cancer_mutation()

         if (is.null(d.nonna)) {
           incProgress(100, detail = "Doing part %")
           plot_ly () %>%
             layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
         } else {
           vline <- function(x = 0, color = "red") {
             list(
               type = "line",
               y0 = 0,
               y1 = 1,
               yref = "paper",
               x0 = x,
               x1 = x,
               line = list(color = color)
             )
           }
           fit <- density(na.omit(d.nonna[,2]))
           p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
             layout(title="the aggregated dataset",
                    height = 480,
                    xaxis = list(title="-log10(p.value)",
                                 titlefont=list(size=20,color="black",family= "Aril black"),
                                 showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    yaxis=list(title="Density",
                               titlefont=list(size=20,color="black",family= "Aril black"),
                               showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    shapes=list(vline(1.3))) %>%
             add_annotations(x = 1.3,
                             y = 1,
                             text = "p.value = 0.05",
                             xref = paste0("x", 1),  # add this
                             yref = paste0("y", 1),  # add this
                             showarrow = FALSE,
                             ax = 20,
                             ay = -40,
                             font=list(size=20,color="black",family= "Aril black"))
           incProgress(100, detail = "Doing part %")
           p
         }
       })
     })



     output$pan_cancer_mutation_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_mutation_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_mutation_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_mutation_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         if (input$sample_pan_cancer=="custom selection") {
           d.nonna <- selecteddatabase_pan_cancer_mutation()
         } else {
           d.nonna <- pan_cancer_prio_precalculated_mutation[[input$sample_pan_cancer]]
         }
         
         if (input$pan_cancer_mutation_type == "Excel (CSV)") {
           write.csv (d.nonna,file,row.names = TRUE)
         } else if (input$pan_cancer_mutation_type == "Text (TSV)") {
           write.table (d.nonna,file,quote=F, row.names = TRUE, sep="\t")
         } else if (input$pan_cancer_mutation_type == "JSON") {
           write(toJSON(d.nonna),file)
         }
       }
     )

     # output$pan_cancer_mutation_table <- DT::renderDataTable({
     #   selecteddatabase_pan_cancer_mutation()
     # })

     observeEvent(input$do_pan_cancer,{
       input$do_pan_cancer
       isolate({
         if (input$sample_pan_cancer=="custom selection") {
           d.nonna.t.p0.05 <- selecteddatabase_pan_cancer_mutation()
         } else {
           d.nonna.t.p0.05 <- pan_cancer_prio_precalculated_mutation[[input$sample_pan_cancer]]
         }
         # d.nonna.t.p0.05 <- selecteddatabase_pan_cancer_mutation()

         shinyInput <- function(FUN, len, id, ...) {
           inputs <- character(len)
           for (i in seq_len(len)) {
             inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
             #inputs[i] <- as.character(FUN(paste0(id,"_",i), ...))
           }
           inputs
         }

         d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
           #Signature=rownames(d.nonna.t.p0.05),
           #p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
           #FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
           signif(d.nonna.t.p0.05[,c(1,3)],2),
           #View=as.character(actionButton("test","View")),
           View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),"pan-cancer_",label="View",onclick='Shiny.onInputChange(\"select_button\",  this.id)'),
           stringsAsFactors = FALSE#,
           #row.names = 1:nrow(d.nonna.t.p0.05)
         ))

         pan_cancer_mutation_tables_data <- d.nonna.t.p0.05.test$data

         output$pan_cancer_mutation_table <- DT::renderDataTable(
           pan_cancer_mutation_tables_data,server = FALSE, escape = FALSE, selection = 'none'
         )
       })

     })

     observeEvent(input$select_button,{
       s <- as.numeric(strsplit(input$select_button, "_")[[1]][3])
       Viewname <- as.character(strsplit(input$select_button, "_")[[1]][2])
       output$pan_cancer_mutation_modals <- renderUI({
         tagList(
           bsModal(paste('pan_cancer_mutation_model', s ,sep=''), Viewname, "select_button", size = "large",

                   plotlyOutput("browse_pan_cancer_mutation_bsModal_plot")
                   #textOutput("browse_pan_cancer_mutation_bsModal_plot")
           )
         )
       })
       toggleModal(session,paste('pan_cancer_mutation_model', s ,sep=''), toggle = Viewname)
       ##Reset the select_button
       #session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_pan_cancer_mutation")
     })

     observeEvent(input$select_button,{
       # output$browse_pan_cancer_mutation_bsModal_plot <- renderText({
       #   input$select_button
       # })
       output$browse_pan_cancer_mutation_bsModal_plot <- renderPlotly({
         isolate({
           input_gene_mutation <- as.character(strsplit(input$select_button, "_")[[1]][2])
           selected.database <- selecteddatabase_pan_cancer_batchcorrected()
           selected.mutations <- all.mutations[all.mutations[,"PatientID"] %in% selected.database[,"PatientID"],]
           if (nrow(selected.database)==0) {
             selected.mutations.function <- selected.database
           } else {
             gene1 <- input_gene_mutation
             selected.mutations.1 <- selected.mutations[selected.mutations[,"gene"] %in% gene1,]
             if (nrow(selected.mutations.1)>0 ) {
               #### important: resharp mutation file to database.test format
               b<-aggregate(variant_class~(PatientID+gene),data=selected.mutations.1,FUN=paste0)
               c <- b %>% spread(gene,variant_class,fill="no mutation")
               c[c=="NULL"]<-"no mutation"

               if ((ncol(c)-1) < length(gene1)) {
                 c.add <- matrix("no mutation",nrow(c),length(gene1[gene1 %ni% colnames(c)[2:ncol(c)]]))
                 colnames(c.add) <- gene1[gene1 %ni% colnames(c)[2:ncol(c)]]
                 c <- cbind(c, c.add)
               }

               c.rest <- data.frame(PatientID=unique(selected.mutations[selected.mutations[,"PatientID"] %ni% selected.mutations.1[,"PatientID"],"PatientID"]))
               c <- rbind.fill(c,c.rest)
               c[c=="NULL"] <- "no mutation"
               c[is.na(c)] <- "no mutation"

               #c <- data.frame(c,category=apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") }))
               c <- data.frame(c,category="no mutation")
               c[,"category"] <- as.character(c[,"category"])
               c[!apply(c,1,function(x) {paste(unique(as.character(x[2:(2+length(gene1))])),collapse = "") == "no mutation" }),"category"]<- "mutated"

               selected.database <- merge(selected.database,c,by="PatientID",all.x=T)
               selected.database[is.na(selected.database[,"category"]),"category"] <- "unknown"
               selected.mutations.function <- as.data.frame(selected.database)
             } else {
               selected.database <- data.frame(selected.database,category="unknown")
               selected.mutations.function <- selected.database
             }
           }

           selected.database <- data.frame(selected.database,Groups=selected.database$PFS.status)
           d <- selected.database #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           d.mutations <- selected.mutations #%>% as.matrix %>% as.data.frame %>% split(.$data_source)
           d.mutations.function <- selected.mutations.function #%>% split(.$data_source)

           d.mutations.function <- d.mutations.function[!is.na(d.mutations.function[,"PFS.status"]) & d.mutations.function[,"category"]!="unknown",]

           d.table <- unlist(table(d.mutations.function[,c("PFS.status","category")]))
           d.table <- d.table[rownames(d.table) %in% c("nonresponse","response"),colnames(d.table) %in% c("mutated","no mutation")]
           try.test <- try(p.v <- fisher.test(d.table)$p.value)
           #try.test <- try(p.v <- fisher.test(unlist(table(d[,c("PFS.status","IND")]))[1:2,])$p.value)
           if (class(try.test)=="try-error" | is.na(try.test)) {
             p.v <- " = NA"
           } else {
             p.v<-paste0(" = ",signif(p.v,2))
           }

           ind <- sort(unique(d[,"Groups"]))
           d.mutations.res <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[2] & !is.na(d[,"Groups"])),"PatientID"] ,]
           d.mutations.nonres <- d.mutations[d.mutations[,"PatientID"] %in% d[(d[,"Groups"]==ind[1] & !is.na(d[,"Groups"])),"PatientID"] ,]

           d.mutations.res<-d.mutations.res[,-1]
           #d.mutations.res<-d.mutations.res[,c("PatientID","gene","variant_class","Protein_change","Chromosome","data_source")]
           d.mutations.res<-d.mutations.res[,c(6,2:5,7)]
           colnames(d.mutations.res)[1] <- "sample"
           if (nrow(d.mutations.res[d.mutations.res[,"gene"] %in% input_gene_mutation,])==0) {
             p1 <- ggplot(d.mutations.res, aes(x=sample,y="NA")) +
               xlab(paste0("response samples (n=",length(unique(d.mutations.res$sample)),")")) +
               theme(axis.text.x=element_text(angle=50, hjust=1,size=4),
                     axis.text.y=element_text(size=8,colour='black'),
                     axis.title.x=element_text(size=11),
                     axis.title.y=element_blank())
             p.bar.res <- data.frame(PFS.status="response",genes=input_gene_mutation,Number=0,n=length(unique(d.mutations.res$sample)))
             p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
           } else {
             p1<-#ggplotly(
               waterfall(d.mutations.res, fileType = "Custom",
                         variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))), ### here all.mutation is important, if not, different data source will have different color enve in the same mutation type
                         plotGenes=unlist(str_split(input_gene_mutation,",")),
                         mainXlabel = TRUE,my_x_label=ind[2],
                         plotMutBurden = FALSE,returnp="p1",
                         mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                       '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                       '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                       '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                       '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                       '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                       '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                       '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
             Number.genes.res <- NULL
             for (j in 1:length(input_gene_mutation)) {
               Number.genes.res <- c(Number.genes.res,length(unique(d.mutations.res[d.mutations.res[,"gene"]==input_gene_mutation[j],"sample"])))
             }
             p.bar.res <- data.frame(PFS.status="response",genes=input_gene_mutation,Number=Number.genes.res,n=length(unique(d.mutations.res$sample)))
             p.bar.res <- p.bar.res[order(p.bar.res$genes,decreasing = TRUE),]
           }
           d.mutations.nonres<-d.mutations.nonres[,-1]
           d.mutations.nonres<-d.mutations.nonres[,c(6,2:5,7)]
           colnames(d.mutations.nonres)[1] <- "sample"
           if (nrow(d.mutations.nonres[d.mutations.nonres[,"gene"] %in% input_gene_mutation,])==0) {
             q1 <- ggplot(d.mutations.nonres, aes(x=sample,y="NA")) +
               xlab(paste0("nonresponse samples (n=",length(unique(d.mutations.nonres$sample)),")")) +
               theme(axis.text.x=element_text(angle=50, hjust=1,size=4),
                     axis.text.y=element_text(size=8,colour='black'),
                     axis.title.x=element_text(size=11),
                     axis.title.y=element_blank())
             p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_mutation,Number=0,n=length(unique(d.mutations.nonres$sample)))
             p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
           } else {
             q1<-#ggplotly(
               waterfall(d.mutations.nonres, fileType = "Custom",
                         variant_class_order=sort(as.character(unique(all.mutations[,"variant_class"]))),
                         plotGenes=unlist(str_split(input_gene_mutation,",")),
                         mainXlabel = TRUE,my_x_label=ind[1],
                         plotMutBurden = FALSE,returnp="p1",
                         mainPalette=c('#4f00A8', '#A80100', '#CF5A59', '#A80079', '#BC2D94',
                                       '#CF59AE', '#000000', '#006666', '#00A8A8', '#009933',
                                       '#ace7b9', '#cdf0d5', '#59CF74', '#002AA8', '#5977CF',
                                       '#F37812', '#F2B079', '#888811', '#FDF31C', '#8C8C8C',
                                       '#CD5C5C', '#E9967A', '#FF0000', '#FF7F50', '#FFD700',
                                       '#FF8C00', '#FFFFE0', '#FFEFD5', '#FFDAB9', '#EEE8AA',
                                       '#7CFC00', '#228B22', '#ADFF2F', '#8FBC8F', '#2E8B57',
                                       '#00FFFF', '#AFEEEE', '#00CED1', '#008080', '#B0C4DE'))#,
             Number.genes.nonres <- NULL
             for (j in 1:length(input_gene_mutation)) {
               Number.genes.nonres <- c(Number.genes.nonres,length(unique(d.mutations.nonres[d.mutations.nonres[,"gene"]==input_gene_mutation[j],"sample"])))
             }
             p.bar.nonres <- data.frame(PFS.status="nonresponse",genes=input_gene_mutation,Number=Number.genes.nonres,n=length(unique(d.mutations.nonres$sample)))
             p.bar.nonres <- p.bar.nonres[order(p.bar.nonres$genes,decreasing = TRUE),]
           }

           p.bar.res <- data.frame(p.bar.res,perc=signif(p.bar.res$Number/p.bar.res$n*100,3))
           p.bar.nonres <- data.frame(p.bar.nonres,perc=signif(p.bar.nonres$Number/p.bar.nonres$n*100,3))

           if (sum(p.bar.res$perc)==0 & sum(p.bar.nonres$perc)!=0 ) {
             p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
             p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
           } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)==0) {
             p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
             p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
           } else if (sum(p.bar.res$perc)!=0 & sum(p.bar.nonres$perc)!=0) {
             p.bar.res <- p.bar.res[p.bar.res[,"perc"]!=0,]
             p.bar.nonres <- p.bar.nonres[p.bar.nonres[,"perc"]!=0,]
           } else {
             p.bar.res <- data.frame(PFS.status="response",genes="no mutation",Number=0,n=p.bar.res[1,"n"],perc=0)
             p.bar.nonres <- data.frame(PFS.status="nonresponse",genes="no mutation",Number=0,n=p.bar.nonres[1,"n"],perc=0)
           }
           #p.bar <- rbind(p.bar.res,p.bar.nonres)
           p.bar <- rbind(p.bar.nonres,p.bar.res)
           p.bar_cumsum <- data.frame(p.bar,label_ypos=unlist(tapply(p.bar$perc,p.bar$PFS.status,FUN=cumsum))-(p.bar$perc/2))
           bar.plot <- ggplot(p.bar_cumsum,aes(fill=genes,y=perc,x=PFS.status)) + geom_bar(position="stack",stat="identity") +
             theme(legend.title=element_blank(),axis.title.x=element_blank()) + ylab("% mutated samples") +
             geom_text(aes(y=label_ypos,label=paste0(genes," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
             #geom_text(aes(y=p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"label_ypos"],label=paste0(p.bar_cumsum[p.bar_cumsum[,"label_ypos"]!=0,"genes"]," (",perc,"%)")),vjust=1.6,color="black",size=3.5) +
             geom_text(aes(x=1.5,y=max(aggregate(perc~PFS.status,p.bar_cumsum,sum)[,"perc"])+2,label=paste0("p-value",p.v)))


           #incProgress(100, detail = "Doing part %")
           subplot(style(bar.plot,showlegend=F),
                   subplot(p1,q1,nrows=2,margin=0.1,titleX=TRUE,titleY=TRUE) %>%
                     layout(#title = as.character(unique(d$data_source)),
                       titlefont=list(size=20,color="black",family= "Aril"),
                       legend=list(font=list(size=10))),
                   #xaxis=list(title=lable_x,side=side_x,titlefont=list(size=15))),
                   widths=c(1/3,2/3),margin=0.05,titleX=TRUE,titleY=TRUE) %>%
             layout(title=as.character(unique(d$data_source)))
         })
       })
     })


     ###############
     #### pan cancer mutational signature
     ###############
     
     selecteddatabase_pan_cancer_mutation_signature <- eventReactive(input$do_pan_cancer,{
       withProgress(message = 'Preparing data', value = 0, {
       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       selected.database.res <- selected.database[selected.database[,"PFS.status"]=="response" & !is.na(selected.database[,"PFS.status"]),]
       selected.database.nonres <- selected.database[selected.database[,"PFS.status"]=="nonresponse" & !is.na(selected.database[,"PFS.status"]),]
       
       subset.p.value <- NULL
       d.nonna <- NULL
       if (nrow(selected.database.res)!=0 & nrow(selected.database.nonres)!=0) {
         signatures <- c("Signature.1","Signature.2","Signature.3","Signature.4","Signature.5","Signature.6","Signature.7","Signature.8","Signature.9","Signature.10","Signature.11","Signature.12","Signature.13","Signature.14","Signature.15","Signature.16","Signature.17","Signature.18","Signature.19","Signature.20","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25","Signature.26","Signature.27","Signature.28","Signature.29","Signature.30","unknown")
         for (j in 1:length(signatures)) {
           incProgress(j/length(signatures), detail = "Doing part %")
           try.test <- try(p.value <- wilcox.test(as.numeric(na.omit(selected.database.res[,signatures[j]])),as.numeric(na.omit(selected.database.nonres[,signatures[j]])))$p.value)
           if ( class(try.test) == "try-error" | is.na(try.test)) {
             p.value <- NA
           # } else if (p.value == 0.000) {
           #   p.value <- 0.0001
           # } else {
           #   p.value <- p.value
             #p.value <- wilcox.test(as.numeric(na.omit(selected.database.res[,immu_cells[j]])),as.numeric(na.omit(selected.database.nonres[,immu_cells[j]])))$p.value
           # } else {
           #   p.value <- signif(p.v,3)
           }
           subset.p.value <-c(subset.p.value,p.value)
         }
         d.nonna <- data.frame(p.value=signif(as.numeric(subset.p.value),2),log.p.value= signif(-log10(as.numeric(subset.p.value)),2),FDR=signif(p.adjust(as.numeric(subset.p.value),"fdr"),2))
         rownames(d.nonna) <- signatures
         d.nonna <- d.nonna[order(d.nonna[,1]),]
       }
       d.nonna
     })
     })
     
     output$pan_cancer_mutation_signature_plots <- renderPlotly({
       withProgress(message = 'Making plot', value = 0, {
         
       d.nonna <- selecteddatabase_pan_cancer_mutation_signature()
           
       
       if (is.null(d.nonna)) {
         incProgress(100, detail = "Doing part %")
         plot_ly () %>%
           layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
           add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
       } else {
         vline <- function(x = 0, color = "red") {
           list(
             type = "line",
             y0 = 0,
             y1 = 1,
             yref = "paper",
             x0 = x,
             x1 = x,
             line = list(color = color)
           )
         }
         fit <- density(na.omit(d.nonna[,2]))
         p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
           layout(title="the aggregated dataset",
                  height = 480,
                  xaxis = list(title="-log10(p.value)",
                               titlefont=list(size=20,color="black",family= "Aril black"),
                               showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                  yaxis=list(title="Density",
                             titlefont=list(size=20,color="black",family= "Aril black"),
                             showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                  shapes=list(vline(1.3))) %>%
           add_annotations(x = 1.3,
                           y = 1,
                           text = "p.value = 0.05",
                           xref = paste0("x", 1),  # add this
                           yref = paste0("y", 1),  # add this
                           showarrow = FALSE,
                           ax = 20,
                           ay = -40,
                           font=list(size=20,color="black",family= "Aril black"))
         incProgress(100, detail = "Doing part %")
         p
       }
     })
    })
       

     
     output$pan_cancer_mutation_signature_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_mutation_signature_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_mutation_signature_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_mutation_signature_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         if (input$pan_cancer_mutation_signature_type == "Excel (CSV)") {
           write.csv (selecteddatabase_pan_cancer_mutation_signature(),file,row.names = FALSE)
         } else if (input$pan_cancer_mutation_signature_type == "Text (TSV)") {
           write.table (selecteddatabase_pan_cancer_mutation_signature(),file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_mutation_signature_type == "JSON") {
           write(toJSON(selecteddatabase_pan_cancer_mutation_signature()),file)
         }
       }
     )
     
     # output$pan_cancer_mutation_signature_table <- DT::renderDataTable({
     #   selecteddatabase_pan_cancer_mutation_signature()
     # })
     observeEvent(input$do_pan_cancer,{
       input$do_pan_cancer
     isolate({
       d.nonna.t.p0.05 <- selecteddatabase_pan_cancer_mutation_signature()
       shinyInput <- function(FUN, len, id, ...) {
         inputs <- character(len)
         for (i in seq_len(len)) {
           inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
         }
         inputs
       }
       d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
         #Signature=rownames(d.nonna.t.p0.05),
         #p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
         #FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
         signif(d.nonna.t.p0.05[,c(1,3)],2),
         View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),"pan-cancer_",label="View",onclick='Shiny.onInputChange(\"select_button_pan_cancer_mutation_signature\",  this.id)'),
         stringsAsFactors = FALSE#,
         #row.names = 1:nrow(d.nonna.t.p0.05)
       ))
       
       pan_cancer_mutation_signature_tables_data <- d.nonna.t.p0.05.test$data
       
       output$pan_cancer_mutation_signature_table <- DT::renderDataTable(
         pan_cancer_mutation_signature_tables_data,server = FALSE, escape = FALSE, selection = 'none'
       )
     })
     })
     
     observeEvent(input$select_button_pan_cancer_mutation_signature,{
       s <- as.numeric(strsplit(input$select_button_pan_cancer_mutation_signature, "_")[[1]][3])
       Viewname <- as.character(strsplit(input$select_button_pan_cancer_mutation_signature, "_")[[1]][2])
       output$pan_cancer_mutation_signature_modals <- renderUI({
         tagList(
           bsModal(paste('pan_cancer_mutation_signature_model', s ,sep=''), Viewname, "select_button_pan_cancer_mutation_signature", size = "large",
                   
                   plotlyOutput("browse_pan_cancer_mutation_signature_bsModal_plot")
                   #textOutput("browse_prio_mutation_signature_bsModal_plot")
           )
         )
       })
       toggleModal(session,paste('pan_cancer_mutation_signature_model', s ,sep=''), toggle = Viewname)
       ##Reset the select_button
       session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_pan_cancer_mutation_signature")
     })
     
     observeEvent(input$select_button_pan_cancer_mutation_signature,{
       signature <- as.character(strsplit(input$select_button_pan_cancer_mutation_signature, "_")[[1]][2])
       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       d <- selected.database[,c("PFS.status",signature)]
       colnames(d)[2] <- "Signature"
       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Signature"]),]
       p<-plot_ly(data=d,x=~PFS.status,
                  y=~Signature,
                  type="violin",#type="box",
                  points="all",#boxpoints="all",
                  jitter=0.3,
                  pointpos=-1.8,
                  #mode="markders",
                  color= ~PFS.status,
                  colors=c("#7CDBD5","#DCB239"),
                  #hoveron="points+boxes",
                  legendgroup=~PFS.status,
                  showlegend=TRUE
       ) %>%
         layout(title = "pan-cancer",titlefont=list(size=20,color="black",family= "Aril"),
                height = 480,
                xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                           showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                yaxis=list(title=signature,
                           titlefont=list(size=20,color="black",family= "Aril"),
                           showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                )) %>%
         add_annotations(x=0.5,
                         y=d[which.max(d[,"Signature"]),"Signature"],
                         
                         text=paste0("p-value = ",signif(wilcox.test(d[d[,"PFS.status"]=="nonresponse","Signature"],d[d[,"PFS.status"]=="response","Signature"])$p.value,2)),
                         xref = paste0("x", 1),  # add this
                         yref = paste0("y", 1),  # add this
                         showarrow = FALSE,
                         ax = 20,
                         ay = -40,
                         font=list(size=20,color="black",family= "Aril black"))
       output$browse_pan_cancer_mutation_signature_bsModal_plot <- renderPlotly(
         p
       )
     })
     
     ###################
     #### pan cancer expression
     ###################

     selecteddatabase_pan_cancer_expression <- eventReactive(input$do_pan_cancer,{
       withProgress(message = 'Preparing data', value = 0, {
       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       selected.database.res <- selected.database[selected.database[,"PFS.status"]=="response" & !is.na(selected.database[,"PFS.status"]),]
       selected.database.nonres <- selected.database[selected.database[,"PFS.status"]=="nonresponse" & !is.na(selected.database[,"PFS.status"]),]

       subset.p.value <- NULL
       d.nonna <- NULL
       if (nrow(selected.database.res)!=0 & nrow(selected.database.nonres)!=0) {
         genes <- colnames(selected.database)[13:ncol(selected.database)]
         genes <- genes[genes %ni% c("Mutation.Load","
                      Signature.1","Signature.2","Signature.3","Signature.4","Signature.5","Signature.6","Signature.7","Signature.8","Signature.9",
                     "Signature.10","Signature.11","Signature.12","Signature.13","Signature.14","Signature.15","Signature.16","Signature.17","Signature.18",
                     "Signature.19","Signature.20","Signature.21","Signature.22","Signature.23","Signature.24","Signature.25","Signature.26","Signature.27",
                     "Signature.28","Signature.29","Signature.30","unknown","B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive",
                     "T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta",
                     "NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting",
                     "Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")]
         # immu_cells <- c("B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive","T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta","NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting","Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")
         for (j in 1:length(genes)) {
           incProgress(j/length(genes), detail = "Doing part %")
           try.test <- try(p.value <- wilcox.test(as.numeric(na.omit(selected.database.res[,genes[j]])),as.numeric(na.omit(selected.database.nonres[,genes[j]])))$p.value)
           if (class(try.test)=="try-error" | is.na(try.test)) {
             p.value <- NA
             # } else if (p.value == 0){
             #   p.value <- 0.0001
             # } else {
             #   p.value <- p.value
             #p.value <- wilcox.test(as.numeric(na.omit(selected.database.res[,immu_cells[j]])),as.numeric(na.omit(selected.database.nonres[,immu_cells[j]])))$p.value
           }
           subset.p.value <-c(subset.p.value,p.value)
         }
         d.nonna <- data.frame(p.value=signif(as.numeric(subset.p.value),2),log.p.value= signif(-log10(as.numeric(subset.p.value)),2),FDR=signif(p.adjust(as.numeric(subset.p.value),"fdr"),2))
         rownames(d.nonna) <- genes
         d.nonna <- d.nonna[order(d.nonna[,1]),]
       }
       d.nonna
     })
     })

     output$pan_cancer_expression_plots <- renderPlotly({
       withProgress(message = 'Making plot', value = 0, {
         if (input$sample_pan_cancer=="custom selection") {
          d.nonna <- selecteddatabase_pan_cancer_expression()
         } else {
           d.nonna <- pan_cancer_prio_precalculated_expression[[input$sample_pan_cancer]]
         }
         if (is.null(d.nonna)) {
           incProgress(100, detail = "Doing part %")
           plot_ly () %>%
             layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
         } else {
           vline <- function(x = 0, color = "red") {
             list(
               type = "line",
               y0 = 0,
               y1 = 1,
               yref = "paper",
               x0 = x,
               x1 = x,
               line = list(color = color)
             )
           }
           fit <- density(na.omit(d.nonna[,2]))
           p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
             layout(title="the aggregated dataset",
                    height = 480,
                    xaxis = list(title="-log10(p.value)",
                                 titlefont=list(size=20,color="black",family= "Aril black"),
                                 showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    yaxis=list(title="Density",
                               titlefont=list(size=20,color="black",family= "Aril black"),
                               showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    shapes=list(vline(1.3))) %>%
             add_annotations(x = 1.3,
                             y = 1,
                             text = "p.value = 0.05",
                             xref = paste0("x", 1),  # add this
                             yref = paste0("y", 1),  # add this
                             showarrow = FALSE,
                             ax = 20,
                             ay = -40,
                             font=list(size=20,color="black",family= "Aril black"))
           incProgress(100, detail = "Doing part %")
           p
         }
       })
     })

     output$pan_cancer_expression_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_expression_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_expression_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_expression_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         if (input$sample_pan_cancer=="custom selection") {
           d.nonna <- selecteddatabase_pan_cancer_expression()
         } else {
           d.nonna <- pan_cancer_prio_precalculated_expression[[input$sample_pan_cancer]]
         }
         
         if (input$pan_cancer_expression_type == "Excel (CSV)") {
           write.csv (d.nonna,file,row.names = FALSE)
         } else if (input$pan_cancer_expression_type == "Text (TSV)") {
           write.table (d.nonna,file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_expression_type == "JSON") {
           write(toJSON(d.nonna),file)
         }
       }
     )

     observeEvent(input$do_pan_cancer,{
       input$do_pan_cancer
       isolate({
         if (input$sample_pan_cancer=="custom selection") {
           d.nonna.t.p0.05 <- selecteddatabase_pan_cancer_expression()
         } else {
           d.nonna.t.p0.05 <- pan_cancer_prio_precalculated_expression[[input$sample_pan_cancer]]
         }
         
         #d.nonna.t.p0.05 <- selecteddatabase_pan_cancer_expression()
         shinyInput <- function(FUN, len, id, ...) {
           inputs <- character(len)
           for (i in seq_len(len)) {
             inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
           }
           inputs
         }
         d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
           #Signature=rownames(d.nonna.t.p0.05),
           #p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
           #FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
           signif(d.nonna.t.p0.05[,c(1,3)],2),
           View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),"pan-cancer_",label="View",onclick='Shiny.onInputChange(\"select_button_pan_cancer_expression\",  this.id)'),
           stringsAsFactors = FALSE#,
           #row.names = 1:nrow(d.nonna.t.p0.05)
         ))

         pan_cancer_expression_tables_data <- d.nonna.t.p0.05.test$data

         output$pan_cancer_expression_table <- DT::renderDataTable(
           pan_cancer_expression_tables_data,server = FALSE, escape = FALSE, selection = 'none'
         )
       })
     })

     observeEvent(input$select_button_pan_cancer_expression,{
       s <- as.numeric(strsplit(input$select_button_pan_cancer_expression, "_")[[1]][3])
       Viewname <- as.character(strsplit(input$select_button_pan_cancer_expression, "_")[[1]][2])
       output$pan_cancer_expression_modals <- renderUI({
         tagList(
           bsModal(paste('pan_cancer_expression_model', s ,sep=''), Viewname, "select_button_pan_cancer_expression", size = "large",

                   plotlyOutput("browse_pan_cancer_expression_bsModal_plot")
                   #textOutput("browse_prio_expression_bsModal_plot")
           )
         )
       })
       toggleModal(session,paste('pan_cancer_expression_model', s ,sep=''), toggle = Viewname)
       ##Reset the select_button
       session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_pan_cancer_expression")
     })

     observeEvent(input$select_button_pan_cancer_expression,{
       genes <- as.character(strsplit(input$select_button_pan_cancer_expression, "_")[[1]][2])
       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       d <- selected.database[,c("PFS.status",genes)]
       colnames(d)[2] <- "Gene"
       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Gene"]),]
       p<-plot_ly(data=d,x=~PFS.status,
                  y=~Gene,
                  type="violin",#type="box",
                  points="all",#boxpoints="all",
                  jitter=0.3,
                  pointpos=-1.8,
                  #mode="markders",
                  color= ~PFS.status,
                  colors=c("#7CDBD5","#DCB239"),
                  #hoveron="points+boxes",
                  legendgroup=~PFS.status,
                  showlegend=TRUE
       ) %>%
         layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                height = 480,
                xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                           showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                yaxis=list(title=genes,
                           titlefont=list(size=20,color="black",family= "Aril"),
                           showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                )) %>%
         add_annotations(x=0.5,
                         y=d[which.max(d[,"Gene"]),"Gene"],

                         text=paste0("p-value = ",signif(wilcox.test(d[d[,"PFS.status"]=="nonresponse","Gene"],d[d[,"PFS.status"]=="response","Gene"])$p.value,2)),
                         xref = paste0("x", 1),  # add this
                         yref = paste0("y", 1),  # add this
                         showarrow = FALSE,
                         ax = 20,
                         ay = -40,
                         font=list(size=20,color="black",family= "Aril black"))
       output$browse_pan_cancer_expression_bsModal_plot <- renderPlotly(
         p
       )
     })



     #################
     #### pan cancer cibersort
     #################
     
     selecteddatabase_pan_cancer_cibersort <- eventReactive(input$do_pan_cancer,{
       withProgress(message = 'Preparing data', value = 0, {
       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       selected.database.res <- selected.database[selected.database[,"PFS.status"]=="response" & !is.na(selected.database[,"PFS.status"]),]
       selected.database.nonres <- selected.database[selected.database[,"PFS.status"]=="nonresponse" & !is.na(selected.database[,"PFS.status"]),]
       
       subset.p.value <- NULL
       d.nonna <- NULL
       if (nrow(selected.database.res)!=0 & nrow(selected.database.nonres)!=0) {
         immu_cells <- c("B.cells.naive","B.cells.memory","Plasma.cells","T.cells.CD8","T.cells.CD4.naive","T.cells.CD4.memory.resting","T.cells.CD4.memory.activated","T.cells.follicular.helper","T.cells.regulatory..Tregs.","T.cells.gamma.delta","NK.cells.resting","NK.cells.activated","Monocytes","Macrophages.M0","Macrophages.M1","Macrophages.M2","Dendritic.cells.resting","Dendritic.cells.activated","Mast.cells.resting","Mast.cells.activated","Eosinophils","Neutrophils")
         for (j in 1:length(immu_cells)) {
           incProgress(j/length(immu_cells), detail = "Doing part %")
           try.test <- try(p.value <- wilcox.test(as.numeric(na.omit(selected.database.res[,immu_cells[j]])),as.numeric(na.omit(selected.database.nonres[,immu_cells[j]])))$p.value)
           if (class(try.test)=="try-error" | is.na(try.test)) {
             p.value <- NA
           # } else if (p.value == 0){
           #   p.value <- 0.0001
           # } else {
           #   p.value <- p.value
             #p.value <- wilcox.test(as.numeric(na.omit(selected.database.res[,immu_cells[j]])),as.numeric(na.omit(selected.database.nonres[,immu_cells[j]])))$p.value
           }
           subset.p.value <-c(subset.p.value,p.value)
         }
         d.nonna <- data.frame(p.value=signif(as.numeric(subset.p.value),2),log.p.value= signif(-log10(as.numeric(subset.p.value)),2),FDR=signif(p.adjust(as.numeric(subset.p.value),"fdr"),2))
         rownames(d.nonna) <- immu_cells
         d.nonna <- d.nonna[order(d.nonna[,1]),]
       }
       d.nonna
     })
     })
     
     output$pan_cancer_cibersort_plots <- renderPlotly({
       withProgress(message = 'Making plot', value = 0, {
         d.nonna <- selecteddatabase_pan_cancer_cibersort()
         
         if (is.null(d.nonna)) {
           incProgress(100, detail = "Doing part %")
           plot_ly () %>%
             layout(title = as.character(unique(d$data_source)),font=list(size=15,family="Aril")) %>%
             add_annotations(x=2,y=2,text="no PFS outcome under this selection",showarrow=F,font=list(size=16,color="purple"))
         } else {
           vline <- function(x = 0, color = "red") {
             list(
               type = "line",
               y0 = 0,
               y1 = 1,
               yref = "paper",
               x0 = x,
               x1 = x,
               line = list(color = color)
             )
           }
           fit <- density(na.omit(d.nonna[,2]))
           p<-plot_ly(x=fit$x, y=fit$y,type="scatter",mode="line",fill="tozeroy",name="Density") %>%
             layout(title="the aggregated dataset",
                    height = 480,
                    xaxis = list(title="-log10(p.value)",
                                 titlefont=list(size=20,color="black",family= "Aril black"),
                                 showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    yaxis=list(title="Density",
                               titlefont=list(size=20,color="black",family= "Aril black"),
                               showticklabels=TRUE,tickmode="auto",tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                    shapes=list(vline(1.3))) %>%
             add_annotations(x = 1.3,
                             y = 1,
                             text = "p.value = 0.05",
                             xref = paste0("x", 1),  # add this
                             yref = paste0("y", 1),  # add this
                             showarrow = FALSE,
                             ax = 20,
                             ay = -40,
                             font=list(size=20,color="black",family= "Aril black"))
           incProgress(100, detail = "Doing part %")
           p
         }
       })
     })
     
     
     
     output$pan_cancer_cibersort_download <- downloadHandler(
       filename = function() {
         if (input$pan_cancer_cibersort_type == "Excel (CSV)") {
           "newfile.csv"
         } else if (input$pan_cancer_cibersort_type == "Text (TSV)") {
           "newfile.txt"
         } else if (input$pan_cancer_cibersort_type == "JSON") {
           "newfile.json"
         }
       },
       content = function(file) {
         if (input$pan_cancer_cibersort_type == "Excel (CSV)") {
           write.csv (selecteddatabase_pan_cancer_cibersort(),file,row.names = FALSE)
         } else if (input$pan_cancer_cibersort_type == "Text (TSV)") {
           write.table (selecteddatabase_pan_cancer_cibersort(),file,quote=F, row.names = FALSE, sep="\t")
         } else if (input$pan_cancer_cibersort_type == "JSON") {
           write(toJSON(selecteddatabase_pan_cancer_cibersort()),file)
         }
       }
     )
     
     observeEvent(input$do_pan_cancer,{
       input$do_pan_cancer
       isolate({
         d.nonna.t.p0.05 <- selecteddatabase_pan_cancer_cibersort()
         shinyInput <- function(FUN, len, id, ...) {
           inputs <- character(len)
           for (i in seq_len(len)) {
             inputs[i] <- as.character(FUN(paste0(id, rownames(d.nonna.t.p0.05)[i],"_",i), ...))
           }
           inputs
         }
         d.nonna.t.p0.05.test <- reactiveValues(data=data.frame(
           #Signature=rownames(d.nonna.t.p0.05),
           #p.value=formatC(d.nonna.t.p0.05[,1],format="e",digits = 2),
           #FDR=formatC(d.nonna.t.p0.05[,3],format="e",digits = 2),
           signif(d.nonna.t.p0.05[,c(1,3)],2),
           View=shinyInput(actionButton,nrow(d.nonna.t.p0.05),"pan-cancer_",label="View",onclick='Shiny.onInputChange(\"select_button_pan_cancer_cibersort\",  this.id)'),
           stringsAsFactors = FALSE#,
           #row.names = 1:nrow(d.nonna.t.p0.05)
         ))
         
         pan_cancer_cibersort_tables_data <- d.nonna.t.p0.05.test$data
         
         output$pan_cancer_cibersort_table <- DT::renderDataTable(
           pan_cancer_cibersort_tables_data,server = FALSE, escape = FALSE, selection = 'none'
         )
       })
     })
     
     observeEvent(input$select_button_pan_cancer_cibersort,{
       s <- as.numeric(strsplit(input$select_button_pan_cancer_cibersort, "_")[[1]][3])
       Viewname <- as.character(strsplit(input$select_button_pan_cancer_cibersort, "_")[[1]][2])
       output$pan_cancer_cibersort_modals <- renderUI({
         tagList(
           bsModal(paste('pan_cancer_cibersort_model', s ,sep=''), Viewname, "select_button_pan_cancer_cibersort", size = "large",
                   
                   plotlyOutput("browse_pan_cancer_cibersort_bsModal_plot")
                   #textOutput("browse_prio_cibersort_bsModal_plot")
           )
         )
       })
       toggleModal(session,paste('pan_cancer_cibersort_model', s ,sep=''), toggle = Viewname)
       ##Reset the select_button
       session$sendCustomMessage(type = 'resetInputValue', message =  "select_button_pan_cancer_cibersort")
     })
     
     observeEvent(input$select_button_pan_cancer_cibersort,{
       immucells <- as.character(strsplit(input$select_button_pan_cancer_cibersort, "_")[[1]][2])
       selected.database <- selecteddatabase_pan_cancer_batchcorrected()
       d <- selected.database[,c("PFS.status",immucells)]
       colnames(d)[2] <- "Immucells"
       d <- d[!is.na(d[,"PFS.status"]) & !is.na(d[,"Immucells"]),]
       p<-plot_ly(data=d,x=~PFS.status,
                  y=~Immucells,
                  type="violin",#type="box",
                  points="all",#boxpoints="all",
                  jitter=0.3,
                  pointpos=-1.8,
                  #mode="markders",
                  color= ~PFS.status,
                  colors=c("#7CDBD5","#DCB239"),
                  #hoveron="points+boxes",
                  legendgroup=~PFS.status,
                  showlegend=TRUE
       ) %>%
         layout(title = "the aggregated dataset",titlefont=list(size=20,color="black",family= "Aril"),
                height = 480,
                xaxis=list(title=" ",titlefont=list(size=20,color="black",family= "Aril"),
                           showticklabels=TRUE,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)),
                yaxis=list(title=immucells,
                           titlefont=list(size=20,color="black",family= "Aril"),
                           showticklabels=TRUE,tickmode="auto",nticks=4,tickcolor="#444",linecolor = "#444",tickfont=list(size=15)
                )) %>%
         add_annotations(x=0.5,
                         y=d[which.max(d[,"Immucells"]),"Immucells"],
                         
                         text=paste0("p-value = ",signif(wilcox.test(d[d[,"PFS.status"]=="nonresponse","Immucells"],d[d[,"PFS.status"]=="response","Immucells"])$p.value,2)),
                         xref = paste0("x", 1),  # add this
                         yref = paste0("y", 1),  # add this
                         showarrow = FALSE,
                         ax = 20,
                         ay = -40,
                         font=list(size=20,color="black",family= "Aril black"))
       output$browse_pan_cancer_cibersort_bsModal_plot <- renderPlotly(
         p
       )
     })
     #######################################################
     #### users' data
     #######################################################
     output$user_clinical_table <- renderTable({
       if (is.null(input$file_clinical)) {
         return(NULL)
       } else {
         
         df_clinical <- user_clinical()
         if (input$display_clinical=="All") {
           df_clinical  
         } else {
           if (nrow(df_clinical)>=10) {
             df_clinical[1:10,]
           } else {
             df_clinical
           }
         }
       }
     })
     output$user_transcriptome_table <- renderTable({
       if (is.null(input$file_transcriptome)) {
         return(NULL)
       } else {
         
         df_transcriptome <- user_transcriptome()
         if (input$display_transcriptome=="All") {
           df_transcriptome  
         } else {
           if (ncol(df_transcriptome)>=10) {
             df_transcriptome[1:10,1:10]
           } else {
             df_transcriptome[1:10,]
           }
         }
       }
     })
     output$user_mutation_table <- renderTable({
       if (is.null(input$file_mutation)) {
         return(NULL)
       } else {
         
         df_mutation <- user_mutation()
         if (input$display_mutation=="All") {
           df_mutation  
         } else {
           if (nrow(df_mutation)>=10) {
             df_mutation[1:10,]
           } else {
             df_mutation
           }
         }
       }
     })
     observeEvent(input$do_user_data, {
       for (i in 1:100) {
         updateProgressBar(
           session = session,
           id = "pb2",
           value = i, total = 100,
           title = paste("Process", trunc(i/10))
         )
         database.test.process()
         all.mutations.process()
       }
     })
     
#       output$video <- renderUI({
# #        #h6("intro video",br(),
# #       #    tags$video(src='demo.mp4',type="video/mp4",width="350px",height="350px",controls="controls"))
#         HTML('<iframe width="560" height="315" src="https://www.youtube.com/watch?v=9qCSmPQAUuw&list=WL&index=17&t=0s" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>')
# #        
#       })
# #     h1("test",br(),HTML('<iframe width="560" height="315" src="https://www.youtube.com/watch?v=9qCSmPQAUuw&list=WL&index=17&t=0s" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>')),
     
  
  shinyjs::hide(id = "loading-content", anim = TRUE, animType = "fade")    
  shinyjs::show("app-content")
#  shinyjs::hide(id = "loading-c", anim = TRUE, animType = "fade")
#  shinyjs::hide(id = "loading-c", anim = TRUE, animType = "fade")  
#  shinyjs::show("app-c")
#  shinyjs::hide(id = "loading-expression-summary", anim = TRUE, animType = "fade")    
#  shinyjs::show("app-expression-summary")
  
})